---
title: "Test Setup"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Test data

I previously created some test data for users to test out the pipeline. I would like to update this test data to include height GWAS from EUR and EAS populations to test out the multi-source functionality.
  
```{r}
library(data.table)

yengo_eur<-fread('/users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_eur.txt')
yengo_eas<-fread('/users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_eas.txt')

ref <- NULL
for(i in 1:22){
  ref <-
    rbind(
      ref,
      readRDS(
        paste0(
          '~/oliverpainfel/Software/MyGit/GenoPred/pipeline/resources/data/ref/ref.chr', i, '.rds'
      )))
}

yengo_eur <- yengo_eur[yengo_eur$variant_id %in% ref$SNP,]
yengo_eas <- yengo_eas[yengo_eas$variant_id %in% ref$SNP,]

# Save subset GWAS sumstats
dir.create('~/test_data/reference/gwas_sumstats', recursive = T)
fwrite(yengo_eur, '~/oliverpainfel/Software/MyGit/GenoPred/pipeline/test_data/reference/gwas_sumstats/height_eur.txt.gz', quote=F, sep=' ', na='NA')
fwrite(yengo_eas, '~/oliverpainfel/Software/MyGit/GenoPred/pipeline/test_data/reference/gwas_sumstats/height_eas.txt.gz', quote=F, sep=' ', na='NA')

```

Test the pipeline using the example config.

```{bash}
cd ~/oliverpainfel/Software/MyGit/GenoPred/pipeline
conda activate genopred
snakemake --profile slurm --configfile=example_input/config.yaml --use-conda output_all -n
```

Make a test config that uses the EUR and EAS height GWAS.

```{r}
setwd('/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline')
library(data.table)

# Create gwas_list
gwas_list <- NULL
gwas_list<-rbind(gwas_list, 
                 data.table(name='height_eur',
                            path = 'test_data/reference/gwas_sumstats/height_eur.txt.gz',
                            population = 'EUR',
                            n = NA,
                            sampling = NA,
                            prevalence = NA,
                            mean = NA,
                            sd = NA,
                            label = "\"Height EUR\""))

gwas_list<-rbind(gwas_list, 
                 data.table(name='height_eas',
                            path = 'test_data/reference/gwas_sumstats/height_eas.txt.gz',
                            population = 'EAS',
                            n = NA,
                            sampling = NA,
                            prevalence = NA,
                            mean = NA,
                            sd = NA,
                            label = "\"Height EAS\""))

write.table(gwas_list, 'example_input/gwas_list.multisource.txt', col.names = T, row.names = F, quote = F, sep = ' ')

# Create gwas_groups
gwas_groups <- data.frame(
  name = 'height',
  gwas = 'height_eur,height_eas',
  label = "\"Height EUR+EAS\""
)

write.table(gwas_groups, 'example_input/gwas_groups.multisource.txt', col.names = T, row.names = F, quote = F, sep = ' ')

# Create config file
conf <- c(
  'outdir: test_data/output/test1',
  'config_file: example_input/config.multisource.yaml',
  'gwas_list: example_input/gwas_list.multisource.txt',
  'gwas_groups: example_input/gwas_groups.multisource.txt',
  'target_list: example_input/target_list.txt',
  "pgs_methods: ['lassosum']",
  "leopard_methods: ['lassosum']",
  'testing: chr22',
  'cores_prep_pgs: 5',
  'cores_target_pgs: 5',
  'mem_target_pgs: 2000'
)

write.table(conf, 'example_input/config.multisource.yaml', col.names = F, row.names = F, quote = F)

```

Now test it using the multi-source set up.

```{bash}
cd ~/oliverpainfel/Software/MyGit/GenoPred/pipeline
conda activate genopred
snakemake --profile slurm --configfile=example_input/config.multisource.yaml --use-conda output_all -n

# There was an error when running leopard_quickprs - running interactively to understand.
```

Now compress the test data, upload to Zenodo, and update the documentation.

```{bash}
cd ~/oliverpainfel/Software/MyGit/GenoPred/pipeline/test_data
tar -czvf archive_name.tar.gz folder_name/
```

***



some target data (in various formats), GWAS summary statistics for two t

I want to set up a series of tests that can be quickly run to check everything is working as it should. 

We can test every step all the time, so lets focus on testing the following scenario:

- Target sample data: Same as GenoPred test data, but restricted to subset of SNPs
- GWAS sumstats: Same as GenoPred test data, but restricted to subset of SNPs
- External sumstats: Same as GenoPred test data, but restricted to subset of SNPs
- Reference data: Same as GenoPred test data, but restricted to subset of SNPs

***

# Create test data

## Reference data

```{r}
setwd('~/oliverpainfel/Software/MyGit/GenoPred/pipeline')
# Read in chromosome 22 data, subset 1000 SNPs
ref_rds<-readRDS('~/oliverpainfel/Software/MyGit/GenoPred/pipeline/resources/data/ref/ref.chr22.rds')
test_snps<-ref_rds$SNP[1:1000]
dir.create('misc/dev/test_data/')
write.table(test_snps, 'misc/dev/test_data/test_snps.txt', col.names = F, row.names = F, quote = F)

# Create folder containing mini version of reference data
dir.create('misc/dev/test_data/ref')
saveRDS(ref_rds[ref_rds$SNP %in% test_snps,], 'misc/dev/test_data/ref/ref.chr22.rds')

system('cp ~/oliverpainfel/Software/MyGit/GenoPred/pipeline/resources/data/ref/ref.pop.txt misc/dev/test_data/ref/')
system('cp ~/oliverpainfel/Software/MyGit/GenoPred/pipeline/resources/data/ref/ref.keep.list misc/dev/test_data/ref/')
system('cp -r ~/oliverpainfel/Software/MyGit/GenoPred/pipeline/resources/data/ref/keep_files misc/dev/test_data/ref/')

pops<-list.files('~/oliverpainfel/Software/MyGit/GenoPred/pipeline/resources/data/ref/freq_files')
for(i in pops){
  dir.create(paste0('misc/dev/test_data/ref/freq_files/',i), recursive = T)
  freq_i<-fread(paste0('~/oliverpainfel/Software/MyGit/GenoPred/pipeline/resources/data/ref/freq_files/',i,'/ref.',i,'.chr22.afreq'))
  freq_i<-freq_i[freq_i$ID %in% test_snps,]
  fwrite(freq_i, paste0('misc/dev/test_data/ref/freq_files/',i,'/ref.',i,'.chr22.afreq'), quote=F, sep=' ', na='NA')
}
```

```{bash}
cd ~/oliverpainfel/Software/MyGit/GenoPred/pipeline

~/oliverpainfel/Software/plink2 \
  --pfile ~/oliverpainfel/Software/MyGit/GenoPred/pipeline/resources/data/ref/ref.chr22 \
  --extract misc/dev/test_data/test_snps.txt \
  --make-pgen \
  --out misc/dev/test_data/ref/ref.chr22
```

***

# Target sample data

```{bash}
mkdir misc/dev/test_data/target

~/oliverpainfel/Software/plink2 \
  --pfile ~/oliverpainfel/Software/MyGit/GenoPred/pipeline/test_data/target/imputed_sample_plink2/example.chr22 \
  --extract misc/dev/test_data/test_snps.txt \
  --make-pgen \
  --out misc/dev/test_data/target/example.chr22
```

***

# GWAS sumstats

```{r}
library(data.table)
ss<-fread('~/oliverpainfel/Software/MyGit/GenoPred/pipeline/test_data/reference/gwas_sumstats/BODY04.gz')
ss<-ss[ss$SNP %in% test_snps,]
dir.create('misc/dev/test_data/gwas')
fwrite(ss, 'misc/dev/test_data/gwas/BODY04.gz', quote=F, sep=' ', na='NA')
```

***

# External score file

```{r}
ext_score<-fread('~/oliverpainfel/Software/MyGit/GenoPred/pipeline/test_data/reference/score_files/PGS002804.txt.gz')
ext_score<-ext_score[ext_score$rsID %in% test_snps,]
dir.create('misc/dev/test_data/ext_score')
fwrite(ext_score, 'misc/dev/test_data/ext_score/PGS002804.txt.gz', quote=F, sep=' ', na='NA')
```

***

# Try running pipeline with test data

## Create config

```{r}
dir.create('misc/dev/test_data/config')

# gwas_list
gwas_list<-data.frame(
  name='BODY04',
  path='misc/dev/test_data/gwas/BODY04.gz',
  population='EUR',
  n=NA,
  sampling=NA,
  prevalence=NA,
  mean=0,
  sd=1,
  label='Body Mass Index'
)

gwas_list$label<-paste0("\"", gwas_list$label, "\"")

write.table(gwas_list, 'misc/dev/test_data/config/gwas_list.txt', col.names = T, row.names = F, quote = F)

# target_list
target_list<-data.frame(
  name='example_plink2',
  path='misc/dev/test_data/target/example',
  type='plink2',
  indiv_report=F
)

write.table(target_list, 'misc/dev/test_data/config/target_list.txt', col.names = T, row.names = F, quote = F)

# score_list
score_list<-data.frame(
  name='PGS002804',
  path='misc/dev/test_data/ext_score/PGS002804.txt.gz',
  label='Height Yengo EUR'
)

score_list$label<-paste0("\"", score_list$label, "\"")

write.table(score_list, 'misc/dev/test_data/config/score_list.txt', col.names = T, row.names = F, quote = F)

# config file
config<-c(
  "outdir: misc/dev/test_data/output",
  "config_file: misc/dev/test_data/config/config.yaml",
  "refdir: misc/dev/test_data/ref",
  "resdir: misc/dev/test_data/output/resources",
  "gwas_list: misc/dev/test_data/config/gwas_list.txt",
  "target_list: misc/dev/test_data/config/target_list.txt",
  "score_list: misc/dev/test_data/config/score_list.txt",
  "pgs_methods: ['ptclump','lassosum']",
  "testing: chr22",
  "cores_target_pgs: 1",
  "mem_target_pgs: 2000"
  )

write.table(config, 'misc/dev/test_data/config/config.yaml', col.names = F, row.names = F, quote = F)
```

***

## Run pipeline

Generate the reference data within the latest GenoPred container to ensure perfect reproduction.

```{bash}

singularity shell \
  --bind /scratch/prj/oliverpainfel:/scratch/prj/oliverpainfel \
  --writable-tmpfs \
  /users/k1806347/oliverpainfel/Software/singularity/genopred_pipeline_latest.sif
  
# Activate env and pull latest version from dev
source /opt/mambaforge/etc/profile.d/conda.sh
conda activate genopred
cd /tools/GenoPred/pipeline
git checkout dev
git pull

# Run pipeline
snakemake \
-j1 \
--use-conda \
output_all \
pc_projection \
misc/dev/test_data/output/reference/target_checks/example_plink2/indiv_report-4_EAS.4_EAS-report.done \
--configfile=misc/dev/test_data/config/config.yaml

# Move output outside the container so it can be pushed to the repo
mkdir /scratch/prj/oliverpainfel/Software/MyGit/GenoPred/pipeline/misc/dev/test_data/output
cp -r misc/dev/test_data/output/* /scratch/prj/oliverpainfel/Software/MyGit/GenoPred/pipeline/misc/dev/test_data/output/

# 
```

***

This output can now be used as a reference to check whether commits to the repo effects the output.

***

# Files for cross-population PGS methods

Make mini versions of the Yengo EUR and EAS GWAS sumstats.

```{r}
library(data.table)

test_snps<-fread('misc/dev/test_data/test_snps.txt', header=F)$V1

for(i in c('eur','eas')){
  ss<-fread(paste0('/users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_',i,'.txt'))
  ss<-ss[ss$variant_id %in% test_snps,]
  fwrite(ss, paste0('misc/dev/test_data/gwas/yengo_2022_height_',i,'.txt.gz'), quote=F, sep=' ', na='NA')
}

```







