<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Evaluating stratified polygenic scores</title>

<script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GenoPred</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Evaluating stratified polygenic scores</h1>

</div>


<style>
p.caption {
  font-size: 1.5em;
}
</style>
<style type="text/css">
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
</style>
<p>Here I will evaluate various stratified polygenic scoring approaches.
Research questions include:</p>
<ul>
<li><p>Adjust effect sizes before or after stratification. PRSet
performs clumping after stratification to increase the variance
explained by a stratified PGS. However, this is computationally
intensive compared to adjusting GWAS once, and then stratifying. The
limitation of adjusting before stratification is the effect within a
locus might be randomly assigned to a variant that is not within the
gene boundary. I would guess this effect would be minimal and
potentially increase the likihood of pulling in genetic effects that are
not actually mediated by the gene/pathway.</p></li>
<li><p>How to link variants to genes. Proximity or annotation based. AKA
PRSset/MAGMA approach GeRS/TWAS approach.</p></li>
</ul>
<p>Initially we will use case control phenotypes and GWAS as we did for
the PRS comparison, but looking more into the design used by Judit (Case
vs Case) would also be interesting.</p>
<ul>
<li>How to select annotations to stratify by. MAGMA GSEA p-value
theshold? Elastic net will handle this to some extent, but an initial
relaxed threshold is probably sensible. The ideal threshold for
inclusion will probably depend on the sample size on the sample size
where PGS are being modelled. UKB is big so can probably handl quite few
PGS.</li>
</ul>
<p>Structure of analysis: - Identify enriched and GO annotations using
MAMGA for each GWAS - Perform PRSet using enriched GO pathways (using
sparse pTs) - Stratify standard pT+clump PRS (i.e. after clumping) -
Stratify MegaPRS pseudoval PRS - Evaluate in UKB</p>
<hr />
<div id="gene-set-enrichment-analysis" class="section level1"
number="1">
<h1><span class="header-section-number">1</span> Gene set enrichment
analysis</h1>
<p>We have already downloaded a range of gene sets from MSigDB. We are
using the same collection used by FUMA, including
c2.all.v7.5.1.entrez.gmt, c5.go.bp.v7.5.1.entrez.gmt,
c5.go.cc.v7.5.1.entrez.gmt and c5.go.mf.v7.5.1.entrez.gmt.</p>
<p>Given polygenic scores will be restricted to HapMap3 variants, we
will also restrict the MAGMA analysis to HapMap3 variants.</p>
<pre class="bash"><code># Combine the gene sets into a single file
mkdir -p /users/k1806347/oliverpainfel/Analyses/stratified_pgs/sets/

cat /users/k1806347/oliverpainfel/Data/MSigDB/c2.all.v7.5.1.entrez.gmt /users/k1806347/oliverpainfel/Data/MSigDB/c5.go.cc.v7.5.1.entrez.gmt /users/k1806347/oliverpainfel/Data/MSigDB/c5.go.bp.v7.5.1.entrez.gmt /users/k1806347/oliverpainfel/Data/MSigDB/c5.go.mf.v7.5.1.entrez.gmt &gt; /users/k1806347/oliverpainfel/Analyses/stratified_pgs/sets/msigdb_sets.gmt
</code></pre>
<hr />
<div id="run-magma-gene-association-analysis" class="section level2"
number="1.1">
<h2><span class="header-section-number">1.1</span> Run MAGMA
gene-association analysis</h2>
<p>We will use some of the reference data used by GenoFuncPipe to avoid
downloading it again.</p>
<pre class="bash"><code>
gwas=$(echo DEPR06 COLL01 HEIG03 BODY04 DIAB05 COAD01 CROH01 SCLE03 RHEU02 PRCA01 BRCA01)

# Decompress GWAS
for gwas_i in ${gwas}; do

mkdir -p /users/k1806347/oliverpainfel/Analyses/stratified_pgs/magma/gene_assoc/${gwas_i}

gzip -f -d -c /users/k1806347/oliverpainfel/Data/GWAS_sumstats/prs_comparison/cleaned/${gwas_i}.cleaned.gz &gt; /users/k1806347/oliverpainfel/Analyses/stratified_pgs/magma/gene_assoc/${gwas_i}/${gwas_i}.cleaned

done

# Run MAGMA
for gwas_i in ${gwas}; do

sbatch -p cpu --mem 10G -n 1 --wrap=&quot;/users/k1806347/oliverpainfel/Software/MyGit/GenoFunc/GenoFuncPipe/resources/software/magma/magma \
      --bfile /users/k1806347/oliverpainfel/Software/MyGit/GenoFunc/GenoFuncPipe/resources/data/magma_ref/g1000_eur \
      --pval /users/k1806347/oliverpainfel/Analyses/stratified_pgs/magma/gene_assoc/${gwas_i}/${gwas_i}.cleaned use=SNP,P ncol=N \
      --gene-annot /users/k1806347/oliverpainfel/Software/MyGit/GenoFunc/GenoFuncPipe/resources/data/magma/NCBI37.3.genes.annot \
      --out /users/k1806347/oliverpainfel/Analyses/stratified_pgs/magma/gene_assoc/${gwas_i}/${gwas_i}_gene_assoc&quot;
      
done

# Delete decompressed sumstats
for gwas_i in ${gwas}; do

rm /users/k1806347/oliverpainfel/Analyses/stratified_pgs/magma/gene_assoc/${gwas_i}/${gwas_i}.cleaned

done
</code></pre>
<hr />
</div>
<div id="run-magma-gene-set-enrichment-analysis" class="section level2"
number="1.2">
<h2><span class="header-section-number">1.2</span> Run MAGMA gene-set
enrichment analysis</h2>
<pre class="bash"><code>
gwas=$(echo DEPR06 COLL01 HEIG03 BODY04 DIAB05 COAD01 CROH01 SCLE03 RHEU02 PRCA01 BRCA01)

# Decompress GWAS
for gwas_i in ${gwas}; do

mkdir -p /users/k1806347/oliverpainfel/Analyses/stratified_pgs/magma/set_assoc/${gwas_i}

sbatch -p cpu --mem 10G -n 1 --wrap=&quot;/users/k1806347/oliverpainfel/Software/MyGit/GenoFunc/GenoFuncPipe/resources/software/magma/magma \
--gene-results /users/k1806347/oliverpainfel/Analyses/stratified_pgs/magma/gene_assoc/${gwas_i}/${gwas_i}_gene_assoc.genes.raw \
--set-annot /users/k1806347/oliverpainfel/Analyses/stratified_pgs/sets/msigdb_sets.gmt \
--out /users/k1806347/oliverpainfel/Analyses/stratified_pgs/magma/set_assoc/${gwas_i}/${gwas_i}_set_assoc&quot;

done
</code></pre>
<hr />
</div>
<div
id="subset-.gmt-files-to-include-only-sets-that-are-enriched-for-each-gwas"
class="section level2" number="1.3">
<h2><span class="header-section-number">1.3</span> Subset .gmt files to
include only sets that are enriched for each GWAS</h2>
<pre class="r"><code>gwas&lt;-c(&#39;DEPR06&#39;,&#39;COLL01&#39;,&#39;HEIG03&#39;,&#39;BODY04&#39;,&#39;DIAB05&#39;,&#39;COAD01&#39;,&#39;CROH01&#39;,&#39;SCLE03&#39;,&#39;RHEU02&#39;,&#39;PRCA01&#39;,&#39;BRCA01&#39;)

library(data.table)

# To help decide set_pTs, count the number of sets achieving each threshold
# Use a range of p-value threshold to select sets
set_pt&lt;-c(0.05, 0.01, 0.001, 0.0001, 0.00001, 0.000001)

gwas_nset&lt;-NULL

for(gwas_i in gwas){
  set_assoc&lt;-read.table(paste0(&#39;/users/k1806347/oliverpainfel/Analyses/stratified_pgs/magma/set_assoc/&#39;,gwas_i,&#39;/&#39;,gwas_i,&#39;_set_assoc.gsa.out&#39;), header=T, stringsAsFactors = F)
  
  for(set_pt_i in set_pt){
    gwas_nset&lt;-rbind(gwas_nset, data.frame(gwas=gwas_i,
                                           set_pT=set_pt_i,
                                           n_set=sum(set_assoc$P &lt; set_pt_i)))
  }
}

# For simplicity, lets try using four thresholds
set_pt&lt;-c(0.05, 0.01, 0.001, 0.0001)

# Create subset .gmt containing significant sets for each GWAS
sets&lt;-readLines(&#39;/users/k1806347/oliverpainfel/Analyses/stratified_pgs/sets/msigdb_sets.gmt&#39;)
set_ids&lt;-sapply(strsplit(sets, &#39;\t&#39;),&quot;[[&quot;,1)

for(gwas_i in gwas){
  dir.create(paste0(&#39;/users/k1806347/oliverpainfel/Analyses/stratified_pgs/sets/&#39;,gwas_i))
  
  set_assoc&lt;-read.table(paste0(&#39;/users/k1806347/oliverpainfel/Analyses/stratified_pgs/magma/set_assoc/&#39;,gwas_i,&#39;/&#39;,gwas_i,&#39;_set_assoc.gsa.out&#39;), header=T, stringsAsFactors = F)
  
  for(set_pt_i in set_pt){
    writeLines(sets[set_ids %in% set_assoc$FULL_NAME[set_assoc$P &lt; set_pt_i]], paste0(&#39;/users/k1806347/oliverpainfel/Analyses/stratified_pgs/sets/&#39;,gwas_i,&#39;/sig_set.set_pt_&#39;,set_pt_i,&#39;.gmt&#39;))

  }
}</code></pre>
<hr />
</div>
</div>
<div id="calculate-stratified-polygenic-scores" class="section level1"
number="2">
<h1><span class="header-section-number">2</span> Calculate stratified
polygenic scores</h1>
<p>Here we will calculate polygenic scores using SNPs linked to genes
with enriched sets. We will link variants to genes using the MAGMA
annotation file (i.e. proximity based). We will use score files
previously create using the pT+clump and MegaPRS pseudoval methods.</p>
<hr />
<div id="create-snp-lists-for-enriched-gene-sets" class="section level2"
number="2.1">
<h2><span class="header-section-number">2.1</span> Create SNP lists for
enriched gene sets</h2>
<pre class="r"><code>gwas&lt;-c(&#39;DEPR06&#39;,&#39;COLL01&#39;,&#39;HEIG03&#39;,&#39;BODY04&#39;,&#39;DIAB05&#39;,&#39;COAD01&#39;,&#39;CROH01&#39;,&#39;SCLE03&#39;,&#39;RHEU02&#39;,&#39;PRCA01&#39;,&#39;BRCA01&#39;)

set_pt&lt;-c(0.05, 0.01, 0.001, 0.0001)

library(data.table)

# Read in MAGMA annot file
annot&lt;-readLines(&#39;/users/k1806347/oliverpainfel/Software/MyGit/GenoFunc/GenoFuncPipe/resources/data/magma/NCBI37.3.genes.annot&#39;)[-1:-2]
annot&lt;-strsplit(annot, &#39;\t&#39;)
annot_ids&lt;-sapply(annot,&quot;[[&quot;,1)

# Read in HapMap3 SNP list
hm3&lt;-fread(&#39;/users/k1806347/oliverpainfel/Data/ldsc/w_hm3.snplist&#39;)

for(gwas_i in gwas){
  for(set_pt_i in set_pt[1]){
    sets&lt;-readLines(paste0(&#39;/users/k1806347/oliverpainfel/Analyses/stratified_pgs/sets/&#39;,gwas_i,&#39;/sig_set.set_pt_&#39;,set_pt_i,&#39;.gmt&#39;))
    
    sets&lt;-strsplit(sets, &#39;\t&#39;)
    
    dir.create(paste0(&#39;/users/k1806347/oliverpainfel/Analyses/stratified_pgs/sets/&#39;,gwas_i,&#39;/snplists&#39;))
    
    for(set_i in 1:length(sets)){
      set_id&lt;-sets[[set_i]][1]
      genes&lt;-sets[[set_i]][-1:-2]
      
      annot_subset&lt;-annot[annot_ids %in% genes]
      
      snps&lt;-unique(do.call(c, annot_subset))
      snps&lt;-snps[grepl(&#39;^rs&#39;, snps)]
      snps&lt;-snps[!is.na(snps)]
      
      snps&lt;-snps[snps %in% hm3$SNP]
  
      write.table(snps, paste0(&#39;/users/k1806347/oliverpainfel/Analyses/stratified_pgs/sets/&#39;,gwas_i,&#39;/snplists/&#39;,set_id,&#39;.snplist&#39;), col.names=F, row.names=F, quote=F)

    }
  }
}</code></pre>
<hr />
</div>
<div id="megaprs" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> MegaPRS</h2>
<hr />
<div id="stratify-megaprs-weights-to-each-set" class="section level3"
number="2.2.1">
<h3><span class="header-section-number">2.2.1</span> Stratify MegaPRS
weights to each set</h3>
<pre class="r"><code>gwas&lt;-c(&#39;DEPR06&#39;,&#39;COLL01&#39;,&#39;HEIG03&#39;,&#39;BODY04&#39;,&#39;DIAB05&#39;,&#39;COAD01&#39;,&#39;CROH01&#39;,&#39;SCLE03&#39;,&#39;RHEU02&#39;,&#39;PRCA01&#39;,&#39;BRCA01&#39;)

set_pt&lt;-c(0.05, 0.01, 0.001, 0.0001)

library(data.table)

for(gwas_i in gwas){
  pseudo&lt;-fread(paste0(&#39;/users/k1806347/oliverpainfel/Data/1KG/Phase3/Score_files_for_polygenic/LDAK/&#39;,gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,gwas_i,&#39;.pseudoval.txt&#39;))

  pseudo_mod&lt;-gsub(&#39;.*_&#39;,&#39;&#39;,pseudo$V1[pseudo$V2 == max(pseudo$V2)][1])
  
  score&lt;-fread(cmd=paste0(&quot;zcat /users/k1806347/oliverpainfel/Data/1KG/Phase3/Score_files_for_polygenic/LDAK/&quot;,gwas_i,&quot;/1KGPhase3.w_hm3.&quot;,gwas_i,&quot;.score.gz | cut -d&#39; &#39; -f 1,2,&quot;,as.numeric(pseudo_mod)+2))
  names(score)&lt;-c(&#39;SNP&#39;,&#39;A1&#39;,&#39;SCORE&#39;)
  
  dir.create(paste0(&#39;/users/k1806347/oliverpainfel/Analyses/stratified_pgs/score_files/&#39;,gwas_i,&#39;/MegaPRS/&#39;), recursive=T)
  
  for(set_pt_i in set_pt[1]){
    sets&lt;-readLines(paste0(&#39;/users/k1806347/oliverpainfel/Analyses/stratified_pgs/sets/&#39;,gwas_i,&#39;/sig_set.set_pt_&#39;,set_pt_i,&#39;.gmt&#39;))
    
    sets&lt;-strsplit(sets, &#39;\t&#39;)
    
      for(set_i in 1:length(sets)){
        set_id&lt;-sets[[set_i]][1]
        snplist&lt;-fread(paste0(&#39;/users/k1806347/oliverpainfel/Analyses/stratified_pgs/sets/&#39;,gwas_i,&#39;/snplists/&#39;,set_id,&#39;.snplist&#39;), header=F)$V1
        
        score[[paste0(&#39;SCORE_&#39;,set_id)]]&lt;-score$SCORE
        score[[paste0(&#39;SCORE_&#39;,set_id)]][!(score$SNP %in% snplist)]&lt;-0
      }
  }
  
  score$SCORE&lt;-NULL
  
  fwrite(score, paste0(&#39;/users/k1806347/oliverpainfel/Analyses/stratified_pgs/score_files/&#39;,gwas_i,&#39;/MegaPRS/1KGPhase3.w_hm3.&#39;,gwas_i,&#39;.pseudo.stratified.score.gz&#39;), sep=&#39; &#39;, quote=F, na=&#39;NA&#39;)
  
}

# Process using 1KG reference to get reference mean and scale
for(gwas_i in gwas){
  dir.create(paste0(&#39;/users/k1806347/oliverpainfel/Analyses/stratified_pgs/score_files/&#39;,gwas_i,&#39;/MegaPRS/processed&#39;))
  system(paste0(&#39;sbatch -p cpu --mem 20G -n 1 /users/k1806347/oliverpainfel/Software/Rscript.sh /users/k1806347/oliverpainfel/Software/MyGit/GenoPred/Scripts/external_score_processor/external_score_processor_plink2.R --ref_plink_chr /users/k1806347/oliverpainfel/Data/1KG/Phase3/1KGPhase3.w_hm3.chr --score /users/k1806347/oliverpainfel/Analyses/stratified_pgs/score_files/&#39;,gwas_i,&#39;/MegaPRS/1KGPhase3.w_hm3.&#39;,gwas_i,&#39;.pseudo.stratified.score.gz --plink2 /users/k1806347/oliverpainfel/Software/plink2 --output /users/k1806347/oliverpainfel/Analyses/stratified_pgs/score_files/&#39;,gwas_i,&#39;/MegaPRS/processed/1KGPhase3.w_hm3.&#39;,gwas_i,&#39;.pseudo.stratified --ref_pop_scale /users/k1806347/oliverpainfel/Data/1KG/Phase3/super_pop_keep.list&#39;))
}</code></pre>
<details>
<summary>
Show code
</summary>
<pre class="bash"><code>
pheno=$(echo Depression Intelligence BMI Height T2D CAD IBD MultiScler RheuArth Breast_Cancer Prostate_Cancer)
gwas=$(echo DEPR06 COLL01 BODY04 HEIG03 DIAB05 COAD01 CROH01 SCLE03 RHEU02 BRCA01 PRCA01)

# Create directory
mkdir -p /users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/MegaPRS

# Create file listing GWAS that haven&#39;t been processed.
&gt; /users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/MegaPRS/todo.txt
for i in $(seq 1 11);do
gwas_i=$(echo ${gwas} | cut -f ${i} -d &#39; &#39;)
pheno_i=$(echo ${pheno} | cut -f ${i} -d &#39; &#39;)

if [ ! -f /users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/MegaPRS/${gwas_i}/UKBB.subset.w_hm3.${gwas_i}.pseudo.stratified.profiles ]; then
echo ${gwas_i} ${pheno_i} &gt;&gt; /users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/MegaPRS/todo.txt
fi
done

# Create shell script to run using sbatch
cat &gt; /users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/MegaPRS/sbatch.sh &lt;&lt; &#39;EOF&#39;
#!/bin/sh

#SBATCH -p cpu
#SBATCH --mem 40G
#SBATCH -n 5
#SBATCH -J strat_pgs
#SBATCH --exclude erc-hpc-comp001

gwas=$(awk -v var=&quot;$SLURM_ARRAY_TASK_ID&quot; &#39;NR == var {print $1}&#39; /users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/MegaPRS/todo.txt)
pheno=$(awk -v var=&quot;$SLURM_ARRAY_TASK_ID&quot; &#39;NR == var {print $2}&#39; /users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/MegaPRS/todo.txt)

echo $gwas
echo $pheno

/users/k1806347/oliverpainfel/Software/Rscript.sh /users/k1806347/oliverpainfel/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer/Scaled_polygenic_scorer_plink2.R \
    --target_plink_chr /users/k1806347/oliverpainfel/Data/UKBB/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr \
    --target_keep /users/k1806347/oliverpainfel/Data/UKBB/Phenotype/PRS_comp_subset/UKBB.${pheno}.txt \
    --ref_score /users/k1806347/oliverpainfel/Analyses/stratified_pgs/score_files/${gwas}/MegaPRS/1KGPhase3.w_hm3.${gwas}.pseudo.stratified.score.gz \
    --ref_scale /users/k1806347/oliverpainfel/Analyses/stratified_pgs/score_files/${gwas}/MegaPRS/processed/1KGPhase3.w_hm3.${gwas}.pseudo.stratified.EUR.scale \
    --ref_freq_chr /users/k1806347/oliverpainfel/Data/1KG/Phase3/freq_files/EUR/1KGPhase3.w_hm3.EUR.chr \
    --plink2 /users/k1806347/oliverpainfel/Software/plink2 \
    --pheno_name ${gwas} \
    --n_cores 5 \
    --output /users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/MegaPRS/${gwas}/UKBB.subset.w_hm3.${gwas}.pseudo.stratified

EOF

sbatch --array 1-$(wc -l /users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/MegaPRS/todo.txt | cut -d&#39; &#39; -f1)%5 /users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/MegaPRS/sbatch.sh
</code></pre>
<hr />
</div>
</div>
</div>
<div id="evaluate-megaprs-polygenic-scores" class="section level1"
number="3">
<h1><span class="header-section-number">3</span> Evaluate MegaPRS
polygenic scores</h1>
<p>Do stratified PGS improve prediction of case control status over
genome-wide PGS? If so, which threshold is best when selecting
significant gene-sets?</p>
<p>Do stratified MegaPRS perform better than stratified pT+clump? Do
stratified pT+clump perform different to PRSet?</p>
<hr />
<div id="test-whether-stratified-pgs-improve-over-genome-wide-pgs"
class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Test whether
stratified PGS improve over genome-wide PGS</h2>
<pre class="r"><code>###
# Create files including pseudovalidated PGS genome-wide, and for gene sets within each significance threshold
### 

library(data.table)

pheno&lt;-c(&#39;Depression&#39;,&#39;Intelligence&#39;,&#39;BMI&#39;,&#39;Height&#39;,&#39;T2D&#39;,&#39;CAD&#39;,&#39;IBD&#39;,&#39;MultiScler&#39;,&#39;RheuArth&#39;,&#39;Breast_Cancer&#39;,&#39;Prostate_Cancer&#39;)
gwas&lt;-c(&#39;DEPR06&#39;,&#39;COLL01&#39;,&#39;BODY04&#39;,&#39;HEIG03&#39;,&#39;DIAB05&#39;,&#39;COAD01&#39;,&#39;CROH01&#39;,&#39;SCLE03&#39;,&#39;RHEU02&#39;,&#39;BRCA01&#39;,&#39;PRCA01&#39;)
prev=c(&#39;0.15&#39;,&#39;NA&#39;,&#39;NA&#39;,&#39;NA&#39;,&#39;0.05&#39;,&#39;0.03&#39;,&#39;0.013&#39;,&#39;0.00164&#39;,&#39;0.005&#39;,&#39;0.125&#39;,&#39;0.125&#39;)


set_pt&lt;-c(0.05, 0.01, 0.001, 0.0001)

for(i in 1:length(pheno)){
    # Read in enrichment results
    set_assoc&lt;-read.table(paste0(&#39;/users/k1806347/oliverpainfel/Analyses/stratified_pgs/magma/set_assoc/&#39;,gwas[i],&#39;/&#39;,gwas[i],&#39;_set_assoc.gsa.out&#39;), header=T, stringsAsFactors = F)

    # Identify pseudovalidated model
    pseudo&lt;-fread(paste0(&#39;/users/k1806347/oliverpainfel/Data/1KG/Phase3/Score_files_for_polygenic/LDAK/&#39;,gwas[i],&#39;/1KGPhase3.w_hm3.&#39;,gwas[i],&#39;.pseudoval.txt&#39;))
    
    # Subset GW profile file to contain the score for this model only
    profile_header&lt;-fread(cmd=paste0(&#39;head -n 3 /users/k1806347/oliverpainfel/Data/UKBB/PRS_for_comparison/1KG_ref/LDAK/&#39;,gwas[i],&#39;/UKBB.subset.w_hm3.&#39;,gwas[i],&#39;.profiles&#39;))
    col_extract&lt;-c(1,2,which(names(profile_header) %in% paste0(gwas[i],&#39;_ldak_Model&#39;,gsub(&#39;.*_&#39;,&#39;&#39;,pseudo$V1[pseudo$V2 == max(pseudo$V2)][1]))))
    
    system(paste0(&quot;cut -f &quot;,paste(col_extract,collapse=&#39;,&#39;),&quot; -d&#39; &#39; /users/k1806347/oliverpainfel/Data/UKBB/PRS_for_comparison/1KG_ref/LDAK/&quot;,gwas[i],&quot;/UKBB.subset.w_hm3.&quot;,gwas[i],&quot;.profiles &gt; /users/k1806347/oliverpainfel/Data/UKBB/PRS_for_comparison/1KG_ref/LDAK/&quot;,gwas[i],&quot;/UKBB.subset.w_hm3.&quot;,gwas[i],&quot;.pseudo.profiles&quot;))
    
    profile_header&lt;-fread(cmd=paste0(&#39;head -n 3 /users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/MegaPRS/&#39;,gwas[i],&#39;/UKBB.subset.w_hm3.&#39;,gwas[i],&#39;.pseudo.stratified.profiles&#39;))

    for(set_pt_i in set_pt){
      col_extract&lt;-c(1,2,which(gsub(paste0(gwas[i],&#39;_&#39;),&#39;&#39;,names(profile_header)) %in% set_assoc$FULL_NAME[set_assoc$P &lt; set_pt_i]))
      
      system(paste0(&quot;cut -f &quot;,paste(col_extract,collapse=&#39;,&#39;),&quot; -d&#39; &#39; /users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/MegaPRS/&quot;,gwas[i],&quot;/UKBB.subset.w_hm3.&quot;,gwas[i],&quot;.pseudo.stratified.profiles &gt; /users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/MegaPRS/&quot;,gwas[i],&quot;/UKBB.subset.w_hm3.&quot;,gwas[i],&quot;.pseudo.stratified.set_pt_&quot;,set_pt_i,&quot;.profiles&quot;))

    # Create predictors file specifying the genome-wide and stratified PGS
    pred_file&lt;-NULL
  
    # Genome-wide PGS
    pred_file&lt;-rbind(pred_file,data.frame( predictors=paste0(&quot;/users/k1806347/oliverpainfel/Data/UKBB/PRS_for_comparison/1KG_ref/LDAK/&quot;,gwas[i],&quot;/UKBB.subset.w_hm3.&quot;,gwas[i],&quot;.pseudo.profiles&quot;), 
  group=&#39;genome_wide&#39;))
    
    # Stratified PGS
    pred_file&lt;-rbind(pred_file,data.frame( predictors=paste0(&quot;/users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/MegaPRS/&quot;,gwas[i],&quot;/UKBB.subset.w_hm3.&quot;,gwas[i],&quot;.pseudo.stratified.set_pt_&quot;,set_pt_i,&quot;.profiles&quot;), 
  group=paste0(&#39;set_pt_&#39;,set_pt_i)))

    dir.create(paste0(&#39;/users/k1806347/oliverpainfel/Analyses/stratified_pgs/assoc/&#39;,pheno[i]), recursive=T)
    write.table(pred_file, paste0(&#39;/users/k1806347/oliverpainfel/Analyses/stratified_pgs/assoc/&#39;,pheno[i],&#39;/UKBB.w_hm3.&#39;,gwas[i],&#39;.EUR-PRSs.GW_vs_strat.set_pt_&#39;,set_pt_i,&#39;.predictor_groups&#39;), row.names=F, col.names=T, quote=F)

    }
}
      

for(i in 1:length(pheno)){
    for(set_pt_i in set_pt){

      system(paste0(&#39;sbatch --mem 30G -n 3 -p cpu /users/k1806347/oliverpainfel/Software/Rscript.sh /users/k1806347/oliverpainfel/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R --pheno /users/k1806347/oliverpainfel/Data/UKBB/Phenotype/PRS_comp_subset/UKBB.&#39;,pheno[i],&#39;.txt --keep /users/k1806347/oliverpainfel/Analyses/PRS_comparison/UKBB_outcomes_for_prediction/ukb18177_glanville_post_qc_id_list.UpdateIDs.fam --out /users/k1806347/oliverpainfel/Analyses/stratified_pgs/assoc/&#39;,pheno[i],&#39;/UKBB.w_hm3.&#39;,gwas[i],&#39;.EUR-PRSs.GW_vs_strat.set_pt_&#39;,set_pt_i,&#39; --n_core 3 --assoc T --outcome_pop_prev &#39;,prev[i],&#39; --predictors /users/k1806347/oliverpainfel/Analyses/stratified_pgs/assoc/&#39;,pheno[i],&#39;/UKBB.w_hm3.&#39;,gwas[i],&#39;.EUR-PRSs.GW_vs_strat.set_pt_&#39;,set_pt_i,&#39;.predictor_groups&#39;))

  }
}</code></pre>
<hr />
</div>
<div id="plot-the-results" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> Plot the results</h2>
<pre class="r"><code>library(data.table)

pheno&lt;-c(&#39;Depression&#39;,&#39;Intelligence&#39;,&#39;BMI&#39;,&#39;Height&#39;,&#39;T2D&#39;,&#39;CAD&#39;,&#39;IBD&#39;,&#39;MultiScler&#39;,&#39;RheuArth&#39;,&#39;Breast_Cancer&#39;,&#39;Prostate_Cancer&#39;)
gwas&lt;-c(&#39;DEPR06&#39;,&#39;COLL01&#39;,&#39;BODY04&#39;,&#39;HEIG03&#39;,&#39;DIAB05&#39;,&#39;COAD01&#39;,&#39;CROH01&#39;,&#39;SCLE03&#39;,&#39;RHEU02&#39;,&#39;BRCA01&#39;,&#39;PRCA01&#39;)
prev=c(&#39;0.15&#39;,&#39;NA&#39;,&#39;NA&#39;,&#39;NA&#39;,&#39;0.05&#39;,&#39;0.03&#39;,&#39;0.013&#39;,&#39;0.00164&#39;,&#39;0.005&#39;,&#39;0.125&#39;,&#39;0.125&#39;)
set_pt&lt;-c(0.05, 0.01, 0.001, 0.0001)

# Read in the evaluation results
res_all&lt;-list()
for(i in 1:length(pheno)){
  for(set_pt_i in set_pt){
    tmp&lt;-fread(paste0(&#39;/users/k1806347/oliverpainfel/Analyses/stratified_pgs/assoc/&#39;,pheno[i],&#39;/UKBB.w_hm3.&#39;,gwas[i],&#39;.EUR-PRSs.GW_vs_strat.set_pt_&#39;,set_pt_i,&#39;.pred_eval.txt&#39;))
    
    tmp&lt;-tmp[,c(&#39;Model&#39;,&#39;R&#39;,&#39;SE&#39;,&#39;P&#39;), with=F]
    
    tmp$Model&lt;-c(&#39;GW&#39;,&#39;Strat&#39;,&#39;Both&#39;)
    tmp$pheno&lt;-pheno[i]
    tmp$set_pt&lt;-set_pt_i
    
    # Insert number of pathways information
    gmt&lt;-readLines(paste0(&#39;/users/k1806347/oliverpainfel/Analyses/stratified_pgs/sets/&#39;,gwas[i],&#39;/sig_set.set_pt_&#39;,set_pt_i,&#39;.gmt&#39;))
    tmp$n_set&lt;-length(gmt)
    
    # Insert the number of SNPs
    snplist&lt;-NULL
    for(k in 1:length(gmt)){
      snplist&lt;-unique(c(snplist, fread(paste0(&#39;/users/k1806347/oliverpainfel/Analyses/stratified_pgs/sets/&#39;,gwas[i],&#39;/snplists/&#39;,gsub(&#39;\t.*&#39;,&#39;&#39;,gmt[k]),&#39;.snplist&#39;), header=F)$V1))
    }
    tmp$n_snp&lt;-length(snplist)
    
    res_all[[paste0(pheno[i],&#39;_&#39;,set_pt_i)]]&lt;-tmp
  }
}

res_all&lt;-do.call(rbind, res_all)
res_all$Model&lt;-factor(res_all$Model, levels=c(&#39;Strat&#39;,&#39;GW&#39;,&#39;Both&#39;))
res_all$set_pt&lt;-factor(res_all$set_pt, levels=unique(res_all$set_pt))

library(ggplot2)
library(cowplot)

# Compare stratified vs GW vs both
png(paste0(&#39;/users/k1806347/oliverpainfel/Analyses/stratified_pgs/assoc/strat_vs_gw_both.png&#39;), units=&#39;px&#39;, width=1750, height=2000, res=300)

  ggplot(res_all, aes(x=pheno, y=R, fill=Model)) +
                          geom_bar(stat=&quot;identity&quot;,width = 0.7, position=position_dodge(width = 0.7)) +
                          geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2, position=position_dodge(width = 0.7)) +
                          labs(y=&quot;Correlation (SE)&quot;, x=&#39;&#39;) +
                          theme_half_open() +
                                  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
                          background_grid(major = &#39;y&#39;, minor = &#39;y&#39;) +
                          facet_grid(rows = vars(set_pt))

dev.off()

# Compare stratfied across set_pt
png(paste0(&#39;/users/k1806347/oliverpainfel/Analyses/stratified_pgs/assoc/strat_across_set_pt.png&#39;), units=&#39;px&#39;, width=2250, height=1200, res=300)
  ggplot(res_all[res_all$Model == &#39;Strat&#39;,], aes(x=pheno, y=R, fill=set_pt)) +
                          geom_bar(stat=&quot;identity&quot;,width = 0.7, position=position_dodge(width = 0.8)) +
                          geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2, position=position_dodge(width = 0.8)) +
                          geom_text(aes(x=pheno, y=R, label = paste0(n_set,&#39;, &#39;, n_snp)), size=3, angle=90, hjust=-0.7, position=position_dodge(width=0.8)) +
                          ylim(NA, max(res_all[res_all$Model == &#39;Strat&#39;,]$R+0.25)) +
                          labs(y=&quot;Correlation (SE)&quot;, x=&#39;&#39;) +
                          theme_half_open() +
                                  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
                          background_grid(major = &#39;y&#39;, minor = &#39;y&#39;)
dev.off()</code></pre>
<hr />
<p><img
src="/scratch/prj/oliverpainfel/Analyses/stratified_pgs/assoc/strat_vs_gw_both.png"
alt="MegaPRS" /> <img
src="/scratch/prj/oliverpainfel/Analyses/stratified_pgs/assoc/strat_across_set_pt.png"
alt="MegaPRS" /></p>
<hr />
</div>
</div>
<div id="try-prset" class="section level1" number="4">
<h1><span class="header-section-number">4</span> Try PRSet</h1>
<p>Here we will generate stratified PRS using PRSet. We will run using
default procedure, and when clumping in advance to compare whether
clumping within each pathway makes a large difference.</p>
<hr />
<div id="download-prsice" class="section level2" number="4.1">
<h2><span class="header-section-number">4.1</span> Download PRSice</h2>
<pre class="bash"><code># Download the latest version of PRSice
mkdir /users/k1806347/oliverpainfel/Software/PRSice
cd /users/k1806347/oliverpainfel/Software/PRSice
wget https://github.com/choishingwan/PRSice/releases/download/2.3.5/PRSice_linux.zip
unzip PRSice_linux.zip
</code></pre>
<hr />
</div>
<div id="download-gtf-file" class="section level2" number="4.2">
<h2><span class="header-section-number">4.2</span> Download GTF
file</h2>
<pre class="r"><code>cd /users/k1806347/oliverpainfel/Data/Gene_Locations
wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_37/GRCh37_mapping/gencode.v37lift37.annotation.gtf.gz
gunzip gencode.v37lift37.annotation.gtf.gz</code></pre>
<hr />
</div>
<div id="create-score-files" class="section level2" number="4.3">
<h2><span class="header-section-number">4.3</span> Create score
files</h2>
<pre class="bash"><code># I already had a script for running PRSice. I need to update it to allow for PRSset.

# Don&#39;t use every threshold

#########
# First run genome-wide
#########
pheno=$(echo Depression Intelligence BMI Height T2D CAD IBD MultiScler RheuArth Breast_Cancer Prostate_Cancer)
gwas=$(echo DEPR06 COLL01 BODY04 HEIG03 DIAB05 COAD01 CROH01 SCLE03 RHEU02 BRCA01 PRCA01)

# Create directory
mkdir -p /users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/PRSset

# Create file listing GWAS that haven&#39;t been processed.
&gt; /users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/PRSset/todo.txt
for i in $(seq 1 11);do
gwas_i=$(echo ${gwas} | cut -f ${i} -d &#39; &#39;)
pheno_i=$(echo ${pheno} | cut -f ${i} -d &#39; &#39;)

if [ ! -f /users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/PRSset/score_files/${gwas}/1KGPhase3.w_hm3.${gwas}.EUR.scale ]; then
echo ${gwas_i} ${pheno_i} &gt;&gt; /users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/PRSset/todo.txt
fi
done

# Create shell script to run using sbatch
cat &gt; /users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/PRSset/sbatch.sh &lt;&lt; &#39;EOF&#39;
#!/bin/sh

#SBATCH -p cpu
#SBATCH --mem 40G
#SBATCH -n 5
#SBATCH -J strat_pgs
#SBATCH --exclude erc-hpc-comp001

gwas=$(awk -v var=&quot;$SLURM_ARRAY_TASK_ID&quot; &#39;NR == var {print $1}&#39; /users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/PRSset/todo.txt)
pheno=$(awk -v var=&quot;$SLURM_ARRAY_TASK_ID&quot; &#39;NR == var {print $2}&#39; /users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/PRSset/todo.txt)

echo $gwas
echo $pheno

/users/k1806347/oliverpainfel/Software/Rscript.sh /users/k1806347/oliverpainfel/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator_dense/polygenic_score_file_creator_sparse_2_prset.R \
  --ref_plink_chr /users/k1806347/oliverpainfel/Data/1KG/Phase3/1KGPhase3.w_hm3.chr \
  --ref_keep /users/k1806347/oliverpainfel/Data/1KG/Phase3/keep_files/EUR_samples.keep \
  --sumstats /users/k1806347/oliverpainfel/Data/GWAS_sumstats/prs_comparison/cleaned/${gwas}.cleaned.gz \
  --plink /users/k1806347/oliverpainfel/Software/plink1.9.sh \
  --prsice_path /users/k1806347/oliverpainfel/Software/PRSice \
  --rscript /users/k1806347/oliverpainfel/Software/Rscript.sh \
  --memory 3000 \
  --fastscore T \
  --gmt /users/k1806347/oliverpainfel/Analyses/stratified_pgs/sets/${gwas}/sig_set.set_pt_0.05.gmt \
  --gtf /users/k1806347/oliverpainfel/Data/Gene_Locations/gencode.v37lift37.annotation.gtf \
  --fastscore T \
  --output /users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/PRSset/score_files/${gwas}_3/1KGPhase3.w_hm3.${gwas} \
  --ref_pop_scale /users/k1806347/oliverpainfel/Data/1KG/Phase3/super_pop_keep.list

EOF

sbatch --array 1-$(wc -l /users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/PRSset/todo.txt | cut -d&#39; &#39; -f1)%5 /users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/PRSset/sbatch.sh


opt$ref_plink_chr&lt;-&#39;/users/k1806347/oliverpainfel/Data/1KG/Phase3/1KGPhase3.w_hm3.chr&#39;
opt$ref_keep&lt;-&#39;/users/k1806347/oliverpainfel/Data/1KG/Phase3/keep_files/EUR_samples.keep&#39;
opt$sumstats&lt;-&#39;/users/k1806347/oliverpainfel/Data/GWAS_sumstats/prs_comparison/cleaned/DEPR06.cleaned.gz&#39;
opt$plink&lt;-&#39;/users/k1806347/oliverpainfel/Software/plink1.9.sh&#39;
opt$prsice_path&lt;-&#39;/users/k1806347/oliverpainfel/Software/PRSice&#39;
opt$rscript&lt;-&#39;/users/k1806347/oliverpainfel/Software/Rscript.sh&#39;
opt$memory&lt;-3000
opt$fastscore&lt;-T
opt$gmt&lt;-&#39;/users/k1806347/oliverpainfel/Analyses/stratified_pgs/sets/DEPR06/sig_set.set_pt_0.05.gmt&#39;
opt$gtf&lt;-&#39;/users/k1806347/oliverpainfel/Data/Gene_Locations/gencode.v37lift37.annotation.gtf&#39;
opt$fastscore&lt;-T
opt$output&lt;-&#39;/users/k1806347/oliverpainfel/Data/UKBB/stratified_pgs/PRSset/score_files/DEPR06_3/1KGPhase3.w_hm3.DEPR06&#39;
opt$ref_pop_scale&lt;-&#39;/users/k1806347/oliverpainfel/Data/1KG/Phase3/super_pop_keep.list&#39;
</code></pre>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
