<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Ancestry-based prediction</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-sm-12 col-md-4 col-lg-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-sm-12 col-md-8 col-lg-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GenoPred</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fas fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Ancestry-based prediction</h1>

</div>


<style>
p.caption {
  font-size: 1.5em;
}
</style>
<style type="text/css">
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
</style>
<hr />
<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>Genotypes vary in frequency between different ancestral groups. As such, it is important to consider an individuals’ ancestry in genetic association studies to avoid population stratification and to avoid false positives. However, there are many aspects of health that do vary between ancestral groups, however we are unable to determine whether a variant is causally associated with an outcome, or if it is associated by chance.</p>
<p>In the context of prediction, where the aim is to explain the maximum amount of variance in a phenotype with no consideration for what is causal, difference in peoples ancestry may be a useful predictor. Therefore, here we explore whether principal components of population structure (mainly ancestry), typically used as covariates in association studies, can be used a predictor for a range of health outcomes.</p>
<p>Principal components are typically estimated within the sample of interest to capture sources of population structure which are unique to the sample. However, in this context we want the principal components (PCs), to capture variance that can be generalised across samples, to improve the external validity of our prediction models. One way to achieve this is to derive the PCs in a reference sample, and then project the derived PCs into subsequent target samples, ensuring that PCs correspond to the same variance across samples.</p>
<p>The allele frequency differences between ancestries are not only captured by PCs, but also by other genotype-based scores. Hence why PCs are used as covariates for inference when using polygenic risk scores. Therefore, it is also of interest to determine whether PCs can improve prediction over and above polygenic scores.</p>
<p><br/></p>
<hr />
</div>
<div id="aims" class="section level1">
<h1><span class="header-section-number">2</span> Aims</h1>
<ol style="list-style-type: decimal">
<li>Test whether reference-projected PCs can significantly predict variance in a range of phenotypes.</li>
<li>Test whether reference-projected PCs can significantly improve prediction over polygenic scores alone.</li>
<li>Compare the predictive utility of ancestry-specific PCs and across-ancestry PCs.</li>
<li>Determine the optimal number of PCs for prediction.</li>
</ol>
<p><br/></p>
<hr />
</div>
<div id="methods" class="section level1">
<h1><span class="header-section-number">3</span> Methods</h1>
<div id="samples" class="section level2">
<h2><span class="header-section-number">3.1</span> Samples</h2>
<ul>
<li>UK Biobank</li>
<li>TEDS</li>
</ul>
</div>
<div id="outcomes" class="section level2">
<h2><span class="header-section-number">3.2</span> Outcomes</h2>
<ul>
<li>UK Biobank
<ul>
<li>Depression (binary)</li>
<li>Intelligence (continuous)</li>
<li>Body mass index (BMI - continuous)</li>
<li>Height (continuous)</li>
<li>Coronary Artery Disease (CAD - Binary)</li>
<li>Type II Diabetes (T2D - Binary)</li>
</ul></li>
<li>TEDS
<ul>
<li>ADHD traits (continuous)</li>
<li>Height (continuous)</li>
<li>Body mass index (BMI - continuous)</li>
<li>GCSE scores (continuous)</li>
</ul></li>
</ul>
</div>
<div id="genotypic-data" class="section level2">
<h2><span class="header-section-number">3.3</span> Genotypic data</h2>
<p>HapMap3 variants were extracted from the HRC imputed genetic data, and converted to hard-call PLINK format with no hard-call threshold to maximise overlap with the HapMap3 SNP list. Individuals were excluded if they had extensive missing data, had non-European ancestry, or were closely-related to other individuals in the sample. This was done prior to this project.</p>
</div>
<div id="ancestry-scoring" class="section level2">
<h2><span class="header-section-number">3.4</span> Ancestry scoring</h2>
<p>Calculation of the first 100 European-specific and 100 all-ancestry PCs was perfomed using a reference standardised approach (see <a href="https://opain.github.io/GenoPred/Pipeline_prep.html#3_ancestry_scoring">here</a>). In brief, PCs were derived in the 1000 Genomes reference, then projected onto the UK Biobank sample, and subsequent factor scores were then centered and scaled according to the mean and standard deviation of the PCs in an ancestry matched sample. The standard deviation of PCs in UKBB and TEDS was estimated for comparison as this may effect the variance that PCs can explain in each target sample.</p>
</div>
<div id="polygenic-scoring" class="section level2">
<h2><span class="header-section-number">3.5</span> Polygenic scoring</h2>
<p>Polygenic scores were derived using PRScs, a Bayesian shrinkage method. Polygenic scores based on a range of global shrinkage parameters were used as this approach was shown to explain the most variance (see <a href="https://opain.github.io/GenoPred/Determine_optimal_polygenic_scoring_approach.html">here</a>). Furthermore, PRScs-based polygenic scores for a range of outcomes were included, as this approach has been shown to increase the variance explained by polygenic scores (see <a href="https://opain.github.io/GenoPred/Determine_optimal_polygenic_scoring_approach.html">here</a>). Maximising the variance explained by polygenic scores will provide more robust evidence that PCs can explain additional variance to polygenic scores, and therefore are a useful predictor.</p>
<p>Polygenic scores were calculated in UK Biobankand TEDS using a reference-standardised approach (see <a href="https://opain.github.io/GenoPred/Pipeline_prep.html#4_polygenic_scoring">here</a>). In brief, all scores were derived using HapMap3 SNPs only, modelling LD based on European individuals within the 1000 Genomes reference. Any HapMap3 missing in the target sample are imputed using the allele frequency measured in the European subset of the 1000 Genomes reference. Score were centreed and scaled based on the mean and standard deviation of scores in an ancestry matched samples.</p>
</div>
<div id="estimating-predictive-ability" class="section level2">
<h2><span class="header-section-number">3.6</span> Estimating predictive ability</h2>
<p>All models were derived using elastic-net regularisation to reduce the likelihood of overfitting and account for multicollinearity when modelling highly correlated predictors. 10-fold cross validation was performed using 80% of individuals to identify optimal parameters, with subsequent test-set validation in the remaining 20% of individuals to estimate the predictive utility among individuals not included in the parameter selection process. Permutation testing was used to determine whether the difference in variance explained was significantly greater than 0.</p>
<p>Model building and evaluation was performed using an Rscript called Model_builder.R (more information <a href="https://github.com/opain/GenoPred/tree/master/Scripts/Model_builder">here</a>).</p>
</div>
<div id="code" class="section level2">
<h2><span class="header-section-number">3.7</span> Code</h2>
<div id="calculate-descriptives-of-pcs-in-teds-and-ukbb" class="section level3">
<h3><span class="header-section-number">3.7.1</span> Calculate descriptives of PCs in TEDS and UKBB</h3>
<details>
<p><summary>Evaluate PCs alone</summary></p>
<pre class="bash"><code>module add general/R/3.5.0
R
library(data.table)

##
# UKBB
##

# Read in the UKBB PCs
UKBB_EUR_PCs&lt;-fread(&#39;/users/k1806347/brc_scratch/Data/UKBB/Projected_PCs/EUR/UKBB.w_hm3.EUR.eigenvec&#39;)
UKBB_All_PCs&lt;-fread(&#39;/users/k1806347/brc_scratch/Data/UKBB/Projected_PCs/EUR/AllAncestry/UKBB.w_hm3.AllAncestry.EUR.eigenvec&#39;)

# Read in the TEDS PCs
TEDS_EUR_PCs&lt;-fread(&#39;/users/k1806347/brc_scratch/Data/TEDS/Projected_PCs/EUR/EUR_specific/TEDS.w_hm3.EUR.EUR.eigenvec&#39;)
TEDS_All_PCs&lt;-fread(&#39;/users/k1806347/brc_scratch/Data/TEDS/Projected_PCs/EUR/AllAncestry/TEDS.w_hm3.AllAncestry.EUR.eigenvec&#39;)

# Extract individuals with European ancestry
UKBB_EUR_keep&lt;-fread(&#39;/users/k1806347/brc_scratch/Data/UKBB/Projected_PCs/Ancestry_idenitfier/UKBB.w_hm3.AllAncestry.EUR.keep&#39;)
UKBB_EUR_PCs&lt;-UKBB_EUR_PCs[(UKBB_EUR_PCs$IID %in% UKBB_EUR_keep$V2),]
UKBB_All_PCs&lt;-UKBB_All_PCs[(UKBB_All_PCs$IID %in% UKBB_EUR_keep$V2),]

TEDS_EUR_keep&lt;-fread(&#39;/users/k1806347/brc_scratch/Data/TEDS/Projected_PCs/Ancestry_idenitfier/TEDS.w_hm3.AllAncestry.EUR.keep&#39;)
TEDS_EUR_PCs&lt;-TEDS_EUR_PCs[(TEDS_EUR_PCs$IID %in% TEDS_EUR_keep$V2),]
TEDS_All_PCs&lt;-TEDS_All_PCs[(TEDS_All_PCs$IID %in% TEDS_EUR_keep$V2),]

# Calculate mean and sd
UKBB_EUR_PCs_desc&lt;-NULL
for(i in names(UKBB_EUR_PCs)[-1:-2]){
  tmp&lt;-data.frame(Mean_UKBB_EUR=mean(UKBB_EUR_PCs[[i]]),
                  SD_UKBB_EUR=sd(UKBB_EUR_PCs[[i]]))
  UKBB_EUR_PCs_desc&lt;-rbind(UKBB_EUR_PCs_desc,tmp)
}

UKBB_All_PCs_desc&lt;-NULL
for(i in names(UKBB_All_PCs)[-1:-2]){
  tmp&lt;-data.frame(Mean_UKBB_All=mean(UKBB_All_PCs[[i]]),
                  SD_UKBB_All=sd(UKBB_All_PCs[[i]]))
  UKBB_All_PCs_desc&lt;-rbind(UKBB_All_PCs_desc,tmp)
}

TEDS_EUR_PCs_desc&lt;-NULL
for(i in names(TEDS_EUR_PCs)[-1:-2]){
  tmp&lt;-data.frame(Mean_TEDS_EUR=mean(TEDS_EUR_PCs[[i]]),
                  SD_TEDS_EUR=sd(TEDS_EUR_PCs[[i]]))
  TEDS_EUR_PCs_desc&lt;-rbind(TEDS_EUR_PCs_desc,tmp)
}

TEDS_All_PCs_desc&lt;-NULL
for(i in names(TEDS_All_PCs)[-1:-2]){
  tmp&lt;-data.frame(Mean_TEDS_All=mean(TEDS_All_PCs[[i]]),
                  SD_TEDS_All=sd(TEDS_All_PCs[[i]]))
  TEDS_All_PCs_desc&lt;-rbind(TEDS_All_PCs_desc,tmp)
}

PCs_desc&lt;-do.call(cbind, list(UKBB_EUR_PCs_desc,UKBB_All_PCs_desc,TEDS_EUR_PCs_desc,TEDS_All_PCs_desc))
PCs_desc&lt;-data.frame(PC=1:dim(PCs_desc)[1], PCs_desc)

write.csv(PCs_desc[c(&#39;PC&#39;,&#39;Mean_UKBB_EUR&#39;,&#39;SD_UKBB_EUR&#39;,&#39;Mean_UKBB_All&#39;,&#39;SD_UKBB_All&#39;)], &#39;/users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/UKBB_PC_descriptives.csv&#39;, col.names=T, row.names=F, quote=F)
write.csv(PCs_desc[c(&#39;PC&#39;,&#39;Mean_TEDS_EUR&#39;,&#39;SD_TEDS_EUR&#39;,&#39;Mean_TEDS_All&#39;,&#39;SD_TEDS_All&#39;)], &#39;/users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/TEDS_PC_descriptives.csv&#39;, col.names=T, row.names=F, quote=F)

q()
n
</code></pre>
</details>
</div>
</div>
<div id="pcs-alone" class="section level2">
<h2><span class="header-section-number">3.8</span> PCs alone</h2>
<div id="uk-biobank" class="section level3">
<h3><span class="header-section-number">3.8.1</span> UK Biobank</h3>
<details>
<p><summary>Evaluate PCs alone</summary></p>
<pre class="bash"><code># Make required directories
for pheno_i in $(echo Depression Intelligence BMI Height T2D CAD);do
mkdir -p /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPCs
done

######
# Create a file to group the predictors
######

module add general/R/3.5.0
R
library(data.table)

groups&lt;-data.frame( predictor=c(&#39;/users/k1806347/brc_scratch/Data/UKBB/Projected_PCs/EUR/UKBB.w_hm3.EUR.eigenvec&#39;,&#39;/users/k1806347/brc_scratch/Data/UKBB/Projected_PCs/EUR/AllAncestry/UKBB.w_hm3.AllAncestry.EUR.eigenvec&#39;),
                    group=c(&#39;EUR_PCs&#39;,&#39;AllAncestry_PCs&#39;))
                    
write.table(groups, &#39;/users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/UKBB-projected-PCs.predictor_groups&#39;, col.names=T, row.names=F, quote=F)
q()
n

######
# Derive and evaluate models
######
pheno=$(echo Depression Intelligence BMI Height T2D CAD)
pheno_file=$(echo ever_depressed_pheno_final.UpdateIDs.txt UKBB_Fluid.intelligence.score.UpdateIDs.pheno UKBB_BMI.score.UpdateIDs.pheno UKBB_Height.score.UpdateIDs.pheno t2d_only_111119.UpdateIDs.txt cad_only_111119.UpdateIDs.txt)
gwas=$(echo DEPR06 COLL01 BODY03 HEIG03 DIAB05 COAD01)
prev=$(echo 0.15 NA NA NA 0.05 0.03)

for i in $(seq 1 6);do
pheno_i=$(echo ${pheno} | cut -f ${i} -d &#39; &#39;)
pheno_file_i=$(echo ${pheno_file} | cut -f ${i} -d &#39; &#39;)
gwas_i=$(echo ${gwas} | cut -f ${i} -d &#39; &#39;)
prev_i=$(echo ${prev} | cut -f ${i} -d &#39; &#39;)

qsub -l h_vmem=10G -pe smp 5 /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/Scripts/Model_builder.R \
--pheno /users/k1806347/brc_scratch/Data/UKBB/Phenotype/${pheno_i}/${pheno_file_i} \
--keep /users/k1806347/brc_scratch/Data/UKBB/Projected_PCs/Ancestry_idenitfier/UKBB.w_hm3.AllAncestry.EUR.keep \
--out /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPCs/UKBB-projected-PCs \
--n_core 5 \
--assoc T \
--outcome_pop_prev ${prev_i} \
--predictors /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/UKBB-projected-PCs.predictor_groups
done
</code></pre>
</details>
</div>
<div id="teds" class="section level3">
<h3><span class="header-section-number">3.8.2</span> TEDS</h3>
<details>
<p><summary>Evaluate PCs alone</summary></p>
<pre class="bash"><code># Make required directories
for pheno_i in $(echo Height21 BMI21 GCSE ADHD);do
mkdir -p /users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/${pheno_i}/Association_withPCs
done

######
# Create a file to group the predictors
######

module add general/R/3.5.0
R
library(data.table)

groups&lt;-data.frame( predictor=c(&#39;/users/k1806347/brc_scratch/Data/TEDS/Projected_PCs/EUR/EUR_specific/TEDS.w_hm3.EUR.EUR.eigenvec&#39;,&#39;/users/k1806347/brc_scratch/Data/TEDS/Projected_PCs/EUR/AllAncestry/TEDS.w_hm3.AllAncestry.EUR.eigenvec&#39;),
                    group=c(&#39;EUR_PCs&#39;,&#39;AllAncestry_PCs&#39;))
                    
write.table(groups, &#39;/users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/TEDS-projected-PCs.predictor_groups&#39;, col.names=T, row.names=F, quote=F)
q()
n

######
# Derive and evaluate models
######
pheno=$(echo Height21 BMI21 GCSE ADHD)
gwas=$(echo HEIG03 BODY11 EDUC03 ADHD04)
prev=$(echo NA NA NA NA)

for i in $(seq 1 4);do
pheno_i=$(echo ${pheno} | cut -f ${i} -d &#39; &#39;)
gwas_i=$(echo ${gwas} | cut -f ${i} -d &#39; &#39;)
prev_i=$(echo ${prev} | cut -f ${i} -d &#39; &#39;)

qsub -l h_vmem=6G -pe smp 3 /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/Scripts/Model_builder.R \
--pheno /users/k1806347/brc_scratch/Data/TEDS/Phenotypic/Derived_outcomes/TEDS_${pheno_i}.txt \
--keep /users/k1806347/brc_scratch/Data/TEDS/Projected_PCs/Ancestry_idenitfier/TEDS.w_hm3.AllAncestry.EUR.keep \
--out /users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/${pheno_i}/Association_withPCs/TEDS-projected-PCs \
--n_core 3 \
--assoc T \
--outcome_pop_prev ${prev_i} \
--predictors /users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/TEDS-projected-PCs.predictor_groups
done
</code></pre>
</details>
<p><br/></p>
</div>
</div>
<div id="pcs-prss" class="section level2">
<h2><span class="header-section-number">3.9</span> PCs + PRSs</h2>
<p>Here we will test PCs can significantly improve prediction over multiple-trait PRSs alone.</p>
</div>
<div id="prss-controlled-for-pcs" class="section level2">
<h2><span class="header-section-number">3.10</span> PRSs controlled for PCs</h2>
<p>Here will test whether PRSs controlled for PCs improve prediction. We will adjust PRSs for PCs using the EUR 1KG reference sample.</p>
<hr />
</div>
</div>
<div id="results" class="section level1">
<h1><span class="header-section-number">4</span> Results</h1>
<div id="descriptives" class="section level2">
<h2><span class="header-section-number">4.1</span> Descriptives</h2>
<details>
<p><summary>Show PC descriptives</summary></p>
</details>
<p><br/></p>
</div>
<div id="pcs-alone-1" class="section level2">
<h2><span class="header-section-number">4.2</span> PCs alone</h2>
<div id="uk-biobank-1" class="section level3">
<h3><span class="header-section-number">4.2.1</span> UK biobank</h3>
<details>
<p><summary>Show UK Biobank results table</summary></p>
</details>
<p><br/></p>
</div>
<div id="teds-1" class="section level3">
<h3><span class="header-section-number">4.2.2</span> TEDS</h3>
<details>
<p><summary>Show TEDS results table</summary></p>
</details>
<p><br/></p>
<p>PCs show significant prediction in the test set of UKBB, but not in TEDS. The estimates of correlation between predicted and observed values are more accurate in UKBB than TEDS due to a larger sample size. TEDS may not have sufficient sample size to detect significant prediction using PCs.</p>
<p><br/></p>
<hr />
</div>
</div>
</div>
<div id="discussion" class="section level1">
<h1><span class="header-section-number">5</span> Discussion</h1>
<p><br/></p>
<hr />
</div>
<div id="conclusion" class="section level1">
<h1><span class="header-section-number">6</span> Conclusion</h1>
<hr />
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
