---
title: "CrossMeta"
output: 
  html_document:
    toc: true
    theme: united
    toc_depth: 3
    number_sections: true
    toc_float: true
    fig_caption: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

<style>
p.caption {
  font-size: 1.5em;
}
</style>

```{css, echo=F}
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
```

***

# Introduction

This script documents the evaluation and comparison of different approaches for combining GWAS from different populations to improve polygenic prediction.

We evaluate and compare the following approaches:
- Modelling PRS from each GWAS, and a PRS from standard inverse-variance meta-GWAS
- Modelling PRS from each GWAS, and a PRS from genome-wide reweighted meta-GWAS
- Modelling PRS from each GWAS, and a PRS from SNP-level reweighted meta-GWAS
- Modelling PRS from PRS-CSx

We focus on depression, combining GWAS summary statistics from the following studies:

- EUR 
-- PGC2 MDD (incl. 23andMe)
-- Biobank Japan (BBJ)
-- CONVERGE
-- 23andMe
-- PGC Cross Population (Karolina)

- AMR 
-- 23andMe

- AFR 
-- 23andMe

We also include a comparison of methods using Height and BMI, using the following GWAS summary statistics:

- Height
-- EUR
--- Wood et al
-- EAS
--- BBJ

- BMI
-- EUR
--- Locke et al
-- EAS
--- BBJ

We use the UK Biobank sample as our target sample, split into the five 1KG reference super-populations.

***

# Methods

***

## UK Biobank (UKB)

***

### Ancestry identification

We used the observed genotype data in UKB to classify individuals into the five 1000 Genomes (1KG) reference super-populations. This process has been previously documented [here](https://opain.github.io/UKB-GenoPrep/ancestry_identification.html). In summary, genetic principal components (PCs) were derived in the 1KG Phase 3 reference, and then first 6 were used to created an elastic net model predicting super-population membership. Then the reference PCs were projected into the UKB sample, and the elastic net model trained in 1KG was the applied in UKB to predict super-population membership. Subsequently, within super-population quality control was performed, which has been previously documented [here](https://opain.github.io/UKB-GenoPrep/quality_control.html). In summary, PCs were generated in each  super-population, k-means clustering was then performed to define centroids, and outliers from centroids were removed. Furthermore, individuals were removed if there were reported by UKB as heterozygosity or missingness outliers, genotype-phenotype sex mismatch, or genotype missingness > 0.02 (after removal of SNPs with missingness > 0.05).

***

### Phenotype definition

Depression: To maximise statistical power, we used a continuous measure of depression, termed the interalising factor, previously described [here](https://pubmed.ncbi.nlm.nih.gov/32519625/). In summary, the internalising factor is the first factor of depressive symptoms assessed in UKB.

```{bash, eval=F, echo=T}
# Create symbolic links to the three phenotype file with codes corresponding to GWAS codes (i.e. MDD, Height, BMI)
mkdir -p /users/k1806347/brc_scratch/Analyses/CrossMeta/Phenotype
ln -s /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/UKB.InternalisingFactor.UpdateIDs.txt /users/k1806347/brc_scratch/Analyses/CrossMeta/Phenotype/MDD.txt
ln -s /users/k1806347/brc_scratch/Data/UKBB/Phenotype/Height/UKBB_Height.score.UpdateIDs.pheno /users/k1806347/brc_scratch/Analyses/CrossMeta/Phenotype/Height.txt
ln -s /users/k1806347/brc_scratch/Data/UKBB/Phenotype/BMI/UKBB_BMI.score.UpdateIDs.pheno /users/k1806347/brc_scratch/Analyses/CrossMeta/Phenotype/BMI.txt

```

***

## GWAS summary statistics

Below is a table listing details of the GWAS summary statistics used in this study:

```{r, eval=F, echo=T}

# Depression
gwas<-data.frame(code=c('CONVERGE_MDD_EAS'),
                 discovery=c('CONVERGE'),
                 phenotype=c('MDD'),
                 population=c('EAS'),
                 pmid=c(26176920),
                 N=c(10640),
                 Ncas=c(5303),
                 Ncon=c(5337))

gwas<-rbind(gwas,data.frame( code=c('BBJ_MDD_EAS'),
                             discovery=c('BBJ'),
                             phenotype=c('MDD'),
                             population=c('EAS'),
                             pmid=c(34594039),
                             N=c(178630),
                             Ncas=c(836),
                             Ncon=c(177794)))

gwas<-rbind(gwas,data.frame( code=c('PGC2_MDD_EUR'),
                             discovery=c('PGC2'),
                             phenotype=c('MDD'),
                             population=c('EUR'),
                             pmid=c(29700475),
                             N=c(143265),
                             Ncas=c(45591),
                             Ncon=c(97674)))

# Height
gwas<-rbind(gwas,data.frame( code=c('Wood_Height_EUR'),
                             discovery=c('Wood'),
                             phenotype=c('Height'),
                             population=c('EUR'),
                             pmid=c(25282103),
                             N=c(253288),
                             Ncas=c(NA),
                             Ncon=c(NA)))

gwas<-rbind(gwas,data.frame( code=c('BBJ_Height_EAS'),
                             discovery=c('BBJ'),
                             phenotype=c('Height'),
                             population=c('EAS'),
                             pmid=c(26367797),
                             N=c(159095),
                             Ncas=c(NA),
                             Ncon=c(NA)))

# BMI
gwas<-rbind(gwas,data.frame( code=c('Locke_BMI_EUR'),
                             discovery=c('Locke'),
                             phenotype=c('BMI'),
                             population=c('EUR'),
                             pmid=c(25673413),
                             N=c(339224),
                             Ncas=c(NA),
                             Ncon=c(NA)))

gwas<-rbind(gwas,data.frame( code=c('BBJ_BMI_EAS'),
                             discovery=c('BBJ'),
                             phenotype=c('BMI'),
                             population=c('EAS'),
                             pmid=c(28892062),
                             N=c(158284),
                             Ncas=c(NA),
                             Ncon=c(NA)))

dir.create('/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats', recursive=T)
write.csv(gwas, '/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/gwas_list.txt', row.names=F, quote=F)

```

```{r,eval=T, echo=F}
library(knitr)

gwas_list<-read.csv('/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/gwas_list.txt')
kable(gwas_list, rownames = FALSE, caption='List of GWAS used')
```

**Note**: PGC2_MDD_EUR does not include data from 23andMe or UKB.

***

### Download and format GWAS

```{bash, eval=F, echo=T}
mkdir /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/
mkdir /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/
mkdir /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/

```

<details><summary>PGC_MDD_EAS</summary>

```{R, eval=F, echo=T}
#####
# PGC MDD EAS
#####
# https://pubmed.ncbi.nlm.nih.gov/34586374/
# Downloaded from https://figshare.com/ndownloader/files/31424374
# I requested access to sumstats exluding UKB from Xiangrui Meng

# I will format them, and then QC
library(data.table)
sum_auto<-fread('/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/jamapsy_Giannakopoulou_2021_exclude_whi_23andMe_ukb.txt.gz')

sum_auto<-sum_auto[,c('MarkerName','chr','pos','Allele1','Allele2','Freq1','Effect','StdErr','P.SE'),with=F]
names(sum_auto)<-c('SNP','CHR','POS','A1','A2','FREQ','BETA','SE','P')

# N is derived from supp etable 2 (194548-(133+366))
sum_auto$N<-194049

sum_auto$A1<-toupper(sum_auto$A1)
sum_auto$A2<-toupper(sum_auto$A2)

fwrite(sum_auto, '/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/PGC_MDD_EAS', sep=' ', quote=F, na='NA')

system('~/brc_scratch/Software/pigz /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/PGC_MDD_EAS')

```

</details>

<details><summary>CONVERGE_MDD_EAS</summary>

```{bash, eval=F, echo=T}
#####
# CONVERGE MDD
#####
# https://pubmed.ncbi.nlm.nih.gov/26176920/
# Downloaded from https://figshare.com/ndownloader/files/10374393

wget --no-check-certificate -O /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/CONVERGE_MDD_EAS https://figshare.com/ndownloader/files/10374393

```

```{R, eval=F, echo=T}
# I will format them, and then QC
library(data.table)
sum_auto<-fread('/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/CONVERGE_MDD_EAS')

# The sumstats README on figshare has labelled the allele incorrectly. FRQ and OR in fact correspond to A2.
names(sum_auto)<-c('CHR','POS','SNP','A2','A1','FREQ','OR','SE','P','REF_FREQ','HWE.P','INFO')

sum_auto$N<-10640

fwrite(sum_auto, '/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/CONVERGE_MDD_EAS', sep=' ', quote=F, na='NA')

system('~/brc_scratch/Software/pigz /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/CONVERGE_MDD_EAS')

```

</details>

<details><summary>BBJ_MDD_EAS</summary>

```{bash, eval=F, echo=T}

wget --no-check-certificate -O /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/BBJ_MDD_EAS.zip https://pheweb.jp/download/Depression

unzip /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/BBJ_MDD_EAS.zip -d /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/BBJ_MDD_EAS

```

```{R, eval=F, echo=T}
# I will format them, and then QC
library(data.table)
sum_auto<-fread('/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/BBJ_MDD_EAS/hum0197.v3.BBJ.Dep.v1/GWASsummary_Depression_Japanese_SakaueKanai2020.auto.txt.gz')
sum_auto<-sum_auto[,c('CHR','POS','SNPID','Allele1','Allele2','AF_Allele2','imputationInfo','N','BETA','SE','p.value','AF.Cases','AF.Controls'),with=F]

# Note. Effect allele is A2. Flip the alleles so A1 is the allele BETA/allele frequencies correspond to.
sum_auto$Allele3<-sum_auto$Allele2
sum_auto$Allele2<-sum_auto$Allele1
sum_auto$Allele1<-sum_auto$Allele3
sum_auto$Allele3<-NULL

names(sum_auto)<-c('CHR','POS','SNP','A1','A2','FREQ','INFO','N','BETA','SE','p.value','FREQ.Cases','FREQ.Controls')

sum_auto$NCAS<-836
sum_auto$NCON<-177794

fwrite(sum_auto, '/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/BBJ_MDD_EAS', sep=' ', quote=F, na='NA')

system('~/brc_scratch/Software/pigz /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/BBJ_MDD_EAS')

```

</details>

<details><summary>PGC2_MDD_EUR</summary>

```{bash, eval=F, echo=T}
#######
# Wray et al. MDD (excl. 23andMe and excl. UKB)
#######
# https://www.nature.com/articles/s41588-018-0090-3

cp /scratch/groups/ukbiobank/sumstats/cleaned/DEPR07.gz /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/PGC2_MDD_EUR.gz

```

</details>

***

<details><summary>BBJ_Height_EAS</summary>

```{bash, eval=F, echo=T}
# Download the sumstats
wget --no-check-certificate -O /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/BBJ_Height_EAS.gz http://jenger.riken.jp/135analysisresult_qtl_download/
```

```{r, eval=F, echo=T}
# Format the sumstats for qc script
library(data.table)

# Read in reference data
ref<-fread('/users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.GW.bim')
ref$V3<-NULL

# We need to insert RSID, N and update column names
sum_auto<-fread('/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/BBJ_Height_EAS.gz')

sum_auto_matched<-merge(sum_auto, ref, by.x=c('CHR','POS','REF','ALT'), by.y=c('V1','V4','V5','V6'))
sum_auto_swapped<-merge(sum_auto, ref, by.x=c('CHR','POS','REF','ALT'), by.y=c('V1','V4','V6','V5'))

sum_auto_withRSID<-rbind(sum_auto_matched, sum_auto_swapped)

sum_auto_withRSID<-sum_auto_withRSID[,c('CHR','POS','REF','ALT','ALT_freq','BETA','SE','P_INF','V2'), with=F]
names(sum_auto_withRSID)<-c('CHR','POS','A2','A1','FREQ','BETA','SE','P','SNP')

sum_auto_withRSID$N<-159095

fwrite(sum_auto_withRSID, '/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/BBJ_Height_EAS', sep=' ', quote=F, na='NA')

system('~/brc_scratch/Software/pigz /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/BBJ_Height_EAS')

```

</details>

<details><summary>Wood_Height_EUR</summary>

```{bash, eval=F, echo=T}

cp /scratch/groups/ukbiobank/sumstats/cleaned/HEIG03.gz /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/Wood_Height_EUR.gz

```

</details>

***

<details><summary>BBJ_BMI_EAS</summary>

```{bash, eval=F, echo=T}
# Download the sumstats
wget --no-check-certificate -O /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/BBJ_BMI_EAS.gz http://jenger.riken.jp/1analysisresult_qtl_download/

```

```{r, eval=F, echo=T}
library(data.table)

# Read in reference data
ref<-fread('/users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.GW.bim')
ref$V3<-NULL

# We need to insert RSID, N and update column names
sum_auto<-fread('/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/BBJ_BMI_EAS.gz')

sum_auto_matched<-merge(sum_auto, ref, by.x=c('CHR','POS','REF','ALT'), by.y=c('V1','V4','V5','V6'))
sum_auto_swapped<-merge(sum_auto, ref, by.x=c('CHR','POS','REF','ALT'), by.y=c('V1','V4','V6','V5'))

sum_auto_withRSID<-rbind(sum_auto_matched, sum_auto_swapped)

sum_auto_withRSID<-sum_auto_withRSID[,c('CHR','POS','REF','ALT','Frq','BETA','SE','P','V2'), with=F]

names(sum_auto_withRSID)<-c('CHR','POS','A2','A1','FREQ','BETA','SE','P','SNP')

sum_auto_withRSID$N<-158284

fwrite(sum_auto_withRSID, '/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/BBJ_BMI_EAS', sep=' ', quote=F, na='NA')

system('~/brc_scratch/Software/pigz /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/BBJ_BMI_EAS')

```

</details>

<details><summary>Locke_BMI_EUR</summary>

```{bash, eval=F, echo=T}

cp /scratch/groups/ukbiobank/sumstats/cleaned/BODY04.gz /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/Locke_BMI_EUR.gz

```

</details>

***

### QC GWAS

<details><summary>Show code</summary>

```{bash, eval=F, echo=T}
# QC the sumstats
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Pipeline_prep.config

pop=EUR
for gwas in $(echo PGC2_MDD_EUR Wood_Height_EUR Locke_BMI_EUR);do
/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/sumstat_cleaner/sumstat_cleaner.R \
  --sumstats /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/${gwas}.gz \
  --ref_plink_chr ${Geno_1KG_dir}/1KGPhase3.w_hm3.chr \
  --ref_freq_chr ${Geno_1KG_dir}/freq_files/${pop}/1KGPhase3.w_hm3.${pop}.chr \
  --output /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/${gwas}
done

pop=EAS
for gwas in $(echo PGC_MDD_EAS CONVERGE_MDD_EAS BBJ_MDD_EAS BBJ_Height_EAS BBJ_BMI_EAS);do
/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/sumstat_cleaner/sumstat_cleaner.R \
  --sumstats /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/${gwas}.gz \
  --ref_plink_chr ${Geno_1KG_dir}/1KGPhase3.w_hm3.chr \
  --ref_freq_chr ${Geno_1KG_dir}/freq_files/${pop}/1KGPhase3.w_hm3.${pop}.chr \
  --output /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/${gwas}
done

```
</details>

***

## Cross Population SNP annotations

We focus on two annotations describing differences in linkage disequilibrium (LD) and allele frequency for a given SNP across populations. These include LDcorr, which is the correlation between LD r2 estimates for a given SNP across two populations, and Fst which is a measure of differences in allele frequency for a given SNP across populations.

***

### Calculate annotations

<details><summary>Show code</summary>
```{bash, echo=T, eval=F}
pop1=EUR

for pop2 in $(echo EAS AMR AFR SAS); do
  for chr in $(seq 1 22);do
    sbatch -p brc,shared --mem 30G -n 1 --nodes 1 /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/cross_pop/cross_pop.R \
                  --ref /scratch/groups/biomarkers-brc-mh/Reference_data/1KG_Phase3/bigsnpr/1000G_phase3_common_norel \
                  --chr ${chr} \
                  --pop1 ${pop1} \
                  --pop2 ${pop2} \
                  --n_cores 1 \
                  --output /scratch/groups/biomarkers-brc-mh/Reference_data/CrossPop/Populations_${pop1}_${pop2}/chr${chr}/res
  done
done

# Create genome-wide files
pop1=EUR
for pop2 in $(echo EAS AMR AFR SAS); do
head -1 /scratch/groups/biomarkers-brc-mh/Reference_data/CrossPop/Populations_${pop1}_${pop2}/chr1/res.chr1.${pop1}.${pop2}.txt > /scratch/groups/biomarkers-brc-mh/Reference_data/CrossPop/Populations_${pop1}_${pop2}/CrossPop.${pop1}.${pop2}.GW.txt
tail -n +2 -q /scratch/groups/biomarkers-brc-mh/Reference_data/CrossPop/Populations_${pop1}_${pop2}/chr*/res.chr*.${pop1}.${pop2}.txt >> /scratch/groups/biomarkers-brc-mh/Reference_data/CrossPop/Populations_${pop1}_${pop2}/CrossPop.${pop1}.${pop2}.GW.txt
done
```

</details>

***

### Annotate GWAS

<details><summary>Show code</summary>
```{r, echo=T, eval=F}
library(data.table)

pop1='EUR'
pop2='EAS'

gwas<-c('PGC2_MDD_EUR','Wood_Height_EUR','Locke_BMI_EUR','CONVERGE_MDD_EAS','BBJ_MDD_EAS','BBJ_Height_EAS','BBJ_BMI_EAS')

gwas_eur<-gwas[grepl(paste0('_',pop1,'$'), gwas)]
gwas_noneur<-gwas[grepl(paste0('_',pop2,'$'), gwas)]

dir.create('/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/annotated')

for(pop2_i in pop2){
  # Read in pop1-pop2_i annotations
  annot<-NULL
  for(i in 1:22){
    annot<-rbind(annot, fread(paste0('/scratch/groups/biomarkers-brc-mh/Reference_data/CrossPop/Populations_',pop1,'_',pop2_i,'/chr',i,'/res.chr',i,'.',pop1,'.',pop2_i,'.txt')))
  }

  for(gwas_i in gwas_eur){
    # Read in gwas_i sumstats
    gwas_i_tmp<-fread(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/',gwas_i,'.gz'))
    
    # Insert BETA column if not present
    if(sum('BETA' %in% names(gwas_i_tmp)) == 0){
      gwas_i_tmp$BETA<-log(gwas_i_tmp$OR)
    }

    # Merge annotations with gwas
    gwas_annot<-merge(gwas_i_tmp, annot[,c('SNP','Fst','cor_score'),with=F], by='SNP')
    
    # Scale Fst and cor_score to be between 0 and 1
    gwas_annot$Fst_scaled<-(gwas_annot$Fst-min(gwas_annot$Fst))/(max(gwas_annot$Fst)-min(gwas_annot$Fst))
    gwas_annot$cor_score_scaled<-(gwas_annot$cor_score-min(gwas_annot$cor_score))/(max(gwas_annot$cor_score)-min(gwas_annot$cor_score))
    
    # Flip Fst so high value indicate similar MAF
    gwas_annot$Fst_scaled<-1-gwas_annot$Fst_scaled
    
    # Reweight SE according to LDcorr and Fst
    for(q in c(0.5,1,1.5,2)){
      gwas_annot[[paste0('SE_Fst_',q)]]<-gwas_annot$SE/sqrt(gwas_annot$Fst_scaled^q)
      gwas_annot[[paste0('SE_LDcorr_',q)]]<-gwas_annot$SE/sqrt(gwas_annot$cor_score_scaled^q)
    }
    
    # Reweight SE genome-wide Ji et al. increase weight in target ancestry GWAS (EAS) by 1:5. For simplicity, we will downweight the non-target ancestry GWAS (EUR) by 1:5.
    for(k in 2:5){
      gwas_annot[[paste0('SE_k_',k)]]<-gwas_annot$SE/sqrt(1/k)
    }

    # Save the results
    fwrite(gwas_annot, paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/annotated/',gwas_i), quote=F, sep=' ', na='NA')
    
    # Compress the sumstats
    if(file.exists(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/annotated/',gwas_i,'.gz'))){
      system(paste0('rm /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/annotated/',gwas_i,'.gz'))
    }
    system(paste0('~/brc_scratch/Software/pigz /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/annotated/',gwas_i))

  }
}

# For non-EUR sumstats, just make sure a BETA column is present
for(gwas_i in gwas_noneur){
  # Read in gwas_i sumstats
  gwas_i_tmp<-fread(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/',gwas_i,'.gz'))
  
  # Insert BETA column if not present
  if(sum('BETA' %in% names(gwas_i_tmp)) == 0){
    print(gwas_i)
    
    gwas_i_tmp$BETA<-log(gwas_i_tmp$OR)
  
    # Save the results
    fwrite(gwas_i_tmp, paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/',gwas_i), quote=F, sep=' ', na='NA')
    
    # Compress the sumstats
    if(file.exists(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/',gwas_i,'.gz'))){
      system(paste0('rm /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/',gwas_i,'.gz'))
    }
    system(paste0('~/brc_scratch/Software/pigz /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/',gwas_i))
  }
}

```

</details>

***

## Characterise associations

<details><summary>Show code</summary>
```{R, eval=F, echo=T}
library(data.table)
library(ggplot2)
library(cowplot)

pheno<-c('MDD','MDD','MDD','Height','BMI')
samp1=c('PGC2','PGC2','PGC2','Wood','Locke')
samp2=c('PGC','CONVERGE','BBJ','BBJ','BBJ')
pop1=c('EUR','EUR','EUR','EUR','EUR')
pop2=c('EAS','EAS','EAS','EAS','EAS')

gwas_eur<-unique(paste0(samp1,'_',pheno,'_',pop1))
gwas_eas<-unique(paste0(samp2,'_',pheno,'_',pop2))

for(i in 1:length(gwas_eur)){
  gwas_1<-gwas_eur[i]
  gwas_1_dat<-fread(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/annotated/',gwas_1,'.gz'))

  for(annot in c('LDcorr','Fst')){
    if(annot == 'LDcorr'){
      var<-'cor_score'
    } else {
      var<-'Fst'
    }
    
    # Measure 5 and 95% quantiles
    quant<-NULL
    for(q in c(0.01,0.05,0.1,0.5,0.9,0.95,0.99)){
      quant<-rbind(quant, data.frame(q=q,
                                     val=quantile(gwas_1_dat[[var]], probs = q)))

      if(q < 0.5){
        dir.create(paste0('/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/', gwas_1,'/',annot,'/bot_',q), recursive=T)
        write.table(gwas_1_dat[gwas_1_dat[[var]] < quant$val[quant$q == q],]$SNP, paste0("/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/", gwas_1,'/',annot,'/bot_',q,'/bot_',q,'_',annot,'.snps.txt'), col.names=F, row.names=F, quote=F)
      } 
      
      if(q > 0.5){
        dir.create(paste0('/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/', gwas_1,'/',annot,'/top_',1-q), recursive=T)
        write.table(gwas_1_dat[gwas_1_dat[[var]] > quant$val[quant$q == q],]$SNP, paste0("/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/", gwas_1,'/',annot,'/top_',1-q,'/top_',1-q,'_',annot,'.snps.txt'), col.names=F, row.names=F, quote=F)
      }
      
      if(q == 0.5){
        dir.create(paste0('/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/', gwas_1,'/',annot,'/top_',q), recursive=T)
        write.table(gwas_1_dat[gwas_1_dat[[var]] > quant$val[quant$q == q],]$SNP, paste0("/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/", gwas_1,'/',annot,'/top_',q,'/top_',q,'_',annot,'.snps.txt'), col.names=F, row.names=F, quote=F)

        dir.create(paste0('/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/', gwas_1,'/',annot,'/bot_',q), recursive=T)
        write.table(gwas_1_dat[gwas_1_dat[[var]] < quant$val[quant$q == q],]$SNP, paste0("/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/", gwas_1,'/',annot,'/bot_',q,'/bot_',q,'_',annot,'.snps.txt'), col.names=F, row.names=F, quote=F)
      }


    dir.create(paste0('/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/', gwas_1,'/',annot,'/all'), recursive=T)

    # Perform LD clumping for these subsets of SNPs
    for(chr in 1:22){
      system(paste0('/users/k1806347/brc_scratch/Software/plink1.9.sh --bfile /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr',chr,' --keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/EUR_samples.keep --extract /scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/', gwas_1,'/',annot,'/top_0.05/top_0.05_',annot,'.snps.txt --clump /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/',gwas_1,'.gz --clump-p1 1 --clump-p2 1 --clump-r2 0.1 --clump-kb 250 --out /scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/', gwas_1,'/',annot,'/top_0.05/top_0.05_',annot,'.snps.chr',chr))

      system(paste0('/users/k1806347/brc_scratch/Software/plink1.9.sh --bfile /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr',chr,' --keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/EUR_samples.keep --clump /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/',gwas_1,'.gz --clump-p1 1 --clump-p2 1 --clump-r2 0.1 --clump-kb 250 --out /scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/', gwas_1,'/',annot,'/all/all.snps.chr',chr))
      
      system(paste0('/users/k1806347/brc_scratch/Software/plink1.9.sh --bfile /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr',chr,' --keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/EUR_samples.keep --extract /scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/', gwas_1,'/',annot,'/bot_0.05/bot_0.05_',annot,'.snps.txt --clump /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/',gwas_1,'.gz --clump-p1 1 --clump-p2 1 --clump-r2 0.1 --clump-kb 250 --out /scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/', gwas_1,'/',annot,'/bot_0.05/bot_0.05_',annot,'.snps.chr',chr))
    }
    
    clump_top5<-NULL
    for(chr in 1:22){
      if(file.exists(paste0('/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/', gwas_1,'/',annot,'/top_0.05/top_0.05_',annot,'.snps.chr',chr,'.clumped'))){
        clumped<-fread(paste0('/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/', gwas_1,'/',annot,'/top_0.05/top_0.05_',annot,'.snps.chr',chr,'.clumped'))
        
        clumped_SNPs<-clumped$SNP
    	  clump_top5<-c(clump_top5, clumped_SNPs)
        
    	  print(chr)
      }
    }
    
    clump_all<-NULL
    for(chr in 1:22){
      if(file.exists(paste0('/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/', gwas_1,'/',annot,'/all/all.snps.chr',chr,'.clumped'))){
        clumped<-fread(paste0('/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/', gwas_1,'/',annot,'/all/all.snps.chr',chr,'.clumped'))
        
        clumped_SNPs<-clumped$SNP
    	  clump_all<-c(clump_all, clumped_SNPs)
        
    	  print(chr)
      }
    }

    clump_bot5<-NULL
    for(chr in 1:22){
    	if(file.exists(paste0('/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/', gwas_1,'/',annot,'/bot_0.05/bot_0.05_',annot,'.snps.chr',chr,'.clumped'))){
clumped<-fread(paste0('/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/', gwas_1,'/',annot,'/bot_0.05/bot_0.05_',annot,'.snps.chr',chr,'.clumped'))
      
    	clumped_SNPs<-clumped$SNP
    	clump_bot5<-c(clump_bot5, clumped_SNPs)
    	
    	print(chr)
    	}
    }
    
    gwas_1_dat$top5_indep<-F
    gwas_1_dat$top5_indep[(gwas_1_dat$SNP %in% clump_top5)]<-T
    gwas_1_dat$all_indep<-F
    gwas_1_dat$all_indep[(gwas_1_dat$SNP %in% clump_all)]<-T
    gwas_1_dat$bot5_indep<-F
    gwas_1_dat$bot5_indep[(gwas_1_dat$SNP %in% clump_bot5)]<-T
    
    # Now estimate pearson correlation of BETAs across populations
    gwas_eas_for_gwas_1<-gwas_eas[grepl(gsub('.*_','',gsub('_EUR','',gwas_1)),gwas_eas)]
    
    for(j in 1:length(gwas_eas_for_gwas_1)){
      gwas_2<-gwas_eas_for_gwas_1[j]
      gwas_2_dat<-fread(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/',gwas_2,'.gz'))

      if(sum(names(gwas_2_dat) == 'BETA') != 1){
        gwas_2_dat$BETA<-log(gwas_2_dat$OR)
      }
      
      both_match<-merge(gwas_1_dat, gwas_2_dat, by=c('SNP','A1','A2'))
      
      thresh_vec<-c(1,0.5,0.1,0.01,0.001,1e-5,5e-8)

      cor_res_all<-NULL
      
      for(pT in thresh_vec){
        both_match_pT<-both_match[both_match$P.x < pT,]
        if(dim(both_match_pT[both_match_pT$bot5_indep == T])[1] > 5 & dim(both_match_pT[both_match_pT$top5_indep == T])[1] > 5){
          
          all_res<-summary(lm(scale(BETA.x) ~ scale(BETA.y), data=both_match_pT[both_match_pT$all_indep == T,]))
          cor_res_all<-rbind(cor_res_all,data.frame(pT=pT,
                                                    type='all',
                                                    cor=coef(all_res)[2,1],
                                                    se=coef(all_res)[2,2]))
  
          bot5_res<-summary(lm(scale(BETA.x) ~ scale(BETA.y), data=both_match_pT[both_match_pT$bot5_indep == T,]))
          cor_res_all<-rbind(cor_res_all,data.frame(pT=pT,
                                                    type='bot5',
                                                    cor=coef(bot5_res)[2,1],
                                                    se=coef(bot5_res)[2,2]))
  
          top5_res<-summary(lm(scale(BETA.x) ~ scale(BETA.y), data=both_match_pT[both_match_pT$top5_indep == T,]))
          cor_res_all<-rbind(cor_res_all,data.frame(pT=pT,
                                                    type='top5',
                                                    cor=coef(top5_res)[2,1],
                                                    se=coef(top5_res)[2,2]))
        }
      }
     
      dir.create(paste0('/scratch/users/k1806347/Analyses/CrossMeta/sign_test/res/',gwas_1,'_vs_',gwas_2,'/',annot), recursive=T)
      write.table(cor_res_all, paste0('/scratch/users/k1806347/Analyses/CrossMeta/sign_test/res/',gwas_1,'_vs_',gwas_2,'/',annot,'/cor_res.txt'), col.names=T, row.names=F, quote=F)
    }
  }
}

cor_res_all_all<-NULL
for(i in 1:length(gwas_eur)){
  gwas_1<-gwas_eur[i]
  for(annot in c('LDcorr','Fst')){
    gwas_eas_for_gwas_1<-gwas_eas[grepl(gsub('.*_','',gsub('_EUR','',gwas_1)),gwas_eas)]

    for(j in 1:length(gwas_eas_for_gwas_1)){
      gwas_2<-gwas_eas_for_gwas_1[j]
      cor_res_all<-fread(paste0('/scratch/users/k1806347/Analyses/CrossMeta/sign_test/res/',gwas_1,'_vs_',gwas_2,'/',annot,'/cor_res.txt'))
      cor_res_all$annot<-annot
      cor_res_all$test<-paste0(gwas_1,'_vs_',gwas_2)
      
      cor_res_all_all<-rbind(cor_res_all_all, cor_res_all)
    }
    
    cor_res_all_all$type<-factor(cor_res_all_all$type, levels=c('bot5','all','top5'))
    cor_res_all_all$pT<-factor(cor_res_all_all$pT, levels=unique(cor_res_all_all$pT))
  }
}

cor_res_all_all$test<-gsub('vs','vs\n',gsub('_',' ',cor_res_all_all$test))
cor_res_all_all$test<-factor(cor_res_all_all$test, levels=unique(cor_res_all_all$test))

png(paste0('/scratch/users/k1806347/Analyses/CrossMeta/sign_test/res/cor_plot_Height_BMI.png'), units='px', res=300, width=2000, height=1500)
  print(ggplot(cor_res_all_all[grepl('BMI|Height',cor_res_all_all$test),], aes(x=pT, y=cor, fill=type)) +
    geom_bar(stat="identity", position=position_dodge(.9)) +
    geom_errorbar(aes(ymin=cor-se, ymax=cor+se), width=.2,position=position_dodge(.9)) +
    theme_half_open() +
    background_grid() +
    labs(y="BETA correlation (SE)", title=paste0(gwas_1,'_vs_',gwas_2,': ',annot)) +
    theme(axis.text.x = element_text(angle = 45, hjust=1)) +
    facet_grid(vars(test), vars(annot), scales = "free"))
dev.off()

png(paste0('/scratch/users/k1806347/Analyses/CrossMeta/sign_test/res/cor_plot_MDD.png'), units='px', res=300, width=2000, height=2100)
  print(ggplot(cor_res_all_all[grepl('MDD',cor_res_all_all$test),], aes(x=pT, y=cor, fill=type)) +
    geom_bar(stat="identity", position=position_dodge(.9)) +
    geom_errorbar(aes(ymin=cor-se, ymax=cor+se), width=.2,position=position_dodge(.9)) +
    theme_half_open() +
    background_grid() +
    labs(y="BETA correlation (SE)", title=paste0(gwas_1,'_vs_',gwas_2,': ',annot)) +
    theme(axis.text.x = element_text(angle = 45, hjust=1)) +
    facet_grid(vars(test), vars(annot), scales = "free"))
dev.off()

```
</details>


***

## Meta-GWAS

We want to meta-analyse the EUR GWAS sumstats with each of the EAS sumstats

***

### Standard inverse variance meta-GWAS

This is performing standard inverse-variance meta-GWAS.

<details><summary>Show code</summary>

```{bash, eval=F, echo=T}

# Create METAL scripts
mkdir -p /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/scripts
mkdir -p /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results

pheno=(MDD MDD MDD Height BMI)
samp1=(PGC2 PGC2 PGC2 Wood Locke)
samp2=(PGC CONVERGE BBJ BBJ BBJ)
pop1=(EUR EUR EUR EUR EUR)
pop2=(EAS EAS EAS EAS EAS)

# use for loop to read all values and indexes
for (( i=0; i<${#pheno[@]}; i++ ));do

cat <<EOF > /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/scripts/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_metal_standard.sh

#!/bin/sh
SCHEME STDERR
CUSTOMVARIABLE TotalSampleSize

MARKER SNP
ALLELE A1 A2
EFFECT BETA
PVALUE P 
LABEL TotalSampleSize as N
STDERR SE 

PROCESS /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/annotated/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}.gz

MARKER SNP
ALLELE A1 A2
EFFECT BETA
PVALUE P 
LABEL TotalSampleSize as N
STDERR SE 

PROCESS /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/${samp2[${i}]}_${pheno[${i}]}_${pop2[${i}]}.gz

OUTFILE /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_metal_standard .tbl
ANALYZE 
QUIT

EOF

done

# Run scripts
for (( i=0; i<${#pheno[@]}; i++ ));do
sbatch -p brc,shared /users/k1806347/brc_scratch/Software/metal.sh /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/scripts/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_metal_standard.sh
done

```

</details>

***

### Genome-wide reweighted meta-GWAS

This involves modifying the SEs of all SNPs from a given GWAS prior to performing standard inverse-variance meta-GWAS.

<details><summary>Show code</summary>

```{bash, eval=F, echo=T}

# Create METAL scripts
pheno=(MDD MDD MDD Height BMI)
samp1=(PGC2 PGC2 PGC2 Wood Locke)
samp2=(PGC CONVERGE BBJ BBJ BBJ)
pop1=(EUR EUR EUR EUR EUR)
pop2=(EAS EAS EAS EAS EAS)

integers=(2 3 4 5)

# use for loop to read all values and indexes
for (( i=0; i<${#pheno[@]}; i++ ));do
for k in ${integers[@]};do

cat <<EOF > /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/scripts/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_metal_integer_${k}.sh

#!/bin/sh
SCHEME STDERR
CUSTOMVARIABLE TotalSampleSize

MARKER SNP
ALLELE A1 A2
EFFECT BETA
PVALUE P 
LABEL TotalSampleSize as N
STDERR SE_k_${k} 

PROCESS /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/annotated/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}.gz

MARKER SNP
ALLELE A1 A2
EFFECT BETA
PVALUE P 
LABEL TotalSampleSize as N
STDERR SE 

PROCESS /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/${samp2[${i}]}_${pheno[${i}]}_${pop2[${i}]}.gz

OUTFILE /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_metal_integer_${k} .tbl
ANALYZE 
QUIT

EOF

done
done

# Run scripts
for (( i=0; i<${#pheno[@]}; i++ ));do
for k in ${integers[@]};do
sbatch -p brc,shared /users/k1806347/brc_scratch/Software/metal.sh /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/scripts/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_metal_integer_${k}.sh
done
done

```

</details>

***

### SNP-level reweighted meta-GWAS

This involves modifying the SE of SNPs according to LDcorr and Fst from a given GWAS prior to performing standard inverse-variance meta-GWAS.

<details><summary>Show code</summary>

```{bash, eval=F, echo=T}

# Create METAL scripts
pheno=(MDD MDD MDD Height BMI)
samp1=(PGC2 PGC2 PGC2 Wood Locke)
samp2=(PGC CONVERGE BBJ BBJ BBJ)
pop1=(EUR EUR EUR EUR EUR)
pop2=(EAS EAS EAS EAS EAS)

annot=(LDcorr Fst)
weight=(0.5 1 1.5 2)

# use for loop to read all values and indexes
for (( i=0; i<${#pheno[@]}; i++ ));do
for j in ${annot[@]};do
for q in ${weight[@]};do

cat <<EOF > /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/scripts/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_metal_${j}_${q}.sh

#!/bin/sh
SCHEME STDERR
CUSTOMVARIABLE TotalSampleSize

MARKER SNP
ALLELE A1 A2
EFFECT BETA
PVALUE P 
LABEL TotalSampleSize as N
STDERR SE_${j}_${q} 

PROCESS /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/annotated/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}.gz

MARKER SNP
ALLELE A1 A2
EFFECT BETA
PVALUE P 
LABEL TotalSampleSize as N
STDERR SE 

PROCESS /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/${samp2[${i}]}_${pheno[${i}]}_${pop2[${i}]}.gz

OUTFILE /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_metal_${j}_${q} .tbl
ANALYZE 
QUIT

EOF

done
done
done

# Run scripts
for (( i=0; i<${#pheno[@]}; i++ ));do
for j in ${annot[@]};do
for q in ${weight[@]};do
sbatch -p brc,shared /users/k1806347/brc_scratch/Software/metal.sh /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/scripts/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_metal_${j}_${q}.sh
done
done
done

```

</details>

***

### Reformat METAL output

```{r, eval=F, echo=T}
library(data.table)

pheno<-c('MDD','MDD','MDD','Height','BMI')
samp1=c('PGC2','PGC2','PGC2','Wood','Locke')
samp2=c('PGC','CONVERGE','BBJ','BBJ','BBJ')
pop1=c('EUR','EUR','EUR','EUR','EUR')
pop2=c('EAS','EAS','EAS','EAS','EAS')

integers=c(2,3,4,5)

annot=c('LDcorr','Fst')
weight=c(0.5,1,1.5,2)

type<-c('standard','integer','annot')

for(type_i in type){
  if(type_i == 'standard'){
    for(i in 1:length(pheno)){
      dat<-fread(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_standard1.tbl'))
      
      names(dat)<-c('SNP','A1','A2','BETA','SE','P','Direction','N')
      dat$A1<-toupper(dat$A1)
      dat$A2<-toupper(dat$A2)
      
      # Remove SNPs which were not present in one of the samples
      dat<-dat[!grepl('\\?',dat$Direction),]
      
      fwrite(dat, paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_standard1.reformat.txt'), sep=' ', na='NA', quote=F)
      
      if(file.exists(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_standard1.reformat.txt.gz'))){
        system(paste0('rm /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_standard1.reformat.txt.gz'))
      }
      system(paste0('~/brc_scratch/Software/pigz /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_standard1.reformat.txt'))
      
      system(paste0('/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/sumstat_cleaner/sumstat_cleaner.R --sumstats /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_standard1.reformat.txt.gz  --ref_plink_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr --out /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_standard1.reformat.txt.cleaned'))
      
    }
  }
    
  if(type_i == 'integer'){
    for(i in 1:length(pheno)){
      for(integers_i in integers){
        dat<-fread(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_integer_',integers_i,'1.tbl'))
        
        names(dat)<-c('SNP','A1','A2','BETA','SE','P','Direction','N')
        dat$A1<-toupper(dat$A1)
        dat$A2<-toupper(dat$A2)
        
        # Remove SNPs which were not present in one of the samples
        dat<-dat[!grepl('\\?',dat$Direction),]
        
        fwrite(dat, paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_integer_',integers_i,'1.reformat.txt'), sep=' ', na='NA', quote=F)
        
        if(file.exists(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_integer_',integers_i,'1.reformat.txt.gz'))){
          system(paste0('rm /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_integer_',integers_i,'1.reformat.txt.gz'))
        }

        system(paste0('~/brc_scratch/Software/pigz /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_integer_',integers_i,'1.reformat.txt'))
        
        system(paste0('/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/sumstat_cleaner/sumstat_cleaner.R --sumstats /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_integer_',integers_i,'1.reformat.txt.gz --ref_plink_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr --out /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_integer_',integers_i,'1.reformat.txt.cleaned'))

      }
    }
  }
    
  if(type_i == 'annot'){
    for(i in 1:length(pheno)){
      for(annot_i in annot){
        for(weight_i in weight){
        dat<-fread(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_',annot_i,'_',weight_i,'1.tbl'))
        
        names(dat)<-c('SNP','A1','A2','BETA','SE','P','Direction','N')
        dat$A1<-toupper(dat$A1)
        dat$A2<-toupper(dat$A2)
        
        # Remove SNPs which were not present in one of the samples
        dat<-dat[!grepl('\\?',dat$Direction),]
        
        fwrite(dat, paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_',annot_i,'_',weight_i,'1.reformat.txt'), sep=' ', na='NA', quote=F)
        
        if(file.exists(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_',annot_i,'_',weight_i,'1.reformat.txt.gz'))){
          system(paste0('rm /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_',annot_i,'_',weight_i,'1.reformat.txt.gz'))
        }

        system(paste0('~/brc_scratch/Software/pigz /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_',annot_i,'_',weight_i,'1.reformat.txt'))
        
        system(paste0('/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/sumstat_cleaner/sumstat_cleaner.R --sumstats /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_',annot_i,'_',weight_i,'1.reformat.txt.gz --ref_plink_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr --out /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_',annot_i,'_',weight_i,'1.reformat.txt.cleaned'))

        }
      }
    }
  }
}

```

***

## Polygenic scoring

***

### Generating SNP-weights

***

#### Meta-GWAS

```{bash eval=F, echo=T}
# Create a reference keep file to specify EUR and EAS individuals
cat /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/EUR_samples.keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/EAS_samples.keep > /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/EUR_EAS_samples.keep

```

```{R, eval=F, echo=T}

pheno<-c('MDD','MDD','MDD','Height','BMI')
samp1=c('PGC2','PGC2','PGC2','Wood','Locke')
samp2=c('PGC','CONVERGE','BBJ','BBJ','BBJ')
pop1=c('EUR','EUR','EUR','EUR','EUR')
pop2=c('EAS','EAS','EAS','EAS','EAS')

integers=c(2,3,4,5)

annot=c('LDcorr','Fst')
weight=c(0.5,1,1.5,2)

type<-c('standard','integer','annot')

dir.create('/users/k1806347/brc_scratch/Analyses/CrossMeta/score_files')

for(type_i in type){
  if(type_i == 'standard'){
    for(i in 1:length(pheno)){
      
      system(paste0('sbatch -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator/polygenic_score_file_creator.R --ref_plink_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr --ref_keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/EUR_EAS_samples.keep --sumstats /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_standard1.reformat.txt.cleaned.gz --plink /users/k1806347/brc_scratch/Software/plink1.9.sh --memory 3000 --output /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/',type_i,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/score --ref_pop_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/super_pop_keep.list'))
  
    }
  }
    
  if(type_i == 'integer'){
    for(i in 1:length(pheno)){
      for(integers_i in integers){
        
              system(paste0('sbatch -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator/polygenic_score_file_creator.R --ref_plink_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr --ref_keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/EUR_EAS_samples.keep --sumstats /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_integer_',integers_i,'1.reformat.txt.cleaned.gz --plink /users/k1806347/brc_scratch/Software/plink1.9.sh --memory 3000 --output /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/',type_i,'/','k_',integers_i,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/score --ref_pop_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/super_pop_keep.list'))

      }
    }
  }
    
  if(type_i == 'annot'){
    for(i in 1:length(pheno)){
      for(annot_i in annot){
        for(weight_i in weight){
          
        system(paste0('sbatch -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator/polygenic_score_file_creator.R --ref_plink_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr --ref_keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/EUR_EAS_samples.keep --sumstats /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'_metal_',annot_i,'_',weight_i,'1.reformat.txt.cleaned.gz --plink /users/k1806347/brc_scratch/Software/plink1.9.sh --memory 3000 --output /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/',type_i,'/',annot_i,'/','q_',weight_i,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/score --ref_pop_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/super_pop_keep.list'))

        }
      }
    }
  }
}

```

***

#### Single sample

```{bash, eval=F, echo=T}
########
# All SNPs
########

pop=EUR
for gwas in $(echo PGC2_MDD_EUR Wood_Height_EUR Locke_BMI_EUR);do
  sbatch -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator/polygenic_score_file_creator.R \
  --ref_plink_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr \
  --ref_keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/${pop}_samples.keep \
  --sumstats /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/${gwas}.gz \
  --plink /users/k1806347/brc_scratch/Software/plink1.9.sh \
  --memory 3000 \
  --output /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/${pop}_only/${gwas}/score \
  --ref_pop_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/super_pop_keep.list
done

pop=EAS
for gwas in $(echo PGC_MDD_EAS CONVERGE_MDD_EAS BBJ_MDD_EAS BBJ_Height_EAS BBJ_BMI_EAS);do
  sbatch -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator/polygenic_score_file_creator.R \
  --ref_plink_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr \
  --ref_keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/${pop}_samples.keep \
  --sumstats /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/${gwas}.gz \
  --plink /users/k1806347/brc_scratch/Software/plink1.9.sh \
  --memory 3000 \
  --output /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/${pop}_only/${gwas}/score \
  --ref_pop_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/super_pop_keep.list
done

########
# Using subsets of LDcor and Fst
########

pop=EUR
for gwas in $(echo PGC2_MDD_EUR Wood_Height_EUR Locke_BMI_EUR);do
for annot in $(echo LDcorr Fst);do
for q in $(echo top_0.01 top_0.05 top_0.1 top_0.5 bot_0.01 bot_0.05 bot_0.1 bot_0.5);do
  sbatch -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator/polygenic_score_file_creator.R \
  --ref_plink_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr \
  --ref_keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/${pop}_samples.keep \
  --sumstats /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/${gwas}.gz \
  --extract /scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/${gwas}/${annot}/${q}/${q}_${annot}.snps.txt \
  --plink /users/k1806347/brc_scratch/Software/plink1.9.sh \
  --memory 3000 \
  --output /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/${pop}_only_subset/${gwas}/${annot}/${q}/score \
  --ref_pop_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/super_pop_keep.list
done
done
done

```

***

#### PRS-CSx

Documentation for generating PRS-CSx score files is [here](https://opain.github.io/GenoPred/Pipeline_prep.html)

***

### Calculating polygenic scores in target

***

#### Meta-GWAS

```{R, eval=F, echo=T}

pheno<-c('MDD','MDD','MDD','Height','BMI')
samp1=c('PGC2','PGC2','PGC2','Wood','Locke')
samp2=c('PGC','CONVERGE','BBJ','BBJ','BBJ')
pop1=c('EUR','EUR','EUR','EUR','EUR')
pop2=c('EAS','EAS','EAS','EAS','EAS')

integers=c(2,3,4,5)

annot=c('LDcorr','Fst')
weight=c(0.5,1,1.5,2)

type<-c('standard','integer','annot')

dir.create('/users/k1806347/brc_scratch/Analyses/CrossMeta/prs')

for(type_i in type){
  if(type_i == 'standard'){
    for(i in 1:length(pheno)){
      for(targ_pop in c(pop1[i],pop2[i])){
      
        system(paste0('sbatch -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer/Scaled_polygenic_scorer.R --target_plink_chr /users/k1806347/brc_scratch/Data/UKBB/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr --target_keep /users/k1806347/brc_scratch/Misc/Toshiki/UKB_keep_files/UKB_QC_UpdateIDs_',targ_pop,'.keep --ref_score /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/',type_i,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/score --ref_scale /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/',type_i,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/score.',targ_pop,'.scale --ref_freq_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/freq_files/',targ_pop,'/1KGPhase3.w_hm3.',targ_pop,'.chr --plink /users/k1806347/brc_scratch/Software/plink1.9.sh --output /users/k1806347/brc_scratch/Analyses/CrossMeta/prs/',targ_pop,'_target/',type_i,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/prs'))

      }
    }
  }
  
  if(type_i == 'integer'){
    for(i in 1:length(pheno)){
      for(integers_i in integers){
        for(targ_pop in c(pop1[i],pop2[i])){
          system(paste0('sbatch -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer/Scaled_polygenic_scorer.R --target_plink_chr /users/k1806347/brc_scratch/Data/UKBB/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr --target_keep /users/k1806347/brc_scratch/Misc/Toshiki/UKB_keep_files/UKB_QC_UpdateIDs_',targ_pop,'.keep --ref_score /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/',type_i,'/','k_',integers_i,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/score --ref_scale /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/',type_i,'/','k_',integers_i,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/score.',targ_pop,'.scale --ref_freq_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/freq_files/',targ_pop,'/1KGPhase3.w_hm3.',targ_pop,'.chr --plink /users/k1806347/brc_scratch/Software/plink1.9.sh --output /users/k1806347/brc_scratch/Analyses/CrossMeta/prs/',targ_pop,'_target/',type_i,'/','k_',integers_i,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/prs'))

        }
      }
    }
  }
    
  if(type_i == 'annot'){
    for(i in 1:length(pheno)){
      for(annot_i in annot){
        for(weight_i in weight){
          for(targ_pop in c(pop1[i],pop2[i])){
            system(paste0('sbatch -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer/Scaled_polygenic_scorer.R --target_plink_chr /users/k1806347/brc_scratch/Data/UKBB/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr --target_keep /users/k1806347/brc_scratch/Misc/Toshiki/UKB_keep_files/UKB_QC_UpdateIDs_',targ_pop,'.keep --ref_score /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/',type_i,'/',annot_i,'/','q_',weight_i,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/score --ref_scale /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/',type_i,'/',annot_i,'/','q_',weight_i,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/score.',targ_pop,'.scale --ref_freq_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/freq_files/',targ_pop,'/1KGPhase3.w_hm3.',targ_pop,'.chr --plink /users/k1806347/brc_scratch/Software/plink1.9.sh --output /users/k1806347/brc_scratch/Analyses/CrossMeta/prs/',targ_pop,'_target/',type_i,'/',annot_i,'/','q_',weight_i,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/prs'))

          }
        }
      }
    }
  }
}

```

***

#### Single sample

```{r, eval=F, echo=T}

########
# All SNPs
########

for(targ_pop in c('EUR','EAS')){
  for(gwas in c('PGC2_MDD_EUR','Wood_Height_EUR','Locke_BMI_EUR','PGC_MDD_EAS','CONVERGE_MDD_EAS','BBJ_MDD_EAS','BBJ_Height_EAS','BBJ_BMI_EAS')){
    pop<-gsub('.*_','',gwas)
      
    system(paste0('sbatch -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer/Scaled_polygenic_scorer.R --target_plink_chr /users/k1806347/brc_scratch/Data/UKBB/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr --target_keep /users/k1806347/brc_scratch/Misc/Toshiki/UKB_keep_files/UKB_QC_UpdateIDs_',targ_pop,'.keep --ref_score /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/',pop,'_only/',gwas,'/score --ref_scale /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/',pop,'_only/',gwas,'/score.',targ_pop,'.scale --ref_freq_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/freq_files/',targ_pop,'/1KGPhase3.w_hm3.',targ_pop,'.chr --plink /users/k1806347/brc_scratch/Software/plink1.9.sh --output /users/k1806347/brc_scratch/Analyses/CrossMeta/prs/',targ_pop,'_target/',pop,'_only/',gwas,'/prs'))

  }
}

########
# Using subsets of LDcor and Fst
########

for(targ_pop in c('EAS')){
  for(gwas in c('PGC2_MDD_EUR','Wood_Height_EUR','Locke_BMI_EUR')){
    for(annot in c('LDcorr','Fst')){
      for(q in c('top_0.01', 'top_0.05', 'top_0.1', 'top_0.5', 'bot_0.01', 'bot_0.05', 'bot_0.1', 'bot_0.5')){

        pop<-gsub('.*_','',gwas)
          
        system(paste0('sbatch -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer/Scaled_polygenic_scorer.R --target_plink_chr /users/k1806347/brc_scratch/Data/UKBB/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr --target_keep /users/k1806347/brc_scratch/Misc/Toshiki/UKB_keep_files/UKB_QC_UpdateIDs_',targ_pop,'.keep --ref_score /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/',pop,'_only_subset/',gwas,'/',annot,'/',q,'/score --ref_scale /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/',pop,'_only_subset/',gwas,'/',annot,'/',q,'/score.',targ_pop,'.scale --ref_freq_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/freq_files/',targ_pop,'/1KGPhase3.w_hm3.',targ_pop,'.chr --plink /users/k1806347/brc_scratch/Software/plink1.9.sh --output /users/k1806347/brc_scratch/Analyses/CrossMeta/prs/',targ_pop,'_target/',pop,'_only_subset/',gwas,'/',annot,'/',q,'/prs'))
      }
    }
  }
}

```

***

#### PRS-CSx

```{r, eval=F, echo=T}
# PLINK v2.00a2LM 64-bit Intel (9 Mar 2019) was getting a file read error when submitting as a batch job but not interactively
# Here I am using PLINK v2.00a2.3LM AVX2 Intel (24 Jan 2020) which works fine.

pheno<-c('MDD','MDD','MDD','Height','BMI')
samp1=c('PGC2','PGC2','PGC2','Wood','Locke')
samp2=c('PGC','CONVERGE','BBJ','BBJ','BBJ')
pop1=c('EUR','EUR','EUR','EUR','EUR')
pop2=c('EAS','EAS','EAS','EAS','EAS')

for(i in 1:length(pheno)){
for(targ_pop in c(pop1[i], pop2[i])){
for(type in c(pop1[i], pop2[i], 'META')){

    system(paste0('sbatch -p brc,shared --mem 20G /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer/Scaled_polygenic_scorer_plink2.R --target_plink_chr /users/k1806347/brc_scratch/Data/UKBB/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr --target_keep /users/k1806347/brc_scratch/Misc/Toshiki/UKB_keep_files/UKB_QC_UpdateIDs_',targ_pop,'.keep --ref_score /users/k1806347/brc_scratch/Data/1KG/Phase3/Score_files_for_polygenic/PRS-CSx/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/1KGPhase3.w_hm3.',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'.',type,'.score.gz --ref_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/Score_files_for_polygenic/PRS-CSx/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/1KGPhase3.w_hm3.',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'.',type,'.',targ_pop,'.scale --ref_freq_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/freq_files/',targ_pop,'/1KGPhase3.w_hm3.',targ_pop,'.chr --plink2 /users/k1806347/brc_scratch/Software/plink2_alpha/plink2 --output /users/k1806347/brc_scratch/Analyses/CrossMeta/prs/',targ_pop,'_target/PRS-CSx/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'.',type,'/prs'))

}
}
}

```

***

## Evaluating polygenic scores

### Model Comparison

Here we will model the PRS from individual GWAS and meta-GWAS to evaluate and compare their predictive utility.

<details><summary>Show code</summary>

```{bash, echo=TRUE, eval=FALSE}
#########
# Estimate variance explained by EUR PRS subset by annotations
#########

pheno=(MDD MDD MDD Height BMI)
samp1=(PGC2 PGC2 PGC2 Wood Locke)
samp2=(PGC CONVERGE BBJ BBJ BBJ)
pop1=(EUR EUR EUR EUR EUR)
pop2=(EAS EAS EAS EAS EAS)

# use for loop to read all values and indexes
for (( i=0; i<${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop2[${i}]}); do
for annot in $(echo LDcorr Fst); do
mkdir -p /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}

cat <<EOF > /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}/prs_1_${annot}.txt
predictors group
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/prs.profiles ${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}_all
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only_subset/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/${annot}/top_0.01/prs.profiles ${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}_${annot}_top_0.01
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only_subset/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/${annot}/top_0.05/prs.profiles ${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}_${annot}_top_0.05
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only_subset/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/${annot}/top_0.1/prs.profiles ${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}_${annot}_top_0.1
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only_subset/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/${annot}/top_0.5/prs.profiles ${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}_${annot}_top_0.5
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only_subset/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/${annot}/bot_0.01/prs.profiles ${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}_${annot}_bot_0.01
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only_subset/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/${annot}/bot_0.05/prs.profiles ${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}_${annot}_bot_0.05
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only_subset/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/${annot}/bot_0.1/prs.profiles ${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}_${annot}_bot_0.1
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only_subset/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/${annot}/bot_0.5/prs.profiles ${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}_${annot}_bot_0.5

EOF

done
done
done

for (( i=0; i<${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop2[${i}]}); do
for annot in $(echo LDcorr Fst); do

# Test association between prs and phenotype
sbatch --mem 10G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R \
--pheno /users/k1806347/brc_scratch/Analyses/CrossMeta/Phenotype/${pheno[${i}]}.txt \
--out /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}/prs_1_${annot}.test \
--assoc T \
--predictors /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}/prs_1_${annot}.txt

done
done
done

#########
# Estimate predictive utility of single GWAS, Standard Meta, Integer Meta, and Annotation-weight Meta
# Per pT results can be seen in the assoc.txt file
#########

pheno=(MDD MDD MDD Height BMI)
samp1=(PGC2 PGC2 PGC2 Wood Locke)
samp2=(PGC CONVERGE BBJ BBJ BBJ)
pop1=(EUR EUR EUR EUR EUR)
pop2=(EAS EAS EAS EAS EAS)

# use for loop to read all values and indexes
for (( i=0; i<${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

mkdir -p /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}

cat <<EOF > /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_1.txt
predictors group
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/prs.profiles ${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}_single
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop2[${i}]}_only/${samp2[${i}]}_${pheno[${i}]}_${pop2[${i}]}/prs.profiles ${samp2[${i}]}_${pheno[${i}]}_${pop2[${i}]}_single
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/standard/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/integer/k_2/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_integer_k_2
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/integer/k_3/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_integer_k_3
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/integer/k_4/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_integer_k_4
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/integer/k_5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_integer_k_5
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/LDcorr/q_0.5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_LDcorr_q_0.5
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/LDcorr/q_1/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_LDcorr_q_1
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/LDcorr/q_1.5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_LDcorr_q_1.5
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/LDcorr/q_2/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_LDcorr_q_2
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/Fst/q_0.5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_Fst_q_0.5
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/Fst/q_1/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_Fst_q_1
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/Fst/q_1.5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_Fst_q_1.5
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/Fst/q_2/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_Fst_q_2


EOF

done
done

for (( i=0; i<${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

# Test association between prs and phenotype
sbatch --mem 10G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R \
--pheno /users/k1806347/brc_scratch/Analyses/CrossMeta/Phenotype/${pheno[${i}]}.txt \
--out /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_1.test \
--assoc T \
--predictors /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_1.txt

done
done

#########
# Compare predictive utility of standard meta + single sample gwas vs standard meta + single sample gwas + integer-weighted meta
#########

pheno=(MDD MDD MDD Height BMI)
samp1=(PGC2 PGC2 PGC2 Wood Locke)
samp2=(PGC CONVERGE BBJ BBJ BBJ)
pop1=(EUR EUR EUR EUR EUR)
pop2=(EAS EAS EAS EAS EAS)

# use for loop to read all values and indexes
for (( i=0; i<${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

mkdir -p /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}

cat <<EOF > /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_2.txt
predictors group
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop2[${i}]}_only/${samp2[${i}]}_${pheno[${i}]}_${pop2[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/standard/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/integer/k_2/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles integer
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/integer/k_3/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles integer
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/integer/k_4/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles integer
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/integer/k_5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles integer

EOF
done
done

for (( i=0; i<${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

# Test association between prs and phenotype
sbatch --mem 10G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R \
--pheno /users/k1806347/brc_scratch/Analyses/CrossMeta/Phenotype/${pheno[${i}]}.txt \
--out /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_2.test \
--assoc T \
--predictors /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_2.txt

done
done

#########
# Compare predictive utility of standard meta + single sample gwas vs standard meta + single sample gwas + LDcorr/Fst-weighted meta
#########

pheno=(MDD MDD MDD Height BMI)
samp1=(PGC2 PGC2 PGC2 Wood Locke)
samp2=(PGC CONVERGE BBJ BBJ BBJ)
pop1=(EUR EUR EUR EUR EUR)
pop2=(EAS EAS EAS EAS EAS)

# use for loop to read all values and indexes
for annot in $(echo LDcorr Fst);do
for (( i=0; i<${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

mkdir -p /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}

cat <<EOF > /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_${annot}.txt
predictors group
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop2[${i}]}_only/${samp2[${i}]}_${pheno[${i}]}_${pop2[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/standard/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/${annot}/q_0.5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${annot}
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/${annot}/q_1/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${annot}
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/${annot}/q_1.5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${annot}
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/${annot}/q_2/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${annot}

EOF
done
done
done

for annot in $(echo LDcorr Fst);do
for (( i=0; i<${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

# Test association between prs and phenotype
sbatch --mem 10G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R \
--pheno /users/k1806347/brc_scratch/Analyses/CrossMeta/Phenotype/${pheno[${i}]}.txt \
--out /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_${annot}.test \
--assoc T \
--predictors /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_${annot}.txt

done
done
done

#########
# Compare predictive utility of standard meta + single sample gwas vs standard meta + single sample gwas + LDcorr and Fst-weighted meta
#########

pheno=(MDD MDD MDD Height BMI)
samp1=(PGC2 PGC2 PGC2 Wood Locke)
samp2=(PGC CONVERGE BBJ BBJ BBJ)
pop1=(EUR EUR EUR EUR EUR)
pop2=(EAS EAS EAS EAS EAS)

# use for loop to read all values and indexes
for (( i=0; i<${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

mkdir -p /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}

cat <<EOF > /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_3.txt
predictors group
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop2[${i}]}_only/${samp2[${i}]}_${pheno[${i}]}_${pop2[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/standard/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/LDcorr/q_0.5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles LDcorr_Fst
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/LDcorr/q_1/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles LDcorr_Fst
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/LDcorr/q_1.5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles LDcorr_Fst
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/LDcorr/q_2/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles LDcorr_Fst
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/Fst/q_0.5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles LDcorr_Fst
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/Fst/q_1/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles LDcorr_Fst
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/Fst/q_1.5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles LDcorr_Fst
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/Fst/q_2/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles LDcorr_Fst

EOF
done
done

for (( i=0; i<${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

# Test association between prs and phenotype
sbatch --mem 10G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R \
--pheno /users/k1806347/brc_scratch/Analyses/CrossMeta/Phenotype/${pheno[${i}]}.txt \
--out /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_3.test \
--assoc T \
--predictors /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_3.txt

done
done

#########
# Compare predictive utility of PRS-CSx single sample PRS vs PRS-CSx single sample PRS + PRS-CSx META PRS 
#########

pheno=(MDD MDD MDD Height BMI)
samp1=(PGC2 PGC2 PGC2 Wood Locke)
samp2=(PGC CONVERGE BBJ BBJ BBJ)
pop1=(EUR EUR EUR EUR EUR)
pop2=(EAS EAS EAS EAS EAS)

# use for loop to read all values and indexes
for (( i=0; i<${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

mkdir -p /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}

cat <<EOF > /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/PRS-CSx_1.txt
predictors group
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/PRS-CSx/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}.${pop1[${i}]}/prs.profiles PRS-CSx_single
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/PRS-CSx/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}.${pop2[${i}]}/prs.profiles PRS-CSx_single
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/PRS-CSx/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}.META/prs.profiles PRS-CSx_meta

EOF
done
done

for (( i=0; i<${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

# Test association between prs and phenotype
sbatch --mem 10G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R \
--pheno /users/k1806347/brc_scratch/Analyses/CrossMeta/Phenotype/${pheno[${i}]}.txt \
--out /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/PRS-CSx_1.test \
--assoc T \
--predictors /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/PRS-CSx_1.txt

done
done

#########
# Compare predictive utility of standard meta + single sample gwas vs standard meta + single sample gwas + all PRS-CSx PRS 
#########

pheno=(MDD MDD MDD Height BMI)
samp1=(PGC2 PGC2 PGC2 Wood Locke)
samp2=(PGC CONVERGE BBJ BBJ BBJ)
pop1=(EUR EUR EUR EUR EUR)
pop2=(EAS EAS EAS EAS EAS)

# use for loop to read all values and indexes
for (( i=0; i<${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

mkdir -p /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}

cat <<EOF > /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/PRS-CSx_2.txt
predictors group
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop2[${i}]}_only/${samp2[${i}]}_${pheno[${i}]}_${pop2[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/standard/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/PRS-CSx/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}.${pop1[${i}]}/prs.profiles PRS-CSx
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/PRS-CSx/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}.${pop2[${i}]}/prs.profiles PRS-CSx
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/PRS-CSx/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}.META/prs.profiles PRS-CSx

EOF
done
done

for (( i=0; i<${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

# Test association between prs and phenotype
sbatch --mem 10G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R \
--pheno /users/k1806347/brc_scratch/Analyses/CrossMeta/Phenotype/${pheno[${i}]}.txt \
--out /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/PRS-CSx_2.test \
--assoc T \
--predictors /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/PRS-CSx_2.txt

done
done

###########
# That should be enough comparisons to see what we want
###########

```

</details>

***

### Plotting results

```{r, echo=T, eval=F}
library(data.table)
library(ggplot2)
library(cowplot)

pheno<-c('MDD','MDD','MDD','Height','BMI')
samp1=c('PGC2','PGC2','PGC2','Wood','Locke')
samp2=c('PGC','CONVERGE','BBJ','BBJ','BBJ')
pop1=c('EUR','EUR','EUR','EUR','EUR')
pop2=c('EAS','EAS','EAS','EAS','EAS')

############
# Plot subset PRS
############

for(i in which(!duplicated(paste0(pheno,samp1,pop1)))){
for(targ_pop in c(pop2[i])){
assoc<-NULL
for(annot in c('LDcorr','Fst')){
  
  assoc<-rbind(assoc,fread(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/',targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'/prs_1_',annot,'.test.assoc.txt')))

}

  assoc$Target<-targ_pop

  assoc$pT<-gsub('e.','e-',gsub('.*SCORE_','',assoc$Predictor))

  assoc$Type<-NA
  assoc$Type[grepl('EUR.all',assoc$Predictor)]<-'All SNPs'
  
  assoc$Type[grepl('EUR.LDcorr',assoc$Predictor)]<-gsub('.*EUR.LDcorr.','LDcorr: ',gsub('.PredFile.*','',assoc$Predictor[grepl('EUR.LDcorr',assoc$Predictor)]))
  
  
  assoc$Type[grepl('EUR.Fst',assoc$Predictor)]<-gsub('.*EUR.Fst.','Fst: ',gsub('.PredFile.*','',assoc$Predictor[grepl('EUR.Fst',assoc$Predictor)]))

  assoc$Type<-paste0(gsub('bot.','Bottom ', gsub('top.','Top ',assoc$Type)),'%')

  assoc$pT<-factor(assoc$pT, levels=unique(assoc$pT))
  assoc$Type<-factor(assoc$Type, levels=c("LDcorr: Bottom 0.01%","LDcorr: Bottom 0.05%","LDcorr: Bottom 0.1%", "LDcorr: Bottom 0.5%","LDcorr: Top 0.01%","LDcorr: Top 0.05%","LDcorr: Top 0.1%","LDcorr: Top 0.5%", "Fst: Bottom 0.01%","Fst: Bottom 0.05%", "Fst: Bottom 0.1%", "Fst: Bottom 0.5%", "Fst: Top 0.01%", "Fst: Top 0.05%", "Fst: Top 0.1%", "Fst: Top 0.5%", "All SNPs%"))

  assoc<-assoc[complete.cases(assoc),]
  
  if(min(assoc$BETA-assoc$SE) < 0){
    ymin=min(assoc$BETA-assoc$SE)
  } else {
    ymin<-0
  }

  assoc_plot<-list()
  
  assoc_plot[['1']]<-ggplot(assoc[grepl('All|LDcorr', assoc$Type),], aes(x=pT, y=BETA, fill=Type)) +
  geom_bar(stat="identity", position=position_dodge(.9)) +
  geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
  theme_half_open() +
  background_grid() +
  ylim(ymin,max(assoc$BETA+assoc$SE)) +
  labs(y="R (SE)") +
  theme(axis.text.x = element_text(angle = 45, hjust=1))

  assoc_plot[['2']]<-ggplot(assoc[grepl('All|Fst', assoc$Type),], aes(x=pT, y=BETA, fill=Type)) +
  geom_bar(stat="identity", position=position_dodge(.9)) +
  geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
  theme_half_open() +
  background_grid() +
  ylim(ymin,max(assoc$BETA+assoc$SE)) +
  labs(y="R (SE)") +
  theme(axis.text.x = element_text(angle = 45, hjust=1))

  png(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/',targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'/prs_1.test.pT_assoc.png'), units='px', res=300, width=2500, height=2500)
    print(plot_grid(plotlist=assoc_plot, ncol=1))
  dev.off()

  eval<-NULL
  for(annot in c('LDcorr','Fst')){
    tmp<-fread(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/',targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'/prs_1_',annot,'.test.pred_eval.txt'))
    
    tmp$Model[tmp$Model == 'All_group']<-paste0('All ',annot)
    
    eval<-rbind(eval, tmp)
  
  }
  
  eval<-eval[!duplicated(eval$Model),]

  eval$Type<-gsub('.*EUR.','',gsub('_group','',eval$Model))
  eval$Type[eval$Type == 'all']<-'All SNPs'
  eval$Type<-gsub('.bot.',': Bottom ', eval$Type)
  eval$Type<-gsub('.top.',': Top ', eval$Type)
  eval$Type<-factor(eval$Type, levels = c("LDcorr: Bottom 0.01","LDcorr: Bottom 0.05","LDcorr: Bottom 0.1","LDcorr: Bottom 0.5","LDcorr: Top 0.01","LDcorr: Top 0.05", "LDcorr: Top 0.1","LDcorr: Top 0.5", "Fst: Bottom 0.01","Fst: Bottom 0.05","Fst: Bottom 0.1","Fst: Bottom 0.5", "Fst: Top 0.01","Fst: Top 0.05","Fst: Top 0.1", "Fst: Top 0.5","All SNPs", "All LDcorr","All Fst"))
  
  if(min(eval$R-eval$SE) < 0){
    ymin=min(eval$R-eval$SE)
  } else {
    ymin<-0
  }
  
  png(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/',targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'/prs_1.test.eval.png'), units='px', res=300, width=2000, height=1500)
  print(ggplot(eval, aes(x=Type, y=R, fill=Type)) +
  geom_bar(stat="identity", position=position_dodge(.9)) +
  geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2,position=position_dodge(.9)) +
  theme_half_open() +
  background_grid() +
  ylim(ymin,max(eval$R+eval$SE)) +
  labs(y="R (SE)") +
  theme(axis.text.x = element_text(angle = 45, hjust=1)))
  dev.off()

  }
}

#################
# Plot CrossMeta results
###############

for(i in 1:length(pheno)){
for(targ_pop in c(pop1[i], pop2[i])){

############
# Plot the predictive utility of each predictor
############

  assoc<-fread(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/',targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/prs_1.test.assoc.txt'))
  assoc$Target<-targ_pop

  assoc$pT<-gsub('e.','e-',gsub('.*SCORE_','',assoc$Predictor))
  
  assoc$Type[grepl(paste0('Group_',samp1[i],'.',pheno[i],'.',pop1[i],'.single'), assoc$Predictor)]<-paste0(pop1[i],' GWAS')
  assoc$Type[grepl(paste0('Group_',samp2[i],'.',pheno[i],'.',pop2[i],'.single'), assoc$Predictor)]<-paste0(pop2[i],' GWAS')

  assoc$Type[grepl(paste0('Group_',pheno[i],'.',samp1[i],'.',pop1[i],'.',samp2[i],'.',pop2[i],'.standard'), assoc$Predictor)]<-'Meta standard'

  assoc$Type[grepl(paste0('Group_',pheno[i],'.',samp1[i],'.',pop1[i],'.',samp2[i],'.',pop2[i],'.integer'), assoc$Predictor)]<-'Meta integer'
  assoc$Type[grepl(paste0('Group_',pheno[i],'.',samp1[i],'.',pop1[i],'.',samp2[i],'.',pop2[i],'.LDcorr'), assoc$Predictor)]<-'Meta LDcorr'
  assoc$Type[grepl(paste0('Group_',pheno[i],'.',samp1[i],'.',pop1[i],'.',samp2[i],'.',pop2[i],'.Fst'), assoc$Predictor)]<-'Meta Fst'

  assoc$Annot_weight<-NA
  assoc$Annot_weight[assoc$Type == 'Meta integer']<-gsub('\\.Pred.*','',gsub('.*integer.k.','',assoc$Predictor[assoc$Type == 'Meta integer']))
  assoc$Annot_weight[assoc$Type == 'Meta Fst']<-gsub('\\.Pred.*','',gsub('.*Fst.q.','',assoc$Predictor[assoc$Type == 'Meta Fst']))
  assoc$Annot_weight[assoc$Type == 'Meta LDcorr']<-gsub('\\.Pred.*','',gsub('.*LDcorr.q.','',assoc$Predictor[assoc$Type == 'Meta LDcorr']))
  
  assoc$Type[!is.na(assoc$Annot_weight) & (assoc$Type == 'Meta Fst' | assoc$Type == 'Meta LDcorr')]<-paste0(assoc$Type[!is.na(assoc$Annot_weight) & (assoc$Type == 'Meta Fst' | assoc$Type == 'Meta LDcorr')], ' q=', assoc$Annot_weight[!is.na(assoc$Annot_weight) & (assoc$Type == 'Meta Fst' | assoc$Type == 'Meta LDcorr')])
  assoc$Type[!is.na(assoc$Annot_weight) & (assoc$Type == 'Meta integer')]<-paste0(assoc$Type[!is.na(assoc$Annot_weight) & (assoc$Type == 'Meta integer')], ' k=', assoc$Annot_weight[!is.na(assoc$Annot_weight) & (assoc$Type == 'Meta integer')])

  assoc$pT<-factor(assoc$pT, levels=unique(assoc$pT))
  assoc$Type<-factor(assoc$Type, levels=unique(assoc$Type))

  if(min(assoc$BETA-assoc$SE) < 0){
    ymin=min(assoc$BETA-assoc$SE)
  } else {
    ymin<-0
  }
  
  assoc_plot<-list()
  
  assoc_plot[['1']]<-ggplot(assoc[(assoc$Type == 'EUR GWAS' | assoc$Type == 'EAS GWAS' | assoc$Type == 'Meta standard')], aes(x=pT, y=BETA, fill=Type)) +
  geom_bar(stat="identity", position=position_dodge(.9)) +
  geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
  theme_half_open() +
  background_grid() +
  ylim(ymin,max(assoc$BETA+assoc$SE)) +
  labs(y="R (SE)") +
  theme(axis.text.x = element_text(angle = 45, hjust=1))

  assoc_plot[['2']]<-ggplot(assoc[(grepl('Meta integer', assoc$Type) | assoc$Type == 'Meta standard')], aes(x=pT, y=BETA, fill=Type)) +
  geom_bar(stat="identity", position=position_dodge(.9)) +
  geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
  theme_half_open() +
  background_grid() +
  ylim(ymin,max(assoc$BETA+assoc$SE)) +
  labs(y="R (SE)") +
  theme(axis.text.x = element_text(angle = 45, hjust=1))

  assoc_plot[['3']]<-ggplot(assoc[(grepl('Meta LDcorr', assoc$Type) | assoc$Type == 'Meta standard')], aes(x=pT, y=BETA, fill=Type)) +
  geom_bar(stat="identity", position=position_dodge(.9)) +
  geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
  theme_half_open() +
  background_grid() +
  ylim(ymin,max(assoc$BETA+assoc$SE)) +
  labs(y="R (SE)") +
  theme(axis.text.x = element_text(angle = 45, hjust=1))
  
  assoc_plot[['4']]<-ggplot(assoc[(grepl('Meta Fst', assoc$Type) | assoc$Type == 'Meta standard')], aes(x=pT, y=BETA, fill=Type)) +
  geom_bar(stat="identity", position=position_dodge(.9)) +
  geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
  theme_half_open() +
  background_grid() +
  ylim(ymin,max(assoc$BETA+assoc$SE)) +
  labs(y="R (SE)") +
  theme(axis.text.x = element_text(angle = 45, hjust=1))

  png(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/',targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/prs_1.test.pT_assoc.png'), units='px', res=300, width=2000, height=3000)
    print(plot_grid(plotlist=assoc_plot, ncol=1))
  dev.off()
  }
}

###############
# Plot comparisons between all models
###############

# Plot the absolute results
for(i in 1:length(pheno)){
for(targ_pop in c(pop1[i], pop2[i])){
  eval_integer<- fread(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/',targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/prs_2.test.pred_eval.txt'))
  eval_integer$Label<-c('Baseline','Meta integer','Baseline +\nMeta integer')

  eval_LDcorr<- fread(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/',targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/prs_LDcorr.test.pred_eval.txt'))
  eval_LDcorr$Label<-c('Baseline','Meta LDcorr','Baseline +\nMeta LDcorr')
  
  eval_Fst<- fread(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/',targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/prs_Fst.test.pred_eval.txt'))
  eval_Fst$Label<-c('Baseline','Meta Fst','Baseline +\nMeta Fst')

  eval_Both<- fread(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/',targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/prs_3.test.pred_eval.txt'))
  eval_Both$Label<-c('Baseline','Meta LDcorr + \nMeta Fst','Baseline +\nMeta LDcorr +\nMeta Fst')

  eval_all<-do.call(rbind, list(eval_integer, eval_LDcorr,eval_Fst,eval_Both))
  eval_all<-eval_all[!duplicated(eval_all),]
  
  label_order<-c('Meta integer','Meta LDcorr','Meta Fst','Meta LDcorr + \nMeta Fst','Baseline','Baseline +\nMeta integer','Baseline +\nMeta LDcorr','Baseline +\nMeta Fst','Baseline +\nMeta LDcorr +\nMeta Fst')
  
  eval_all<-eval_all[match(label_order, eval_all$Label),]
  eval_all$Label<-factor(eval_all$Label, levels=eval_all$Label)

  eval_all$Target<-targ_pop

  eval_all<-eval_all[grepl('Baseline', eval_all$Label),]
  
  eval_plot<-ggplot(eval_all, aes(x=Label, y=R, colour=Label, group = 1)) +
  geom_point(stat="identity", position=position_dodge(.9), size=5) +
  geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2,position=position_dodge(.9)) +
  theme_half_open() +
  background_grid() +
  ylim(min(eval_all$R-eval_all$SE),max(eval_all$R+eval_all$SE)+0.075) +
  labs(y="R (SE)", x='', title=paste0(targ_pop,' target: ',pheno[i],'-',samp1[i],' + ',pop1[i],'-',samp2[i],'_',pop2[i])) +
  theme(axis.text.x = element_text(angle = 45, hjust=1), legend.position = "none")

  comp_integer<- fread(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/',targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/prs_2.test.pred_comp.txt'))
  comp_integer<-comp_integer[comp_integer$Model_1 == 'All' & comp_integer$Model_2 == 'single.and.standard' ]

  comp_LDcorr<- fread(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/',targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/prs_LDcorr.test.pred_comp.txt'))
  comp_LDcorr<-comp_LDcorr[comp_LDcorr$Model_1 == 'All' & comp_LDcorr$Model_2 == 'single.and.standard' ]

  comp_Fst<- fread(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/',targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/prs_Fst.test.pred_comp.txt'))
  comp_Fst<-comp_Fst[comp_Fst$Model_1 == 'All' & comp_Fst$Model_2 == 'single.and.standard' ]

  comp_BOTH<- fread(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/',targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/prs_3.test.pred_comp.txt'))
  comp_BOTH<-comp_BOTH[comp_BOTH$Model_1 == 'All' & comp_BOTH$Model_2 == 'single.and.standard' ]

  paths_integer<-data.frame(x=c(1,1,2,2),
                           y=c(max(eval_all$R+eval_all$SE)+0.005,max(eval_all$R+eval_all$SE)+0.01,max(eval_all$R+eval_all$SE)+0.01,max(eval_all$R+eval_all$SE)+0.005))
  
  paths_LDcorr<-paths_integer
  paths_LDcorr$x[paths_LDcorr$x != 1]<-paths_LDcorr$x[paths_LDcorr$x != 1]+1
  paths_LDcorr$y<-paths_LDcorr$y+0.015

  paths_Fst<-paths_LDcorr
  paths_Fst$x[paths_Fst$x != 1]<-paths_Fst$x[paths_Fst$x != 1]+1
  paths_Fst$y<-paths_Fst$y+0.015

  paths_both<-paths_Fst
  paths_both$x[paths_both$x != 1]<-paths_both$x[paths_both$x != 1]+1
  paths_both$y<-paths_both$y+0.015
  
  eval_plot<-eval_plot + 
    geom_path(data=paths_integer, aes(x=x,y=y), colour='black') +
    geom_path(data=paths_LDcorr, aes(x=x,y=y), colour='black') +
    geom_path(data=paths_Fst, aes(x=x,y=y), colour='black') +
    geom_path(data=paths_both, aes(x=x,y=y), colour='black')
    
    
  eval_plot<-eval_plot + 
    annotate("text",x=1.5,y=max(eval_all$R+eval_all$SE)+0.015,label=paste0('diff=', round(abs(comp_integer$R_diff/comp_integer$Model_2_R) * 100,1),'%; p=', format(comp_LDcorr$R_diff_pval, scientific = TRUE, digits = 2)), size=4) +
    annotate("text",x=2,y=max(eval_all$R+eval_all$SE)+0.03,label=paste0('diff=', round(abs(comp_LDcorr$R_diff/comp_LDcorr$Model_2_R) * 100,1),'%; p=', format(comp_LDcorr$R_diff_pval, scientific = TRUE, digits = 2)), size=4) +
    annotate("text",x=2.5,y=max(eval_all$R+eval_all$SE)+0.045,label=paste0('diff=', round(abs(comp_Fst$R_diff/comp_Fst$Model_2_R) * 100,1),'%; p=', format(comp_Fst$R_diff_pval, scientific = TRUE, digits = 2)), size=4) +
    annotate("text",x=3,y=max(eval_all$R+eval_all$SE)+0.06,label=paste0('diff=', round(abs(comp_BOTH$R_diff/comp_BOTH$Model_2_R) * 100,1),'%; p=', format(comp_BOTH$R_diff_pval, scientific = TRUE, digits = 2)), size=4)

  png(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/',targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/prs_all.test.comp.png'), units='px', res=300, width=2200, height=1500)
    print(eval_plot)
  dev.off()
  
  }
}

##########################
# Plot PRS-CSx results
##########################

for(i in 1:length(pheno)){
for(targ_pop in c(pop1[i], pop2[i])){

############
# Plot the predictive utility of each predictor
############

# Plot the PRS-CSx results across phi parameters
prscsx_assoc<-fread(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/',targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/PRS-CSx_1.test.assoc.txt'))
prscsx_assoc$Target<-targ_pop

prscsx_assoc$phi<-gsub('1e-00','1',gsub('e.','e-',gsub('.*phi_','',prscsx_assoc$Predictor)))

prscsx_assoc$Type[grepl('PredFile1', prscsx_assoc$Predictor)]<-paste0('PRS-CSx ', pop1[i])
prscsx_assoc$Type[grepl('PredFile2', prscsx_assoc$Predictor)]<-paste0('PRS-CSx ', pop2[i])
prscsx_assoc$Type[grepl('PredFile3', prscsx_assoc$Predictor)]<-paste0('PRS-CSx ', 'meta')

prscsx_assoc$Type<-factor(prscsx_assoc$Type, levels=unique(prscsx_assoc$Type))
prscsx_assoc$Phi<-factor(prscsx_assoc$Phi, levels=unique(prscsx_assoc$Phi))

if(min(prscsx_assoc$BETA-prscsx_assoc$SE) < 0){
  ymin=min(prscsx_assoc$BETA-prscsx_assoc$SE)
} else {
  ymin<-0
}

assoc_plot<-ggplot(prscsx_assoc, aes(x=phi, y=BETA, fill=Type)) +
  geom_bar(stat="identity", position=position_dodge(.9)) +
  geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
  theme_half_open() +
  background_grid() +
  ylim(ymin,max(prscsx_assoc$BETA+prscsx_assoc$SE)) +
  labs(y="R (SE)") +
  theme(axis.text.x = element_text(angle = 45, hjust=1))

png(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/',targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/PRS-CSx_1.test.phi_assoc.png'), units='px', res=300, width=2000, height=1000)
  print(assoc_plot)
dev.off()

# Plot the PRS-CSx models combined with standard meta-analysis
eval_prscsx_1<-fread(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/',targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/PRS-CSx_1.test.pred_eval.txt'))
eval_prscsx_1$Label<-c('PRS-CSx singles','PRS-CSx meta','PRS-CSx singles +\nPRS-CSx meta')

eval_prscsx_2<-fread(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/',targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/PRS-CSx_2.test.pred_eval.txt'))
eval_prscsx_2$Label<-c('Baseline','PRS-CSx singles +\nPRS-CSx meta','Baseline +\nPRS-CSx singles +\nPRS-CSx meta')
  
eval_prscx_all<-do.call(rbind, list(eval_prscsx_1, eval_prscsx_2))
eval_prscx_all<-eval_prscx_all[!duplicated(eval_prscx_all$Label),]

label_order<-c("PRS-CSx singles","PRS-CSx meta","PRS-CSx singles +\nPRS-CSx meta","Baseline", "Baseline +\nPRS-CSx singles +\nPRS-CSx meta")

eval_prscx_all<-eval_prscx_all[match(label_order, eval_prscx_all$Label),]
eval_prscx_all$Label<-factor(eval_prscx_all$Label, levels=eval_prscx_all$Label)

eval_prscx_all$Target<-targ_pop

eval_prscx_plot<-ggplot(eval_prscx_all, aes(x=Label, y=R, colour=Label, group = 1)) +
geom_point(stat="identity", position=position_dodge(.9), size=5) +
geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2,position=position_dodge(.9)) +
theme_half_open() +
background_grid() +
ylim(min(eval_prscx_all$R-eval_prscx_all$SE),max(eval_prscx_all$R+eval_prscx_all$SE)+0.075) +
labs(y="R (SE)", x='', title=paste0(targ_pop,' target: ',pheno[i],'-',samp1[i],' + ',pop1[i],'-',samp2[i])) +
theme(axis.text.x = element_text(angle = 45, hjust=1), legend.position = "none")

comp_prscsx_1<- fread(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/',targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/PRS-CSx_1.test.pred_comp.txt'))
comp_prscsx_2<- fread(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/',targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/PRS-CSx_2.test.pred_comp.txt'))

comp_prscsx_all<-rbind(comp_prscsx_1, comp_prscsx_2)

paths_1<-data.frame(x=c(1,1,2,2),
                   y=c(max(eval_prscx_all$R+eval_prscx_all$SE)+0.005,max(eval_prscx_all$R+eval_prscx_all$SE)+0.01,max(eval_prscx_all$R+eval_prscx_all$SE)+0.01,max(eval_prscx_all$R+eval_prscx_all$SE)+0.005))
  
paths_2<-paths_1
paths_2$x[paths_2$x != 1]<-paths_2$x[paths_2$x != 1]+0.975
paths_2$y<-paths_2$y+0.015

paths_3<-paths_1
paths_3$x<-c(3,3,3.975,3.975)

paths_4<-paths_1
paths_4$x<-c(3.025,3.025,5,5)
paths_4$y<-paths_1$y+0.015

paths_5<-paths_1
paths_5$x<-c(4.025,4.025,5,5)

eval_prscx_plot<-eval_prscx_plot + 
  geom_path(data=paths_1, aes(x=x,y=y), colour='black') +
  geom_path(data=paths_2, aes(x=x,y=y), colour='black') +
  geom_path(data=paths_3, aes(x=x,y=y), colour='black') +
  geom_path(data=paths_4, aes(x=x,y=y), colour='black') +
  geom_path(data=paths_5, aes(x=x,y=y), colour='black')

  eval_prscx_plot<-eval_prscx_plot + 
    annotate("text",x=1.5,y=max(eval_prscx_all$R+eval_prscx_all$SE)+0.015,label=paste0('diff=', round(abs(comp_prscsx_1$R_diff/comp_prscsx_1$Model_2_R) * 100,1)[2],'%; p=', format(comp_prscsx_1$R_diff_pval[2], scientific = TRUE, digits = 2)), size=3.5) +
    annotate("text",x=2,y=max(eval_prscx_all$R+eval_prscx_all$SE)+0.03,label=paste0('diff=', round(abs(comp_prscsx_1$R_diff/comp_prscsx_1$Model_2_R) * 100,1)[7],'%; p=', format(comp_prscsx_1$R_diff_pval[7], scientific = TRUE, digits = 2)), size=3.5) +
    annotate("text",x=3.5,y=max(eval_prscx_all$R+eval_prscx_all$SE)+0.015,label=paste0('diff=', round(abs(comp_prscsx_2$R_diff/comp_prscsx_2$Model_2_R) * 100,1)[2],'%; p=', format(comp_prscsx_2$R_diff_pval[2], scientific = TRUE, digits = 2)), size=3.5) +
    annotate("text",x=4,y=max(eval_prscx_all$R+eval_prscx_all$SE)+0.03,label=paste0('diff=', round(abs(comp_prscsx_2$R_diff/comp_prscsx_2$Model_2_R) * 100,1)[8],'%; p=', format(comp_prscsx_2$R_diff_pval[8], scientific = TRUE, digits = 2)), size=3.5) +
    annotate("text",x=4.5,y=max(eval_prscx_all$R+eval_prscx_all$SE)+0.015,label=paste0('diff=', round(abs(comp_prscsx_2$R_diff/comp_prscsx_2$Model_2_R) * 100,1)[7],'%; p=', format(comp_prscsx_2$R_diff_pval[7], scientific = TRUE, digits = 2)), size=3.5)

png(paste0('/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/',targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],'/PRS-CSx_2.test.comp.png'), units='px', res=300, width=2500, height=1500)
  print(eval_prscx_plot)
dev.off()
  

}
}

```

</details>

***

# Results

```{bash, eval=T, echo = F}
mkdir -p Images/CrossMeta

pheno=(MDD MDD MDD Height BMI)
samp1=(PGC2 PGC2 PGC2 Wood Locke)
samp2=(PGC CONVERGE BBJ BBJ BBJ)
pop1=(EUR EUR EUR EUR EUR)
pop2=(EAS EAS EAS EAS EAS)

# use for loop to read all values and indexes
for (( i=0; i<${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

mkdir -p Images/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}

cp /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_1.test.pT_assoc.png Images/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}

cp /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_all.test.comp.png Images/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}

cp /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/PRS-CSx_1.test.phi_assoc.png Images/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}

cp /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/PRS-CSx_2.test.comp.png Images/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}

done
done

```

***

## CrossMeta

***

### Per pT results

```{r,  results='asis', echo=FALSE}
pheno<-c('MDD','MDD','MDD','Height','BMI')
samp1=c('PGC2','PGC2','PGC2','Wood','Locke')
samp2=c('PGC','CONVERGE','BBJ','BBJ','BBJ')
pop1=c('EUR','EUR','EUR','EUR','EUR')
pop2=c('EAS','EAS','EAS','EAS','EAS')

for(i in 1:length(pheno)){
  for(targ_pop in c(pop1[i], pop2[i])){
    cat(paste0("<details><summary>", targ_pop," target, GWAS1: ",pheno[i],"_",samp1[i],", GWAS2: ",pop1[i],"_",samp2[i],"_",pop2[i],"</summary>\n\n"))
    cat(paste0("![](Images/CrossMeta/assoc/",targ_pop,"/",pheno[i],"_",samp1[i],"_",pop1[i],"_",samp2[i],"_",pop2[i],"/prs_1.test.pT_assoc.png)\n\n"))
    cat("</details>\n\n")
    cat(' \n \n')
  }
}
```

***

### Model comparison results

```{r,  results='asis', echo=FALSE}
pheno<-c('MDD','MDD','MDD','Height','BMI')
samp1=c('PGC2','PGC2','PGC2','Wood','Locke')
samp2=c('PGC','CONVERGE','BBJ','BBJ','BBJ')
pop1=c('EUR','EUR','EUR','EUR','EUR')
pop2=c('EAS','EAS','EAS','EAS','EAS')

for(i in 1:length(pheno)){
  for(targ_pop in c(pop1[i], pop2[i])){
    cat(paste0("<details><summary>", targ_pop," target, GWAS1: ",pheno[i],"_",samp1[i],", GWAS2: ",pop1[i],"_",samp2[i],"_",pop2[i],"</summary>\n\n"))
    cat(paste0("![](Images/CrossMeta/assoc/",targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],"/prs_all.test.comp.png)\n\n"))
    cat("</details>\n\n")
    cat(' \n \n')
  }
}
```

***

## PRS-CSx

***

### Per phi results

```{r,  results='asis', echo=FALSE}
pheno<-c('MDD','MDD','MDD','Height','BMI')
samp1=c('PGC2','PGC2','PGC2','Wood','Locke')
samp2=c('PGC','CONVERGE','BBJ','BBJ','BBJ')
pop1=c('EUR','EUR','EUR','EUR','EUR')
pop2=c('EAS','EAS','EAS','EAS','EAS')

for(i in 1:length(pheno)){
  for(targ_pop in c(pop1[i], pop2[i])){
    cat(paste0("<details><summary>", targ_pop," target, GWAS1: ",pheno[i],"_",samp1[i],", GWAS2: ",pop1[i],"_",samp2[i],"_",pop2[i],"</summary>\n\n"))
    cat(paste0("![](Images/CrossMeta/assoc/",targ_pop,"/",pheno[i],"_",samp1[i],"_",pop1[i],"_",samp2[i],"_",pop2[i],"/PRS-CSx_1.test.phi_assoc.png)\n\n"))
    cat("</details>\n\n")
    cat(' \n \n')
  }
}
```

***

### Model comparison results

```{r,  results='asis', echo=FALSE}
pheno<-c('MDD','MDD','MDD','Height','BMI')
samp1=c('PGC2','PGC2','PGC2','Wood','Locke')
samp2=c('PGC','CONVERGE','BBJ','BBJ','BBJ')
pop1=c('EUR','EUR','EUR','EUR','EUR')
pop2=c('EAS','EAS','EAS','EAS','EAS')

for(i in 1:length(pheno)){
  for(targ_pop in c(pop1[i], pop2[i])){
    cat(paste0("<details><summary>", targ_pop," target, GWAS1: ",pheno[i],"_",samp1[i],", GWAS2: ",pop1[i],"_",samp2[i],"_",pop2[i],"</summary>\n\n"))
    cat(paste0("![](Images/CrossMeta/assoc/",targ_pop,'/',pheno[i],'_',samp1[i],'_',pop1[i],'_',samp2[i],'_',pop2[i],"/PRS-CSx_2.test.comp.png)\n\n"))
    cat("</details>\n\n")
    cat(' \n \n')
  }
}
```



