---
title: "Genotype-based Prediction (GenoPred)"
output: 
  html_document:
    toc: true
    theme: united
---

***

Welcome!

This website provides details of a project evaluating methods for genotype-based prediction. The website will be updated continuously as the project develops.

Please see our [GenoPred pipeline (GenoPredPipe)](https://github.com/opain/GenoPred/tree/master/GenoPredPipe), providing a one-liner for ancestry inference and polygenic score calculation.

***

#### Contents:

* Data Preparation
  * Genetic
    * [Reference genetic: European 503 subset of 1KG](Pipeline_prep.html)
    * [Reference genetic: European 10K subset of UK Biobank](Pipeline_prep_withUKBB_ref.html)
    * [Target genetic](Genotype-based_scoring_in_target_samples.html)
  * Phenotypic
    * [Target phenotypic](Target_sample_phenotype_prep.html)
    
* Main Research Questions
  * [Comparison of polygenic scoring methods](Determine_optimal_polygenic_scoring_approach.html)
    * [See here for updated results including LDAK's MegaPRS method](Determine_optimal_polygenic_scoring_approach_update_21102021.html)
  * [Evaluating gene expression risk scores](Functionally_informed_prediction.html)
  * [Converting polygenic scores to the absolute scale](Absolute_Conversion.html)
    * [Link to Shiny app](PRS_to_Abs_tool.html)
  * [Local genetic correlation weighted polygenic scores](local_rG_weighted_pgs.html)
  * [Improve polygenic prediction in diverse populations](DiverseAncestry.html) (under development)

* Side Questions 
  * [Standardised vs. non-standardised](standardised_vs_nonstandarised.html)
  * [Estimating significance of model comparisons](Z-test_vs_ANOVA_V2.html)

* Future developments
  * Stratification using pathway/tissue-specific genetic scores
  * Gene expression risk scores using observed expression

***

#### Help

This website was created by Oliver Pain (oliver.pain@kcl.ac.uk). Please get in touch if you have any questions or comments. Also, follow me on [twitter](https://twitter.com/ollie_pain) and [GitHub](https://github.com/opain) for updates.

***

#### Acknowledgments

I would like to thank the following people who have contributed to this project so far:

* **Cathryn Lewis**
* Lasse Folkersen
* Kylie Glanville
* Saskia Hagenaars
* Saskia Selzam
* Anna F&uuml;rtjes
* Helena Gaspar
* Jonathon Coleman
* Gerome Breen
* Robert Plomin
* Alex Gillett
* Ken Hanscombe
* Ryan Arathimos
* Toshiki Kono

***

#### Funding:

This work is funded by the MRC (MR/N015746/1), and the National Institute for Health Research (NIHR) Biomedical Research Centre at South London and Maudsley NHS Foundation Trust and King’s College London. The authors acknowledge use of the research computing facility at King’s College London, Rosalind (https://rosalind.kcl.ac.uk), which is delivered in partnership with the NIHR Maudsley BRC, and part-funded by capital equipment grants from the Maudsley Charity (award 980) and Guy’s & St. Thomas’ Charity (TR130505). The views expressed are those of the authors and not necessarily those of the NHS, the NIHR or the Department of Health and Social Care. 

***

#### Change log:

**24/06/2020**:

* Change log started
* Updated Ancestry_identifier.R to use proper grid search in enet, allow other modelling methods to be specified, and improve model performance reporting.

**13/08/2020**:

* Add comparison between standardised and non-standardised polygenic scores.

**10/02/2021**:

* PRS comparison: Added LDPred2 and DBSLMM methods. Added prostate and breast cancer. Add polygenicity and ancestry senstivity analysis.
* GeRS comparison: Minor formatting changes.

**13/04/2021**

* PRS comparison: Evaluated new version of GCTB v2.03
* PRS conversion to absolute scale: Updated html

**22/08/2021**

* Added GenoPredPipe documentation

**22/08/2021**

* Added LDAK's MegaPRS to polygenic score methods comparison
* Benchmarked the runtime of LDpred2 with new SFMB format

**14/03/2022**

* Added local genetic correlation weighted polygenic score analysis

