---
title: "GenoPredPipe"
params:
  name: ""
  output: ""
output:
  html_document:
    toc: true
    theme: united
    toc_depth: 3
    toc_float: true
    fig_caption: yes

---

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '../')

library(data.table)
library(knitr)
library(ggplot2)
library(cowplot)
library(kableExtra)
library(stringr)

knit_hooks$set(optipng = hook_optipng)
knit_hooks$set(pngquant = hook_pngquant)

```

***

<font size="4"> 
ID: `r gsub('_',' ',params$name)`
</font>

***

# Descriptives

```{R, eval=T, echo=F}

imputation_log<-readLines(paste0(params$output,'/',params$name,'/',params$name,'.23andMe_imputer.log'))

nvar_before_imp<-gsub(' variants','',gsub('.*contains ','',imputation_log[grepl('Input target data contains', imputation_log)]))

nvar_after_imp<-gsub(' SNPs.','',gsub('.*contain ','',imputation_log[grepl('.gen files contain', imputation_log)]))

nvar_hard_after_imp<-gsub(' SNPs.','',gsub('.*contain ','',imputation_log[grepl('.plink files contain', imputation_log)]))

ref_bim<-NULL
for(i in 1:22){
  ref_bim<-rbind(ref_bim, fread(paste0('resources/data/ref/ref.chr',i,'.bim')))
}
nvar_in_ref<-nrow(ref_bim)

harmoniser_log<-readLines(paste0(params$output,'/',params$name,'/',params$name,'.ref.chr.hm3_harmoniser.log'))

ref_overlap_after_imp<-gsub('%.*','%',harmoniser_log[grepl('of variants in reference are in the target.',harmoniser_log)])

```

Observed genotype data contained `r nvar_before_imp` variants. After imputation, `r nvar_after_imp` variants were available, of which `r nvar_hard_after_imp` had a probability >0.9. Of the `r nvar_in_ref` variants in the reference, `r ref_overlap_after_imp` were available as hard calls in the target.

***

# Ancestry

## Population Classification

```{R, eval=T, echo=F, message=F, warning=F}

model_pred<-fread(paste0(params$output,'/',params$name,'/ancestry/ancestry_all/',params$name,'.Ancestry.model_pred'))
pred_pop<-names(model_pred)[model_pred[1,] == max(model_pred[1,-1:-2])]  

model_pred_melt<-melt(model_pred[,-1:-2])
names(model_pred_melt)<-c('code','prob')

```

The ancestry model found the individual most likely to be of `r pred_pop` ancestry, with a probability of `r paste0(round(max(model_pred[1,-1:-2])*100,1),'%')`.

***

```{R, eval=T, echo=F, message=F, warning=F, fig.cap="Population Probability", fig.dim = c(8, 4)}
ggplot(model_pred_melt, aes(x=code, y=prob*100, fill=code)) +
  geom_bar(stat="identity") +
  labs(x='Population', y="Probability (%)") +
  theme_minimal_hgrid() +
  theme(legend.position = "none") +
  ylim(0,100)

```

***

<details><summary>Show principal component plots</summary>

```{r, results='asis', echo=F}
cat(paste0("![Principal Component Scores Compared to Reference Populations](",normalizePath(params$output, mustWork = FALSE),'/',params$name,'/ancestry/ancestry_all/', params$name,".Ancestry.PCs_plot_POP.png)"))
```

</details>

