---
title: "GenoPredPipe"
params:
  name: ""
  output: ""
output:
  html_document:
    toc: true
    theme: united
    toc_depth: 3
    toc_float: true
    fig_caption: yes

---

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '../')

library(data.table)
library(knitr)
library(ggplot2)
library(cowplot)
library(ggchicklet)
library(kableExtra)
library(stringr)

knit_hooks$set(optipng = hook_optipng)
knit_hooks$set(pngquant = hook_pngquant)

```

***

<font size="4"> 
ID: `r gsub('_',' ',params$name)`
</font>

***

# Descriptives

```{R, eval=T, echo=F}

imputation_log<-readLines(paste0(params$output,'/',params$name,'/',params$name,'.23andMe_imputer.log'))

nvar_before_imp<-gsub(' variants','',gsub('.*contains ','',imputation_log[grepl('Input target data contains', imputation_log)]))

nvar_after_imp<-gsub(' SNPs.','',gsub('.*contain ','',imputation_log[grepl('.gen files contain', imputation_log)]))

nvar_hard_after_imp<-gsub(' SNPs.','',gsub('.*contain ','',imputation_log[grepl('.plink files contain', imputation_log)]))

ref_bim<-NULL
for(i in 1:22){
  ref_bim<-rbind(ref_bim, fread(paste0('resources/data/1kg/1KGPhase3.w_hm3.chr',i,'.bim')))
}
nvar_in_ref<-nrow(ref_bim)

harmoniser_log<-readLines(paste0(params$output,'/',params$name,'/',params$name,'.1KGphase3.hm3.chr.hm3_harmoniser.log'))

ref_overlap_after_imp<-gsub('%.*','%',harmoniser_log[grepl('of variants in reference are in the target.',harmoniser_log)])

```

Observed genotype data contained `r nvar_before_imp` variants. After imputation, `r nvar_after_imp` variants were available, of which `r nvar_hard_after_imp` had a probability >0.9. Of the `r nvar_in_ref` variants in the reference, `r ref_overlap_after_imp` were available as hard calls in the target.

***

# Ancestry

## Super Population Classification

```{R, eval=T, echo=F, message=F, warning=F}
# Read in 1KG population codes
pop_codes<-fread('resources/data/1kg/1kg_pop_codes.tsv')
pop_codes<-pop_codes[,1:3]
names(pop_codes)<-c('pop','code','super_code')
pop_codes<-pop_codes[pop_codes$code != '',]

super_pop_codes<-fread('resources/data/1kg/1kg_super_pop_codes.tsv')
names(super_pop_codes)<-c('pop','code')

model_pred<-fread(paste0(params$output,'/',params$name,'/ancestry/ancestry_all/',params$name,'.Ancestry.model_pred'))
pred_super_pop<-names(model_pred)[model_pred[1,] == max(model_pred[1,-1:-2])]  

model_pred_melt<-melt(model_pred[,-1:-2])
names(model_pred_melt)<-c('code','prob')
model_pred_melt<-merge(model_pred_melt, super_pop_codes, by='code')

```

The ancestry model found the individual most likely to be of `r super_pop_codes$pop[super_pop_codes$code == pred_super_pop]` ancestry, with a probability of `r paste0(round(max(model_pred[1,-1:-2])*100,1),'%')`.

***

```{R, eval=T, echo=F, message=F, warning=F, fig.cap="Super Population Probability", fig.dim = c(8, 4)}
ggplot(model_pred_melt, aes(x=pop, y=prob*100, fill=pop)) +
  geom_bar(stat="identity") +
  labs(x='Super Population', y="Probability (%)") +
  theme_minimal_hgrid() +
  theme(legend.position = "none") +
  ylim(0,100)

```

***

<details><summary>Show principal component plots</summary>

```{r, echo=FALSE, fig.cap="Principal Component Scores Compared to Reference Super Populations", out.width = '100%'}
knitr::include_graphics(normalizePath(paste0(params$output,'/',params$name,'/ancestry/ancestry_all/', params$name,'.Ancestry.PCs_plot_super_pop.png')))
```

```{r, eval=T, echo=F}
kable(super_pop_codes, align = "c") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "responsive"))
```

</details>

## `r super_pop_codes$pop[super_pop_codes$code == pred_super_pop]` Population Classification

```{R, eval=T, echo=F, message=F, warning=F}

model_pred<-fread(paste0(params$output,'/',params$name,'/ancestry/ancestry_',pred_super_pop,'/',params$name,'.Ancestry.model_pred'))
pred_pop<-names(model_pred)[model_pred[1,] == max(model_pred[1,-1:-2])]  

model_pred_melt<-melt(model_pred[,-1:-2])
names(model_pred_melt)<-c('code','prob')
model_pred_melt<-merge(model_pred_melt, pop_codes, by='code')
model_pred_melt$pop_wrap = str_wrap(model_pred_melt$pop, width = 15)

```

The ancestry model found the individual most likely to be of `r pred_pop` ancestry, with a probability of `r paste0(round(max(model_pred[1,-1:-2])*100,1),'%')`.

***

```{R, eval=T, echo=F, message=F, warning=F, fig.cap="Super Population Probability", fig.dim = c(8, 4)}
ggplot(model_pred_melt, aes(x=pop_wrap, y=prob*100, fill=pop_wrap)) +
  geom_bar(stat="identity") +
  labs(x='Population', y="Probability (%)") +
  theme_minimal_hgrid() +
  theme(legend.position = "none") +
  ylim(0,100)

```

***

<details><summary>Show principal component plots</summary>

```{r, echo=FALSE, fig.cap="Principal Component Scores Compared to Reference Super Populations", out.width = '100%'}
knitr::include_graphics(normalizePath(paste0(params$output,'/',params$name,'/ancestry/ancestry_',pred_super_pop,'/', params$name,'.Ancestry.PCs_plot_pop.png')))
```

```{r, eval=T, echo=F}
kable(pop_codes[pop_codes$super_code == pred_super_pop, 1:2], align = "c") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "responsive"))
```

</details>

***

# Polygenic Scores {.tabset}

```{R, eval=T, echo=F, message=F, warning=F}
config<-read.table('config.yaml')

# Read in gwas_list
gwas_list<-fread(as.character(config$V2[config$V1 == 'gwas_list:']))

# Read in prs
dbslmm_prs<-NULL
for(gwas_i in gwas_list$name){
  prs_i<-fread(paste0(params$output,'/',params$name,'/prs/',pred_super_pop,'/dbslmm/',gwas_i,'/',params$name,'.',gwas_i,'.',pred_super_pop,'.profiles'))
  dbslmm_prs<-rbind(dbslmm_prs, data.frame(gwas=gwas_i,
                                           prs_z=prs_i[[paste0(gwas_i,'_DBSLMM')]]))
}

##
# Convert to absolute scale
##

# Read in absolute scale functions
source('scripts/absolute_scale_conversions.R')

# Read in pseudoval results
pseudoval<-NULL
for(gwas_i in gwas_list$name){
  pseudoval_log<-read.table(paste0('resources/data/1kg/prs_pseudoval/',gwas_i,'/lassosum_pseudo_',gwas_i,'.log'), sep='&', header=F)$V1
  pseudoval_log_r<-as.numeric(gsub('value = ','',pseudoval_log[grepl('value = ', pseudoval_log)]))
  
  if(is.na(gwas_list$sampling[gwas_list$name == gwas_i])){
    pseudoval_log_r2<-pseudoval_log_r^2
    
    pseudoval<-rbind(pseudoval, data.frame(gwas=gwas_i,
                                           pseudoval_r2obs=pseudoval_log_r2,
                                           pseudoval_d=NA,
                                           pseudoval_auc=NA))

  } else {
    n_case<-gwas_list$sampling[gwas_list$name == gwas_i]
    n_con<-1-gwas_list$sampling[gwas_list$name == gwas_i]
  
    a<-(n_case+n_con)^2/(n_case*n_con)
  
    pseudoval_log_r2<-pseudoval_log_r^2
    pseudoval_log_d<-sqrt(a)*pseudoval_log_r/sqrt(1-pseudoval_log_r^2)
    pseudoval_log_auc <- pnorm(pseudoval_log_d/sqrt(2), 0, 1)
    
    pseudoval<-rbind(pseudoval, data.frame(gwas=gwas_i,
                                           pseudoval_r2obs=pseudoval_log_r2,
                                           pseudoval_d=pseudoval_log_d,
                                           pseudoval_auc=pseudoval_log_auc))

  }
}

# Calculate absolute risk
abs_res<-NULL
n_quant<-1000
for(gwas_i in gwas_list$name){
  if(is.na(gwas_list$prevalence[gwas_list$name == gwas_i])){
    quant<-which_quant(PRS_z_score = dbslmm_prs$prs_z[dbslmm_prs$gwas == gwas_i], n_quantile = n_quant)
        
    indiv_result<-mean_sd_quant.f(PRS_R2=pseudoval$pseudoval_r2obs[pseudoval$gwas == gwas_i], Outcome_mean=0, Outcome_sd=1, n_quantile=n_quant, quant=quant)
    
    abs_res<-rbind(abs_res, data.frame(gwas=gwas_i,
                                       prev=NA,
                                       abs_prob=NA,
                                       mean=indiv_result$x_mean,
                                       sd=indiv_result$x_sd))
  } else {
    risk_quantiles<-ccprobs.f(d=pseudoval$pseudoval_d[pseudoval$gwas == gwas_i], prev=gwas_list$prevalence[gwas_list$name == gwas_i], n_quantile=n_quant)
    indiv_result_all<-risk_quantiles[dbslmm_prs$prs_z[dbslmm_prs$gwas == gwas_i] > risk_quantiles$q_min & dbslmm_prs$prs_z[dbslmm_prs$gwas == gwas_i] < risk_quantiles$q_max,]
    indiv_result<-indiv_result_all[, c('p_case', 'p_control')]
    
    abs_res<-rbind(abs_res, data.frame(gwas=gwas_i,
                                   prev=gwas_list$prevalence[gwas_list$name == gwas_i],
                                   abs_prob=indiv_result$p_case,
                                   mean=NA,
                                   sd=NA))
  }
}

# Create plots
plots<-list()
for(gwas_i in gwas_list$name){
  PRS_z_score<-dbslmm_prs$prs_z[dbslmm_prs$gwas == gwas_i]
  
  # Create relative plot
  rel_plot<-ggplot(data = data.frame(x = c(-4, 4)), aes(x=x)) +
  stat_function(fun = dnorm, n = 101, args = list(mean = 0, sd = 1)) +
  stat_function(fun = dnorm, args = list(mean = 0, sd = 1), xlim = c(PRS_z_score, -4),
                geom = "area", fill = "#CC66FF", alpha = .4) +
  stat_function(fun = dnorm, args = list(mean = 0, sd = 1), xlim = c(PRS_z_score, 4),
                geom = "area", fill = "#FF6633", alpha = .4) +
  geom_vline(xintercept=PRS_z_score, linetype='dashed') +
  geom_text(label=paste0(round(pnorm(PRS_z_score)*100,1),"% have lower \npolygenic scores"), mapping=aes(x=PRS_z_score-0.1, y=0.5), colour='#CC66FF', hjust='right', vjust=0.8, size=5) +
  geom_text(label=paste0(round(100-(pnorm(PRS_z_score)*100),1),"% have higher \npolygenic scores"), mapping=aes(x=PRS_z_score+0.1, y=0.5), colour='#FF6633', hjust='left', vjust=0.8, size=5) +
  scale_y_continuous(breaks = NULL) +
  theme_half_open() +
  xlim(-5,5) +
  labs(y='Number of people', x='Polygenic Score', title='Distribution of polygenic scores') +
  theme(plot.title = element_text(hjust = 0.5))

  # Create absolute plot
  if(is.na(gwas_list$prevalence[gwas_list$name == gwas_i])){
    # Define parameters
    PRS_R2=pseudoval$pseudoval_r2obs[pseudoval$gwas == gwas_i]
    Outcome_mean=gwas_list$mean[gwas_list$name == gwas_i]
    Outcome_sd=gwas_list$sd[gwas_list$name == gwas_i]
    conf_int=95/100
    n_quant<-1000
    
    # Run analysis
    quant<-which_quant(PRS_z_score = PRS_z_score, n_quantile = n_quant)
    
    risk_quantiles<-mean_sd_quant.f(PRS_R2=PRS_R2, Outcome_mean=Outcome_mean, Outcome_sd=Outcome_sd, n_quantile=n_quant, quant=quant)
    indiv_result_all<-risk_quantiles[PRS_z_score > risk_quantiles$q_min & PRS_z_score <= risk_quantiles$q_max,]
    
    ref<-dnorm(seq(-4.5,4.5,0.01), 0, 1)
    ref_plot<-data.frame(x=(seq(-4.5,4.5,0.01)*Outcome_sd)+Outcome_mean,
                         y=ref,
                         Group='General public')
    
    indiv<-dnorm(seq(-4.5,4.5,0.01), 0, 1)
    indiv_plot<-data.frame(x=(seq(-4.5,4.5,0.01)*indiv_result_all$x_sd)+indiv_result_all$x_mean,
                           y=indiv,
                           Group='People like you')
    
    plot_dat<-rbind(ref_plot, indiv_plot)
    plot_dat$Group<-factor(plot_dat$Group, levels=c('General public','People like you'))
    
    # Calculate 95CI for target individual
    lowCI<-qnorm((1-conf_int)/2,indiv_result_all$x_mean,indiv_result_all$x_sd)
    highCI<-qnorm(1-((1-conf_int)/2),indiv_result_all$x_mean,indiv_result_all$x_sd)
    
    abs_plot<-ggplot(plot_dat, aes(x=x, y=y, fill=Group)) +
        geom_area(alpha=0.4, colour='black') +
        scale_fill_manual(values=c("#84CA72","#0066CC")) +
        labs(y='Number of people', x='Trait', title='You compared to general population', fill=NULL) +
        geom_segment(aes(x = indiv_result_all$x_mean , y = 0, xend = indiv_result_all$x_mean, yend = 0.4), color="black") +
        geom_segment(aes(x = lowCI , y = 0, xend = lowCI, yend = 0.38), color="#0066CC", linetype="dashed") +
        geom_segment(aes(x = highCI , y = 0, xend = highCI, yend = 0.38), color="#0066CC", linetype="dashed") +
        geom_text(label=paste0('Estimate = ',round(indiv_result_all$x_mean,2)), mapping=aes(x=Outcome_mean+Outcome_sd, y=0.5), colour='black', hjust='left', vjust=0.8, size=5, check_overlap = TRUE) +
        geom_text(label=paste0(conf_int*100,'% CI = ',round(lowCI,2),' â€“ ',round(highCI,2)), mapping=aes(x=Outcome_mean+Outcome_sd, y=0.5), colour='#0066CC', hjust='left', vjust=2.5, size=5, check_overlap = TRUE) +
        scale_y_continuous(breaks = NULL) +
        theme_half_open() +
        theme(plot.title = element_text(hjust = 0.5)) +
        theme(legend.position=c(0.01,0.95), legend.box = "horizontal")
    
    plot_text<-ggdraw() + draw_text(paste0('Your PRS Z-score for ',gwas_list$label[gwas_list$name == gwas_i],' is ', dbslmm_prs$prs_z[dbslmm_prs$gwas == gwas_i],' which\nis higher than ', paste0(round(100*pnorm(dbslmm_prs$prs_z[dbslmm_prs$gwas == gwas_i]),1),'%'),' of other ', super_pop_codes$pop[super_pop_codes$code == pred_super_pop],' individuals.\n\nGiven the PRS explains ', round(100*pseudoval$pseudoval_r2obs[pseudoval$gwas == gwas_i],1),'% of the variance in ', gwas_list$label[gwas_list$name == gwas_i],',\n and the mean and SD of ', gwas_list$label[gwas_list$name == gwas_i],' in the general population is ',gwas_list$mean[gwas_list$name == gwas_i],' and ',gwas_list$sd[gwas_list$name == gwas_i],' respectively,\n on average people with your PRS have a ', gwas_list$label[gwas_list$name == gwas_i],' of ', round(abs_res$mean[abs_res$gwas == gwas_i],2), " (95% CI=",round(lowCI,2),'-',round(highCI,2),")."), x = 0.5, y = 0.5)

    plots[[gwas_i]]<-plot_grid(rel_plot, abs_plot, plot_text, ncol = 1, rel_heights = c(1, 1,0.5))
  } else {
    d<-pseudoval$pseudoval_d[pseudoval$gwas == gwas_i]
    prev<-gwas_list$prevalence[gwas_list$name == gwas_i]*100
    E_PRS <- d*(prev/100)
    varPRS <- (prev/100)*(1+(d^2) - (d*(prev/100))^2) + (1-(prev/100))*(1 - (d*(prev/100))^2)
        
    tmp<-dnorm_2_new(x=seq(-4,4,length.out=101), mean_1 = (0-(E_PRS))/sqrt(varPRS), sd_1 = 1/sqrt(varPRS), mean_2 = (d-(E_PRS))/sqrt(varPRS), sd_2 = 1/sqrt(varPRS), p_2=(prev/100))
        
    # Define parameters
    risk_quantiles<-ccprobs.f(d=d, prev=(prev/100), n_quantile=1000)
    
    indiv_result_all<-risk_quantiles[PRS_z_score > risk_quantiles$q_min & PRS_z_score < risk_quantiles$q_max,]
    indiv_result<-indiv_result_all[, c('p_case', 'p_control')]
    
    indiv_result<-melt(indiv_result)
    indiv_result$variable<-c('Case','Control')
    indiv_result$variable<-factor(indiv_result$variable,levels = c('Control','Case'))
        
    # Create a grid of dots, some red and some blue
    n_case_1<-round(indiv_result$value*100,1)[1]
    n_control_1<-round(indiv_result$value*100,1)[2]
    n_case<-round(indiv_result$value*100)[1]
    n_control<-round(indiv_result$value*100)[2]
    
    # Create data for general population example
    pop_case_1<-round(prev,1)
    pop_case<-round(prev)
    pop_control_1<-round(100-prev,1)
    pop_control<-round(100-prev)
    pop_result<-data.frame(variable=c('Case','Control'),
                           value=c((prev/100),1-(prev/100)))
    
    pop_result$variable<-factor(pop_result$variable, levels = c('Control','Case'))
    
    plot_dat_pop<-(matrix(c(rep('Case',pop_case),rep('Control',100-pop_case)),nrow=10, ncol=10))
    plot_dat_pop<-data.frame(melt(plot_dat_pop))
    plot_dat_pop$value<-factor(plot_dat_pop$value, levels = c('Control','Case'))
    
    bar_chart<-ggplot(data.frame(x=1,y=0:1), aes(x=x, y=y)) +
        geom_chicklet(radius = grid::unit(1, 'mm'), data=indiv_result, mapping=aes(x=1, y=value, fill=variable), stat="identity",position='stack') +
        scale_fill_manual(values=c("#84CA72","#0066CC"), drop = F) +
        annotate("text", x=1.5, y=((((n_control_1)/2))+n_case_1)/100, label=paste0(n_control_1,"%\ndo not have trait"), colour = '#84CA72', hjust=0, size=6) +
        annotate("text", x=1.5, y=((n_case_1/2))/100, label=paste0(n_case_1,'%\nhave trait'), colour = '#0066CC', hjust=0, size=6) +
        ylim(-0.1,1.05) +
        theme_half_open() +
        labs(title='Of people with your genetics,') +
        xlim(0.25,2.5) +
        theme(axis.line=element_blank(),axis.text.x=element_blank(),
              axis.text.y=element_blank(),axis.ticks=element_blank(),
              axis.title.x=element_blank(),
              axis.title.y=element_blank(),
              panel.background=element_blank(),panel.border=element_blank(),panel.grid.major=element_blank(),
              panel.grid.minor=element_blank(),plot.background=element_blank(), plot.title = element_text(hjust = 0.5), legend.position = "none")
        
    bar_chart_pop<-ggplot(data.frame(x=1,y=0:1), aes(x=x, y=y)) +
        geom_chicklet(radius = grid::unit(1, 'mm'), data=pop_result, mapping=aes(x=1, y=value, fill=variable), stat="identity",position='stack') +
        scale_fill_manual(values=c("#84CA72","#0066CC"), drop = F) +
        annotate("text", x=1.5, y=(((pop_control_1/2))+pop_case_1)/100, label=paste0(pop_control_1,"%\ndo not have trait"), colour = '#84CA72', hjust=0, size=6) +
        annotate("text", x=1.5, y=((pop_case_1/2))/100, label=paste0(pop_case_1,'%\nhave trait'), colour = '#0066CC', hjust=0, size=6) +
        ylim(-0.1,1.05) +
        theme_half_open() +
        labs(title='In the general population,') +
        xlim(0.25,2.5) +
        theme(axis.line=element_blank(),axis.text.x=element_blank(),
              axis.text.y=element_blank(),axis.ticks=element_blank(),
              axis.title.x=element_blank(),
              axis.title.y=element_blank(),
              panel.background=element_blank(),panel.border=element_blank(),panel.grid.major=element_blank(),
              panel.grid.minor=element_blank(),plot.background=element_blank(), plot.title = element_text(hjust = 0.5), legend.position = "none")
        
        bar_chart_grid<-plot_grid(bar_chart, bar_chart_pop, labels = NULL, nrow = 1)
        
        plot_text<-ggdraw() + draw_text(paste0('Your PRS Z-score for ',gwas_list$label[gwas_list$name == gwas_i],' is ', dbslmm_prs$prs_z[dbslmm_prs$gwas == gwas_i],' which\nis higher than ', paste0(round(100*pnorm(dbslmm_prs$prs_z[dbslmm_prs$gwas == gwas_i]),1),'%'),' of other ', super_pop_codes$pop[super_pop_codes$code == pred_super_pop],' individuals.\n\nGiven the PRS has an AUC of ', round(pseudoval$pseudoval_auc[pseudoval$gwas == gwas_i],3),',\n and the prevelance of ', gwas_list$label[gwas_list$name == gwas_i],' in the general population is ',gwas_list$prevalence[gwas_list$name == gwas_i],',\n on average ',round(100*abs_res$abs_prob[abs_res$gwas == gwas_i],1),'% of people with your PRS have ', gwas_list$label[gwas_list$name == gwas_i],'.'), x = 0.5, y = 0.5)

        plots[[gwas_i]]<-plot_grid(rel_plot, bar_chart_grid, plot_text, ncol = 1, rel_heights = c(1, 1, 0.5))
  }
}

```

```{r, eval=T, echo=F}
all_dat<-Reduce(function(dtf1, dtf2) merge(dtf1, dtf2, by = "gwas", all.x = TRUE), list(dbslmm_prs, pseudoval,abs_res))

gwas_list_brief<-gwas_list[,c('name','label','mean','sd')]
names(gwas_list_brief)<-c('gwas','label','pop_mean','pop_sd')

all_dat<-merge(all_dat, gwas_list_brief, by='gwas')

all_dat_brief<-all_dat[,c('label','pseudoval_r2obs','pseudoval_auc','prev','pop_mean','pop_sd','prs_z','abs_prob','mean','sd')]

names(all_dat_brief)<-c('Outcome','PRS R-squared','PRS AUC','Prevelance','Mean','SD','PRS Z-score','Prevalence','Mean','SD')
all_dat_brief[,-1]<-round(all_dat_brief[,-1],3)

kable(all_dat_brief, align = "c") %>%
  add_header_above(c(" " = 1, "PRS Descriptives" = 2, "Distribution in General\nPopulation"=3, "Distribution in People\nLike You" = 4)) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "responsive"))

```

```{r,  results='asis', echo=FALSE, fig.dim=c(10,9), dpi = 70, optipng = '-o7'}

for(gwas_i in gwas_list$name){
  cat('##',gwas_list$label[gwas_list$name == gwas_i],' \n')
  print(plots[[gwas_i]])
  
  cat(' \n \n')
}


```

***

</div>


