<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Evaluate local rG weighted polygenic scores</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GenoPred</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Evaluate local rG weighted polygenic scores</h1>

</div>


<style>
p.caption {
  font-size: 1.5em;
}
</style>
<style type="text/css">
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
</style>
<hr />
<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>Here we will test whether weighting PGS by local heritability and genetic correlations with the target phenotype improves prediction. We will use three target outcomes: Depression, BMI, and Intelligence. We will collect GWAS for traits geneticaly correlated with the target outcomes, derive polygenic scores using the leading PRS method (LDAK MegaPRS), and then evaluate their predictive utility before and after weighting by local genetic correlation (estimated using HESS or LAVA).</p>
<p>A preprint of this work can be found <a href="https://doi.org/10.1101/2022.03.10.483736">here</a>.</p>
<hr />
</div>
<div id="methods" class="section level1">
<h1><span class="header-section-number">2</span> Methods</h1>
<hr />
<div id="gwas-sumstats" class="section level2">
<h2><span class="header-section-number">2.1</span> GWAS sumstats</h2>
<hr />
<div id="depression" class="section level3">
<h3><span class="header-section-number">2.1.1</span> Depression</h3>
<p>Target trait GWAS: Wray et al. excl UKB</p>
<p>Secondary GWAS: Schizophreina (PGC2), Bipolar disorder (PGC2), Autism, Smoking, ADHD, Educationtional attainment, Anxiety, age at menarche</p>
<hr />
</div>
<div id="bmi" class="section level3">
<h3><span class="header-section-number">2.1.2</span> BMI</h3>
<p>Target trait GWAS: Locke et al.</p>
<p>Secondary GWAS: Obesity, Childhood obesity, Waist circumference, coronary artery disease, age at menarche, Educationtional attainment, type 2 diabetes, fasting insulin.</p>
<hr />
</div>
<div id="intelligence" class="section level3">
<h3><span class="header-section-number">2.1.3</span> Intelligence</h3>
<p>Target trait GWAS: Intelligence</p>
<p>Secondary GWAS: Educational attainment, high IQ, Childhood IQ, ADHD, Depression, Schizophrenia, Autism, Longevity</p>
<hr />
<details>
<p><summary>Remove sample overlap between UKB and Savage et al. Intelligence GWAS</summary></p>
<p>Jeanne Savage has kindly sent me the UKB only sumstats. The UKB data was derived from two versions of the same fluid IQ test, one carried out in-person on a touchscreen device (‘ts’) and one web-based (‘wb’). There are two separate GWAS files, one for each of these variables, which were meta-analyzed alongside the other datasets.</p>
<p>To avoid sample overlap, we will first meta-analyse the two UKB sumstats using syntax provided by Jeanne, and then substract the UKB meta-GWAS sumstats from the Savage et al. sumstats.</p>
<pre class="bash"><code>
# Run METAL
/users/k1806347/brc_scratch/Software/metal.sh

SCHEME SAMPLESIZE
MARKER SNP
WEIGHT N
ALLELE EFF_ALL NONEFF_ALL
EFFECT BETA
STDERR SE
PVALUE P
PROCESS /users/k1806347/brc_scratch/Data/GWAS_sumstats/Savage/ts_clean_mac100_info60.sumstats.gz
PROCESS /users/k1806347/brc_scratch/Data/GWAS_sumstats/Savage/wb_clean_mac100_info60.sumstats.gz
MINWEIGHT 50000
OUTFILE /users/k1806347/brc_scratch/Data/GWAS_sumstats/Savage/ukb_meta .tbl

ANALYZE

QUIT
</code></pre>
<pre class="r"><code># Format the METAL output
library(data.table)

# Read in the sumstats
ukb_meta&lt;-fread(&#39;/users/k1806347/brc_scratch/Data/GWAS_sumstats/Savage/ukb_meta1.tbl&#39;)

# Format for QC using GenoFunc sumstat_cleaner.R
names(ukb_meta)&lt;-c(&#39;SNP&#39;,&#39;A1&#39;,&#39;A2&#39;,&#39;N&#39;,&#39;BETA&#39;,&#39;P&#39;,&#39;Direction&#39;)
ukb_meta$CHR&lt;-gsub(&#39;:.*&#39;,&#39;&#39;,ukb_meta$SNP)
ukb_meta$ORIGBP&lt;-gsub(&#39;.*:&#39;,&#39;&#39;,ukb_meta$SNP)
ukb_meta$A1&lt;-toupper(ukb_meta$A1)
ukb_meta$A2&lt;-toupper(ukb_meta$A2)
ukb_meta$SE&lt;-1
ukb_meta&lt;-ukb_meta[,c(&#39;CHR&#39;,&#39;ORIGBP&#39;,&#39;A1&#39;,&#39;A2&#39;,&#39;BETA&#39;,&#39;SE&#39;,&#39;N&#39;,&#39;P&#39;),with=F]

fwrite(ukb_meta, &#39;/users/k1806347/brc_scratch/Data/GWAS_sumstats/Savage/ukb_meta1_reformat.tbl.gz&#39;, quote=F, sep=&#39; &#39;, na=&#39;NA&#39;)</code></pre>
<pre class="bash"><code># Run GenoFunc sumstat_cleaner on UKB meta-analysis
cd /users/k1806347/brc_scratch/Software/MyGit/GenoFunc/GenoFuncPipe

/users/k1806347/brc_scratch/Software/Rscript.sh scripts/sumstat_cleaner.R \
  --sumstats /users/k1806347/brc_scratch/Data/GWAS_sumstats/Savage/ukb_meta1_reformat.tbl.gz \
  --ref_chr resources/data/1kg/1KG.Phase3.EUR.MAF_001.chr \
  --output /users/k1806347/brc_scratch/Data/GWAS_sumstats/Savage/ukb_meta1_cleaned

# Run GenoFunc sumstat_cleaner on full Savage meta-analysis
cd /users/k1806347/brc_scratch/Software/MyGit/GenoFunc/GenoFuncPipe

/users/k1806347/brc_scratch/Software/Rscript.sh scripts/sumstat_cleaner.R \
  --sumstats /scratch/groups/ukbiobank/sumstats/cleaned/INTE03.gz \
  --ref_chr resources/data/1kg/1KG.Phase3.EUR.MAF_001.chr \
  --output /users/k1806347/brc_scratch/Data/GWAS_sumstats/Savage/INTE03_cleaned
</code></pre>
<pre class="r"><code># Read in the full and ukb only sumstats
library(data.table)

full&lt;-fread(&#39;/users/k1806347/brc_scratch/Data/GWAS_sumstats/Savage/INTE03_cleaned.gz&#39;)
ukb&lt;-fread(&#39;/users/k1806347/brc_scratch/Data/GWAS_sumstats/Savage/ukb_meta1_cleaned.gz&#39;)

# Calculate Z
full$Z&lt;-full$BETA/full$SE
ukb$Z&lt;-ukb$BETA/ukb$SE

# Merge together
both_matched&lt;-merge(full[,c(&#39;CHR&#39;,&#39;BP&#39;,&#39;SNP&#39;,&#39;A1&#39;,&#39;A2&#39;,&#39;Z&#39;,&#39;N&#39;), with=F], ukb[,c(&#39;SNP&#39;,&#39;A1&#39;,&#39;A2&#39;,&#39;Z&#39;,&#39;N&#39;), with=F], by=c(&#39;SNP&#39;,&#39;A1&#39;,&#39;A2&#39;))

# Calculate weight of full and ukb cohorts
both_matched$W.x&lt;-sqrt(both_matched$N.x)
both_matched$W.y&lt;-sqrt(both_matched$N.y)

# Perform the meta-analysis in reverse to calculate the sumstats based on all cohorts except UK Biobank (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2922887/)
# Subtract the UKB Z (Z.y) from the full Z (Z.x)
both_matched$Z.noukb&lt;-((both_matched$Z.x*both_matched$W.x)-(both_matched$Z.y*both_matched$W.y)) / sqrt((both_matched$W.x^2)-(both_matched$W.y^2))

# Calcualte N of meta-analysis without UKB
both_matched$N.noukb&lt;-both_matched$N.x - both_matched$N.y

# Subset relevent columns and format for GenoPredPipe QC
both_matched&lt;-both_matched[,c(&#39;CHR&#39;,&#39;BP&#39;,&#39;SNP&#39;,&#39;A1&#39;,&#39;A2&#39;,&#39;Z.noukb&#39;,&#39;N.noukb&#39;), with=T]
both_matched$P&lt;-2*pnorm(-abs(both_matched$Z.noukb))
names(both_matched)&lt;-c(&#39;CHR&#39;,&#39;ORIGBP&#39;,&#39;SNP&#39;,&#39;A1&#39;,&#39;A2&#39;,&#39;BETA&#39;,&#39;N&#39;,&#39;P&#39;)

fwrite(both_matched, &#39;/users/k1806347/brc_scratch/Data/GWAS_sumstats/Savage/INTE03_noUKB.gz&#39;, quote=F, sep=&#39; &#39;, na=&#39;NA&#39;)</code></pre>
</details>
<p>index</p>
<details>
<p><summary>Create table listing GWAS</summary></p>
<pre class="r"><code># I need the latest csv of the database to create the table
gwas&lt;-c(&#39;DEPR06&#39;,&#39;SCHI02&#39;,&#39;BIPO02&#39;,&#39;AUTI07&#39;,&#39;SMOK04&#39;,&#39;ADHD05&#39;,&#39;COLL01&#39;,&#39;DIAB05&#39;,&#39;ANXI02&#39;,&#39;MENA01F&#39;,&#39;OBES01&#39;,&#39;WAIS01&#39;,&#39;COAD01&#39;,&#39;GLYC05&#39;,&#39;INTE01&#39;,&#39;INTE03&#39;,&#39;BODY04&#39;)
pheno&lt;-read.csv(&#39;/users/k1806347/brc_scratch/Data/GWAS_sumstats/QC_sumstats_list_031218.csv&#39;, stringsAsFactors=F)

pheno&lt;-pheno[pheno$Code %in% gwas,]
pheno&lt;-pheno[c(&#39;Code&#39;,&#39;sample_size_discovery&#39;,&#39;Ncases&#39;,&#39;Ncontrols&#39;,&#39;PMID&#39;)]
pheno&lt;-pheno[match(gwas, pheno$Code),]
pheno$Code&lt;-gwas
pheno$Phenotype&lt;-c(&#39;Major Depression&#39;,&#39;Schizophrenia&#39;,&#39;Bipolar Disorder&#39;,&#39;Autism&#39;,&#39;Smoking (Ever/Never Regular)&#39;,&#39;Attention Deficit Hyperactivity Disorder&#39;,&#39;College Completion&#39;,&#39;Type-2 Diabetes&#39;,&#39;Anxiety (Factor Score)&#39;,&#39;Age at Menarche&#39;,&#39;Obesity Class 1&#39;,&#39;Waist Circumference&#39;,&#39;Coronary Arety Disease&#39;,&#39;Fasting Insulin (Age- &amp; Sex-adjusted)&#39;,&#39;Childhood Intelligence&#39;,&#39;Intelligence&#39;,&#39;BMI&#39;)
pheno$PMID[pheno$Code == &#39;BIPO02&#39;]&lt;-&#39;31043756&#39;
pheno$PMID[pheno$Code == &#39;AUTI07&#39;]&lt;-&#39;30804558&#39;
pheno$PMID[pheno$Code == &#39;ADHD05&#39;]&lt;-&#39;30478444&#39;
pheno$PMID[pheno$Code == &#39;INTE03&#39;]&lt;-&#39;29942086&#39;

# Update INTE03 information to account for removal of UKB
pheno$sample_size_discovery[pheno$Code== &#39;INTE03&#39;]&lt;-74214

names(pheno)&lt;-c(&#39;Code&#39;,&#39;N&#39;,&#39;Ncase&#39;,&#39;Ncon&#39;,&#39;PMID&#39;,&#39;Phenotype&#39;)

pheno$N&lt;-gsub(&#39;,&#39;,&#39;&#39;,pheno$N)
pheno$Ncase&lt;-gsub(&#39;,&#39;,&#39;&#39;,pheno$Ncase)
pheno$Ncon&lt;-gsub(&#39;,&#39;,&#39;&#39;,pheno$Ncon)

pheno$Ncase[pheno$Code == &#39;BIPO02&#39;]&lt;-20352
pheno$Ncon[pheno$Code == &#39;BIPO02&#39;]&lt;-31358
pheno$Ncase[pheno$Code == &#39;AUTI07&#39;]&lt;-18381
pheno$Ncon[pheno$Code == &#39;AUTI07&#39;]&lt;-27969

pheno$N&lt;-as.numeric(pheno$N)
pheno$Ncase&lt;-as.numeric(pheno$Ncase)
pheno$Ncon&lt;-as.numeric(pheno$Ncon)

pheno$N[is.na(pheno$N)]&lt;-pheno$Ncase[is.na(pheno$N)]+pheno$Ncon[is.na(pheno$N)]

pheno&lt;-pheno[,c(&#39;Code&#39;,&#39;Phenotype&#39;,&#39;N&#39;,&#39;Ncase&#39;,&#39;Ncon&#39;,&#39;PMID&#39;)]

write.csv(pheno, &#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/gwas_lis.csv&#39;, row.names=F, quote=F)</code></pre>
</details>
<p>index</p>
<details>
<p><summary>Show table of GWAS sumstats</summary></p>
<table>
<thead>
<tr class="header">
<th align="left">Code</th>
<th align="left">Phenotype</th>
<th align="right">N</th>
<th align="right">Ncase</th>
<th align="right">Ncon</th>
<th align="right">PMID</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">DEPR06</td>
<td align="left">Major Depression</td>
<td align="right">431394</td>
<td align="right">116404</td>
<td align="right">314990</td>
<td align="right">29700475</td>
</tr>
<tr class="even">
<td align="left">SCHI02</td>
<td align="left">Schizophrenia</td>
<td align="right">77096</td>
<td align="right">33640</td>
<td align="right">43456</td>
<td align="right">25056061</td>
</tr>
<tr class="odd">
<td align="left">BIPO02</td>
<td align="left">Bipolar Disorder</td>
<td align="right">51710</td>
<td align="right">20352</td>
<td align="right">31358</td>
<td align="right">31043756</td>
</tr>
<tr class="even">
<td align="left">AUTI07</td>
<td align="left">Autism</td>
<td align="right">46350</td>
<td align="right">18381</td>
<td align="right">27969</td>
<td align="right">30804558</td>
</tr>
<tr class="odd">
<td align="left">SMOK04</td>
<td align="left">Smoking (Ever/Never Regular)</td>
<td align="right">74035</td>
<td align="right">41969</td>
<td align="right">32066</td>
<td align="right">20418890</td>
</tr>
<tr class="even">
<td align="left">ADHD05</td>
<td align="left">Attention Deficit Hyperactivity Disorder</td>
<td align="right">53293</td>
<td align="right">19099</td>
<td align="right">34194</td>
<td align="right">30478444</td>
</tr>
<tr class="odd">
<td align="left">COLL01</td>
<td align="left">College Completion</td>
<td align="right">95427</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">23722424</td>
</tr>
<tr class="even">
<td align="left">DIAB05</td>
<td align="left">Type-2 Diabetes</td>
<td align="right">159208</td>
<td align="right">26676</td>
<td align="right">132532</td>
<td align="right">28566273</td>
</tr>
<tr class="odd">
<td align="left">ANXI02</td>
<td align="left">Anxiety (Factor Score)</td>
<td align="right">17310</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">26754954</td>
</tr>
<tr class="even">
<td align="left">MENA01F</td>
<td align="left">Age at Menarche</td>
<td align="right">132989</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">25231870</td>
</tr>
<tr class="odd">
<td align="left">OBES01</td>
<td align="left">Obesity Class 1</td>
<td align="right">98697</td>
<td align="right">32858</td>
<td align="right">65839</td>
<td align="right">23563607</td>
</tr>
<tr class="even">
<td align="left">WAIS01</td>
<td align="left">Waist Circumference</td>
<td align="right">231927</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">25673412</td>
</tr>
<tr class="odd">
<td align="left">COAD01</td>
<td align="left">Coronary Arety Disease</td>
<td align="right">184305</td>
<td align="right">60801</td>
<td align="right">123504</td>
<td align="right">26343387</td>
</tr>
<tr class="even">
<td align="left">GLYC05</td>
<td align="left">Fasting Insulin (Age- &amp; Sex-adjusted)</td>
<td align="right">38238</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">20081858</td>
</tr>
<tr class="odd">
<td align="left">INTE01</td>
<td align="left">Childhood Intelligence</td>
<td align="right">12441</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">23358156</td>
</tr>
<tr class="even">
<td align="left">INTE03</td>
<td align="left">Intelligence</td>
<td align="right">74214</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">29942086</td>
</tr>
<tr class="odd">
<td align="left">BODY04</td>
<td align="left">BMI</td>
<td align="right">322154</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">25673413</td>
</tr>
</tbody>
</table>
</details>
<hr />
</div>
<div id="gwas-quality-control" class="section level3">
<h3><span class="header-section-number">2.1.4</span> GWAS Quality Control</h3>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code>. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Pipeline_prep.config

# Create directory
mkdir -p /users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats
</code></pre>
<pre class="r"><code>gwas&lt;-c(&#39;/scratch/groups/ukbiobank/sumstats/cleaned/DEPR06&#39;,&#39;/scratch/groups/ukbiobank/sumstats/cleaned/SCHI02&#39;,&#39;/scratch/groups/ukbiobank/sumstats/cleaned/BIPO02&#39;,&#39;/scratch/groups/ukbiobank/sumstats/cleaned/AUTI07&#39;,&#39;/scratch/groups/ukbiobank/sumstats/cleaned/SMOK04&#39;,&#39;/scratch/groups/ukbiobank/sumstats/cleaned/ADHD05&#39;,&#39;/scratch/groups/ukbiobank/sumstats/cleaned/COLL01&#39;,&#39;/scratch/groups/ukbiobank/sumstats/cleaned/DIAB05&#39;,&#39;/scratch/groups/ukbiobank/sumstats/cleaned/ANXI02&#39;,&#39;/scratch/groups/ukbiobank/sumstats/cleaned/MENA01F&#39;,&#39;/scratch/groups/ukbiobank/sumstats/cleaned/OBES01&#39;,&#39;/scratch/groups/ukbiobank/sumstats/cleaned/WAIS01&#39;,&#39;/scratch/groups/ukbiobank/sumstats/cleaned/COAD01&#39;,&#39;/scratch/groups/ukbiobank/sumstats/cleaned/GLYC05&#39;,&#39;/scratch/groups/ukbiobank/sumstats/cleaned/INTE01&#39;,&#39;/users/k1806347/brc_scratch/Data/GWAS_sumstats/Savage/INTE03_noUKB&#39;,&#39;/scratch/groups/ukbiobank/sumstats/cleaned/BODY04&#39;)

gwas_id&lt;-gsub(&#39;_.*&#39;,&#39;&#39;,gsub(&#39;.*/&#39;,&#39;&#39;,gwas))

to_do&lt;-data.frame(file=gwas, id=gwas_id)
to_do$file&lt;-paste0(to_do$file,&#39;.gz&#39;)

# Remove from list if already cleaned
for(i in to_do$id){
  if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/&#39;,i,&#39;.cleaned.gz&#39;))){
    to_do$cleaned&lt;-T
  } else {
    to_do$cleaned&lt;-F
  }
}

to_do&lt;-to_do[to_do$cleaned == F,]
to_do$cleaned&lt;-NULL

write.table(to_do, &#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/todo.txt&#39;, col.names=F, row.names=F, quote=F)</code></pre>
<pre class="bash"><code>
# Create shell script to run using sbatch
cat &gt; /users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/sbatch.sh &lt;&lt; &#39;EOF&#39;
#!/bin/sh

#SBATCH -p shared,brc
#SBATCH --mem 5G

. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Pipeline_prep.config

file=$(awk -v var=&quot;$SLURM_ARRAY_TASK_ID&quot; &#39;NR == var {print $1}&#39; /users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/todo.txt)
id=$(awk -v var=&quot;$SLURM_ARRAY_TASK_ID&quot; &#39;NR == var {print $2}&#39; /users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/todo.txt)

/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/sumstat_cleaner/sumstat_cleaner.R \
  --sumstats ${file} \
  --ref_plink_chr ${Geno_1KG_dir}/1KGPhase3.w_hm3.chr \
  --ref_freq_chr ${Geno_1KG_dir}/freq_files/EUR/1KGPhase3.w_hm3.EUR.chr \
  --output /users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/${id}.cleaned

EOF

sbatch --array 1-$(wc -l /users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/todo.txt | cut -d&#39; &#39; -f1)%5 /users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/sbatch.sh

# GLYC05 contains MAF so FREQ doesn&#39;t correspond to particular allele. As a result we are loosing many SNPs. Remove FREQ column and re-QC.

cp ${gwas_rep_cleaned}/GLYC05.gz /users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/GLYC05_tmp.gz
zcat /users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/GLYC05_tmp.gz | cut -f 1-3,5-8 | gzip &gt; /users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/GLYC05_tmp2.gz

/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/sumstat_cleaner/sumstat_cleaner.R \
  --sumstats /users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/GLYC05_tmp2.gz \
  --ref_plink_chr ${Geno_1KG_dir}/1KGPhase3.w_hm3.chr \
  --ref_freq_chr ${Geno_1KG_dir}/freq_files/EUR/1KGPhase3.w_hm3.EUR.chr \
  --output /users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/GLYC05.cleaned

rm /users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/GLYC05_tmp.gz
rm /users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/GLYC05_tmp2.gz</code></pre>
</details>
<hr />
</div>
</div>
<div id="estimate-genome-wide-and-local-genetic-correlation" class="section level2">
<h2><span class="header-section-number">2.2</span> Estimate genome-wide and local genetic correlation</h2>
<details>
<p><summary>Show code</summary></p>
<pre class="r"><code># Create phenotype data.frame
# Leave number of cases and control as NA as when set as bainry LAVA seems to take a lot longer.
# We can convert the heritability estimatest to the liability model later if we want to.

gwas&lt;-c(&#39;DEPR06&#39;,&#39;SCHI02&#39;,&#39;BIPO02&#39;,&#39;AUTI07&#39;,&#39;SMOK04&#39;,&#39;ADHD05&#39;,&#39;COLL01&#39;,&#39;DIAB05&#39;,&#39;ANXI02&#39;,&#39;MENA01F&#39;,&#39;OBES01&#39;,&#39;WAIS01&#39;,&#39;COAD01&#39;,&#39;GLYC05&#39;,&#39;INTE01&#39;,&#39;INTE03&#39;,&#39;BODY04&#39;)

pheno&lt;-data.frame(phenotype=gwas,
                  cases=NA,
                  controls=NA,
                  prevalence=NA,
                  filename=paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/&#39;, gwas,&#39;.cleaned.gz&#39;))

dir.create(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/lava/&#39;)
write.table(pheno, &#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/lava/input.info.txt&#39;, col.names=T, row.names=F, quote=F)</code></pre>
<pre class="bash"><code>##########
# Prepare sample overlap file
##########

# Munge all sumstats
for gwas in $(echo DEPR06 SCHI02 BIPO02 AUTI07 SMOK04 ADHD05 COLL01 DIAB05 ANXI02 MENA01F OBES01 WAIS01 COAD01 GLYC05 INTE01 INTE03 BODY04); do
  sbatch -p brc,shared ~/brc_scratch/Software/munge_sumstats.sh \
   --sumstats /users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/${gwas}.cleaned.gz \
   --merge-alleles /users/k1806347/brc_scratch/Data/ldsc/w_hm3.snplist \
   --out /users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/${gwas}.cleaned
done

# Run bivariate LDSC for all traits
mkdir /users/k1806347/brc_scratch/Analyses/local_rg_pgs/lava/sample_overlap

for gwas1 in $(echo DEPR06 SCHI02 BIPO02 AUTI07 SMOK04 ADHD05 COLL01 DIAB05 ANXI02 MENA01F OBES01 WAIS01 COAD01 GLYC05 INTE01 INTE03 BODY04); do
for gwas2 in $(echo DEPR06 SCHI02 BIPO02 AUTI07 SMOK04 ADHD05 COLL01 DIAB05 ANXI02 MENA01F OBES01 WAIS01 COAD01 GLYC05 INTE01 INTE03 BODY04); do
   sbatch -p brc,shared ~/brc_scratch/Software/ldsc.sh \
     --rg /users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/${gwas1}.cleaned.sumstats.gz,/users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/${gwas2}.cleaned.sumstats.gz \
     --ref-ld-chr /users/k1806347/brc_scratch/Data/ldsc/eur_w_ld_chr/ \
     --w-ld-chr /users/k1806347/brc_scratch/Data/ldsc/eur_w_ld_chr/ \
     --out /users/k1806347/brc_scratch/Analyses/local_rg_pgs/lava/sample_overlap/${gwas1}_${gwas2}_rg
done
done

# Collate results
cd /users/k1806347/brc_scratch/Analyses/local_rg_pgs/lava/sample_overlap/
FILES=($(ls *_rg.log))
N=$(echo ${#FILES[@]})
for I in ${FILES[@]}; do
        PHEN=$(echo $I | sed &#39;s/_rg\.log//&#39;)
        
        # subset log files to relevant output
        tail -n 5 $I | head -n 2 &gt; $PHEN.rg
        
        # add to single data set
        if [[ $I == ${FILES[0]} ]]; then
            cat $PHEN.rg &gt; all.rg
        else
            cat $PHEN.rg | sed &#39;1d&#39; &gt;&gt; all.rg
        fi
done
</code></pre>
<pre class="r"><code>scor = read.table(&quot;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/lava/sample_overlap/all.rg&quot;,header=T)              # read in
scor = scor[,c(&quot;p1&quot;,&quot;p2&quot;,&quot;gcov_int&quot;)]             # retain key headers
scor$p1 = gsub(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/&#39;,&#39;&#39;,gsub(&quot;.cleaned.sumstats.gz&quot;,&quot;&quot;,scor$p1))
scor$p2 = gsub(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/&#39;,&#39;&#39;,gsub(&quot;.cleaned.sumstats.gz&quot;,&quot;&quot;,scor$p2))
phen = unique(scor$p1)
n = length(phen)
mat = matrix(NA,n,n)                    # create matrix
rownames(mat) = colnames(mat) = phen    # set col/rownames
for (i in phen) {
        for (j in phen) {
                mat[i,j] = subset(scor, p1==i &amp; p2==j)$gcov_int
        }
}
if (!all(t(mat)==mat)) { mat[lower.tri(mat)] = t(mat)[lower.tri(mat)] }  # sometimes there might be small differences in gcov_int depending on which phenotype was analysed as the outcome / predictor
mat = round(cov2cor(mat),5)                       # standardise
write.table(mat, &quot;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/lava/sample_overlap/sample.overlap.txt&quot;, quote=F)   # save</code></pre>
<pre class="bash"><code>#####
# Create R script to run LAVA across genome
#####

cat &gt; /users/k1806347/brc_scratch/Analyses/local_rg_pgs/lava/lava.R &lt;&lt; &#39;EOF&#39;
# command line arguments, specifying input/output file names and phenotype subset
arg = commandArgs(T); ref.prefix = arg[1]; loc.file = arg[2]; info.file = arg[3]; sample.overlap.file = arg[4]; phenos = unlist(strsplit(arg[5],&quot;,&quot;)); out.fname = arg[6]

### Load package
library(LAVA)

### Read in data
loci = read.loci(loc.file); n.loc = nrow(loci)
input = process.input(info.file, sample.overlap.file, ref.prefix, phenos)

print(paste(&quot;Starting LAVA analysis for&quot;,n.loc,&quot;loci&quot;))
progress = ceiling(quantile(1:n.loc, seq(.05,1,.05)))   # (if you want to print the progress)

### Analyse
u=b=list()
for (i in 1:n.loc) {
        if (i %in% progress) print(paste(&quot;..&quot;,names(progress[which(progress==i)])))     # (printing progress)
        locus = process.locus(loci[i,], input)                                          # process locus
        
        # It is possible that the locus cannot be defined for various reasons (e.g. too few SNPs), so the !is.null(locus) check is necessary before calling the analysis functions.
        if (!is.null(locus)) {
                # extract some general locus info for the output
                loc.info = data.frame(locus = locus$id, chr = locus$chr, start = locus$start, stop = locus$stop, n.snps = locus$n.snps, n.pcs = locus$K)
                
                # run the univariate and bivariate tests
                loc.out = run.univ.bivar(locus)
                u[[i]] = cbind(loc.info, loc.out$univ)
                if(!is.null(loc.out$bivar)) b[[i]] = cbind(loc.info, loc.out$bivar)
        }
}

# save the output
write.table(do.call(rbind,u), paste0(out.fname,&quot;.univ.lava&quot;), row.names=F,quote=F,col.names=T)
write.table(do.call(rbind,b), paste0(out.fname,&quot;.bivar.lava&quot;), row.names=F,quote=F,col.names=T)

print(paste0(&quot;Done! Analysis output written to &quot;,out.fname,&quot;.*.lava&quot;))

EOF

########
# Run LAVA for each target GWAS
########

for target_gwas in $(echo DEPR06 BODY04 INTE03); do
for secondary_gwas in $(echo DEPR06 SCHI02 BIPO02 AUTI07 SMOK04 ADHD05 COLL01 DIAB05 ANXI02 MENA01F OBES01 WAIS01 COAD01 GLYC05 INTE01 INTE03 BODY04); do
if [ ${target_gwas} != ${secondary_gwas} ]; then
mkdir -p /users/k1806347/brc_scratch/Analyses/local_rg_pgs/lava/results/${target_gwas}
sbatch -p brc,shared --mem 10G /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Analyses/local_rg_pgs/lava/lava.R \
  /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.GW \
  /users/k1806347/brc_scratch/Data/LAVA/blocks_s2500_m25_f1_w200.GRCh37_hg19.locfile \
  /users/k1806347/brc_scratch/Analyses/local_rg_pgs/lava/input.info.txt \
  /users/k1806347/brc_scratch/Analyses/local_rg_pgs/lava/sample_overlap/sample.overlap.txt \
  ${target_gwas},${secondary_gwas} \
  /users/k1806347/brc_scratch/Analyses/local_rg_pgs/lava/results/${target_gwas}/${target_gwas}.${secondary_gwas}
fi
done

sleep 9000
done

# Notes: If the target phenotype GWAS does not have a large sample, then heritability within a locus may not be significant, and therefore rho cannot be accuratly estimated. This means there may also be an advantage to using unweighted secondary PGS as they may contain information from shared loci that cannot be identified using LAVA.
# This takes approximately 2.5 hours per pair of traits</code></pre>
</details>
<hr />
</div>
<div id="characterise-the-genome-wide-and-local-genetic-correlation-results" class="section level2">
<h2><span class="header-section-number">2.3</span> Characterise the genome-wide and local genetic correlation results</h2>
<details>
<p><summary>Show code</summary></p>
<pre class="r"><code>library(data.table)
library(cowplot)
library(ggplot2)

target_gwas&lt;-c(&#39;DEPR06&#39;,&#39;BODY04&#39;,&#39;INTE03&#39;)
secondary_gwas&lt;-c(&#39;DEPR06&#39;,&#39;SCHI02&#39;,&#39;BIPO02&#39;,&#39;AUTI07&#39;,&#39;SMOK04&#39;,&#39;ADHD05&#39;,&#39;COLL01&#39;,&#39;DIAB05&#39;,&#39;ANXI02&#39;,&#39;MENA01F&#39;,&#39;OBES01&#39;,&#39;WAIS01&#39;,&#39;COAD01&#39;,&#39;GLYC05&#39;,&#39;INTE01&#39;,&#39;INTE03&#39;,&#39;BODY04&#39;)

ncase&lt;-c(116404,33640,20352,18381,41969,19099,NA,26676,NA,NA,32858,NA,60801,NA,NA,NA,NA)
ncon&lt;-c(314990,43456,31358,27969,32066,34194,NA,132532,NA,NA,65839,NA,123504,NA,NA,NA,NA)
prev&lt;-c(0.15,0.01,0.015,0.012,0.5,0.05,NA,0.05,NA,NA,0.13,NA,0.03,NA,NA,NA,NA)

des&lt;-NULL
all_res&lt;-NULL
for(target_gwas_i in target_gwas){
  for(secondary_gwas_i in secondary_gwas[secondary_gwas != target_gwas_i]){
    univ_res&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/lava/results/&#39;,target_gwas_i,&#39;/&#39;,target_gwas_i,&#39;.&#39;,secondary_gwas_i,&#39;.univ.lava&#39;))
    bivar_res&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/lava/results/&#39;,target_gwas_i,&#39;/&#39;,target_gwas_i,&#39;.&#39;,secondary_gwas_i,&#39;.bivar.lava&#39;))
    
    # Convert heritability to liability scale
    h2l_R2 &lt;- function(k, r2, p) {
      x= qnorm(1-k)
      z= dnorm(x)
      i=z/k
      C= k*(1-k)*k*(1-k)/(z^2*p*(1-p))
      theta= i*((p-k)/(1-k))*(i*((p-k)/(1-k))-x)
      h2l_R2 = C*r2 / (1 + C*theta*r2)
      return(h2l_R2)
    }
    
    univ_res$h2.liab&lt;-NA
    for(trait in c(target_gwas_i, secondary_gwas_i)){
      if(!is.na(ncase[secondary_gwas == trait])){
        univ_res$h2.liab[univ_res$phen == trait]&lt;-h2l_R2(k=prev[secondary_gwas == trait], r2=univ_res$h2.obs[univ_res$phen == trait], p=ncase[secondary_gwas == trait]/(ncase[secondary_gwas == trait]+ncon[secondary_gwas == trait]))
      } else {
        univ_res$h2.liab[univ_res$phen == trait]&lt;-univ_res$h2.obs[univ_res$phen == trait]
      }
    }
    
    # Calculate SE for h2
    univ_res$z&lt;-abs(qnorm(univ_res$p))
    univ_res$se&lt;-univ_res$h2.liab/univ_res$z
    
    bivar_res&lt;-bivar_res[!is.na(bivar_res$p),]
    
    bivar_res$Direction&lt;-&#39;+&#39;
    bivar_res$Direction[bivar_res$rho &lt; 0]&lt;-&#39;-&#39;
    
    bivar_res$p.fdr&lt;-p.adjust(bivar_res$p, method=&#39;fdr&#39;)
    
    univ_res_wide&lt;-merge(univ_res[univ_res$phen == target_gwas_i,c(&#39;locus&#39;,&#39;chr&#39;,&#39;start&#39;,&#39;stop&#39;,&#39;h2.obs&#39;,&#39;se&#39;,&#39;z&#39;,&#39;p&#39;)],univ_res[univ_res$phen == secondary_gwas_i,c(&#39;locus&#39;,&#39;h2.obs&#39;,&#39;se&#39;,&#39;z&#39;,&#39;p&#39;)], by=&#39;locus&#39;)
      
    tmp&lt;-merge(univ_res_wide, bivar_res[,c(&#39;locus&#39;,&#39;rho&#39;,&#39;p&#39;,&#39;phen1&#39;,&#39;phen2&#39;)], by=&#39;locus&#39;)
    
    all_res&lt;-rbind(all_res, tmp)
    
    des&lt;-rbind(des, data.frame(target=target_gwas_i,
                               secondary=secondary_gwas_i,
                               n_loci=nrow(bivar_res),
                               n_snps=sum(bivar_res$n.snps),
                               n_sig_loci=nrow(bivar_res[bivar_res$p &lt; 0.05,]),
                               n_sig_snps=sum(bivar_res[bivar_res$p &lt; 0.05,]$n.snps),
                               n_sig_loci_pos=nrow(bivar_res[bivar_res$p &lt; 0.05 &amp; bivar_res$Direction == &#39;+&#39;,]),
                               n_sig_loci_neg=nrow(bivar_res[bivar_res$p &lt; 0.05 &amp; bivar_res$Direction == &#39;-&#39;,]),
                               n_sig_fdr_loci=nrow(bivar_res[bivar_res$p.fdr &lt; 0.05,]),
                               n_sig_fdr_snps=sum(bivar_res[bivar_res$p.fdr &lt; 0.05,]$n.snps),
                               n_sig_fdr_loci_pos=nrow(bivar_res[bivar_res$p.fdr &lt; 0.05 &amp; bivar_res$Direction == &#39;+&#39;,]),
                               n_sig_fdr_loci_neg=nrow(bivar_res[bivar_res$p.fdr &lt; 0.05 &amp; bivar_res$Direction == &#39;-&#39;,])))
  }
}

write.table(des, paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/lava/results/lava_descript.csv&#39;), row.names = F, quote=F)
write.table(all_res, paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/lava/results/all_res.csv&#39;), row.names = F, quote=F)</code></pre>
</details>
<hr />
</div>
<div id="generate-polygenic-scores" class="section level2">
<h2><span class="header-section-number">2.4</span> Generate polygenic scores</h2>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code>. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Pipeline_prep.config

gwas=$(echo DEPR06 SCHI02 BIPO02 AUTI07 SMOK04 ADHD05 COLL01 DIAB05 ANXI02 MENA01F OBES01 WAIS01 COAD01 GLYC05 INTE01 INTE03 BODY04)

# Create directory
mkdir -p /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK

# Create file listing GWAS that haven&#39;t been processed.
&gt; /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/todo.txt
for gwas_i in $(echo DEPR06 SCHI02 BIPO02 AUTI07 SMOK04 ADHD05 COLL01 DIAB05 ANXI02 MENA01F OBES01 WAIS01 COAD01 GLYC05 INTE01 INTE03 BODY04);do
if [ ! -f /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/${gwas_i}/1KGPhase3.w_hm3.${gwas_i}.EUR.scale ]; then
echo ${gwas_i} &gt;&gt; /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/todo.txt
fi
done

# Create shell script to run using sbatch
cat &gt; /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/sbatch.sh &lt;&lt; &#39;EOF&#39;
#!/bin/sh

#SBATCH -p shared,brc
#SBATCH --mem 20G
#SBATCH -n 5
#SBATCH --nodes 1
#SBATCH -J LDAK

. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Pipeline_prep.config

gwas=$(awk -v var=&quot;$SLURM_ARRAY_TASK_ID&quot; &#39;NR == var {print $1}&#39; /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/todo.txt)

echo ${gwas}

/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/ldak_mega_prs/ldak_mega_prs.R \
--ref_plink ${Geno_1KG_dir}/1KGPhase3.w_hm3.GW \
--ref_keep ${Geno_1KG_dir}/keep_files/EUR_samples.keep \
--sumstats /users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/${gwas}.cleaned.gz \
--plink1 ${plink1_9} \
--plink2 ${plink2} \
--ldak /users/k1806347/brc_scratch/Software/ldak5.1.linux \
--ldak_map /users/k1806347/brc_scratch/Data/LDAK/genetic_map_b37 \
--ldak_tag /users/k1806347/brc_scratch/Data/LDAK/bld/ \
--ldak_highld /users/k1806347/brc_scratch/Data/LDAK/highld.txt \
--memory 5000 \
--n_cores 5 \
--output /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/${gwas}/1KGPhase3.w_hm3.${gwas} \
--ref_pop_scale ${Geno_1KG_dir}/super_pop_keep.list

EOF

sbatch --array 1-$(wc -l /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/todo.txt | cut -d&#39; &#39; -f1)%6 /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/sbatch.sh
</code></pre>
</details>
<hr />
</div>
<div id="reweight-polygenic-scores" class="section level2">
<h2><span class="header-section-number">2.5</span> Reweight polygenic scores</h2>
<details>
<p><summary>Show code</summary></p>
<pre class="r"><code>library(data.table)

# Read in reference bim for RSID coordinates
ref_bim&lt;-fread(&#39;/users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.GW.bim&#39;)
names(ref_bim)&lt;-c(&#39;CHR&#39;,&#39;SNP&#39;,&#39;POS&#39;,&#39;BP&#39;,&#39;A1&#39;,&#39;A2&#39;)

# Read in the rho estimates
target_gwas&lt;-c(&#39;DEPR06&#39;,&#39;BODY04&#39;,&#39;INTE03&#39;)
secondary_gwas&lt;-c(&#39;DEPR06&#39;,&#39;SCHI02&#39;,&#39;BIPO02&#39;,&#39;AUTI07&#39;,&#39;SMOK04&#39;,&#39;ADHD05&#39;,&#39;COLL01&#39;,&#39;DIAB05&#39;,&#39;ANXI02&#39;,&#39;MENA01F&#39;,&#39;OBES01&#39;,&#39;WAIS01&#39;,&#39;COAD01&#39;,&#39;GLYC05&#39;,&#39;INTE01&#39;,&#39;INTE03&#39;,&#39;BODY04&#39;)

ncase&lt;-c(116404,33640,20352,18381,41969,19099,NA,26676,NA,NA,32858,NA,60801,NA,NA,NA,NA)
ncon&lt;-c(314990,43456,31358,27969,32066,34194,NA,132532,NA,NA,65839,NA,123504,NA,NA,NA,NA)
prev&lt;-c(0.15,0.01,0.015,0.012,0.5,0.05,NA,0.05,NA,NA,0.13,NA,0.03,NA,NA,NA,NA)

univ_res_all&lt;-list()
for(target_gwas_i in target_gwas){
  for(secondary_gwas_i in secondary_gwas[secondary_gwas != target_gwas_i]){
    univ_res&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/lava/results/&#39;,target_gwas_i,&#39;/&#39;,target_gwas_i,&#39;.&#39;,secondary_gwas_i,&#39;.univ.lava&#39;))
    
    # Remove loci with non-significant heritability
    univ_res&lt;-univ_res[univ_res$p &lt; 0.05,]
    
    # Convert heritability to liability scale
    h2l_R2 &lt;- function(k, r2, p) {
      x= qnorm(1-k)
      z= dnorm(x)
      i=z/k
      C= k*(1-k)*k*(1-k)/(z^2*p*(1-p))
      theta= i*((p-k)/(1-k))*(i*((p-k)/(1-k))-x)
      h2l_R2 = C*r2 / (1 + C*theta*r2)
      return(h2l_R2)
    }
    
    univ_res$h2.liab&lt;-NA
    for(trait in c(target_gwas_i, secondary_gwas_i)){
      if(!is.na(ncase[secondary_gwas == trait])){
        univ_res$h2.liab[univ_res$phen == trait]&lt;-h2l_R2(k=prev[secondary_gwas == trait], r2=univ_res$h2.obs[univ_res$phen == trait], p=ncase[secondary_gwas == trait]/(ncase[secondary_gwas == trait]+ncon[secondary_gwas == trait]))
      } else {
        univ_res$h2.liab[univ_res$phen == trait]&lt;-univ_res$h2.obs[univ_res$phen == trait]
      }
    }

    univ_res_all[[paste0(target_gwas_i,&#39;_&#39;,secondary_gwas_i)]]&lt;-univ_res
  }
}

bivar_res_all&lt;-NULL
for(target_gwas_i in target_gwas){
  for(secondary_gwas_i in secondary_gwas[secondary_gwas != target_gwas_i]){
    bivar_res&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/lava/results/&#39;,target_gwas_i,&#39;/&#39;,target_gwas_i,&#39;.&#39;,secondary_gwas_i,&#39;.bivar.lava&#39;))
    bivar_res&lt;-bivar_res[!is.na(bivar_res$p),]
    bivar_res$p.fdr&lt;-p.adjust(bivar_res$p, method=&#39;fdr&#39;)
    bivar_res_all&lt;-rbind(bivar_res_all, bivar_res)
  }
}

# Read in the score files and reweight by rho
for(target_gwas_i in target_gwas){
  for(secondary_gwas_i in secondary_gwas[secondary_gwas != target_gwas_i]){
    
    # Write out the pseudoval model for each GWAS
    pseudo&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&#39;.pseudoval.txt&#39;))
    
    pseudo$V1[pseudo$V2 == max(pseudo$V2)][1]
    
    score&lt;-fread(paste0(&#39;zcat /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.score.gz | cut -d&#39; &#39; -f 1,2,&quot;,as.numeric(gsub(&#39;Score_&#39;,&#39;&#39;,pseudo$V1[pseudo$V2 == max(pseudo$V2)][1]))+2))
    names(score)&lt;-c(&#39;SNP&#39;,&#39;A1&#39;,&#39;SCORE&#39;)
    
    fwrite(score[,c(&#39;SNP&#39;,&#39;A1&#39;,&#39;SCORE&#39;),with=F], paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.score&quot;), quote=F, sep=&#39; &#39;, na=&#39;NA&#39;)
    if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.score.gz&quot;))){
      system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.score.gz&quot;))
    }
    
    system(paste0(&#39;gzip /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.score&quot;))
    
    # Now reweight the score for by rho and h2
    score_bim&lt;-merge(score, ref_bim[,c(&#39;SNP&#39;,&#39;CHR&#39;,&#39;BP&#39;),with=F], by=&#39;SNP&#39;)
    
    univ_res&lt;-univ_res_all[[paste0(target_gwas_i,&#39;_&#39;, secondary_gwas_i)]]
    
    bivar_res&lt;-bivar_res_all[bivar_res_all$phen1 == target_gwas_i &amp; bivar_res_all$phen2 == secondary_gwas_i,]
    
    score_bim_reweight&lt;-NULL
    for(locus in unique(bivar_res$locus)){
      tmp&lt;-score_bim[score_bim$CHR == bivar_res$chr[bivar_res$locus == locus] &amp; score_bim$BP &gt;= bivar_res$start[bivar_res$locus == locus] &amp; score_bim$BP &lt;= bivar_res$stop[bivar_res$locus == locus],]
      tmp$SCORE_rho&lt;-tmp$SCORE*bivar_res$rho[bivar_res$locus == locus]
      tmp$SCORE_rho_h2&lt;-(tmp$SCORE*bivar_res$rho[bivar_res$locus == locus])/(univ_res$h2.liab[univ_res$locus == locus &amp; univ_res$phen == secondary_gwas_i] / univ_res$h2.liab[univ_res$locus == locus &amp; univ_res$phen == target_gwas_i])
      
      # Create score only considering nominally significant rho
      if(bivar_res$p[bivar_res$locus == locus] &lt; 0.05){
        tmp$SCORE_sig&lt;-tmp$SCORE
      } else {
        tmp$SCORE_sig&lt;-0
      }

      if(bivar_res$p[bivar_res$locus == locus] &lt; 0.05){
        tmp$SCORE_rho_sig&lt;-tmp$SCORE_rho
      } else {
        tmp$SCORE_rho_sig&lt;-0
      }
      
      if(bivar_res$p[bivar_res$locus == locus] &lt; 0.05){
        tmp$SCORE_rho_sig_h2&lt;-tmp$SCORE_rho_h2
      } else {
        tmp$SCORE_rho_sig_h2&lt;-0
      }

      # Create score only considering fdr significant rho
      if(bivar_res$p.fdr[bivar_res$locus == locus] &lt; 0.05){
        tmp$SCORE_fdr&lt;-tmp$SCORE
      } else {
        tmp$SCORE_fdr&lt;-0
      }

      if(bivar_res$p.fdr[bivar_res$locus == locus] &lt; 0.05){
        tmp$SCORE_rho_fdr&lt;-tmp$SCORE_rho
      } else {
        tmp$SCORE_rho_fdr&lt;-0
      }
      
      if(bivar_res$p.fdr[bivar_res$locus == locus] &lt; 0.05){
        tmp$SCORE_rho_fdr_h2&lt;-tmp$SCORE_rho_h2
      } else {
        tmp$SCORE_rho_fdr_h2&lt;-0
      }
    
      tmp&lt;-tmp[,c(&#39;SNP&#39;,&#39;A1&#39;,&#39;SCORE&#39;,&#39;SCORE_rho&#39;,&#39;SCORE_rho_h2&#39;,&#39;SCORE_sig&#39;,&#39;SCORE_rho_sig&#39;,&#39;SCORE_rho_sig_h2&#39;,&#39;SCORE_fdr&#39;,&#39;SCORE_rho_fdr&#39;,&#39;SCORE_rho_fdr_h2&#39;), with=F]
      score_bim_reweight&lt;-rbind(score_bim_reweight, tmp)
    }
    
    # Write out sig h2 restricted but unweighted
    fwrite(score_bim_reweight[,c(&#39;SNP&#39;,&#39;A1&#39;,&#39;SCORE&#39;),with=F], paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_unweighted_&quot;,target_gwas_i,&#39;.score&#39;), quote=F, sep=&#39; &#39;, na=&#39;NA&#39;)
    if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_unweighted_&quot;,target_gwas_i,&#39;.score.gz&#39;))){
      system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_unweighted_&quot;,target_gwas_i,&#39;.score.gz&#39;))
    }
    
    system(paste0(&#39;gzip /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_unweighted_&quot;,target_gwas_i,&#39;.score&#39;))

    # Write out sig rho restricted but unweighted
    fwrite(score_bim_reweight[,c(&#39;SNP&#39;,&#39;A1&#39;,&#39;SCORE_sig&#39;),with=F], paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_unweighted_&quot;,target_gwas_i,&#39;.score&#39;), quote=F, sep=&#39; &#39;, na=&#39;NA&#39;)
    if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_unweighted_&quot;,target_gwas_i,&#39;.score.gz&#39;))){
      system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_unweighted_&quot;,target_gwas_i,&#39;.score.gz&#39;))
    }
    
    system(paste0(&#39;gzip /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_unweighted_&quot;,target_gwas_i,&#39;.score&#39;))

    # Write out fdr sig rho restricted but unweighted
    fwrite(score_bim_reweight[,c(&#39;SNP&#39;,&#39;A1&#39;,&#39;SCORE_fdr&#39;),with=F], paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_unweighted_&quot;,target_gwas_i,&#39;.score&#39;), quote=F, sep=&#39; &#39;, na=&#39;NA&#39;)
    if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_unweighted_&quot;,target_gwas_i,&#39;.score.gz&#39;))){
      system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_unweighted_&quot;,target_gwas_i,&#39;.score.gz&#39;))
    }
    
    system(paste0(&#39;gzip /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_unweighted_&quot;,target_gwas_i,&#39;.score&#39;))

    # Write out sig h2 restricted and rho weighted
    fwrite(score_bim_reweight[,c(&#39;SNP&#39;,&#39;A1&#39;,&#39;SCORE_rho&#39;),with=F], paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_rho_weighted_&quot;,target_gwas_i,&#39;.score&#39;), quote=F, sep=&#39; &#39;, na=&#39;NA&#39;)
    if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_rho_weighted_&quot;,target_gwas_i,&#39;.score.gz&#39;))){
      system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_rho_weighted_&quot;,target_gwas_i,&#39;.score.gz&#39;))
    }
    
    system(paste0(&#39;gzip /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_rho_weighted_&quot;,target_gwas_i,&#39;.score&#39;))

    # Write out sig rho restricted and rho weighted
    fwrite(score_bim_reweight[,c(&#39;SNP&#39;,&#39;A1&#39;,&#39;SCORE_rho_sig&#39;),with=F], paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_rho_weighted_&quot;,target_gwas_i,&#39;.score&#39;), quote=F, sep=&#39; &#39;, na=&#39;NA&#39;)
    if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_rho_weighted_&quot;,target_gwas_i,&#39;.score.gz&#39;))){
      system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_rho_weighted_&quot;,target_gwas_i,&#39;.score.gz&#39;))
    }
    
    system(paste0(&#39;gzip /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_rho_weighted_&quot;,target_gwas_i,&#39;.score&#39;))
    
    # Write out fdr sig rho restricted and rho weighted
    fwrite(score_bim_reweight[,c(&#39;SNP&#39;,&#39;A1&#39;,&#39;SCORE_rho_fdr&#39;),with=F], paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_rho_weighted_&quot;,target_gwas_i,&#39;.score&#39;), quote=F, sep=&#39; &#39;, na=&#39;NA&#39;)
    if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_rho_weighted_&quot;,target_gwas_i,&#39;.score.gz&#39;))){
      system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_rho_weighted_&quot;,target_gwas_i,&#39;.score.gz&#39;))
    }
    
    system(paste0(&#39;gzip /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_rho_weighted_&quot;,target_gwas_i,&#39;.score&#39;))

    # Write out sig h2 restricted and rho and h2 weighted
    fwrite(score_bim_reweight[,c(&#39;SNP&#39;,&#39;A1&#39;,&#39;SCORE_rho_h2&#39;),with=F], paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_rho_h2_weighted_&quot;,target_gwas_i,&#39;.score&#39;), quote=F, sep=&#39; &#39;, na=&#39;NA&#39;)
    if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_rho_h2_weighted_&quot;,target_gwas_i,&#39;.score.gz&#39;))){
      system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_rho_h2_weighted_&quot;,target_gwas_i,&#39;.score.gz&#39;))
    }
    
    system(paste0(&#39;gzip /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_rho_h2_weighted_&quot;,target_gwas_i,&#39;.score&#39;))
    
    # Write out sig rho restricted and rho and h2 weighted
    fwrite(score_bim_reweight[,c(&#39;SNP&#39;,&#39;A1&#39;,&#39;SCORE_rho_sig_h2&#39;),with=F], paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_rho_h2_weighted_&quot;,target_gwas_i,&#39;.score&#39;), quote=F, sep=&#39; &#39;, na=&#39;NA&#39;)
    if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_rho_h2_weighted_&quot;,target_gwas_i,&#39;.score.gz&#39;))){
      system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_rho_h2_weighted_&quot;,target_gwas_i,&#39;.score.gz&#39;))
    }
    
    system(paste0(&#39;gzip /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_rho_h2_weighted_&quot;,target_gwas_i,&#39;.score&#39;))

    # Write out fdr rho restricted and rho and h2 weighted
    fwrite(score_bim_reweight[,c(&#39;SNP&#39;,&#39;A1&#39;,&#39;SCORE_rho_fdr_h2&#39;),with=F], paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_rho_h2_weighted_&quot;,target_gwas_i,&#39;.score&#39;), quote=F, sep=&#39; &#39;, na=&#39;NA&#39;)
    if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_rho_h2_weighted_&quot;,target_gwas_i,&#39;.score.gz&#39;))){
      system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_rho_h2_weighted_&quot;,target_gwas_i,&#39;.score.gz&#39;))
    }
    
    system(paste0(&#39;gzip /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_rho_h2_weighted_&quot;,target_gwas_i,&#39;.score&#39;))

    # Combine reweighted PGS with full PGS
    score_full_reweight&lt;-score
    for(type in names(score_bim_reweight)[grepl(&#39;SCORE&#39;,names(score_bim_reweight)) &amp; grepl(&#39;rho&#39;,names(score_bim_reweight))]){
     tmp&lt;-merge(score, score_bim_reweight[,c(&#39;SNP&#39;,type), with=F], by=&#39;SNP&#39;, all=T)
     tmp[[type]][which(tmp[[type]] == 0 | is.na(tmp[[type]]))]&lt;-tmp[[&#39;SCORE&#39;]][which(tmp[[type]] == 0 | is.na(tmp[[type]]))]
     tmp&lt;-tmp[match(score_full_reweight$SNP, tmp$SNP),]
     score_full_reweight[[type]]&lt;-tmp[[type]]
    }
    
    # Write out sig h2 restricted and rho weighted with unadjusted SNPs
    fwrite(score_full_reweight[,c(&#39;SNP&#39;,&#39;A1&#39;,&#39;SCORE_rho&#39;),with=F], paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_rho_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score&#39;), quote=F, sep=&#39; &#39;, na=&#39;NA&#39;)
    if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_rho_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score.gz&#39;))){
      system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_rho_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score.gz&#39;))
    }
    
    system(paste0(&#39;gzip /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_rho_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score&#39;))

    # Write out sig rho restricted and rho weighted with unadjusted SNPs
    fwrite(score_full_reweight[,c(&#39;SNP&#39;,&#39;A1&#39;,&#39;SCORE_rho_sig&#39;),with=F], paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_rho_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score&#39;), quote=F, sep=&#39; &#39;, na=&#39;NA&#39;)
    if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_rho_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score.gz&#39;))){
      system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_rho_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score.gz&#39;))
    }
    
    system(paste0(&#39;gzip /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_rho_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score&#39;))
    
    # Write out fdr sig rho restricted and rho weighted with unadjusted SNPs
    fwrite(score_full_reweight[,c(&#39;SNP&#39;,&#39;A1&#39;,&#39;SCORE_rho_fdr&#39;),with=F], paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_rho_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score&#39;), quote=F, sep=&#39; &#39;, na=&#39;NA&#39;)
    if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_rho_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score.gz&#39;))){
      system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_rho_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score.gz&#39;))
    }
    
    system(paste0(&#39;gzip /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_rho_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score&#39;))

    # Write out sig h2 restricted and rho and h2 weighted with unadjusted SNPs
    fwrite(score_full_reweight[,c(&#39;SNP&#39;,&#39;A1&#39;,&#39;SCORE_rho_h2&#39;),with=F], paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_rho_h2_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score&#39;), quote=F, sep=&#39; &#39;, na=&#39;NA&#39;)
    if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_rho_h2_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score.gz&#39;))){
      system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_rho_h2_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score.gz&#39;))
    }
    
    system(paste0(&#39;gzip /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_rho_h2_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score&#39;))
    
    # Write out sig rho restricted and rho and h2 weighted with unadjusted SNPs
    fwrite(score_full_reweight[,c(&#39;SNP&#39;,&#39;A1&#39;,&#39;SCORE_rho_sig_h2&#39;),with=F], paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_rho_h2_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score&#39;), quote=F, sep=&#39; &#39;, na=&#39;NA&#39;)
    if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_rho_h2_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score.gz&#39;))){
      system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_rho_h2_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score.gz&#39;))
    }
    
    system(paste0(&#39;gzip /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_rho_h2_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score&#39;))

    # Write out fdr rho restricted and rho and h2 weighted with unadjusted SNPs
    fwrite(score_full_reweight[,c(&#39;SNP&#39;,&#39;A1&#39;,&#39;SCORE_rho_fdr_h2&#39;),with=F], paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_rho_h2_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score&#39;), quote=F, sep=&#39; &#39;, na=&#39;NA&#39;)
    if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_rho_h2_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score.gz&#39;))){
      system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_rho_h2_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score.gz&#39;))
    }
    
    system(paste0(&#39;gzip /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_rho_h2_weighted_with_unadjusted_&quot;,target_gwas_i,&#39;.score&#39;))

    # Make score files for SNPs that aren&#39;t reweighted
    score_unadjust&lt;-score
    for(type in names(score_bim_reweight)[grepl(&#39;SCORE&#39;,names(score_bim_reweight)) &amp; !(grepl(&#39;rho&#39;,names(score_bim_reweight)))]){
      tmp&lt;-score$SCORE
      tmp[score$SNP %in% score_bim_reweight$SNP[which(score_bim_reweight[[type]] != 0)]]&lt;-0
      
      score_unadjust[[paste0(type,&#39;_remainder&#39;)]]&lt;-tmp
    }
    
    # Write out unadjusted for SNPs not in sig h2 loci
    fwrite(score_unadjust[,c(&#39;SNP&#39;,&#39;A1&#39;,&#39;SCORE_remainder&#39;),with=F], paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_remainder_&quot;,target_gwas_i,&#39;.score&#39;), quote=F, sep=&#39; &#39;, na=&#39;NA&#39;)
    if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_remainder_&quot;,target_gwas_i,&#39;.score.gz&#39;))){
      system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_remainder_&quot;,target_gwas_i,&#39;.score.gz&#39;))
    }
    
    system(paste0(&#39;gzip /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.h2_restricted_remainder_&quot;,target_gwas_i,&#39;.score&#39;))

    # Write out unadjusted for SNPs not in sig rho loci
    fwrite(score_unadjust[,c(&#39;SNP&#39;,&#39;A1&#39;,&#39;SCORE_sig_remainder&#39;),with=F], paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_remainder_&quot;,target_gwas_i,&#39;.score&#39;), quote=F, sep=&#39; &#39;, na=&#39;NA&#39;)
    if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_remainder_&quot;,target_gwas_i,&#39;.score.gz&#39;))){
      system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_remainder_&quot;,target_gwas_i,&#39;.score.gz&#39;))
    }
    
    system(paste0(&#39;gzip /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_restricted_remainder_&quot;,target_gwas_i,&#39;.score&#39;))

    # Write out unadjusted for SNPs not in fdr sig rho loci
    fwrite(score_unadjust[,c(&#39;SNP&#39;,&#39;A1&#39;,&#39;SCORE_fdr_remainder&#39;),with=F], paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_remainder_&quot;,target_gwas_i,&#39;.score&#39;), quote=F, sep=&#39; &#39;, na=&#39;NA&#39;)
    if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_remainder_&quot;,target_gwas_i,&#39;.score.gz&#39;))){
      system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_remainder_&quot;,target_gwas_i,&#39;.score.gz&#39;))
    }
    
    system(paste0(&#39;gzip /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&quot;.pseudo.rho_fdr_restricted_remainder_&quot;,target_gwas_i,&#39;.score&#39;))

  }
}

# Process new score files with external_scorer script to create scale files
# Read in the rho estimates
target_gwas&lt;-c(&#39;DEPR06&#39;,&#39;BODY04&#39;,&#39;INTE03&#39;)
secondary_gwas&lt;-c(&#39;DEPR06&#39;,&#39;SCHI02&#39;,&#39;BIPO02&#39;,&#39;AUTI07&#39;,&#39;SMOK04&#39;,&#39;ADHD05&#39;,&#39;COLL01&#39;,&#39;DIAB05&#39;,&#39;ANXI02&#39;,&#39;MENA01F&#39;,&#39;OBES01&#39;,&#39;WAIS01&#39;,&#39;COAD01&#39;,&#39;GLYC05&#39;,&#39;INTE01&#39;,&#39;INTE03&#39;,&#39;BODY04&#39;)

for(target_gwas_i in target_gwas){
  for(secondary_gwas_i in secondary_gwas[secondary_gwas != target_gwas_i]){
    for(type in c(&#39;rho_restricted_unweighted&#39;,&#39;rho_fdr_restricted_unweighted&#39;,&#39;h2_restricted_unweighted&#39;,&#39;rho_restricted_rho_weighted&#39;,&#39;rho_fdr_restricted_rho_weighted&#39;,&#39;h2_restricted_rho_weighted&#39;,&#39;rho_restricted_rho_h2_weighted&#39;,&#39;rho_fdr_restricted_rho_h2_weighted&#39;,&#39;h2_restricted_rho_h2_weighted&#39;,&#39;rho_restricted_rho_weighted_with_unadjusted&#39;,&#39;rho_fdr_restricted_rho_weighted_with_unadjusted&#39;,&#39;h2_restricted_rho_weighted_with_unadjusted&#39;,&#39;rho_restricted_rho_h2_weighted_with_unadjusted&#39;,&#39;rho_fdr_restricted_rho_h2_weighted_with_unadjusted&#39;,&#39;h2_restricted_rho_h2_weighted_with_unadjusted&#39;,&#39;rho_restricted_remainder&#39;,&#39;rho_fdr_restricted_remainder&#39;,&#39;h2_restricted_remainder&#39;)){
      dir.create(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/processed/&#39;,secondary_gwas_i,&#39;.pseudo.&#39;,type,&#39;_&#39;,target_gwas_i), recursive=T)
      system(paste0(&#39;sbatch -p brc,shared --mem 10G  /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/external_score_processor/external_score_processor_plink2.R --ref_plink_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr --score /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&#39;.pseudo.&#39;,type,&#39;_&#39;,target_gwas_i,&#39;.score.gz --plink2 /users/k1806347/brc_scratch/Software/plink2 --output /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/processed/&#39;,secondary_gwas_i,&#39;.pseudo.&#39;,type,&#39;_&#39;,target_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&#39;.pseudo.&#39;,type,&#39;_&#39;,target_gwas_i,&#39; --ref_pop_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/super_pop_keep.list&#39;))
    }
  }
}

for(secondary_gwas_i in secondary_gwas){
  dir.create(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/processed/&#39;,secondary_gwas_i,&#39;.pseudo&#39;), recursive=T)
  system(paste0(&#39;/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/external_score_processor/external_score_processor_plink2.R --ref_plink_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr --score /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&#39;.pseudo.score.gz --plink2 /users/k1806347/brc_scratch/Software/plink2 --output /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/processed/&#39;,secondary_gwas_i,&#39;.pseudo/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&#39;.pseudo --ref_pop_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/super_pop_keep.list&#39;))
}</code></pre>
</details>
<hr />
</div>
<div id="perform-polygenic-scoring" class="section level2">
<h2><span class="header-section-number">2.6</span> Perform polygenic scoring</h2>
<details>
<p><summary>Show code</summary></p>
<pre class="r"><code># Read in the rho estimates
target_gwas&lt;-c(&#39;DEPR06&#39;,&#39;BODY04&#39;,&#39;INTE03&#39;)
target_pheno&lt;-c(&#39;Depression&#39;,&#39;BMI&#39;,&#39;Intelligence&#39;)
secondary_gwas&lt;-c(&#39;DEPR06&#39;,&#39;SCHI02&#39;,&#39;BIPO02&#39;,&#39;AUTI07&#39;,&#39;SMOK04&#39;,&#39;ADHD05&#39;,&#39;COLL01&#39;,&#39;DIAB05&#39;,&#39;ANXI02&#39;,&#39;MENA01F&#39;,&#39;OBES01&#39;,&#39;WAIS01&#39;,&#39;COAD01&#39;,&#39;GLYC05&#39;,&#39;INTE01&#39;,&#39;INTE03&#39;,&#39;BODY04&#39;)

for(i in 1:length(target_gwas)){
  target_gwas_i&lt;-target_gwas[i]
  target_pheno_i&lt;-target_pheno[i]
  for(secondary_gwas_i in secondary_gwas[secondary_gwas != target_gwas_i]){
    for(type in c(&#39;rho_restricted_unweighted&#39;,&#39;rho_fdr_restricted_unweighted&#39;,&#39;h2_restricted_unweighted&#39;,&#39;rho_restricted_rho_weighted&#39;,&#39;rho_fdr_restricted_rho_weighted&#39;,&#39;h2_restricted_rho_weighted&#39;,&#39;rho_restricted_rho_h2_weighted&#39;,&#39;rho_fdr_restricted_rho_h2_weighted&#39;,&#39;h2_restricted_rho_h2_weighted&#39;,&#39;rho_restricted_rho_weighted_with_unadjusted&#39;,&#39;rho_fdr_restricted_rho_weighted_with_unadjusted&#39;,&#39;h2_restricted_rho_weighted_with_unadjusted&#39;,&#39;rho_restricted_rho_h2_weighted_with_unadjusted&#39;,&#39;rho_fdr_restricted_rho_h2_weighted_with_unadjusted&#39;,&#39;h2_restricted_rho_h2_weighted_with_unadjusted&#39;,&#39;rho_restricted_remainder&#39;,&#39;rho_fdr_restricted_remainder&#39;,&#39;h2_restricted_remainder&#39;)){
      if(!file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,secondary_gwas_i,&#39;/&#39;,target_gwas_i,&#39;/&#39;,type,&#39;/prs.profiles&#39;))){
        system(paste0(&#39;sbatch -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer/Scaled_polygenic_scorer_plink2.R --target_plink_chr /users/k1806347/brc_scratch/Data/UKBB/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr --target_keep /users/k1806347/brc_scratch/Data/UKBB/Phenotype/PRS_comp_subset/UKBB.&#39;,target_pheno_i,&#39;.txt --ref_score /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&#39;.pseudo.&#39;,type,&#39;_&#39;,target_gwas_i,&#39;.score.gz --ref_scale /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/processed/&#39;,secondary_gwas_i,&#39;.pseudo.&#39;,type,&#39;_&#39;,target_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&#39;.pseudo.&#39;,type,&#39;_&#39;,target_gwas_i,&#39;.EUR.scale --ref_freq_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/freq_files/EUR/1KGPhase3.w_hm3.EUR.chr --plink2 /users/k1806347/brc_scratch/Software/plink2_alpha/plink2 --output /users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,secondary_gwas_i,&#39;/&#39;,target_gwas_i,&#39;/&#39;,type,&#39;/prs&#39;))
      }
    }
    Sys.sleep(900)
  }
}

for(secondary_gwas_i in secondary_gwas){
  system(paste0(&#39;sbatch -p brc,shared --mem 20G /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer/Scaled_polygenic_scorer_plink2.R --target_plink_chr /users/k1806347/brc_scratch/Data/UKBB/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr --target_keep /users/k1806347/brc_scratch/Data/UKBB/Projected_PCs/Ancestry_idenitfier/UKBB.w_hm3.AllAncestry.EUR.keep --ref_score /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&#39;.pseudo.score.gz --ref_scale /users/k1806347/brc_scratch/Analyses/local_rg_pgs/score_files/LDAK/&#39;,secondary_gwas_i,&#39;/processed/&#39;,secondary_gwas_i,&#39;.pseudo/1KGPhase3.w_hm3.&#39;,secondary_gwas_i,&#39;.pseudo.EUR.scale --ref_freq_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/freq_files/EUR/1KGPhase3.w_hm3.EUR.chr --plink2 /users/k1806347/brc_scratch/Software/plink2_alpha/plink2 --output /users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,secondary_gwas_i,&#39;/prs&#39;))
}</code></pre>
</details>
<hr />
</div>
<div id="evaluate-polygenic-scores" class="section level2">
<h2><span class="header-section-number">2.7</span> Evaluate polygenic scores</h2>
<details>
<p><summary>Show code</summary></p>
<pre class="r"><code>############
# First test whether inclusion of secondary PRS improve prediction over target phenotype gwas (i.e. replicate Krapohl et al.)
############

target_gwas&lt;-c(&#39;DEPR06&#39;,&#39;BODY04&#39;,&#39;INTE03&#39;)
target_pheno&lt;-c(&#39;Depression&#39;,&#39;BMI&#39;,&#39;Intelligence&#39;)
target_pop_prev&lt;-c(0.15,NA,NA)
secondary_gwas&lt;-c(&#39;DEPR06&#39;,&#39;SCHI02&#39;,&#39;BIPO02&#39;,&#39;AUTI07&#39;,&#39;SMOK04&#39;,&#39;ADHD05&#39;,&#39;COLL01&#39;,&#39;DIAB05&#39;,&#39;ANXI02&#39;,&#39;MENA01F&#39;,&#39;OBES01&#39;,&#39;WAIS01&#39;,&#39;COAD01&#39;,&#39;GLYC05&#39;,&#39;INTE01&#39;,&#39;INTE03&#39;,&#39;BODY04&#39;)

# use for loop to read all values and indexes
for(i in 1:length(target_gwas)){
  pred_file&lt;-NULL
  target_gwas_i&lt;-target_gwas[i]
  target_pheno_i&lt;-target_pheno[i]
  target_pop_prev_i&lt;-target_pop_prev[i]
  # Add in prs for target phenotype
  tmp&lt;-data.frame(predictors=paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,target_gwas_i,&#39;/prs.profiles&#39;),
                  group=&#39;target&#39;)
  
  pred_file&lt;-rbind(pred_file, tmp)
  
  for(secondary_gwas_i in secondary_gwas[secondary_gwas != target_gwas_i]){
    # Add in prs for target phenotype
    tmp&lt;-data.frame(predictors=paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,secondary_gwas_i,&#39;/prs.profiles&#39;),
                    group=&#39;secondary&#39;)
    
    pred_file&lt;-rbind(pred_file, tmp)
  }
  
  dir.create(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i), recursive=T)
  write.table(pred_file, paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_1.txt&#39;), col.names=T, row.names=F, quote=F)

  system(paste0(&#39;sbatch --mem 10G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R --pheno /users/k1806347/brc_scratch/Data/UKBB/Phenotype/PRS_comp_subset/UKBB.&#39;,target_pheno_i,&#39;.txt --out /users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_1 --assoc T --outcome_pop_prev &#39;,target_pop_prev_i,&#39; --predictors /users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_1.txt&#39;))

}

############
# Compare prs restricted by h2 and rho significance, and unweighted and weighted by rho, and rho and h2
############

target_gwas&lt;-c(&#39;DEPR06&#39;,&#39;BODY04&#39;,&#39;INTE03&#39;)
target_pheno&lt;-c(&#39;Depression&#39;,&#39;BMI&#39;,&#39;Intelligence&#39;)
target_pop_prev&lt;-c(0.15,NA,NA)
secondary_gwas&lt;-c(&#39;DEPR06&#39;,&#39;SCHI02&#39;,&#39;BIPO02&#39;,&#39;AUTI07&#39;,&#39;SMOK04&#39;,&#39;ADHD05&#39;,&#39;COLL01&#39;,&#39;DIAB05&#39;,&#39;ANXI02&#39;,&#39;MENA01F&#39;,&#39;OBES01&#39;,&#39;WAIS01&#39;,&#39;COAD01&#39;,&#39;GLYC05&#39;,&#39;INTE01&#39;,&#39;INTE03&#39;,&#39;BODY04&#39;)

# use for loop to read all values and indexes
for(i in 1:length(target_gwas)){
  pred_file&lt;-NULL
  target_gwas_i&lt;-target_gwas[i]
  target_pheno_i&lt;-target_pheno[i]
  target_pop_prev_i&lt;-target_pop_prev[i]
  for(secondary_gwas_i in secondary_gwas[secondary_gwas != target_gwas_i]){
    for(type in c(&#39;rho_restricted_unweighted&#39;,&#39;rho_fdr_restricted_unweighted&#39;,&#39;h2_restricted_unweighted&#39;,&#39;rho_restricted_rho_weighted&#39;,&#39;rho_fdr_restricted_rho_weighted&#39;,&#39;h2_restricted_rho_weighted&#39;,&#39;rho_restricted_rho_h2_weighted&#39;,&#39;rho_fdr_restricted_rho_h2_weighted&#39;,&#39;h2_restricted_rho_h2_weighted&#39;)){
      # Add in prs for target phenotype
      tmp&lt;-data.frame(predictors=paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,secondary_gwas_i,&#39;/&#39;,target_gwas_i,&#39;/&#39;,type,&#39;/prs.profiles&#39;),
                      group=type)
      
      pred_file&lt;-rbind(pred_file, tmp)
    }
  }

  write.table(pred_file, paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_2.txt&#39;), col.names=T, row.names=F, quote=F)

  system(paste0(&#39;sbatch --mem 10G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R --pheno /users/k1806347/brc_scratch/Data/UKBB/Phenotype/PRS_comp_subset/UKBB.&#39;,target_pheno_i,&#39;.txt --out /users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_2 --assoc T --outcome_pop_prev &#39;,target_pop_prev_i,&#39; --predictors /users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_2.txt&#39;))

}

############
# Test whether adding weighted PRS improves prediction over the standard approach (i.e. Krapohl et al.) 
############

target_gwas&lt;-c(&#39;DEPR06&#39;,&#39;BODY04&#39;,&#39;INTE03&#39;)
target_pheno&lt;-c(&#39;Depression&#39;,&#39;BMI&#39;,&#39;Intelligence&#39;)
target_pop_prev&lt;-c(0.15,NA,NA)
secondary_gwas&lt;-c(&#39;DEPR06&#39;,&#39;SCHI02&#39;,&#39;BIPO02&#39;,&#39;AUTI07&#39;,&#39;SMOK04&#39;,&#39;ADHD05&#39;,&#39;COLL01&#39;,&#39;DIAB05&#39;,&#39;ANXI02&#39;,&#39;MENA01F&#39;,&#39;OBES01&#39;,&#39;WAIS01&#39;,&#39;COAD01&#39;,&#39;GLYC05&#39;,&#39;INTE01&#39;,&#39;INTE03&#39;,&#39;BODY04&#39;)

# use for loop to read all values and indexes
for(i in 1:length(target_gwas)){
  for(type in c(&#39;rho_restricted_unweighted&#39;,&#39;rho_fdr_restricted_unweighted&#39;,&#39;h2_restricted_unweighted&#39;,&#39;rho_restricted_rho_weighted&#39;,&#39;rho_fdr_restricted_rho_weighted&#39;,&#39;h2_restricted_rho_weighted&#39;,&#39;rho_restricted_rho_h2_weighted&#39;,&#39;rho_fdr_restricted_rho_h2_weighted&#39;,&#39;h2_restricted_rho_h2_weighted&#39;)){
    pred_file&lt;-NULL
    target_gwas_i&lt;-target_gwas[i]
    target_pheno_i&lt;-target_pheno[i]
    target_pop_prev_i&lt;-target_pop_prev[i]
    
    # Add in prs for target phenotype
    tmp&lt;-data.frame(predictors=paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,target_gwas_i,&#39;/prs.profiles&#39;),
                    group=&#39;baseline&#39;)
    
    pred_file&lt;-rbind(pred_file, tmp)
  
    for(secondary_gwas_i in secondary_gwas[secondary_gwas != target_gwas_i]){
      # Add in prs for target phenotype
      tmp&lt;-data.frame(predictors=paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,secondary_gwas_i,&#39;/prs.profiles&#39;),
                      group=&#39;baseline&#39;)
      
      pred_file&lt;-rbind(pred_file, tmp)
      
      # Add in weighted prs for target phenotype
      tmp&lt;-data.frame(predictors=paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,secondary_gwas_i,&#39;/&#39;,target_gwas_i,&#39;/&#39;,type,&#39;/prs.profiles&#39;),
                      group=type)
      
      pred_file&lt;-rbind(pred_file, tmp)
    }
  
    write.table(pred_file, paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_3_&#39;,type,&#39;.txt&#39;), col.names=T, row.names=F, quote=F)
  
    system(paste0(&#39;sbatch --mem 10G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R --pheno /users/k1806347/brc_scratch/Data/UKBB/Phenotype/PRS_comp_subset/UKBB.&#39;,target_pheno_i,&#39;.txt --out /users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_3_&#39;,type,&#39; --assoc T --outcome_pop_prev &#39;,target_pop_prev_i,&#39; --predictors /users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_3_&#39;,type,&#39;.txt&#39;))
  
  }
}

############
# Compare unadjusted PRS to reweighted PRS with unadjusted SNPs 
############

target_gwas&lt;-c(&#39;DEPR06&#39;,&#39;BODY04&#39;,&#39;INTE03&#39;)
target_pheno&lt;-c(&#39;Depression&#39;,&#39;BMI&#39;,&#39;Intelligence&#39;)
target_pop_prev&lt;-c(0.15,NA,NA)
secondary_gwas&lt;-c(&#39;DEPR06&#39;,&#39;SCHI02&#39;,&#39;BIPO02&#39;,&#39;AUTI07&#39;,&#39;SMOK04&#39;,&#39;ADHD05&#39;,&#39;COLL01&#39;,&#39;DIAB05&#39;,&#39;ANXI02&#39;,&#39;MENA01F&#39;,&#39;OBES01&#39;,&#39;WAIS01&#39;,&#39;COAD01&#39;,&#39;GLYC05&#39;,&#39;INTE01&#39;,&#39;INTE03&#39;,&#39;BODY04&#39;)

# use for loop to read all values and indexes
for(i in 1:length(target_gwas)){
    for(type in c(&#39;rho_restricted_rho_weighted_with_unadjusted&#39;,&#39;rho_fdr_restricted_rho_weighted_with_unadjusted&#39;,&#39;h2_restricted_rho_weighted_with_unadjusted&#39;,&#39;rho_restricted_rho_h2_weighted_with_unadjusted&#39;,&#39;rho_fdr_restricted_rho_h2_weighted_with_unadjusted&#39;,&#39;h2_restricted_rho_h2_weighted_with_unadjusted&#39;)){
    pred_file&lt;-NULL
    target_gwas_i&lt;-target_gwas[i]
    target_pheno_i&lt;-target_pheno[i]
    target_pop_prev_i&lt;-target_pop_prev[i]
    
    for(secondary_gwas_i in secondary_gwas[secondary_gwas != target_gwas_i]){
      # Add in prs for target phenotype
      tmp&lt;-data.frame(predictors=paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,secondary_gwas_i,&#39;/prs.profiles&#39;),
                      group=&#39;standard&#39;)
      
      pred_file&lt;-rbind(pred_file, tmp)
      
      # Add in weighted prs for target phenotype
      tmp&lt;-data.frame(predictors=paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,secondary_gwas_i,&#39;/&#39;,target_gwas_i,&#39;/&#39;,type,&#39;/prs.profiles&#39;),
                      group=type)
      
      pred_file&lt;-rbind(pred_file, tmp)
    }
  
    write.table(pred_file, paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_4_&#39;,type,&#39;.txt&#39;), col.names=T, row.names=F, quote=F)
  
    system(paste0(&#39;sbatch --mem 10G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R --pheno /users/k1806347/brc_scratch/Data/UKBB/Phenotype/PRS_comp_subset/UKBB.&#39;,target_pheno_i,&#39;.txt --out /users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_4_&#39;,type,&#39; --assoc T --outcome_pop_prev &#39;,target_pop_prev_i,&#39; --predictors /users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_4_&#39;,type,&#39;.txt&#39;))
  
  }
}

############
# Compare unadjusted PRS with target PRS to reweighted PRS with unadjusted SNPs and target PRS
############

target_gwas&lt;-c(&#39;DEPR06&#39;,&#39;BODY04&#39;,&#39;INTE03&#39;)
target_pheno&lt;-c(&#39;Depression&#39;,&#39;BMI&#39;,&#39;Intelligence&#39;)
target_pop_prev&lt;-c(0.15,NA,NA)
secondary_gwas&lt;-c(&#39;DEPR06&#39;,&#39;SCHI02&#39;,&#39;BIPO02&#39;,&#39;AUTI07&#39;,&#39;SMOK04&#39;,&#39;ADHD05&#39;,&#39;COLL01&#39;,&#39;DIAB05&#39;,&#39;ANXI02&#39;,&#39;MENA01F&#39;,&#39;OBES01&#39;,&#39;WAIS01&#39;,&#39;COAD01&#39;,&#39;GLYC05&#39;,&#39;INTE01&#39;,&#39;INTE03&#39;,&#39;BODY04&#39;)

# use for loop to read all values and indexes
for(i in 1:length(target_gwas)){
    for(type in c(&#39;rho_restricted_rho_weighted_with_unadjusted&#39;,&#39;rho_fdr_restricted_rho_weighted_with_unadjusted&#39;,&#39;h2_restricted_rho_weighted_with_unadjusted&#39;,&#39;rho_restricted_rho_h2_weighted_with_unadjusted&#39;,&#39;rho_fdr_restricted_rho_h2_weighted_with_unadjusted&#39;,&#39;h2_restricted_rho_h2_weighted_with_unadjusted&#39;)){
    pred_file&lt;-NULL
    target_gwas_i&lt;-target_gwas[i]
    target_pheno_i&lt;-target_pheno[i]
    target_pop_prev_i&lt;-target_pop_prev[i]
    
    # Add in prs for target phenotype
    tmp&lt;-data.frame(predictors=paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,target_gwas_i,&#39;/prs.profiles&#39;),
                    group=&#39;baseline&#39;)
    
    pred_file&lt;-rbind(pred_file, tmp)
  
    for(secondary_gwas_i in secondary_gwas[secondary_gwas != target_gwas_i]){
      # Add in prs for target phenotype
      tmp&lt;-data.frame(predictors=paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,secondary_gwas_i,&#39;/prs.profiles&#39;),
                      group=&#39;baseline&#39;)
      
      pred_file&lt;-rbind(pred_file, tmp)
      
      # Add in weighted prs for target phenotype
      tmp&lt;-data.frame(predictors=paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,secondary_gwas_i,&#39;/&#39;,target_gwas_i,&#39;/&#39;,type,&#39;/prs.profiles&#39;),
                      group=type)
      
      pred_file&lt;-rbind(pred_file, tmp)
    }
  
    write.table(pred_file, paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_5_&#39;,type,&#39;.txt&#39;), col.names=T, row.names=F, quote=F)
  
    system(paste0(&#39;sbatch --mem 10G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R --pheno /users/k1806347/brc_scratch/Data/UKBB/Phenotype/PRS_comp_subset/UKBB.&#39;,target_pheno_i,&#39;.txt --out /users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_5_&#39;,type,&#39; --assoc T --outcome_pop_prev &#39;,target_pop_prev_i,&#39; --predictors /users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_5_&#39;,type,&#39;.txt&#39;))
  
  }
}

############
# Compare secondary PGS weighted and the remainder (seperately) to unadjusted secondary PGS
############

target_gwas&lt;-c(&#39;DEPR06&#39;,&#39;BODY04&#39;,&#39;INTE03&#39;)
target_pheno&lt;-c(&#39;Depression&#39;,&#39;BMI&#39;,&#39;Intelligence&#39;)
target_pop_prev&lt;-c(0.15,NA,NA)
secondary_gwas&lt;-c(&#39;DEPR06&#39;,&#39;SCHI02&#39;,&#39;BIPO02&#39;,&#39;AUTI07&#39;,&#39;SMOK04&#39;,&#39;ADHD05&#39;,&#39;COLL01&#39;,&#39;DIAB05&#39;,&#39;ANXI02&#39;,&#39;MENA01F&#39;,&#39;OBES01&#39;,&#39;WAIS01&#39;,&#39;COAD01&#39;,&#39;GLYC05&#39;,&#39;INTE01&#39;,&#39;INTE03&#39;,&#39;BODY04&#39;)

# use for loop to read all values and indexes
for(i in 1:length(target_gwas)){
  for(type in c(&#39;rho_restricted_unweighted&#39;,&#39;rho_fdr_restricted_unweighted&#39;,&#39;h2_restricted_unweighted&#39;,&#39;rho_restricted_rho_weighted&#39;,&#39;rho_fdr_restricted_rho_weighted&#39;,&#39;h2_restricted_rho_weighted&#39;,&#39;rho_restricted_rho_h2_weighted&#39;,&#39;rho_fdr_restricted_rho_h2_weighted&#39;,&#39;h2_restricted_rho_h2_weighted&#39;)){
      
    pred_file&lt;-NULL
    target_gwas_i&lt;-target_gwas[i]
    target_pheno_i&lt;-target_pheno[i]
    target_pop_prev_i&lt;-target_pop_prev[i]
    
    for(secondary_gwas_i in secondary_gwas[secondary_gwas != target_gwas_i]){
      # Add in prs for target phenotype
      tmp&lt;-data.frame(predictors=paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,secondary_gwas_i,&#39;/prs.profiles&#39;),
                      group=&#39;standard&#39;)
      
      pred_file&lt;-rbind(pred_file, tmp)
      
      # Add in weighted prs for target phenotype
      tmp&lt;-data.frame(predictors=paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,secondary_gwas_i,&#39;/&#39;,target_gwas_i,&#39;/&#39;,type,&#39;/prs.profiles&#39;),
                      group=type)

      pred_file&lt;-rbind(pred_file, tmp)
      
      # Add in remainder of weighted prs for target phenotype
      tmp&lt;-data.frame(predictors=paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,secondary_gwas_i,&#39;/&#39;,target_gwas_i,&#39;/&#39;,gsub(&#39;restricted_.*&#39;,&#39;restricted_remainder&#39;,type),&#39;/prs.profiles&#39;),
                      group=type)
      
      pred_file&lt;-rbind(pred_file, tmp)
    }
  
    write.table(pred_file, paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_6_&#39;,type,&#39;.txt&#39;), col.names=T, row.names=F, quote=F)
  
    system(paste0(&#39;sbatch --mem 10G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R --pheno /users/k1806347/brc_scratch/Data/UKBB/Phenotype/PRS_comp_subset/UKBB.&#39;,target_pheno_i,&#39;.txt --out /users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_6_&#39;,type,&#39; --assoc T --outcome_pop_prev &#39;,target_pop_prev_i,&#39; --predictors /users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_6_&#39;,type,&#39;.txt&#39;))
  
  }
}

############
# Compare secondary PGS weighted and the remainder (seperately) +target PGS to unadjusted secondary PGS + target PGS
############

target_gwas&lt;-c(&#39;DEPR06&#39;,&#39;BODY04&#39;,&#39;INTE03&#39;)
target_pheno&lt;-c(&#39;Depression&#39;,&#39;BMI&#39;,&#39;Intelligence&#39;)
target_pop_prev&lt;-c(0.15,NA,NA)
secondary_gwas&lt;-c(&#39;DEPR06&#39;,&#39;SCHI02&#39;,&#39;BIPO02&#39;,&#39;AUTI07&#39;,&#39;SMOK04&#39;,&#39;ADHD05&#39;,&#39;COLL01&#39;,&#39;DIAB05&#39;,&#39;ANXI02&#39;,&#39;MENA01F&#39;,&#39;OBES01&#39;,&#39;WAIS01&#39;,&#39;COAD01&#39;,&#39;GLYC05&#39;,&#39;INTE01&#39;,&#39;INTE03&#39;,&#39;BODY04&#39;)

# use for loop to read all values and indexes
for(i in 1:length(target_gwas)){
  for(type in c(&#39;rho_restricted_unweighted&#39;,&#39;rho_fdr_restricted_unweighted&#39;,&#39;h2_restricted_unweighted&#39;,&#39;rho_restricted_rho_weighted&#39;,&#39;rho_fdr_restricted_rho_weighted&#39;,&#39;h2_restricted_rho_weighted&#39;,&#39;rho_restricted_rho_h2_weighted&#39;,&#39;rho_fdr_restricted_rho_h2_weighted&#39;,&#39;h2_restricted_rho_h2_weighted&#39;)){
      
    pred_file&lt;-NULL
    target_gwas_i&lt;-target_gwas[i]
    target_pheno_i&lt;-target_pheno[i]
    target_pop_prev_i&lt;-target_pop_prev[i]
    
    # Add in prs for target phenotype
    tmp&lt;-data.frame(predictors=paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,target_gwas_i,&#39;/prs.profiles&#39;),
                    group=&#39;baseline&#39;)
    
    pred_file&lt;-rbind(pred_file, tmp)

    # Add in prs for target phenotype
    tmp&lt;-data.frame(predictors=paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,target_gwas_i,&#39;/prs.profiles&#39;),
                    group=type)
    
    pred_file&lt;-rbind(pred_file, tmp)

    for(secondary_gwas_i in secondary_gwas[secondary_gwas != target_gwas_i]){
      # Add in prs for target phenotype
      tmp&lt;-data.frame(predictors=paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,secondary_gwas_i,&#39;/prs.profiles&#39;),
                      group=&#39;baseline&#39;)
      
      pred_file&lt;-rbind(pred_file, tmp)
      
      # Add in weighted prs for target phenotype
      tmp&lt;-data.frame(predictors=paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,secondary_gwas_i,&#39;/&#39;,target_gwas_i,&#39;/&#39;,type,&#39;/prs.profiles&#39;),
                      group=type)

      pred_file&lt;-rbind(pred_file, tmp)
      
      # Add in remainder of weighted prs for target phenotype
      tmp&lt;-data.frame(predictors=paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,secondary_gwas_i,&#39;/&#39;,target_gwas_i,&#39;/&#39;,gsub(&#39;restricted_.*&#39;,&#39;restricted_remainder&#39;,type),&#39;/prs.profiles&#39;),
                      group=type)
      
      pred_file&lt;-rbind(pred_file, tmp)
    }
  
    write.table(pred_file, paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_7_&#39;,type,&#39;.txt&#39;), col.names=T, row.names=F, quote=F)
  
    system(paste0(&#39;sbatch --mem 10G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R --pheno /users/k1806347/brc_scratch/Data/UKBB/Phenotype/PRS_comp_subset/UKBB.&#39;,target_pheno_i,&#39;.txt --out /users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_7_&#39;,type,&#39; --assoc T --outcome_pop_prev &#39;,target_pop_prev_i,&#39; --predictors /users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_7_&#39;,type,&#39;.txt&#39;))
  
  }
}</code></pre>
</details>
<hr />
</div>
</div>
<div id="results" class="section level1">
<h1><span class="header-section-number">3</span> Results</h1>
<hr />
<div id="heritabilitygenetic-correlation-results" class="section level2">
<h2><span class="header-section-number">3.1</span> Heritability/Genetic correlation results</h2>
<details>
<p><summary>Show code</summary></p>
<pre class="r"><code>target_gwas&lt;-c(&#39;DEPR06&#39;,&#39;BODY04&#39;,&#39;INTE03&#39;)
secondary_gwas&lt;-c(&#39;DEPR06&#39;,&#39;SCHI02&#39;,&#39;BIPO02&#39;,&#39;AUTI07&#39;,&#39;SMOK04&#39;,&#39;ADHD05&#39;,&#39;COLL01&#39;,&#39;DIAB05&#39;,&#39;ANXI02&#39;,&#39;MENA01F&#39;,&#39;OBES01&#39;,&#39;WAIS01&#39;,&#39;COAD01&#39;,&#39;GLYC05&#39;,&#39;INTE01&#39;,&#39;INTE03&#39;,&#39;BODY04&#39;)
ncase&lt;-c(116404,33640,20352,18381,41969,19099,NA,26676,NA,NA,32858,NA,60801,NA,NA,NA,NA)
ncon&lt;-c(314990,43456,31358,27969,32066,34194,NA,132532,NA,NA,65839,NA,123504,NA,NA,NA,NA)
prev&lt;-c(0.15,0.01,0.015,0.012,0.5,0.05,NA,0.05,NA,NA,0.13,NA,0.03,NA,NA,NA,NA)

# Read in LDSC results
scor = read.table(&quot;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/lava/sample_overlap/all.rg&quot;,header=T)

###
# Tabulate SNP-heritability results
###
hsq&lt;-scor[,c(&quot;p1&quot;,&quot;p2&quot;,&quot;h2_obs&quot;,&quot;h2_obs_se&quot;)] 
hsq&lt;-hsq[hsq$p1 == hsq$p2,]
hsq$p2&lt;-NULL
hsq$p1&lt;-gsub(&#39;.cleaned.*&#39;,&#39;&#39;,gsub(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/&#39;,&#39;&#39;,hsq$p1))

# Calculate p-value
hsq$p&lt;-pnorm(-abs(hsq$h2_obs/hsq$h2_obs_se))
  
# Convert heritability to liability scale
h2l_R2 &lt;- function(k, r2, p) {
  x= qnorm(1-k)
  z= dnorm(x)
  i=z/k
  C= k*(1-k)*k*(1-k)/(z^2*p*(1-p))
  theta= i*((p-k)/(1-k))*(i*((p-k)/(1-k))-x)
  h2l_R2 = C*r2 / (1 + C*theta*r2)
  return(h2l_R2)
}

se_h2l_R2 &lt;- function(k,h2,se, p) {
  x= qnorm(1-k)
  z= dnorm(x)
  i=z/k
  C= k*(1-k)*k*(1-k)/(z^2*p*(1-p))
  theta= i*((p-k)/(1-k))*(i*((p-k)/(1-k))-x)
  se_h2l_R2 = C*(1-h2*theta)*se
  return(se_h2l_R2)
}

hsq&lt;-hsq[match(secondary_gwas, hsq$p1),]

hsq$h2_liab&lt;-hsq$h2_obs
hsq$h2_liab_se&lt;-hsq$h2_obs_se

for(i in which(!is.na(prev))){
  hsq$h2_liab[i]&lt;-h2l_R2(k=prev[i], r2 = hsq$h2_obs[i], p = (ncase/(ncase+ncon))[i])
  hsq$h2_liab_se[i]&lt;-se_h2l_R2(k=prev[i], h2 = hsq$h2_obs[i], se=hsq$h2_obs_se[i], p = (ncase/(ncase+ncon))[i])
}

hsq$label&lt;-paste0(round(hsq$h2_liab,3), &quot;\n(&quot;, round(hsq$h2_liab_se, 3), &quot;)&quot;)

###
# Tabulate rG estimates
###
rg&lt;-scor[,c(&quot;p1&quot;,&quot;p2&quot;,&quot;rg&quot;,&quot;se&quot;)] 
rg$p1&lt;-gsub(&#39;.cleaned.*&#39;,&#39;&#39;,gsub(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/&#39;,&#39;&#39;,rg$p1))
rg$p2&lt;-gsub(&#39;.cleaned.*&#39;,&#39;&#39;,gsub(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/GWAS_sumstats/&#39;,&#39;&#39;,rg$p2))

rg$label&lt;-paste0(round(rg$rg,3), &quot;\n(&quot;, round(rg$se, 3), &quot;)&quot;)

# Calculate p
rg$p&lt;-2*pnorm(-abs(rg$rg/rg$se))

###
# Tabulate local rG estimates
###

lava_res&lt;-fread(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/lava/results/lava_descript.csv&#39;)

lava_res$label&lt;-paste0(lava_res$n_sig_fdr_loci_pos, &#39;/&#39;, lava_res$n_sig_fdr_loci_neg)

lava_res$prop_conc&lt;-lava_res$n_sig_fdr_loci_pos/lava_res$n_sig_fdr_loci
lava_res$prop_conc[lava_res$prop_conc &lt; 0.5]&lt;-1-lava_res$prop_conc[lava_res$prop_conc &lt; 0.5]

###
# Create plot
###

library(ggplot2)
library(cowplot)


hsq_plot&lt;-ggplot(data = hsq, aes(x = &#39;SNP-h2&#39;, y = p1)) +
    theme_bw() +
    geom_tile(aes(fill = h2_liab), colour = &#39;black&#39;) +
    scale_fill_gradientn(colours=c(&quot;white&quot;,&quot;green&quot;), na.value = &#39;grey&#39;,name = &quot;hsq&quot;) +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),plot.title = element_text(hjust = 0.5), legend.position = &quot;none&quot;) +
    geom_text(aes(label=label), color=&quot;black&quot;, size=3) +
    labs(x =&quot;&quot;, y = &quot;&quot;, title=&#39;SNP-h2&#39;) +
  coord_fixed()

rg_tmp&lt;-rg[rg$p1 %in% target_gwas,]
rg_tmp&lt;-rg_tmp[rg_tmp$p1 != rg_tmp$p2,]

rg_plot&lt;-ggplot(data = rg_tmp, aes(x = p1, y = p2)) +
    theme_bw() +
    geom_tile(aes(fill = rg), colour = &#39;black&#39;) +
    scale_fill_gradientn(colours=c(&quot;dodgerblue2&quot;,&quot;white&quot;,&quot;red&quot;), na.value = &#39;grey&#39;,name = &quot;rG&quot;) +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),plot.title = element_text(hjust = 0.5), legend.position = &quot;none&quot;) +
    geom_text(aes(label=label), color=&quot;black&quot;, size=3) +
    geom_text(data = rg_tmp[rg_tmp$p&lt;0.05,], aes(x = p1, y = p2, label=label), color=&quot;black&quot;, size=3, fontface = &quot;bold&quot;) +
    labs(x =&quot;&quot;, y = &quot;&quot;, title=&#39;Genome-wide rG&#39;) +
  coord_fixed()

local_rg_plot&lt;-ggplot(data = lava_res, aes(x = target, y = secondary)) +
    theme_bw() +
    geom_tile(aes(fill = prop_conc), colour = &#39;black&#39;) +
    scale_fill_gradientn(colours=c(&quot;white&quot;,&quot;purple&quot;), na.value = &#39;grey&#39;,name = &quot;Direction\nconcordance&quot;) +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),plot.title = element_text(hjust = 0.5), legend.position = &quot;none&quot;) +
    geom_text(aes(label=label), color=&quot;black&quot;, size=3) +
    labs(x =&quot;&quot;, y = &quot;&quot;, title=&#39;Local rG&#39;) +
  coord_fixed()

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/lava/results/hsq_rg_local.png&#39;), units=&#39;px&#39;, res=300, width=1900, height=2500)
  plot_grid(plotlist=list(hsq_plot, rg_plot, local_rg_plot), rel_widths=c(0.9,1.2,1.2), nrow=1)
dev.off()</code></pre>
</details>
<details>
<p><summary>Show SNP-h2, rG and local rG figure</summary></p>
<p><img src="Images/local_rg_weighted_pgs/hsq_rg_local.png" /></p>
</details>
<hr />
</div>
<div id="comparison-of-models" class="section level2">
<h2><span class="header-section-number">3.2</span> Comparison of models</h2>
<details>
<p><summary>Show code</summary></p>
<pre class="r"><code>library(data.table)
library(ggplot2)
library(cowplot)

target_gwas&lt;-c(&#39;DEPR06&#39;,&#39;BODY04&#39;,&#39;INTE03&#39;)
target_pheno&lt;-c(&#39;Depression&#39;,&#39;BMI&#39;,&#39;Intelligence&#39;)
target_pop_prev&lt;-c(0.15,NA,NA)
secondary_gwas&lt;-c(&#39;DEPR06&#39;,&#39;SCHI02&#39;,&#39;BIPO02&#39;,&#39;AUTI07&#39;,&#39;SMOK04&#39;,&#39;ADHD05&#39;,&#39;COLL01&#39;,&#39;DIAB05&#39;,&#39;ANXI02&#39;,&#39;MENA01F&#39;,&#39;OBES01&#39;,&#39;WAIS01&#39;,&#39;COAD01&#39;,&#39;GLYC05&#39;,&#39;INTE01&#39;,&#39;INTE03&#39;,&#39;BODY04&#39;)

#########
# Plot results testing gain from seconadary PGS over target PGS alone
#########

test&lt;-1  
eval_plot_list&lt;-list()

for(i in 1:length(target_gwas)){
  target_pheno_i&lt;-target_pheno[i]
  tmp_eval&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_&#39;,test,&#39;.pred_eval.txt&#39;))
  tmp_comp&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_&#39;,test,&#39;.pred_comp.txt&#39;))
  
  tmp_eval$Model&lt;-c(&#39;Target\nPGS&#39;,&#39;Secondary\nPGS&#39;,&#39;All&#39;)
  tmp_eval$Model&lt;-factor(tmp_eval$Model, levels=tmp_eval$Model)

  tmp_comp$R_diff_pval&lt;-format(tmp_comp$R_diff_pval, scientific = TRUE, digits = 2)
  tmp_comp$R_diff_pval[tmp_comp$R_diff_pval != &quot;0.0e+00&quot;]&lt;-paste0(&#39;=&#39;,tmp_comp$R_diff_pval[tmp_comp$R_diff_pval != &quot;0.0e+00&quot;])
  tmp_comp$R_diff_pval[tmp_comp$R_diff_pval == &quot;0.0e+00&quot;]&lt;-&#39;&lt;1e-300&#39;
  
  eval_plot&lt;-ggplot(tmp_eval, aes(x=Model, y=R)) +
    geom_bar(data=tmp_eval, aes(fill=Model), stat=&quot;identity&quot;, position=position_dodge()) +
    geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2, position=position_dodge(.9)) +
    labs(title=paste0(&#39;Target: &#39;,target_pheno_i), y=&quot;R (SE)&quot;, x=&#39;&#39;) +
    theme_half_open() +
    background_grid() +
    theme(legend.position = &quot;none&quot;)

  if(min(tmp_eval$R-tmp_eval$SE) &lt; 0){
    ylim_range&lt;-max(tmp_eval$R+tmp_eval$SE)-min(tmp_eval$R-tmp_eval$SE)
  } else {
    ylim_range&lt;-max(tmp_eval$R+tmp_eval$SE)
  }
  
  paths_1&lt;-data.frame(x=c(1,1,1.95,1.95),
                      y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20),
                          max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                          max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                          max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)))
  paths_2&lt;-data.frame(x=c(2.05,2.05,3,3),
                      y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20),
                          max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                          max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                          max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)))
  paths_3&lt;-data.frame(x=c(1,1,3,3),
                      y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)+(ylim_range/5),
                          max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5),
                          max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5),
                          max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)+(ylim_range/5)))
  
  eval_plot&lt;-eval_plot + 
    geom_path(data=paths_1, aes(x=x,y=y), colour=&#39;black&#39;) +
    geom_path(data=paths_2, aes(x=x,y=y), colour=&#39;black&#39;) +
    geom_path(data=paths_3, aes(x=x,y=y), colour=&#39;black&#39;)

  eval_plot&lt;-eval_plot + 
    annotate(&quot;text&quot;,x=1.5,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[4] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[2]), size=4) +
    annotate(&quot;text&quot;,x=2.5,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[8] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[6]), size=4) +
    annotate(&quot;text&quot;,x=2,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[7] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[7]), size=4)

  eval_plot_list[[paste0(i,&#39;_&#39;,test)]]&lt;-eval_plot
  
}

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/test_&#39;,test,&#39;.png&#39;), units=&#39;px&#39;, res=300, width=4600, height=1500)
  print(plot_grid(plotlist=eval_plot_list, nrow=1))
dev.off()

#########
# Compare results from stratified/weighted secondary PGS
#########

test&lt;-2
eval_plot_list&lt;-list()
assoc_plot_brief_list&lt;-list()
assoc_plot_brief_unres_list&lt;-list()
lava_plot_list&lt;-list()
for(i in 1:length(target_gwas)){
  target_pheno_i&lt;-target_pheno[i]
  target_gwas_i&lt;-target_gwas[i]
  tmp_assoc&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_&#39;,test,&#39;.assoc.txt&#39;))
  
  tmp_assoc$Filter&lt;-gsub(&#39;.restricted.*&#39;,&#39;&#39;,gsub(&#39;Group_&#39;,&#39;&#39;,tmp_assoc$Predictor))
  tmp_assoc$Filter[tmp_assoc$Filter == &#39;rho&#39;]&lt;-&#39;rho-p &lt; 0.05&#39;
  tmp_assoc$Filter[tmp_assoc$Filter == &#39;rho.fdr&#39;]&lt;-&#39;rho-fdr.p &lt; 0.05&#39;
  tmp_assoc$Filter[tmp_assoc$Filter == &#39;h2&#39;]&lt;-&#39;hsq-p &lt; 0.05&#39;
  tmp_assoc$Filter[tmp_assoc$Filter == &#39;All_group&#39;]&lt;-&#39;All&#39;
  tmp_assoc$Filter&lt;-factor(tmp_assoc$Filter, levels=c(&#39;hsq-p &lt; 0.05&#39;,&#39;rho-p &lt; 0.05&#39;,&#39;rho-fdr.p &lt; 0.05&#39;,&#39;All&#39;))
  
  tmp_assoc$Weighting&lt;-gsub(&#39;.PredFile.*&#39;,&#39;&#39;,gsub(&#39;.*.restricted.&#39;,&#39;&#39;,tmp_assoc$Predictor))
  tmp_assoc$Weighting[tmp_assoc$Weighting == &#39;unweighted&#39;]&lt;-&#39;Unweighted&#39;
  tmp_assoc$Weighting[tmp_assoc$Weighting == &#39;rho.weighted&#39;]&lt;-&#39;rho*BETA2&#39;
  tmp_assoc$Weighting[tmp_assoc$Weighting == &#39;rho.h2.weighted&#39;]&lt;-&quot;rho*BETA2/(hsq2/hsq1)&quot;
  tmp_assoc$Weighting&lt;-factor(tmp_assoc$Weighting, levels=c(&#39;Unweighted&#39;,&#39;rho*BETA2&#39;,&#39;rho*BETA2/(hsq2/hsq1)&#39;,&#39;All&#39;))
  
  tmp_assoc$Test&lt;-paste0(tmp_assoc$Filter,&#39;: &#39;,tmp_assoc$Weighting)
  tmp_assoc$Test&lt;-factor(tmp_assoc$Test, levels=c(&quot;hsq-p &lt; 0.05: Unweighted&quot;,&quot;rho-p &lt; 0.05: Unweighted&quot;,&quot;rho-fdr.p &lt; 0.05: Unweighted&quot;,&quot;hsq-p &lt; 0.05: rho*BETA2&quot;,&quot;rho-p &lt; 0.05: rho*BETA2&quot;,&quot;rho-fdr.p &lt; 0.05: rho*BETA2&quot;,&quot;hsq-p &lt; 0.05: rho*BETA2/(hsq2/hsq1)&quot;,&quot;rho-p &lt; 0.05: rho*BETA2/(hsq2/hsq1)&quot;,&quot;rho-fdr.p &lt; 0.05: rho*BETA2/(hsq2/hsq1)&quot;))
  
  tmp_pred&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_&#39;,test,&#39;.txt&#39;))
  gsub(&#39;/&#39;,&#39;&#39;,gsub(paste0(target_gwas_i,&#39;.*&#39;),&#39;&#39;,gsub(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,&#39;&#39;,tmp_pred$predictors)))
  tmp_assoc$GWAS&lt;-gsub(&#39;/&#39;,&#39;&#39;,gsub(paste0(target_gwas_i,&#39;.*&#39;),&#39;&#39;,gsub(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,&#39;&#39;,tmp_pred$predictors)))
  tmp_assoc$GWAS&lt;-factor(tmp_assoc$GWAS, levels=unique(tmp_assoc$GWAS))
  
  assoc_plot&lt;-ggplot(tmp_assoc, aes(x=GWAS, y=abs(BETA), fill=Weighting)) +
    geom_bar(stat=&quot;identity&quot;, position=position_dodge()) +
    geom_errorbar(aes(ymin=abs(BETA)-SE, ymax=abs(BETA)+SE), width=.2, position=position_dodge(.9)) +
    labs(title=paste0(&#39;Target: &#39;,target_pheno_i), y=&quot;Absolute-R (SE)&quot;, x=&#39;&#39;) +
    theme_half_open() +
    background_grid() +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
    facet_grid(rows = vars(Filter))

  png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_&#39;,test,&#39;.png&#39;), units=&#39;px&#39;, res=300, width=4500, height=2000)
  print(assoc_plot)
  dev.off()

  assoc_plot_brief&lt;-ggplot(tmp_assoc[tmp_assoc$Filter == &#39;rho-p &lt; 0.05&#39;,], aes(x=GWAS, y=abs(BETA), fill=Weighting)) +
    geom_bar(stat=&quot;identity&quot;, position=position_dodge()) +
    geom_errorbar(aes(ymin=abs(BETA)-SE, ymax=abs(BETA)+SE), width=.2, position=position_dodge(.9)) +
    labs(title=paste0(&#39;Target: &#39;,target_pheno_i), y=&quot;Absolute-R (SE)&quot;, x=&#39;&#39;) +
    theme_half_open() +
    background_grid() +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

  assoc_plot_brief_list[[i]]&lt;-assoc_plot_brief
  
  # Compare with unrestricted PGS associations
  tmp_assoc_unres&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_&#39;,1,&#39;.assoc.txt&#39;))
  tmp_pred_unres&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_&#39;,1,&#39;.txt&#39;))
  tmp_assoc_unres$GWAS&lt;-gsub(&#39;/prs.profiles&#39;,&#39;&#39;,gsub(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/prs/&#39;,&#39;&#39;,tmp_pred_unres$predictors))
  tmp_assoc_unres$GWAS&lt;-factor(tmp_assoc_unres$GWAS, levels=unique(tmp_assoc_unres$GWAS))
  tmp_assoc_unres$Filter&lt;-&#39;Standard&#39;
  tmp_assoc_unres$Weighting&lt;-&#39;Standard&#39;
  tmp_assoc_unres$Test&lt;-&#39;Standard&#39;
  tmp_assoc_unres&lt;-tmp_assoc_unres[tmp_assoc_unres$GWAS != target_gwas_i,]
  
  tmp_assoc_both&lt;-rbind(tmp_assoc_unres,tmp_assoc)
  
  assoc_plot_brief_unres&lt;-ggplot(tmp_assoc_both[tmp_assoc_both$Filter == &#39;rho-p &lt; 0.05&#39; | tmp_assoc_both$Filter == &#39;Standard&#39;,], aes(x=GWAS, y=abs(BETA), fill=Weighting)) +
    geom_bar(stat=&quot;identity&quot;, position=position_dodge()) +
    geom_errorbar(aes(ymin=abs(BETA)-SE, ymax=abs(BETA)+SE), width=.2, position=position_dodge(.9)) +
    labs(title=paste0(&#39;Target: &#39;,target_pheno_i), y=&quot;Absolute-R (SE)&quot;, x=&#39;&#39;, fill=&#39;Type&#39;) +
    theme_half_open() +
    background_grid() +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

  assoc_plot_brief_unres_list[[i]]&lt;-assoc_plot_brief_unres

  lava_res&lt;-fread(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/lava/results/lava_descript.csv&#39;)
  lava_res&lt;-lava_res[lava_res$target == target_gwas_i,]

  tmp_assoc_com&lt;-merge(tmp_assoc[tmp_assoc$Test == &#39;rho-p &lt; 0.05: Unweighted&#39;,],tmp_assoc[tmp_assoc$Test == &#39;rho-p &lt; 0.05: rho*BETA2&#39;,], by=&#39;GWAS&#39;)
  tmp_assoc_com$R_diff&lt;-abs(tmp_assoc_com$BETA.y)-abs(tmp_assoc_com$BETA.x)
  tmp_assoc_com&lt;-tmp_assoc_com[,c(&#39;GWAS&#39;,&#39;R_diff&#39;),with=F]
  lava_res&lt;-merge(tmp_assoc_com, lava_res, by.x=&#39;GWAS&#39;, by.y=&#39;secondary&#39;)
  
  lava_res$prop_conc&lt;-lava_res$n_sig_loci_pos/lava_res$n_sig_loci
  lava_res$prop_conc[lava_res$prop_conc &lt; 0.5]&lt;-1-lava_res$prop_conc[lava_res$prop_conc &lt; 0.5]
  lava_res$non_uniform&lt;-1-lava_res$prop_conc
  
  lava_plot&lt;-ggplot(lava_res, aes(x=non_uniform, y=R_diff)) +
    geom_point(aes(size=lava_res$n_sig_loci)) +
    scale_size_continuous(range = c(1, 3)) +
    labs(title=paste0(&#39;Target: &#39;,target_pheno_i), y=&quot;R improvement\nwhen weighting&quot;, x=&#39;Proportion Inconsistent&#39;, size=&#39;N loci&#39;) +
    theme_half_open() +
    background_grid()
  
  lava_plot_list[[i]]&lt;-lava_plot
  
  tmp_eval&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_&#39;,test,&#39;.pred_eval.txt&#39;))
  tmp_comp&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_&#39;,test,&#39;.pred_comp.txt&#39;))
  
  tmp_eval$Filter&lt;-gsub(&#39;.restricted.*&#39;,&#39;&#39;,tmp_eval$Model)
  tmp_eval$Filter[tmp_eval$Filter == &#39;rho&#39;]&lt;-&#39;rho-p &lt; 0.05&#39;
  tmp_eval$Filter[tmp_eval$Filter == &#39;rho.fdr&#39;]&lt;-&#39;rho-fdr.p &lt; 0.05&#39;
  tmp_eval$Filter[tmp_eval$Filter == &#39;h2&#39;]&lt;-&#39;hsq-p &lt; 0.05&#39;
  tmp_eval$Filter[tmp_eval$Filter == &#39;All_group&#39;]&lt;-&#39;All&#39;
  tmp_eval$Filter&lt;-factor(tmp_eval$Filter, levels=c(&#39;hsq-p &lt; 0.05&#39;,&#39;rho-p &lt; 0.05&#39;,&#39;rho-fdr.p &lt; 0.05&#39;,&#39;All&#39;))
  
  tmp_eval$Weighting&lt;-gsub(&#39;_group&#39;,&#39;&#39;,gsub(&#39;.*.restricted.&#39;,&#39;&#39;,tmp_eval$Model))
  tmp_eval$Weighting[tmp_eval$Weighting == &#39;unweighted&#39;]&lt;-&#39;Unweighted&#39;
  tmp_eval$Weighting[tmp_eval$Weighting == &#39;rho.weighted&#39;]&lt;-&#39;rho*BETA2&#39;
  tmp_eval$Weighting[tmp_eval$Weighting == &#39;rho.h2.weighted&#39;]&lt;-&quot;rho*BETA2/(hsq2/hsq1)&quot;
  tmp_eval$Weighting&lt;-factor(tmp_eval$Weighting, levels=c(&#39;Unweighted&#39;,&#39;rho*BETA2&#39;,&#39;rho*BETA2/(hsq2/hsq1)&#39;,&#39;All&#39;))
  
  eval_plot&lt;-ggplot(tmp_eval, aes(x=Filter, y=R)) +
    geom_bar(data=tmp_eval, aes(fill=Weighting), stat=&quot;identity&quot;, position=position_dodge()) +
    geom_errorbar(data=tmp_eval, aes(ymin=R-SE, ymax=R+SE, group=Weighting), width=.2, position=position_dodge(.9)) +
    labs(title=paste0(&#39;Target: &#39;,target_pheno_i), y=&quot;R (SE)&quot;, x=&#39;&#39;) +
    theme_half_open() +
    background_grid()

  eval_plot_list[[paste0(i,&#39;_&#39;,test)]]&lt;-eval_plot
}

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/test_&#39;,test,&#39;.png&#39;), units=&#39;px&#39;, res=300, width=2500, height=2500)
  print(plot_grid(plotlist=eval_plot_list, ncol=1))
dev.off()

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/test_&#39;,test,&#39;_assoc.png&#39;), units=&#39;px&#39;, res=300, width=2200, height=2500)
  print(plot_grid(plotlist=assoc_plot_brief_list, ncol=1))
dev.off()

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/test_&#39;,test,&#39;_assoc_with_unres.png&#39;), units=&#39;px&#39;, res=300, width=2200, height=2500)
  print(plot_grid(plotlist=assoc_plot_brief_unres_list, ncol=1))
dev.off()

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/test_&#39;,test,&#39;_lava.png&#39;), units=&#39;px&#39;, res=300, width=1500, height=2000)
  print(plot_grid(plotlist=lava_plot_list, ncol=1))
dev.off()

#########
# Plot results testing gain from seconadary PGS over target PGS alone
#########

test&lt;-3  
eval_plot_all_list&lt;-list()
tmp_eval_all&lt;-list()
tmp_comp_all&lt;-list()
for(i in 1:length(target_gwas)){
  eval_plot_list&lt;-list()
  tmp_eval_i&lt;-list()
  tmp_comp_i&lt;-list()
  for(type in c(&#39;rho_restricted_unweighted&#39;,&#39;rho_fdr_restricted_unweighted&#39;,&#39;h2_restricted_unweighted&#39;,&#39;rho_restricted_rho_weighted&#39;,&#39;rho_fdr_restricted_rho_weighted&#39;,&#39;h2_restricted_rho_weighted&#39;,&#39;rho_restricted_rho_h2_weighted&#39;,&#39;rho_fdr_restricted_rho_h2_weighted&#39;,&#39;h2_restricted_rho_h2_weighted&#39;)){
    target_pheno_i&lt;-target_pheno[i]
    tmp_eval&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_&#39;,test,&#39;_&#39;,type,&#39;.pred_eval.txt&#39;))
    tmp_comp&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_&#39;,test,&#39;_&#39;,type,&#39;.pred_comp.txt&#39;))
    
    tmp_comp$R_diff_pval&lt;-format(tmp_comp$R_diff_pval, scientific = TRUE, digits = 2)
    tmp_comp$R_diff_pval[as.numeric(tmp_comp$R_diff_pval) != 0]&lt;-paste0(&#39;=&#39;,tmp_comp$R_diff_pval[as.numeric(tmp_comp$R_diff_pval) != 0])
    tmp_comp$R_diff_pval[as.numeric(tmp_comp$R_diff_pval) == 0]&lt;-&#39;&lt;1e-300&#39;
    
    Filter&lt;-gsub(&#39;_restricted.*&#39;,&#39;&#39;, type)
    if(Filter == &#39;rho&#39;){Filter&lt;-&#39;rho-p &lt; 0.05&#39;}
    if(Filter == &#39;rho.fdr&#39;){Filter&lt;-&#39;rho-fdr.p &lt; 0.05&#39;}
    if(Filter == &#39;h2&#39;){Filter&lt;-&#39;hsq-p &lt; 0.05&#39;}

    Weighting&lt;-gsub(&#39;_group&#39;,&#39;&#39;,gsub(&#39;.*restricted_&#39;,&#39;&#39;, type))
    if(Weighting == &#39;unweighted&#39;){Weighting&lt;-&#39;Unweighted&#39;}
    if(Weighting == &#39;rho_weighted&#39;){Weighting&lt;-&#39;local rg\nweighted&#39;}
    if(Weighting == &#39;rho_h2_weighted&#39;){Weighting&lt;-&#39;local rg +\nhsq weighted&#39;}

    tmp_eval$Model&lt;-c(&#39;Target PGS +\nSecondary PGS&#39;,paste0(Filter,&#39;\n&#39;,Weighting),&#39;All&#39;)
    tmp_eval$Model&lt;-factor(tmp_eval$Model, levels=tmp_eval$Model)
  
    eval_plot&lt;-ggplot(tmp_eval, aes(x=Model, y=R)) +
      geom_bar(data=tmp_eval, aes(fill=Model), stat=&quot;identity&quot;, position=position_dodge()) +
      geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2, position=position_dodge(.9)) +
      labs(title=paste0(&#39;Target: &#39;,target_pheno_i), y=&quot;R (SE)&quot;, x=&#39;&#39;) +
      theme_half_open() +
      background_grid() +
      theme(legend.position = &quot;none&quot;)
  
    if(min(tmp_eval$R-tmp_eval$SE) &lt; 0){
      ylim_range&lt;-max(tmp_eval$R+tmp_eval$SE)-min(tmp_eval$R-tmp_eval$SE)
    } else {
      ylim_range&lt;-max(tmp_eval$R+tmp_eval$SE)
    }
    
    paths_1&lt;-data.frame(x=c(1,1,1.95,1.95),
                        y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)))
    paths_2&lt;-data.frame(x=c(2.05,2.05,3,3),
                        y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)))
    paths_3&lt;-data.frame(x=c(1,1,3,3),
                        y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)+(ylim_range/5),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)+(ylim_range/5)))
    
    eval_plot&lt;-eval_plot + 
      geom_path(data=paths_1, aes(x=x,y=y), colour=&#39;black&#39;) +
      geom_path(data=paths_2, aes(x=x,y=y), colour=&#39;black&#39;) +
      geom_path(data=paths_3, aes(x=x,y=y), colour=&#39;black&#39;)
  
    eval_plot&lt;-eval_plot + 
      annotate(&quot;text&quot;,x=1.5,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[4] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[4],&#39;   &#39;), size=4) +
      annotate(&quot;text&quot;,x=2.5,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[8] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[8]), size=4) +
      annotate(&quot;text&quot;,x=2,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[7] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[7]), size=4)
  
    eval_plot_list[[paste0(i,&#39;_&#39;,test,&#39;_&#39;,type)]]&lt;-eval_plot
    
    if(grepl(&#39;rho_restricted&#39;,type)){
      tmp_eval$Model&lt;-gsub(&#39;rho-p &lt; 0.05\n&#39;,&#39;&#39;,tmp_eval$Model)  
      tmp_eval$Model&lt;-factor(tmp_eval$Model, levels=tmp_eval$Model)
      
      eval_plot&lt;-ggplot(tmp_eval, aes(x=Model, y=R)) +
        geom_bar(data=tmp_eval, aes(fill=Model), stat=&quot;identity&quot;, position=position_dodge()) +
        geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2, position=position_dodge(.9)) +
        labs(title=paste0(&#39;Target: &#39;,target_pheno_i), y=&quot;R (SE)&quot;, x=&#39;&#39;) +
        theme_half_open() +
        background_grid() +
        theme(legend.position = &quot;none&quot;)
    
      if(min(tmp_eval$R-tmp_eval$SE) &lt; 0){
        ylim_range&lt;-max(tmp_eval$R+tmp_eval$SE)-min(tmp_eval$R-tmp_eval$SE)
      } else {
        ylim_range&lt;-max(tmp_eval$R+tmp_eval$SE)
      }
      
      paths_1&lt;-data.frame(x=c(1,1,1.95,1.95),
                          y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)))
      paths_2&lt;-data.frame(x=c(2.05,2.05,3,3),
                          y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)))
      paths_3&lt;-data.frame(x=c(1,1,3,3),
                          y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)+(ylim_range/5),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)+(ylim_range/5)))
      
      eval_plot&lt;-eval_plot + 
        geom_path(data=paths_1, aes(x=x,y=y), colour=&#39;black&#39;) +
        geom_path(data=paths_2, aes(x=x,y=y), colour=&#39;black&#39;) +
        geom_path(data=paths_3, aes(x=x,y=y), colour=&#39;black&#39;)
    
      eval_plot&lt;-eval_plot + 
        annotate(&quot;text&quot;,x=1.5,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[4] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[4],&#39;   &#39;), size=4) +
        annotate(&quot;text&quot;,x=2.5,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[8] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[8]), size=4) +
        annotate(&quot;text&quot;,x=2,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[7] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[7]), size=4)
        
      eval_plot_all_list[[paste0(i,&#39;_&#39;,test,&#39;_&#39;,type)]]&lt;-eval_plot
      
      tmp_eval$Phenotype&lt;-target_pheno_i
      tmp_eval$Type&lt;-type
      tmp_eval_i[[paste0(i,&#39;_&#39;,test,&#39;_&#39;,type)]]&lt;-tmp_eval
      
      tmp_comp$Phenotype&lt;-target_pheno_i
      tmp_comp$Type&lt;-type
      tmp_comp_i[[paste0(i,&#39;_&#39;,test,&#39;_&#39;,type)]]&lt;-tmp_comp
    }
  }
  png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_&#39;,test,&#39;.png&#39;), units=&#39;px&#39;, res=300, width=5000, height=3000)
  print(plot_grid(plotlist=eval_plot_list, ncol=3))
  dev.off()
  
  tmp_eval_all[[i]]&lt;-do.call(rbind,tmp_eval_i)
  tmp_eval_all[[i]]$Ncase&lt;-NULL
  tmp_eval_all[[i]]$Ncont&lt;-NULL
  tmp_eval_all[[i]]$R2o&lt;-NULL
  tmp_eval_all[[i]]$R2l&lt;-NULL
  
  tmp_comp_all[[i]]&lt;-do.call(rbind,tmp_comp_i)
}

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/test_&#39;,test,&#39;.png&#39;), units=&#39;px&#39;, res=300, width=4750, height=3000)
print(plot_grid(plotlist=eval_plot_all_list, ncol=3))
dev.off()

###
# Combine h2_resricted results and make simplified plot
###
tmp_eval_all_tab&lt;-do.call(rbind, tmp_eval_all)
tmp_comp_all_tab&lt;-do.call(rbind, tmp_comp_all)

# Remove secondary PGS only rows
tmp_eval_all_tab&lt;-tmp_eval_all_tab[tmp_eval_all_tab$Model == &#39;Target PGS +\nSecondary PGS&#39; | tmp_eval_all_tab$Model == &#39;All&#39; ,]
tmp_comp_all_tab&lt;-tmp_comp_all_tab[(tmp_comp_all_tab$Model_1 == &#39;All&#39; &amp; tmp_comp_all_tab$Model_2 == &#39;baseline&#39;),]

# Rename &#39;All&#39; Models
tmp_eval_all_tab$Model&lt;-as.character(tmp_eval_all_tab$Model)
tmp_eval_all_tab$Model[tmp_eval_all_tab$Model == &#39;All&#39;]&lt;-paste0(tmp_eval_all_tab$Model[tmp_eval_all_tab$Model == &#39;All&#39;],&#39;_&#39;,tmp_eval_all_tab$Type[tmp_eval_all_tab$Model == &#39;All&#39;])

# Remove duplicated Target PGS + Secondary PGS results
tmp_eval_all_tab&lt;-tmp_eval_all_tab[!duplicated(paste0(tmp_eval_all_tab$Model,tmp_eval_all_tab$Phenotype)),]

# Rename models
tmp_eval_all_tab$Model&lt;-rep(c(&#39;Baseline&#39;,&#39;Unweighted&#39;,&#39;Local rG\nWeighted&#39;,&#39;Local rG +\nhsq Weighted&#39;),3)
tmp_eval_all_tab$Model&lt;-factor(tmp_eval_all_tab$Model, levels=c(&#39;Baseline&#39;,&#39;Unweighted&#39;,&#39;Local rG\nWeighted&#39;,&#39;Local rG +\nhsq Weighted&#39;))

eval_plot_brief_all_list&lt;-NULL
for(i in 1:length(target_gwas)){
  target_pheno_i&lt;-target_pheno[i]
  tmp_eval_all_tab_i&lt;-tmp_eval_all_tab[tmp_eval_all_tab$Phenotype == target_pheno_i,]
  tmp_comp_all_tab_i&lt;-tmp_comp_all_tab[tmp_comp_all_tab$Phenotype == target_pheno_i,]
  
  eval_plot&lt;-ggplot(tmp_eval_all_tab_i, aes(x=Model, y=R)) +
    geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2, position=position_dodge(.9)) +
    geom_point(aes(fill=Model), stat=&quot;identity&quot;, position=position_dodge(1), size=5, shape=23, colour=&#39;black&#39;) +
    labs(title=paste0(&#39;Target: &#39;,target_pheno_i), y=&quot;R (SE)&quot;, x=&#39;&#39;) +
    theme_half_open() +
    background_grid() +
    theme(legend.position = &quot;none&quot;, axis.text.x = element_text(angle = 45, hjust=1))

    ylim_range&lt;-max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)-min(tmp_eval_all_tab_i$R-tmp_eval_all_tab_i$SE)

    paths_1&lt;-data.frame(x=c(1,1,2,2),
                        y=c(max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/20),
                            max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/10),
                            max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/10),
                            max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/20)))
    paths_2&lt;-data.frame(x=c(1,1,3,3),
                        y=c(max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/20)+(1*(ylim_range/5)),
                            max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/10)+(1*(ylim_range/5)),
                            max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/10)+(1*(ylim_range/5)),
                            max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/20)+(1*(ylim_range/5))))
    paths_3&lt;-data.frame(x=c(1,1,4,4),
                        y=c(max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/20)+(2*(ylim_range/5)),
                            max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/10)+(2*(ylim_range/5)),
                            max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/10)+(2*(ylim_range/5)),
                            max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/20)+(2*(ylim_range/5))))

    eval_plot&lt;-eval_plot + 
      geom_path(data=paths_1, aes(x=x,y=y), colour=&#39;black&#39;) +
      geom_path(data=paths_2, aes(x=x,y=y), colour=&#39;black&#39;) +
      geom_path(data=paths_3, aes(x=x,y=y), colour=&#39;black&#39;)
  
    eval_plot&lt;-eval_plot + 
      annotate(&quot;text&quot;,x=1.5,y=max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/10)+(ylim_range/15),label=paste0(round((tmp_comp_all_tab_i$R_diff/tmp_comp_all_tab_i$Model_2_R)[1] * 100,1),&#39;%; p&#39;, tmp_comp_all_tab_i$R_diff_pval[1],&#39;   &#39;), size=4) +
      annotate(&quot;text&quot;,x=2,y=max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/10)+(ylim_range/15)+(1*(ylim_range/5)),label=paste0(round((tmp_comp_all_tab_i$R_diff/tmp_comp_all_tab_i$Model_2_R)[2] * 100,1),&#39;%; p&#39;, tmp_comp_all_tab_i$R_diff_pval[2]), size=4) +
      annotate(&quot;text&quot;,x=2.5,y=max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/10)+(ylim_range/15)+(2*(ylim_range/5)),label=paste0(round((tmp_comp_all_tab_i$R_diff/tmp_comp_all_tab_i$Model_2_R)[3] * 100,1),&#39;%; p&#39;, tmp_comp_all_tab_i$R_diff_pval[3]), size=4)
    
    eval_plot_brief_all_list[[i]]&lt;-eval_plot

}

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/test_&#39;,test,&#39;_brief.png&#39;), units=&#39;px&#39;, res=300, width=3750, height=1300)
print(plot_grid(plotlist=eval_plot_brief_all_list, ncol=3))
dev.off()

#########
# Plot results testing gain from reweighted with unadjusted seconadary PGS
#########

for(test in 4:5){
eval_plot_all_list&lt;-list()
for(i in 1:length(target_gwas)){
  eval_plot_list&lt;-list()
    for(type in c(&#39;rho_restricted_rho_weighted_with_unadjusted&#39;,&#39;rho_fdr_restricted_rho_weighted_with_unadjusted&#39;,&#39;h2_restricted_rho_weighted_with_unadjusted&#39;,&#39;rho_restricted_rho_h2_weighted_with_unadjusted&#39;,&#39;rho_fdr_restricted_rho_h2_weighted_with_unadjusted&#39;,&#39;h2_restricted_rho_h2_weighted_with_unadjusted&#39;)){
    target_pheno_i&lt;-target_pheno[i]
    tmp_eval&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_&#39;,test,&#39;_&#39;,type,&#39;.pred_eval.txt&#39;))
    tmp_comp&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_&#39;,test,&#39;_&#39;,type,&#39;.pred_comp.txt&#39;))
    
    tmp_comp$R_diff_pval&lt;-format(tmp_comp$R_diff_pval, scientific = TRUE, digits = 2)
    tmp_comp$R_diff_pval[as.numeric(tmp_comp$R_diff_pval) != 0]&lt;-paste0(&#39;=&#39;,tmp_comp$R_diff_pval[as.numeric(tmp_comp$R_diff_pval) != 0])
    tmp_comp$R_diff_pval[as.numeric(tmp_comp$R_diff_pval) == 0]&lt;-&#39;&lt;1e-300&#39;
    
    Filter&lt;-gsub(&#39;_restricted.*&#39;,&#39;&#39;, type)
    if(Filter == &#39;rho&#39;){Filter&lt;-&#39;rho-p &lt; 0.05&#39;}
    if(Filter == &#39;rho.fdr&#39;){Filter&lt;-&#39;rho-fdr.p &lt; 0.05&#39;}
    if(Filter == &#39;h2&#39;){Filter&lt;-&#39;hsq-p &lt; 0.05&#39;}

    Weighting&lt;-gsub(&#39;_with_unadjusted&#39;,&#39;&#39;,gsub(&#39;.*restricted_&#39;,&#39;&#39;, type))
    if(Weighting == &#39;unweighted&#39;){Weighting&lt;-&#39;Unweighted&#39;}
    if(Weighting == &#39;rho_weighted&#39;){Weighting&lt;-&#39;local rg\nweighted&#39;}
    if(Weighting == &#39;rho_h2_weighted&#39;){Weighting&lt;-&#39;local rg +\nhsq weighted&#39;}

    tmp_eval$Model&lt;-c(&#39;Standard&#39;,paste0(Filter,&#39;\n&#39;,Weighting),&#39;All&#39;)
    tmp_eval$Model&lt;-factor(tmp_eval$Model, levels=tmp_eval$Model)
  
    eval_plot&lt;-ggplot(tmp_eval, aes(x=Model, y=R)) +
      geom_bar(data=tmp_eval, aes(fill=Model), stat=&quot;identity&quot;, position=position_dodge()) +
      geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2, position=position_dodge(.9)) +
      labs(title=paste0(&#39;Target: &#39;,target_pheno_i), y=&quot;R (SE)&quot;, x=&#39;&#39;) +
      theme_half_open() +
      background_grid() +
      theme(legend.position = &quot;none&quot;)
  
    if(min(tmp_eval$R-tmp_eval$SE) &lt; 0){
      ylim_range&lt;-max(tmp_eval$R+tmp_eval$SE)-min(tmp_eval$R-tmp_eval$SE)
    } else {
      ylim_range&lt;-max(tmp_eval$R+tmp_eval$SE)
    }
    
    paths_1&lt;-data.frame(x=c(1,1,1.95,1.95),
                        y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)))
    paths_2&lt;-data.frame(x=c(2.05,2.05,3,3),
                        y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)))
    paths_3&lt;-data.frame(x=c(1,1,3,3),
                        y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)+(ylim_range/5),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)+(ylim_range/5)))
    
    eval_plot&lt;-eval_plot + 
      geom_path(data=paths_1, aes(x=x,y=y), colour=&#39;black&#39;) +
      geom_path(data=paths_2, aes(x=x,y=y), colour=&#39;black&#39;) +
      geom_path(data=paths_3, aes(x=x,y=y), colour=&#39;black&#39;)
  
    eval_plot&lt;-eval_plot + 
      annotate(&quot;text&quot;,x=1.5,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[4] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[4],&#39;   &#39;), size=4) +
      annotate(&quot;text&quot;,x=2.5,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[8] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[8]), size=4) +
      annotate(&quot;text&quot;,x=2,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[7] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[7]), size=4)
  
    eval_plot_list[[paste0(i,&#39;_&#39;,test,&#39;_&#39;,type)]]&lt;-eval_plot
    
    if(grepl(&#39;rho_restricted&#39;,type)){
      tmp_eval$Model&lt;-gsub(&#39;rho-p &lt; 0.05\n&#39;,&#39;&#39;,tmp_eval$Model)  
      tmp_eval$Model&lt;-factor(tmp_eval$Model, levels=tmp_eval$Model)
      
      eval_plot&lt;-ggplot(tmp_eval, aes(x=Model, y=R)) +
        geom_bar(data=tmp_eval, aes(fill=Model), stat=&quot;identity&quot;, position=position_dodge()) +
        geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2, position=position_dodge(.9)) +
        labs(title=paste0(&#39;Target: &#39;,target_pheno_i), y=&quot;R (SE)&quot;, x=&#39;&#39;) +
        theme_half_open() +
        background_grid() +
        theme(legend.position = &quot;none&quot;)
    
      if(min(tmp_eval$R-tmp_eval$SE) &lt; 0){
        ylim_range&lt;-max(tmp_eval$R+tmp_eval$SE)-min(tmp_eval$R-tmp_eval$SE)
      } else {
        ylim_range&lt;-max(tmp_eval$R+tmp_eval$SE)
      }
      
      paths_1&lt;-data.frame(x=c(1,1,1.95,1.95),
                          y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)))
      paths_2&lt;-data.frame(x=c(2.05,2.05,3,3),
                          y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)))
      paths_3&lt;-data.frame(x=c(1,1,3,3),
                          y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)+(ylim_range/5),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)+(ylim_range/5)))
      
      eval_plot&lt;-eval_plot + 
        geom_path(data=paths_1, aes(x=x,y=y), colour=&#39;black&#39;) +
        geom_path(data=paths_2, aes(x=x,y=y), colour=&#39;black&#39;) +
        geom_path(data=paths_3, aes(x=x,y=y), colour=&#39;black&#39;)
    
      eval_plot&lt;-eval_plot + 
        annotate(&quot;text&quot;,x=1.5,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[4] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[4],&#39;   &#39;), size=4) +
        annotate(&quot;text&quot;,x=2.5,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[8] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[8]), size=4) +
        annotate(&quot;text&quot;,x=2,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[7] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[7]), size=4)
        
      eval_plot_all_list[[paste0(i,&#39;_&#39;,test,&#39;_&#39;,type)]]&lt;-eval_plot
    }
  }
  png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_&#39;,test,&#39;.png&#39;), units=&#39;px&#39;, res=300, width=5000, height=3000)
  print(plot_grid(plotlist=eval_plot_list, ncol=3))
  dev.off()
}

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/test_&#39;,test,&#39;.png&#39;), units=&#39;px&#39;, res=300, width=4750, height=3000)
print(plot_grid(plotlist=eval_plot_all_list, ncol=3))
dev.off()
}

#########
# Plot results testing gain from seconadary PGS over target PGS alone
#########
  
test&lt;-6  
eval_plot_all_list&lt;-list()
tmp_eval_all&lt;-list()
tmp_comp_all&lt;-list()
for(i in 1:length(target_gwas)){
  eval_plot_list&lt;-list()
  tmp_eval_i&lt;-list()
  tmp_comp_i&lt;-list()
  for(type in c(&#39;rho_restricted_unweighted&#39;,&#39;rho_fdr_restricted_unweighted&#39;,&#39;h2_restricted_unweighted&#39;,&#39;rho_restricted_rho_weighted&#39;,&#39;rho_fdr_restricted_rho_weighted&#39;,&#39;h2_restricted_rho_weighted&#39;,&#39;rho_restricted_rho_h2_weighted&#39;,&#39;rho_fdr_restricted_rho_h2_weighted&#39;,&#39;h2_restricted_rho_h2_weighted&#39;)){
    target_pheno_i&lt;-target_pheno[i]
    tmp_eval&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_&#39;,test,&#39;_&#39;,type,&#39;.pred_eval.txt&#39;))
    tmp_comp&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_&#39;,test,&#39;_&#39;,type,&#39;.pred_comp.txt&#39;))
    
    tmp_comp$R_diff_pval&lt;-format(tmp_comp$R_diff_pval, scientific = TRUE, digits = 2)
    tmp_comp$R_diff_pval[as.numeric(tmp_comp$R_diff_pval) != 0]&lt;-paste0(&#39;=&#39;,tmp_comp$R_diff_pval[as.numeric(tmp_comp$R_diff_pval) != 0])
    tmp_comp$R_diff_pval[as.numeric(tmp_comp$R_diff_pval) == 0]&lt;-&#39;&lt;1e-300&#39;
    
    Filter&lt;-gsub(&#39;_restricted.*&#39;,&#39;&#39;, type)
    if(Filter == &#39;rho&#39;){Filter&lt;-&#39;rho-p &lt; 0.05&#39;}
    if(Filter == &#39;rho.fdr&#39;){Filter&lt;-&#39;rho-fdr.p &lt; 0.05&#39;}
    if(Filter == &#39;h2&#39;){Filter&lt;-&#39;hsq-p &lt; 0.05&#39;}

    Weighting&lt;-gsub(&#39;_group&#39;,&#39;&#39;,gsub(&#39;.*restricted_&#39;,&#39;&#39;, type))
    if(Weighting == &#39;unweighted&#39;){Weighting&lt;-&#39;Unweighted&#39;}
    if(Weighting == &#39;rho_weighted&#39;){Weighting&lt;-&#39;local rg\nweighted&#39;}
    if(Weighting == &#39;rho_h2_weighted&#39;){Weighting&lt;-&#39;local rg +\nhsq weighted&#39;}

    tmp_eval$Model&lt;-c(&#39;Standard&#39;,paste0(Filter,&#39;\n&#39;,Weighting),&#39;All&#39;)
    tmp_eval$Model&lt;-factor(tmp_eval$Model, levels=tmp_eval$Model)
  
    eval_plot&lt;-ggplot(tmp_eval, aes(x=Model, y=R)) +
      geom_bar(data=tmp_eval, aes(fill=Model), stat=&quot;identity&quot;, position=position_dodge()) +
      geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2, position=position_dodge(.9)) +
      labs(title=paste0(&#39;Target: &#39;,target_pheno_i), y=&quot;R (SE)&quot;, x=&#39;&#39;) +
      theme_half_open() +
      background_grid() +
      theme(legend.position = &quot;none&quot;)
  
    if(min(tmp_eval$R-tmp_eval$SE) &lt; 0){
      ylim_range&lt;-max(tmp_eval$R+tmp_eval$SE)-min(tmp_eval$R-tmp_eval$SE)
    } else {
      ylim_range&lt;-max(tmp_eval$R+tmp_eval$SE)
    }
    
    paths_1&lt;-data.frame(x=c(1,1,1.95,1.95),
                        y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)))
    paths_2&lt;-data.frame(x=c(2.05,2.05,3,3),
                        y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)))
    paths_3&lt;-data.frame(x=c(1,1,3,3),
                        y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)+(ylim_range/5),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)+(ylim_range/5)))
    
    eval_plot&lt;-eval_plot + 
      geom_path(data=paths_1, aes(x=x,y=y), colour=&#39;black&#39;) +
      geom_path(data=paths_2, aes(x=x,y=y), colour=&#39;black&#39;) +
      geom_path(data=paths_3, aes(x=x,y=y), colour=&#39;black&#39;)
  
    eval_plot&lt;-eval_plot + 
      annotate(&quot;text&quot;,x=1.5,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[4] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[4],&#39;   &#39;), size=4) +
      annotate(&quot;text&quot;,x=2.5,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[8] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[8]), size=4) +
      annotate(&quot;text&quot;,x=2,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[7] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[7]), size=4)
  
    eval_plot_list[[paste0(i,&#39;_&#39;,test,&#39;_&#39;,type)]]&lt;-eval_plot
    
    if(grepl(&#39;rho_restricted&#39;,type)){
      tmp_eval$Model&lt;-gsub(&#39;rho-p &lt; 0.05\n&#39;,&#39;&#39;,tmp_eval$Model)  
      tmp_eval$Model&lt;-factor(tmp_eval$Model, levels=tmp_eval$Model)
      
      eval_plot&lt;-ggplot(tmp_eval, aes(x=Model, y=R)) +
        geom_bar(data=tmp_eval, aes(fill=Model), stat=&quot;identity&quot;, position=position_dodge()) +
        geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2, position=position_dodge(.9)) +
        labs(title=paste0(&#39;Target: &#39;,target_pheno_i), y=&quot;R (SE)&quot;, x=&#39;&#39;) +
        theme_half_open() +
        background_grid() +
        theme(legend.position = &quot;none&quot;)
    
      if(min(tmp_eval$R-tmp_eval$SE) &lt; 0){
        ylim_range&lt;-max(tmp_eval$R+tmp_eval$SE)-min(tmp_eval$R-tmp_eval$SE)
      } else {
        ylim_range&lt;-max(tmp_eval$R+tmp_eval$SE)
      }
      
      paths_1&lt;-data.frame(x=c(1,1,1.95,1.95),
                          y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)))
      paths_2&lt;-data.frame(x=c(2.05,2.05,3,3),
                          y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)))
      paths_3&lt;-data.frame(x=c(1,1,3,3),
                          y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)+(ylim_range/5),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)+(ylim_range/5)))
      
      eval_plot&lt;-eval_plot + 
        geom_path(data=paths_1, aes(x=x,y=y), colour=&#39;black&#39;) +
        geom_path(data=paths_2, aes(x=x,y=y), colour=&#39;black&#39;) +
        geom_path(data=paths_3, aes(x=x,y=y), colour=&#39;black&#39;)
    
      eval_plot&lt;-eval_plot + 
        annotate(&quot;text&quot;,x=1.5,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[4] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[4],&#39;   &#39;), size=4) +
        annotate(&quot;text&quot;,x=2.5,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[8] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[8]), size=4) +
        annotate(&quot;text&quot;,x=2,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[7] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[7]), size=4)
        
      eval_plot_all_list[[paste0(i,&#39;_&#39;,test,&#39;_&#39;,type)]]&lt;-eval_plot
      
      tmp_eval$Phenotype&lt;-target_pheno_i
      tmp_eval$Type&lt;-type
      tmp_eval_i[[paste0(i,&#39;_&#39;,test,&#39;_&#39;,type)]]&lt;-tmp_eval
      
      tmp_comp$Phenotype&lt;-target_pheno_i
      tmp_comp$Type&lt;-type
      tmp_comp_i[[paste0(i,&#39;_&#39;,test,&#39;_&#39;,type)]]&lt;-tmp_comp
    }
  }
  png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_&#39;,test,&#39;.png&#39;), units=&#39;px&#39;, res=300, width=5000, height=3000)
  print(plot_grid(plotlist=eval_plot_list, ncol=3))
  dev.off()
  
  tmp_eval_all[[i]]&lt;-do.call(rbind,tmp_eval_i)
  tmp_eval_all[[i]]$Ncase&lt;-NULL
  tmp_eval_all[[i]]$Ncont&lt;-NULL
  tmp_eval_all[[i]]$R2o&lt;-NULL
  tmp_eval_all[[i]]$R2l&lt;-NULL
  
  tmp_comp_all[[i]]&lt;-do.call(rbind,tmp_comp_i)
}

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/test_&#39;,test,&#39;.png&#39;), units=&#39;px&#39;, res=300, width=4750, height=3000)
print(plot_grid(plotlist=eval_plot_all_list, ncol=3))
dev.off()

# Make brief version of the plot
tmp_eval_all_tab&lt;-do.call(rbind, tmp_eval_all)
tmp_comp_all_tab&lt;-do.call(rbind, tmp_comp_all)

# Remove secondary PGS only rows
tmp_eval_all_tab&lt;-tmp_eval_all_tab[tmp_eval_all_tab$Model == &#39;Standard&#39; | tmp_eval_all_tab$Model == &#39;All&#39; ,]
tmp_comp_all_tab&lt;-tmp_comp_all_tab[(tmp_comp_all_tab$Model_1 == &#39;All&#39; &amp; tmp_comp_all_tab$Model_2 == &#39;standard&#39;),]

# Rename &#39;All&#39; Models
tmp_eval_all_tab$Model&lt;-as.character(tmp_eval_all_tab$Model)
tmp_eval_all_tab$Model[tmp_eval_all_tab$Model == &#39;All&#39;]&lt;-paste0(tmp_eval_all_tab$Model[tmp_eval_all_tab$Model == &#39;All&#39;],&#39;_&#39;,tmp_eval_all_tab$Type[tmp_eval_all_tab$Model == &#39;All&#39;])

# Remove duplicated Target PGS + Secondary PGS results
tmp_eval_all_tab&lt;-tmp_eval_all_tab[!duplicated(paste0(tmp_eval_all_tab$Model,tmp_eval_all_tab$Phenotype)),]

# Rename models
tmp_eval_all_tab$Model&lt;-rep(c(&#39;Standard&#39;,&#39;Unweighted&#39;,&#39;Local rG\nWeighted&#39;,&#39;Local rG +\nhsq Weighted&#39;),3)
tmp_eval_all_tab$Model&lt;-factor(tmp_eval_all_tab$Model, levels=c(&#39;Standard&#39;,&#39;Unweighted&#39;,&#39;Local rG\nWeighted&#39;,&#39;Local rG +\nhsq Weighted&#39;))

eval_plot_brief_all_list&lt;-NULL
for(i in 1:length(target_gwas)){
  target_pheno_i&lt;-target_pheno[i]
  tmp_eval_all_tab_i&lt;-tmp_eval_all_tab[tmp_eval_all_tab$Phenotype == target_pheno_i,]
  tmp_comp_all_tab_i&lt;-tmp_comp_all_tab[tmp_comp_all_tab$Phenotype == target_pheno_i,]
  
  eval_plot&lt;-ggplot(tmp_eval_all_tab_i, aes(x=Model, y=R)) +
    geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2, position=position_dodge(.9)) +
    geom_point(aes(fill=Model), stat=&quot;identity&quot;, position=position_dodge(1), size=5, shape=23, colour=&#39;black&#39;) +
    labs(title=paste0(&#39;Target: &#39;,target_pheno_i), y=&quot;R (SE)&quot;, x=&#39;&#39;) +
    theme_half_open() +
    background_grid() +
    theme(legend.position = &quot;none&quot;, axis.text.x = element_text(angle = 45, hjust=1))

    ylim_range&lt;-max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)-min(tmp_eval_all_tab_i$R-tmp_eval_all_tab_i$SE)

    paths_1&lt;-data.frame(x=c(1,1,2,2),
                        y=c(max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/20),
                            max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/10),
                            max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/10),
                            max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/20)))
    paths_2&lt;-data.frame(x=c(1,1,3,3),
                        y=c(max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/20)+(1*(ylim_range/5)),
                            max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/10)+(1*(ylim_range/5)),
                            max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/10)+(1*(ylim_range/5)),
                            max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/20)+(1*(ylim_range/5))))
    paths_3&lt;-data.frame(x=c(1,1,4,4),
                        y=c(max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/20)+(2*(ylim_range/5)),
                            max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/10)+(2*(ylim_range/5)),
                            max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/10)+(2*(ylim_range/5)),
                            max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/20)+(2*(ylim_range/5))))

    eval_plot&lt;-eval_plot + 
      geom_path(data=paths_1, aes(x=x,y=y), colour=&#39;black&#39;) +
      geom_path(data=paths_2, aes(x=x,y=y), colour=&#39;black&#39;) +
      geom_path(data=paths_3, aes(x=x,y=y), colour=&#39;black&#39;)
  
    eval_plot&lt;-eval_plot + 
      annotate(&quot;text&quot;,x=1.5,y=max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/10)+(ylim_range/15),label=paste0(round((tmp_comp_all_tab_i$R_diff/tmp_comp_all_tab_i$Model_2_R)[1] * 100,1),&#39;%; p&#39;, tmp_comp_all_tab_i$R_diff_pval[1],&#39;   &#39;), size=4) +
      annotate(&quot;text&quot;,x=2,y=max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/10)+(ylim_range/15)+(1*(ylim_range/5)),label=paste0(round((tmp_comp_all_tab_i$R_diff/tmp_comp_all_tab_i$Model_2_R)[2] * 100,1),&#39;%; p&#39;, tmp_comp_all_tab_i$R_diff_pval[2]), size=4) +
      annotate(&quot;text&quot;,x=2.5,y=max(tmp_eval_all_tab_i$R+tmp_eval_all_tab_i$SE)+(ylim_range/10)+(ylim_range/15)+(2*(ylim_range/5)),label=paste0(round((tmp_comp_all_tab_i$R_diff/tmp_comp_all_tab_i$Model_2_R)[3] * 100,1),&#39;%; p&#39;, tmp_comp_all_tab_i$R_diff_pval[3]), size=4)
    
    eval_plot_brief_all_list[[i]]&lt;-eval_plot

}

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/test_&#39;,test,&#39;_brief.png&#39;), units=&#39;px&#39;, res=300, width=3750, height=1300)
print(plot_grid(plotlist=eval_plot_brief_all_list, ncol=3))
dev.off()

#########
# Plot results testing gain from seconadary PGS over target PGS alone
#########

test&lt;-7  
eval_plot_all_list&lt;-list()
for(i in 1:length(target_gwas)){
  eval_plot_list&lt;-list()
  for(type in c(&#39;rho_restricted_unweighted&#39;,&#39;rho_fdr_restricted_unweighted&#39;,&#39;h2_restricted_unweighted&#39;,&#39;rho_restricted_rho_weighted&#39;,&#39;rho_fdr_restricted_rho_weighted&#39;,&#39;h2_restricted_rho_weighted&#39;,&#39;rho_restricted_rho_h2_weighted&#39;,&#39;rho_fdr_restricted_rho_h2_weighted&#39;,&#39;h2_restricted_rho_h2_weighted&#39;)){
    target_pheno_i&lt;-target_pheno[i]
    tmp_eval&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_&#39;,test,&#39;_&#39;,type,&#39;.pred_eval.txt&#39;))
    tmp_comp&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_&#39;,test,&#39;_&#39;,type,&#39;.pred_comp.txt&#39;))
    
    tmp_comp$R_diff_pval&lt;-format(tmp_comp$R_diff_pval, scientific = TRUE, digits = 2)
    tmp_comp$R_diff_pval[as.numeric(tmp_comp$R_diff_pval) != 0]&lt;-paste0(&#39;=&#39;,tmp_comp$R_diff_pval[as.numeric(tmp_comp$R_diff_pval) != 0])
    tmp_comp$R_diff_pval[as.numeric(tmp_comp$R_diff_pval) == 0]&lt;-&#39;&lt;1e-300&#39;
    
    Filter&lt;-gsub(&#39;_restricted.*&#39;,&#39;&#39;, type)
    if(Filter == &#39;rho&#39;){Filter&lt;-&#39;rho-p &lt; 0.05&#39;}
    if(Filter == &#39;rho.fdr&#39;){Filter&lt;-&#39;rho-fdr.p &lt; 0.05&#39;}
    if(Filter == &#39;h2&#39;){Filter&lt;-&#39;hsq-p &lt; 0.05&#39;}

    Weighting&lt;-gsub(&#39;_group&#39;,&#39;&#39;,gsub(&#39;.*restricted_&#39;,&#39;&#39;, type))
    if(Weighting == &#39;unweighted&#39;){Weighting&lt;-&#39;Unweighted&#39;}
    if(Weighting == &#39;rho_weighted&#39;){Weighting&lt;-&#39;local rg\nweighted&#39;}
    if(Weighting == &#39;rho_h2_weighted&#39;){Weighting&lt;-&#39;local rg +\nhsq weighted&#39;}

    tmp_eval$Model&lt;-c(&#39;Target PGS +\nSecondary PGS&#39;,paste0(Filter,&#39;\n&#39;,Weighting),&#39;All&#39;)
    tmp_eval$Model&lt;-factor(tmp_eval$Model, levels=tmp_eval$Model)
  
    eval_plot&lt;-ggplot(tmp_eval, aes(x=Model, y=R)) +
      geom_bar(data=tmp_eval, aes(fill=Model), stat=&quot;identity&quot;, position=position_dodge()) +
      geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2, position=position_dodge(.9)) +
      labs(title=paste0(&#39;Target: &#39;,target_pheno_i), y=&quot;R (SE)&quot;, x=&#39;&#39;) +
      theme_half_open() +
      background_grid() +
      theme(legend.position = &quot;none&quot;)
  
    if(min(tmp_eval$R-tmp_eval$SE) &lt; 0){
      ylim_range&lt;-max(tmp_eval$R+tmp_eval$SE)-min(tmp_eval$R-tmp_eval$SE)
    } else {
      ylim_range&lt;-max(tmp_eval$R+tmp_eval$SE)
    }
    
    paths_1&lt;-data.frame(x=c(1,1,1.95,1.95),
                        y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)))
    paths_2&lt;-data.frame(x=c(2.05,2.05,3,3),
                        y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)))
    paths_3&lt;-data.frame(x=c(1,1,3,3),
                        y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)+(ylim_range/5),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5),
                            max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)+(ylim_range/5)))
    
    eval_plot&lt;-eval_plot + 
      geom_path(data=paths_1, aes(x=x,y=y), colour=&#39;black&#39;) +
      geom_path(data=paths_2, aes(x=x,y=y), colour=&#39;black&#39;) +
      geom_path(data=paths_3, aes(x=x,y=y), colour=&#39;black&#39;)
  
    eval_plot&lt;-eval_plot + 
      annotate(&quot;text&quot;,x=1.5,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[4] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[4],&#39;   &#39;), size=4) +
      annotate(&quot;text&quot;,x=2.5,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[8] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[8]), size=4) +
      annotate(&quot;text&quot;,x=2,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[7] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[7]), size=4)
  
    eval_plot_list[[paste0(i,&#39;_&#39;,test,&#39;_&#39;,type)]]&lt;-eval_plot
    
    if(grepl(&#39;rho_restricted&#39;,type)){
      tmp_eval$Model&lt;-gsub(&#39;rho-p &lt; 0.05\n&#39;,&#39;&#39;,tmp_eval$Model)  
      tmp_eval$Model&lt;-factor(tmp_eval$Model, levels=tmp_eval$Model)
      
      eval_plot&lt;-ggplot(tmp_eval, aes(x=Model, y=R)) +
        geom_bar(data=tmp_eval, aes(fill=Model), stat=&quot;identity&quot;, position=position_dodge()) +
        geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2, position=position_dodge(.9)) +
        labs(title=paste0(&#39;Target: &#39;,target_pheno_i), y=&quot;R (SE)&quot;, x=&#39;&#39;) +
        theme_half_open() +
        background_grid() +
        theme(legend.position = &quot;none&quot;)
    
      if(min(tmp_eval$R-tmp_eval$SE) &lt; 0){
        ylim_range&lt;-max(tmp_eval$R+tmp_eval$SE)-min(tmp_eval$R-tmp_eval$SE)
      } else {
        ylim_range&lt;-max(tmp_eval$R+tmp_eval$SE)
      }
      
      paths_1&lt;-data.frame(x=c(1,1,1.95,1.95),
                          y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)))
      paths_2&lt;-data.frame(x=c(2.05,2.05,3,3),
                          y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)))
      paths_3&lt;-data.frame(x=c(1,1,3,3),
                          y=c(max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)+(ylim_range/5),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5),
                              max(tmp_eval$R+tmp_eval$SE)+(ylim_range/20)+(ylim_range/5)))
      
      eval_plot&lt;-eval_plot + 
        geom_path(data=paths_1, aes(x=x,y=y), colour=&#39;black&#39;) +
        geom_path(data=paths_2, aes(x=x,y=y), colour=&#39;black&#39;) +
        geom_path(data=paths_3, aes(x=x,y=y), colour=&#39;black&#39;)
    
      eval_plot&lt;-eval_plot + 
        annotate(&quot;text&quot;,x=1.5,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[4] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[4],&#39;   &#39;), size=4) +
        annotate(&quot;text&quot;,x=2.5,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[8] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[8]), size=4) +
        annotate(&quot;text&quot;,x=2,y=max(tmp_eval$R+tmp_eval$SE)+(ylim_range/10)+(ylim_range/5)+(ylim_range/10),label=paste0(round((tmp_comp$R_diff/tmp_comp$Model_2_R)[7] * 100,1),&#39;%; p&#39;, tmp_comp$R_diff_pval[7]), size=4)
        
      eval_plot_all_list[[paste0(i,&#39;_&#39;,test,&#39;_&#39;,type)]]&lt;-eval_plot
    }
  }
  png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/&#39;,target_pheno_i,&#39;/test_&#39;,test,&#39;.png&#39;), units=&#39;px&#39;, res=300, width=5000, height=3000)
  print(plot_grid(plotlist=eval_plot_list, ncol=3))
  dev.off()
}

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/local_rg_pgs/assoc/test_&#39;,test,&#39;.png&#39;), units=&#39;px&#39;, res=300, width=4750, height=3000)
print(plot_grid(plotlist=eval_plot_all_list, ncol=3))
dev.off()</code></pre>
</details>
<details>
<p><summary>Show gain provided by secondary PGS over target PGS</summary></p>
<p><img src="Images/local_rg_weighted_pgs/test_1.png" /></p>
</details>
<details>
<p><summary>Show association between phenotype and each reweighted PGS</summary></p>
<p><img src="Images/local_rg_weighted_pgs/test_2_assoc_with_unres.png" /></p>
</details>
<details>
<p><summary>Show relationship between gain in secondary PGS prediction and inconsistency of local rG directions</summary></p>
<p><img src="Images/local_rg_weighted_pgs/test_2_lava.png" /></p>
</details>
<details>
<p><summary>Show gain from local hsq/rG weighted secondary PGS over standard secondary PGS</summary></p>
<p><img src="Images/local_rg_weighted_pgs/test_6_brief.png" /></p>
</details>
<details>
<p><summary>Show gain from local hsq/rG weighted PGS over target PGS and standard secondary PGS</summary></p>
<p><img src="Images/local_rg_weighted_pgs/test_3_brief.png" /></p>
</details>
<hr />
<hr />
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
