<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>CrossMeta</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GenoPred</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">CrossMeta</h1>

</div>


<style>
p.caption {
  font-size: 1.5em;
}
</style>
<style type="text/css">
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
</style>
<hr />
<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>This script documents the evaluation and comparison of different approaches for combining GWAS from different populations to improve polygenic prediction.</p>
<p>We evaluate and compare the following approaches: - Modelling PRS from each GWAS, and a PRS from standard inverse-variance meta-GWAS - Modelling PRS from each GWAS, and a PRS from genome-wide reweighted meta-GWAS - Modelling PRS from each GWAS, and a PRS from SNP-level reweighted meta-GWAS - Modelling PRS from PRS-CSx</p>
<p>We focus on depression, combining GWAS summary statistics from the following studies:</p>
<ul>
<li>EUR
<ul>
<li>PGC2 MDD (incl. 23andMe)</li>
<li>Biobank Japan (BBJ)</li>
<li>CONVERGE</li>
<li>23andMe</li>
<li>PGC Cross Population (Karolina)</li>
</ul></li>
<li>AMR
<ul>
<li>23andMe</li>
</ul></li>
<li>AFR
<ul>
<li>23andMe</li>
</ul></li>
</ul>
<p>We also include a comparison of methods using Height and BMI, using the following GWAS summary statistics:</p>
<ul>
<li>Height
<ul>
<li>EUR
<ul>
<li>Wood et al</li>
</ul></li>
<li>EAS
<ul>
<li>BBJ</li>
</ul></li>
</ul></li>
<li>BMI
<ul>
<li>EUR
<ul>
<li>Locke et al</li>
</ul></li>
<li>EAS
<ul>
<li>BBJ</li>
</ul></li>
</ul></li>
</ul>
<p>We use the UK Biobank sample as our target sample, split into the five 1KG reference super-populations.</p>
<hr />
</div>
<div id="methods" class="section level1">
<h1><span class="header-section-number">2</span> Methods</h1>
<hr />
<div id="uk-biobank-ukb" class="section level2">
<h2><span class="header-section-number">2.1</span> UK Biobank (UKB)</h2>
<hr />
<div id="ancestry-identification" class="section level3">
<h3><span class="header-section-number">2.1.1</span> Ancestry identification</h3>
<p>We used the observed genotype data in UKB to classify individuals into the five 1000 Genomes (1KG) reference super-populations. This process has been previously documented <a href="https://opain.github.io/UKB-GenoPrep/ancestry_identification.html">here</a>. In summary, genetic principal components (PCs) were derived in the 1KG Phase 3 reference, and then first 6 were used to created an elastic net model predicting super-population membership. Then the reference PCs were projected into the UKB sample, and the elastic net model trained in 1KG was the applied in UKB to predict super-population membership. Subsequently, within super-population quality control was performed, which has been previously documented <a href="https://opain.github.io/UKB-GenoPrep/quality_control.html">here</a>. In summary, PCs were generated in each super-population, k-means clustering was then performed to define centroids, and outliers from centroids were removed. Furthermore, individuals were removed if there were reported by UKB as heterozygosity or missingness outliers, genotype-phenotype sex mismatch, or genotype missingness &gt; 0.02 (after removal of SNPs with missingness &gt; 0.05).</p>
<hr />
</div>
<div id="phenotype-definition" class="section level3">
<h3><span class="header-section-number">2.1.2</span> Phenotype definition</h3>
<p>Depression: To maximise statistical power, we used a continuous measure of depression, termed the interalising factor, previously described <a href="https://pubmed.ncbi.nlm.nih.gov/32519625/">here</a>. In summary, the internalising factor is the first factor of depressive symptoms assessed in UKB.</p>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code># Create symbolic links to the three phenotype file with codes corresponding to GWAS codes (i.e. MDD, Height, BMI)
mkdir -p /users/k1806347/brc_scratch/Analyses/CrossMeta/Phenotype
ln -s /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/UKB.InternalisingFactor.UpdateIDs.txt /users/k1806347/brc_scratch/Analyses/CrossMeta/Phenotype/MDD.txt
ln -s /users/k1806347/brc_scratch/Data/UKBB/Phenotype/Height/UKBB_Height.score.UpdateIDs.pheno /users/k1806347/brc_scratch/Analyses/CrossMeta/Phenotype/Height.txt
ln -s /users/k1806347/brc_scratch/Data/UKBB/Phenotype/BMI/UKBB_BMI.score.UpdateIDs.pheno /users/k1806347/brc_scratch/Analyses/CrossMeta/Phenotype/BMI.txt
</code></pre>
</details>
<hr />
</div>
</div>
<div id="gwas-summary-statistics" class="section level2">
<h2><span class="header-section-number">2.2</span> GWAS summary statistics</h2>
<p>Below is a table listing details of the GWAS summary statistics used in this study:</p>
<details>
<p><summary>Show code</summary></p>
<pre class="r"><code># Depression
gwas&lt;-data.frame(code=c(&#39;CONVERGE_MDD_EAS&#39;),
                 discovery=c(&#39;CONVERGE&#39;),
                 phenotype=c(&#39;MDD&#39;),
                 population=c(&#39;EAS&#39;),
                 pmid=c(26176920),
                 N=c(10640),
                 Ncas=c(5303),
                 Ncon=c(5337))

gwas&lt;-rbind(gwas,data.frame( code=c(&#39;BBJ_MDD_EAS&#39;),
                             discovery=c(&#39;BBJ&#39;),
                             phenotype=c(&#39;MDD&#39;),
                             population=c(&#39;EAS&#39;),
                             pmid=c(34594039),
                             N=c(178630),
                             Ncas=c(836),
                             Ncon=c(177794)))

gwas&lt;-rbind(gwas,data.frame( code=c(&#39;PGC2_MDD_EUR&#39;),
                             discovery=c(&#39;PGC2&#39;),
                             phenotype=c(&#39;MDD&#39;),
                             population=c(&#39;EUR&#39;),
                             pmid=c(29700475),
                             N=c(143265),
                             Ncas=c(45591),
                             Ncon=c(97674)))

# Height
gwas&lt;-rbind(gwas,data.frame( code=c(&#39;Wood_Height_EUR&#39;),
                             discovery=c(&#39;Wood&#39;),
                             phenotype=c(&#39;Height&#39;),
                             population=c(&#39;EUR&#39;),
                             pmid=c(25282103),
                             N=c(253288),
                             Ncas=c(NA),
                             Ncon=c(NA)))

gwas&lt;-rbind(gwas,data.frame( code=c(&#39;BBJ_Height_EAS&#39;),
                             discovery=c(&#39;BBJ&#39;),
                             phenotype=c(&#39;Height&#39;),
                             population=c(&#39;EAS&#39;),
                             pmid=c(26367797),
                             N=c(159095),
                             Ncas=c(NA),
                             Ncon=c(NA)))

# BMI
gwas&lt;-rbind(gwas,data.frame( code=c(&#39;Locke_BMI_EUR&#39;),
                             discovery=c(&#39;Locke&#39;),
                             phenotype=c(&#39;BMI&#39;),
                             population=c(&#39;EUR&#39;),
                             pmid=c(25673413),
                             N=c(339224),
                             Ncas=c(NA),
                             Ncon=c(NA)))

gwas&lt;-rbind(gwas,data.frame( code=c(&#39;BBJ_BMI_EAS&#39;),
                             discovery=c(&#39;BBJ&#39;),
                             phenotype=c(&#39;BMI&#39;),
                             population=c(&#39;EAS&#39;),
                             pmid=c(28892062),
                             N=c(158284),
                             Ncas=c(NA),
                             Ncon=c(NA)))

dir.create(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats&#39;, recursive=T)
write.csv(gwas, &#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/gwas_list.txt&#39;, row.names=F, quote=F)</code></pre>
<table>
<caption>List of GWAS used</caption>
<thead>
<tr class="header">
<th align="left">code</th>
<th align="left">discovery</th>
<th align="left">phenotype</th>
<th align="left">population</th>
<th align="right">pmid</th>
<th align="right">N</th>
<th align="right">Ncas</th>
<th align="right">Ncon</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">CONVERGE_MDD_EAS</td>
<td align="left">CONVERGE</td>
<td align="left">MDD</td>
<td align="left">EAS</td>
<td align="right">26176920</td>
<td align="right">10640</td>
<td align="right">5303</td>
<td align="right">5337</td>
</tr>
<tr class="even">
<td align="left">BBJ_MDD_EAS</td>
<td align="left">BBJ</td>
<td align="left">MDD</td>
<td align="left">EAS</td>
<td align="right">34594039</td>
<td align="right">178630</td>
<td align="right">836</td>
<td align="right">177794</td>
</tr>
<tr class="odd">
<td align="left">PGC2_MDD_EUR</td>
<td align="left">PGC2</td>
<td align="left">MDD</td>
<td align="left">EUR</td>
<td align="right">29700475</td>
<td align="right">143265</td>
<td align="right">45591</td>
<td align="right">97674</td>
</tr>
<tr class="even">
<td align="left">Wood_Height_EUR</td>
<td align="left">Wood</td>
<td align="left">Height</td>
<td align="left">EUR</td>
<td align="right">25282103</td>
<td align="right">253288</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
<tr class="odd">
<td align="left">BBJ_Height_EAS</td>
<td align="left">BBJ</td>
<td align="left">Height</td>
<td align="left">EAS</td>
<td align="right">26367797</td>
<td align="right">159095</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
<tr class="even">
<td align="left">Locke_BMI_EUR</td>
<td align="left">Locke</td>
<td align="left">BMI</td>
<td align="left">EUR</td>
<td align="right">25673413</td>
<td align="right">339224</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
<tr class="odd">
<td align="left">BBJ_BMI_EAS</td>
<td align="left">BBJ</td>
<td align="left">BMI</td>
<td align="left">EAS</td>
<td align="right">28892062</td>
<td align="right">158284</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
</tbody>
</table>
<p><strong>Note</strong>: PGC2_MDD_EUR does not include data from 23andMe or UKB.</p>
</details>
<hr />
<div id="download-and-format-gwas" class="section level3">
<h3><span class="header-section-number">2.2.1</span> Download and format GWAS</h3>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code>mkdir /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/
mkdir /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/
mkdir /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/
</code></pre>
<details>
<p><summary>PGC_MDD_EAS</summary></p>
<pre class="r"><code>#####
# PGC MDD EAS
#####
# https://pubmed.ncbi.nlm.nih.gov/34586374/
# Downloaded from https://figshare.com/ndownloader/files/31424374
# I requested access to sumstats exluding UKB from Xiangrui Meng

# I will format them, and then QC
library(data.table)
sum_auto&lt;-fread(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/jamapsy_Giannakopoulou_2021_exclude_whi_23andMe_ukb.txt.gz&#39;)

sum_auto&lt;-sum_auto[,c(&#39;MarkerName&#39;,&#39;chr&#39;,&#39;pos&#39;,&#39;Allele1&#39;,&#39;Allele2&#39;,&#39;Freq1&#39;,&#39;Effect&#39;,&#39;StdErr&#39;,&#39;P.SE&#39;),with=F]
names(sum_auto)&lt;-c(&#39;SNP&#39;,&#39;CHR&#39;,&#39;POS&#39;,&#39;A1&#39;,&#39;A2&#39;,&#39;FREQ&#39;,&#39;BETA&#39;,&#39;SE&#39;,&#39;P&#39;)

# N is derived from supp etable 2 (194548-(133+366))
sum_auto$N&lt;-194049

sum_auto$A1&lt;-toupper(sum_auto$A1)
sum_auto$A2&lt;-toupper(sum_auto$A2)

fwrite(sum_auto, &#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/PGC_MDD_EAS&#39;, sep=&#39; &#39;, quote=F, na=&#39;NA&#39;)

system(&#39;~/brc_scratch/Software/pigz /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/PGC_MDD_EAS&#39;)</code></pre>
</details>
<details>
<p><summary>CONVERGE_MDD_EAS</summary></p>
<pre class="bash"><code>#####
# CONVERGE MDD
#####
# https://pubmed.ncbi.nlm.nih.gov/26176920/
# Downloaded from https://figshare.com/ndownloader/files/10374393

wget --no-check-certificate -O /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/CONVERGE_MDD_EAS https://figshare.com/ndownloader/files/10374393
</code></pre>
<pre class="r"><code># I will format them, and then QC
library(data.table)
sum_auto&lt;-fread(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/CONVERGE_MDD_EAS&#39;)

# The sumstats README on figshare has labelled the allele incorrectly. FRQ and OR in fact correspond to A2.
names(sum_auto)&lt;-c(&#39;CHR&#39;,&#39;POS&#39;,&#39;SNP&#39;,&#39;A2&#39;,&#39;A1&#39;,&#39;FREQ&#39;,&#39;OR&#39;,&#39;SE&#39;,&#39;P&#39;,&#39;REF_FREQ&#39;,&#39;HWE.P&#39;,&#39;INFO&#39;)

sum_auto$N&lt;-10640

fwrite(sum_auto, &#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/CONVERGE_MDD_EAS&#39;, sep=&#39; &#39;, quote=F, na=&#39;NA&#39;)

system(&#39;~/brc_scratch/Software/pigz /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/CONVERGE_MDD_EAS&#39;)</code></pre>
</details>
<details>
<p><summary>BBJ_MDD_EAS</summary></p>
<pre class="bash"><code>
wget --no-check-certificate -O /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/BBJ_MDD_EAS.zip https://pheweb.jp/download/Depression

unzip /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/BBJ_MDD_EAS.zip -d /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/BBJ_MDD_EAS
</code></pre>
<pre class="r"><code># I will format them, and then QC
library(data.table)
sum_auto&lt;-fread(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/BBJ_MDD_EAS/hum0197.v3.BBJ.Dep.v1/GWASsummary_Depression_Japanese_SakaueKanai2020.auto.txt.gz&#39;)
sum_auto&lt;-sum_auto[,c(&#39;CHR&#39;,&#39;POS&#39;,&#39;SNPID&#39;,&#39;Allele1&#39;,&#39;Allele2&#39;,&#39;AF_Allele2&#39;,&#39;imputationInfo&#39;,&#39;N&#39;,&#39;BETA&#39;,&#39;SE&#39;,&#39;p.value&#39;,&#39;AF.Cases&#39;,&#39;AF.Controls&#39;),with=F]

# Note. Effect allele is A2. Flip the alleles so A1 is the allele BETA/allele frequencies correspond to.
sum_auto$Allele3&lt;-sum_auto$Allele2
sum_auto$Allele2&lt;-sum_auto$Allele1
sum_auto$Allele1&lt;-sum_auto$Allele3
sum_auto$Allele3&lt;-NULL

names(sum_auto)&lt;-c(&#39;CHR&#39;,&#39;POS&#39;,&#39;SNP&#39;,&#39;A1&#39;,&#39;A2&#39;,&#39;FREQ&#39;,&#39;INFO&#39;,&#39;N&#39;,&#39;BETA&#39;,&#39;SE&#39;,&#39;p.value&#39;,&#39;FREQ.Cases&#39;,&#39;FREQ.Controls&#39;)

sum_auto$NCAS&lt;-836
sum_auto$NCON&lt;-177794

fwrite(sum_auto, &#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/BBJ_MDD_EAS&#39;, sep=&#39; &#39;, quote=F, na=&#39;NA&#39;)

system(&#39;~/brc_scratch/Software/pigz /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/BBJ_MDD_EAS&#39;)</code></pre>
</details>
<details>
<p><summary>PGC2_MDD_EUR</summary></p>
<pre class="bash"><code>#######
# Wray et al. MDD (excl. 23andMe and excl. UKB)
#######
# https://www.nature.com/articles/s41588-018-0090-3

cp /scratch/groups/ukbiobank/sumstats/cleaned/DEPR07.gz /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/PGC2_MDD_EUR.gz
</code></pre>
</details>
<hr />
<details>
<p><summary>BBJ_Height_EAS</summary></p>
<pre class="bash"><code># Download the sumstats
wget --no-check-certificate -O /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/BBJ_Height_EAS.gz http://jenger.riken.jp/135analysisresult_qtl_download/</code></pre>
<pre class="r"><code># Format the sumstats for qc script
library(data.table)

# Read in reference data
ref&lt;-fread(&#39;/users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.GW.bim&#39;)
ref$V3&lt;-NULL

# We need to insert RSID, N and update column names
sum_auto&lt;-fread(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/BBJ_Height_EAS.gz&#39;)

sum_auto_matched&lt;-merge(sum_auto, ref, by.x=c(&#39;CHR&#39;,&#39;POS&#39;,&#39;REF&#39;,&#39;ALT&#39;), by.y=c(&#39;V1&#39;,&#39;V4&#39;,&#39;V5&#39;,&#39;V6&#39;))
sum_auto_swapped&lt;-merge(sum_auto, ref, by.x=c(&#39;CHR&#39;,&#39;POS&#39;,&#39;REF&#39;,&#39;ALT&#39;), by.y=c(&#39;V1&#39;,&#39;V4&#39;,&#39;V6&#39;,&#39;V5&#39;))

sum_auto_withRSID&lt;-rbind(sum_auto_matched, sum_auto_swapped)

sum_auto_withRSID&lt;-sum_auto_withRSID[,c(&#39;CHR&#39;,&#39;POS&#39;,&#39;REF&#39;,&#39;ALT&#39;,&#39;ALT_freq&#39;,&#39;BETA&#39;,&#39;SE&#39;,&#39;P_INF&#39;,&#39;V2&#39;), with=F]
names(sum_auto_withRSID)&lt;-c(&#39;CHR&#39;,&#39;POS&#39;,&#39;A2&#39;,&#39;A1&#39;,&#39;FREQ&#39;,&#39;BETA&#39;,&#39;SE&#39;,&#39;P&#39;,&#39;SNP&#39;)

sum_auto_withRSID$N&lt;-159095

fwrite(sum_auto_withRSID, &#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/BBJ_Height_EAS&#39;, sep=&#39; &#39;, quote=F, na=&#39;NA&#39;)

system(&#39;~/brc_scratch/Software/pigz /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/BBJ_Height_EAS&#39;)</code></pre>
</details>
<details>
<p><summary>Wood_Height_EUR</summary></p>
<pre class="bash"><code>
cp /scratch/groups/ukbiobank/sumstats/cleaned/HEIG03.gz /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/Wood_Height_EUR.gz
</code></pre>
</details>
<hr />
<details>
<p><summary>BBJ_BMI_EAS</summary></p>
<pre class="bash"><code># Download the sumstats
wget --no-check-certificate -O /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/BBJ_BMI_EAS.gz http://jenger.riken.jp/1analysisresult_qtl_download/
</code></pre>
<pre class="r"><code>library(data.table)

# Read in reference data
ref&lt;-fread(&#39;/users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.GW.bim&#39;)
ref$V3&lt;-NULL

# We need to insert RSID, N and update column names
sum_auto&lt;-fread(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/raw/BBJ_BMI_EAS.gz&#39;)

sum_auto_matched&lt;-merge(sum_auto, ref, by.x=c(&#39;CHR&#39;,&#39;POS&#39;,&#39;REF&#39;,&#39;ALT&#39;), by.y=c(&#39;V1&#39;,&#39;V4&#39;,&#39;V5&#39;,&#39;V6&#39;))
sum_auto_swapped&lt;-merge(sum_auto, ref, by.x=c(&#39;CHR&#39;,&#39;POS&#39;,&#39;REF&#39;,&#39;ALT&#39;), by.y=c(&#39;V1&#39;,&#39;V4&#39;,&#39;V6&#39;,&#39;V5&#39;))

sum_auto_withRSID&lt;-rbind(sum_auto_matched, sum_auto_swapped)

sum_auto_withRSID&lt;-sum_auto_withRSID[,c(&#39;CHR&#39;,&#39;POS&#39;,&#39;REF&#39;,&#39;ALT&#39;,&#39;Frq&#39;,&#39;BETA&#39;,&#39;SE&#39;,&#39;P&#39;,&#39;V2&#39;), with=F]

names(sum_auto_withRSID)&lt;-c(&#39;CHR&#39;,&#39;POS&#39;,&#39;A2&#39;,&#39;A1&#39;,&#39;FREQ&#39;,&#39;BETA&#39;,&#39;SE&#39;,&#39;P&#39;,&#39;SNP&#39;)

sum_auto_withRSID$N&lt;-158284

fwrite(sum_auto_withRSID, &#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/BBJ_BMI_EAS&#39;, sep=&#39; &#39;, quote=F, na=&#39;NA&#39;)

system(&#39;~/brc_scratch/Software/pigz /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/BBJ_BMI_EAS&#39;)</code></pre>
</details>
<details>
<p><summary>Locke_BMI_EUR</summary></p>
<pre class="bash"><code>
cp /scratch/groups/ukbiobank/sumstats/cleaned/BODY04.gz /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/Locke_BMI_EUR.gz
</code></pre>
</details>
<hr />
</div>
<div id="qc-gwas" class="section level3">
<h3><span class="header-section-number">2.2.2</span> QC GWAS</h3>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code># QC the sumstats
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Pipeline_prep.config

pop=EUR
for gwas in $(echo PGC2_MDD_EUR Wood_Height_EUR Locke_BMI_EUR);do
/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/sumstat_cleaner/sumstat_cleaner.R \
  --sumstats /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/${gwas}.gz \
  --ref_plink_chr ${Geno_1KG_dir}/1KGPhase3.w_hm3.chr \
  --ref_freq_chr ${Geno_1KG_dir}/freq_files/${pop}/1KGPhase3.w_hm3.${pop}.chr \
  --output /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/${gwas}
done

pop=EAS
for gwas in $(echo PGC_MDD_EAS CONVERGE_MDD_EAS BBJ_MDD_EAS BBJ_Height_EAS BBJ_BMI_EAS);do
/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/sumstat_cleaner/sumstat_cleaner.R \
  --sumstats /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/processed/${gwas}.gz \
  --ref_plink_chr ${Geno_1KG_dir}/1KGPhase3.w_hm3.chr \
  --ref_freq_chr ${Geno_1KG_dir}/freq_files/${pop}/1KGPhase3.w_hm3.${pop}.chr \
  --output /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/${gwas}
done
</code></pre>
</details>
<hr />
</div>
</div>
<div id="cross-population-snp-annotations" class="section level2">
<h2><span class="header-section-number">2.3</span> Cross Population SNP annotations</h2>
<p>We focus on two annotations describing differences in linkage disequilibrium (LD) and allele frequency for a given SNP across populations. These include LDcorr, which is the correlation between LD r2 estimates for a given SNP across two populations, and Fst which is a measure of differences in allele frequency for a given SNP across populations.</p>
<hr />
<div id="calculate-annotations" class="section level3">
<h3><span class="header-section-number">2.3.1</span> Calculate annotations</h3>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code>pop1=EUR

for pop2 in $(echo EAS AMR AFR SAS); do
  for chr in $(seq 1 22);do
    sbatch -p brc,shared --mem 30G -n 1 --nodes 1 /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/cross_pop/cross_pop.R \
                  --ref /scratch/groups/biomarkers-brc-mh/Reference_data/1KG_Phase3/bigsnpr/1000G_phase3_common_norel \
                  --chr ${chr} \
                  --pop1 ${pop1} \
                  --pop2 ${pop2} \
                  --n_cores 1 \
                  --output /scratch/groups/biomarkers-brc-mh/Reference_data/CrossPop/Populations_${pop1}_${pop2}/chr${chr}/res
  done
done

# Create genome-wide files
pop1=EUR
for pop2 in $(echo EAS AMR AFR SAS); do
head -1 /scratch/groups/biomarkers-brc-mh/Reference_data/CrossPop/Populations_${pop1}_${pop2}/chr1/res.chr1.${pop1}.${pop2}.txt &gt; /scratch/groups/biomarkers-brc-mh/Reference_data/CrossPop/Populations_${pop1}_${pop2}/CrossPop.${pop1}.${pop2}.GW.txt
tail -n +2 -q /scratch/groups/biomarkers-brc-mh/Reference_data/CrossPop/Populations_${pop1}_${pop2}/chr*/res.chr*.${pop1}.${pop2}.txt &gt;&gt; /scratch/groups/biomarkers-brc-mh/Reference_data/CrossPop/Populations_${pop1}_${pop2}/CrossPop.${pop1}.${pop2}.GW.txt
done</code></pre>
</details>
<hr />
</div>
<div id="annotate-gwas" class="section level3">
<h3><span class="header-section-number">2.3.2</span> Annotate GWAS</h3>
<details>
<p><summary>Show code</summary></p>
<pre class="r"><code>library(data.table)

pop1=&#39;EUR&#39;
pop2=&#39;EAS&#39;

gwas&lt;-c(&#39;PGC2_MDD_EUR&#39;,&#39;Wood_Height_EUR&#39;,&#39;Locke_BMI_EUR&#39;,&#39;CONVERGE_MDD_EAS&#39;,&#39;BBJ_MDD_EAS&#39;,&#39;BBJ_Height_EAS&#39;,&#39;BBJ_BMI_EAS&#39;)

gwas_eur&lt;-gwas[grepl(paste0(&#39;_&#39;,pop1,&#39;$&#39;), gwas)]
gwas_noneur&lt;-gwas[grepl(paste0(&#39;_&#39;,pop2,&#39;$&#39;), gwas)]

dir.create(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/annotated&#39;)

for(pop2_i in pop2){
  # Read in pop1-pop2_i annotations
  annot&lt;-NULL
  for(i in 1:22){
    annot&lt;-rbind(annot, fread(paste0(&#39;/scratch/groups/biomarkers-brc-mh/Reference_data/CrossPop/Populations_&#39;,pop1,&#39;_&#39;,pop2_i,&#39;/chr&#39;,i,&#39;/res.chr&#39;,i,&#39;.&#39;,pop1,&#39;.&#39;,pop2_i,&#39;.txt&#39;)))
  }

  for(gwas_i in gwas_eur){
    # Read in gwas_i sumstats
    gwas_i_tmp&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/&#39;,gwas_i,&#39;.gz&#39;))
    
    # Insert BETA column if not present
    if(sum(&#39;BETA&#39; %in% names(gwas_i_tmp)) == 0){
      gwas_i_tmp$BETA&lt;-log(gwas_i_tmp$OR)
    }

    # Merge annotations with gwas
    gwas_annot&lt;-merge(gwas_i_tmp, annot[,c(&#39;SNP&#39;,&#39;Fst&#39;,&#39;cor_score&#39;),with=F], by=&#39;SNP&#39;)
    
    # Scale Fst and cor_score to be between 0 and 1
    gwas_annot$Fst_scaled&lt;-(gwas_annot$Fst-min(gwas_annot$Fst))/(max(gwas_annot$Fst)-min(gwas_annot$Fst))
    gwas_annot$cor_score_scaled&lt;-(gwas_annot$cor_score-min(gwas_annot$cor_score))/(max(gwas_annot$cor_score)-min(gwas_annot$cor_score))
    
    # Flip Fst so high value indicate similar MAF
    gwas_annot$Fst_scaled&lt;-1-gwas_annot$Fst_scaled
    
    # Reweight SE according to LDcorr and Fst
    for(q in c(0.5,1,1.5,2)){
      gwas_annot[[paste0(&#39;SE_Fst_&#39;,q)]]&lt;-gwas_annot$SE/sqrt(gwas_annot$Fst_scaled^q)
      gwas_annot[[paste0(&#39;SE_LDcorr_&#39;,q)]]&lt;-gwas_annot$SE/sqrt(gwas_annot$cor_score_scaled^q)
    }
    
    # Reweight SE genome-wide Ji et al. increase weight in target ancestry GWAS (EAS) by 1:5. For simplicity, we will downweight the non-target ancestry GWAS (EUR) by 1:5.
    for(k in 2:5){
      gwas_annot[[paste0(&#39;SE_k_&#39;,k)]]&lt;-gwas_annot$SE/sqrt(1/k)
    }

    # Save the results
    fwrite(gwas_annot, paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/annotated/&#39;,gwas_i), quote=F, sep=&#39; &#39;, na=&#39;NA&#39;)
    
    # Compress the sumstats
    if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/annotated/&#39;,gwas_i,&#39;.gz&#39;))){
      system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/annotated/&#39;,gwas_i,&#39;.gz&#39;))
    }
    system(paste0(&#39;~/brc_scratch/Software/pigz /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/annotated/&#39;,gwas_i))

  }
}

# For non-EUR sumstats, just make sure a BETA column is present
for(gwas_i in gwas_noneur){
  # Read in gwas_i sumstats
  gwas_i_tmp&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/&#39;,gwas_i,&#39;.gz&#39;))
  
  # Insert BETA column if not present
  if(sum(&#39;BETA&#39; %in% names(gwas_i_tmp)) == 0){
    print(gwas_i)
    
    gwas_i_tmp$BETA&lt;-log(gwas_i_tmp$OR)
  
    # Save the results
    fwrite(gwas_i_tmp, paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/&#39;,gwas_i), quote=F, sep=&#39; &#39;, na=&#39;NA&#39;)
    
    # Compress the sumstats
    if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/&#39;,gwas_i,&#39;.gz&#39;))){
      system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/&#39;,gwas_i,&#39;.gz&#39;))
    }
    system(paste0(&#39;~/brc_scratch/Software/pigz /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/&#39;,gwas_i))
  }
}</code></pre>
</details>
<hr />
</div>
</div>
<div id="characterise-associations" class="section level2">
<h2><span class="header-section-number">2.4</span> Characterise associations</h2>
<details>
<p><summary>Show code</summary></p>
<pre class="r"><code>library(data.table)
library(ggplot2)
library(cowplot)

pheno&lt;-c(&#39;MDD&#39;,&#39;MDD&#39;,&#39;MDD&#39;,&#39;Height&#39;,&#39;BMI&#39;)
samp1=c(&#39;PGC2&#39;,&#39;PGC2&#39;,&#39;PGC2&#39;,&#39;Wood&#39;,&#39;Locke&#39;)
samp2=c(&#39;PGC&#39;,&#39;CONVERGE&#39;,&#39;BBJ&#39;,&#39;BBJ&#39;,&#39;BBJ&#39;)
pop1=c(&#39;EUR&#39;,&#39;EUR&#39;,&#39;EUR&#39;,&#39;EUR&#39;,&#39;EUR&#39;)
pop2=c(&#39;EAS&#39;,&#39;EAS&#39;,&#39;EAS&#39;,&#39;EAS&#39;,&#39;EAS&#39;)

gwas_eur&lt;-unique(paste0(samp1,&#39;_&#39;,pheno,&#39;_&#39;,pop1))
gwas_eas&lt;-unique(paste0(samp2,&#39;_&#39;,pheno,&#39;_&#39;,pop2))

for(i in 1:length(gwas_eur)){
  gwas_1&lt;-gwas_eur[i]
  gwas_1_dat&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/annotated/&#39;,gwas_1,&#39;.gz&#39;))

  for(annot in c(&#39;LDcorr&#39;,&#39;Fst&#39;)){
    if(annot == &#39;LDcorr&#39;){
      var&lt;-&#39;cor_score&#39;
    } else {
      var&lt;-&#39;Fst&#39;
    }
    
    # Measure 5 and 95% quantiles
    quant&lt;-NULL
    for(q in c(0.01,0.05,0.1,0.5,0.9,0.95,0.99)){
      quant&lt;-rbind(quant, data.frame(q=q,
                                     val=quantile(gwas_1_dat[[var]], probs = q)))

      if(q &lt; 0.5){
        dir.create(paste0(&#39;/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/&#39;, gwas_1,&#39;/&#39;,annot,&#39;/bot_&#39;,q), recursive=T)
        write.table(gwas_1_dat[gwas_1_dat[[var]] &lt; quant$val[quant$q == q],]$SNP, paste0(&quot;/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/&quot;, gwas_1,&#39;/&#39;,annot,&#39;/bot_&#39;,q,&#39;/bot_&#39;,q,&#39;_&#39;,annot,&#39;.snps.txt&#39;), col.names=F, row.names=F, quote=F)
      } 
      
      if(q &gt; 0.5){
        dir.create(paste0(&#39;/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/&#39;, gwas_1,&#39;/&#39;,annot,&#39;/top_&#39;,1-q), recursive=T)
        write.table(gwas_1_dat[gwas_1_dat[[var]] &gt; quant$val[quant$q == q],]$SNP, paste0(&quot;/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/&quot;, gwas_1,&#39;/&#39;,annot,&#39;/top_&#39;,1-q,&#39;/top_&#39;,1-q,&#39;_&#39;,annot,&#39;.snps.txt&#39;), col.names=F, row.names=F, quote=F)
      }
      
      if(q == 0.5){
        dir.create(paste0(&#39;/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/&#39;, gwas_1,&#39;/&#39;,annot,&#39;/top_&#39;,q), recursive=T)
        write.table(gwas_1_dat[gwas_1_dat[[var]] &gt; quant$val[quant$q == q],]$SNP, paste0(&quot;/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/&quot;, gwas_1,&#39;/&#39;,annot,&#39;/top_&#39;,q,&#39;/top_&#39;,q,&#39;_&#39;,annot,&#39;.snps.txt&#39;), col.names=F, row.names=F, quote=F)

        dir.create(paste0(&#39;/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/&#39;, gwas_1,&#39;/&#39;,annot,&#39;/bot_&#39;,q), recursive=T)
        write.table(gwas_1_dat[gwas_1_dat[[var]] &lt; quant$val[quant$q == q],]$SNP, paste0(&quot;/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/&quot;, gwas_1,&#39;/&#39;,annot,&#39;/bot_&#39;,q,&#39;/bot_&#39;,q,&#39;_&#39;,annot,&#39;.snps.txt&#39;), col.names=F, row.names=F, quote=F)
      }

    dir.create(paste0(&#39;/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/&#39;, gwas_1,&#39;/&#39;,annot,&#39;/all&#39;), recursive=T)

    # Perform LD clumping for these subsets of SNPs
    for(chr in 1:22){
      system(paste0(&#39;/users/k1806347/brc_scratch/Software/plink1.9.sh --bfile /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr&#39;,chr,&#39; --keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/EUR_samples.keep --extract /scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/&#39;, gwas_1,&#39;/&#39;,annot,&#39;/top_0.05/top_0.05_&#39;,annot,&#39;.snps.txt --clump /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/&#39;,gwas_1,&#39;.gz --clump-p1 1 --clump-p2 1 --clump-r2 0.1 --clump-kb 250 --out /scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/&#39;, gwas_1,&#39;/&#39;,annot,&#39;/top_0.05/top_0.05_&#39;,annot,&#39;.snps.chr&#39;,chr))

      system(paste0(&#39;/users/k1806347/brc_scratch/Software/plink1.9.sh --bfile /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr&#39;,chr,&#39; --keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/EUR_samples.keep --clump /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/&#39;,gwas_1,&#39;.gz --clump-p1 1 --clump-p2 1 --clump-r2 0.1 --clump-kb 250 --out /scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/&#39;, gwas_1,&#39;/&#39;,annot,&#39;/all/all.snps.chr&#39;,chr))
      
      system(paste0(&#39;/users/k1806347/brc_scratch/Software/plink1.9.sh --bfile /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr&#39;,chr,&#39; --keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/EUR_samples.keep --extract /scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/&#39;, gwas_1,&#39;/&#39;,annot,&#39;/bot_0.05/bot_0.05_&#39;,annot,&#39;.snps.txt --clump /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/&#39;,gwas_1,&#39;.gz --clump-p1 1 --clump-p2 1 --clump-r2 0.1 --clump-kb 250 --out /scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/&#39;, gwas_1,&#39;/&#39;,annot,&#39;/bot_0.05/bot_0.05_&#39;,annot,&#39;.snps.chr&#39;,chr))
    }
    
    clump_top5&lt;-NULL
    for(chr in 1:22){
      if(file.exists(paste0(&#39;/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/&#39;, gwas_1,&#39;/&#39;,annot,&#39;/top_0.05/top_0.05_&#39;,annot,&#39;.snps.chr&#39;,chr,&#39;.clumped&#39;))){
        clumped&lt;-fread(paste0(&#39;/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/&#39;, gwas_1,&#39;/&#39;,annot,&#39;/top_0.05/top_0.05_&#39;,annot,&#39;.snps.chr&#39;,chr,&#39;.clumped&#39;))
        
        clumped_SNPs&lt;-clumped$SNP
          clump_top5&lt;-c(clump_top5, clumped_SNPs)
        
          print(chr)
      }
    }
    
    clump_all&lt;-NULL
    for(chr in 1:22){
      if(file.exists(paste0(&#39;/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/&#39;, gwas_1,&#39;/&#39;,annot,&#39;/all/all.snps.chr&#39;,chr,&#39;.clumped&#39;))){
        clumped&lt;-fread(paste0(&#39;/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/&#39;, gwas_1,&#39;/&#39;,annot,&#39;/all/all.snps.chr&#39;,chr,&#39;.clumped&#39;))
        
        clumped_SNPs&lt;-clumped$SNP
          clump_all&lt;-c(clump_all, clumped_SNPs)
        
          print(chr)
      }
    }

    clump_bot5&lt;-NULL
    for(chr in 1:22){
        if(file.exists(paste0(&#39;/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/&#39;, gwas_1,&#39;/&#39;,annot,&#39;/bot_0.05/bot_0.05_&#39;,annot,&#39;.snps.chr&#39;,chr,&#39;.clumped&#39;))){
clumped&lt;-fread(paste0(&#39;/scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/&#39;, gwas_1,&#39;/&#39;,annot,&#39;/bot_0.05/bot_0.05_&#39;,annot,&#39;.snps.chr&#39;,chr,&#39;.clumped&#39;))
      
        clumped_SNPs&lt;-clumped$SNP
        clump_bot5&lt;-c(clump_bot5, clumped_SNPs)
        
        print(chr)
        }
    }
    
    gwas_1_dat$top5_indep&lt;-F
    gwas_1_dat$top5_indep[(gwas_1_dat$SNP %in% clump_top5)]&lt;-T
    gwas_1_dat$all_indep&lt;-F
    gwas_1_dat$all_indep[(gwas_1_dat$SNP %in% clump_all)]&lt;-T
    gwas_1_dat$bot5_indep&lt;-F
    gwas_1_dat$bot5_indep[(gwas_1_dat$SNP %in% clump_bot5)]&lt;-T
    
    # Now estimate pearson correlation of BETAs across populations
    gwas_eas_for_gwas_1&lt;-gwas_eas[grepl(gsub(&#39;.*_&#39;,&#39;&#39;,gsub(&#39;_EUR&#39;,&#39;&#39;,gwas_1)),gwas_eas)]
    
    for(j in 1:length(gwas_eas_for_gwas_1)){
      gwas_2&lt;-gwas_eas_for_gwas_1[j]
      gwas_2_dat&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/&#39;,gwas_2,&#39;.gz&#39;))

      if(sum(names(gwas_2_dat) == &#39;BETA&#39;) != 1){
        gwas_2_dat$BETA&lt;-log(gwas_2_dat$OR)
      }
      
      both_match&lt;-merge(gwas_1_dat, gwas_2_dat, by=c(&#39;SNP&#39;,&#39;A1&#39;,&#39;A2&#39;))
      
      thresh_vec&lt;-c(1,0.5,0.1,0.01,0.001,1e-5,5e-8)

      cor_res_all&lt;-NULL
      
      for(pT in thresh_vec){
        both_match_pT&lt;-both_match[both_match$P.x &lt; pT,]
        if(dim(both_match_pT[both_match_pT$bot5_indep == T])[1] &gt; 5 &amp; dim(both_match_pT[both_match_pT$top5_indep == T])[1] &gt; 5){
          
          all_res&lt;-summary(lm(scale(BETA.x) ~ scale(BETA.y), data=both_match_pT[both_match_pT$all_indep == T,]))
          cor_res_all&lt;-rbind(cor_res_all,data.frame(pT=pT,
                                                    type=&#39;all&#39;,
                                                    cor=coef(all_res)[2,1],
                                                    se=coef(all_res)[2,2]))
  
          bot5_res&lt;-summary(lm(scale(BETA.x) ~ scale(BETA.y), data=both_match_pT[both_match_pT$bot5_indep == T,]))
          cor_res_all&lt;-rbind(cor_res_all,data.frame(pT=pT,
                                                    type=&#39;bot5&#39;,
                                                    cor=coef(bot5_res)[2,1],
                                                    se=coef(bot5_res)[2,2]))
  
          top5_res&lt;-summary(lm(scale(BETA.x) ~ scale(BETA.y), data=both_match_pT[both_match_pT$top5_indep == T,]))
          cor_res_all&lt;-rbind(cor_res_all,data.frame(pT=pT,
                                                    type=&#39;top5&#39;,
                                                    cor=coef(top5_res)[2,1],
                                                    se=coef(top5_res)[2,2]))
        }
      }
     
      dir.create(paste0(&#39;/scratch/users/k1806347/Analyses/CrossMeta/sign_test/res/&#39;,gwas_1,&#39;_vs_&#39;,gwas_2,&#39;/&#39;,annot), recursive=T)
      write.table(cor_res_all, paste0(&#39;/scratch/users/k1806347/Analyses/CrossMeta/sign_test/res/&#39;,gwas_1,&#39;_vs_&#39;,gwas_2,&#39;/&#39;,annot,&#39;/cor_res.txt&#39;), col.names=T, row.names=F, quote=F)
      }
    }
  }
}

cor_res_all_all&lt;-NULL
for(i in 1:length(gwas_eur)){
  gwas_1&lt;-gwas_eur[i]
  for(annot in c(&#39;LDcorr&#39;,&#39;Fst&#39;)){
    gwas_eas_for_gwas_1&lt;-gwas_eas[grepl(gsub(&#39;.*_&#39;,&#39;&#39;,gsub(&#39;_EUR&#39;,&#39;&#39;,gwas_1)),gwas_eas)]

    for(j in 1:length(gwas_eas_for_gwas_1)){
      gwas_2&lt;-gwas_eas_for_gwas_1[j]
      cor_res_all&lt;-fread(paste0(&#39;/scratch/users/k1806347/Analyses/CrossMeta/sign_test/res/&#39;,gwas_1,&#39;_vs_&#39;,gwas_2,&#39;/&#39;,annot,&#39;/cor_res.txt&#39;))
      cor_res_all$annot&lt;-annot
      cor_res_all$test&lt;-paste0(gwas_1,&#39;_vs_&#39;,gwas_2)
      
      cor_res_all_all&lt;-rbind(cor_res_all_all, cor_res_all)
    }
    
    cor_res_all_all$type&lt;-factor(cor_res_all_all$type, levels=c(&#39;bot5&#39;,&#39;all&#39;,&#39;top5&#39;))
    cor_res_all_all$pT&lt;-factor(cor_res_all_all$pT, levels=unique(cor_res_all_all$pT))
  }
}

cor_res_all_all$test&lt;-gsub(&#39;vs&#39;,&#39;vs\n&#39;,gsub(&#39;_&#39;,&#39; &#39;,cor_res_all_all$test))
cor_res_all_all$test&lt;-factor(cor_res_all_all$test, levels=rev(unique(cor_res_all_all$test)))

cor_res_all_all$test_brief&lt;-NA
cor_res_all_all$test_brief[grepl(&#39;Height&#39;, cor_res_all_all$test)]&lt;-&#39;Height&#39;
cor_res_all_all$test_brief[grepl(&#39;BMI&#39;, cor_res_all_all$test)]&lt;-&#39;BMI&#39;
cor_res_all_all$test_brief&lt;-factor(cor_res_all_all$test_brief, levels=rev(unique(cor_res_all_all$test_brief)))
cor_res_all_all$pT&lt;-factor(cor_res_all_all$pT, levels=rev(unique(cor_res_all_all$pT)))

png(paste0(&#39;/scratch/users/k1806347/Analyses/CrossMeta/sign_test/res/cor_plot_Height_BMI.png&#39;), units=&#39;px&#39;, res=300, width=2000, height=1500)
  print(ggplot(cor_res_all_all[grepl(&#39;BMI|Height&#39;,cor_res_all_all$test),], aes(x=pT, y=cor, fill=type)) +
    geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
    geom_errorbar(aes(ymin=cor-se, ymax=cor+se), width=.2,position=position_dodge(.9)) +
    theme_half_open() +
    background_grid() +
    labs(y=&quot;BETA correlation (SE)&quot;, fill=&#39;&#39;) +
    theme(axis.text.x = element_text(angle = 45, hjust=1)) +
    facet_grid(vars(test_brief), vars(annot), scales = &quot;free&quot;))
dev.off()

png(paste0(&#39;/scratch/users/k1806347/Analyses/CrossMeta/sign_test/res/cor_plot_MDD.png&#39;), units=&#39;px&#39;, res=300, width=2000, height=2100)
  print(ggplot(cor_res_all_all[grepl(&#39;MDD&#39;,cor_res_all_all$test),], aes(x=pT, y=cor, fill=type)) +
    geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
    geom_errorbar(aes(ymin=cor-se, ymax=cor+se), width=.2,position=position_dodge(.9)) +
    theme_half_open() +
    background_grid() +
    labs(y=&quot;BETA correlation (SE)&quot;, fill=&#39;&#39;) +
    theme(axis.text.x = element_text(angle = 45, hjust=1)) +
    facet_grid(vars(test), vars(annot), scales = &quot;free&quot;))
dev.off()</code></pre>
</details>
<hr />
</div>
<div id="meta-gwas" class="section level2">
<h2><span class="header-section-number">2.5</span> Meta-GWAS</h2>
<p>We want to meta-analyse the EUR GWAS sumstats with each of the EAS sumstats</p>
<hr />
<div id="standard-inverse-variance-meta-gwas" class="section level3">
<h3><span class="header-section-number">2.5.1</span> Standard inverse variance meta-GWAS</h3>
<p>This is performing standard inverse-variance meta-GWAS.</p>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code>
# Create METAL scripts
mkdir -p /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/scripts
mkdir -p /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results

pheno=(MDD MDD MDD Height BMI)
samp1=(PGC2 PGC2 PGC2 Wood Locke)
samp2=(PGC CONVERGE BBJ BBJ BBJ)
pop1=(EUR EUR EUR EUR EUR)
pop2=(EAS EAS EAS EAS EAS)

# use for loop to read all values and indexes
for (( i=0; i&lt;${#pheno[@]}; i++ ));do

cat &lt;&lt;EOF &gt; /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/scripts/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_metal_standard.sh

#!/bin/sh
SCHEME STDERR
CUSTOMVARIABLE TotalSampleSize

MARKER SNP
ALLELE A1 A2
EFFECT BETA
PVALUE P 
LABEL TotalSampleSize as N
STDERR SE 

PROCESS /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/annotated/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}.gz

MARKER SNP
ALLELE A1 A2
EFFECT BETA
PVALUE P 
LABEL TotalSampleSize as N
STDERR SE 

PROCESS /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/${samp2[${i}]}_${pheno[${i}]}_${pop2[${i}]}.gz

OUTFILE /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_metal_standard .tbl
ANALYZE 
QUIT

EOF

done

# Run scripts
for (( i=0; i&lt;${#pheno[@]}; i++ ));do
sbatch -p brc,shared /users/k1806347/brc_scratch/Software/metal.sh /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/scripts/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_metal_standard.sh
done
</code></pre>
</details>
<hr />
</div>
<div id="genome-wide-reweighted-meta-gwas" class="section level3">
<h3><span class="header-section-number">2.5.2</span> Genome-wide reweighted meta-GWAS</h3>
<p>This involves modifying the SEs of all SNPs from a given GWAS prior to performing standard inverse-variance meta-GWAS.</p>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code>
# Create METAL scripts
pheno=(MDD MDD MDD Height BMI)
samp1=(PGC2 PGC2 PGC2 Wood Locke)
samp2=(PGC CONVERGE BBJ BBJ BBJ)
pop1=(EUR EUR EUR EUR EUR)
pop2=(EAS EAS EAS EAS EAS)

integers=(2 3 4 5)

# use for loop to read all values and indexes
for (( i=0; i&lt;${#pheno[@]}; i++ ));do
for k in ${integers[@]};do

cat &lt;&lt;EOF &gt; /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/scripts/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_metal_integer_${k}.sh

#!/bin/sh
SCHEME STDERR
CUSTOMVARIABLE TotalSampleSize

MARKER SNP
ALLELE A1 A2
EFFECT BETA
PVALUE P 
LABEL TotalSampleSize as N
STDERR SE_k_${k} 

PROCESS /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/annotated/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}.gz

MARKER SNP
ALLELE A1 A2
EFFECT BETA
PVALUE P 
LABEL TotalSampleSize as N
STDERR SE 

PROCESS /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/${samp2[${i}]}_${pheno[${i}]}_${pop2[${i}]}.gz

OUTFILE /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_metal_integer_${k} .tbl
ANALYZE 
QUIT

EOF

done
done

# Run scripts
for (( i=0; i&lt;${#pheno[@]}; i++ ));do
for k in ${integers[@]};do
sbatch -p brc,shared /users/k1806347/brc_scratch/Software/metal.sh /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/scripts/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_metal_integer_${k}.sh
done
done
</code></pre>
</details>
<hr />
</div>
<div id="snp-level-reweighted-meta-gwas" class="section level3">
<h3><span class="header-section-number">2.5.3</span> SNP-level reweighted meta-GWAS</h3>
<p>This involves modifying the SE of SNPs according to LDcorr and Fst from a given GWAS prior to performing standard inverse-variance meta-GWAS.</p>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code>
# Create METAL scripts
pheno=(MDD MDD MDD Height BMI)
samp1=(PGC2 PGC2 PGC2 Wood Locke)
samp2=(PGC CONVERGE BBJ BBJ BBJ)
pop1=(EUR EUR EUR EUR EUR)
pop2=(EAS EAS EAS EAS EAS)

annot=(LDcorr Fst)
weight=(0.5 1 1.5 2)

# use for loop to read all values and indexes
for (( i=0; i&lt;${#pheno[@]}; i++ ));do
for j in ${annot[@]};do
for q in ${weight[@]};do

cat &lt;&lt;EOF &gt; /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/scripts/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_metal_${j}_${q}.sh

#!/bin/sh
SCHEME STDERR
CUSTOMVARIABLE TotalSampleSize

MARKER SNP
ALLELE A1 A2
EFFECT BETA
PVALUE P 
LABEL TotalSampleSize as N
STDERR SE_${j}_${q} 

PROCESS /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/annotated/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}.gz

MARKER SNP
ALLELE A1 A2
EFFECT BETA
PVALUE P 
LABEL TotalSampleSize as N
STDERR SE 

PROCESS /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/${samp2[${i}]}_${pheno[${i}]}_${pop2[${i}]}.gz

OUTFILE /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_metal_${j}_${q} .tbl
ANALYZE 
QUIT

EOF

done
done
done

# Run scripts
for (( i=0; i&lt;${#pheno[@]}; i++ ));do
for j in ${annot[@]};do
for q in ${weight[@]};do
sbatch -p brc,shared /users/k1806347/brc_scratch/Software/metal.sh /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/scripts/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_metal_${j}_${q}.sh
done
done
done
</code></pre>
</details>
<hr />
</div>
<div id="reformat-metal-output" class="section level3">
<h3><span class="header-section-number">2.5.4</span> Reformat METAL output</h3>
<details>
<p><summary>Show code</summary></p>
<pre class="r"><code>library(data.table)

pheno&lt;-c(&#39;MDD&#39;,&#39;MDD&#39;,&#39;MDD&#39;,&#39;Height&#39;,&#39;BMI&#39;)
samp1=c(&#39;PGC2&#39;,&#39;PGC2&#39;,&#39;PGC2&#39;,&#39;Wood&#39;,&#39;Locke&#39;)
samp2=c(&#39;PGC&#39;,&#39;CONVERGE&#39;,&#39;BBJ&#39;,&#39;BBJ&#39;,&#39;BBJ&#39;)
pop1=c(&#39;EUR&#39;,&#39;EUR&#39;,&#39;EUR&#39;,&#39;EUR&#39;,&#39;EUR&#39;)
pop2=c(&#39;EAS&#39;,&#39;EAS&#39;,&#39;EAS&#39;,&#39;EAS&#39;,&#39;EAS&#39;)

integers=c(2,3,4,5)

annot=c(&#39;LDcorr&#39;,&#39;Fst&#39;)
weight=c(0.5,1,1.5,2)

type&lt;-c(&#39;standard&#39;,&#39;integer&#39;,&#39;annot&#39;)

for(type_i in type){
  if(type_i == &#39;standard&#39;){
    for(i in 1:length(pheno)){
      dat&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_standard1.tbl&#39;))
      
      names(dat)&lt;-c(&#39;SNP&#39;,&#39;A1&#39;,&#39;A2&#39;,&#39;BETA&#39;,&#39;SE&#39;,&#39;P&#39;,&#39;Direction&#39;,&#39;N&#39;)
      dat$A1&lt;-toupper(dat$A1)
      dat$A2&lt;-toupper(dat$A2)
      
      # Remove SNPs which were not present in one of the samples
      dat&lt;-dat[!grepl(&#39;\\?&#39;,dat$Direction),]
      
      fwrite(dat, paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_standard1.reformat.txt&#39;), sep=&#39; &#39;, na=&#39;NA&#39;, quote=F)
      
      if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_standard1.reformat.txt.gz&#39;))){
        system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_standard1.reformat.txt.gz&#39;))
      }
      system(paste0(&#39;~/brc_scratch/Software/pigz /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_standard1.reformat.txt&#39;))
      
      system(paste0(&#39;/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/sumstat_cleaner/sumstat_cleaner.R --sumstats /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_standard1.reformat.txt.gz  --ref_plink_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr --out /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_standard1.reformat.txt.cleaned&#39;))
      
    }
  }
    
  if(type_i == &#39;integer&#39;){
    for(i in 1:length(pheno)){
      for(integers_i in integers){
        dat&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_integer_&#39;,integers_i,&#39;1.tbl&#39;))
        
        names(dat)&lt;-c(&#39;SNP&#39;,&#39;A1&#39;,&#39;A2&#39;,&#39;BETA&#39;,&#39;SE&#39;,&#39;P&#39;,&#39;Direction&#39;,&#39;N&#39;)
        dat$A1&lt;-toupper(dat$A1)
        dat$A2&lt;-toupper(dat$A2)
        
        # Remove SNPs which were not present in one of the samples
        dat&lt;-dat[!grepl(&#39;\\?&#39;,dat$Direction),]
        
        fwrite(dat, paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_integer_&#39;,integers_i,&#39;1.reformat.txt&#39;), sep=&#39; &#39;, na=&#39;NA&#39;, quote=F)
        
        if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_integer_&#39;,integers_i,&#39;1.reformat.txt.gz&#39;))){
          system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_integer_&#39;,integers_i,&#39;1.reformat.txt.gz&#39;))
        }

        system(paste0(&#39;~/brc_scratch/Software/pigz /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_integer_&#39;,integers_i,&#39;1.reformat.txt&#39;))
        
        system(paste0(&#39;/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/sumstat_cleaner/sumstat_cleaner.R --sumstats /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_integer_&#39;,integers_i,&#39;1.reformat.txt.gz --ref_plink_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr --out /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_integer_&#39;,integers_i,&#39;1.reformat.txt.cleaned&#39;))

      }
    }
  }
    
  if(type_i == &#39;annot&#39;){
    for(i in 1:length(pheno)){
      for(annot_i in annot){
        for(weight_i in weight){
        dat&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_&#39;,annot_i,&#39;_&#39;,weight_i,&#39;1.tbl&#39;))
        
        names(dat)&lt;-c(&#39;SNP&#39;,&#39;A1&#39;,&#39;A2&#39;,&#39;BETA&#39;,&#39;SE&#39;,&#39;P&#39;,&#39;Direction&#39;,&#39;N&#39;)
        dat$A1&lt;-toupper(dat$A1)
        dat$A2&lt;-toupper(dat$A2)
        
        # Remove SNPs which were not present in one of the samples
        dat&lt;-dat[!grepl(&#39;\\?&#39;,dat$Direction),]
        
        fwrite(dat, paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_&#39;,annot_i,&#39;_&#39;,weight_i,&#39;1.reformat.txt&#39;), sep=&#39; &#39;, na=&#39;NA&#39;, quote=F)
        
        if(file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_&#39;,annot_i,&#39;_&#39;,weight_i,&#39;1.reformat.txt.gz&#39;))){
          system(paste0(&#39;rm /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_&#39;,annot_i,&#39;_&#39;,weight_i,&#39;1.reformat.txt.gz&#39;))
        }

        system(paste0(&#39;~/brc_scratch/Software/pigz /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_&#39;,annot_i,&#39;_&#39;,weight_i,&#39;1.reformat.txt&#39;))
        
        system(paste0(&#39;/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/sumstat_cleaner/sumstat_cleaner.R --sumstats /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_&#39;,annot_i,&#39;_&#39;,weight_i,&#39;1.reformat.txt.gz --ref_plink_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr --out /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_&#39;,annot_i,&#39;_&#39;,weight_i,&#39;1.reformat.txt.cleaned&#39;))

        }
      }
    }
  }
}</code></pre>
</details>
<hr />
</div>
</div>
<div id="polygenic-scoring" class="section level2">
<h2><span class="header-section-number">2.6</span> Polygenic scoring</h2>
<hr />
<div id="generating-snp-weights" class="section level3">
<h3><span class="header-section-number">2.6.1</span> Generating SNP-weights</h3>
<hr />
<div id="meta-gwas-1" class="section level4">
<h4><span class="header-section-number">2.6.1.1</span> Meta-GWAS</h4>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code># Create a reference keep file to specify EUR and EAS individuals
cat /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/EUR_samples.keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/EAS_samples.keep &gt; /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/EUR_EAS_samples.keep
</code></pre>
<pre class="r"><code>pheno&lt;-c(&#39;MDD&#39;,&#39;MDD&#39;,&#39;MDD&#39;,&#39;Height&#39;,&#39;BMI&#39;)
samp1=c(&#39;PGC2&#39;,&#39;PGC2&#39;,&#39;PGC2&#39;,&#39;Wood&#39;,&#39;Locke&#39;)
samp2=c(&#39;PGC&#39;,&#39;CONVERGE&#39;,&#39;BBJ&#39;,&#39;BBJ&#39;,&#39;BBJ&#39;)
pop1=c(&#39;EUR&#39;,&#39;EUR&#39;,&#39;EUR&#39;,&#39;EUR&#39;,&#39;EUR&#39;)
pop2=c(&#39;EAS&#39;,&#39;EAS&#39;,&#39;EAS&#39;,&#39;EAS&#39;,&#39;EAS&#39;)

integers=c(2,3,4,5)

annot=c(&#39;LDcorr&#39;,&#39;Fst&#39;)
weight=c(0.5,1,1.5,2)

type&lt;-c(&#39;standard&#39;,&#39;integer&#39;,&#39;annot&#39;)

dir.create(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/score_files&#39;)

for(type_i in type){
  if(type_i == &#39;standard&#39;){
    for(i in 1:length(pheno)){
      
      system(paste0(&#39;sbatch -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator/polygenic_score_file_creator.R --ref_plink_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr --ref_keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/EUR_EAS_samples.keep --sumstats /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_standard1.reformat.txt.cleaned.gz --plink /users/k1806347/brc_scratch/Software/plink1.9.sh --memory 3000 --output /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/&#39;,type_i,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/score --ref_pop_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/super_pop_keep.list&#39;))
  
    }
  }
    
  if(type_i == &#39;integer&#39;){
    for(i in 1:length(pheno)){
      for(integers_i in integers){
        
              system(paste0(&#39;sbatch -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator/polygenic_score_file_creator.R --ref_plink_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr --ref_keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/EUR_EAS_samples.keep --sumstats /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_integer_&#39;,integers_i,&#39;1.reformat.txt.cleaned.gz --plink /users/k1806347/brc_scratch/Software/plink1.9.sh --memory 3000 --output /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/&#39;,type_i,&#39;/&#39;,&#39;k_&#39;,integers_i,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/score --ref_pop_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/super_pop_keep.list&#39;))

      }
    }
  }
    
  if(type_i == &#39;annot&#39;){
    for(i in 1:length(pheno)){
      for(annot_i in annot){
        for(weight_i in weight){
          
        system(paste0(&#39;sbatch -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator/polygenic_score_file_creator.R --ref_plink_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr --ref_keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/EUR_EAS_samples.keep --sumstats /users/k1806347/brc_scratch/Analyses/CrossMeta/metal/results/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;_metal_&#39;,annot_i,&#39;_&#39;,weight_i,&#39;1.reformat.txt.cleaned.gz --plink /users/k1806347/brc_scratch/Software/plink1.9.sh --memory 3000 --output /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/&#39;,type_i,&#39;/&#39;,annot_i,&#39;/&#39;,&#39;q_&#39;,weight_i,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/score --ref_pop_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/super_pop_keep.list&#39;))

        }
      }
    }
  }
}</code></pre>
</details>
<hr />
</div>
<div id="single-sample" class="section level4">
<h4><span class="header-section-number">2.6.1.2</span> Single sample</h4>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code>########
# All SNPs
########

pop=EUR
for gwas in $(echo PGC2_MDD_EUR Wood_Height_EUR Locke_BMI_EUR);do
  sbatch -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator/polygenic_score_file_creator.R \
  --ref_plink_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr \
  --ref_keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/${pop}_samples.keep \
  --sumstats /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/${gwas}.gz \
  --plink /users/k1806347/brc_scratch/Software/plink1.9.sh \
  --memory 3000 \
  --output /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/${pop}_only/${gwas}/score \
  --ref_pop_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/super_pop_keep.list
done

pop=EAS
for gwas in $(echo PGC_MDD_EAS CONVERGE_MDD_EAS BBJ_MDD_EAS BBJ_Height_EAS BBJ_BMI_EAS);do
  sbatch -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator/polygenic_score_file_creator.R \
  --ref_plink_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr \
  --ref_keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/${pop}_samples.keep \
  --sumstats /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/${gwas}.gz \
  --plink /users/k1806347/brc_scratch/Software/plink1.9.sh \
  --memory 3000 \
  --output /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/${pop}_only/${gwas}/score \
  --ref_pop_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/super_pop_keep.list
done

########
# Using subsets of LDcor and Fst
########

pop=EUR
for gwas in $(echo PGC2_MDD_EUR Wood_Height_EUR Locke_BMI_EUR);do
for annot in $(echo LDcorr Fst);do
for q in $(echo top_0.01 top_0.05 top_0.1 top_0.5 bot_0.01 bot_0.05 bot_0.1 bot_0.5);do
  sbatch -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator/polygenic_score_file_creator.R \
  --ref_plink_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr \
  --ref_keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/${pop}_samples.keep \
  --sumstats /users/k1806347/brc_scratch/Analyses/CrossMeta/GWAS_sumstats/cleaned/${gwas}.gz \
  --extract /scratch/users/k1806347/Analyses/CrossMeta/sign_test/clump/${gwas}/${annot}/${q}/${q}_${annot}.snps.txt \
  --plink /users/k1806347/brc_scratch/Software/plink1.9.sh \
  --memory 3000 \
  --output /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/${pop}_only_subset/${gwas}/${annot}/${q}/score \
  --ref_pop_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/super_pop_keep.list
done
done
done
</code></pre>
</details>
<hr />
</div>
<div id="prs-csx" class="section level4">
<h4><span class="header-section-number">2.6.1.3</span> PRS-CSx</h4>
<p>Documentation for generating PRS-CSx score files is <a href="https://opain.github.io/GenoPred/Pipeline_prep.html">here</a></p>
<hr />
</div>
</div>
<div id="calculating-polygenic-scores-in-target" class="section level3">
<h3><span class="header-section-number">2.6.2</span> Calculating polygenic scores in target</h3>
<hr />
<div id="meta-gwas-2" class="section level4">
<h4><span class="header-section-number">2.6.2.1</span> Meta-GWAS</h4>
<details>
<p><summary>Show code</summary></p>
<pre class="r"><code>pheno&lt;-c(&#39;MDD&#39;,&#39;MDD&#39;,&#39;MDD&#39;,&#39;Height&#39;,&#39;BMI&#39;)
samp1=c(&#39;PGC2&#39;,&#39;PGC2&#39;,&#39;PGC2&#39;,&#39;Wood&#39;,&#39;Locke&#39;)
samp2=c(&#39;PGC&#39;,&#39;CONVERGE&#39;,&#39;BBJ&#39;,&#39;BBJ&#39;,&#39;BBJ&#39;)
pop1=c(&#39;EUR&#39;,&#39;EUR&#39;,&#39;EUR&#39;,&#39;EUR&#39;,&#39;EUR&#39;)
pop2=c(&#39;EAS&#39;,&#39;EAS&#39;,&#39;EAS&#39;,&#39;EAS&#39;,&#39;EAS&#39;)

integers=c(2,3,4,5)

annot=c(&#39;LDcorr&#39;,&#39;Fst&#39;)
weight=c(0.5,1,1.5,2)

type&lt;-c(&#39;standard&#39;,&#39;integer&#39;,&#39;annot&#39;)

dir.create(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/prs&#39;)

for(type_i in type){
  if(type_i == &#39;standard&#39;){
    for(i in 1:length(pheno)){
      for(targ_pop in c(pop1[i],pop2[i])){
      
        system(paste0(&#39;sbatch -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer/Scaled_polygenic_scorer.R --target_plink_chr /users/k1806347/brc_scratch/Data/UKBB/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr --target_keep /users/k1806347/brc_scratch/Misc/Toshiki/UKB_keep_files/UKB_QC_UpdateIDs_&#39;,targ_pop,&#39;.keep --ref_score /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/&#39;,type_i,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/score --ref_scale /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/&#39;,type_i,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/score.&#39;,targ_pop,&#39;.scale --ref_freq_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/freq_files/&#39;,targ_pop,&#39;/1KGPhase3.w_hm3.&#39;,targ_pop,&#39;.chr --plink /users/k1806347/brc_scratch/Software/plink1.9.sh --output /users/k1806347/brc_scratch/Analyses/CrossMeta/prs/&#39;,targ_pop,&#39;_target/&#39;,type_i,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/prs&#39;))

      }
    }
  }
  
  if(type_i == &#39;integer&#39;){
    for(i in 1:length(pheno)){
      for(integers_i in integers){
        for(targ_pop in c(pop1[i],pop2[i])){
          system(paste0(&#39;sbatch -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer/Scaled_polygenic_scorer.R --target_plink_chr /users/k1806347/brc_scratch/Data/UKBB/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr --target_keep /users/k1806347/brc_scratch/Misc/Toshiki/UKB_keep_files/UKB_QC_UpdateIDs_&#39;,targ_pop,&#39;.keep --ref_score /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/&#39;,type_i,&#39;/&#39;,&#39;k_&#39;,integers_i,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/score --ref_scale /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/&#39;,type_i,&#39;/&#39;,&#39;k_&#39;,integers_i,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/score.&#39;,targ_pop,&#39;.scale --ref_freq_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/freq_files/&#39;,targ_pop,&#39;/1KGPhase3.w_hm3.&#39;,targ_pop,&#39;.chr --plink /users/k1806347/brc_scratch/Software/plink1.9.sh --output /users/k1806347/brc_scratch/Analyses/CrossMeta/prs/&#39;,targ_pop,&#39;_target/&#39;,type_i,&#39;/&#39;,&#39;k_&#39;,integers_i,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/prs&#39;))

        }
      }
    }
  }
    
  if(type_i == &#39;annot&#39;){
    for(i in 1:length(pheno)){
      for(annot_i in annot){
        for(weight_i in weight){
          for(targ_pop in c(pop1[i],pop2[i])){
            system(paste0(&#39;sbatch -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer/Scaled_polygenic_scorer.R --target_plink_chr /users/k1806347/brc_scratch/Data/UKBB/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr --target_keep /users/k1806347/brc_scratch/Misc/Toshiki/UKB_keep_files/UKB_QC_UpdateIDs_&#39;,targ_pop,&#39;.keep --ref_score /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/&#39;,type_i,&#39;/&#39;,annot_i,&#39;/&#39;,&#39;q_&#39;,weight_i,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/score --ref_scale /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/&#39;,type_i,&#39;/&#39;,annot_i,&#39;/&#39;,&#39;q_&#39;,weight_i,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/score.&#39;,targ_pop,&#39;.scale --ref_freq_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/freq_files/&#39;,targ_pop,&#39;/1KGPhase3.w_hm3.&#39;,targ_pop,&#39;.chr --plink /users/k1806347/brc_scratch/Software/plink1.9.sh --output /users/k1806347/brc_scratch/Analyses/CrossMeta/prs/&#39;,targ_pop,&#39;_target/&#39;,type_i,&#39;/&#39;,annot_i,&#39;/&#39;,&#39;q_&#39;,weight_i,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/prs&#39;))

          }
        }
      }
    }
  }
}</code></pre>
</details>
<hr />
</div>
<div id="single-sample-1" class="section level4">
<h4><span class="header-section-number">2.6.2.2</span> Single sample</h4>
<details>
<p><summary>Show code</summary></p>
<pre class="r"><code>########
# All SNPs
########

for(targ_pop in c(&#39;EUR&#39;,&#39;EAS&#39;)){
  for(gwas in c(&#39;PGC2_MDD_EUR&#39;,&#39;Wood_Height_EUR&#39;,&#39;Locke_BMI_EUR&#39;,&#39;PGC_MDD_EAS&#39;,&#39;CONVERGE_MDD_EAS&#39;,&#39;BBJ_MDD_EAS&#39;,&#39;BBJ_Height_EAS&#39;,&#39;BBJ_BMI_EAS&#39;)){
    pop&lt;-gsub(&#39;.*_&#39;,&#39;&#39;,gwas)
      
    system(paste0(&#39;sbatch -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer/Scaled_polygenic_scorer.R --target_plink_chr /users/k1806347/brc_scratch/Data/UKBB/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr --target_keep /users/k1806347/brc_scratch/Misc/Toshiki/UKB_keep_files/UKB_QC_UpdateIDs_&#39;,targ_pop,&#39;.keep --ref_score /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/&#39;,pop,&#39;_only/&#39;,gwas,&#39;/score --ref_scale /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/&#39;,pop,&#39;_only/&#39;,gwas,&#39;/score.&#39;,targ_pop,&#39;.scale --ref_freq_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/freq_files/&#39;,targ_pop,&#39;/1KGPhase3.w_hm3.&#39;,targ_pop,&#39;.chr --plink /users/k1806347/brc_scratch/Software/plink1.9.sh --output /users/k1806347/brc_scratch/Analyses/CrossMeta/prs/&#39;,targ_pop,&#39;_target/&#39;,pop,&#39;_only/&#39;,gwas,&#39;/prs&#39;))

  }
}

########
# Using subsets of LDcor and Fst
########

for(targ_pop in c(&#39;EUR&#39;,&#39;EAS&#39;)){
  for(gwas in c(&#39;PGC2_MDD_EUR&#39;,&#39;Wood_Height_EUR&#39;,&#39;Locke_BMI_EUR&#39;)){
    for(annot in c(&#39;LDcorr&#39;,&#39;Fst&#39;)){
      for(q in c(&#39;top_0.01&#39;, &#39;top_0.05&#39;, &#39;top_0.1&#39;, &#39;top_0.5&#39;, &#39;bot_0.01&#39;, &#39;bot_0.05&#39;, &#39;bot_0.1&#39;, &#39;bot_0.5&#39;)){

        pop&lt;-gsub(&#39;.*_&#39;,&#39;&#39;,gwas)
          
        if(!file.exists(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/&#39;,targ_pop,&#39;_target/&#39;,pop,&#39;_only_subset/&#39;,gwas,&#39;/&#39;,annot,&#39;/&#39;,q,&#39;/prs.profiles&#39;)))

        system(paste0(&#39;sbatch -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer/Scaled_polygenic_scorer.R --target_plink_chr /users/k1806347/brc_scratch/Data/UKBB/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr --target_keep /users/k1806347/brc_scratch/Misc/Toshiki/UKB_keep_files/UKB_QC_UpdateIDs_&#39;,targ_pop,&#39;.keep --ref_score /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/&#39;,pop,&#39;_only_subset/&#39;,gwas,&#39;/&#39;,annot,&#39;/&#39;,q,&#39;/score --ref_scale /users/k1806347/brc_scratch/Analyses/CrossMeta/score_files/&#39;,pop,&#39;_only_subset/&#39;,gwas,&#39;/&#39;,annot,&#39;/&#39;,q,&#39;/score.&#39;,targ_pop,&#39;.scale --ref_freq_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/freq_files/&#39;,targ_pop,&#39;/1KGPhase3.w_hm3.&#39;,targ_pop,&#39;.chr --plink /users/k1806347/brc_scratch/Software/plink1.9.sh --output /users/k1806347/brc_scratch/Analyses/CrossMeta/prs/&#39;,targ_pop,&#39;_target/&#39;,pop,&#39;_only_subset/&#39;,gwas,&#39;/&#39;,annot,&#39;/&#39;,q,&#39;/prs&#39;))
        
      }
    }
  }
}</code></pre>
</details>
<hr />
</div>
<div id="prs-csx-1" class="section level4">
<h4><span class="header-section-number">2.6.2.3</span> PRS-CSx</h4>
<details>
<p><summary>Show code</summary></p>
<pre class="r"><code># PLINK v2.00a2LM 64-bit Intel (9 Mar 2019) was getting a file read error when submitting as a batch job but not interactively
# Here I am using PLINK v2.00a2.3LM AVX2 Intel (24 Jan 2020) which works fine.

pheno&lt;-c(&#39;MDD&#39;,&#39;MDD&#39;,&#39;MDD&#39;,&#39;Height&#39;,&#39;BMI&#39;)
samp1=c(&#39;PGC2&#39;,&#39;PGC2&#39;,&#39;PGC2&#39;,&#39;Wood&#39;,&#39;Locke&#39;)
samp2=c(&#39;PGC&#39;,&#39;CONVERGE&#39;,&#39;BBJ&#39;,&#39;BBJ&#39;,&#39;BBJ&#39;)
pop1=c(&#39;EUR&#39;,&#39;EUR&#39;,&#39;EUR&#39;,&#39;EUR&#39;,&#39;EUR&#39;)
pop2=c(&#39;EAS&#39;,&#39;EAS&#39;,&#39;EAS&#39;,&#39;EAS&#39;,&#39;EAS&#39;)

for(i in 1:length(pheno)){
for(targ_pop in c(pop1[i], pop2[i])){
for(type in c(pop1[i], pop2[i], &#39;META&#39;)){

    system(paste0(&#39;sbatch -p brc,shared --mem 20G /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer/Scaled_polygenic_scorer_plink2.R --target_plink_chr /users/k1806347/brc_scratch/Data/UKBB/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr --target_keep /users/k1806347/brc_scratch/Misc/Toshiki/UKB_keep_files/UKB_QC_UpdateIDs_&#39;,targ_pop,&#39;.keep --ref_score /users/k1806347/brc_scratch/Data/1KG/Phase3/Score_files_for_polygenic/PRS-CSx/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/1KGPhase3.w_hm3.&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;.&#39;,type,&#39;.score.gz --ref_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/Score_files_for_polygenic/PRS-CSx/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/1KGPhase3.w_hm3.&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;.&#39;,type,&#39;.&#39;,targ_pop,&#39;.scale --ref_freq_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/freq_files/&#39;,targ_pop,&#39;/1KGPhase3.w_hm3.&#39;,targ_pop,&#39;.chr --plink2 /users/k1806347/brc_scratch/Software/plink2_alpha/plink2 --output /users/k1806347/brc_scratch/Analyses/CrossMeta/prs/&#39;,targ_pop,&#39;_target/PRS-CSx/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;.&#39;,type,&#39;/prs&#39;))

}
}
}</code></pre>
</details>
<hr />
</div>
</div>
</div>
<div id="evaluating-polygenic-scores" class="section level2">
<h2><span class="header-section-number">2.7</span> Evaluating polygenic scores</h2>
<div id="model-comparison" class="section level3">
<h3><span class="header-section-number">2.7.1</span> Model Comparison</h3>
<p>Here we will model the PRS from individual GWAS and meta-GWAS to evaluate and compare their predictive utility.</p>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code>#########
# Estimate variance explained by EUR PRS subset by annotations
#########

pheno=(MDD MDD MDD Height BMI)
samp1=(PGC2 PGC2 PGC2 Wood Locke)
samp2=(PGC CONVERGE BBJ BBJ BBJ)
pop1=(EUR EUR EUR EUR EUR)
pop2=(EAS EAS EAS EAS EAS)

# use for loop to read all values and indexes
for (( i=0; i&lt;${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop2[${i}]} ${pop1[${i}]}); do
for annot in $(echo LDcorr Fst); do
mkdir -p /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}

cat &lt;&lt;EOF &gt; /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}/prs_1_${annot}.txt
predictors group
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/prs.profiles ${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}_all
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only_subset/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/${annot}/top_0.01/prs.profiles ${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}_${annot}_top_0.01
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only_subset/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/${annot}/top_0.05/prs.profiles ${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}_${annot}_top_0.05
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only_subset/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/${annot}/top_0.1/prs.profiles ${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}_${annot}_top_0.1
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only_subset/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/${annot}/top_0.5/prs.profiles ${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}_${annot}_top_0.5
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only_subset/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/${annot}/bot_0.01/prs.profiles ${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}_${annot}_bot_0.01
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only_subset/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/${annot}/bot_0.05/prs.profiles ${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}_${annot}_bot_0.05
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only_subset/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/${annot}/bot_0.1/prs.profiles ${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}_${annot}_bot_0.1
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only_subset/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/${annot}/bot_0.5/prs.profiles ${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}_${annot}_bot_0.5

EOF

done
done
done

for (( i=0; i&lt;${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop2[${i}]} ${pop1[${i}]}); do
for annot in $(echo LDcorr Fst); do

# Test association between prs and phenotype
sbatch --mem 10G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R \
--pheno /users/k1806347/brc_scratch/Analyses/CrossMeta/Phenotype/${pheno[${i}]}.txt \
--out /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}/prs_1_${annot}.test \
--assoc T \
--predictors /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}/prs_1_${annot}.txt

done
done
done

#########
# Estimate predictive utility of single GWAS, Standard Meta, Integer Meta, and Annotation-weight Meta
# Per pT results can be seen in the assoc.txt file
#########

pheno=(MDD MDD MDD Height BMI)
samp1=(PGC2 PGC2 PGC2 Wood Locke)
samp2=(PGC CONVERGE BBJ BBJ BBJ)
pop1=(EUR EUR EUR EUR EUR)
pop2=(EAS EAS EAS EAS EAS)

# use for loop to read all values and indexes
for (( i=0; i&lt;${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

mkdir -p /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}

cat &lt;&lt;EOF &gt; /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_1.txt
predictors group
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/prs.profiles ${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}_single
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop2[${i}]}_only/${samp2[${i}]}_${pheno[${i}]}_${pop2[${i}]}/prs.profiles ${samp2[${i}]}_${pheno[${i}]}_${pop2[${i}]}_single
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/standard/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/integer/k_2/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_integer_k_2
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/integer/k_3/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_integer_k_3
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/integer/k_4/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_integer_k_4
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/integer/k_5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_integer_k_5
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/LDcorr/q_0.5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_LDcorr_q_0.5
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/LDcorr/q_1/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_LDcorr_q_1
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/LDcorr/q_1.5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_LDcorr_q_1.5
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/LDcorr/q_2/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_LDcorr_q_2
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/Fst/q_0.5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_Fst_q_0.5
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/Fst/q_1/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_Fst_q_1
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/Fst/q_1.5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_Fst_q_1.5
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/Fst/q_2/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}_Fst_q_2


EOF

done
done

for (( i=0; i&lt;${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

# Test association between prs and phenotype
sbatch --mem 10G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R \
--pheno /users/k1806347/brc_scratch/Analyses/CrossMeta/Phenotype/${pheno[${i}]}.txt \
--out /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_1.test \
--assoc T \
--predictors /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_1.txt

done
done

#########
# Compare predictive utility of standard meta + single sample gwas vs standard meta + single sample gwas + integer-weighted meta
#########

pheno=(MDD MDD MDD Height BMI)
samp1=(PGC2 PGC2 PGC2 Wood Locke)
samp2=(PGC CONVERGE BBJ BBJ BBJ)
pop1=(EUR EUR EUR EUR EUR)
pop2=(EAS EAS EAS EAS EAS)

# use for loop to read all values and indexes
for (( i=0; i&lt;${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

mkdir -p /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}

cat &lt;&lt;EOF &gt; /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_2.txt
predictors group
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop2[${i}]}_only/${samp2[${i}]}_${pheno[${i}]}_${pop2[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/standard/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/integer/k_2/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles integer
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/integer/k_3/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles integer
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/integer/k_4/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles integer
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/integer/k_5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles integer

EOF
done
done

for (( i=0; i&lt;${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

# Test association between prs and phenotype
sbatch --mem 10G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R \
--pheno /users/k1806347/brc_scratch/Analyses/CrossMeta/Phenotype/${pheno[${i}]}.txt \
--out /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_2.test \
--assoc T \
--predictors /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_2.txt

done
done

#########
# Compare predictive utility of standard meta + single sample gwas vs standard meta + single sample gwas + LDcorr/Fst-weighted meta
#########

pheno=(MDD MDD MDD Height BMI)
samp1=(PGC2 PGC2 PGC2 Wood Locke)
samp2=(PGC CONVERGE BBJ BBJ BBJ)
pop1=(EUR EUR EUR EUR EUR)
pop2=(EAS EAS EAS EAS EAS)

# use for loop to read all values and indexes
for annot in $(echo LDcorr Fst);do
for (( i=0; i&lt;${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

mkdir -p /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}

cat &lt;&lt;EOF &gt; /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_${annot}.txt
predictors group
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop2[${i}]}_only/${samp2[${i}]}_${pheno[${i}]}_${pop2[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/standard/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/${annot}/q_0.5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${annot}
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/${annot}/q_1/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${annot}
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/${annot}/q_1.5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${annot}
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/${annot}/q_2/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles ${annot}

EOF
done
done
done

for annot in $(echo LDcorr Fst);do
for (( i=0; i&lt;${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

# Test association between prs and phenotype
sbatch --mem 10G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R \
--pheno /users/k1806347/brc_scratch/Analyses/CrossMeta/Phenotype/${pheno[${i}]}.txt \
--out /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_${annot}.test \
--assoc T \
--predictors /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_${annot}.txt

done
done
done

#########
# Compare predictive utility of standard meta + single sample gwas vs standard meta + single sample gwas + LDcorr and Fst-weighted meta
#########

pheno=(MDD MDD MDD Height BMI)
samp1=(PGC2 PGC2 PGC2 Wood Locke)
samp2=(PGC CONVERGE BBJ BBJ BBJ)
pop1=(EUR EUR EUR EUR EUR)
pop2=(EAS EAS EAS EAS EAS)

# use for loop to read all values and indexes
for (( i=0; i&lt;${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

mkdir -p /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}

cat &lt;&lt;EOF &gt; /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_3.txt
predictors group
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop2[${i}]}_only/${samp2[${i}]}_${pheno[${i}]}_${pop2[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/standard/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/LDcorr/q_0.5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles LDcorr_Fst
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/LDcorr/q_1/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles LDcorr_Fst
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/LDcorr/q_1.5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles LDcorr_Fst
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/LDcorr/q_2/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles LDcorr_Fst
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/Fst/q_0.5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles LDcorr_Fst
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/Fst/q_1/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles LDcorr_Fst
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/Fst/q_1.5/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles LDcorr_Fst
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/annot/Fst/q_2/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles LDcorr_Fst

EOF
done
done

for (( i=0; i&lt;${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

# Test association between prs and phenotype
sbatch --mem 10G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R \
--pheno /users/k1806347/brc_scratch/Analyses/CrossMeta/Phenotype/${pheno[${i}]}.txt \
--out /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_3.test \
--assoc T \
--predictors /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs_3.txt

done
done

#########
# Compare predictive utility of PRS-CSx single sample PRS vs PRS-CSx single sample PRS + PRS-CSx META PRS 
#########

pheno=(MDD MDD MDD Height BMI)
samp1=(PGC2 PGC2 PGC2 Wood Locke)
samp2=(PGC CONVERGE BBJ BBJ BBJ)
pop1=(EUR EUR EUR EUR EUR)
pop2=(EAS EAS EAS EAS EAS)

# use for loop to read all values and indexes
for (( i=0; i&lt;${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

mkdir -p /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}

cat &lt;&lt;EOF &gt; /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/PRS-CSx_1.txt
predictors group
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/PRS-CSx/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}.${pop1[${i}]}/prs.profiles PRS-CSx_single
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/PRS-CSx/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}.${pop2[${i}]}/prs.profiles PRS-CSx_single
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/PRS-CSx/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}.META/prs.profiles PRS-CSx_meta

EOF
done
done

for (( i=0; i&lt;${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

# Test association between prs and phenotype
sbatch --mem 10G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R \
--pheno /users/k1806347/brc_scratch/Analyses/CrossMeta/Phenotype/${pheno[${i}]}.txt \
--out /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/PRS-CSx_1.test \
--assoc T \
--predictors /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/PRS-CSx_1.txt

done
done

#########
# Compare predictive utility of standard meta + single sample gwas vs standard meta + single sample gwas + all PRS-CSx PRS 
#########

pheno=(MDD MDD MDD Height BMI)
samp1=(PGC2 PGC2 PGC2 Wood Locke)
samp2=(PGC CONVERGE BBJ BBJ BBJ)
pop1=(EUR EUR EUR EUR EUR)
pop2=(EAS EAS EAS EAS EAS)

# use for loop to read all values and indexes
for (( i=0; i&lt;${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

mkdir -p /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}

cat &lt;&lt;EOF &gt; /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/PRS-CSx_2.txt
predictors group
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop1[${i}]}_only/${samp1[${i}]}_${pheno[${i}]}_${pop1[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/${pop2[${i}]}_only/${samp2[${i}]}_${pheno[${i}]}_${pop2[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/standard/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/prs.profiles single_and_standard
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/PRS-CSx/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}.${pop1[${i}]}/prs.profiles PRS-CSx
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/PRS-CSx/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}.${pop2[${i}]}/prs.profiles PRS-CSx
/users/k1806347/brc_scratch/Analyses/CrossMeta/prs/${targ_pop}_target/PRS-CSx/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}.META/prs.profiles PRS-CSx

EOF
done
done

for (( i=0; i&lt;${#pheno[@]}; i++ ));do
for targ_pop in $(echo ${pop1[${i}]} ${pop2[${i}]}); do

# Test association between prs and phenotype
sbatch --mem 10G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R \
--pheno /users/k1806347/brc_scratch/Analyses/CrossMeta/Phenotype/${pheno[${i}]}.txt \
--out /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/PRS-CSx_2.test \
--assoc T \
--predictors /users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/${targ_pop}/${pheno[${i}]}_${samp1[${i}]}_${pop1[${i}]}_${samp2[${i}]}_${pop2[${i}]}/PRS-CSx_2.txt

done
done

###########
# That should be enough comparisons to see what we want
###########
</code></pre>
</details>
<hr />
</div>
<div id="plotting-results" class="section level3">
<h3><span class="header-section-number">2.7.2</span> Plotting results</h3>
<details>
<p><summary>Show code</summary></p>
<pre class="r"><code>library(data.table)
library(ggplot2)
library(cowplot)

pheno&lt;-c(&#39;MDD&#39;,&#39;MDD&#39;,&#39;MDD&#39;,&#39;Height&#39;,&#39;BMI&#39;)
samp1=c(&#39;PGC2&#39;,&#39;PGC2&#39;,&#39;PGC2&#39;,&#39;Wood&#39;,&#39;Locke&#39;)
samp2=c(&#39;PGC&#39;,&#39;CONVERGE&#39;,&#39;BBJ&#39;,&#39;BBJ&#39;,&#39;BBJ&#39;)
pop1=c(&#39;EUR&#39;,&#39;EUR&#39;,&#39;EUR&#39;,&#39;EUR&#39;,&#39;EUR&#39;)
pop2=c(&#39;EAS&#39;,&#39;EAS&#39;,&#39;EAS&#39;,&#39;EAS&#39;,&#39;EAS&#39;)

############
# Plot portability of subset PRS
############

assoc_all_pop_all_pheno&lt;-NULL
for(i in which(!duplicated(paste0(pheno,samp1,pop1)))){
assoc_all_pop&lt;-NULL
  for(targ_pop in c(pop1[i], pop2[i])){
    assoc&lt;-NULL
    for(annot in c(&#39;LDcorr&#39;,&#39;Fst&#39;)){
  
      #assoc&lt;-rbind(assoc,fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;/prs_1_&#39;,annot,&#39;.test.assoc.txt&#39;)))
      
      tmp&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;/prs_1_&#39;,annot,&#39;.test.assoc.txt&#39;))
      tmp$Annot&lt;-annot
      assoc&lt;-rbind(assoc,tmp)

    }
  
    assoc$Label&lt;-paste0(pheno[i],&#39; &#39;)
    
    assoc$Target&lt;-targ_pop
  
    assoc$pT&lt;-gsub(&#39;e.&#39;,&#39;e-&#39;,gsub(&#39;.*SCORE_&#39;,&#39;&#39;,assoc$Predictor))
  
    assoc$Type&lt;-NA
    assoc$Type[grepl(&#39;EUR.all&#39;,assoc$Predictor)]&lt;-&#39;All SNPs&#39;
    
    assoc$Type[grepl(&#39;EUR.LDcorr&#39;,assoc$Predictor)]&lt;-gsub(&#39;.*EUR.LDcorr.&#39;,&#39;LDcorr: &#39;,gsub(&#39;.PredFile.*&#39;,&#39;&#39;,assoc$Predictor[grepl(&#39;EUR.LDcorr&#39;,assoc$Predictor)]))
    
    
    assoc$Type[grepl(&#39;EUR.Fst&#39;,assoc$Predictor)]&lt;-gsub(&#39;.*EUR.Fst.&#39;,&#39;Fst: &#39;,gsub(&#39;.PredFile.*&#39;,&#39;&#39;,assoc$Predictor[grepl(&#39;EUR.Fst&#39;,assoc$Predictor)]))
  
    assoc$Type&lt;-paste0(gsub(&#39;bot.&#39;,&#39;Bottom &#39;, gsub(&#39;top.&#39;,&#39;Top &#39;,assoc$Type)),&#39;%&#39;)
  
    assoc$pT&lt;-factor(assoc$pT, levels=unique(assoc$pT))
    assoc$Type&lt;-factor(assoc$Type, levels=c(&quot;LDcorr: Bottom 0.01%&quot;,&quot;LDcorr: Bottom 0.05%&quot;,&quot;LDcorr: Bottom 0.1%&quot;, &quot;LDcorr: Bottom 0.5%&quot;,&quot;LDcorr: Top 0.01%&quot;,&quot;LDcorr: Top 0.05%&quot;,&quot;LDcorr: Top 0.1%&quot;,&quot;LDcorr: Top 0.5%&quot;, &quot;Fst: Bottom 0.01%&quot;,&quot;Fst: Bottom 0.05%&quot;, &quot;Fst: Bottom 0.1%&quot;, &quot;Fst: Bottom 0.5%&quot;, &quot;Fst: Top 0.01%&quot;, &quot;Fst: Top 0.05%&quot;, &quot;Fst: Top 0.1%&quot;, &quot;Fst: Top 0.5%&quot;, &quot;All SNPs%&quot;))
  
    assoc&lt;-assoc[complete.cases(assoc),]
    assoc_all_pop&lt;-rbind(assoc_all_pop, assoc)
  }
  assoc_all_pop$Phenotype&lt;-pheno[i]
  assoc_all_pop_all_pheno&lt;-rbind(assoc_all_pop_all_pheno, assoc_all_pop)
}

# Extract results corresponding to bot5%, top5% and all variants
assoc_all_pop_all_pheno_subset&lt;-assoc_all_pop_all_pheno[grepl(&#39;Bottom 0.05|Top 0.05|All SNPs&#39;, assoc_all_pop_all_pheno$Type),]

# Calculate portability
assoc_all_pop_all_pheno_subset&lt;-assoc_all_pop_all_pheno_subset[!duplicated(paste0(assoc_all_pop_all_pheno_subset$Predictor,&#39;-&#39;,assoc_all_pop_all_pheno_subset$Target,&#39;-&#39;,assoc_all_pop_all_pheno_subset$Annot,&#39;-&#39;,assoc_all_pop_all_pheno_subset$Label)),]
assoc_all_pop_all_pheno_subset_EUR&lt;-assoc_all_pop_all_pheno_subset[assoc_all_pop_all_pheno_subset$Target == &#39;EUR&#39;,]
assoc_all_pop_all_pheno_subset_EAS&lt;-assoc_all_pop_all_pheno_subset[assoc_all_pop_all_pheno_subset$Target == &#39;EAS&#39;,]

assoc_all_pop_all_pheno_subset_wide&lt;-merge(assoc_all_pop_all_pheno_subset_EUR, assoc_all_pop_all_pheno_subset_EAS, by=c(&#39;Phenotype&#39;,&#39;Type&#39;,&#39;pT&#39;,&#39;Annot&#39;,&#39;Label&#39;))

assoc_all_pop_all_pheno_subset_wide$Portability&lt;-assoc_all_pop_all_pheno_subset_wide$BETA.y/assoc_all_pop_all_pheno_subset_wide$BETA.x

# Scale SE of BETA in EAS by BETA.y/portability
assoc_all_pop_all_pheno_subset_wide$Portability_SE&lt;-assoc_all_pop_all_pheno_subset_wide$SE.y/(assoc_all_pop_all_pheno_subset_wide$BETA.y/assoc_all_pop_all_pheno_subset_wide$Portability)

assoc_all_pop_all_pheno_subset_wide$Test&lt;-NA
assoc_all_pop_all_pheno_subset_wide$Test[grepl(&#39;Bottom&#39;, assoc_all_pop_all_pheno_subset_wide$Type)]&lt;-&#39;bot5&#39;
assoc_all_pop_all_pheno_subset_wide$Test[grepl(&#39;Top&#39;, assoc_all_pop_all_pheno_subset_wide$Type)]&lt;-&#39;top5&#39;
assoc_all_pop_all_pheno_subset_wide$Test[grepl(&#39;All&#39;, assoc_all_pop_all_pheno_subset_wide$Type)]&lt;-&#39;all&#39;

assoc_all_pop_all_pheno_subset_wide$Test&lt;-factor(assoc_all_pop_all_pheno_subset_wide$Test, levels=c(&#39;bot5&#39;,&#39;all&#39;,&#39;top5&#39;))

# Subset results to only show pT=0.5,0.1,0.01,1e-4,1e-8 to avoid redundancy
assoc_all_pop_all_pheno_subset_wide&lt;-assoc_all_pop_all_pheno_subset_wide[assoc_all_pop_all_pheno_subset_wide$pT %in% c(&#39;1e-08&#39;,&#39;1e-06&#39;,&#39;1e-04&#39;,&#39;0.01&#39;,&#39;0.1&#39;,&#39;0.5&#39;),]

# Plot results
png(paste0(&#39;/scratch/users/k1806347/Analyses/CrossMeta/assoc/Top_Bot_Port_BMI_Height.png&#39;), units=&#39;px&#39;, res=300, width=2000, height=1500)
port_plot_height_bmi&lt;-  ggplot(assoc_all_pop_all_pheno_subset_wide[grepl(&#39;BMI|Height&#39;,assoc_all_pop_all_pheno_subset_wide$Phenotype),], aes(x=pT, y=Portability, fill=Test)) +
    geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
    geom_errorbar(aes(ymin=Portability-Portability_SE, ymax=Portability+Portability_SE), width=.2,position=position_dodge(.9)) +
    theme_half_open() +
    background_grid() +
    labs(y=&quot;Portability&quot;) +
    theme(axis.text.x = element_text(angle = 45, hjust=1)) +
    facet_grid(vars(Label), vars(Annot), scales = &quot;free&quot;)

port_plot_height_bmi

dev.off()

# Plot results
png(paste0(&#39;/scratch/users/k1806347/Analyses/CrossMeta/assoc/Top_Bot_Port_MDD.png&#39;), units=&#39;px&#39;, res=300, width=2000, height=1000)
port_plot_mdd&lt;-ggplot(assoc_all_pop_all_pheno_subset_wide[grepl(&#39;MDD&#39;,assoc_all_pop_all_pheno_subset_wide$Phenotype),], aes(x=pT, y=Portability, fill=Test)) +
    geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
    geom_errorbar(aes(ymin=Portability-Portability_SE, ymax=Portability+Portability_SE), width=.2,position=position_dodge(.9)) +
    theme_half_open() +
    background_grid() +
    labs(y=&quot;Portability&quot;) +
    theme(axis.text.x = element_text(angle = 45, hjust=1)) +
    facet_grid(vars(Label), vars(Annot), scales = &quot;free&quot;)

port_plot_mdd

dev.off()

############
# Plot subset PRS
############

assoc_brief_all&lt;-NULL
eval_all&lt;-NULL
comp_all&lt;-NULL

for(i in which(!duplicated(paste0(pheno,samp1,pop1)))){
for(targ_pop in c(pop2[i])){
assoc&lt;-NULL
for(annot in c(&#39;LDcorr&#39;,&#39;Fst&#39;)){
  
  tmp&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;/prs_1_&#39;,annot,&#39;.test.assoc.txt&#39;))
  tmp$Annot&lt;-annot
  assoc&lt;-rbind(assoc,tmp)
  
}
  
  assoc$Target&lt;-targ_pop

  assoc$pT&lt;-gsub(&#39;e.&#39;,&#39;e-&#39;,gsub(&#39;.*SCORE_&#39;,&#39;&#39;,assoc$Predictor))

  assoc$Type&lt;-NA
  assoc$Type[grepl(&#39;EUR.all&#39;,assoc$Predictor)]&lt;-&#39;All SNPs&#39;
  
  assoc$Type[grepl(&#39;EUR.LDcorr&#39;,assoc$Predictor)]&lt;-gsub(&#39;.*EUR.LDcorr.&#39;,&#39;LDcorr: &#39;,gsub(&#39;.PredFile.*&#39;,&#39;&#39;,assoc$Predictor[grepl(&#39;EUR.LDcorr&#39;,assoc$Predictor)]))
  
  
  assoc$Type[grepl(&#39;EUR.Fst&#39;,assoc$Predictor)]&lt;-gsub(&#39;.*EUR.Fst.&#39;,&#39;Fst: &#39;,gsub(&#39;.PredFile.*&#39;,&#39;&#39;,assoc$Predictor[grepl(&#39;EUR.Fst&#39;,assoc$Predictor)]))

  assoc$Type&lt;-gsub(&#39;bot.&#39;,&#39;Bottom &#39;, gsub(&#39;top.&#39;,&#39;Top &#39;,assoc$Type))

  assoc$pT&lt;-factor(assoc$pT, levels=unique(assoc$pT))
  assoc$Type&lt;-factor(assoc$Type, levels=c(&quot;LDcorr: Bottom 0.01&quot;,&quot;LDcorr: Bottom 0.05&quot;,&quot;LDcorr: Bottom 0.1&quot;, &quot;LDcorr: Bottom 0.5&quot;,&quot;LDcorr: Top 0.01&quot;,&quot;LDcorr: Top 0.05&quot;,&quot;LDcorr: Top 0.1&quot;,&quot;LDcorr: Top 0.5&quot;, &quot;Fst: Bottom 0.01&quot;,&quot;Fst: Bottom 0.05&quot;, &quot;Fst: Bottom 0.1&quot;, &quot;Fst: Bottom 0.5&quot;, &quot;Fst: Top 0.01&quot;, &quot;Fst: Top 0.05&quot;, &quot;Fst: Top 0.1&quot;, &quot;Fst: Top 0.5&quot;, &quot;All SNPs&quot;))

  assoc&lt;-assoc[complete.cases(assoc),]
  
  if(min(assoc$BETA-assoc$SE) &lt; 0){
    ymin=min(assoc$BETA-assoc$SE)
  } else {
    ymin&lt;-0
  }

  assoc_plot&lt;-list()
  
  assoc_plot[[&#39;1&#39;]]&lt;-ggplot(assoc[grepl(&#39;All|LDcorr&#39;, assoc$Type),], aes(x=pT, y=BETA, fill=Type)) +
  geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9), colour=&#39;black&#39;) +
  geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
  theme_half_open() +
  background_grid() +
  ylim(ymin,max(assoc$BETA+assoc$SE)) +
  labs(y=&quot;R (SE)&quot;) +
  theme(axis.text.x = element_text(angle = 45, hjust=1)) +
  scale_fill_manual(values=c(&quot;#660000&quot;,&quot;#990000&quot;, &quot;#CC0000&quot;, &quot;#FF0000&quot;,&quot;#003300&quot;,&quot;#006600&quot;, &quot;#33CC33&quot;, &quot;#66FF00&quot;,&quot;#0066CC&quot;,&quot;#CC33CC&quot;))

  assoc_plot[[&#39;2&#39;]]&lt;-ggplot(assoc[grepl(&#39;All|Fst&#39;, assoc$Type),], aes(x=pT, y=BETA, fill=Type)) +
  geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
  geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
  theme_half_open() +
  background_grid() +
  ylim(ymin,max(assoc$BETA+assoc$SE)) +
  labs(y=&quot;R (SE)&quot;) +
  theme(axis.text.x = element_text(angle = 45, hjust=1)) +
  scale_fill_manual(values=c(&quot;#660000&quot;,&quot;#990000&quot;, &quot;#CC0000&quot;, &quot;#FF0000&quot;,&quot;#003300&quot;,&quot;#006600&quot;, &quot;#33CC33&quot;, &quot;#66FF00&quot;,&quot;#0066CC&quot;,&quot;#CC33CC&quot;))

  png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;/prs_1.test.pT_assoc.png&#39;), units=&#39;px&#39;, res=300, width=2500, height=2500)
    print(plot_grid(plotlist=assoc_plot, ncol=1))
  dev.off()
  
  ###
  # Make a brief version of the plot containing only top and bottom 5%
  # Note showing the top 10% makes the results clearer for Height due to lower SE. Consider changing the BETA comparison threshold
  ###
  assoc_brief&lt;-assoc[grepl(&#39;0.05|All&#39;,assoc$Type),]
  assoc_brief$Type&lt;-factor(assoc_brief$Type, levels=c(&#39;LDcorr: Bottom 0.05&#39;,&#39;Fst: Bottom 0.05&#39;,&#39;All SNPs&#39;,&#39;LDcorr: Top 0.05&#39;,&#39;Fst: Top 0.05&#39;))
    
  assoc_plot_brief&lt;-list()
  
  assoc_plot[[paste0(pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_1&#39;)]]&lt;-ggplot(assoc_brief[grepl(&#39;All|LDcorr&#39;, assoc_brief$Type),], aes(x=pT, y=BETA, fill=Type)) +
  geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
  geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
  theme_half_open() +
  background_grid() +
  ylim(ymin,max(assoc_brief$BETA+assoc_brief$SE)) +
  labs(y=&quot;R (SE)&quot;) +
  theme(axis.text.x = element_text(angle = 45, hjust=1))

  assoc_plot[[paste0(pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_2&#39;)]]&lt;-ggplot(assoc_brief[grepl(&#39;All|Fst&#39;, assoc_brief$Type),], aes(x=pT, y=BETA, fill=Type)) +
  geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
  geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
  theme_half_open() +
  background_grid() +
  ylim(ymin,max(assoc_brief$BETA+assoc_brief$SE)) +
  labs(y=&quot;R (SE)&quot;) +
  theme(axis.text.x = element_text(angle = 45, hjust=1))

  png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;/prs_1.test_brief.pT_assoc.png&#39;), units=&#39;px&#39;, res=300, width=2500, height=2500)
    print(plot_grid(plotlist=assoc_plot, ncol=1))
  dev.off()

  assoc_brief$Phenotype&lt;-pheno[i]
  assoc_brief_all&lt;-rbind(assoc_brief_all, assoc_brief)

  #####
  # Plot elastic net model results
  #####
  eval&lt;-NULL
  for(annot in c(&#39;LDcorr&#39;,&#39;Fst&#39;)){
    tmp&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;/prs_1_&#39;,annot,&#39;.test.pred_eval.txt&#39;))
    
    tmp$Model[tmp$Model == &#39;All_group&#39;]&lt;-paste0(&#39;All SNPs + All &#39;,annot)
    
    tmp$Annot&lt;-annot
    
    eval&lt;-rbind(eval, tmp)
  
  }

  comp&lt;-NULL
  for(annot in c(&#39;LDcorr&#39;,&#39;Fst&#39;)){
    tmp&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;/prs_1_&#39;,annot,&#39;.test.pred_comp.txt&#39;))
    
    tmp$Phenotype&lt;-pheno[i]
    tmp$Annot&lt;-annot
    
    comp&lt;-rbind(comp, tmp)
  
  }
  comp_all&lt;-rbind(comp_all, comp)

  eval$Type&lt;-gsub(&#39;.*EUR.&#39;,&#39;&#39;,gsub(&#39;_group&#39;,&#39;&#39;,eval$Model))
  eval$Type[eval$Type == &#39;all&#39;]&lt;-&#39;All SNPs&#39;
  eval$Type&lt;-gsub(&#39;.bot.&#39;,&#39;: Bottom &#39;, eval$Type)
  eval$Type&lt;-gsub(&#39;.top.&#39;,&#39;: Top &#39;, eval$Type)
  eval$Type&lt;-factor(eval$Type, levels = c(&quot;LDcorr: Bottom 0.01&quot;,&quot;LDcorr: Bottom 0.05&quot;,&quot;LDcorr: Bottom 0.1&quot;,&quot;LDcorr: Bottom 0.5&quot;,&quot;LDcorr: Top 0.01&quot;,&quot;LDcorr: Top 0.05&quot;, &quot;LDcorr: Top 0.1&quot;,&quot;LDcorr: Top 0.5&quot;, &quot;Fst: Bottom 0.01&quot;,&quot;Fst: Bottom 0.05&quot;,&quot;Fst: Bottom 0.1&quot;,&quot;Fst: Bottom 0.5&quot;, &quot;Fst: Top 0.01&quot;,&quot;Fst: Top 0.05&quot;,&quot;Fst: Top 0.1&quot;, &quot;Fst: Top 0.5&quot;,&quot;All SNPs&quot;, &quot;All SNPs + All LDcorr&quot;,&quot;All SNPs + All Fst&quot;))
  
  if(min(eval$R-eval$SE) &lt; 0){
    ymin=min(eval$R-eval$SE)
  } else {
    ymin&lt;-0
  }
  
  eval_plot&lt;-list()

  eval_plot[[1]]&lt;-ggplot(eval[grepl(&#39;LDcorr|SNPs$&#39;, eval$Type),], aes(x=Type, y=R, fill=Type)) +
    geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
    geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2,position=position_dodge(.9)) +
    theme_half_open() +
    background_grid() +
    ylim(ymin,max(eval$R+eval$SE)) +
    labs(y=&quot;R (SE)&quot;) +
    theme(axis.text.x = element_text(angle = 45, hjust=1))
  
  eval_plot[[2]]&lt;-ggplot(eval[grepl(&#39;Fst|SNPs$&#39;, eval$Type),], aes(x=Type, y=R, fill=Type)) +
    geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
    geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2,position=position_dodge(.9)) +
    theme_half_open() +
    background_grid() +
    ylim(ymin,max(eval$R+eval$SE)) +
    labs(y=&quot;R (SE)&quot;) +
    theme(axis.text.x = element_text(angle = 45, hjust=1))
  
  png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;/prs_1.test.eval.png&#39;), units=&#39;px&#39;, res=300, width=2500, height=2500)
    print(plot_grid(plotlist=eval_plot, ncol=1))
  dev.off()

  # Plot the data differently
  eval$Quant&lt;-gsub(&#39;.* &#39;,&#39;&#39;,eval$Type)
  eval$Quant[grepl(&#39;All&#39;, eval$Type)]&lt;-as.character(eval$Type[grepl(&#39;All&#39;, eval$Type)])
  eval$Quant&lt;-gsub(&#39;\\+ &#39;,&#39;\\+\n&#39;,eval$Quant)
  
  eval$TopBot&lt;-NA
  eval$TopBot[grepl(&#39;Top&#39;, eval$Type)]&lt;-&#39;Top&#39;
  eval$TopBot[grepl(&#39;Bottom&#39;, eval$Type)]&lt;-&#39;Bottom&#39;
  eval$TopBot[grepl(&#39;All&#39;, eval$Type)]&lt;-as.character(eval$Type[grepl(&#39;All&#39;, eval$Type)])
  eval$TopBot&lt;-gsub(&#39;\\+ &#39;,&#39;\\+\n&#39;,eval$TopBot)

    eval$Phenotype&lt;-pheno[i]
    eval_all&lt;-rbind(eval_all, eval)

  }
}

# Make plot containing brief assoc results for Height and BMI
assoc_brief_all$Test[grepl(&#39;Bottom&#39;, assoc_brief_all$Type)]&lt;-&#39;bot5&#39;
assoc_brief_all$Test[grepl(&#39;Top&#39;, assoc_brief_all$Type)]&lt;-&#39;top5&#39;
assoc_brief_all$Test[grepl(&#39;All&#39;, assoc_brief_all$Type)]&lt;-&#39;all&#39;

assoc_brief_all$Test&lt;-factor(assoc_brief_all$Test, levels=c(&#39;bot5&#39;,&#39;all&#39;,&#39;top5&#39;))

# Subset results to only show pT=0.5,0.1,0.01,1e-4,1e-8 to avoid redundancy
assoc_brief_all&lt;-assoc_brief_all[assoc_brief_all$pT %in% c(&#39;1e-08&#39;,&#39;1e-06&#39;,&#39;1e-04&#39;,&#39;0.01&#39;,&#39;0.1&#39;,&#39;0.5&#39;),]

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/prs_1.test_brief.Height_BMI.pT_assoc.png&#39;), units=&#39;px&#39;, res=300, width=2000, height=1500)
assoc_plot_height_bmi&lt;-ggplot(assoc_brief_all[grepl(&#39;Height|BMI&#39;, assoc_brief_all$Phenotype),], aes(x=pT, y=BETA, fill=Test)) +
  geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
  geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
  theme_half_open() +
  background_grid() +
  labs(y=&quot;R (SE)&quot;) +
  theme(axis.text.x = element_text(angle = 45, hjust=1)) +
  facet_grid(vars(Phenotype), vars(Annot), scales = &quot;free&quot;)

assoc_plot_height_bmi

dev.off()

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/prs_1.test_brief.MDD.pT_assoc.png&#39;), units=&#39;px&#39;, res=300, width=2000, height=1000)
assoc_plot_mdd&lt;-ggplot(assoc_brief_all[grepl(&#39;MDD&#39;, assoc_brief_all$Phenotype),], aes(x=pT, y=BETA, fill=Test)) +
  geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
  geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
  theme_half_open() +
  background_grid() +
  labs(y=&quot;R (SE)&quot;) +
  theme(axis.text.x = element_text(angle = 45, hjust=1)) +
  facet_grid(vars(Phenotype), vars(Annot), scales = &quot;free&quot;)

assoc_plot_mdd
dev.off()

# Make a figure combining the assoc and port plots for the manuscript
png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/Top_Bot_Port_Assoc_BMI_Height.png&#39;), units=&#39;px&#39;, res=300, width=1850, height=2200)
plot_grid(assoc_plot_height_bmi, port_plot_height_bmi, labels = c(&#39;A&#39;, &#39;B&#39;), nrow=2)
dev.off()

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/Top_Bot_Port_Assoc_MDD.png&#39;), units=&#39;px&#39;, res=300, width=1850, height=1400)
plot_grid(assoc_plot_mdd, port_plot_mdd, labels = c(&#39;A&#39;, &#39;B&#39;), nrow=2)
dev.off()

###
# Make a combined plot of elastic net models results
###

# Make one plot showing results for each quantile
# and another plot comparing the All SNP and all quantile models
png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/EAS/quant_pred_eval_BMI_Height.png&#39;), units=&#39;px&#39;, res=300, width=2200, height=1600)
tmp&lt;-eval_all[grepl(&#39;BMI|Height&#39;, eval_all$Phenotype) &amp; !grepl(&#39;All&#39;, eval_all$Type),]
tmp_comp&lt;-comp_all[grepl(&#39;BMI|Height&#39;, comp_all$Phenotype) &amp; !grepl(&#39;All|all&#39;, comp_all$Model_1) &amp; !grepl(&#39;All|all&#39;, comp_all$Model_2),]

tmp_comp$TopBot_1&lt;-gsub(&#39;.*LDcorr\\.&#39;, &#39;&#39;, tmp_comp$Model_1)
tmp_comp$TopBot_1&lt;-gsub(&#39;.*Fst\\.&#39;, &#39;&#39;, tmp_comp$TopBot_1)
tmp_comp$TopBot_1&lt;-gsub(&#39;\\..*&#39;,&#39;&#39;, tmp_comp$TopBot_1)
tmp_comp$TopBot_1&lt;-paste0(toupper(substr(tmp_comp$TopBot_1, 1, 1)), substr(tmp_comp$TopBot_1, 2, nchar(tmp_comp$TopBot_1)))

tmp_comp$TopBot_2&lt;-gsub(&#39;.*LDcorr\\.&#39;, &#39;&#39;, tmp_comp$Model_2)
tmp_comp$TopBot_2&lt;-gsub(&#39;.*Fst\\.&#39;, &#39;&#39;, tmp_comp$TopBot_2)
tmp_comp$TopBot_2&lt;-gsub(&#39;\\..*&#39;,&#39;&#39;, tmp_comp$TopBot_2)
tmp_comp$TopBot_2&lt;-paste0(toupper(substr(tmp_comp$TopBot_2, 1, 1)), substr(tmp_comp$TopBot_2, 2, nchar(tmp_comp$TopBot_2)))

tmp_comp$Quant_1&lt;-paste0(&#39;0.&#39;,gsub(&#39;.*\\.&#39;,&#39;&#39;,tmp_comp$Model_1))
tmp_comp$Quant_2&lt;-paste0(&#39;0.&#39;,gsub(&#39;.*\\.&#39;,&#39;&#39;,tmp_comp$Model_2))

tmp_comp$Quant&lt;-tmp_comp$Quant_1
tmp_comp$TopBot&lt;-tmp_comp$TopBot_1

tmp_comp&lt;-tmp_comp[tmp_comp$Quant_1 == tmp_comp$Quant_2,]
tmp_comp&lt;-tmp_comp[tmp_comp$TopBot_1 != tmp_comp$TopBot_2,]

tmp_comp$R_diff_pval&lt;-format(tmp_comp$R_diff_pval, scientific = TRUE, digits = 2)
tmp_comp$R_diff_pval[tmp_comp$R_diff_pval != &quot;0.0e+00&quot;]&lt;-paste0(&#39;=&#39;,tmp_comp$R_diff_pval[tmp_comp$R_diff_pval != &quot;0.0e+00&quot;])
tmp_comp$R_diff_pval[tmp_comp$R_diff_pval == &quot;0.0e+00&quot;]&lt;-&#39;&lt;1e-320&#39;
tmp_comp$Label&lt;-paste0(round(abs(tmp_comp$R_diff/tmp_comp$Model_2_R) * 100,1),&#39;%\np&#39;, tmp_comp$R_diff_pval)

tmp&lt;-tmp[,c(&#39;Quant&#39;,&#39;R&#39;,&#39;SE&#39;,&#39;TopBot&#39;,&#39;Annot&#39;,&#39;Phenotype&#39;), with=F]
tmp_comp&lt;-tmp_comp[,c(&#39;Quant&#39;,&#39;Phenotype&#39;,&#39;Annot&#39;,&#39;Label&#39;,&#39;TopBot&#39;), with=F]

tmp_both&lt;-merge(tmp, tmp_comp, by=c(&#39;Quant&#39;,&#39;Annot&#39;,&#39;Phenotype&#39;,&#39;TopBot&#39;), all.x=T)

if(min(tmp_both$R-tmp_both$SE) &lt; 0){
    ylim_range&lt;-max(tmp_both$R+tmp_both$SE)-min(tmp_both$R-tmp_both$SE)
  } else {
    ylim_range&lt;-max(tmp_both$R+tmp_both$SE)
}
  
eval_plot_bmi_height&lt;-ggplot(tmp_both, aes(x=Quant, y=R)) +
    geom_bar(data=tmp_both, aes(fill=TopBot), stat=&quot;identity&quot;, position=position_dodge(.9)) +
    geom_errorbar(aes(ymin=R-SE, ymax=R+SE, group=TopBot), width=.2,position=position_dodge(.9)) +
    geom_text(aes(y=max(tmp$R+tmp$SE)+(ylim_range/10)), label=tmp_both$Label, size=3) +
    theme_half_open() +
    ylim(c(min(tmp_both$R-tmp_both$SE),max(tmp_both$R+tmp_both$SE)+(ylim_range/10)+(ylim_range/10))) +
    background_grid() +
    labs(y=&quot;R (SE)&quot;, fill=&#39;&#39;, x=&#39;Annotation Quantile&#39;) +
    theme(axis.text.x = element_text(angle = 45, hjust=1)) +
    panel_border() +
    facet_grid(vars(Phenotype), vars(Annot), scales = &quot;free&quot;)

  eval_plot_bmi_height
dev.off()

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/EAS/quant_pred_eval_MDD.png&#39;), units=&#39;px&#39;, res=300, width=2200, height=1000)
tmp&lt;-eval_all[grepl(&#39;MDD&#39;, eval_all$Phenotype) &amp; !grepl(&#39;All&#39;, eval_all$Type),]
tmp_comp&lt;-comp_all[grepl(&#39;MDD&#39;, comp_all$Phenotype) &amp; !grepl(&#39;All|all&#39;, comp_all$Model_1) &amp; !grepl(&#39;All|all&#39;, comp_all$Model_2),]

tmp_comp$TopBot_1&lt;-gsub(&#39;.*LDcorr\\.&#39;, &#39;&#39;, tmp_comp$Model_1)
tmp_comp$TopBot_1&lt;-gsub(&#39;.*Fst\\.&#39;, &#39;&#39;, tmp_comp$TopBot_1)
tmp_comp$TopBot_1&lt;-gsub(&#39;\\..*&#39;,&#39;&#39;, tmp_comp$TopBot_1)
tmp_comp$TopBot_1&lt;-paste0(toupper(substr(tmp_comp$TopBot_1, 1, 1)), substr(tmp_comp$TopBot_1, 2, nchar(tmp_comp$TopBot_1)))

tmp_comp$TopBot_2&lt;-gsub(&#39;.*LDcorr\\.&#39;, &#39;&#39;, tmp_comp$Model_2)
tmp_comp$TopBot_2&lt;-gsub(&#39;.*Fst\\.&#39;, &#39;&#39;, tmp_comp$TopBot_2)
tmp_comp$TopBot_2&lt;-gsub(&#39;\\..*&#39;,&#39;&#39;, tmp_comp$TopBot_2)
tmp_comp$TopBot_2&lt;-paste0(toupper(substr(tmp_comp$TopBot_2, 1, 1)), substr(tmp_comp$TopBot_2, 2, nchar(tmp_comp$TopBot_2)))

tmp_comp$Quant_1&lt;-paste0(&#39;0.&#39;,gsub(&#39;.*\\.&#39;,&#39;&#39;,tmp_comp$Model_1))
tmp_comp$Quant_2&lt;-paste0(&#39;0.&#39;,gsub(&#39;.*\\.&#39;,&#39;&#39;,tmp_comp$Model_2))

tmp_comp$Quant&lt;-tmp_comp$Quant_1
tmp_comp$TopBot&lt;-tmp_comp$TopBot_1

tmp_comp&lt;-tmp_comp[tmp_comp$Quant_1 == tmp_comp$Quant_2,]
tmp_comp&lt;-tmp_comp[tmp_comp$TopBot_1 != tmp_comp$TopBot_2,]

tmp_comp$R_diff_pval&lt;-format(tmp_comp$R_diff_pval, scientific = TRUE, digits = 2)
tmp_comp$R_diff_pval[tmp_comp$R_diff_pval != &quot;0.0e+00&quot;]&lt;-paste0(&#39;=&#39;,tmp_comp$R_diff_pval[tmp_comp$R_diff_pval != &quot;0.0e+00&quot;])
tmp_comp$R_diff_pval[tmp_comp$R_diff_pval == &quot;0.0e+00&quot;]&lt;-&#39;&lt;1e-320&#39;
tmp_comp$Label&lt;-paste0(round(abs(tmp_comp$R_diff/tmp_comp$Model_2_R) * 100,1),&#39;%\np&#39;, tmp_comp$R_diff_pval)

tmp&lt;-tmp[,c(&#39;Quant&#39;,&#39;R&#39;,&#39;SE&#39;,&#39;TopBot&#39;,&#39;Annot&#39;,&#39;Phenotype&#39;), with=F]
tmp_comp&lt;-tmp_comp[,c(&#39;Quant&#39;,&#39;Phenotype&#39;,&#39;Annot&#39;,&#39;Label&#39;,&#39;TopBot&#39;), with=F]

tmp_both&lt;-merge(tmp, tmp_comp, by=c(&#39;Quant&#39;,&#39;Annot&#39;,&#39;Phenotype&#39;,&#39;TopBot&#39;), all.x=T)

if(min(tmp_both$R-tmp_both$SE) &lt; 0){
    ylim_range&lt;-max(tmp_both$R+tmp_both$SE)-min(tmp_both$R-tmp_both$SE)
  } else {
    ylim_range&lt;-max(tmp_both$R+tmp_both$SE)
}
  
eval_plot_mdd&lt;-ggplot(tmp_both, aes(x=Quant, y=R)) +
    geom_bar(data=tmp_both, aes(fill=TopBot), stat=&quot;identity&quot;, position=position_dodge(.9)) +
    geom_errorbar(aes(ymin=R-SE, ymax=R+SE, group=TopBot), width=.2,position=position_dodge(.9)) +
    geom_text(aes(y=max(tmp$R+tmp$SE)+(ylim_range/10)), label=tmp_both$Label, size=3) +
    theme_half_open() +
    ylim(c(min(tmp_both$R-tmp_both$SE),max(tmp_both$R+tmp_both$SE)+(ylim_range/10)+(ylim_range/10))) +
    background_grid() +
    labs(y=&quot;R (SE)&quot;, fill=&#39;&#39;, x=&#39;Annotation Quantile&#39;) +
    theme(axis.text.x = element_text(angle = 45, hjust=1)) +
    panel_border() +
    facet_grid(vars(Phenotype), vars(Annot), scales = &quot;free&quot;)

eval_plot_mdd

dev.off()

###
# Plot all SNP model comparison
###
png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/EAS/all_snp_pred_eval_BMI_Height.png&#39;), units=&#39;px&#39;, res=300, width=1600, height=1600)

tmp&lt;-eval_all[grepl(&#39;BMI|Height&#39;, eval_all$Phenotype) &amp; grepl(&#39;All&#39;, eval_all$Type),]
tmp_comp&lt;-comp_all[grepl(&#39;BMI|Height&#39;, comp_all$Phenotype) &amp; grepl(&#39;All|all&#39;, comp_all$Model_1) &amp; grepl(&#39;All|all&#39;, comp_all$Model_2),]

tmp_comp$Model_1[tmp_comp$Model_1 != &#39;All&#39;]&lt;-&#39;All SNPs&#39;
tmp_comp$Model_2[tmp_comp$Model_2 != &#39;All&#39;]&lt;-&#39;All SNPs&#39;

tmp_comp$Model_1[tmp_comp$Model_1 == &#39;All&#39;]&lt;-paste0(&#39;All SNPs +\nAll &#39;,tmp_comp$Annot[tmp_comp$Model_1 == &#39;All&#39;])
tmp_comp$Model_2[tmp_comp$Model_2 == &#39;All&#39;]&lt;-paste0(&#39;All SNPs +\nAll &#39;,tmp_comp$Annot[tmp_comp$Model_2 == &#39;All&#39;])

tmp_comp&lt;-tmp_comp[tmp_comp$Model_1 != tmp_comp$Model_2,]

tmp_comp$R_diff_pval&lt;-format(tmp_comp$R_diff_pval, scientific = TRUE, digits = 2)
tmp_comp$R_diff_pval[tmp_comp$R_diff_pval != &quot;0.0e+00&quot;]&lt;-paste0(&#39;=&#39;,tmp_comp$R_diff_pval[tmp_comp$R_diff_pval != &quot;0.0e+00&quot;])
tmp_comp$R_diff_pval[tmp_comp$R_diff_pval == &quot;0.0e+00&quot;]&lt;-&#39;&lt;1e-320&#39;
tmp_comp$Label&lt;-paste0(round(abs(tmp_comp$R_diff/tmp_comp$Model_2_R) * 100,1),&#39;%\np&#39;, tmp_comp$R_diff_pval)

tmp&lt;-tmp[,c(&#39;Type&#39;,&#39;R&#39;,&#39;SE&#39;,&#39;Annot&#39;,&#39;Phenotype&#39;), with=F]
tmp_comp&lt;-tmp_comp[,c(&#39;Model_1&#39;,&#39;Model_2&#39;,&#39;Phenotype&#39;,&#39;Annot&#39;,&#39;Label&#39;), with=F]

tmp_both&lt;-merge(tmp, tmp_comp, by.x=c(&#39;Type&#39;,&#39;Annot&#39;,&#39;Phenotype&#39;), by.y=c(&#39;Model_2&#39;,&#39;Annot&#39;,&#39;Phenotype&#39;), all.x=T)

if(min(tmp_both$R-tmp_both$SE) &lt; 0){
    ylim_range&lt;-max(tmp_both$R+tmp_both$SE)-min(tmp_both$R-tmp_both$SE)
  } else {
    ylim_range&lt;-max(tmp_both$R+tmp_both$SE)
}

tmp_both$Type&lt;-gsub(&#39;\\+ &#39;,&#39;\\+\n&#39;,tmp_both$Type)

y_lim&lt;-c(min(tmp_both$R-tmp_both$SE),max(tmp_both$R+tmp_both$SE)+(ylim_range/10)+(ylim_range/10)+(ylim_range/10))
if(y_lim[1] != 0){
  y_lim&lt;-c(0,y_lim[2])
}
  
ggplot(tmp_both, aes(x=Type, y=R, fill=Type)) +
    geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
    geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2,position=position_dodge(.9)) +
    geom_text(data=tmp_both, aes(x=Model_1, y=max(tmp_both$R+tmp_both$SE)+(ylim_range/10)+(ylim_range/10)), label=tmp_both$Label, size=4) +
    ylim(y_lim) +
    theme_half_open() +
    background_grid() +
    labs(y=&quot;R (SE)&quot;, fill=&#39;&#39;, x=&#39;Model&#39;) +
    theme(axis.text.x = element_text(angle = 45, hjust=1), legend.position=&#39;none&#39;) +
    panel_border() +
    facet_grid(vars(Phenotype), scales = &quot;free&quot;)
dev.off()

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/EAS/all_snp_pred_eval_MDD.png&#39;), units=&#39;px&#39;, res=300, width=1600, height=1000)

tmp&lt;-eval_all[grepl(&#39;MDD&#39;, eval_all$Phenotype) &amp; grepl(&#39;All&#39;, eval_all$Type),]
tmp_comp&lt;-comp_all[grepl(&#39;MDD&#39;, comp_all$Phenotype) &amp; grepl(&#39;All|all&#39;, comp_all$Model_1) &amp; grepl(&#39;All|all&#39;, comp_all$Model_2),]

tmp_comp$Model_1[tmp_comp$Model_1 != &#39;All&#39;]&lt;-&#39;All SNPs&#39;
tmp_comp$Model_2[tmp_comp$Model_2 != &#39;All&#39;]&lt;-&#39;All SNPs&#39;

tmp_comp$Model_1[tmp_comp$Model_1 == &#39;All&#39;]&lt;-paste0(&#39;All SNPs +\nAll &#39;,tmp_comp$Annot[tmp_comp$Model_1 == &#39;All&#39;])
tmp_comp$Model_2[tmp_comp$Model_2 == &#39;All&#39;]&lt;-paste0(&#39;All SNPs +\nAll &#39;,tmp_comp$Annot[tmp_comp$Model_2 == &#39;All&#39;])

tmp_comp&lt;-tmp_comp[tmp_comp$Model_1 != tmp_comp$Model_2,]

tmp_comp$R_diff_pval&lt;-format(tmp_comp$R_diff_pval, scientific = TRUE, digits = 2)
tmp_comp$R_diff_pval[tmp_comp$R_diff_pval != &quot;0.0e+00&quot;]&lt;-paste0(&#39;=&#39;,tmp_comp$R_diff_pval[tmp_comp$R_diff_pval != &quot;0.0e+00&quot;])
tmp_comp$R_diff_pval[tmp_comp$R_diff_pval == &quot;0.0e+00&quot;]&lt;-&#39;&lt;1e-320&#39;
tmp_comp$Label&lt;-paste0(round(abs(tmp_comp$R_diff/tmp_comp$Model_2_R) * 100,1),&#39;%\np&#39;, tmp_comp$R_diff_pval)

tmp&lt;-tmp[,c(&#39;Type&#39;,&#39;R&#39;,&#39;SE&#39;,&#39;Annot&#39;,&#39;Phenotype&#39;), with=F]
tmp_comp&lt;-tmp_comp[,c(&#39;Model_1&#39;,&#39;Model_2&#39;,&#39;Phenotype&#39;,&#39;Annot&#39;,&#39;Label&#39;), with=F]

tmp_both&lt;-merge(tmp, tmp_comp, by.x=c(&#39;Type&#39;,&#39;Annot&#39;,&#39;Phenotype&#39;), by.y=c(&#39;Model_2&#39;,&#39;Annot&#39;,&#39;Phenotype&#39;), all.x=T)

if(min(tmp_both$R-tmp_both$SE) &lt; 0){
    ylim_range&lt;-max(tmp_both$R+tmp_both$SE)-min(tmp_both$R-tmp_both$SE)
  } else {
    ylim_range&lt;-max(tmp_both$R+tmp_both$SE)
}

tmp_both$Type&lt;-gsub(&#39;\\+ &#39;,&#39;\\+\n&#39;,tmp_both$Type)

y_lim&lt;-c(min(tmp_both$R-tmp_both$SE),max(tmp_both$R+tmp_both$SE)+(ylim_range/10)+(ylim_range/10)+(ylim_range/10))
if(y_lim[1] != 0 &amp; !(y_lim[1] &lt; 0)){
  y_lim&lt;-c(0,y_lim[2])
}
if(y_lim[2] &lt; 0){
  y_lim[2]&lt;-0
}

ggplot(tmp_both, aes(x=Type, y=R, fill=Type)) +
    geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
    geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2,position=position_dodge(.9)) +
    geom_text(data=tmp_both, aes(x=Model_1, y=max(tmp_both$R+tmp_both$SE)+(ylim_range/10)+(ylim_range/10)), label=tmp_both$Label, size=4) +
    ylim(y_lim) +
    theme_half_open() +
    background_grid() +
    labs(y=&quot;R (SE)&quot;, fill=&#39;&#39;, x=&#39;Model&#39;) +
    theme(axis.text.x = element_text(angle = 45, hjust=1), legend.position=&#39;none&#39;) +
    panel_border() +
    facet_grid(vars(Phenotype), scales = &quot;free&quot;)
dev.off()

#################
# Plot CrossMeta results
###############

assoc_all&lt;-NULL
for(i in 1:length(pheno)){
for(targ_pop in c(pop1[i], pop2[i])){

############
# Plot the predictive utility of each predictor
############

  assoc&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/prs_1.test.assoc.txt&#39;))
  assoc$Target&lt;-targ_pop

  assoc$pT&lt;-gsub(&#39;e.&#39;,&#39;e-&#39;,gsub(&#39;.*SCORE_&#39;,&#39;&#39;,assoc$Predictor))
  
  assoc$Type[grepl(paste0(&#39;Group_&#39;,samp1[i],&#39;.&#39;,pheno[i],&#39;.&#39;,pop1[i],&#39;.single&#39;), assoc$Predictor)]&lt;-paste0(pop1[i],&#39; GWAS&#39;)
  assoc$Type[grepl(paste0(&#39;Group_&#39;,samp2[i],&#39;.&#39;,pheno[i],&#39;.&#39;,pop2[i],&#39;.single&#39;), assoc$Predictor)]&lt;-paste0(pop2[i],&#39; GWAS&#39;)

  assoc$Type[grepl(paste0(&#39;Group_&#39;,pheno[i],&#39;.&#39;,samp1[i],&#39;.&#39;,pop1[i],&#39;.&#39;,samp2[i],&#39;.&#39;,pop2[i],&#39;.standard&#39;), assoc$Predictor)]&lt;-&#39;Meta standard&#39;

  assoc$Type[grepl(paste0(&#39;Group_&#39;,pheno[i],&#39;.&#39;,samp1[i],&#39;.&#39;,pop1[i],&#39;.&#39;,samp2[i],&#39;.&#39;,pop2[i],&#39;.integer&#39;), assoc$Predictor)]&lt;-&#39;Meta integer&#39;
  assoc$Type[grepl(paste0(&#39;Group_&#39;,pheno[i],&#39;.&#39;,samp1[i],&#39;.&#39;,pop1[i],&#39;.&#39;,samp2[i],&#39;.&#39;,pop2[i],&#39;.LDcorr&#39;), assoc$Predictor)]&lt;-&#39;Meta LDcorr&#39;
  assoc$Type[grepl(paste0(&#39;Group_&#39;,pheno[i],&#39;.&#39;,samp1[i],&#39;.&#39;,pop1[i],&#39;.&#39;,samp2[i],&#39;.&#39;,pop2[i],&#39;.Fst&#39;), assoc$Predictor)]&lt;-&#39;Meta Fst&#39;

  assoc$Annot_weight&lt;-NA
  assoc$Annot_weight[assoc$Type == &#39;Meta integer&#39;]&lt;-gsub(&#39;\\.Pred.*&#39;,&#39;&#39;,gsub(&#39;.*integer.k.&#39;,&#39;&#39;,assoc$Predictor[assoc$Type == &#39;Meta integer&#39;]))
  assoc$Annot_weight[assoc$Type == &#39;Meta Fst&#39;]&lt;-gsub(&#39;\\.Pred.*&#39;,&#39;&#39;,gsub(&#39;.*Fst.q.&#39;,&#39;&#39;,assoc$Predictor[assoc$Type == &#39;Meta Fst&#39;]))
  assoc$Annot_weight[assoc$Type == &#39;Meta LDcorr&#39;]&lt;-gsub(&#39;\\.Pred.*&#39;,&#39;&#39;,gsub(&#39;.*LDcorr.q.&#39;,&#39;&#39;,assoc$Predictor[assoc$Type == &#39;Meta LDcorr&#39;]))
  
  assoc$Type[!is.na(assoc$Annot_weight) &amp; (assoc$Type == &#39;Meta Fst&#39; | assoc$Type == &#39;Meta LDcorr&#39;)]&lt;-paste0(assoc$Type[!is.na(assoc$Annot_weight) &amp; (assoc$Type == &#39;Meta Fst&#39; | assoc$Type == &#39;Meta LDcorr&#39;)], &#39; q=&#39;, assoc$Annot_weight[!is.na(assoc$Annot_weight) &amp; (assoc$Type == &#39;Meta Fst&#39; | assoc$Type == &#39;Meta LDcorr&#39;)])
  assoc$Type[!is.na(assoc$Annot_weight) &amp; (assoc$Type == &#39;Meta integer&#39;)]&lt;-paste0(assoc$Type[!is.na(assoc$Annot_weight) &amp; (assoc$Type == &#39;Meta integer&#39;)], &#39; k=&#39;, assoc$Annot_weight[!is.na(assoc$Annot_weight) &amp; (assoc$Type == &#39;Meta integer&#39;)])

  assoc$pT&lt;-factor(assoc$pT, levels=unique(assoc$pT))
  assoc$Type&lt;-factor(assoc$Type, levels=unique(assoc$Type))

  # Subset results to only show pT=0.5,0.1,0.01,1e-4,1e-8 to avoid redundancy
  assoc&lt;-assoc[assoc$pT %in% c(&#39;1e-08&#39;,&#39;1e-06&#39;,&#39;1e-04&#39;,&#39;0.01&#39;,&#39;0.1&#39;,&#39;0.5&#39;),]

  if(min(assoc$BETA-assoc$SE) &lt; 0){
    ymin=min(assoc$BETA-assoc$SE)
  } else {
    ymin&lt;-0
  }
  
  assoc_plot&lt;-list()
  
  assoc_plot[[&#39;1&#39;]]&lt;-ggplot(assoc[(assoc$Type == &#39;EUR GWAS&#39; | assoc$Type == &#39;EAS GWAS&#39; | assoc$Type == &#39;Meta standard&#39;)], aes(x=pT, y=BETA, fill=Type)) +
    geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
    geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
    theme_half_open() +
    background_grid() +
    ylim(ymin,max(assoc$BETA+assoc$SE)) +
    labs(y=&quot;R (SE)&quot;) +
    theme(axis.text.x = element_text(angle = 45, hjust=1))

  assoc_plot[[&#39;2&#39;]]&lt;-ggplot(assoc[(grepl(&#39;Meta integer&#39;, assoc$Type) | assoc$Type == &#39;Meta standard&#39;)], aes(x=pT, y=BETA, fill=Type)) +
  geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
  geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
  theme_half_open() +
  background_grid() +
  ylim(ymin,max(assoc$BETA+assoc$SE)) +
  labs(y=&quot;R (SE)&quot;) +
  theme(axis.text.x = element_text(angle = 45, hjust=1))

  assoc_plot[[&#39;3&#39;]]&lt;-ggplot(assoc[(grepl(&#39;Meta LDcorr&#39;, assoc$Type) | assoc$Type == &#39;Meta standard&#39;)], aes(x=pT, y=BETA, fill=Type)) +
  geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
  geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
  theme_half_open() +
  background_grid() +
  ylim(ymin,max(assoc$BETA+assoc$SE)) +
  labs(y=&quot;R (SE)&quot;) +
  theme(axis.text.x = element_text(angle = 45, hjust=1))
  
  assoc_plot[[&#39;4&#39;]]&lt;-ggplot(assoc[(grepl(&#39;Meta Fst&#39;, assoc$Type) | assoc$Type == &#39;Meta standard&#39;)], aes(x=pT, y=BETA, fill=Type)) +
  geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
  geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
  theme_half_open() +
  background_grid() +
  ylim(ymin,max(assoc$BETA+assoc$SE)) +
  labs(y=&quot;R (SE)&quot;) +
  theme(axis.text.x = element_text(angle = 45, hjust=1))

  png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/prs_1.test.pT_assoc.png&#39;), units=&#39;px&#39;, res=300, width=2000, height=3000)
    print(plot_grid(plotlist=assoc_plot, ncol=1))
  dev.off()
  
  assoc$Phenotype&lt;-pheno[i]
  assoc$GWAS_1&lt;-paste0(pheno[i],&#39; &#39;,samp1[i],&#39; &#39;,pop1[i])
  assoc$GWAS_2&lt;-paste0(pheno[i],&#39; &#39;,samp2[i],&#39; &#39;,pop2[i])
  assoc$Target&lt;-targ_pop
  assoc_all&lt;-rbind(assoc_all, assoc)
  }
}

# Make a plot summarising the result when using EUR, EAS and meta-GWAS across p-value thresholds
png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/prs_1.test.pT_assoc.all_targ.bmi_height.png&#39;), units=&#39;px&#39;, res=300, width=2700, height=1800)
ggplot(assoc_all[(assoc_all$Type == &#39;EUR GWAS&#39; | assoc_all$Type == &#39;EAS GWAS&#39; | assoc_all$Type == &#39;Meta standard&#39;) &amp; grepl(&#39;BMI|Height&#39;, assoc_all$Phenotype)], aes(x=pT, y=BETA, fill=Type)) +
    geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
    geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
    theme_half_open() +
    background_grid() +
    labs(y=&quot;R (SE)&quot;) +
    theme(axis.text.x = element_text(angle = 45, hjust=1)) +
    panel_border() +
    facet_grid(vars(Phenotype), vars(Target), scales = &quot;free&quot;)
dev.off()

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/prs_1.test.pT_assoc.all_targ.mdd.png&#39;), units=&#39;px&#39;, res=300, width=2700, height=2200)
ggplot(assoc_all[(assoc_all$Type == &#39;EUR GWAS&#39; | assoc_all$Type == &#39;EAS GWAS&#39; | assoc_all$Type == &#39;Meta standard&#39;) &amp; grepl(&#39;MDD&#39;, assoc_all$Phenotype)], aes(x=pT, y=BETA, fill=Type)) +
    geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
    geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
    theme_half_open() +
    background_grid() +
    labs(y=&quot;R (SE)&quot;) +
    theme(axis.text.x = element_text(angle = 45, hjust=1)) +
    panel_border() +
    facet_grid(vars(paste0(GWAS_1,&#39; vs.\n&#39;,GWAS_2)), vars(Target), scales = &quot;free&quot;)
dev.off()

# Make plot summarising results when reweighting meta-analysis
int&lt;-ggplot(assoc_all[(grepl(&#39;Meta integer&#39;, assoc_all$Type) | assoc_all$Type == &#39;Meta standard&#39;) &amp; grepl(&#39;BMI|Height&#39;, assoc_all$Phenotype)], aes(x=pT, y=BETA, fill=Type)) +
    geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
    geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
    theme_half_open() +
    background_grid() +
    labs(y=&quot;R (SE)&quot;) +
    theme(axis.text.x = element_text(angle = 45, hjust=1)) +
    panel_border() +
    facet_grid(vars(Phenotype), vars(Target), scales = &quot;free&quot;)

ldcor&lt;-ggplot(assoc_all[(grepl(&#39;Meta LDcorr&#39;, assoc_all$Type) | assoc_all$Type == &#39;Meta standard&#39;) &amp; grepl(&#39;BMI|Height&#39;, assoc_all$Phenotype)], aes(x=pT, y=BETA, fill=Type)) +
    geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
    geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
    theme_half_open() +
    background_grid() +
    labs(y=&quot;R (SE)&quot;) +
    theme(axis.text.x = element_text(angle = 45, hjust=1)) +
    panel_border() +
    facet_grid(vars(Phenotype), vars(Target), scales = &quot;free&quot;)

fst&lt;-ggplot(assoc_all[(grepl(&#39;Meta Fst&#39;, assoc_all$Type) | assoc_all$Type == &#39;Meta standard&#39;) &amp; grepl(&#39;BMI|Height&#39;, assoc_all$Phenotype)], aes(x=pT, y=BETA, fill=Type)) +
    geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
    geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
    theme_half_open() +
    background_grid() +
    labs(y=&quot;R (SE)&quot;) +
    theme(axis.text.x = element_text(angle = 45, hjust=1)) +
    panel_border() +
    facet_grid(vars(Phenotype), vars(Target), scales = &quot;free&quot;)

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/prs_1.test.pT_assoc.all_targ.bmi_height.meta.png&#39;), units=&#39;px&#39;, res=300, width=2700, height=4000)
plot_grid(int, ldcor, fst, labels= c(&#39;A&#39;, &#39;B&#39;,&#39;C&#39;), nrow=3)
dev.off()

int&lt;-ggplot(assoc_all[(grepl(&#39;Meta integer&#39;, assoc_all$Type) | assoc_all$Type == &#39;Meta standard&#39;) &amp; grepl(&#39;MDD&#39;, assoc_all$Phenotype)], aes(x=pT, y=BETA, fill=Type)) +
    geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
    geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
    theme_half_open() +
    background_grid() +
    labs(y=&quot;R (SE)&quot;) +
    theme(axis.text.x = element_text(angle = 45, hjust=1)) +
    panel_border() +
    facet_grid(vars(gsub(&#39;MDD &#39;,&#39;&#39;,paste0(GWAS_1,&#39; vs.\n&#39;,GWAS_2))), vars(Target), scales = &quot;free&quot;)

ldcor&lt;-ggplot(assoc_all[(grepl(&#39;Meta LDcorr&#39;, assoc_all$Type) | assoc_all$Type == &#39;Meta standard&#39;) &amp; grepl(&#39;MDD&#39;, assoc_all$Phenotype)], aes(x=pT, y=BETA, fill=Type)) +
    geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
    geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
    theme_half_open() +
    background_grid() +
    labs(y=&quot;R (SE)&quot;) +
    theme(axis.text.x = element_text(angle = 45, hjust=1)) +
    panel_border() +
    facet_grid(vars(gsub(&#39;MDD &#39;,&#39;&#39;,paste0(GWAS_1,&#39; vs.\n&#39;,GWAS_2))), vars(Target), scales = &quot;free&quot;)

fst&lt;-ggplot(assoc_all[(grepl(&#39;Meta Fst&#39;, assoc_all$Type) | assoc_all$Type == &#39;Meta standard&#39;) &amp; grepl(&#39;MDD&#39;, assoc_all$Phenotype)], aes(x=pT, y=BETA, fill=Type)) +
    geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
    geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
    theme_half_open() +
    background_grid() +
    labs(y=&quot;R (SE)&quot;) +
    theme(axis.text.x = element_text(angle = 45, hjust=1)) +
    panel_border() +
    facet_grid(vars(gsub(&#39;MDD &#39;,&#39;&#39;,paste0(GWAS_1,&#39; vs.\n&#39;,GWAS_2))), vars(Target), scales = &quot;free&quot;)

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/prs_1.test.pT_assoc.all_targ.mdd.meta.png&#39;), units=&#39;px&#39;, res=300, width=2700, height=5000)
plot_grid(int, ldcor, fst, labels= c(&#39;A&#39;, &#39;B&#39;,&#39;C&#39;), nrow=3)
dev.off()

###############
# Plot comparisons between all models
###############

# Plot the absolute results
eval_all_all&lt;-list()

for(i in 1:length(pheno)){
for(targ_pop in c(pop1[i], pop2[i])){
  eval_integer&lt;- fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/prs_2.test.pred_eval.txt&#39;))
  eval_integer$Label&lt;-c(&#39;Baseline&#39;,&#39;Meta integer&#39;,&#39;Baseline +\nMeta integer&#39;)

  eval_LDcorr&lt;- fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/prs_LDcorr.test.pred_eval.txt&#39;))
  eval_LDcorr$Label&lt;-c(&#39;Baseline&#39;,&#39;Meta LDcorr&#39;,&#39;Baseline +\nMeta LDcorr&#39;)
  
  eval_Fst&lt;- fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/prs_Fst.test.pred_eval.txt&#39;))
  eval_Fst$Label&lt;-c(&#39;Baseline&#39;,&#39;Meta Fst&#39;,&#39;Baseline +\nMeta Fst&#39;)

  eval_Both&lt;- fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/prs_3.test.pred_eval.txt&#39;))
  eval_Both$Label&lt;-c(&#39;Baseline&#39;,&#39;Meta LDcorr + \nMeta Fst&#39;,&#39;Baseline +\nMeta LDcorr +\nMeta Fst&#39;)

  eval_all&lt;-do.call(rbind, list(eval_integer, eval_LDcorr,eval_Fst,eval_Both))
  eval_all&lt;-eval_all[!duplicated(eval_all),]
  
  label_order&lt;-c(&#39;Meta integer&#39;,&#39;Meta LDcorr&#39;,&#39;Meta Fst&#39;,&#39;Meta LDcorr + \nMeta Fst&#39;,&#39;Baseline&#39;,&#39;Baseline +\nMeta integer&#39;,&#39;Baseline +\nMeta LDcorr&#39;,&#39;Baseline +\nMeta Fst&#39;,&#39;Baseline +\nMeta LDcorr +\nMeta Fst&#39;)
  
  eval_all&lt;-eval_all[match(label_order, eval_all$Label),]
  eval_all$Label&lt;-factor(eval_all$Label, levels=eval_all$Label)

  eval_all$Target&lt;-targ_pop

  eval_all&lt;-eval_all[grepl(&#39;Baseline&#39;, eval_all$Label),]
  
  eval_plot&lt;-ggplot(eval_all, aes(x=Label, y=R, colour=Label, group = 1)) +
  geom_point(stat=&quot;identity&quot;, position=position_dodge(.9), size=5) +
  geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2,position=position_dodge(.9)) +
  theme_half_open() +
  background_grid() +
  ylim(min(eval_all$R-eval_all$SE),max(eval_all$R+eval_all$SE)+0.075) +
  labs(y=&quot;R (SE)&quot;, x=&#39;&#39;, title=paste0(targ_pop,&#39; target: &#39;,pheno[i],&#39;-&#39;,samp1[i],&#39;-&#39;,pop1[i],&#39; + &#39;,pheno[i],&#39;-&#39;,samp1[i],&#39;-&#39;,pop1[i])) +
  theme(axis.text.x = element_text(angle = 45, hjust=1), legend.position = &quot;none&quot;)

  comp_integer&lt;- fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/prs_2.test.pred_comp.txt&#39;))
  comp_integer&lt;-comp_integer[comp_integer$Model_1 == &#39;All&#39; &amp; comp_integer$Model_2 == &#39;single.and.standard&#39; ]

  comp_LDcorr&lt;- fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/prs_LDcorr.test.pred_comp.txt&#39;))
  comp_LDcorr&lt;-comp_LDcorr[comp_LDcorr$Model_1 == &#39;All&#39; &amp; comp_LDcorr$Model_2 == &#39;single.and.standard&#39; ]

  comp_Fst&lt;- fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/prs_Fst.test.pred_comp.txt&#39;))
  comp_Fst&lt;-comp_Fst[comp_Fst$Model_1 == &#39;All&#39; &amp; comp_Fst$Model_2 == &#39;single.and.standard&#39; ]

  comp_BOTH&lt;- fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/prs_3.test.pred_comp.txt&#39;))
  comp_BOTH&lt;-comp_BOTH[comp_BOTH$Model_1 == &#39;All&#39; &amp; comp_BOTH$Model_2 == &#39;single.and.standard&#39; ]

  paths_integer&lt;-data.frame(x=c(1,1,2,2),
                           y=c(max(eval_all$R+eval_all$SE)+0.005,max(eval_all$R+eval_all$SE)+0.01,max(eval_all$R+eval_all$SE)+0.01,max(eval_all$R+eval_all$SE)+0.005))
  
  paths_LDcorr&lt;-paths_integer
  paths_LDcorr$x[paths_LDcorr$x != 1]&lt;-paths_LDcorr$x[paths_LDcorr$x != 1]+1
  paths_LDcorr$y&lt;-paths_LDcorr$y+0.015

  paths_Fst&lt;-paths_LDcorr
  paths_Fst$x[paths_Fst$x != 1]&lt;-paths_Fst$x[paths_Fst$x != 1]+1
  paths_Fst$y&lt;-paths_Fst$y+0.015

  paths_both&lt;-paths_Fst
  paths_both$x[paths_both$x != 1]&lt;-paths_both$x[paths_both$x != 1]+1
  paths_both$y&lt;-paths_both$y+0.015
  
  eval_plot&lt;-eval_plot + 
    geom_path(data=paths_integer, aes(x=x,y=y), colour=&#39;black&#39;) +
    geom_path(data=paths_LDcorr, aes(x=x,y=y), colour=&#39;black&#39;) +
    geom_path(data=paths_Fst, aes(x=x,y=y), colour=&#39;black&#39;) +
    geom_path(data=paths_both, aes(x=x,y=y), colour=&#39;black&#39;)
    
  eval_plot&lt;-eval_plot + 
    annotate(&quot;text&quot;,x=1.5,y=max(eval_all$R+eval_all$SE)+0.015,label=paste0(&#39;diff=&#39;, round(abs(comp_integer$R_diff/comp_integer$Model_2_R) * 100,1),&#39;%; p=&#39;, format(comp_integer$R_diff_pval, scientific = TRUE, digits = 2)), size=4) +
    annotate(&quot;text&quot;,x=2,y=max(eval_all$R+eval_all$SE)+0.03,label=paste0(&#39;diff=&#39;, round(abs(comp_LDcorr$R_diff/comp_LDcorr$Model_2_R) * 100,1),&#39;%; p=&#39;, format(comp_LDcorr$R_diff_pval, scientific = TRUE, digits = 2)), size=4) +
    annotate(&quot;text&quot;,x=2.5,y=max(eval_all$R+eval_all$SE)+0.045,label=paste0(&#39;diff=&#39;, round(abs(comp_Fst$R_diff/comp_Fst$Model_2_R) * 100,1),&#39;%; p=&#39;, format(comp_Fst$R_diff_pval, scientific = TRUE, digits = 2)), size=4) +
    annotate(&quot;text&quot;,x=3,y=max(eval_all$R+eval_all$SE)+0.06,label=paste0(&#39;diff=&#39;, round(abs(comp_BOTH$R_diff/comp_BOTH$Model_2_R) * 100,1),&#39;%; p=&#39;, format(comp_BOTH$R_diff_pval, scientific = TRUE, digits = 2)), size=4)

  png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/prs_all.test.comp.png&#39;), units=&#39;px&#39;, res=300, width=2200, height=1500)
    print(eval_plot)
  dev.off()
  
  eval_all$text&lt;-c(NA, 
                    paste0(&#39;diff=&#39;, round(abs(comp_integer$R_diff/comp_integer$Model_2_R) * 100,1),&#39;%\np=&#39;, format(comp_integer$R_diff_pval, scientific = TRUE, digits = 2)),
                    paste0(&#39;diff=&#39;, round(abs(comp_LDcorr$R_diff/comp_LDcorr$Model_2_R) * 100,1),&#39;%\np=&#39;, format(comp_LDcorr$R_diff_pval, scientific = TRUE, digits = 2)),
                    paste0(&#39;diff=&#39;, round(abs(comp_Fst$R_diff/comp_Fst$Model_2_R) * 100,1),&#39;%\np=&#39;, format(comp_Fst$R_diff_pval, scientific = TRUE, digits = 2)),
                    paste0(&#39;diff=&#39;, round(abs(comp_BOTH$R_diff/comp_BOTH$Model_2_R) * 100,1),&#39;%\np=&#39;, format(comp_BOTH$R_diff_pval, scientific = TRUE, digits = 2)))
  
  eval_all$Test&lt;-paste0(pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i])
  eval_all$Phenotype&lt;-pheno[i]
  
  eval_all_all&lt;-rbind(eval_all_all, eval_all)
  }
}

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/EUR/prs_all.test.comp.eur.height_bmi.png&#39;), units=&#39;px&#39;, res=300, width=2200, height=1800)

tmp&lt;-eval_all_all[eval_all_all$Target == &#39;EUR&#39; &amp; grepl(&#39;BMI|Height&#39;, eval_all_all$Test),]

tmp$text[is.na(tmp$text)]&lt;-&#39;&#39;

ylim_range&lt;-max(tmp$R+tmp$SE)-min(tmp$R-tmp$SE)

ggplot(tmp, aes(x=Label, y=R, colour=Label, group = 1)) +
  geom_point(stat=&quot;identity&quot;, position=position_dodge(.9), size=5) +
  geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2,position=position_dodge(.9)) +
  geom_text(data=tmp, aes(x=Label, y=tmp$R+tmp$SE+(ylim_range/100)), label=paste0(&#39;\n&#39;,tmp$text), colour=&#39;black&#39;, size=3) +
  theme_half_open() +
  background_grid() +
  labs(y=&quot;R (SE)&quot;, x=&#39;&#39;, title=&#39;EUR target&#39;) +
  theme(axis.text.x = element_text(angle = 45, hjust=1), legend.position = &quot;none&quot;) +
  panel_border() +
  facet_grid(vars(Phenotype), scales = &quot;free&quot;)

dev.off()


png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/EUR/prs_all.test.comp.eur.mdd.png&#39;), units=&#39;px&#39;, res=300, width=2200, height=2200)

tmp&lt;-eval_all_all[eval_all_all$Target == &#39;EUR&#39; &amp; grepl(&#39;MDD&#39;, eval_all_all$Test),]

tmp$text[is.na(tmp$text)]&lt;-&#39;&#39;
tmp$Name&lt;-gsub(&#39;EUR &#39;,&#39;EUR \\+\n&#39;,gsub(&#39;_&#39;, &#39; &#39;,gsub(&#39;MDD_&#39;,&#39;&#39;, tmp$Test)))

ylim_range&lt;-max(tmp$R+tmp$SE)-min(tmp$R-tmp$SE)

ggplot(tmp, aes(x=Label, y=R, colour=Label, group = 1)) +
  geom_point(stat=&quot;identity&quot;, position=position_dodge(.9), size=5) +
  geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2,position=position_dodge(.9)) +
  geom_text(data=tmp, aes(x=Label, y=tmp$R+tmp$SE+(ylim_range/5)), label=paste0(&#39;\n&#39;,tmp$text), colour=&#39;black&#39;, size=3) +
  theme_half_open() +
  background_grid() +
  labs(y=&quot;R (SE)&quot;, x=&#39;&#39;, title=&#39;EUR target&#39;) +
  theme(axis.text.x = element_text(angle = 45, hjust=1), legend.position = &quot;none&quot;) +
  panel_border() +
  facet_grid(vars(Name), scales = &quot;free&quot;)

dev.off()

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/EAS/prs_all.test.comp.eas.height_bmi.png&#39;), units=&#39;px&#39;, res=300, width=2200, height=1800)

tmp&lt;-eval_all_all[eval_all_all$Target == &#39;EAS&#39; &amp; grepl(&#39;BMI|Height&#39;, eval_all_all$Test),]

tmp$text[is.na(tmp$text)]&lt;-&#39;&#39;

ylim_range&lt;-max(tmp$R+tmp$SE)-min(tmp$R-tmp$SE)

ggplot(tmp, aes(x=Label, y=R, colour=Label, group = 1)) +
  geom_point(stat=&quot;identity&quot;, position=position_dodge(.9), size=5) +
  geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2,position=position_dodge(.9)) +
  geom_text(data=tmp, aes(x=Label, y=tmp$R+tmp$SE+(ylim_range/5)), label=paste0(&#39;\n&#39;,tmp$text), colour=&#39;black&#39;, size=3) +
  theme_half_open() +
  background_grid() +
  labs(y=&quot;R (SE)&quot;, x=&#39;&#39;, title=&#39;EAS target&#39;) +
  theme(axis.text.x = element_text(angle = 45, hjust=1), legend.position = &quot;none&quot;) +
  panel_border() +
  facet_grid(vars(Phenotype), scales = &quot;free&quot;)

dev.off()


png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/EAS/prs_all.test.comp.eas.mdd.png&#39;), units=&#39;px&#39;, res=300, width=2200, height=2200)

tmp&lt;-eval_all_all[eval_all_all$Target == &#39;EAS&#39; &amp; grepl(&#39;MDD&#39;, eval_all_all$Test),]

tmp$text[is.na(tmp$text)]&lt;-&#39;&#39;
tmp$Name&lt;-gsub(&#39;EUR &#39;,&#39;EUR \\+\n&#39;,gsub(&#39;_&#39;, &#39; &#39;,gsub(&#39;MDD_&#39;,&#39;&#39;, tmp$Test)))

ylim_range&lt;-max(tmp$R+tmp$SE)-min(tmp$R-tmp$SE)

ggplot(tmp, aes(x=Label, y=R, colour=Label, group = 1)) +
  geom_point(stat=&quot;identity&quot;, position=position_dodge(.9), size=5) +
  geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2,position=position_dodge(.9)) +
  geom_text(data=tmp, aes(x=Label, y=tmp$R+tmp$SE+(ylim_range/5)), label=paste0(&#39;\n&#39;,tmp$text), colour=&#39;black&#39;, size=3) +
  theme_half_open() +
  background_grid() +
  labs(y=&quot;R (SE)&quot;, x=&#39;&#39;, title=&#39;EAS target&#39;) +
  theme(axis.text.x = element_text(angle = 45, hjust=1), legend.position = &quot;none&quot;) +
  panel_border() +
  facet_grid(vars(Name), scales = &quot;free&quot;)

dev.off()

##########################
# Plot PRS-CSx results
##########################

prscsx_assoc_all&lt;-NULL
eval_prscx_plot_list&lt;-list()

for(i in 1:length(pheno)){
for(targ_pop in c(pop1[i], pop2[i])){

############
# Plot the predictive utility of each predictor
############

# Plot the PRS-CSx results across phi parameters
prscsx_assoc&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/PRS-CSx_1.test.assoc.txt&#39;))
prscsx_assoc$Target&lt;-targ_pop

prscsx_assoc$phi&lt;-gsub(&#39;1e-00&#39;,&#39;1&#39;,gsub(&#39;e.&#39;,&#39;e-&#39;,gsub(&#39;.*phi_&#39;,&#39;&#39;,prscsx_assoc$Predictor)))

prscsx_assoc$Type[grepl(&#39;PredFile1&#39;, prscsx_assoc$Predictor)]&lt;-paste0(&#39;PRS-CSx &#39;, pop1[i])
prscsx_assoc$Type[grepl(&#39;PredFile2&#39;, prscsx_assoc$Predictor)]&lt;-paste0(&#39;PRS-CSx &#39;, pop2[i])
prscsx_assoc$Type[grepl(&#39;PredFile3&#39;, prscsx_assoc$Predictor)]&lt;-paste0(&#39;PRS-CSx &#39;, &#39;meta&#39;)

prscsx_assoc$Type&lt;-factor(prscsx_assoc$Type, levels=unique(prscsx_assoc$Type))
prscsx_assoc$Phi&lt;-factor(prscsx_assoc$Phi, levels=unique(prscsx_assoc$Phi))

if(min(prscsx_assoc$BETA-prscsx_assoc$SE) &lt; 0){
  ymin=min(prscsx_assoc$BETA-prscsx_assoc$SE)
} else {
  ymin&lt;-0
}

assoc_plot&lt;-ggplot(prscsx_assoc, aes(x=phi, y=BETA, fill=Type)) +
  geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
  geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
  theme_half_open() +
  background_grid() +
  ylim(ymin,max(prscsx_assoc$BETA+prscsx_assoc$SE)) +
  labs(y=&quot;R (SE)&quot;) +
  theme(axis.text.x = element_text(angle = 45, hjust=1))

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/PRS-CSx_1.test.phi_assoc.png&#39;), units=&#39;px&#39;, res=300, width=2000, height=1000)
  print(assoc_plot)
dev.off()

prscsx_assoc$Phenotype&lt;-pheno[i]
prscsx_assoc$Target&lt;-targ_pop
prscsx_assoc$Test&lt;-paste0(samp1[i],&#39; &#39;,pop1[i],&#39; +\n&#39;,samp2[i],&#39; &#39;,pop2[i])
prscsx_assoc_all&lt;-rbind(prscsx_assoc_all, prscsx_assoc)

# Plot the PRS-CSx models combined with standard meta-analysis
eval_prscsx_1&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/PRS-CSx_1.test.pred_eval.txt&#39;))
eval_prscsx_1$Label&lt;-c(&#39;PRS-CSx singles&#39;,&#39;PRS-CSx meta&#39;,&#39;PRS-CSx singles +\nPRS-CSx meta&#39;)

eval_prscsx_2&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/PRS-CSx_2.test.pred_eval.txt&#39;))
eval_prscsx_2$Label&lt;-c(&#39;Baseline&#39;,&#39;PRS-CSx singles +\nPRS-CSx meta&#39;,&#39;Baseline +\nPRS-CSx singles +\nPRS-CSx meta&#39;)
  
eval_prscx_all&lt;-do.call(rbind, list(eval_prscsx_1, eval_prscsx_2))
eval_prscx_all&lt;-eval_prscx_all[!duplicated(eval_prscx_all$Label),]

label_order&lt;-c(&quot;PRS-CSx singles&quot;,&quot;PRS-CSx meta&quot;,&quot;PRS-CSx singles +\nPRS-CSx meta&quot;,&quot;Baseline&quot;, &quot;Baseline +\nPRS-CSx singles +\nPRS-CSx meta&quot;)

eval_prscx_all&lt;-eval_prscx_all[match(label_order, eval_prscx_all$Label),]
eval_prscx_all$Label&lt;-factor(eval_prscx_all$Label, levels=eval_prscx_all$Label)

eval_prscx_all$Target&lt;-targ_pop

ylim_range&lt;-max(eval_prscx_all$R+eval_prscx_all$SE)-min(eval_prscx_all$R-eval_prscx_all$SE)
  
eval_prscx_plot&lt;-ggplot(eval_prscx_all, aes(x=Label, y=R, colour=Label, group = 1)) +
geom_point(stat=&quot;identity&quot;, position=position_dodge(.9), size=5) +
geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2,position=position_dodge(.9)) +
theme_half_open() +
background_grid() +
ylim(min(eval_prscx_all$R-eval_prscx_all$SE),max(eval_prscx_all$R+eval_prscx_all$SE)+(ylim_range/1.5)) +
labs(y=&quot;R (SE)&quot;, x=&#39;&#39;, title=paste0(targ_pop,&#39; target, &#39;,pheno[i],&#39;: &#39;,samp1[i],&#39; &#39;,pop1[i],&#39; + &#39;,samp2[i],&#39;-&#39;,pop2[i])) +
theme(legend.position = &quot;none&quot;)

comp_prscsx_1&lt;- fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/PRS-CSx_1.test.pred_comp.txt&#39;))
comp_prscsx_2&lt;- fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/PRS-CSx_2.test.pred_comp.txt&#39;))

comp_prscsx_all&lt;-rbind(comp_prscsx_1, comp_prscsx_2)

paths_1&lt;-data.frame(x=c(1,1,2,2),
                   y=c(max(eval_prscx_all$R+eval_prscx_all$SE)+((ylim_range/10)*1),max(eval_prscx_all$R+eval_prscx_all$SE)+((ylim_range/10)*2),max(eval_prscx_all$R+eval_prscx_all$SE)+((ylim_range/10)*2),max(eval_prscx_all$R+eval_prscx_all$SE)+((ylim_range/10)*1)))
  
paths_2&lt;-paths_1
paths_2$x[paths_2$x != 1]&lt;-paths_2$x[paths_2$x != 1]+0.975
paths_2$y&lt;-c(max(eval_prscx_all$R+eval_prscx_all$SE)+((ylim_range/10)*3),max(eval_prscx_all$R+eval_prscx_all$SE)+((ylim_range/10)*4),max(eval_prscx_all$R+eval_prscx_all$SE)+((ylim_range/10)*4),max(eval_prscx_all$R+eval_prscx_all$SE)+((ylim_range/10)*3))

paths_3&lt;-paths_1
paths_3$x&lt;-c(3,3,3.975,3.975)

paths_4&lt;-paths_1
paths_4$x&lt;-c(3.025,3.025,5,5)
paths_4$y&lt;-c(max(eval_prscx_all$R+eval_prscx_all$SE)+((ylim_range/10)*3),max(eval_prscx_all$R+eval_prscx_all$SE)+((ylim_range/10)*4),max(eval_prscx_all$R+eval_prscx_all$SE)+((ylim_range/10)*4),max(eval_prscx_all$R+eval_prscx_all$SE)+((ylim_range/10)*3))

paths_5&lt;-paths_1
paths_5$x&lt;-c(4.025,4.025,5,5)

eval_prscx_plot&lt;-eval_prscx_plot + 
  geom_path(data=paths_1, aes(x=x,y=y), colour=&#39;black&#39;) +
  geom_path(data=paths_2, aes(x=x,y=y), colour=&#39;black&#39;) +
  geom_path(data=paths_3, aes(x=x,y=y), colour=&#39;black&#39;) +
  geom_path(data=paths_4, aes(x=x,y=y), colour=&#39;black&#39;) +
  geom_path(data=paths_5, aes(x=x,y=y), colour=&#39;black&#39;)

comp_prscsx_1$R_diff_pval&lt;-format(comp_prscsx_1$R_diff_pval, scientific = TRUE, digits = 2)
comp_prscsx_2$R_diff_pval&lt;-format(comp_prscsx_2$R_diff_pval, scientific = TRUE, digits = 2)
comp_prscsx_1$R_diff_pval[comp_prscsx_1$R_diff_pval == &quot;0.0e+00&quot;]&lt;-&#39;&lt;1e-320&#39;
comp_prscsx_2$R_diff_pval[comp_prscsx_2$R_diff_pval == &quot;0.0e+00&quot;]&lt;-&#39;&lt;1e-320&#39;

eval_prscx_plot&lt;-eval_prscx_plot + 
  annotate(&quot;text&quot;,x=1.5,y=max(eval_prscx_all$R+eval_prscx_all$SE)+((ylim_range/10)*3),label=paste0(round(abs(comp_prscsx_1$R_diff/comp_prscsx_1$Model_2_R) * 100,1)[2],&#39;%; p=&#39;, comp_prscsx_1$R_diff_pval[2]), size=3.5) +
  annotate(&quot;text&quot;,x=2,y=max(eval_prscx_all$R+eval_prscx_all$SE)+((ylim_range/10)*5),label=paste0(round(abs(comp_prscsx_1$R_diff/comp_prscsx_1$Model_2_R) * 100,1)[7],&#39;%; p=&#39;, comp_prscsx_1$R_diff_pval[7]), size=3.5) +
  annotate(&quot;text&quot;,x=3.5,y=max(eval_prscx_all$R+eval_prscx_all$SE)+((ylim_range/10)*3),label=paste0(round(abs(comp_prscsx_2$R_diff/comp_prscsx_2$Model_2_R) * 100,1)[2],&#39;%; p=&#39;, comp_prscsx_2$R_diff_pval[2]), size=3.5) +
  annotate(&quot;text&quot;,x=4,y=max(eval_prscx_all$R+eval_prscx_all$SE)+((ylim_range/10)*5),label=paste0(round(abs(comp_prscsx_2$R_diff/comp_prscsx_2$Model_2_R) * 100,1)[8],&#39;%; p=&#39;, comp_prscsx_2$R_diff_pval[8]), size=3.5) +
  annotate(&quot;text&quot;,x=4.5,y=max(eval_prscx_all$R+eval_prscx_all$SE)+((ylim_range/10)*3),label=paste0(round(abs(comp_prscsx_2$R_diff/comp_prscsx_2$Model_2_R) * 100,1)[7],&#39;%; p=&#39;, comp_prscsx_2$R_diff_pval[7]), size=3.5)

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/&#39;,targ_pop,&#39;/&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i],&#39;/PRS-CSx_2.test.comp.png&#39;), units=&#39;px&#39;, res=300, width=2500, height=1500)
  print(eval_prscx_plot)
dev.off()
  
eval_prscx_plot_list[[paste0(targ_pop,&#39;_&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i])]]&lt;-eval_prscx_plot
}
}

# Make a condensed plot containing per phi results across EUR and EAS
png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/PRS-CSx_2.per_phi.height_bmi.png&#39;), units=&#39;px&#39;, res=300, width=2200, height=1500)
ggplot(prscsx_assoc_all[grepl(&#39;Height|BMI&#39;, prscsx_assoc_all$Phenotype),], aes(x=phi, y=BETA, fill=Type)) +
  geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
  geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
  theme_half_open() +
  background_grid() +
  labs(y=&quot;R (SE)&quot;) +
  theme(axis.text.x = element_text(angle = 45, hjust=1)) +
  panel_border() +
  facet_grid(vars(Phenotype), vars(Target), scales = &quot;free&quot;)
dev.off()

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/PRS-CSx_2.per_phi.mdd.png&#39;), units=&#39;px&#39;, res=300, width=2200, height=2000)
ggplot(prscsx_assoc_all[grepl(&#39;MDD&#39;, prscsx_assoc_all$Phenotype),], aes(x=phi, y=BETA, fill=Type)) +
  geom_bar(stat=&quot;identity&quot;, position=position_dodge(.9)) +
  geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2,position=position_dodge(.9)) +
  theme_half_open() +
  background_grid() +
  labs(y=&quot;R (SE)&quot;) +
  theme(axis.text.x = element_text(angle = 45, hjust=1)) +
  panel_border() +
  facet_grid(vars(Test), vars(Target), scales = &quot;free&quot;)
dev.off()

# Make combined plot of PRS-CSx model results
mode_plot_list_height_bmi_eur&lt;-list()
mode_plot_list_height_bmi_eas&lt;-list()
mode_plot_list_mdd_eur&lt;-list()
mode_plot_list_mdd_eas&lt;-list()
for(i in 1:length(pheno)){
for(targ_pop in c(pop1[i], pop2[i])){
  if(grepl(&#39;Height|BMI&#39;, pheno[i]) &amp; targ_pop == &#39;EUR&#39;){
    mode_plot_list_height_bmi_eur[[paste0(targ_pop,&#39;_&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i])]]&lt;-eval_prscx_plot_list[[paste0(targ_pop,&#39;_&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i])]]
  }
  if(grepl(&#39;Height|BMI&#39;, pheno[i]) &amp; targ_pop == &#39;EAS&#39;){
    mode_plot_list_height_bmi_eas[[paste0(targ_pop,&#39;_&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i])]]&lt;-eval_prscx_plot_list[[paste0(targ_pop,&#39;_&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i])]]
  }
  if(grepl(&#39;MDD&#39;, pheno[i]) &amp; targ_pop == &#39;EUR&#39;){
    mode_plot_list_mdd_eur[[paste0(targ_pop,&#39;_&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i])]]&lt;-eval_prscx_plot_list[[paste0(targ_pop,&#39;_&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i])]]
  }
  if(grepl(&#39;MDD&#39;, pheno[i]) &amp; targ_pop == &#39;EAS&#39;){
    mode_plot_list_mdd_eas[[paste0(targ_pop,&#39;_&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i])]]&lt;-eval_prscx_plot_list[[paste0(targ_pop,&#39;_&#39;,pheno[i],&#39;_&#39;,samp1[i],&#39;_&#39;,pop1[i],&#39;_&#39;,samp2[i],&#39;_&#39;,pop2[i])]]
  }
}
}

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/PRS-CSx_2.model_comp.height_bmi.eur.png&#39;), units=&#39;px&#39;, res=300, width=2500, height=2000)
plot_grid(plotlist=mode_plot_list_height_bmi_eur, ncol=1)
dev.off()

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/PRS-CSx_2.model_comp.height_bmi.eas.png&#39;), units=&#39;px&#39;, res=300, width=2500, height=2000)
plot_grid(plotlist=mode_plot_list_height_bmi_eas, ncol=1)
dev.off()

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/PRS-CSx_2.model_comp.mdd.eur.png&#39;), units=&#39;px&#39;, res=300, width=2500, height=3000)
plot_grid(plotlist=mode_plot_list_mdd_eur, ncol=1)
dev.off()

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/CrossMeta/assoc/PRS-CSx_2.model_comp.mdd.eas.png&#39;), units=&#39;px&#39;, res=300, width=2500, height=3000)
plot_grid(plotlist=mode_plot_list_mdd_eas, ncol=1)
dev.off()</code></pre>
</details>
<hr />
</div>
</div>
</div>
<div id="results" class="section level1">
<h1><span class="header-section-number">3</span> Results</h1>
<hr />
<div id="characterising-ldcorr-and-fst-across-populations" class="section level2">
<h2><span class="header-section-number">3.1</span> Characterising LDcorr and Fst across populations</h2>
<details>
<p><summary>BMI and Height</summary></p>
<p><img src="Images/CrossMeta/sign_test/res/cor_plot_Height_BMI.png" /></p>
</details>
<details>
<p><summary>MDD</summary></p>
<p><img src="Images/CrossMeta/sign_test/res/cor_plot_MDD.png" /></p>
</details>
<hr />
</div>
<div id="single-gwas" class="section level2">
<h2><span class="header-section-number">3.2</span> Single GWAS</h2>
<hr />
<div id="per-pt-results" class="section level3">
<h3><span class="header-section-number">3.2.1</span> Per pT results</h3>
<details>
<p><summary>EAS target, GWAS1: MDD_PGC2</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/MDD_PGC2_EUR/prs_1.test.pT_assoc.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: Height_Wood</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/Height_Wood_EUR/prs_1.test.pT_assoc.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: BMI_Locke</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/BMI_Locke_EUR/prs_1.test.pT_assoc.png" /></p>
</details>
<hr />
</div>
<div id="model-results" class="section level3">
<h3><span class="header-section-number">3.2.2</span> Model results</h3>
<details>
<p><summary>EAS target, GWAS1: MDD_PGC2</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/MDD_PGC2_EUR/prs_1.test.eval.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: Height_Wood</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/Height_Wood_EUR/prs_1.test.eval.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: BMI_Locke</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/BMI_Locke_EUR/prs_1.test.eval.png" /></p>
</details>
<hr />
</div>
</div>
<div id="crossmeta" class="section level2">
<h2><span class="header-section-number">3.3</span> CrossMeta</h2>
<hr />
<div id="per-pt-results-1" class="section level3">
<h3><span class="header-section-number">3.3.1</span> Per pT results</h3>
<details>
<p><summary>EUR target, GWAS1: MDD_PGC2, GWAS2: EUR_PGC_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EUR/MDD_PGC2_EUR_PGC_EAS/prs_1.test.pT_assoc.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: MDD_PGC2, GWAS2: EUR_PGC_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/MDD_PGC2_EUR_PGC_EAS/prs_1.test.pT_assoc.png" /></p>
</details>
<details>
<p><summary>EUR target, GWAS1: MDD_PGC2, GWAS2: EUR_CONVERGE_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EUR/MDD_PGC2_EUR_CONVERGE_EAS/prs_1.test.pT_assoc.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: MDD_PGC2, GWAS2: EUR_CONVERGE_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/MDD_PGC2_EUR_CONVERGE_EAS/prs_1.test.pT_assoc.png" /></p>
</details>
<details>
<p><summary>EUR target, GWAS1: MDD_PGC2, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EUR/MDD_PGC2_EUR_BBJ_EAS/prs_1.test.pT_assoc.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: MDD_PGC2, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/MDD_PGC2_EUR_BBJ_EAS/prs_1.test.pT_assoc.png" /></p>
</details>
<details>
<p><summary>EUR target, GWAS1: Height_Wood, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EUR/Height_Wood_EUR_BBJ_EAS/prs_1.test.pT_assoc.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: Height_Wood, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/Height_Wood_EUR_BBJ_EAS/prs_1.test.pT_assoc.png" /></p>
</details>
<details>
<p><summary>EUR target, GWAS1: BMI_Locke, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EUR/BMI_Locke_EUR_BBJ_EAS/prs_1.test.pT_assoc.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: BMI_Locke, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/BMI_Locke_EUR_BBJ_EAS/prs_1.test.pT_assoc.png" /></p>
</details>
<hr />
</div>
<div id="model-comparison-results" class="section level3">
<h3><span class="header-section-number">3.3.2</span> Model comparison results</h3>
<details>
<p><summary>EUR target, GWAS1: MDD_PGC2, GWAS2: EUR_PGC_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EUR/MDD_PGC2_EUR_PGC_EAS/prs_all.test.comp.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: MDD_PGC2, GWAS2: EUR_PGC_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/MDD_PGC2_EUR_PGC_EAS/prs_all.test.comp.png" /></p>
</details>
<details>
<p><summary>EUR target, GWAS1: MDD_PGC2, GWAS2: EUR_CONVERGE_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EUR/MDD_PGC2_EUR_CONVERGE_EAS/prs_all.test.comp.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: MDD_PGC2, GWAS2: EUR_CONVERGE_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/MDD_PGC2_EUR_CONVERGE_EAS/prs_all.test.comp.png" /></p>
</details>
<details>
<p><summary>EUR target, GWAS1: MDD_PGC2, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EUR/MDD_PGC2_EUR_BBJ_EAS/prs_all.test.comp.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: MDD_PGC2, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/MDD_PGC2_EUR_BBJ_EAS/prs_all.test.comp.png" /></p>
</details>
<details>
<p><summary>EUR target, GWAS1: Height_Wood, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EUR/Height_Wood_EUR_BBJ_EAS/prs_all.test.comp.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: Height_Wood, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/Height_Wood_EUR_BBJ_EAS/prs_all.test.comp.png" /></p>
</details>
<details>
<p><summary>EUR target, GWAS1: BMI_Locke, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EUR/BMI_Locke_EUR_BBJ_EAS/prs_all.test.comp.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: BMI_Locke, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/BMI_Locke_EUR_BBJ_EAS/prs_all.test.comp.png" /></p>
</details>
<hr />
</div>
</div>
<div id="prs-csx-2" class="section level2">
<h2><span class="header-section-number">3.4</span> PRS-CSx</h2>
<hr />
<div id="per-phi-results" class="section level3">
<h3><span class="header-section-number">3.4.1</span> Per phi results</h3>
<details>
<p><summary>EUR target, GWAS1: MDD_PGC2, GWAS2: EUR_PGC_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EUR/MDD_PGC2_EUR_PGC_EAS/PRS-CSx_1.test.phi_assoc.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: MDD_PGC2, GWAS2: EUR_PGC_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/MDD_PGC2_EUR_PGC_EAS/PRS-CSx_1.test.phi_assoc.png" /></p>
</details>
<details>
<p><summary>EUR target, GWAS1: MDD_PGC2, GWAS2: EUR_CONVERGE_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EUR/MDD_PGC2_EUR_CONVERGE_EAS/PRS-CSx_1.test.phi_assoc.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: MDD_PGC2, GWAS2: EUR_CONVERGE_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/MDD_PGC2_EUR_CONVERGE_EAS/PRS-CSx_1.test.phi_assoc.png" /></p>
</details>
<details>
<p><summary>EUR target, GWAS1: MDD_PGC2, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EUR/MDD_PGC2_EUR_BBJ_EAS/PRS-CSx_1.test.phi_assoc.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: MDD_PGC2, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/MDD_PGC2_EUR_BBJ_EAS/PRS-CSx_1.test.phi_assoc.png" /></p>
</details>
<details>
<p><summary>EUR target, GWAS1: Height_Wood, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EUR/Height_Wood_EUR_BBJ_EAS/PRS-CSx_1.test.phi_assoc.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: Height_Wood, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/Height_Wood_EUR_BBJ_EAS/PRS-CSx_1.test.phi_assoc.png" /></p>
</details>
<details>
<p><summary>EUR target, GWAS1: BMI_Locke, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EUR/BMI_Locke_EUR_BBJ_EAS/PRS-CSx_1.test.phi_assoc.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: BMI_Locke, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/BMI_Locke_EUR_BBJ_EAS/PRS-CSx_1.test.phi_assoc.png" /></p>
</details>
<hr />
</div>
<div id="model-comparison-results-1" class="section level3">
<h3><span class="header-section-number">3.4.2</span> Model comparison results</h3>
<details>
<p><summary>EUR target, GWAS1: MDD_PGC2, GWAS2: EUR_PGC_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EUR/MDD_PGC2_EUR_PGC_EAS/PRS-CSx_2.test.comp.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: MDD_PGC2, GWAS2: EUR_PGC_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/MDD_PGC2_EUR_PGC_EAS/PRS-CSx_2.test.comp.png" /></p>
</details>
<details>
<p><summary>EUR target, GWAS1: MDD_PGC2, GWAS2: EUR_CONVERGE_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EUR/MDD_PGC2_EUR_CONVERGE_EAS/PRS-CSx_2.test.comp.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: MDD_PGC2, GWAS2: EUR_CONVERGE_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/MDD_PGC2_EUR_CONVERGE_EAS/PRS-CSx_2.test.comp.png" /></p>
</details>
<details>
<p><summary>EUR target, GWAS1: MDD_PGC2, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EUR/MDD_PGC2_EUR_BBJ_EAS/PRS-CSx_2.test.comp.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: MDD_PGC2, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/MDD_PGC2_EUR_BBJ_EAS/PRS-CSx_2.test.comp.png" /></p>
</details>
<details>
<p><summary>EUR target, GWAS1: Height_Wood, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EUR/Height_Wood_EUR_BBJ_EAS/PRS-CSx_2.test.comp.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: Height_Wood, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/Height_Wood_EUR_BBJ_EAS/PRS-CSx_2.test.comp.png" /></p>
</details>
<details>
<p><summary>EUR target, GWAS1: BMI_Locke, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EUR/BMI_Locke_EUR_BBJ_EAS/PRS-CSx_2.test.comp.png" /></p>
</details>
<details>
<p><summary>EAS target, GWAS1: BMI_Locke, GWAS2: EUR_BBJ_EAS</summary></p>
<p><img src="Images/CrossMeta/assoc/EAS/BMI_Locke_EUR_BBJ_EAS/PRS-CSx_2.test.comp.png" /></p>
</details>
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
