<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>GenoPredPipe</title>

<script src="site_libs/header-attrs-2.23/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.0/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.0/css/v4-shims.min.css" rel="stylesheet" />
<!-- header.html -->
<style>
  .navbar {
    min-height: 80px; /* Adjust as needed */
  }

  .navbar-brand {
    padding: 10px 15px; /* Adjust as needed */
  }
</style>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img id="logo" style="height: 60px;" src="Horizontal.png" /></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="research_index.html">Research</a>
</li>
<li>
  <a href="tools_index.html">Tools</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">GenoPredPipe</h1>

</div>


<hr />
<p>This is a snakemake pipeline for running the GenoPred scripts. The
pipeline currently identifies the ancestry of each individual,
calculates ancestry-matched reference-projected principal components,
and calculates polygenic scores using a range of methods standardised
using an ancestry matched reference. Finally, the pipeline provides a
report of the results either for each individual or the sample
overall.</p>
<p>The pipeline’s ability to calculate polygenic scores has been
validated <a
href="https://opain.github.io/GenoPred/Determine_optimal_polygenic_scoring_approach_GenoPredPipe.html">here</a>
in UK Biobank.</p>
<hr />
<div id="getting-started" class="section level1">
<h1>Getting started</h1>
<div id="step-1" class="section level2">
<h2>Step 1</h2>
<p>Clone the repository</p>
<pre class="bash"><code>git clone https://github.com/opain/GenoPred.git
cd GenoPred/GenoPredPipe</code></pre>
<hr />
</div>
<div id="step-2" class="section level2">
<h2>Step 2</h2>
<p>Install <a
href="https://conda.io/en/latest/miniconda.html">Anaconda</a>.</p>
<p><strong>Linux:</strong></p>
<pre class="bash"><code>wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
sh Miniconda3-latest-Linux-x86_64.sh</code></pre>
<p>Install Python 3.8, Snakemake 5.32, and the basic project
dependencies. Note. I am installing these packages in an environment
called ‘base’, if you already have an environment called ‘base’, you may
need to create a new environment to avoid conflicts.</p>
<pre class="bash"><code>conda activate base
conda install python=3.8
conda install -c conda-forge mamba
mamba install -c bioconda -c conda-forge snakemake-minimal==5.32.2
mamba install dropbox
mamba install pandas</code></pre>
<p>Note. If ghostscript is not already installed on your system, you
will need to install it. You can check this by typing ‘ghostscript’ into
the terminal.</p>
<hr />
</div>
<div id="step-3" class="section level2">
<h2>Step 3</h2>
<p>Prepare a snakemake profile for parallel computing. I have provided
an example for users using a slurm scheduler called
‘example_slurm_profile_config.yaml’. Slurm users should create a folder
called ‘slurm’ in ‘$HOME/.config/snakemake’, and then copy in the
example_slurm_profile_config.yaml, renaming it to config.yaml. More
information about profiles in snakemake can be found <a
href="https://snakemake.readthedocs.io/en/stable/executing/cli.html#profiles">here</a>.</p>
<hr />
</div>
</div>
<div id="running-pipeline" class="section level1">
<h1>Running pipeline</h1>
<div id="example-using-test-data" class="section level2">
<h2>Example using test data</h2>
<div id="step-1-1" class="section level3">
<h3>Step 1</h3>
<p>Download test data.</p>
<pre class="bash"><code>wget -O test_data.tar.gz https://zenodo.org/record/6093584/files/test_data.tar.gz?download=1
tar -xf test_data.tar.gz
rm test_data.tar.gz</code></pre>
<hr />
</div>
<div id="step-2-1" class="section level3">
<h3>Step 2</h3>
<p>Run full pipeline for all target samples.</p>
<pre class="bash"><code>snakemake --profile slurm --use-conda run_create_reports</code></pre>
<p>Note. If you receive an error saying ‘MissingOutputException’, you
should try adding ‘–latency-wait 20’ to the snakemake command, which
tells the pipeline to wait 20 seconds between steps, thereby allowing
filesystem latency.</p>
<p>Note. Please be patient when running the pipeline for the first time.
Expect the ‘downloading and installing remote packages’ to take ~1 hour.
It has to create the conda environment in first instance, which involves
installing python and R and many packages. Expect this to take ~1
hour.</p>
<hr />
</div>
</div>
</div>
<div id="using-your-own-data" class="section level1">
<h1>Using your own data</h1>
<p>You must specify a file listing target samples (target_list) and a
file listing of GWAS summary statistics (gwas_list) for the pipeline to
use. The location of these files should be specified in the config.yaml
file. The column names for the gwas_list and target_list files should be
as follows:</p>
<hr />
<div id="target_list-required" class="section level2">
<h2>target_list (required)</h2>
<ul>
<li>name: Name of the target sample</li>
<li>path: File path to the target genotype data. For type ‘23andMe’,
provide full file name either zipped (.zip) or uncompressed (.txt). For
types ‘samp_imp_plink1’, ‘samp_imp_bgen’, and ‘samp_imp_vcf’,
per-chromosome genotype data should be provided with the following
filename format:
&lt;prefix&gt;.chr&lt;1-22&gt;.&lt;.bed/.bim/.fam/.bgen/.vcf.gz&gt;. If
type ‘samp_imp_bgen’, the sample file should be called
&lt;prefix&gt;.sample, and each .bgen file should have a corresponding
.bgi file.</li>
<li>type: Either ‘23andMe’, ‘samp_imp_plink1’, ‘samp_imp_bgen’, or
‘samp_imp_vcf’. ‘23andMe’ = 23andMe formatted data for an individual.
‘samp_imp_plink1’ = Preimputed PLINK1 binary format data
(.bed/.bim/.fam) for a group of individuals. ‘samp_imp_bgen’ =
Preimputed Oxford format data (.bgen/.sample) for a group of
individuals. ‘samp_imp_vcf’ = Preimputed gzipped VCF format data
(.vcf.gz) for a group of individuals.</li>
<li>output: The desired output directory for target sample. The <name>
of the target sample will automically be appended to the end. i.e. when
output=‘test’ and name=‘Joe_Bloggs’, the output will written in a folder
called ‘test/Joe_Bloggs’.</li>
<li>indiv_report: Either ‘T’ or ‘F’. ‘T’ = Individual-specific ancestry
and polygenic score reports will be created. Use with caution if target
sample contains many individuals, as it will create an .html report for
each individual.</li>
</ul>
<p>Note. This file should be space delimited.</p>
<hr />
</div>
<div id="gwas_list-required" class="section level2">
<h2>gwas_list (required)</h2>
<ul>
<li>name: Short name for the GWAS</li>
<li>path: File path to the GWAS summary statistics (uncompressed or
gzipped)</li>
<li>population: The super population that the GWAS was performed in
(AFR/AMR/EAS/EUR/SAS)</li>
<li>sampling: The proportion of the GWAS sample that were cases (if
binary, otherwise NA)</li>
<li>prevalence: The population prevelance of the phenotype (if binary,
otherwise NA)</li>
<li>mean: The phenotype mean in the general population (if continuous,
otherwise NA)</li>
<li>sd: The phenotype sd in the general population (if continuous,
otherwise NA)</li>
<li>label: A human readable name for the GWAS phenotype. Wrap in quotes
if multiple words. For example, “Body Mass Index”.</li>
</ul>
<p>Note. This file should be space delimited.</p>
<hr />
<div id="gwas-summary-statistic-format" class="section level3">
<h3>GWAS summary statistic format</h3>
<p>The following column names are expected in the GWAS summary
statistics files:</p>
<ul>
<li>SNP: RSID for variant (required)</li>
<li>A1: Allele 1 (effect allele) (required)</li>
<li>A2: Allele 2 (required)</li>
<li>P: P-value of association (required)</li>
<li>OR: Odds ratio effect size (required if binary)</li>
<li>BETA: BETA effect size (required if continuous)</li>
<li>SE: Standard error of log(OR) or BETA (required)</li>
<li>N: Total sample size (required)</li>
<li>FREQ: Allele frequency in GWAS sample (optional)</li>
<li>INFO: Imputation quality (optional)</li>
</ul>
<hr />
</div>
</div>
<div id="score_list-required" class="section level2">
<h2>score_list (required)</h2>
<p>It is also possible to provide a list of external created score files
(score_list) to be used for polygenic scoring. The score_list should
have the same format as gwas_list, except the path column should
indicate the location of the score file. Caution: GenoPredPipe assumes
the externally provided score files are restricted and harmonised to the
HapMap3 SNP-list.</p>
<p>If no externally created score files are present, please still
provide a file containing the correct header.</p>
<p>Note. This file should be space delimited.</p>
<hr />
<div id="score-file-format" class="section level3">
<h3>Score file format</h3>
<p>The following column names are expected in the score files:</p>
<ul>
<li>SNP: RSID for variant (required)</li>
<li>A1: Allele 1 (effect allele) (required)</li>
</ul>
<p>Then one or more columns indicating the effect size of each variant,
for example:</p>
<pre><code>SNP         A1  SCORE_phi_1e-06   SCORE_phi_1e-04
rs3131972   A   4.502864e-07      -0.0002017462
rs3131969   A   1.48949e-05       8.115991e-06 
rs1048488   C   -1.371439e-05     4.194312e-05 </code></pre>
<hr />
</div>
</div>
</div>
<div id="running-parts-of-the-pipeline" class="section level1">
<h1>Running parts of the pipeline</h1>
<p>It is possible to only run specific parts of the pipeline for all
target samples. For example:</p>
<pre class="bash"><code># Harmonise with the reference only
snakemake --profile slurm --use-conda run_harmonise_target_with_ref

# Create ancestry report only
snakemake --profile slurm --use-conda run_create_ancestry_reports

# Calculate reference-projected principal components only
snakemake --profile slurm --use-conda run_target_pc_all

# Calculate pT+clump polygenic scores only
snakemake --profile slurm --use-conda run_target_prs_pt_clump_all_name

# Calculate DBSLMM polygenic scores only
snakemake --profile slurm --use-conda run_target_prs_dbslmm_all_name
</code></pre>
<p>Furthermore, you can request specific outputs for specific target
samples. For example:</p>
<pre class="bash"><code># Run full pipeline for specific target sample. 
# Replace &lt;name&gt; with name of target sample in the target_list file.
snakemake --profile slurm --use-conda resources/data/target_checks/&lt;name&gt;/create_sample_report.done

# Calculate pT+clump polygenic scores for specific population, in a specific target sample, and specific GWAS. 
# Replace &lt;name&gt; with name of a target sample in the target_list file. 
# Replace &lt;gwas&gt; with the name of a gwas in the gwas_list file.
# Replace &lt;population&gt; with the super population ancestry code (e.g. EUR, EAS, AMR, AFR, SAS).
snakemake --profile slurm --use-conda resources/data/target_checks/&lt;name&gt;/target_prs_pt_clump_&lt;population&gt;_&lt;gwas&gt;.done


# To generate polygenic scores using other methods, replace pt_clump in the previous command to either dbslmm, prscs, lassosum, sbayesr, ldpred2, megaprs, or external. For example, to calculate megaprs polygenic scores based on the BODY04 GWAS in the European (EUR) subset of the target sample called example_plink:
snakemake --profile slurm --use-conda resources/data/target_checks/example_plink1/target_prs_megaprs_EUR_BODY04.done

# Look inside the rules/GenoPredPipe.smk file to see all available rules.
</code></pre>
<hr />
</div>
<div id="output" class="section level1">
<h1>Output</h1>
<p>Potentially useful GenoPredPipe outputs can be found in the following
locations:</p>
<ul>
<li>1KG Phase 3 reference data restricted to HapMap3 variants in PLINK1
binary format (.bed/.bim/.fam)
<ul>
<li>Per chromosome:
resources/data/1kg/1KGPhase3.w_hm3.chr&lt;chr&gt;.&lt;bed/bim/fam&gt;</li>
<li>Genome-wide:
resources/data/1kg/1KGPhase3.w_hm3.GW.&lt;.bed/.bim/.fam&gt;</li>
<li>Super population and population keep files:
resources/data/1kg/keep_files/&lt;pop&gt;_sample.keep</li>
<li>Allele frequency files split by population:
resources/data/1kg/freq_files/&lt;pop&gt;/1KGPhase3.w_hm3.&lt;pop&gt;.chr&lt;chr&gt;.frq</li>
</ul></li>
</ul>
<p><br/></p>
<ul>
<li><p>Quality controlled and formatted GWAS summary statistics:
resources/data/gwas_sumstat/&lt;gwas&gt;/&lt;gwas&gt;.cleaned.gz</p></li>
<li><p>Lassosum pseudovalidation results:
resources/data/1kg/prs_pseudoval/&lt;gwas&gt;</p></li>
<li><p>Projected principal component .score files:
resources/data/1kg/prs_score_files/&lt;method&gt;/&lt;gwas&gt;</p></li>
<li><p>Polygenic score .score files:
resources/data/1kg/prs_score_files/&lt;method&gt;/&lt;gwas&gt;</p></li>
<li><p>Target sample outputs (&lt;output&gt; corresponds to the output
specified in the target_list file)</p>
<ul>
<li>Imputed genotype data (if 23andMe input):
&lt;output&gt;/&lt;name&gt;/&lt;name&gt;.&lt;gen/sample&gt;</li>
<li>Genotype data restricted to HapMap3 SNPs and harmonised with
reference:
&lt;output&gt;/&lt;name&gt;/&lt;name&gt;.1KGPhase3.w_hm3.chr&lt;chr&gt;.&lt;bed/bim/fam&gt;</li>
<li>Super population ancestry results:
&lt;output&gt;/&lt;name&gt;/ancestry</li>
<li>Within super population ancestry results:
&lt;output&gt;/&lt;name&gt;/ancestry/ancestry_&lt;population&gt;</li>
<li>Projected principal components:
&lt;output&gt;/&lt;name&gt;/projected_pcs/&lt;population&gt;</li>
<li>Polygenic scores:
&lt;output&gt;/&lt;name&gt;/prs/&lt;population&gt;/&lt;method&gt;/&lt;gwas&gt;</li>
<li>Individual-level report: &lt;output&gt;/&lt;name&gt;/reports</li>
<li>Sample-level report: &lt;output&gt;/&lt;name&gt;/reports</li>
</ul></li>
</ul>
<hr />
</div>
<div id="troubleshooting" class="section level1">
<h1>Troubleshooting</h1>
<p>If using the –profile flag, the log files will be saved in the
GenoPredPipe/logs folder. If running interactively (i.e. -j1), the error
should be printed on the screen.</p>
<hr />
<p>Please contact Oliver Pain (<a href="mailto:oliver.pain@kcl.ac.uk"
class="email">oliver.pain@kcl.ac.uk</a>) for any questions or
comments.</p>
</div>

<!-- footer.html -->
<hr/>
<div style="display: flex; justify-content: center; align-items: center;">
  <img src="kcl.svg" alt="Kings College London" style="height: 60px; padding-left: 5px; padding-right: 5px;">
  <img src="Wellcome_logo.png" alt="The Wellcome Trust" style="height: 60px; padding-left: 5px; padding-right: 5px;">
  <img src="nihr-logo.png" alt="National Institute for Health and Care Research" style="height: 60px; padding-left: 5px; padding-right: 5px;">
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
