<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Genotype-based Prediction (GenoPred)</title>

<script src="site_libs/header-attrs-2.23/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.0/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.0/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GenoPred</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Genotype-based Prediction (GenoPred)</h1>

</div>


<hr />
<p>Welcome!</p>
<p>This website provides details of a project evaluating methods for
genotype-based prediction. The website will be updated continuously as
the project develops.</p>
<p>Please see our <a
href="https://github.com/opain/GenoPred/tree/master/GenoPredPipe">GenoPred
pipeline (GenoPredPipe)</a>, providing a one-liner for ancestry
inference and polygenic score calculation.</p>
<hr />
<div id="contents" class="section level4">
<h4>Contents:</h4>
<ul>
<li>Data Preparation
<ul>
<li>Genetic
<ul>
<li><a href="Pipeline_prep.html">Reference genetic: European 503 subset
of 1KG</a></li>
<li><a href="Pipeline_prep_withUKBB_ref.html">Reference genetic:
European 10K subset of UK Biobank</a></li>
<li><a href="Genotype-based_scoring_in_target_samples.html">Target
genetic</a></li>
</ul></li>
<li>Phenotypic
<ul>
<li><a href="Target_sample_phenotype_prep.html">Target
phenotypic</a></li>
</ul></li>
</ul></li>
<li>Main Research Questions
<ul>
<li><a
href="Determine_optimal_polygenic_scoring_approach.html">Comparison of
polygenic scoring methods</a>
<ul>
<li><a
href="Determine_optimal_polygenic_scoring_approach_update_21102021.html">See
here for updated results including LDAK’s MegaPRS method</a></li>
</ul></li>
<li><a href="Functionally_informed_prediction.html">Evaluating gene
expression risk scores</a></li>
<li><a href="Absolute_Conversion.html">Converting polygenic scores to
the absolute scale</a>
<ul>
<li><a href="PRS_to_Abs_tool.html">Link to Shiny app</a></li>
</ul></li>
<li><a href="local_rG_weighted_pgs.html">Local genetic correlation
weighted polygenic scores</a></li>
<li><a href="DiverseAncestry.html">Improve polygenic prediction in
diverse populations</a> (under development)</li>
</ul></li>
<li>Side Questions
<ul>
<li><a href="standardised_vs_nonstandarised.html">Standardised
vs. non-standardised</a></li>
<li><a href="Z-test_vs_ANOVA_V2.html">Estimating significance of model
comparisons</a></li>
</ul></li>
<li>Future developments
<ul>
<li>Stratification using pathway/tissue-specific genetic scores</li>
<li>Gene expression risk scores using observed expression</li>
</ul></li>
</ul>
<hr />
</div>
<div id="help" class="section level4">
<h4>Help</h4>
<p>This website was created by Oliver Pain (<a
href="mailto:oliver.pain@kcl.ac.uk"
class="email">oliver.pain@kcl.ac.uk</a>). Please get in touch if you
have any questions or comments. Also, follow me on <a
href="https://twitter.com/ollie_pain">twitter</a> and <a
href="https://github.com/opain">GitHub</a> for updates.</p>
<hr />
</div>
<div id="acknowledgments" class="section level4">
<h4>Acknowledgments</h4>
<p>I would like to thank the following people who have contributed to
this project so far:</p>
<ul>
<li><strong>Cathryn Lewis</strong></li>
<li>Lasse Folkersen</li>
<li>Kylie Glanville</li>
<li>Saskia Hagenaars</li>
<li>Saskia Selzam</li>
<li>Anna Fürtjes</li>
<li>Helena Gaspar</li>
<li>Jonathon Coleman</li>
<li>Gerome Breen</li>
<li>Robert Plomin</li>
<li>Alex Gillett</li>
<li>Ken Hanscombe</li>
<li>Ryan Arathimos</li>
<li>Toshiki Kono</li>
</ul>
<hr />
</div>
<div id="funding" class="section level4">
<h4>Funding:</h4>
<p>This work is funded by the MRC (MR/N015746/1), and the National
Institute for Health Research (NIHR) Biomedical Research Centre at South
London and Maudsley NHS Foundation Trust and King’s College London. The
authors acknowledge use of the research computing facility at King’s
College London, Rosalind (<a href="https://rosalind.kcl.ac.uk"
class="uri">https://rosalind.kcl.ac.uk</a>), which is delivered in
partnership with the NIHR Maudsley BRC, and part-funded by capital
equipment grants from the Maudsley Charity (award 980) and Guy’s &amp;
St. Thomas’ Charity (TR130505). The views expressed are those of the
authors and not necessarily those of the NHS, the NIHR or the Department
of Health and Social Care.</p>
<hr />
</div>
<div id="change-log" class="section level4">
<h4>Change log:</h4>
<p><strong>24/06/2020</strong>:</p>
<ul>
<li>Change log started</li>
<li>Updated Ancestry_identifier.R to use proper grid search in enet,
allow other modelling methods to be specified, and improve model
performance reporting.</li>
</ul>
<p><strong>13/08/2020</strong>:</p>
<ul>
<li>Add comparison between standardised and non-standardised polygenic
scores.</li>
</ul>
<p><strong>10/02/2021</strong>:</p>
<ul>
<li>PRS comparison: Added LDPred2 and DBSLMM methods. Added prostate and
breast cancer. Add polygenicity and ancestry senstivity analysis.</li>
<li>GeRS comparison: Minor formatting changes.</li>
</ul>
<p><strong>13/04/2021</strong></p>
<ul>
<li>PRS comparison: Evaluated new version of GCTB v2.03</li>
<li>PRS conversion to absolute scale: Updated html</li>
</ul>
<p><strong>22/08/2021</strong></p>
<ul>
<li>Added GenoPredPipe documentation</li>
</ul>
<p><strong>22/08/2021</strong></p>
<ul>
<li>Added LDAK’s MegaPRS to polygenic score methods comparison</li>
<li>Benchmarked the runtime of LDpred2 with new SFMB format</li>
</ul>
<p><strong>14/03/2022</strong></p>
<ul>
<li>Added local genetic correlation weighted polygenic score
analysis</li>
</ul>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
