---
title: Genetic analysis of Mirocals trial
output: 
  html_document:
    theme: cosmo
    toc: true
    toc_float: true
    toc_depth: 2
    css: styles/styles.css
    includes:
      in_header: header.html
      after_body: footer.html

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(eval = FALSE)
```

```{css, echo=F}
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
```

***

# Proposal

The Mirocals trial tested the efficacy of low dose IL-2 for treatment of ALS. The trial showed IL-2 + riluzole treatment provided a significant improvement in survival compared to those only receiving placebo + riluzole. Low dose IL-2 has been shown to increase the number of regulatory T-cells (Tregs).

This trial has collected whole genome sequence data and methylation data (across several time points). This presents an opportunity to identify genetic predictors/biomarkers of response to IL-2 treatment. The sample size of the trial is limited (N drug = 110, N placebo = 110), which means we need to test specific hypotheses and leverage larger external datasets.

***

## Outcome

There are two key outcome variables from the trial, including survival and regulatory T-cell numbers. It seems survival is the more meaningful outcome for patients, but perhaps there are benefits to using T-cell numbers that I am unaware of. If we are using survival, then we should using a cox regression, but if use T cell numbers linear regression.

***

## Polygenic scores (PGS)

List of polygenic scores that might predict response to IL-2:

* ALS risk, survival, age at onset
* Risk of other neurodegenerative diseases
* Interleukin-2 receptor subunit alpha (IL2RA) protein levels
* Regulatory T-cell molecule levels
* Autoimmune disease risk
* Other inflammatory markers (CRP, IL-6, etc.)

I would suggest using SBayesRC as the method to generate the PGS. If GWAS from multiple populations are available we can using SBayesRC-multi.

We could also further stratify the polygenic scores by functional annotations, such as pathway membership. This could be particularly useful, as it may be that only certain pathways are relevant to IL-2 efficacy. For example, ALS risk mediated by inflammatory pathways may be particularly relevant for IL-2 efficacy. 

***

## Methylation scores

We can use previously published EWAS summary statistics to do the same thing. Alternatively, we can use 

***

# Calculate PGS

```{bash}
mkdir -p /users/k1806347/oliverpainfel/Data/miraclestrial/original

# Create index for each batch
module add bcftools/1.12-gcc-13.2.0-python-3.11.6

####
# Format for GenoPred
####
# Split into per chromosome files

##
# Batch 1
## 
for chr in $(seq 1 22);do
    sbatch --mem 10G -n 1 -p interruptible_cpu -t 1:00:00 --wrap="/users/k1806347/oliverpainfel/Software/plink2 \
        --vcf /scratch/prj/mirocalstrial/recovered/DNAscan/variantcalling/merged_strelka_stringent.vcf.gz \
        --chr ${chr} \
        --out /users/k1806347/oliverpainfel/Data/miraclestrial/processed/miracles_batch1.chr${chr} \
        --allow-extra-chr \
        --export vcf bgz"
done

# The file is truncated mid chr16, and missing chr 17-22

# Recreate merged dataset 
module add bcftools/1.12-gcc-13.2.0-python-3.11.6

bcftools merge \
  -l sample_vcf_list.txt \
  --missing-to-ref \
  -o /users/k1806347/oliverpainfel/Data/miraclestrial/original/miracles_batch1.vcf.gz

# Replace corrupt version
cp /users/k1806347/oliverpainfel/Data/miraclestrial/original/miracles_batch1.vcf.gz /scratch/prj/mirocalstrial/recovered/DNAscan/variantcalling/merged_strelka_stringent_fixed.vcf.gz

# Split by chromosome
for chr in $(seq 1 22);do
    sbatch --mem 10G -n 1 -p interruptible_cpu -t 1:00:00 --wrap="/users/k1806347/oliverpainfel/Software/plink2 \
        --vcf /users/k1806347/oliverpainfel/Data/miraclestrial/original/miracles_batch1.vcf.gz \
        --chr ${chr} \
        --out /users/k1806347/oliverpainfel/Data/miraclestrial/processed/miracles_batch1.chr${chr} \
        --allow-extra-chr \
        --export vcf bgz"
done

##
# Batch 2
## 
for chr in $(seq 1 22);do
    sbatch --mem 10G -n 1 -p interruptible_cpu -t 1:00:00 --wrap="/users/k1806347/oliverpainfel/Software/plink2 \
        --vcf /scratch/prj/mirocalstrial/recovered/DNAscan_batch2/variantcalling/merged_strelka_stringent.vcf.gz \
        --chr ${chr} \
        --out /users/k1806347/oliverpainfel/Data/miraclestrial/processed/miracles_batch2.chr${chr} \
        --allow-extra-chr \
        --export vcf bgz"
done

# Batch 2 seems fine.

```

```{r}
library(data.table)

dir.create('/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/output', recursive = T)
dir.create('/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config', recursive = T)

# Create target_list
target_list <- NULL
target_list <- rbind(target_list, data.table(
  name = 'miracles_batch1',
  path = '/users/k1806347/oliverpainfel/Data/miraclestrial/processed/miracles_batch1',
  type = 'vcf',
  indiv_report = F))

target_list <- rbind(target_list, data.table(
  name = 'miracles_batch2',
  path = '/users/k1806347/oliverpainfel/Data/miraclestrial/processed/miracles_batch2',
  type = 'vcf',
  indiv_report = F))

write.table(
  target_list,
  '/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/target_list.txt',
  col.names = T,
  row.names = F,
  quote = F,
  sep = ' '
)

# Create gwas_list
gwas_list <- data.table(
  name='yengo_eur',
  path = '/users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_eur.txt',
  population = 'EUR',
  n = NA,
  sampling = NA,
  prevalence = NA,
  mean = NA,
  sd = NA,
  label = "\"Yengo 2022 Height EUR\"")

write.table(
  gwas_list,
  '/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/gwas_list.txt',
  col.names = T,
  row.names = F,
  quote = F,
  sep = ' '
)

# Create config file
conf <- c(
  'outdir: /users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/output',
  'config_file: /users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/config.yaml',
  'target_list: /users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/target_list.txt',
  'gwas_list: /users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/gwas_list.txt',
  "pgs_methods: ['dbslmm']"
)

write.table(
  conf,
  '/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/config.yaml',
  col.names = F,
  row.names = F,
  quote = F
)

```

```{bash}
cd /users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline
conda activate genopred

snakemake \
  --profile slurm \
  --use-conda \
  --configfile=/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/config.yaml \
  output_all -n 
```

