---
title: Genetic analysis of Mirocals trial
output: 
  html_document:
    theme: cosmo
    toc: true
    toc_float: true
    toc_depth: 2
    css: styles/styles.css
    includes:
      in_header: header.html
      after_body: footer.html

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(eval = FALSE)
```

```{css, echo=F}
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
```

***

# Proposal

The Mirocals trial tested the efficacy of low dose IL-2 for treatment of ALS. The trial showed IL-2 + riluzole treatment provided a significant improvement in survival compared to those only receiving placebo + riluzole. Low dose IL-2 has been shown to increase the number of regulatory T-cells (Tregs).

This trial has collected whole genome sequence data and methylation data (across several time points). This presents an opportunity to identify genetic predictors/biomarkers of response to IL-2 treatment. The sample size of the trial is limited (N drug = 110, N placebo = 110), which means we need to test specific hypotheses and leverage larger external datasets.

***

## Outcome

There are two key outcome variables from the trial, including survival and regulatory T-cell numbers. It seems survival is the more meaningful outcome for patients, but perhaps there are benefits to using T-cell numbers that I am unaware of. If we are using survival, then we should using a cox regression, but if use T cell numbers linear regression.

***

## Polygenic scores (PGS)

List of polygenic scores that might predict response to IL-2:

* ALS risk, survival, age at onset
* Risk of other neurodegenerative diseases
* Interleukin-2 receptor subunit alpha (IL2RA) protein levels
* Regulatory T-cell molecule levels
* Autoimmune disease risk
* Other inflammatory markers (CRP, IL-6, etc.)

I would suggest using SBayesRC as the method to generate the PGS. If GWAS from multiple populations are available we can using SBayesRC-multi.

We could also further stratify the polygenic scores by functional annotations, such as pathway membership. This could be particularly useful, as it may be that only certain pathways are relevant to IL-2 efficacy. For example, ALS risk mediated by inflammatory pathways may be particularly relevant for IL-2 efficacy. 

***

## Methylation scores

We can use previously published EWAS summary statistics to do the same thing.

***

# Calculate PGS

***

## Prepare target data

```{bash}
mkdir -p /users/k1806347/oliverpainfel/Data/miraclestrial/original

# Create index for each batch
module add bcftools/1.12-gcc-13.2.0-python-3.11.6

####
# Format for GenoPred
####
# Split into per chromosome files

##
# Batch 1
## 
for chr in $(seq 1 22);do
    sbatch --mem 10G -n 1 -p interruptible_cpu -t 1:00:00 --wrap="/users/k1806347/oliverpainfel/Software/plink2 \
        --vcf /scratch/prj/mirocalstrial/recovered/DNAscan/variantcalling/merged_strelka_stringent.vcf.gz \
        --chr ${chr} \
        --out /users/k1806347/oliverpainfel/Data/miraclestrial/processed/miracles_batch1.chr${chr} \
        --allow-extra-chr \
        --export vcf bgz"
done

# The file is truncated mid chr16, and missing chr 17-22

# Recreate merged dataset 
module add bcftools/1.12-gcc-13.2.0-python-3.11.6

bcftools merge \
  -l sample_vcf_list.txt \
  --missing-to-ref \
  -o /users/k1806347/oliverpainfel/Data/miraclestrial/original/miracles_batch1.vcf.gz

# Replace corrupt version
cp /users/k1806347/oliverpainfel/Data/miraclestrial/original/miracles_batch1.vcf.gz /scratch/prj/mirocalstrial/recovered/DNAscan/variantcalling/merged_strelka_stringent_fixed.vcf.gz

# Split by chromosome
for chr in $(seq 1 22);do
    sbatch --mem 10G -n 1 -p interruptible_cpu -t 1:00:00 --wrap="/users/k1806347/oliverpainfel/Software/plink2 \
        --vcf /users/k1806347/oliverpainfel/Data/miraclestrial/original/miracles_batch1.vcf.gz \
        --chr ${chr} \
        --out /users/k1806347/oliverpainfel/Data/miraclestrial/processed/miracles_batch1.chr${chr} \
        --allow-extra-chr \
        --export vcf bgz"
done

##
# Batch 2
## 
for chr in $(seq 1 22);do
    sbatch --mem 10G -n 1 -p interruptible_cpu -t 1:00:00 --wrap="/users/k1806347/oliverpainfel/Software/plink2 \
        --vcf /scratch/prj/mirocalstrial/recovered/DNAscan_batch2/variantcalling/merged_strelka_stringent.vcf.gz \
        --chr ${chr} \
        --out /users/k1806347/oliverpainfel/Data/miraclestrial/processed/miracles_batch2.chr${chr} \
        --allow-extra-chr \
        --export vcf bgz"
done

# Batch 2 seems fine.

```

***

## Download GWAS

```{r}
# Read in GWAS catalogue metadata
gwas_cat <- fread('~/oliverpainfel/Data/GWAS_sumstats/gwas_catalog_v1.0.2.1-studies_r2025-08-24.tsv')

# Remove GWAS without full sumstats
gwas_cat<-gwas_cat[gwas_cat$`FULL SUMMARY STATISTICS` == 'yes',]

# Remove columns we don't want
gwas_cat <- gwas_cat[, c(
  "PUBMEDID",
  "FIRST AUTHOR",
  "DATE",
  "DISEASE/TRAIT",
  "INITIAL SAMPLE SIZE",
  "REPLICATION SAMPLE SIZE",
  "ASSOCIATION COUNT",
  "MAPPED_TRAIT",
  "STUDY ACCESSION",
  "SUMMARY STATS LOCATION"
), with = F]

# List all the GWAS catalogue IDs that you want to keep
gwas_ids<-c(
  'GCST90027164',
  'GCST009325',
  'GCST007320',
  'GCST90001390',
  'GCST90444478',
  'GCST004455',
  'GCST90274806',
  'GCST90241611',
  'GCST004455',
  'GCST90088221',
  'GCST90469588',
  'GCST90428414',
  'GCST004454',
  'GCST90428415',
  'GCST90274811',
  'GCST90469586',
  'GCST90241613',
  'GCST90469585',
  'GCST90179335',
  'GCST90277655',
  'GCST90019431',
  'GCST90088237',
  'GCST90241612',
  'GCST90001960',
  'GCST90001506',
  'GCST90001505',
  'GCST90001504',
  'GCST90001511',
  'GCST90001510',
  'GCST90001940',
  'GCST90001935',
  'GCST90475282'
)

gwas_cat_select <- gwas_cat[gwas_cat$`STUDY ACCESSION` %in% gwas_ids,]
gwas_cat_select$url <- NA
  
# Download the summary statistics
dir.create('~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/')

for(i in 1:nrow(gwas_cat_select)){
  print(i)
  if(!file.exists(paste0('~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/',gwas_cat_select$`STUDY ACCESSION`[i], '.gz'))){
    con <- curl::curl(gwas_cat_select$`SUMMARY STATS LOCATION`[i])
    lines <- readLines(con)
    close(con)
    
    # extract all href targets
    files <- regmatches(lines, gregexpr('href="([^"]+)"', lines))
    files <- unlist(lapply(files, function(x) sub('href="([^"]+)"', '\\1', x)))
    files <- files[!files %in% c("../")]

    if('harmonised/' %in% files){
      con <- curl::curl(paste0(gwas_cat_select$`SUMMARY STATS LOCATION`[i],'/harmonised'))
      lines <- readLines(con)
      close(con)
      
      # extract all href targets
      files <- regmatches(lines, gregexpr('href="([^"]+)"', lines))
      files <- unlist(lapply(files, function(x) sub('href="([^"]+)"', '\\1', x)))
      files <- files[!files %in% c("../")]
      
      url <- files[grepl('.h.tsv.gz$', files)]
      url <- paste0(gwas_cat_select$`SUMMARY STATS LOCATION`[i],'/harmonised/', url)
      
      gwas_cat_select$url[i] <- url
      
      log <- system(
        paste0(
          'wget -q --no-check-certificate -O ~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/',gwas_cat_select$`STUDY ACCESSION`[i], '.gz ', url))
    } else {
      url <- files[grepl('.tsv$|.tsv.gz', files)]
      url <- paste0(gwas_cat_select$`SUMMARY STATS LOCATION`[i],'/', url)
      ending <- ifelse(any(grepl('.gz$', url)), '.gz', '')

      gwas_cat_select$url[i] <- url

      log <- system(
        paste0(
          'wget -q --no-check-certificate -O ~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/',gwas_cat_select$`STUDY ACCESSION`[i], ending, ' ', url))
    }
    
    if(log != 0){
      system(paste0('rm ~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/',gwas_cat_select$`STUDY ACCESSION`[i], '.gz'))
    }
  }
}

write.csv(gwas_cat_select, '~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/gwas_info.csv', row.names = F)

# The Alzheimer's GWAS has lost important information during harmonisation (NEFF)
# Download the original sumstats instead of the harmonised
system('rm ~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/GCST007320.gz')
system(paste0("wget -q --no-check-certificate -O ~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/GCST007320.gz http://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST007001-GCST008000/GCST007320/AD_sumstats_Jansenetal_2019sept.txt.gz"))
gwas_cat_select$url[gwas_cat_select$`STUDY ACCESSION` == 'GCST007320'] <- 'http://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST007001-GCST008000/GCST007320/AD_sumstats_Jansenetal_2019sept.txt.gz'

# Set Neff column to N
ss<-fread('~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/GCST007320.gz')
ss$N<-ss$Neff
ss$Neff<-NULL
ss$Nsum<-NULL
fwrite(ss, '~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/GCST007320.gz', quote = F, na ='NA', sep=' ')
```

***

## Prepare config

```{r}
library(data.table)

dir.create('/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/output', recursive = T)
dir.create('/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config', recursive = T)

# Create target_list
target_list <- NULL
target_list <- rbind(target_list, data.table(
  name = 'miracles_batch1',
  path = '/users/k1806347/oliverpainfel/Data/miraclestrial/processed/miracles_batch1',
  type = 'vcf',
  indiv_report = F))

target_list <- rbind(target_list, data.table(
  name = 'miracles_batch2',
  path = '/users/k1806347/oliverpainfel/Data/miraclestrial/processed/miracles_batch2',
  type = 'vcf',
  indiv_report = F))

write.table(
  target_list,
  '/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/target_list.txt',
  col.names = T,
  row.names = F,
  quote = F,
  sep = ' '
)

# Create gwas_list
gwas_cat_select <- fread('~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/gwas_info.csv')

# Remove the Qatari GWAS
gwas_cat_select <- gwas_cat_select[gwas_cat_select$`STUDY ACCESSION` != 'GCST90161640',]

gwas_list <- data.table(
  name=gwas_cat_select$`STUDY ACCESSION`,
  path = paste0('/users/k1806347/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/',gwas_cat_select$`STUDY ACCESSION`,'.gz'),
  population = 'EUR',
  n = NA,
  sampling = NA,
  prevalence = NA,
  mean = NA,
  sd = NA,
  label = paste0("\"", gwas_cat_select$`FIRST AUTHOR`, ' - ', gwas_cat_select$DATE, ' - ', gwas_cat_select$`DISEASE/TRAIT`, "\""))

gwas_list$n <- c(
  NA,
  3475,
  3677,
  NA,
  NA,
  NA,
  NA,
  NA,
  NA,
  NA,
  NA,
  NA,
  5366 ,
  5361,
  6618,
  NA,
  138086,
  NA,
  NA,
  NA,
  NA,
  NA,
  496,
  NA,
  3301,
  3301,
  3301,
  2923,
  39034,
  39236,
  NA
)

gwas_list$sampling [15] <- 2591 / (2591 + 4027)
gwas_list$sampling [17] <- 27205 / (27205 + 110881)

gwas_list$prevalence [15] <- 0.004
gwas_list$prevalence [17] <- 0.00005

write.table(
  gwas_list,
  '/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/gwas_list.txt',
  col.names = T,
  row.names = F,
  quote = F,
  sep = ' '
)

# Create config file
conf <- c(
  'outdir: /users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/output',
  'config_file: /users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/config.yaml',
  'target_list: /users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/target_list.txt',
  'gwas_list: /users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/gwas_list.txt',
  "pgs_methods: ['sbayesrc']"
)

write.table(
  conf,
  '/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/config.yaml',
  col.names = F,
  row.names = F,
  quote = F
)

```

***

## Run pipeline

```{bash}
cd /users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline
conda activate genopred

snakemake \
  --profile slurm \
  --use-conda \
  --configfile=/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/config.yaml \
  ldsc -n 
```

***

# Have a look at the LDSC results

```{r}
setwd('/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline/')

source('../functions/misc.R')
source_all('../functions')

config <- '/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/config.yaml'
gwas_list<-read_param(config = config, param = 'gwas_list')
outdir<-read_param(config = config, param = 'outdir', return_obj = F)

ldsc<-NULL
for(i in gwas_list$name){
  h2_log <- readLines(paste0(outdir,'/reference/ldsc/', i,'/', i, '.log'))
  h2_log <- h2_log[grepl('SNP-heritability estimate on the observed scale', h2_log)]
  h2_log <- gsub('SNP-heritability estimate on the observed scale = ','',h2_log)
  hsq <- as.numeric(gsub(' .*','',h2_log))
  h2_se <- as.numeric(gsub("\\).", '', gsub(".* \\(", '', h2_log)))
  hsq.pv <- pnorm(hsq / h2_se, lower.tail = FALSE)
  ldsc<-rbind(
    ldsc,
    data.table(
      name = i,
      est = hsq,
      se = h2_se,
      z = hsq/h2_se,
      p = hsq.pv
    )
  )
}

gwas_cat_select <- fread('~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/gwas_info.csv')

gwas_cat_select<-merge(gwas_cat_select, ldsc, by.x = 'STUDY ACCESSION', by.y='name')
ldsc <- gwas_cat_select[, c('STUDY ACCESSION','FIRST AUTHOR','DATE','DISEASE/TRAIT','est','se','z','p'), with = F]

write.csv(
  ldsc,
  '/users/k1806347/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/ldsc.csv',
  row.names = F
)

gwas_list<-gwas_list[name %in% ldsc$`STUDY ACCESSION`[ldsc$z > 2],]
gwas_list$label <- paste0("\"", gwas_list$label, "\"")

write.table(
  gwas_list,
  '/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/gwas_list.txt',
  col.names = T,
  row.names = F,
  quote = F,
  sep = ' '
)

```

```{bash}
cd /users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline
conda activate genopred

snakemake \
  --profile slurm \
  --use-conda \
  --configfile=/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/config.yaml \
  output_all outlier_detection -n 
```

***

The vast majority of the clinical trial participants are of European ancestry (Non-EUR N = 15, EUR N = 267). We should restrict the primary analysis to EUR, and then rerun key analyses in full sample. 

***

Analyses:

1. Replicate primary analysis of clinical trial (is this published/can I see details on this)
2. Then test for interaction between PGS and IL-22 on primary endpoint
  - We will have very little power to detect main/interaction effects at statistical significance
  - Could test PGS for risk as primary analysis, as previous studies have done, and then do a hypothesis-free analysis across many PGS, controlling for multiple testing?
    - ALS risk GWAS Could build multi-trait and multi-population model. Build elastic net containing many scores using ProjectMine and independent GWAS
    

What about Jiajing's PGS? seemed to predict risk very highly.

I can generate the PGS and hand them over, or do the analysis? Depends on how Ahmad wants to structure the papers.

***

# Clinical data

I have been sent some preliminary clinical data for the trial. Lets have a look at it:

```{r}
library(data.table)
library(readxl)

data <- read_excel("~/oliverpainfel/Data/miraclestrial/original/Survival-MIR.xlsx")
data <- data.table(data)

# Rename variables
original_names<-names(data)
names(data) <- c('reference','label','id','c9','unc13a','diagnosis','country','site','reference_2','sex','age_at_diagnosis','age_at_onset','site_of_onset','arm','sex_2','familial','alfrs','svc','age_at_death','age_at_death_rounded_down','randomisation','reference_3','gender','age_c','site_of_onset_2','arm_2')

# Overview of data
str(data)
summary(data)

# Check references are identical
all(data$reference == data$reference_2)
all(data$reference == data$reference_3) # We can delete reference_3 as identical to reference
data$reference_3 <- NULL
data[data$reference != data$reference_2,] # There is an reference value matches different row in reference_2.

# Check sex are identical
all(data$sex == data$sex_2)
all(data$sex == data$gender)
all(data$sex_2 == data$gender) # We can delete gender as identical to sex_2
data$gender <- NULL
data[data$sex != data$sex_2,] # The same rows with discordant reference values have discordant sex values.

# Check site of onset
all(data$site_of_onset == data$site_of_onset_2)
data[data$site_of_onset != data$site_of_onset_2,] # One individual has different site of onset (also mismatched for reference)

# Check arm
all(data$arm == data$arm_2, na.rm = T) # They are identical
data$arm_2 <- NULL
data$arm

# Check age variables
# age_c seems to be age_at_diagnosis, just rounded to 1 decimal point, so delete age_c
data$age_c <- NULL

# Create variable indicating death before end of trial
data$age_at_death_rounded_down <- NULL
data$died_during_trial <- ifelse(grepl('Alive at cut off date', data$age_at_death), F, T)
data$died_during_trial[grepl('NA', data$age_at_death)] <- NA
data$age_at_death <- as.numeric(data$age_at_death)

# Check site of onset
all(data$site_of_onset == data$site_of_onset_2)

#####
# Run cox regression
#####
library(survival)

# Subset those in the trial
trial_data <- data[!is.na(data$arm),]
trial_data <- trial_data[trial_data$randomisation == 'RN',]
nrow(trial_data) # 218 - This is two less than the publication

# We need to create a time to event variable
# We can assume age at diagnosis is the date of enrolment
# The trial ran for 640 days, so we can insert time_to_event for those that survived
trial_data$time_to_event <- trial_data$age_at_death - trial_data$age_at_diagnosis
trial_data$time_to_event[trial_data$died_during_trial == F] <- trial_data$age_at_death[trial_data$died_during_trial == F] - trial_data$age_at_diagnosis[trial_data$died_during_trial == F] + (640/365)
trial_data$status <- as.numeric(trial_data$died_during_trial)
trial_data$arm_numeric <- ifelse(trial_data$arm == 'IL2', 1, 0)

cox_model <- coxph(Surv(time_to_event, status) ~ arm_numeric, data = trial_data)
summary(cox_model)

fit <- survfit(Surv(time_to_event, status) ~ arm_numeric, data = trial_data)
plot(fit, col = c("blue", "red"), xlab = "Time (years)", ylab = "Survival probability")
legend("bottomleft", legend = c("Control", "Treatment"), col = c("blue", "red"), lty = 1)

# Shows non-significant increase in survival in those given IL-2
# The real trial used prognostic covariates which we do not have access to currently.

####
# PGS interactions
####

setwd('/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline/')

source('../functions/misc.R')
source_all('../functions')

config <- '/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/config.yaml'

# Read in the PGS and within sample PCs
pgs <- read_pgs(config = config)
gwas_list <- read_param(config = config, param = 'gwas_list')

# Read in PCs
pcs <-list()
pcs[['miracles_batch1']] <- fread('~/oliverpainfel/Data/miraclestrial/GenoPred/output/miracles_batch1/pcs/within_sample/miracles_batch1.outlier_detection.EUR.PCs.txt')
pcs[['miracles_batch2']] <- fread('~/oliverpainfel/Data/miraclestrial/GenoPred/output/miracles_batch2/pcs/within_sample/miracles_batch2.outlier_detection.EUR.PCs.txt')

# Combine the data across batches
pgs_both <- NULL
for(i in 1:2){
  a<-pgs[[paste0('miracles_batch', i)]]$TRANS
  merged_df <- Reduce(function(x, y) merge(x, y, all = TRUE), lapply(a, function(x) x$sbayesrc))
  names(merged_df) <- gsub('_SBayesRC','', names(merged_df))
  pgs_both <- rbind(pgs_both, merged_df)
}

pcs_both <- NULL
for(i in 1:2){
  a <- pcs[[paste0('miracles_batch', i)]]
  a$batch <- paste0('miracles_batch', i)
  pcs_both <- rbind(pcs_both, a)
}

genetics <- merge(pgs_both, pcs_both, by = c('FID','IID'))
trial_data_gen <- merge(trial_data, genetics, by.x = 'id', by.y = 'IID')

cox_model <- coxph(Surv(time_to_event, status) ~ arm_numeric, data = trial_data_gen)
summary(cox_model)

fit <- survfit(Surv(time_to_event, status) ~ arm_numeric, data = trial_data)
plot(fit, col = c("blue", "red"), xlab = "Time (years)", ylab = "Survival probability")
legend("bottomleft", legend = c("Control", "Treatment"), col = c("blue", "red"), lty = 1)

# Test again with PGS but split by batch and with PCs 
int_res_no_pgs <- NULL
for(j in unique(trial_data_gen$batch)){
  cox_model <- coxph(as.formula(paste0("Surv(time_to_event, status) ~ arm_numeric + ", paste0('PC', 1:10, collapse = ' + '))), 
                     data = trial_data_gen[trial_data_gen$batch == j,])
  cox_model_sum <- summary(cox_model)
  int_res_no_pgs<-rbind(
    int_res_no_pgs,
    data.table(
      Batch = j,
      Term = row.names(coef(cox_model_sum)),
      coef = coef(cox_model_sum)[,1],
      se = coef(cox_model_sum)[,3],
      p = coef(cox_model_sum)[,5]
    )
  )
}

# Meta analyse across batches
library(meta)

int_res_no_pgs_meta <- NULL
for(term in unique(int_res_no_pgs$Term)){
  m <- metagen(TE = int_res_no_pgs$coef[int_res_no_pgs$Term == term], seTE = int_res_no_pgs$se[int_res_no_pgs$Term == term], sm = "HR",)
  
  int_res_no_pgs_meta<-rbind(
    int_res_no_pgs_meta,
    data.table(
      Term = term,
      coef = m$TE.common,
      se = m$seTE.common,
      p = m$pval.common
    )
  )
}

# Test interaction betwen each PGS and the trial arm
int_res <- NULL
for(j in unique(trial_data_gen$batch)){
  for(i in gwas_list$name){
    cox_model <- coxph(as.formula(paste0("Surv(time_to_event, status) ~ arm_numeric * ", i, " + ", paste0('PC', 1:10, collapse = ' + '))), 
                       data = trial_data_gen[trial_data_gen$batch == j,])
    cox_model_sum <- summary(cox_model)
    int_res<-rbind(
      int_res,
      data.table(
        Batch = j,
        PGS = i,
        Term = row.names(coef(cox_model_sum)),
        coef = coef(cox_model_sum)[,1],
        se = coef(cox_model_sum)[,3],
        p = coef(cox_model_sum)[,5]
      )
    )
  }
}

# Meta analyse across batches
library(meta)

int_res_meta <- NULL
for(i in gwas_list$name){
  tmp <- int_res[int_res$PGS == i,]
  for(term in unique(tmp$Term)){
    m <- metagen(TE = tmp$coef[tmp$Term == term], seTE = tmp$se[tmp$Term == term], sm = "HR",)
    
    int_res_meta<-rbind(
      int_res_meta,
      data.table(
        PGS = i,
        Term = term,
        coef = m$TE.common,
        se = m$seTE.common,
        p = m$pval.common
      )
    )
  }
}


```




