---
title: Genetic analysis of Mirocals trial
output: 
  html_document:
    theme: cosmo
    toc: true
    toc_float: true
    toc_depth: 2
    css: styles/styles.css
    includes:
      in_header: header.html
      after_body: footer.html

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(eval = FALSE)
```

```{css, echo=F}
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
```

***

# Proposal

The Mirocals trial tested the efficacy of low dose IL-2 for treatment of ALS. The trial showed IL-2 + riluzole treatment provided a significant improvement in survival compared to those only receiving placebo + riluzole. Low dose IL-2 has been shown to increase the number of regulatory T-cells (Tregs).

This trial has collected whole genome sequence data and methylation data (across several time points). This presents an opportunity to identify genetic predictors/biomarkers of response to IL-2 treatment. The sample size of the trial is limited (N drug = 110, N placebo = 110), which means we need to test specific hypotheses and leverage larger external datasets.

***

## Outcome

There are two key outcome variables from the trial, including survival and regulatory T-cell numbers. It seems survival is the more meaningful outcome for patients, but perhaps there are benefits to using T-cell numbers that I am unaware of. If we are using survival, then we should using a cox regression, but if use T cell numbers linear regression.

***

## Polygenic scores (PGS)

List of polygenic scores that might predict response to IL-2:

* ALS risk, survival, age at onset
* Risk of other neurodegenerative diseases
* Interleukin-2 receptor subunit alpha (IL2RA) protein levels
* Regulatory T-cell molecule levels
* Autoimmune disease risk
* Other inflammatory markers (CRP, IL-6, etc.)

I would suggest using SBayesRC as the method to generate the PGS. If GWAS from multiple populations are available we can using SBayesRC-multi.

We could also further stratify the polygenic scores by functional annotations, such as pathway membership. This could be particularly useful, as it may be that only certain pathways are relevant to IL-2 efficacy. For example, ALS risk mediated by inflammatory pathways may be particularly relevant for IL-2 efficacy. 

***

## Methylation scores

We can use previously published EWAS summary statistics to do the same thing.

***

# Calculate PGS

***

## Prepare target data

```{bash}
mkdir -p /users/k1806347/oliverpainfel/Data/miraclestrial/original

# Create index for each batch
module add bcftools/1.12-gcc-13.2.0-python-3.11.6

####
# Format for GenoPred
####
# Split into per chromosome files

##
# Batch 1
## 
for chr in $(seq 1 22);do
    sbatch --mem 10G -n 1 -p interruptible_cpu -t 1:00:00 --wrap="/users/k1806347/oliverpainfel/Software/plink2 \
        --vcf /scratch/prj/mirocalstrial/recovered/DNAscan/variantcalling/merged_strelka_stringent.vcf.gz \
        --chr ${chr} \
        --out /users/k1806347/oliverpainfel/Data/miraclestrial/processed/miracles_batch1.chr${chr} \
        --allow-extra-chr \
        --export vcf bgz"
done

# The file is truncated mid chr16, and missing chr 17-22

# Recreate merged dataset 
module add bcftools/1.12-gcc-13.2.0-python-3.11.6

bcftools merge \
  -l sample_vcf_list.txt \
  --missing-to-ref \
  -o /users/k1806347/oliverpainfel/Data/miraclestrial/original/miracles_batch1.vcf.gz

# Replace corrupt version
cp /users/k1806347/oliverpainfel/Data/miraclestrial/original/miracles_batch1.vcf.gz /scratch/prj/mirocalstrial/recovered/DNAscan/variantcalling/merged_strelka_stringent_fixed.vcf.gz

# Split by chromosome
for chr in $(seq 1 22);do
    sbatch --mem 10G -n 1 -p interruptible_cpu -t 1:00:00 --wrap="/users/k1806347/oliverpainfel/Software/plink2 \
        --vcf /users/k1806347/oliverpainfel/Data/miraclestrial/original/miracles_batch1.vcf.gz \
        --chr ${chr} \
        --out /users/k1806347/oliverpainfel/Data/miraclestrial/processed/miracles_batch1.chr${chr} \
        --allow-extra-chr \
        --export vcf bgz"
done

##
# Batch 2
## 
for chr in $(seq 1 22);do
    sbatch --mem 10G -n 1 -p interruptible_cpu -t 1:00:00 --wrap="/users/k1806347/oliverpainfel/Software/plink2 \
        --vcf /scratch/prj/mirocalstrial/recovered/DNAscan_batch2/variantcalling/merged_strelka_stringent.vcf.gz \
        --chr ${chr} \
        --out /users/k1806347/oliverpainfel/Data/miraclestrial/processed/miracles_batch2.chr${chr} \
        --allow-extra-chr \
        --export vcf bgz"
done

# Batch 2 seems fine.

```

***

## Download GWAS

```{r}
# Read in GWAS catalogue metadata
gwas_cat <- fread('~/oliverpainfel/Data/GWAS_sumstats/gwas_catalog_v1.0.2.1-studies_r2025-08-24.tsv')

# Remove GWAS without full sumstats
gwas_cat<-gwas_cat[gwas_cat$`FULL SUMMARY STATISTICS` == 'yes',]

# Remove columns we don't want
gwas_cat <- gwas_cat[, c(
  "PUBMEDID",
  "FIRST AUTHOR",
  "DATE",
  "DISEASE/TRAIT",
  "INITIAL SAMPLE SIZE",
  "REPLICATION SAMPLE SIZE",
  "ASSOCIATION COUNT",
  "MAPPED_TRAIT",
  "STUDY ACCESSION",
  "SUMMARY STATS LOCATION"
), with = F]

# List all the GWAS catalogue IDs that you want to keep
gwas_ids<-c(
  'GCST90027164',
  'GCST009325',
  'GCST007320',
  'GCST90001390',
  'GCST90444478',
  'GCST004455',
  'GCST90274806',
  'GCST90241611',
  'GCST004455',
  'GCST90088221',
  'GCST90469588',
  'GCST90428414',
  'GCST004454',
  'GCST90428415',
  'GCST90274811',
  'GCST90469586',
  'GCST90241613',
  'GCST90469585',
  'GCST90179335',
  'GCST90277655',
  'GCST90019431',
  'GCST90088237',
  'GCST90241612',
  'GCST90001960',
  'GCST90001506',
  'GCST90001505',
  'GCST90001504',
  'GCST90001511',
  'GCST90001510',
  'GCST90001940',
  'GCST90001935',
  'GCST90475282',
  'GCST006250'
)

gwas_cat_select <- gwas_cat[gwas_cat$`STUDY ACCESSION` %in% gwas_ids,]
gwas_cat_select$url <- NA
  
# Download the summary statistics
dir.create('~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/')

# Get the URL for each GWAS
for(i in 1:nrow(gwas_cat_select)){
  print(i)
  if(is.na(gwas_cat_select$url[i])){
    con <- curl::curl(gwas_cat_select$`SUMMARY STATS LOCATION`[i])
    lines <- readLines(con)
    close(con)
    
    # extract all href targets
    files <- regmatches(lines, gregexpr('href="([^"]+)"', lines))
    files <- unlist(lapply(files, function(x) sub('href="([^"]+)"', '\\1', x)))
    files <- files[!files %in% c("../")]
  
    if('harmonised/' %in% files){
      con <- curl::curl(paste0(gwas_cat_select$`SUMMARY STATS LOCATION`[i],'/harmonised'))
      lines <- readLines(con)
      close(con)
      
      # extract all href targets
      files <- regmatches(lines, gregexpr('href="([^"]+)"', lines))
      files <- unlist(lapply(files, function(x) sub('href="([^"]+)"', '\\1', x)))
      files <- files[!files %in% c("../")]
      
      url <- files[grepl('.h.tsv.gz$', files)]
      url <- paste0(gwas_cat_select$`SUMMARY STATS LOCATION`[i],'/harmonised/', url)
      gwas_cat_select$url[i] <- url
    
    } else {
      url <- files[grepl('.tsv$|.tsv.gz', files)]
      url <- paste0(gwas_cat_select$`SUMMARY STATS LOCATION`[i],'/', url)
      gwas_cat_select$url[i] <- url
    }
  }
}

# Download the sumstats
for(i in 1:nrow(gwas_cat_select)){
  print(i)
  if(!file.exists(paste0('~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/',gwas_cat_select$`STUDY ACCESSION`[i], '.gz'))){
    ending <- ifelse(any(grepl('.gz$', gwas_cat_select$url[i])), '.gz', '')
    log <- system(
      paste0(
        'wget -q --no-check-certificate -O ~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/',gwas_cat_select$`STUDY ACCESSION`[i], ending, ' ', url))
    if(log != 0){
      system(paste0('rm ~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/',gwas_cat_select$`STUDY ACCESSION`[i], '.gz'))
    }
  }
}

write.csv(gwas_cat_select, '~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/gwas_info.csv', row.names = F)

# The Alzheimer's GWAS has lost important information during harmonisation (NEFF)
# Download the original sumstats instead of the harmonised
system('rm ~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/GCST007320.gz')
system(paste0("wget -q --no-check-certificate -O ~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/GCST007320.gz http://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST007001-GCST008000/GCST007320/AD_sumstats_Jansenetal_2019sept.txt.gz"))
gwas_cat_select$url[gwas_cat_select$`STUDY ACCESSION` == 'GCST007320'] <- 'http://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST007001-GCST008000/GCST007320/AD_sumstats_Jansenetal_2019sept.txt.gz'

# Set Neff column to N
ss<-fread('~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/GCST007320.gz')
ss$N<-ss$Neff
ss$Neff<-NULL
ss$Nsum<-NULL
fwrite(ss, '~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/GCST007320.gz', quote = F, na ='NA', sep=' ')
```

Download the SCZ and FTD sumstats, which are not in the GWAS catalogue.

```{bash}
###
# FTD - This can't be downloaded using wget
###
wget https://rdr.ucl.ac.uk/ndownloader/files/45643512 -O ~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/ftd.manzoni.txt

###
# SCZ
###
wget https://figshare.com/ndownloader/files/34517828 -O ~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/scz.trubetskoy.gz

```

```{r}
# Insert A2 information into FTD sumstats
ss<-fread('~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/ftd.manzoni.txt')

var_info<-data.table(do.call(rbind, strsplit(ss$SNP, ':')))
names(var_info)<-c('CHR','BP','A1','A2')
ss$A2<-NA
ss$A2[ss$A1 == var_info$A1] <- var_info$A2[ss$A1 == var_info$A1]
ss$A2[ss$A1 == var_info$A2] <- var_info$A1[ss$A1 == var_info$A2]
ss <- ss[, c('CHR','BP','A1','A2','OR','SE','PVALUE'), with=F]
fwrite(ss, '~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/ftd.manzoni.txt', sep = ' ', quote = F, na = 'NA')
```

```{r}
gwas_cat_select <- fread('~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/gwas_info.csv')

gwas_cat_select<-apply(gwas_cat_select, 2, as.character)

gwas_cat_select <- rbind(
  gwas_cat_select, 
  data.table(
    PUBMEDID = '38889728',
    `FIRST AUTHOR` = 'Manzoni C',
    DATE = '2024-07-11',
    `DISEASE/TRAIT` = 'Sporadic frontotemporal dementia',
    `INITIAL SAMPLE SIZE` = '4,685 sFTD cases and 15,308 controls',
    `REPLICATION SAMPLE SIZE` = NA,
    `ASSOCIATION COUNT` = NA,
    `MAPPED_TRAIT`  = NA,
    `STUDY ACCESSION` = NA,
    `SUMMARY STATS LOCATION` =  'https://rdr.ucl.ac.uk/ndownloader/files/24952874',
    url = 'https://rdr.ucl.ac.uk/ndownloader/files/24952874'
  ))

gwas_cat_select <- rbind(
  gwas_cat_select, 
  data.table(
    PUBMEDID = '35396580',
    `FIRST AUTHOR` = 'Trubetskoy V',
    DATE = '2022-04-08',
    `DISEASE/TRAIT` = 'Schizophrenia',
    `INITIAL SAMPLE SIZE` = '76,755 individuals with schizophrenia and 243,649 control individuals',
    `REPLICATION SAMPLE SIZE` = NA,
    `ASSOCIATION COUNT` = NA,
    `MAPPED_TRAIT`  = NA,
    `STUDY ACCESSION` = NA,
    `SUMMARY STATS LOCATION` =  'https://figshare.com/ndownloader/files/34517828',
    url = 'https://figshare.com/ndownloader/files/34517828'
  ))

  
write.csv(gwas_cat_select, '~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/gwas_info_2.csv', row.names = F)

```

***

## Prepare config

```{r}
library(data.table)

dir.create('/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/output', recursive = T)
dir.create('/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config', recursive = T)

# Create target_list
target_list <- NULL
target_list <- rbind(target_list, data.table(
  name = 'miracles_batch1',
  path = '/users/k1806347/oliverpainfel/Data/miraclestrial/processed/miracles_batch1',
  type = 'vcf',
  indiv_report = F))

target_list <- rbind(target_list, data.table(
  name = 'miracles_batch2',
  path = '/users/k1806347/oliverpainfel/Data/miraclestrial/processed/miracles_batch2',
  type = 'vcf',
  indiv_report = F))

write.table(
  target_list,
  '/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/target_list.txt',
  col.names = T,
  row.names = F,
  quote = F,
  sep = ' '
)

# Create gwas_list
gwas_cat_select <- fread('~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/gwas_info_2.csv')

gwas_list <- data.table(
  name=gwas_cat_select$`STUDY ACCESSION`,
  path = paste0('/users/k1806347/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/',gwas_cat_select$`STUDY ACCESSION`,'.gz'),
  population = 'EUR',
  n = NA,
  sampling = NA,
  prevalence = NA,
  mean = NA,
  sd = NA,
  label = paste0("\"", gwas_cat_select$`FIRST AUTHOR`, ' - ', gwas_cat_select$DATE, ' - ', gwas_cat_select$`DISEASE/TRAIT`, "\""))

gwas_list$path[is.na(gwas_list$name)] <- c(
  paste0('/users/k1806347/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/ftd.manzoni.txt'),
  paste0('/users/k1806347/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/scz.trubetskoy.gz'))
gwas_list$name[is.na(gwas_list$name)] <- c('ftd.manzoni','scz.trubetskoy')

gwas_list$n <- NA
gwas_list$n[gwas_list$name == 'GCST004455'] <- 3475
gwas_list$n[gwas_list$name == 'GCST004454'] <- 3677
gwas_list$n[gwas_list$name == 'GCST90088221'] <- 5366
gwas_list$n[gwas_list$name == 'GCST90088237'] <- 5361
gwas_list$n[gwas_list$name == 'GCST90001390'] <- 6618
gwas_list$n[gwas_list$name == 'GCST90027164'] <- 138086
gwas_list$n[gwas_list$name == 'GCST90277655'] <- 496
gwas_list$n[gwas_list$name == 'GCST90241611'] <- 3301
gwas_list$n[gwas_list$name == 'GCST90241612'] <- 3301
gwas_list$n[gwas_list$name == 'GCST90241613'] <- 3301
gwas_list$n[gwas_list$name == 'GCST90179335'] <- 2923
gwas_list$n[gwas_list$name == 'GCST90428414'] <- 39034
gwas_list$n[gwas_list$name == 'GCST90428415'] <- 39236
gwas_list$n[gwas_list$name == 'ftd.manzoni'] <- 14989

gwas_list$sampling[gwas_list$name == 'GCST90001390'] <- 2591 / (2591 + 4027)
gwas_list$sampling[gwas_list$name == 'GCST90027164'] <- 27205 / (27205 + 110881)
gwas_list$sampling[gwas_list$name == 'ftd.manzoni'] <- 3756 / (3756 + 11233)
gwas_list$sampling[gwas_list$name == 'scz.trubetskoy'] <- 76755 / (76755 + 243649)
gwas_list$prevalence[gwas_list$name == 'GCST90001390'] <- 0.004
gwas_list$prevalence[gwas_list$name == 'GCST90027164'] <- 0.00005
gwas_list$prevalence[gwas_list$name == 'ftd.manzoni'] <- 0.0002
gwas_list$prevalence[gwas_list$name == 'scz.trubetskoy'] <- 0.01

write.table(
  gwas_list,
  '/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/gwas_list.txt',
  col.names = T,
  row.names = F,
  quote = F,
  sep = ' '
)

# Create config file
conf <- c(
  'outdir: /users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/output',
  'config_file: /users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/config.yaml',
  'target_list: /users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/target_list.txt',
  'gwas_list: /users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/gwas_list.txt',
  "pgs_methods: ['sbayesrc']"
)

write.table(
  conf,
  '/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/config.yaml',
  col.names = F,
  row.names = F,
  quote = F
)

```

***

## Run pipeline

```{bash}
cd /users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline
conda activate genopred

snakemake \
  --profile slurm \
  --use-conda \
  --configfile=/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/config.yaml \
  ldsc -n 
```

***

# Have a look at the LDSC results

```{r}
setwd('/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline/')

source('../functions/misc.R')
source_all('../functions')

config <- '/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/config.yaml'
gwas_list<-read_param(config = config, param = 'gwas_list')
outdir<-read_param(config = config, param = 'outdir', return_obj = F)

ldsc<-NULL
for(i in gwas_list$name){
  h2_log <- readLines(paste0(outdir,'/reference/ldsc/', i,'/', i, '.log'))
  h2_log <- h2_log[grepl('SNP-heritability estimate on the observed scale', h2_log)]
  h2_log <- gsub('SNP-heritability estimate on the observed scale = ','',h2_log)
  hsq <- as.numeric(gsub(' .*','',h2_log))
  h2_se <- as.numeric(gsub("\\).", '', gsub(".* \\(", '', h2_log)))
  hsq.pv <- pnorm(hsq / h2_se, lower.tail = FALSE)
  ldsc<-rbind(
    ldsc,
    data.table(
      name = i,
      est = hsq,
      se = h2_se,
      z = hsq/h2_se,
      p = hsq.pv
    )
  )
}

gwas_cat_select <- fread('~/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/gwas_info_2.csv')
gwas_cat_select$ID <- gwas_cat_select$`STUDY ACCESSION`
gwas_cat_select$ID[is.na(gwas_cat_select$ID)] <- c('ftd.manzoni','scz.trubetskoy')

gwas_cat_select<-merge(gwas_cat_select, ldsc, by.x = 'ID', by.y='name')
ldsc <- gwas_cat_select[, c('ID','STUDY ACCESSION','FIRST AUTHOR','DATE','DISEASE/TRAIT','est','se','z','p'), with = F]

write.csv(
  ldsc,
  '/users/k1806347/oliverpainfel/Data/GWAS_sumstats/mirocalstrial/ldsc.csv',
  row.names = F
)

gwas_list<-gwas_list[name %in% ldsc$ID[ldsc$z > 2],]
gwas_list$label <- paste0("\"", gwas_list$label, "\"")

write.table(
  gwas_list,
  '/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/gwas_list.txt',
  col.names = T,
  row.names = F,
  quote = F,
  sep = ' '
)

```

```{bash}
cd /users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline
conda activate genopred

snakemake \
  --profile slurm \
  --use-conda \
  --configfile=/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/config.yaml \
  output_all outlier_detection -n 
```

***

The vast majority of the clinical trial participants are of European ancestry (Non-EUR N = 15, EUR N = 267). We should restrict the primary analysis to EUR, and then rerun key analyses in full sample. 

***

Analyses:

1. Replicate primary analysis of clinical trial (is this published/can I see details on this)
2. Then test for interaction between PGS and IL-22 on primary endpoint
  - We will have very little power to detect main/interaction effects at statistical significance
  - Could test PGS for risk as primary analysis, as previous studies have done, and then do a hypothesis-free analysis across many PGS, controlling for multiple testing?
    - ALS risk GWAS Could build multi-trait and multi-population model. Build elastic net containing many scores using ProjectMine and independent GWAS
    

What about Jiajing's PGS? seemed to predict risk very highly.

I can generate the PGS and hand them over, or do the analysis? Depends on how Ahmad wants to structure the papers.

***

# Clinical data

The original paper found no effect of the treatment in the unadjusted analysis, but did find a significant effect when including the following covariates: age, ALSFRS-R, CSF-pNFH levels, plasma-CCL2 levels, absolute number of Tregs, and Treatment by CSF-pNFH interaction.

I have been sent some preliminary clinical data for the trial. Lets have a look at it:

```{r}
library(data.table)
library(readxl)

data <- read_excel("~/oliverpainfel/Data/miraclestrial/original/Survival-MIR.xlsx")
data <- data.table(data)

# Rename variables
original_names<-names(data)
names(data) <- c('reference','label','id','c9','unc13a','diagnosis','country','site','reference_2','sex','age_at_diagnosis','age_at_onset','site_of_onset','arm','sex_2','familial','alfrs','svc','age_at_death','age_at_death_rounded_down','randomisation','reference_3','gender','age_c','site_of_onset_2','arm_2')

# Overview of data
str(data)
summary(data)

# Check references are identical
all(data$reference == data$reference_2)
all(data$reference == data$reference_3) # We can delete reference_3 as identical to reference
data$reference_3 <- NULL
data[data$reference != data$reference_2,] # There is an reference value matches different row in reference_2.

# Check sex are identical
all(data$sex == data$sex_2)
all(data$sex == data$gender)
all(data$sex_2 == data$gender) # We can delete gender as identical to sex_2
data$gender <- NULL
data[data$sex != data$sex_2,] # The same rows with discordant reference values have discordant sex values.

# Check site of onset
all(data$site_of_onset == data$site_of_onset_2)
data[data$site_of_onset != data$site_of_onset_2,] # One individual has different site of onset (also mismatched for reference)

# Check arm
all(data$arm == data$arm_2, na.rm = T) # They are identical
data$arm_2 <- NULL
data$arm

# Check age variables
# age_c seems to be age_at_diagnosis, just rounded to 1 decimal point, so delete age_c
data$age_c <- NULL

# Create variable indicating death before end of trial
data$age_at_death_rounded_down <- NULL
data$died_during_trial <- ifelse(grepl('Alive at cut off date', data$age_at_death), F, T)
data$died_during_trial[grepl('NA', data$age_at_death)] <- NA
data$age_at_death <- as.numeric(data$age_at_death)

# Check site of onset
all(data$site_of_onset == data$site_of_onset_2)

#####
# Run cox regression
#####
library(survival)

# Subset those in the trial
trial_data <- data[!is.na(data$arm),]
trial_data <- trial_data[trial_data$randomisation == 'RN',]
nrow(trial_data) # 218 - This is two less than the publication

# We need to create a time to event variable
# We can assume age at diagnosis is the date of enrolment
# The trial ran for 640 days, so we can insert time_to_event for those that survived
trial_data$time_to_event <- trial_data$age_at_death - trial_data$age_at_diagnosis
trial_data$time_to_event[trial_data$died_during_trial == F] <- trial_data$age_at_death[trial_data$died_during_trial == F] - trial_data$age_at_diagnosis[trial_data$died_during_trial == F] + (640/365)
trial_data$status <- as.numeric(trial_data$died_during_trial)
trial_data$arm_numeric <- ifelse(trial_data$arm == 'IL2', 1, 0)

cox_model <- coxph(Surv(time_to_event, status) ~ arm_numeric, data = trial_data)
summary(cox_model)

fit <- survfit(Surv(time_to_event, status) ~ arm_numeric, data = trial_data)
plot(fit, col = c("blue", "red"), xlab = "Time (years)", ylab = "Survival probability")
legend("bottomleft", legend = c("Control", "Treatment"), col = c("blue", "red"), lty = 1)

# Shows non-significant increase in survival in those given IL-2
# The real trial used prognostic covariates which we do not have access to currently.

####
# PGS interactions
####

setwd('/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline/')

source('../functions/misc.R')
source_all('../functions')

config <- '/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/config.yaml'

# Read in the PGS and within sample PCs
pgs <- read_pgs(config = config)
gwas_list <- read_param(config = config, param = 'gwas_list')

# Read in PCs
pcs <-list()
pcs[['miracles_batch1']] <- fread('~/oliverpainfel/Data/miraclestrial/GenoPred/output/miracles_batch1/pcs/within_sample/miracles_batch1.outlier_detection.EUR.PCs.txt')
pcs[['miracles_batch2']] <- fread('~/oliverpainfel/Data/miraclestrial/GenoPred/output/miracles_batch2/pcs/within_sample/miracles_batch2.outlier_detection.EUR.PCs.txt')

# Combine the data across batches
pgs_both <- NULL
for(i in 1:2){
  a<-pgs[[paste0('miracles_batch', i)]]$TRANS
  merged_df <- Reduce(function(x, y) merge(x, y, all = TRUE), lapply(a, function(x) x$sbayesrc))
  names(merged_df) <- gsub('_SBayesRC','', names(merged_df))
  pgs_both <- rbind(pgs_both, merged_df)
}

pcs_both <- NULL
for(i in 1:2){
  a <- pcs[[paste0('miracles_batch', i)]]
  a$batch <- paste0('miracles_batch', i)
  pcs_both <- rbind(pcs_both, a)
}

genetics <- merge(pgs_both, pcs_both, by = c('FID','IID'))
trial_data_gen <- merge(trial_data, genetics, by.x = 'id', by.y = 'IID')

cox_model <- coxph(Surv(time_to_event, status) ~ arm_numeric, data = trial_data_gen)
summary(cox_model)

fit <- survfit(Surv(time_to_event, status) ~ arm_numeric, data = trial_data)
plot(fit, col = c("blue", "red"), xlab = "Time (years)", ylab = "Survival probability")
legend("bottomleft", legend = c("Control", "Treatment"), col = c("blue", "red"), lty = 1)

# Test again with PGS but split by batch and with PCs 
int_res_no_pgs <- NULL
for(j in unique(trial_data_gen$batch)){
  cox_model <- coxph(as.formula(paste0("Surv(time_to_event, status) ~ arm_numeric + ", paste0('PC', 1:10, collapse = ' + '))), 
                     data = trial_data_gen[trial_data_gen$batch == j,])
  cox_model_sum <- summary(cox_model)
  int_res_no_pgs<-rbind(
    int_res_no_pgs,
    data.table(
      Batch = j,
      Term = row.names(coef(cox_model_sum)),
      coef = coef(cox_model_sum)[,1],
      se = coef(cox_model_sum)[,3],
      p = coef(cox_model_sum)[,5]
    )
  )
}

# Meta analyse across batches
library(meta)

int_res_no_pgs_meta <- NULL
for(term in unique(int_res_no_pgs$Term)){
  m <- metagen(TE = int_res_no_pgs$coef[int_res_no_pgs$Term == term], seTE = int_res_no_pgs$se[int_res_no_pgs$Term == term], sm = "HR",)
  
  int_res_no_pgs_meta<-rbind(
    int_res_no_pgs_meta,
    data.table(
      Term = term,
      coef = m$TE.common,
      se = m$seTE.common,
      p = m$pval.common
    )
  )
}

# Test interaction betwen each PGS and the trial arm
int_res <- NULL
for(j in unique(trial_data_gen$batch)){
  for(i in gwas_list$name){
    cox_model <- coxph(as.formula(paste0("Surv(time_to_event, status) ~ arm_numeric * ", i, " + ", paste0('PC', 1:10, collapse = ' + '))), 
                       data = trial_data_gen[trial_data_gen$batch == j,])
    cox_model_sum <- summary(cox_model)
    int_res<-rbind(
      int_res,
      data.table(
        Batch = j,
        PGS = i,
        Term = row.names(coef(cox_model_sum)),
        coef = coef(cox_model_sum)[,1],
        se = coef(cox_model_sum)[,3],
        p = coef(cox_model_sum)[,5]
      )
    )
  }
}

# Meta analyse across batches
library(meta)

int_res_meta <- NULL
for(i in gwas_list$name){
  tmp <- int_res[int_res$PGS == i,]
  for(term in unique(tmp$Term)){
    m <- metagen(TE = tmp$coef[tmp$Term == term], seTE = tmp$se[tmp$Term == term], sm = "HR",)
    
    int_res_meta<-rbind(
      int_res_meta,
      data.table(
        PGS = i,
        Term = term,
        coef = m$TE.common,
        se = m$seTE.common,
        p = m$pval.common
      )
    )
  }
}


```

***

I have been sent new clinical data. Rerun the analyses.

```{r}
library(data.table)
library(readxl)

# Read in data
data <- read_excel("~/oliverpainfel/Data/miraclestrial/original/MIROCALS_AAK_unlocked.xlsx")
data <- data.table(data)

# There are multiple columns with duplicate headers
all(data$`COUNTRY_ID...1` == data$COUNTRY_ID...9, na.rm = T)
sum(is.na(data$`COUNTRY_ID...1`))
sum(is.na(data$`COUNTRY_ID...9`))
data$`COUNTRY_ID...9` <- NULL
names(data)[names(data) == 'COUNTRY_ID...1']<-'COUNTRY_ID'

all(data$SITE_ID...2 == data$SITE_ID...10, na.rm = T)
sum(is.na(data$`SITE_ID...2`))
sum(is.na(data$`SITE_ID...10`))
data$`SITE_ID...10` <- NULL
names(data)[names(data) == 'SITE_ID...2']<-'SITE_ID'

all(data$SUBJECT_REF...3 == data$SUBJECT_REF...6, na.rm = T)
sum(is.na(data$`SUBJECT_REF...3`))
sum(is.na(data$`SUBJECT_REF...6`))
data$`SUBJECT_REF...6` <- NULL
names(data)[names(data) == 'SUBJECT_REF...3']<-'SUBJECT_REF'

all(data$MIRCODE...4 == data$MIRCODE...11, na.rm = T)
sum(is.na(data$`MIRCODE...4`)) # There were no NAs in this
sum(is.na(data$`MIRCODE...11`)) 
data$`MIRCODE...11` <- NULL
names(data)[names(data) == 'MIRCODE...4']<-'MIRCODE'

all(data$ID...5 == data$ID...12, na.rm = T)
sum(is.na(data$`ID...5`)) # There were no NAs in this
sum(is.na(data$`ID...12`)) 
data$`ID...12` <- NULL
names(data)[names(data) == 'ID...5']<-'ID'

all(data$RANDO_num...13 == data$RANDO_num...24, na.rm = T)
sum(is.na(data$`RANDO_num...13`)) 
sum(is.na(data$`RANDO_num...24`)) 
data$`RANDO_num...24` <- NULL
names(data)[names(data) == 'RANDO_num...13']<-'RANDO_num'

all(data$SEX...14 == data$SEX...25, na.rm = T)
sum(is.na(data$`SEX...14`)) 
sum(is.na(data$`SEX...25`)) 
data$`SEX...25` <- NULL
names(data)[names(data) == 'SEX...14']<-'SEX'

all(data$AGEONSET_C...16 == data$AGEONSET_C...26, na.rm = T)
sum(is.na(data$`AGEONSET_C...16`)) 
sum(is.na(data$`AGEONSET_C...26`)) 
data$`AGEONSET_C...26` <- NULL
names(data)[names(data) == 'AGEONSET_C...16']<-'AGEONSET_C'

# Run unadjusted cox regression to test effect of treatment on survival
cox_model <- coxph(Surv(RECLIFE, STATUS) ~ ARM_C, data = data)
summary(cox_model)
# 0.331

fit <- survfit(Surv(RECLIFE, STATUS) ~ ARM_C, data = data)
plot(fit, col = c("blue", "red"), xlab = "Time (years)", ylab = "Survival probability")
legend("bottomleft", legend = c("Control", "Treatment"), col = c("blue", "red"), lty = 1)
# These results look identical to those in the paper.

# Test effect of age, ALSFRS-R, CSF-pNFH levels, plasma-CCL2 levels, absolute number of Tregs
cox_model <- coxph(
  Surv(RECLIFE, STATUS) ~ AGEINC_C + ALSFRS_RND + PNFHCSF_IMP_RDN + CCL2PLASMA_RDN + NTREG_RDN_IMP,
  data = data
)
summary(cox_model)
# All show significant effect as in paper

# Test for interaction between ARM_C * PNFHCSF_IMP_RDN
cox_model <- coxph(
  Surv(RECLIFE, STATUS) ~ ARM_C * PNFHCSF_IMP_RDN,
  data = data
)
summary(cox_model)
# Significant interaction as in paper

# Test effect of treatment with these prognostic factors: age, ALSFRS-R, CSF-pNFH levels, plasma-CCL2 levels, absolute number of Tregs, and Treatment by CSF-pNFH interaction
cox_model <- coxph(
  Surv(RECLIFE, STATUS) ~ AGEINC_C + ALSFRS_RND + PNFHCSF_IMP_RDN + CCL2PLASMA_RDN + NTREG_RDN_IMP + ARM_C * PNFHCSF_IMP_RDN,
  data = data
)
summary(cox_model)
# treatment p = 0.00696, which is consistent with the manuscript.

####
# Check overlap between clinical data and genetic data
####

# Restrict clinical data to the 220 individuals who meet all criteria
data_clean <- data[!is.na(data$RECLIFE),]

# Read in the ancestry inference results
setwd('/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline/')
source('../functions/misc.R')
source_all('../functions')
config <- '/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/config.yaml'

ancestry <- list()
ancestry[['miracles_batch1']] <- read_ancestry(config = config, name = 'miracles_batch1')
ancestry[['miracles_batch2']] <- read_ancestry(config = config, name = 'miracles_batch2')

pop_mem <- NULL
for(i in 1:length(ancestry)){
  tmp <- ancestry[[i]]
  tmp$model_pred$population <- 'Unassigned'
  for(j in 1:length(tmp$keep_files)){
    tmp$model_pred$population[tmp$model_pred$IID %in% tmp$keep_files[[j]]$V1] <- names(tmp$keep_files)[j]
  }
  tmp <- tmp$model_pred[, c('FID','IID','population'), with = F]
  pop_mem <- rbind(pop_mem, tmp)
}

# Combine the data
data_clean_pop <- merge(data_clean, pop_mem, by.x = 'ID', by.y = 'IID')
nrow(data_clean_pop) # 208

# Test whether treatment effect varies across populations
cox_model <- coxph(
  Surv(RECLIFE, STATUS) ~ ARM_C,
  data = data_clean_pop
)
summary(cox_model)
# treatment p = 0.326, similar to unadjusted analysis in full sample

cox_model_full <- coxph(
  Surv(RECLIFE, STATUS) ~ AGEINC_C + ALSFRS_RND + PNFHCSF_IMP_RDN + CCL2PLASMA_RDN + NTREG_RDN_IMP + ARM_C * PNFHCSF_IMP_RDN,
  data = data_clean_pop
)
summary(cox_model_full)
# treatment p = 0.00595, which is consistent with the manuscript.

# Now run within EUR only
cox_model_eur <- coxph(
  Surv(RECLIFE, STATUS) ~ ARM_C,
  data = data_clean_pop[data_clean_pop$population == 'EUR',]
)
summary(cox_model_eur)

cox_model_full_eur <- coxph(
  Surv(RECLIFE, STATUS) ~ AGEINC_C + ALSFRS_RND + PNFHCSF_IMP_RDN + CCL2PLASMA_RDN + NTREG_RDN_IMP + ARM_C * PNFHCSF_IMP_RDN,
  data = data_clean_pop[data_clean_pop$population == 'EUR',]
)
summary(cox_model_full_eur)
# The results are very similar.

####
# PGS interactions
####

setwd('/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline/')

source('../functions/misc.R')
source_all('../functions')

config <- '/users/k1806347/oliverpainfel/Data/miraclestrial/GenoPred/config/config.yaml'

# Read in the PGS and within sample PCs
pgs <- read_pgs(config = config)
gwas_list <- read_param(config = config, param = 'gwas_list')

# Read in PCs
pcs <-list()
pcs[['miracles_batch1']] <- fread('~/oliverpainfel/Data/miraclestrial/GenoPred/output/miracles_batch1/pcs/within_sample/miracles_batch1.outlier_detection.EUR.PCs.txt')
pcs[['miracles_batch2']] <- fread('~/oliverpainfel/Data/miraclestrial/GenoPred/output/miracles_batch2/pcs/within_sample/miracles_batch2.outlier_detection.EUR.PCs.txt')

# Combine the data across batches
pgs_both <- NULL
for(i in 1:2){
  a<-pgs[[paste0('miracles_batch', i)]]$TRANS
  merged_df <- Reduce(function(x, y) merge(x, y, all = TRUE), lapply(a, function(x) x$sbayesrc))
  names(merged_df) <- gsub('_SBayesRC','', names(merged_df))
  pgs_both <- rbind(pgs_both, merged_df)
}

pcs_both <- NULL
for(i in 1:2){
  a <- pcs[[paste0('miracles_batch', i)]]
  a$batch <- paste0('miracles_batch', i)
  pcs_both <- rbind(pcs_both, a)
}

genetics <- merge(pgs_both, pcs_both, by = c('FID','IID'))
trial_data_gen <- merge(data, genetics, by.x = 'ID', by.y = 'IID')

# Rerun unadjusted cox model to see if anything has changed
cox_model <- coxph(Surv(RECLIFE, STATUS) ~ ARM_C, data = trial_data_gen)
summary(cox_model)
# I am loosing 12 individuals due to merge with genetic data. That means there are 12 individuals that were included in the trial that do not have genetic data. There are also 10 people that are not EUR, so don't have PCs.

# Test again with PGS but split by batch and with PCs 
int_res_no_pgs <- NULL
for(j in unique(trial_data_gen$batch)){
  cox_model <- coxph(as.formula(paste0("Surv(RECLIFE, STATUS) ~ ARM_C + ", paste0('PC', 1:10, collapse = ' + '))), 
                     data = trial_data_gen[trial_data_gen$batch == j,])
  cox_model_sum <- summary(cox_model)
  int_res_no_pgs<-rbind(
    int_res_no_pgs,
    data.table(
      Batch = j,
      Term = row.names(coef(cox_model_sum)),
      coef = coef(cox_model_sum)[,1],
      se = coef(cox_model_sum)[,3],
      p = coef(cox_model_sum)[,5]
    )
  )
}

# Meta analyse across batches
library(meta)

int_res_no_pgs_meta <- NULL
for(term in unique(int_res_no_pgs$Term)){
  m <- metagen(TE = int_res_no_pgs$coef[int_res_no_pgs$Term == term], seTE = int_res_no_pgs$se[int_res_no_pgs$Term == term], sm = "HR",)
  
  int_res_no_pgs_meta<-rbind(
    int_res_no_pgs_meta,
    data.table(
      Term = term,
      coef = m$TE.common,
      se = m$seTE.common,
      p = m$pval.common
    )
  )
}

# Test interaction between each PGS and the trial arm
int_res <- NULL
for(j in unique(trial_data_gen$batch)){
  for(i in gwas_list$name){
    cox_model <- coxph(as.formula(paste0("Surv(RECLIFE, STATUS) ~ ARM_C * ", i, " + ", paste0('PC', 1:10, collapse = ' + '))), 
                       data = trial_data_gen[trial_data_gen$batch == j,])
    cox_model_sum <- summary(cox_model)
    int_res<-rbind(
      int_res,
      data.table(
        Batch = j,
        PGS = i,
        Term = row.names(coef(cox_model_sum)),
        coef = coef(cox_model_sum)[,1],
        se = coef(cox_model_sum)[,3],
        p = coef(cox_model_sum)[,5]
      )
    )
  }
}

# Meta analyse across batches
library(meta)

int_res_meta <- NULL
for(i in gwas_list$name){
  tmp <- int_res[int_res$PGS == i,]
  for(term in unique(tmp$Term)){
    m <- metagen(TE = tmp$coef[tmp$Term == term], seTE = tmp$se[tmp$Term == term], sm = "HR",)
    
    int_res_meta<-rbind(
      int_res_meta,
      data.table(
        PGS = i,
        Term = term,
        coef = m$TE.common,
        se = m$seTE.common,
        p = m$pval.common
      )
    )
  }
}


```
