<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>OpenSNP Benchmark</title>

<script src="site_libs/header-attrs-2.23/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.0/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.0/css/v4-shims.min.css" rel="stylesheet" />
<!-- header.html -->
<style>
  .navbar {
    min-height: 80px; /* Adjust as needed */
  }

  .navbar-brand {
    padding: 10px 15px; /* Adjust as needed */
  }
</style>

<link rel="icon" type="image/x-icon" href="favicon.ico"/>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img id="logo" style="height: 60px;" src="Horizontal.png" /></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="research_index.html">Research</a>
</li>
<li>
  <a href="tools_index.html">Tools</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">OpenSNP Benchmark</h1>

</div>


<hr />
<div id="overview" class="section level1">
<h1>Overview</h1>
<p>We used the publicly available OpenSNP dataset to test the GenoPred
pipeline. We used GWAS sumstats for Height from the Yengo et al.Â GWAS.
We used results derived using European and East Asian individuals. We
ran the original version of GenoPred (V1) and the updated version of
GenoPred (V2) to check results were consistent.</p>
<p>As can be seen in the figure below, the GenoPred returned polygenic
scores performing as expected, and results were consistent across
version of GenoPred.</p>
<div class="figure">
<img src="Images/OpenSNP/genopred-v1-comp_strict_eur-yengo_eur.png"
alt="" />
<p class="caption">Performance of PGS methods using European Yengo GWAS
in OpenSNP European individuals</p>
</div>
<p>The code used to perform this analysis is below.</p>
<hr />
</div>
<div id="genopred-v2" class="section level1">
<h1>GenoPred V2</h1>
<div id="prepare-input" class="section level2">
<h2>Prepare input</h2>
<hr />
<div id="download-genotypes" class="section level3">
<h3>Download genotypes</h3>
<details>
<summary>
Show code
</summary>
<pre class="bash"><code>mkdir -p /users/k1806347/oliverpainfel/Data/OpenSNP/raw
cd /users/k1806347/oliverpainfel/Data/OpenSNP/raw
wget https://zenodo.org/records/1442755/files/CrowdAI_v3.tar.gz?download=1 
mv &#39;CrowdAI_v3.tar.gz?download=1&#39; CrowdAI_v3.tar.gz
tar -xvzf CrowdAI_v3.tar.gz
rm CrowdAI_v3.tar.gz

# Use fullset (training) data as the target for GenoPred, as the sample size is larger</code></pre>
</details>
<hr />
</div>
<div id="prepare-phenotype" class="section level3">
<h3>Prepare phenotype</h3>
<details>
<summary>
Show code
</summary>
<pre class="r"><code>library(data.table)

dat &lt;- fread(&#39;/users/k1806347/oliverpainfel/Data/OpenSNP/raw/CrowdAI_v3/training_set_details.txt&#39;)
dat$FID &lt;- 0
dat$IID &lt;- dat$id
dat &lt;- dat[, c(&#39;FID&#39;, &#39;IID&#39;, &#39;height&#39;), with = F]

dir.create(&#39;/users/k1806347/oliverpainfel/Data/OpenSNP/processed/pheno&#39;, recursive = T)
write.table(
  dat,
  &#39;/users/k1806347/oliverpainfel/Data/OpenSNP/processed/pheno/height.txt&#39;,
  col.names = T,
  row.names = F,
  quote = F
)</code></pre>
</details>
<hr />
</div>
<div id="split-vcf-by-chromosome" class="section level3">
<h3>Split VCF by chromosome</h3>
<details>
<summary>
Show code
</summary>
<pre class="bash"><code># Create index
module add bcftools/1.14-gcc-10.3.0-python3+-chk-version
bcftools index /users/k1806347/oliverpainfel/Data/OpenSNP/raw/CrowdAI_v3/fullset/genotyping_data_fullset_train.vcf.gz

# Now, split by chromosome using plink2
# Run on the command line within pipeline conda environment
mkdir /users/k1806347/oliverpainfel/Data/OpenSNP/processed/geno
for chr in $(seq 1 22);do
    plink2 \
        --vcf /users/k1806347/oliverpainfel/Data/OpenSNP/raw/CrowdAI_v3/fullset/genotyping_data_fullset_train.vcf.gz \
        --chr ${chr} \
        --out /users/k1806347/oliverpainfel/Data/OpenSNP/processed/geno/opensnp_train.chr${chr} \
        --export vcf bgz
done</code></pre>
</details>
<hr />
</div>
<div id="download-height-gwas" class="section level3">
<h3>Download height GWAS</h3>
<details>
<summary>
Show code
</summary>
<pre class="bash"><code># These are from the Yengo 2022 paper
mkdir -p /users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test
wget --no-check-certificate -O /users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_eur.txt https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST90245001-GCST90246000/GCST90245992/GCST90245992_buildGRCh37.tsv
wget --no-check-certificate -O /users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_eas.txt https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST90245001-GCST90246000/GCST90245991/GCST90245991_buildGRCh37.tsv</code></pre>
</details>
<hr />
</div>
<div id="create-gwas_list-target_list-and-config"
class="section level3">
<h3>Create gwas_list, target_list and config</h3>
<details>
<summary>
Show code
</summary>
<pre class="r"><code>setwd(&#39;/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/GenoPredPipe&#39;)
library(data.table)

# Create config file
conf &lt;- readLines(&#39;config.yaml&#39;)
conf &lt;- gsub(&#39;outdir: test_data/output/test1&#39;, &#39;outdir: /users/k1806347/oliverpainfel/Data/OpenSNP/GenoPred/test1&#39;, conf)
conf &lt;- gsub(&#39;config_file: config.yaml&#39;, &#39;config_file: misc/opensnp/config.yaml&#39;, conf)
conf &lt;- gsub(&#39;gwas_list: example_input/gwas_list_example.txt&#39;, &#39;gwas_list: misc/opensnp/gwas_list.txt&#39;, conf)
conf &lt;- gsub(&#39;target_list: example_input/target_list_example.txt&#39;, &#39;target_list: misc/opensnp/target_list.txt&#39;, conf)
conf &lt;- conf[conf != &#39;&#39;]
conf &lt;- conf[!grepl(&#39;\\#&#39;, conf)]
conf &lt;- conf[!grepl(&#39;score_list&#39;, conf)]

dir.create(&#39;misc/opensnp/&#39;, recursive = T)
dir.create(&#39;/users/k1806347/oliverpainfel/Data/OpenSNP/GenoPred&#39;, recursive = T)
write.table(conf, &#39;misc/opensnp/config.yaml&#39;, col.names = F, row.names = F, quote = F)

# Create target_list
target_list &lt;- fread(&#39;example_input/target_list_example.txt&#39;)
target_list &lt;- rbind(target_list,
                     data.table(name = &#39;opensnp&#39;,
                                path = &#39;/users/k1806347/oliverpainfel/Data/OpenSNP/processed/geno/opensnp_train&#39;,
                                type = &#39;samp_imp_vcf&#39;,
                                indiv_report = F))

target_list &lt;- target_list[target_list$name == &#39;opensnp&#39;, ]

write.table(target_list, &#39;misc/opensnp/target_list.txt&#39;, col.names = T, row.names = F, quote = F, sep = &#39; &#39;)

# Create gwas_list
gwas_list &lt;- fread(&#39;example_input/gwas_list_example.txt&#39;)
gwas_list&lt;-rbind(gwas_list, 
                 data.table(name=&#39;yengo_eur&#39;,
                            path = &#39;/users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_eur.txt&#39;,
                            population = &#39;EUR&#39;,
                            n = NA,
                            sampling = NA,
                            prevalence = NA,
                            mean = NA,
                            sd = NA,
                            label = &quot;\&quot;Yengo 2022 Height EUR\&quot;&quot;))

gwas_list &lt;- rbind(gwas_list,
                   data.table(name = &#39;yengo_eas&#39;, 
                              path = &#39;/users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_eas.txt&#39;,
                              population = &#39;EAS&#39;,
                              n = NA,
                              sampling = NA,
                              prevalence = NA,
                              mean = NA,
                              sd = NA,
                              label = &quot;\&quot;Yengo 2022 Height EAS\&quot;&quot;))


gwas_list &lt;- gwas_list[gwas_list$name %in% c(&#39;yengo_eur&#39;, &#39;yengo_eas&#39;), ]

write.table(gwas_list, &#39;misc/opensnp/gwas_list.txt&#39;, col.names = T, row.names = F, quote = F, sep = &#39; &#39;)</code></pre>
</details>
<hr />
</div>
<div id="download-a-precomputed-score-file" class="section level3">
<h3>Download a precomputed score file</h3>
<details>
<summary>
Show code
</summary>
<pre class="bash"><code>wget --no-check-certificate -O /users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_eur_pgs.gz https://cnsgenomics.com/data/giant_2022/GIANT_HEIGHT_YENGO_2022_PGS_WEIGHTS_EUR.gz
# Going to pause on this for now until I have figured out a way to harmonise score files.</code></pre>
</details>
<hr />
</div>
</div>
<div id="run-genopred" class="section level2">
<h2>Run GenoPred</h2>
<details>
<summary>
Show code
</summary>
<pre class="bash"><code>snakemake --profile slurm --use-conda --configfile misc/opensnp/config.yaml output_all</code></pre>
</details>
<hr />
</div>
<div id="evaluate-pgs" class="section level2">
<h2>Evaluate PGS</h2>
<details>
<summary>
Show code
</summary>
<pre class="r"><code># Test correlation between PGS and observed height

setwd(&#39;/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/GenoPredPipe/&#39;)
library(data.table)
library(ggplot2)
library(cowplot)

source(&#39;../Scripts/functions/misc.R&#39;)

# Read in pheno data
pheno &lt;- fread(&#39;/users/k1806347/oliverpainfel/Data/OpenSNP/processed/pheno/height.txt&#39;)
pheno$FID&lt;-0

# Read in ancestry data
keep_list &lt;- fread(&#39;/users/k1806347/oliverpainfel/Data/OpenSNP/GenoPred/test1/opensnp/ancestry/keep_list.txt&#39;)

# Read in pgs
gwas_list &lt;- fread(&#39;misc/opensnp/gwas_list.txt&#39;)

# Define pgs_methods used
pgs_methods &lt;- c(&#39;ptclump&#39;, &#39;dbslmm&#39;, &#39;prscs&#39;, &#39;sbayesr&#39;, &#39;lassosum&#39;, &#39;ldpred2&#39;, &#39;megaprs&#39;)

# Define pgs_methods applied to non-EUR GWAS
pgs_methods_eur &lt;- c(&#39;ptclump&#39;,&#39;lassosum&#39;,&#39;megaprs&#39;)

# Read in PGS
pgs &lt;- read_pgs(config = &#39;misc/opensnp/config.yaml&#39;, name = &#39;opensnp&#39;)$opensnp

# Estimate correlation between pheno and pgs
cor &lt;- NULL
for(pop_i in names(pgs)){
  for(gwas_i in names(pgs[[pop_i]])){
    for(pgs_method_i in names(pgs[[pop_i]][[gwas_i]])){
      pgs_i &lt;- pgs[[pop_i]][[gwas_i]][[pgs_method_i]]
      pheno_pgs&lt;-merge(pheno, pgs_i, by = c(&#39;FID&#39;,&#39;IID&#39;))
      
      for(model_i in names(pgs_i)[-1:-2]){
        y &lt;- scale(pheno_pgs$height)
        x &lt;- scale(pheno_pgs[[model_i]])
        
        if(all(is.na(x))){
          next
        }
        
        coef_i &lt;- coef(summary(mod &lt;- lm(y ~ x)))
        
        tmp &lt;- data.table(
          pop = pop_i,
          gwas = gwas_i,
          pgs_method = pgs_method_i,
          name = model_i,
          r = coef_i[2,1],
          se = coef_i[2,2],
          p = coef_i[2,4],
          n = nobs(mod))
      cor &lt;- rbind(cor, tmp)
      
      }
    }
  }
}

# The European sample is the only one large enough for interpretable results
# Subset EUR results
cor_eur &lt;- cor[cor$pop == &#39;EUR&#39;, ]

# Restrict to best and and pseudoval only
cor_eur_subset &lt;- NULL
for(pop_i in unique(cor_eur$pop)){
  for(gwas_i in unique(cor_eur$gwas[cor_eur$pop == pop_i])){
    for(pgs_method_i in unique(cor_eur$pgs_method[cor_eur$pop == pop_i &amp; cor_eur$gwas == gwas_i])){
      
      # Subset relevant results
      cor_eur_i &lt;- cor_eur[
        cor_eur$pop == pop_i &amp;
        cor_eur$gwas == gwas_i &amp;
        cor_eur$pgs_method == pgs_method_i,]
  
      # Top R
      if(pgs_method_i %in% c(&#39;ptclump&#39;,&#39;ldpred2&#39;,&#39;megaprs&#39;,&#39;prscs&#39;,&#39;lassosum&#39;)){
        top_i &lt;- cor_eur_i[which(cor_eur_i$r == max(cor_eur_i$r, na.rm = T))[1],]
        top_i$model &lt;- &#39;Top&#39;
        cor_eur_subset &lt;- rbind(cor_eur_subset, top_i)
      }
  
      # PseudoVal
      if(pgs_method_i %in% c(&#39;ptclump&#39;,&#39;sbayesr&#39;,&#39;ldpred2&#39;,&#39;megaprs&#39;,&#39;prscs&#39;,&#39;lassosum&#39;,&#39;dbslmm&#39;)){
        cor_eur_i$name &lt;- gsub(paste0(gwas_i, &#39;_&#39;), &#39;&#39;, cor_eur_i$name)
        pseudo_param &lt;- find_pseudo(config = &#39;misc/opensnp/config.yaml&#39;, gwas = gwas_i, pgs_method = pgs_method_i)
        pseudo_i &lt;- cor_eur_i[cor_eur_i$name == pseudo_param,]
        pseudo_i$model &lt;- &#39;Pseudo&#39;
        cor_eur_subset &lt;- rbind(cor_eur_subset, pseudo_i)
      }
    }
  }
}

# Plot the results
cor_eur_subset$model &lt;- factor(cor_eur_subset$model, levels = c(&#39;Top&#39;,&#39;Pseudo&#39;))
dir.create(&#39;/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/docs/Images/OpenSNP&#39;)

# yengo_eur
png(&#39;/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/docs/Images/OpenSNP/genopred-yengo_eur.png&#39;,
    units = &#39;px&#39;,
    width = 2000,
    height = 1000,
    res = 300)

  ggplot(cor_eur_subset[cor_eur_subset$gwas == &#39;yengo_eur&#39;,], aes(x = pgs_method, y = r, fill = model)) +
    geom_bar(stat = &quot;identity&quot;, position = position_dodge2(preserve = &quot;single&quot;), width = 0.7) +
    geom_errorbar(
      aes(ymin = r - se, ymax = r + se),
      width = .2,
      position = position_dodge(width = 0.7)
    ) +
    labs(
      y = &quot;Correlation (SE)&quot;,
      x = &#39;PGS Method&#39;,
      fill = &#39;Model&#39;,
      title = paste0(&quot;Yengo - EUR vs OpenSNP - EUR\n(N = &quot;, cor_eur_subset$n[1], &quot;)&quot;)
    ) +
    theme_half_open() +
    background_grid() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),
          plot.title = element_text(hjust = 0.5, size=12))

dev.off()

# yengo_eas
png(&#39;/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/docs/Images/OpenSNP/genopred-yengo_eas.png&#39;,
    units = &#39;px&#39;,
    width = 1200,
    height = 1000,
    res = 300)

  ggplot(cor_eur_subset[cor_eur_subset$gwas == &#39;yengo_eas&#39;,], aes(x = pgs_method, y = r, fill = model)) +
    geom_bar(stat = &quot;identity&quot;, position = position_dodge2(preserve = &quot;single&quot;), width = 0.7) +
    geom_errorbar(
      aes(ymin = r - se, ymax = r + se),
      width = .2,
      position = position_dodge(width = 0.7)
    ) +
    labs(
      y = &quot;Correlation (SE)&quot;,
      x = &#39;PGS Method&#39;,
      fill = &#39;Model&#39;,
      title = paste0(&quot;Yengo - EAS vs OpenSNP - EUR\n(N = &quot;, cor_eur_subset$n[1], &quot;)&quot;)
    ) +
    theme_half_open() +
    background_grid() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),
          plot.title = element_text(hjust = 0.5, size=12))

dev.off()

# Save the results
dir.create(&#39;/scratch/prj/oliverpainfel/Data/OpenSNP/assoc&#39;)
write.csv(
  cor,
  &#39;/scratch/prj/oliverpainfel/Data/OpenSNP/assoc/genopred-yengo-assoc.csv&#39;,
  row.names = F
)</code></pre>
</details>
<hr />
<details>
<summary>
Show results in OpenSNP
</summary>
<center>
<p><img src="Images/OpenSNP/genopred-yengo_eur.png"
style="width:60.0%" /></p>
<hr />
<p><img src="Images/OpenSNP/genopred-yengo_eas.png"
style="width:40.0%" /></p>
</details>
<hr />
</div>
</div>
<div id="genopred-v1-comparison" class="section level1">
<h1>GenoPred V1 comparison</h1>
<p>We will run OpenSNP through GenoPred v1, to check whether updates
have altered the output.</p>
<details>
<summary>
Show code
</summary>
<pre class="bash"><code># Go to another version of the repo on CREATE
cd /users/k1806347/oliverpainfel/test/GenoPred

# Checkout to the v1 version of the repo
git checkout v1
</code></pre>
</details>
<hr />
<div id="prepare-input-1" class="section level3">
<h3>Prepare input</h3>
<details>
<summary>
Show code
</summary>
<pre class="r"><code>library(data.table)
setwd(&#39;/users/k1806347/oliverpainfel/test/GenoPred/GenoPredPipe&#39;)

# target_list
target_list &lt;- fread(&#39;target_list_example.txt&#39;)
target_list &lt;- data.frame(
  name = &#39;opensnp&#39;,
  path=&#39;/users/k1806347/oliverpainfel/Data/OpenSNP/processed/geno/opensnp_train&#39;,
  type = &#39;samp_imp_vcf&#39;,
  output = &#39;/users/k1806347/oliverpainfel/Data/OpenSNP/GenoPred_v1/test1&#39;,
  indiv_report = F)

dir.create(&#39;misc/opensnp&#39;, recursive = T)
write.table(target_list, &#39;misc/opensnp/target_list.txt&#39;, col.names=T, row.names=F, quote=F, sep=&#39; &#39;)

# gwas_list
yengo_eur &lt;- fread(&#39;/users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_eur.txt&#39;)
yengo_eur &lt;- yengo_eur[, c(&#39;variant_id&#39;,&#39;effect_allele&#39;,&#39;other_allele&#39;,&#39;beta&#39;,&#39;standard_error&#39;,&#39;effect_allele_frequency&#39;,&#39;p_value&#39;,&#39;n&#39;), with=F]
names(yengo_eur) &lt;- c(&#39;SNP&#39;,&#39;A1&#39;,&#39;A2&#39;,&#39;BETA&#39;,&#39;SE&#39;,&#39;FREQ&#39;,&#39;P&#39;,&#39;N&#39;)
fwrite(yengo_eur, &#39;/users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_eur.format.txt&#39;, sep = &#39; &#39;, quote = F, na = &#39;NA&#39;)

yengo_eas &lt;- fread(&#39;/users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_eas.txt&#39;)
yengo_eas &lt;- yengo_eas[, c(&#39;variant_id&#39;,&#39;effect_allele&#39;,&#39;other_allele&#39;,&#39;beta&#39;,&#39;standard_error&#39;,&#39;effect_allele_frequency&#39;,&#39;p_value&#39;,&#39;n&#39;), with=F]
names(yengo_eas) &lt;- c(&#39;SNP&#39;,&#39;A1&#39;,&#39;A2&#39;,&#39;BETA&#39;,&#39;SE&#39;,&#39;FREQ&#39;,&#39;P&#39;,&#39;N&#39;)
fwrite(yengo_eas, &#39;/users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_eas.format.txt&#39;, sep = &#39; &#39;, quote = F, na = &#39;NA&#39;)

gwas_list &lt;- fread(&#39;gwas_list_example.txt&#39;)
gwas_list&lt;-rbind(gwas_list, data.table(
    name=&#39;yengoeur&#39;,
    path=&#39;/users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_eur.format.txt&#39;,
    population=&#39;EUR&#39;,
    sampling=NA,
    prevalence=NA,
    mean=NA,
    sd=NA,
    label=&quot;\&quot;Yengo 2022 Height EUR\&quot;&quot;))

gwas_list&lt;-rbind(gwas_list, data.table(
    name=&#39;yengoeas&#39;,
    path=&#39;/users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_eas.format.txt&#39;,
    population=&#39;EAS&#39;,
    sampling=NA,
    prevalence=NA,
    mean=NA,
    sd=NA,
    label=&quot;\&quot;Yengo 2022 Height EAS\&quot;&quot;))

gwas_list&lt;-gwas_list[gwas_list$name %in% c(&#39;yengoeur&#39;,&#39;yengoeas&#39;),]

write.table(gwas_list, &#39;misc/opensnp/gwas_list.txt&#39;, col.names=T, row.names=F, quote=F, sep=&#39; &#39;)

# score_list
score_list &lt;- fread(&#39;score_list_example.txt&#39;)
score_list &lt;- score_list[-1,]
write.table(score_list, &#39;misc/opensnp/score_list.txt&#39;, col.names=T, row.names=F, quote=F, sep=&#39; &#39;)

# config
config &lt;- c(
  &quot;gwas_list: misc/opensnp/gwas_list.txt&quot;,
  &quot;target_list: misc/opensnp/target_list.txt&quot;,
  &quot;score_list: misc/opensnp/score_list.txt&quot;   
)
write.table(config, &#39;misc/opensnp/config.yaml&#39;, col.names=F, row.names=F, quote=F, sep=&#39; &#39;)</code></pre>
</details>
<hr />
</div>
<div id="run-genopred-v1" class="section level3">
<h3>Run GenoPred v1</h3>
<details>
<summary>
Show code
</summary>
<pre class="bash"><code># Calculate score using all methods
snakemake -n --profile slurm --configfile=misc/opensnp/config.yaml --use-conda run_target_prs_all
</code></pre>
</details>
<hr />
</div>
<div id="evaluate-pgs-1" class="section level2">
<h2>Evaluate PGS</h2>
<details>
<summary>
Show code
</summary>
<pre class="r"><code># Test correlation between PGS and phenotype

setwd(&#39;/scratch//prj/oliverpainfel/test/GenoPred/GenoPredPipe/&#39;)
library(data.table)
library(Hmisc)

# Read in pheno data
pheno &lt;- fread(&#39;/users/k1806347/oliverpainfel/Data/OpenSNP/processed/pheno/height.txt&#39;)
pheno$FID&lt;-0

# Read in ancestry data
keep_list &lt;- fread(&#39;/users/k1806347/oliverpainfel/Data/OpenSNP/GenoPred/test1/opensnp/ancestry/keep_list.txt&#39;)

# Read in pgs
gwas_list &lt;- fread(&#39;misc/opensnp/gwas_list.txt&#39;)
pgs_methods &lt;- c(&#39;pt_clump&#39;,&#39;dbslmm&#39;,&#39;prscs&#39;,&#39;sbayesr&#39;,&#39;lassosum&#39;,&#39;ldpred2&#39;,&#39;megaprs&#39;)
pgs_methods_eur &lt;- c(&#39;pt_clump&#39;,&#39;lassosum&#39;,&#39;megaprs&#39;)

pgs &lt;- list()
for(pop_i in keep_list$POP){
  pgs[[pop_i]] &lt;- list()
  for(gwas_i in gwas_list$name){
    pgs[[pop_i]][[gwas_i]] &lt;- list()
    for(pgs_method_i in pgs_methods){
      if(gwas_list$population[gwas_list$name == gwas_i] == &#39;EUR&#39; | (gwas_list$population[gwas_list$name == gwas_i] != &#39;EUR&#39; &amp; (pgs_method_i %in% pgs_methods_eur))){
        pgs[[pop_i]][[gwas_i]][[pgs_method_i]] &lt;- fread(paste0(&#39;/users/k1806347/oliverpainfel/Data/OpenSNP/GenoPred_v1/test1/opensnp/prs/&#39;,pop_i,&#39;/&#39;,pgs_method_i,&#39;/&#39;,gwas_i,&#39;/opensnp.&#39;,gwas_i,&#39;.&#39;,pop_i,&#39;.profiles&#39;))
      }
    }
  }
}

# Estimate correlation between pheno and pgs
cor &lt;- NULL
for(pop_i in names(pgs)){
  for(gwas_i in names(pgs[[pop_i]])){
    for(pgs_method_i in names(pgs[[pop_i]][[gwas_i]])){
      pgs_i &lt;- pgs[[pop_i]][[gwas_i]][[pgs_method_i]]
      pheno_pgs&lt;-merge(pheno, pgs_i, by = c(&#39;FID&#39;,&#39;IID&#39;))
      
      for(model_i in names(pgs_i)[-1:-2]){
        y &lt;- scale(pheno_pgs$height)
        x &lt;- scale(pheno_pgs[[model_i]])
        
        if(all(is.na(x))){
          next
        }
        
        coef_i &lt;- coef(summary(mod &lt;- lm(y ~ x)))
        
        tmp &lt;- data.table(
          pop = pop_i,
          gwas = gwas_i,
          pgs_method = pgs_method_i,
          name = model_i,
          r = coef_i[2,1],
          se = coef_i[2,2],
          p = coef_i[2,4],
          n = nobs(mod))
      cor &lt;- rbind(cor, tmp)
      
      }
    }
  }
}

# Save the results
dir.create(&#39;/scratch/prj/oliverpainfel/Data/OpenSNP/assoc&#39;)
write.csv(
  cor,
  &#39;/scratch/prj/oliverpainfel/Data/OpenSNP/assoc/genopred-v1-yengo-assoc.csv&#39;,
  row.names = F
)

# Restrict to best only
cor_subset &lt;- NULL
for(pop_i in unique(cor$pop)){
  for(gwas_i in unique(cor$gwas[cor$pop == pop_i])){
    for(pgs_method_i in unique(cor$pgs_method[cor$pop == pop_i &amp; cor$gwas == gwas_i])){
      
      # Subset relevant results
      cor_i &lt;- cor[
        cor$pop == pop_i &amp;
        cor$gwas == gwas_i &amp;
        cor$pgs_method == pgs_method_i,]
  
      # Top R
      top_i &lt;- cor_i[which(cor_i$r == max(cor_i$r, na.rm = T))[1],]
      cor_subset &lt;- rbind(cor_subset, top_i)
    }
  }
}

# Read in the associations for the best PGS when using the new version of GenoPred
cor_new &lt;- fread(&#39;/scratch/prj/oliverpainfel/Data/OpenSNP/assoc/genopred-yengo-assoc.csv&#39;)

cor_new_subset &lt;- NULL
for(pop_i in unique(cor_new$pop)){
  for(gwas_i in unique(cor_new$gwas[cor_new$pop == pop_i])){
    for(pgs_method_i in unique(cor_new$pgs_method[cor_new$pop == pop_i &amp; cor_new$gwas == gwas_i])){
      
      # Subset relevant results
      cor_new_i &lt;- cor_new[
        cor_new$pop == pop_i &amp;
        cor_new$gwas == gwas_i &amp;
        cor_new$pgs_method == pgs_method_i,]
  
      # Top R
      top_i &lt;- cor_new_i[which(cor_new_i$r == max(cor_new_i$r, na.rm = T))[1],]
      cor_new_subset &lt;- rbind(cor_new_subset, top_i)
    }
  }
}

# Compare the results
cor_subset$Version &lt;- &#39;GenoPred V1&#39;
cor_subset$pgs_method&lt;-gsub(&#39;_&#39;, &#39;&#39;, cor_subset$pgs_method)
cor_new_subset$Version &lt;- &#39;GenoPred V2&#39;
cor_new_subset$gwas&lt;-gsub(&#39;_&#39;, &#39;&#39;, cor_new_subset$gwas)

cor_both &lt;- rbind(cor_subset, cor_new_subset)

cor_both$pop &lt;- paste0(cor_both$pop, &quot;\n N = &quot;, cor_both$n)

##
# Plot the results
##

# yengo_eur
tmp &lt;- cor_both[cor_both$gwas == &#39;yengoeur&#39;,]
y_lim &lt;- c(min(tmp$r - tmp$se), max(tmp$r + tmp$se))
                  
v1_plot &lt;-
  ggplot(cor_both[cor_both$gwas == &#39;yengoeur&#39; &amp; cor_both$Version == &#39;GenoPred V1&#39;, ], 
         aes(x = pgs_method, y = r)) +
  geom_hline(yintercept = 0) +
    geom_bar(stat = &quot;identity&quot;, position = position_dodge(), width = 0.7, fill = &#39;#F8766D&#39;) +
    geom_errorbar(
      aes(ymin = r - se, ymax = r + se),
      width = .2,
      position = position_dodge(width = 0.7)
    ) +
    labs(
      y = &quot;Correlation (SE)&quot;,
      x = &#39;PGS Method&#39;,
      title = &#39;Yengo - EUR vs OpenSNP&#39;
    ) +
    ylim(y_lim) +
    theme_half_open() +
    background_grid() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),
          plot.title = element_text(hjust = 0.5, size=12)) +
    facet_grid(pop ~ Version) +
    panel_border()

v2_plot &lt;-
  ggplot(cor_both[cor_both$gwas == &#39;yengoeur&#39; &amp; cor_both$Version == &#39;GenoPred V2&#39;, ], 
         aes(x = pgs_method, y = r)) +
  geom_hline(yintercept = 0) +
    geom_bar(stat = &quot;identity&quot;, position = position_dodge(), width = 0.7, fill = &#39;#00BFC4&#39;) +
    geom_errorbar(
      aes(ymin = r - se, ymax = r + se),
      width = .2,
      position = position_dodge(width = 0.7)
    ) +
    labs(
      y = &quot;Correlation (SE)&quot;,
      x = &#39;PGS Method&#39;,
      title = &#39;Yengo - EUR vs OpenSNP&#39;
    ) +
    ylim(y_lim) +
    theme_half_open() +
    background_grid() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),
          plot.title = element_text(hjust = 0.5, size=12)) +
    facet_grid(pop ~ Version) +
    panel_border()

png(&#39;/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/docs/Images/OpenSNP/genopred-v1-comp-yengo_eur.png&#39;,
    units = &#39;px&#39;,
    width = 3000,
    height = 3000,
    res = 300)

  plot_grid(v1_plot, v2_plot, labels = NULL)
  
dev.off()

###
# Compare again using same ancestry classification threshold
###

# Read in ancestry predictions
model_pred &lt;- fread(&#39;/users/k1806347/oliverpainfel/Data/OpenSNP/GenoPred_v1/test1/opensnp/ancestry/ancestry_all/opensnp.Ancestry.model_pred&#39;)

pgs_strict &lt;- list()
for(pop_i in keep_list$POP){
  pop_i_keep &lt;- model_pred[model_pred[[pop_i]] &gt; 0.95, ]
  pgs_strict[[pop_i]] &lt;- list()
  for(gwas_i in gwas_list$name){
    pgs_strict[[pop_i]][[gwas_i]] &lt;- list()
    for(pgs_method_i in pgs_methods){
      if(gwas_list$population[gwas_list$name == gwas_i] == &#39;EUR&#39; | (gwas_list$population[gwas_list$name == gwas_i] != &#39;EUR&#39; &amp; (pgs_method_i %in% pgs_methods_eur))){
        pgs_strict[[pop_i]][[gwas_i]][[pgs_method_i]] &lt;- merge(pgs[[pop_i]][[gwas_i]][[pgs_method_i]], pop_i_keep[, c(&#39;FID&#39;,&#39;IID&#39;), with = F], by = c(&#39;FID&#39;,&#39;IID&#39;))
      }
    }
  }
}

# Estimate correlation between pheno and pgs
cor &lt;- NULL
for(pop_i in names(pgs_strict)){
  for(gwas_i in names(pgs_strict[[pop_i]])){
    for(pgs_strict_method_i in names(pgs_strict[[pop_i]][[gwas_i]])){
      pgs_strict_i &lt;- pgs_strict[[pop_i]][[gwas_i]][[pgs_strict_method_i]]
      pheno_pgs_strict&lt;-merge(pheno, pgs_strict_i, by = c(&#39;FID&#39;,&#39;IID&#39;))
      
      for(model_i in names(pgs_strict_i)[-1:-2]){
        y &lt;- scale(pheno_pgs_strict$height)
        x &lt;- scale(pheno_pgs_strict[[model_i]])
        
        if(all(is.na(x))){
          next
        }
        
        coef_i &lt;- coef(summary(mod &lt;- lm(y ~ x)))
        
        tmp &lt;- data.table(
          pop = pop_i,
          gwas = gwas_i,
          pgs_method = pgs_strict_method_i,
          name = model_i,
          r = coef_i[2,1],
          se = coef_i[2,2],
          p = coef_i[2,4],
          n = nobs(mod))
      cor &lt;- rbind(cor, tmp)
      
      }
    }
  }
}

# Save the results
dir.create(&#39;/scratch/prj/oliverpainfel/Data/OpenSNP/assoc&#39;)
write.csv(
  cor,
  &#39;/scratch/prj/oliverpainfel/Data/OpenSNP/assoc/genopred-v1-strict-yengo-assoc.csv&#39;,
  row.names = F
)

# Restrict to best only
cor_subset &lt;- NULL
for(pop_i in unique(cor$pop)){
  for(gwas_i in unique(cor$gwas[cor$pop == pop_i])){
    for(pgs_method_i in unique(cor$pgs_method[cor$pop == pop_i &amp; cor$gwas == gwas_i])){
      
      # Subset relevant results
      cor_i &lt;- cor[
        cor$pop == pop_i &amp;
        cor$gwas == gwas_i &amp;
        cor$pgs_method == pgs_method_i,]
  
      # Top R
      top_i &lt;- cor_i[which(cor_i$r == max(cor_i$r, na.rm = T))[1],]
      cor_subset &lt;- rbind(cor_subset, top_i)
    }
  }
}

# Compare the results
cor_subset$Version &lt;- &#39;GenoPred V1&#39;
cor_subset$pgs_method&lt;-gsub(&#39;_&#39;, &#39;&#39;, cor_subset$pgs_method)
cor_new_subset$Version &lt;- &#39;GenoPred V2&#39;
cor_new_subset$gwas&lt;-gsub(&#39;_&#39;, &#39;&#39;, cor_new_subset$gwas)

cor_both &lt;- rbind(cor_subset, cor_new_subset)

cor_both$pop &lt;- paste0(cor_both$pop, &quot;\n N = &quot;, cor_both$n)

##
# Plot the results
##

# yengo_eur
tmp &lt;- cor_both[cor_both$gwas == &#39;yengoeur&#39;,]
y_lim &lt;- c(min(tmp$r - tmp$se), max(tmp$r + tmp$se))
                  
v1_plot &lt;-
  ggplot(cor_both[cor_both$gwas == &#39;yengoeur&#39; &amp; cor_both$Version == &#39;GenoPred V1&#39;, ], 
         aes(x = pgs_method, y = r)) +
  geom_hline(yintercept = 0) +
    geom_bar(stat = &quot;identity&quot;, position = position_dodge(), width = 0.7, fill = &#39;#F8766D&#39;) +
    geom_errorbar(
      aes(ymin = r - se, ymax = r + se),
      width = .2,
      position = position_dodge(width = 0.7)
    ) +
    labs(
      y = &quot;Correlation (SE)&quot;,
      x = &#39;PGS Method&#39;,
      title = &#39;Yengo - EUR vs OpenSNP&#39;
    ) +
    ylim(y_lim) +
    theme_half_open() +
    background_grid() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),
          plot.title = element_text(hjust = 0.5, size=12)) +
    facet_grid(pop ~ Version) +
    panel_border()

v2_plot &lt;-
  ggplot(cor_both[cor_both$gwas == &#39;yengoeur&#39; &amp; cor_both$Version == &#39;GenoPred V2&#39;, ], 
         aes(x = pgs_method, y = r)) +
  geom_hline(yintercept = 0) +
    geom_bar(stat = &quot;identity&quot;, position = position_dodge(), width = 0.7, fill = &#39;#00BFC4&#39;) +
    geom_errorbar(
      aes(ymin = r - se, ymax = r + se),
      width = .2,
      position = position_dodge(width = 0.7)
    ) +
    labs(
      y = &quot;Correlation (SE)&quot;,
      x = &#39;PGS Method&#39;,
      title = &#39;Yengo - EUR vs OpenSNP&#39;
    ) +
    ylim(y_lim) +
    theme_half_open() +
    background_grid() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),
          plot.title = element_text(hjust = 0.5, size=12)) +
    facet_grid(pop ~ Version) +
    panel_border()

png(&#39;/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/docs/Images/OpenSNP/genopred-v1-comp_strict-yengo_eur.png&#39;,
    units = &#39;px&#39;,
    width = 3000,
    height = 3000,
    res = 300)

  plot_grid(v1_plot, v2_plot, labels = NULL)
  
dev.off()

######
# Plot the same plot only using the EUR target population in OpenSNP
######

tmp &lt;- cor_both[cor_both$gwas == &#39;yengoeur&#39; &amp; cor_both$pop == &#39;EUR\n N = 653&#39;, ]
y_lim &lt;- c(min(tmp$r - tmp$se), max(tmp$r + tmp$se))

v1_plot &lt;-
  ggplot(tmp[tmp$Version == &#39;GenoPred V1&#39;, ], 
         aes(x = pgs_method, y = r)) +
    geom_hline(yintercept = 0) +
    geom_errorbar(
      aes(ymin = r - se, ymax = r + se),
      width = .2,
      position = position_dodge(width = 0.7)
    ) +
    geom_point(stat = &quot;identity&quot;, position = position_dodge(), size = 5, colour = &#39;#F8766D&#39;) +
    labs(
      y = &quot;Correlation (SE)&quot;,
      x = &#39;PGS Method&#39;,
      title = &#39;Yengo - EUR vs OpenSNP&#39;
    ) +
    ylim(y_lim) +
    theme_half_open() +
    background_grid() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),
          plot.title = element_text(hjust = 0.5, size=12)) +
    facet_grid(pop ~ Version) +
    panel_border()

v2_plot &lt;-
  ggplot(tmp[tmp$Version == &#39;GenoPred V2&#39;, ], 
         aes(x = pgs_method, y = r)) +
    geom_hline(yintercept = 0) +
    geom_errorbar(
      aes(ymin = r - se, ymax = r + se),
      width = .2,
      position = position_dodge(width = 0.7)
    ) +
    geom_point(stat = &quot;identity&quot;, position = position_dodge(), size = 5, colour = &#39;#00BFC4&#39;) +
    labs(
      y = &quot;Correlation (SE)&quot;,
      x = &#39;PGS Method&#39;,
      title = &#39;Yengo - EUR vs OpenSNP&#39;
    ) +
    ylim(y_lim) +
    theme_half_open() +
    background_grid() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),
          plot.title = element_text(hjust = 0.5, size=12)) +
    facet_grid(pop ~ Version) +
    panel_border()

png(&#39;/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/docs/Images/OpenSNP/genopred-v1-comp_strict_eur-yengo_eur.png&#39;,
    units = &#39;px&#39;,
    width = 3000,
    height = 1500,
    res = 300)

  plot_grid(v1_plot, v2_plot, labels = NULL)
  
dev.off()</code></pre>
</details>
<hr />
<details>
<summary>
Show comparison results
</summary>
<center>
<p><img src="../docs/Images/OpenSNP/genopred-v1-comp-yengo_eur.png"
alt="Out-of-the-box comparison" /> <img
src="../docs/Images/OpenSNP/genopred-v1-comp_strict-yengo_eur.png"
alt="Using same ancestry classification threshold" /></p>
<div class="figure">
<img
src="../docs/Images/OpenSNP/genopred-v1-comp_strict_eur-yengo_eur.png"
alt="" />
<p class="caption">Showing results in European OpenSNP data only</p>
</div>
</details>
<hr />
</div>
</div>
<div id="conclusion" class="section level1">
<h1>Conclusion</h1>
<p>These results are as expected. PGS associations with height in
OpenSNP are highly concordant across version of GenoPred. Small
differences occur due to a more stringent ancestry classification
threshold in GenoPred v2, typically increasing the correlation between
the PGS and observed height.</p>
</div>

<!-- footer.html -->
<hr/>
<div style="display: flex; justify-content: center; align-items: center;">
  <img src="kcl.svg" alt="Kings College London" style="height: 60px; padding-left: 5px; padding-right: 5px;">
  <img src="Wellcome_logo.png" alt="The Wellcome Trust" style="height: 60px; padding-left: 5px; padding-right: 5px;">
  <img src="nihr-logo.png" alt="National Institute for Health and Care Research" style="height: 60px; padding-left: 5px; padding-right: 5px;">
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
