<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>OpenSNP Benchmark</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />
<link rel="shortcut icon" href="Images/logo/favicon.ico">

<link rel="stylesheet" href="styles/night-mode.css" id="nightModeStylesheet">

<script>
function toggleNightMode() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (stylesheet.disabled) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
}
</script>

<label class="switch">
  <input type="checkbox" id="toggleNightMode" checked>
  <span class="slider round"></span>
</label>

<script>
document.getElementById('toggleNightMode').addEventListener('change', function() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (this.checked) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
});
</script>

<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-YR18ZB3PR3"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-YR18ZB3PR3');
</script>

<!-- Osano Cookie Consent -->
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/cookieconsent@3/build/cookieconsent.min.css" />
<script src="https://cdn.jsdelivr.net/npm/cookieconsent@3/build/cookieconsent.min.js"></script>
<script>
window.addEventListener("load", function(){
  window.cookieconsent.initialise({
    palette: {
      popup: { background: "#000" },
      button: { background: "#f1d600" }
    },
    theme: "classic",
    position: "bottom-right",
    content: {
      message: "We use a cookie for Google Analytics to understand how people use this site. This helps us improve GenoPred and demonstrate its impact. Please click 'Accept' to help us with this.",
      dismiss: "Accept",
      link: "Google's privacy info",
      href: "https://policies.google.com/technologies/partner-sites"
    },
    onStatusChange: function(status) {
      if (status === 'allow') {
        // Now load Google Analytics only if user consents
        var gtagScript = document.createElement('script');
        gtagScript.setAttribute('async', '');
        gtagScript.src = 'https://www.googletagmanager.com/gtag/js?id=G-YR18ZB3PR3';
        document.head.appendChild(gtagScript);

        window.dataLayer = window.dataLayer || [];
        function gtag(){ dataLayer.push(arguments); }
        gtag('js', new Date());
        gtag('config', 'G-YR18ZB3PR3');
      }
    }
  });
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="styles/styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img class="logo-img" src="Images/logo/Horizontal_white.png" style="height: 42px;" /></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pipeline
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="pipeline_overview.html">Overview</a>
    </li>
    <li>
      <a href="pipeline_readme.html">Instructions</a>
    </li>
    <li>
      <a href="pipeline_technical.html">Technical documentation</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Research
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="research_index.html">Overview</a>
    </li>
    <li>
      <a href="comparison_of_methods_summary.html">Polygenic Scoring Methods Comparison</a>
    </li>
    <li>
      <a href="Functionally_informed_prediction.html">Quantifying Polygenic Signal Mediated by Altered Gene Expression</a>
    </li>
    <li>
      <a href="Absolute_Conversion.html">Translating Polygenic Scores onto the Absolute Scale</a>
    </li>
    <li>
      <a href="CrossPop_summary.html">Cross-Ancestry Polygenic Prediction</a>
    </li>
  </ul>
</li>
<li>
  <a href="more_index.html">More</a>
</li>
<li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">OpenSNP Benchmark</h1>

</div>


<hr />
<p>This document will test multi-source PGS methods using height GWAS
with OpenSNP as the target sample. A previous demonstration of GenoPred
using the OpenSNP data is shown <a
href="opensnp_benchmark.html">here</a>, which shows some additional
information regarding where the OpenSNP data was downloaded.</p>
<hr />
<div id="download-the-sumstats" class="section level1">
<h1>Download the sumstats</h1>
<p>We will use height GWAS sumstats from the Yengo paper, including UKB
and for all populations.</p>
<details>
<summary>
Show code
</summary>
<pre class="bash"><code># These are from the Yengo 2022 paper
mkdir -p /users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test
wget --no-check-certificate -O /users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_eur.txt https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST90245001-GCST90246000/GCST90245992/GCST90245992_buildGRCh37.tsv
wget --no-check-certificate -O /users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_eas.txt https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST90245001-GCST90246000/GCST90245991/GCST90245991_buildGRCh37.tsv
wget --no-check-certificate -O /users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_afr.txt https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST90245001-GCST90246000/GCST90245989/GCST90245989_buildGRCh37.tsv
wget --no-check-certificate -O /users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_amr.txt https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST90245001-GCST90246000/GCST90245993/GCST90245993_buildGRCh37.tsv
wget --no-check-certificate -O /users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_sas.txt https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST90245001-GCST90246000/GCST90245994/GCST90245994_buildGRCh37.tsv
wget --no-check-certificate -O /users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_all.txt https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST90245001-GCST90246000/GCST90245990/GCST90245990_buildGRCh37.tsv</code></pre>
</details>
<hr />
</div>
<div id="create-configuration" class="section level1">
<h1>Create configuration</h1>
<p>We want to compare PGS derived using each GWAS seperately, PGS
derived using GWAS from multiple populations.</p>
<p>We will only use QuickPRS, with LEOPARD to linearly combine across
GWAS populations. Neither of these approaches require individual-level
data to tune the PGS.</p>
<details>
<summary>
Show code
</summary>
<pre class="r"><code>setwd(&#39;/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline&#39;)

# Create gwas_list
gwas_list &lt;- NULL

for(i in c(&#39;EUR&#39;,&#39;EAS&#39;,&#39;CSA&#39;,&#39;AMR&#39;,&#39;AFR&#39;)){
  gwas_list &lt;- rbind(gwas_list, data.table(
    name = paste0(&#39;yengo_&#39;, tolower(i)),
    path = paste0(
      &#39;/users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_&#39;,
      ifelse(i == &#39;CSA&#39;, &#39;sas&#39;, tolower(i)),
      &#39;.txt&#39;
    ),
    population = i,
    n = NA,
    sampling = NA,
    prevalence = NA,
    mean = NA,
    sd = NA,
    label = paste0(&#39;&quot;Yengo 2022 Height &#39;, i,&#39;\&quot;&#39;)))
}

# Insert GWAS from all, and assign EUR as the population
gwas_list &lt;- rbind(gwas_list, data.table(
  name = &#39;yengo_all&#39;,
  path = &#39;/users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_all.txt&#39;,
  population = &#39;EUR&#39;,
  n = NA,
  sampling = NA,
  prevalence = NA,
  mean = NA,
  sd = NA,
  label = &quot;\&quot;Yengo 2022 Height All\&quot;&quot;))

write.table(gwas_list, &#39;misc/opensnp/gwas_list_cross_pop.txt&#39;, col.names = T, row.names = F, quote = F, sep = &#39; &#39;)

# Create gwas_groups file
gwas_groups &lt;- NULL
gwas_groups &lt;- rbind(gwas_groups, data.frame(
  name = &#39;height_eur_eas&#39;,
  gwas = &#39;yengo_eur,yengo_eas&#39;,
  label = &quot;\&quot;Yengo 2022 Height EUR+EAS\&quot;&quot;
))

gwas_groups &lt;- rbind(gwas_groups, data.frame(
  name = &#39;height_eur_eas_afr&#39;,
  gwas = &#39;yengo_eur,yengo_eas,yengo_afr&#39;,
  label = &quot;\&quot;Yengo 2022 Height EUR+EAS+AFR\&quot;&quot;
))

gwas_groups &lt;- rbind(gwas_groups, data.frame(
  name = &#39;height_eur_eas_afr_csa&#39;,
  gwas = &#39;yengo_eur,yengo_eas,yengo_afr,yengo_csa&#39;,
  label = &quot;\&quot;Yengo 2022 Height EUR+EAS+AFR+CSA\&quot;&quot;
))

gwas_groups &lt;- rbind(gwas_groups, data.frame(
  name = &#39;height_eur_eas_afr_csa_amr&#39;,
  gwas = &#39;yengo_eur,yengo_eas,yengo_afr,yengo_csa,yengo_amr&#39;,
  label = &quot;\&quot;Yengo 2022 Height EUR+EAS+AFR+CSA+AMR\&quot;&quot;
))

write.table(gwas_groups, &#39;misc/opensnp/gwas_groups.txt&#39;, col.names = T, row.names = F, quote = F, sep = &#39; &#39;)

# Create config file
config &lt;- readLines(&#39;misc/opensnp/config.yaml&#39;)

config[grepl(&#39;^config_file:&#39;, config)]&lt;- &#39;config_file: misc/opensnp/config_cross_pop.yaml&#39;
config &lt;- config[!grepl(&#39;^score_list:&#39;, config)]
config[grepl(&#39;^outdir:&#39;, config)]&lt;- &#39;outdir: /users/k1806347/oliverpainfel/Data/OpenSNP/GenoPred/test_cross_pop_2&#39;
config[grepl(&#39;^pgs_methods:&#39;, config)]&lt;- &quot;pgs_methods: [&#39;quickprs&#39;]&quot;
config[grepl(&#39;^gwas_list:&#39;, config)]&lt;- &quot;gwas_list: misc/opensnp/gwas_list_cross_pop.txt&quot;
config&lt;-c(config, &#39;gwas_groups: misc/opensnp/gwas_groups.txt&#39;)
config&lt;-c(config, &quot;leopard_methods: [&#39;quickprs&#39;]&quot;)
config&lt;-c(config, &quot;pgs_scaling: [&#39;continuous&#39;, &#39;discrete&#39;]&quot;)

write.table(config, &#39;misc/opensnp/config_cross_pop.yaml&#39;, col.names = F, row.names = F, quote = F)</code></pre>
<div class="note-box">
<p><strong>Note:</strong> The original configfile for OpenSNP, and the
target_list that it refers to, were created in the OpenSNP benchmark
document (<a href="opensnp_benchmark.html">here</a>).</p>
</div>
</details>
<hr />
</div>
<div id="run-pipeline" class="section level1">
<h1>Run pipeline</h1>
<pre class="bash"><code>snakemake --profile slurm --use-conda --configfile=misc/opensnp/config_cross_pop.yaml output_all</code></pre>
<hr />
</div>
<div id="evaluate-pgs" class="section level1">
<h1>Evaluate PGS</h1>
<details>
<summary>
Show code
</summary>
<pre class="r"><code># Test correlation between PGS and observed height

setwd(&#39;/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline/&#39;)
library(data.table)
library(ggplot2)
library(cowplot)

source(&#39;../functions/misc.R&#39;)
source_all(&#39;../functions&#39;)

# Read in pheno data
pheno &lt;- fread(&#39;/users/k1806347/oliverpainfel/Data/OpenSNP/processed/pheno/height.txt&#39;)

# Read in PGS
pgs &lt;- read_pgs(config = &#39;misc/opensnp/config_cross_pop.yaml&#39;, name = &#39;opensnp&#39;)$opensnp

# Read in ancestry
ancestry &lt;- read_ancestry(config = &#39;misc/opensnp/config_cross_pop_gw.yaml&#39;, name = &#39;opensnp&#39;)

# Estimate correlation between pheno and pgs
cor &lt;- NULL
for(pop_i in names(pgs)){
  for(gwas_i in names(pgs[[pop_i]])){
    for(pgs_method_i in names(pgs[[pop_i]][[gwas_i]])){
      pgs_i &lt;- pgs[[pop_i]][[gwas_i]][[pgs_method_i]]
      pheno_pgs&lt;-merge(pheno, pgs_i, by = c(&#39;FID&#39;,&#39;IID&#39;))
      
      for(model_i in names(pgs_i)[-1:-2]){
        y &lt;- scale(pheno_pgs$height)
        x &lt;- scale(pheno_pgs[[model_i]])
        
        if(all(is.na(x))){
          next
        }
        
        coef_i &lt;- coef(summary(mod &lt;- lm(y ~ x)))
        
        tmp &lt;- data.table(
          pop = pop_i,
          gwas = gwas_i,
          pgs_method = pgs_method_i,
          name = model_i,
          r = coef_i[2,1],
          se = coef_i[2,2],
          p = coef_i[2,4],
          n = nobs(mod))
      cor &lt;- rbind(cor, tmp)
      
      }
    }
  }
}

library(ggplot2)
library(cowplot)
ggplot(cor, aes(x = name, y = r, group = gwas)) +
  geom_bar(stat = &quot;identity&quot;, position = position_dodge2(preserve = &quot;single&quot;), width = 0.7) +
  geom_errorbar(
    aes(ymin = r - se, ymax = r + se),
    width = .2,
    position = position_dodge(width = 0.7)
  ) +
  facet_wrap(. ~ pop, scales = &#39;free&#39;) +
  theme_half_open() +
  background_grid() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        plot.title = element_text(hjust = 0.5, size=12))

ggplot(cor[cor$pop %in% c(&#39;EUR&#39;),], aes(x = name, y = r, group = gwas)) +
  geom_bar(stat = &quot;identity&quot;, position = position_dodge2(preserve = &quot;single&quot;), width = 0.7) +
  geom_errorbar(
    aes(ymin = r - se, ymax = r + se),
    width = .2,
    position = position_dodge(width = 0.7)
  ) +
  facet_wrap(. ~ pop) +
  theme_half_open() +
  background_grid() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        plot.title = element_text(hjust = 0.5, size=12))

tmp &lt;- cor[cor$pop %in% c(&#39;EUR&#39;) &amp; grepl(&#39;yengo|EUR_weighted&#39;, cor$name) ,]
tmp$name &lt;- gsub(&#39;yengo_|_quickprs|height_|_targ.*&#39;, &#39;&#39;, tmp$name)
tmp$name &lt;- gsub(&#39;_&#39;,&#39;+&#39;,tmp$name)
tmp$name &lt;- toupper(tmp$name)
tmp$name[tmp$name == &#39;ALL&#39;] &lt;- &quot;ALL (META)&quot;
tmp$name &lt;- factor(tmp$name, levels = tmp$name[order(tmp$r)])
tmp$pop &lt;- &#39;Target = EUR&#39;

png(&#39;~/oliverpainfel/Software/MyGit/GenoPred/docs/Images/OpenSNP/cross_pop_eur.png&#39;, res = 200, width = 1500, height = 1500, units = &#39;px&#39;)
ggplot(tmp, aes(x = name, y = r, fill = gwas)) +
  geom_bar(stat = &quot;identity&quot;, position = position_dodge2(preserve = &quot;single&quot;), width = 0.7) +
  geom_errorbar(
    aes(ymin = r - se, ymax = r + se),
    width = .2,
    position = position_dodge(width = 0.7)
  ) +
  geom_text(
    aes(y = r + se + 0.05,label = round(r, 3)),
    position = position_dodge2(width = 0.7, preserve = &quot;single&quot;),
  ) +
  labs(x = &#39;GWAS populations&#39;, y = &quot;R (SE)&quot;) +
  ylim(c(0,0.6)) +
  facet_wrap(. ~ pop) +
  theme_half_open() +
  background_grid() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        plot.title = element_text(hjust = 0.5, size=12),
        legend.position = &quot;none&quot;)
dev.off()</code></pre>
</details>
<hr />
</div>
<div id="result" class="section level1">
<h1>Result</h1>
<p>EUR is the only population of sufficient sample size to accurately
estimate the performance of the PGS.</p>
<p>Observations:</p>
<ul>
<li>The PGS based on EUR is the best single-source PGS</li>
<li>The PGS based on meta-analysis of each GWAS, with a EUR LD reference
performs worse than the PGS based on EUR GWAS alone.</li>
<li>The PGS based on GWAS from multiple populations (multi-source)
improved prediction slightly over the PGS based on EUR GWAS alone
(single-source), with the prediction improving as more populations were
considered.</li>
</ul>
<div class="centered-container">
<div class="rounded-image-container" style="width: 75%;">
<div class="figure">
<img src="Images/OpenSNP/cross_pop_eur.png" alt="" />
<p class="caption">Correlation between PGS and height</p>
</div>
</div>
</div>
</div>

<!-- footer.html -->
<hr/>

<div class="centered-container">
<div class="rounded-image-container" style="width: 500px;">
<img src="Images/logo/sponsors.png">
</div>
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
