<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>OpenSNP Benchmark with dense SNP list</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />
<link rel="shortcut icon" href="Images/logo/favicon.ico">

<link rel="stylesheet" href="styles/night-mode.css" id="nightModeStylesheet">

<script>
function toggleNightMode() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (stylesheet.disabled) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
}
</script>

<label class="switch">
  <input type="checkbox" id="toggleNightMode" checked>
  <span class="slider round"></span>
</label>

<script>
document.getElementById('toggleNightMode').addEventListener('change', function() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (this.checked) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
});
</script>

<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-YR18ZB3PR3"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-YR18ZB3PR3');
</script>

<!-- Osano Cookie Consent -->
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/cookieconsent@3/build/cookieconsent.min.css" />
<script src="https://cdn.jsdelivr.net/npm/cookieconsent@3/build/cookieconsent.min.js"></script>
<script>
window.addEventListener("load", function(){
  window.cookieconsent.initialise({
    palette: {
      popup: { background: "#000" },
      button: { background: "#f1d600" }
    },
    theme: "classic",
    position: "bottom-right",
    content: {
      message: "We use a cookie for Google Analytics to understand how people use this site. This helps us improve GenoPred and demonstrate its impact. Please click 'Accept' to help us with this.",
      dismiss: "Accept",
      link: "Google's privacy info",
      href: "https://policies.google.com/technologies/partner-sites"
    },
    onStatusChange: function(status) {
      if (status === 'allow') {
        // Now load Google Analytics only if user consents
        var gtagScript = document.createElement('script');
        gtagScript.setAttribute('async', '');
        gtagScript.src = 'https://www.googletagmanager.com/gtag/js?id=G-YR18ZB3PR3';
        document.head.appendChild(gtagScript);

        window.dataLayer = window.dataLayer || [];
        function gtag(){ dataLayer.push(arguments); }
        gtag('js', new Date());
        gtag('config', 'G-YR18ZB3PR3');
      }
    }
  });
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="styles/styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img class="logo-img" src="Images/logo/Horizontal_white.png" style="height: 42px;" /></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pipeline
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="pipeline_overview.html">Overview</a>
    </li>
    <li>
      <a href="pipeline_readme.html">Instructions</a>
    </li>
    <li>
      <a href="pipeline_technical.html">Technical documentation</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Research
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="research_index.html">Overview</a>
    </li>
    <li>
      <a href="comparison_of_methods_summary.html">Polygenic Scoring Methods Comparison</a>
    </li>
    <li>
      <a href="Functionally_informed_prediction.html">Quantifying Polygenic Signal Mediated by Altered Gene Expression</a>
    </li>
    <li>
      <a href="Absolute_Conversion.html">Translating Polygenic Scores onto the Absolute Scale</a>
    </li>
    <li>
      <a href="CrossPop_summary.html">Cross-Ancestry Polygenic Prediction</a>
    </li>
  </ul>
</li>
<li>
  <a href="more_index.html">More</a>
</li>
<li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">OpenSNP Benchmark with dense SNP list</h1>

</div>


<hr />
<div id="overview" class="section level1">
<h1>Overview</h1>
<p>We have previously used OpenSNP as a dataset for testing GenoPred. An
important constraint within the default GenoPred implementation is its
restriction to HapMap3 variants. This may negatively impact the R2 of
PGS, particularly those leveraging functional genomic annotations. This
is also particularly problematic when using externally derived score
files, such as from the PGS catalogue, which may include variant
variants that are not in the HapMap3 SNPlist. To remedy this we can use
a reference dataset that includes more variant, and that is what we will
test here.</p>
<p>We will run GenoPred on the OpenSNP data using the default reference
data (restricted to HapMap3) and a denser reference (all common variants
in the 1KG+HGDP dataset - dbSNP common variant list with MAF &gt; 0.01
in at least of the 1KG populations). We will calculate PGS based on
height GWAS summary statistics, but also a range of height PGS score
files on the PGS catalogue, and see the difference in performance.</p>
<p>Note, the yengo et al 2022 GWAS of height was restricted to HapMap3
variants. To evaluate the impact of using the denser reference, we will
need to find GWAS and PGS score files that are not restricted to HapMap3
variants.</p>
<hr />
</div>
<div id="prepare-input" class="section level1">
<h1>Prepare input</h1>
<hr />
<div id="download-genotypes" class="section level2">
<h2>Download genotypes</h2>
<details>
<summary>
Show code
</summary>
<pre class="bash"><code>mkdir -p /users/k1806347/oliverpainfel/Data/OpenSNP/raw
cd /users/k1806347/oliverpainfel/Data/OpenSNP/raw
wget https://zenodo.org/records/1442755/files/CrowdAI_v3.tar.gz?download=1 
mv &#39;CrowdAI_v3.tar.gz?download=1&#39; CrowdAI_v3.tar.gz
tar -xvzf CrowdAI_v3.tar.gz
rm CrowdAI_v3.tar.gz

# Use both training and testing data to maximise sample size for testing.</code></pre>
</details>
<hr />
</div>
<div id="prepare-phenotype" class="section level2">
<h2>Prepare phenotype</h2>
<details>
<summary>
Show code
</summary>
<pre class="r"><code>library(data.table)

# Training 
train_dat &lt;- fread(&#39;/users/k1806347/oliverpainfel/Data/OpenSNP/raw/CrowdAI_v3/training_set_details.txt&#39;)
train_dat$FID &lt;- train_dat$id
train_dat$IID &lt;- train_dat$id
train_dat &lt;- train_dat[, c(&#39;FID&#39;, &#39;IID&#39;, &#39;height&#39;), with = F]

# Test 
test_dat &lt;- fread(&#39;/users/k1806347/oliverpainfel/Data/OpenSNP/raw/CrowdAI_v3/test_set_details_SECRET.txt&#39;)
test_dat$FID &lt;- test_dat$id
test_dat$IID &lt;- test_dat$id
test_dat &lt;- test_dat[, c(&#39;FID&#39;, &#39;IID&#39;, &#39;height&#39;), with = F]

all_dat&lt;-rbind(train_dat, test_dat)
dir.create(&#39;/users/k1806347/oliverpainfel/Data/OpenSNP/processed/pheno&#39;, recursive = T)
write.table(
  all_dat,
  &#39;/users/k1806347/oliverpainfel/Data/OpenSNP/processed/pheno/height.txt&#39;,
  col.names = T,
  row.names = F,
  quote = F
)</code></pre>
</details>
<hr />
</div>
<div id="split-vcf-by-chromosome" class="section level2">
<h2>Split VCF by chromosome</h2>
<details>
<summary>
Show code
</summary>
<pre class="bash"><code>module add bcftools/1.12-gcc-13.2.0-python-3.11.6

# Create index
bcftools index /users/k1806347/oliverpainfel/Data/OpenSNP/raw/CrowdAI_v3/fullset/genotyping_data_fullset_train.vcf.gz

bcftools index /users/k1806347/oliverpainfel/Data/OpenSNP/raw/CrowdAI_v3/fullset/genotyping_data_fullset_test.vcf.gz

# Merge the vcfs for training and testing data
bcftools merge \
  /users/k1806347/oliverpainfel/Data/OpenSNP/raw/CrowdAI_v3/fullset/genotyping_data_fullset_train.vcf.gz \
  /users/k1806347/oliverpainfel/Data/OpenSNP/raw/CrowdAI_v3/fullset/genotyping_data_fullset_test.vcf.gz \
  -o /users/k1806347/oliverpainfel/Data/OpenSNP/raw/CrowdAI_v3/fullset/genotyping_data_fullset_merged.vcf.gz

# Now, split by chromosome using plink2
# Run on the command line within pipeline conda environment
mkdir /users/k1806347/oliverpainfel/Data/OpenSNP/processed/geno
for chr in $(seq 1 22);do
    /users/k1806347/oliverpainfel/Software/plink2 \
        --vcf /users/k1806347/oliverpainfel/Data/OpenSNP/raw/CrowdAI_v3/fullset/genotyping_data_fullset_merged.vcf.gz \
        --chr ${chr} \
        --out /users/k1806347/oliverpainfel/Data/OpenSNP/processed/geno/opensnp_merged.chr${chr} \
        --export vcf bgz
done</code></pre>
</details>
<hr />
</div>
<div id="download-height-gwas" class="section level2">
<h2>Download height GWAS</h2>
<details>
<summary>
Show code
</summary>
<pre class="bash"><code>mkdir -p /users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test

# Yengo 2022
wget --no-check-certificate -O /users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_eur.txt https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST90245001-GCST90246000/GCST90245992/GCST90245992_buildGRCh37.tsv

# Loh 2018
wget --no-check-certificate -O /users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/loh_2018_height_eur.txt http://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST90029001-GCST90030000/GCST90029008/GCST90029008_buildGRCh37.tsv

# Verma 2024
wget --no-check-certificate -O /users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/verma_2024_height_eur.txt http://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST90475001-GCST90476000/GCST90475359/GCST90475359.tsv.gz</code></pre>
</details>
<hr />
</div>
<div id="create-gwas_list-target_list-and-config"
class="section level2">
<h2>Create gwas_list, target_list and config</h2>
<details>
<summary>
Show code
</summary>
<pre class="r"><code>setwd(&#39;/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline&#39;)
library(data.table)

# Create config file for sparse configuration
conf &lt;- c(
  &#39;outdir: /users/k1806347/oliverpainfel/Data/OpenSNP/GenoPred/sparse_test&#39;,
  &#39;config_file: misc/opensnp/config_sparse.yaml&#39;,
  &#39;gwas_list: misc/opensnp/gwas_list_dense_test.txt&#39;,
  &#39;score_list: misc/opensnp/score_list_dense_test.txt&#39;,
  &#39;target_list: misc/opensnp/target_list.txt&#39;,
  &quot;pgs_scaling: [&#39;discrete&#39;,&#39;continuous&#39;]&quot;,
  &quot;target_populations: [&#39;EUR&#39;]&quot;,
  &quot;pgs_methods: [&#39;ptclump&#39;,&#39;megaprs&#39;]&quot;
)

dir.create(&#39;misc/opensnp/&#39;, recursive = T)
dir.create(&#39;/users/k1806347/oliverpainfel/Data/OpenSNP/GenoPred&#39;, recursive = T)
write.table(conf, &#39;misc/opensnp/config_sparse.yaml&#39;, col.names = F, row.names = F, quote = F)

# Create config file for sparse configuration but restricted to target variants
conf &lt;- c(
  &#39;outdir: /users/k1806347/oliverpainfel/Data/OpenSNP/GenoPred/sparse_restricted_test&#39;,
  &#39;config_file: misc/opensnp/config_sparse_restricted.yaml&#39;,
  &#39;restrict_to_target_variants: T&#39;,
  &#39;gwas_list: misc/opensnp/gwas_list_dense_test.txt&#39;,
  &#39;score_list: misc/opensnp/score_list_dense_test.txt&#39;,
  &#39;target_list: misc/opensnp/target_list.txt&#39;,
  &quot;pgs_scaling: [&#39;discrete&#39;,&#39;continuous&#39;]&quot;,
  &quot;target_populations: [&#39;EUR&#39;]&quot;,
  &quot;pgs_methods: [&#39;ptclump&#39;,&#39;megaprs&#39;]&quot;
)

dir.create(&#39;misc/opensnp/&#39;, recursive = T)
dir.create(&#39;/users/k1806347/oliverpainfel/Data/OpenSNP/GenoPred&#39;, recursive = T)
write.table(conf, &#39;misc/opensnp/config_sparse_restricted.yaml&#39;, col.names = F, row.names = F, quote = F)

# Create config file for dense configuration
conf &lt;- c(
  &#39;outdir: /users/k1806347/oliverpainfel/Data/OpenSNP/GenoPred/dense_test&#39;,
  &#39;dense_reference: T&#39;,
  &#39;config_file: misc/opensnp/config_dense.yaml&#39;,
  &#39;gwas_list: misc/opensnp/gwas_list_dense_test.txt&#39;,
  &#39;score_list: misc/opensnp/score_list_dense_test.txt&#39;,
  &#39;target_list: misc/opensnp/target_list.txt&#39;,
  &quot;pgs_scaling: [&#39;discrete&#39;,&#39;continuous&#39;]&quot;,
  &quot;target_populations: [&#39;EUR&#39;]&quot;,
  &quot;pgs_methods: [&#39;ptclump&#39;,&#39;megaprs&#39;]&quot;
)

dir.create(&#39;misc/opensnp/&#39;, recursive = T)
dir.create(&#39;/users/k1806347/oliverpainfel/Data/OpenSNP/GenoPred&#39;, recursive = T)
write.table(conf, &#39;misc/opensnp/config_dense.yaml&#39;, col.names = F, row.names = F, quote = F)

# Create target_list
target_list &lt;- fread(&#39;example_input/target_list.txt&#39;)
target_list &lt;- rbind(target_list,
                     data.table(name = &#39;opensnp&#39;,
                                path = &#39;/users/k1806347/oliverpainfel/Data/OpenSNP/processed/geno/opensnp_merged&#39;,
                                type = &#39;vcf&#39;,
                                indiv_report = F))

target_list &lt;- target_list[target_list$name == &#39;opensnp&#39;, ]

write.table(target_list, &#39;misc/opensnp/target_list.txt&#39;, col.names = T, row.names = F, quote = F, sep = &#39; &#39;)

# Create gwas_list
gwas_list &lt;- NULL
gwas_list&lt;-rbind(gwas_list, 
                 data.table(name=&#39;yengo_eur&#39;,
                            path = &#39;/users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_eur.txt&#39;,
                            population = &#39;EUR&#39;,
                            n = NA,
                            sampling = NA,
                            prevalence = NA,
                            mean = NA,
                            sd = NA,
                            label = &quot;\&quot;Yengo 2022 Height EUR\&quot;&quot;))

gwas_list&lt;-rbind(gwas_list, 
                 data.table(name=&#39;loh_eur.txt&#39;,
                            path = &#39;/users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/loh_2018_height_eur.txt&#39;,
                            population = &#39;EUR&#39;,
                            n = NA,
                            sampling = NA,
                            prevalence = NA,
                            mean = NA,
                            sd = NA,
                            label = &quot;\&quot;Loh 2018 Height EUR\&quot;&quot;))

gwas_list&lt;-rbind(gwas_list, 
                 data.table(name=&#39;verma_eur&#39;,
                            path = &#39;/users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/verma_2024_height_eur.txt&#39;,
                            population = &#39;EUR&#39;,
                            n = NA,
                            sampling = NA,
                            prevalence = NA,
                            mean = NA,
                            sd = NA,
                            label = &quot;\&quot;Verma 2024 Height EUR\&quot;&quot;))

write.table(gwas_list, &#39;misc/opensnp/gwas_list_dense_test.txt&#39;, col.names = T, row.names = F, quote = F, sep = &#39; &#39;)

# Create score_list
score_list &lt;- data.frame(
  name = &#39;PGS002804&#39;,
  path = NA,
  label = &quot;\&quot;Yengo 2022 Height EUR PGSC\&quot;&quot;
)

score_list &lt;- rbind(score_list,
                   data.table(
                     name = &#39;PGS002972&#39;,
                      path = NA,
                      label = &quot;\&quot;ExPRSweb Height raw DBSLMM\&quot;&quot;))

score_list &lt;- rbind(score_list,
                   data.table(
                     name = &#39;PGS004781&#39;,
                      path = NA,
                      label = &quot;\&quot;PRSmixPlus Height EUR\&quot;&quot;))

score_list &lt;- rbind(score_list,
                   data.table(
                     name = &#39;PGS004996&#39;,
                      path = NA,
                      label = &quot;\&quot;Gunn 2024 Height ldpred EUR\&quot;&quot;))

write.table(score_list, &#39;misc/opensnp/score_list_dense_test.txt&#39;, col.names=T, row.names=F, quote=F, sep=&#39; &#39;)</code></pre>
</details>
<hr />
</div>
</div>
<div id="run-genopred" class="section level1">
<h1>Run GenoPred</h1>
<details>
<summary>
Show code
</summary>
<pre class="bash"><code>snakemake --profile slurm --use-conda --configfile=misc/opensnp/config_sparse.yaml output_all -n 
snakemake --profile slurm --use-conda --configfile=misc/opensnp/config_sparse_restricted.yaml output_all -n 
snakemake --profile slurm --use-conda --configfile=misc/opensnp/config_dense.yaml output_all -n </code></pre>
</details>
<hr />
<div id="evaluate-pgs" class="section level2">
<h2>Evaluate PGS</h2>
<p>Note. ~5M dense reference variants were present in OpenSNP. This
leads to the external score files that go beyond HapMap3 variants to be
dropped due to &lt;75% of variants being present. This highlights that
this issue is not just a consequence of using the default HapMap3
restricted reference data, but generally poor generalisability across
target datasets due to missing variants.</p>
<details>
<summary>
Show code
</summary>
<pre class="r"><code># Test correlation between PGS and observed height

setwd(&#39;/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline/&#39;)
library(data.table)
library(ggplot2)
library(cowplot)

source(&#39;../functions/misc.R&#39;)
source_all(&#39;../functions&#39;)

# Read in pheno data
pheno &lt;- fread(&#39;/users/k1806347/oliverpainfel/Data/OpenSNP/processed/pheno/height.txt&#39;)

# Read in PGS
pgs &lt;- list(
  sparse = read_pgs(config = &#39;misc/opensnp/config_sparse.yaml&#39;, name = &#39;opensnp&#39;)$opensnp,
  sparse_restricted = read_pgs(config = &#39;misc/opensnp/config_sparse_restricted.yaml&#39;, name = &#39;opensnp&#39;)$opensnp,
  dense = read_pgs(config = &#39;misc/opensnp/config_dense.yaml&#39;, name = &#39;opensnp&#39;)$opensnp
)

# Estimate correlation between pheno and pgs
cor &lt;- NULL
for(test_i in names(pgs)){
  for(pop_i in names(pgs[[test_i]])){
    for(gwas_i in names(pgs[[test_i]][[pop_i]])){
      for(pgs_method_i in names(pgs[[test_i]][[pop_i]][[gwas_i]])){
        pgs_i &lt;- pgs[[test_i]][[pop_i]][[gwas_i]][[pgs_method_i]]
        pheno_pgs&lt;-merge(pheno, pgs_i, by = c(&#39;FID&#39;,&#39;IID&#39;))
        
        for(model_i in names(pgs_i)[-1:-2]){
          y &lt;- scale(pheno_pgs$height)
          x &lt;- scale(pheno_pgs[[model_i]])
          
          if(all(is.na(x))){
            next
          }
          
          coef_i &lt;- coef(summary(mod &lt;- lm(y ~ x)))
          
          tmp &lt;- data.table(
            test = test_i,
            pop = pop_i,
            gwas = gwas_i,
            pgs_method = pgs_method_i,
            name = model_i,
            r = coef_i[2,1],
            se = coef_i[2,2],
            p = coef_i[2,4],
            n = nobs(mod))
        cor &lt;- rbind(cor, tmp)
        
        }
      }
    }
  }
}

# Save the results
dir.create(&#39;/scratch/prj/oliverpainfel/Data/OpenSNP/assoc&#39;)
write.csv(
  cor,
  &#39;/scratch/prj/oliverpainfel/Data/OpenSNP/assoc/genopred-dense-assoc.csv&#39;,
  row.names = F
)

# The European sample is the only one large enough for interpretable results. All other populations are &lt;20 individuals
# Subset EUR results
cor_eur &lt;- cor[cor$pop == &#39;EUR&#39;, ]

# Restrict to best and and pseudoval only
cor_eur_subset &lt;- NULL
for(test_i in unique(cor_eur$test)){
  for(pop_i in unique(cor_eur$pop)){
    for(gwas_i in unique(cor_eur$gwas[cor_eur$pop == pop_i])){
      for(pgs_method_i in unique(cor_eur$pgs_method[cor_eur$pop == pop_i &amp; cor_eur$gwas == gwas_i])){
        
        # Subset relevant results
        cor_eur_i &lt;- cor_eur[
          cor_eur$test == test_i &amp;
          cor_eur$pop == pop_i &amp;
          cor_eur$gwas == gwas_i &amp;
          cor_eur$pgs_method == pgs_method_i,]
    
        # Top R
        if(pgs_method_i %in% c(&#39;ptclump&#39;,&#39;ldpred2&#39;,&#39;megaprs&#39;,&#39;prscs&#39;,&#39;lassosum&#39;,&#39;dbslmm&#39;)){
          top_i &lt;- cor_eur_i[which(cor_eur_i$r == max(cor_eur_i$r, na.rm = T))[1],]
          top_i$model &lt;- &#39;Top&#39;
          cor_eur_subset &lt;- rbind(cor_eur_subset, top_i)
        }

        # PseudoVal
        if(pgs_method_i %in% c(&#39;ptclump&#39;,&#39;sbayesr&#39;,&#39;ldpred2&#39;,&#39;megaprs&#39;,&#39;prscs&#39;,&#39;lassosum&#39;,&#39;dbslmm&#39;,&#39;prscsx&#39;)){
          pseudo_param &lt;- find_pseudo(config = paste0(&#39;misc/opensnp/config_&#39;,test_i,&#39;.yaml&#39;), gwas = gwas_i, pgs_method = pgs_method_i)
          pseudo_i &lt;- cor_eur_i[grepl(paste0(pseudo_param,&#39;$&#39;), cor_eur_i$name),]
          pseudo_i$model &lt;- &#39;Pseudo&#39;
          if(pgs_method_i %in% c(&#39;prscsx&#39;)){
            pseudo_i$pgs_method&lt;-paste0(pseudo_i$pgs_method, &quot; (&quot;, gsub(&#39;.*_&#39;,&#39;&#39;, gsub(paste0(&#39;_&#39;,pseudo_param), &#39;&#39;, pseudo_i$name)), &quot;)&quot;)
          }
          cor_eur_subset &lt;- rbind(cor_eur_subset, pseudo_i)
        }
        
        # External
        external_tmp&lt;-cor_eur_i[cor_eur_i$pgs_method == &#39;external&#39;,]
        external_tmp$model &lt;- &#39;External&#39;
        cor_eur_subset &lt;- rbind(cor_eur_subset, external_tmp)
        
      }
    }
  }
}

# Plot the results
cor_eur_subset$model &lt;- factor(cor_eur_subset$model, levels = c(&#39;Top&#39;,&#39;Pseudo&#39;,&#39;External&#39;))
cor_eur_subset$test &lt;-
  factor(
    cor_eur_subset$test,
    levels = c(&#39;sparse&#39;, &#39;sparse_restricted&#39;, &#39;dense&#39;),
    labels = c(&#39;HapMap3&#39;, &quot;HapMap3\n(restricted)&quot;, &quot;Dense&quot;)
  )

cor_eur_subset$gwas &lt;-
  factor(
    cor_eur_subset$gwas,
    levels = c(
      &quot;loh_eur.txt&quot;,
      &quot;verma_eur&quot;,
      &quot;yengo_eur&quot;,
      &quot;PGS004996&quot;,
      &quot;PGS002804&quot;
    ),
    labels = c(&quot;Loh 2018&quot;, &quot;Verma 2024&quot;, &quot;Yengo 2022&quot;, &quot;Gunn 2024&quot;, &quot;Yengo 2022\n(incl. 23andMe)&quot;)
  )

dir.create(&#39;/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/docs/Images/OpenSNP&#39;)

plot_obj_eur &lt;- 
  ggplot(cor_eur_subset, aes(x = pgs_method, y = r, fill = model)) +
  geom_bar(stat = &quot;identity&quot;, position = position_dodge2(preserve = &quot;single&quot;), width = 0.7) +
  geom_errorbar(
    aes(ymin = r - se, ymax = r + se),
    width = .2,
    position = position_dodge(width = 0.7)
  ) +
  labs(
    y = &quot;Correlation (SE)&quot;,
    x = &#39;PGS Method&#39;,
    fill = &#39;Model&#39;
  ) +
  coord_cartesian(ylim = c(0.2, NA)) +
  theme_half_open() +
  background_grid() +
  panel_border() +
  facet_grid(gwas ~ test) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        plot.title = element_text(hjust = 0.5, size=12))

png(&#39;/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/docs/Images/OpenSNP/genopred-dense_eur.png&#39;,
    units = &#39;px&#39;,
    width = 2000,
    height = 2500,
    res = 300)

  plot_obj_eur

dev.off()</code></pre>
</details>
<hr />
</div>
</div>
<div id="results" class="section level1">
<h1>Results</h1>
<p>In summary, the performance of the PGS are very similar whether using
the HapMap3-only reference with or without restricting to variants in
the target sample, and when using the dense reference. The external
score files that are not restricted to HapMap3 variants are still
dropped due to poor overlap with the target sample variants,
highlighting that it is not just the density/coverage provided by the
reference that is the issue.</p>
<details>
<summary>
Show results in OpenSNP
</summary>
<div class="centered-container">
<div class="rounded-image-container" style="width: 60%;">
<div class="figure">
<img src="Images/OpenSNP/genopred-dense_eur.png" alt="" />
<p class="caption">Yengo - EUR PGS correlation with height in EUR
OpenSNP</p>
</div>
</div>
</div>
</details>
<details>
<summary>
Show GWAS QC info
</summary>
<table style="width:100%;">
<colgroup>
<col width="26%" />
<col width="27%" />
<col width="13%" />
<col width="17%" />
<col width="13%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">Label</th>
<th align="left">Label_GWAS</th>
<th align="left">Population</th>
<th align="right">NSNP_Original</th>
<th align="right">NSNP_Final</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Dense</td>
<td align="left">Yengo 2022 Height EUR</td>
<td align="left">EUR</td>
<td align="right">1372608</td>
<td align="right">956956</td>
</tr>
<tr class="even">
<td align="left">Dense</td>
<td align="left">Loh 2018 Height EUR</td>
<td align="left">EUR</td>
<td align="right">12007535</td>
<td align="right">4536885</td>
</tr>
<tr class="odd">
<td align="left">Dense</td>
<td align="left">Verma 2024 Height EUR</td>
<td align="left">EUR</td>
<td align="right">19718047</td>
<td align="right">4961041</td>
</tr>
<tr class="even">
<td align="left">HapMap3 (restricted)</td>
<td align="left">Yengo 2022 Height EUR</td>
<td align="left">EUR</td>
<td align="right">1372608</td>
<td align="right">943912</td>
</tr>
<tr class="odd">
<td align="left">HapMap3 (restricted)</td>
<td align="left">Loh 2018 Height EUR</td>
<td align="left">EUR</td>
<td align="right">12007535</td>
<td align="right">1073008</td>
</tr>
<tr class="even">
<td align="left">HapMap3 (restricted)</td>
<td align="left">Verma 2024 Height EUR</td>
<td align="left">EUR</td>
<td align="right">19718047</td>
<td align="right">1076986</td>
</tr>
<tr class="odd">
<td align="left">HapMap3</td>
<td align="left">Yengo 2022 Height EUR</td>
<td align="left">EUR</td>
<td align="right">1372608</td>
<td align="right">1021951</td>
</tr>
<tr class="even">
<td align="left">HapMap3</td>
<td align="left">Loh 2018 Height EUR</td>
<td align="left">EUR</td>
<td align="right">12007535</td>
<td align="right">1159849</td>
</tr>
<tr class="odd">
<td align="left">HapMap3</td>
<td align="left">Verma 2024 Height EUR</td>
<td align="left">EUR</td>
<td align="right">19718047</td>
<td align="right">1167206</td>
</tr>
</tbody>
</table>
</details>
<details>
<summary>
Show score file info
</summary>
<table>
<colgroup>
<col width="26%" />
<col width="35%" />
<col width="17%" />
<col width="13%" />
<col width="7%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">Label</th>
<th align="left">GWAS</th>
<th align="right">NSNP_Original</th>
<th align="right">NSNP_Final</th>
<th align="left">Pass</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Dense</td>
<td align="left">Yengo 2022 Height EUR PGSC</td>
<td align="right">1097307</td>
<td align="right">934335</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td align="left">Dense</td>
<td align="left">ExPRSweb Height raw DBSLMM</td>
<td align="right">12</td>
<td align="right">0</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td align="left">Dense</td>
<td align="left">PRSmixPlus Height EUR</td>
<td align="right">47507</td>
<td align="right">1</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td align="left">Dense</td>
<td align="left">Gunn 2024 Height ldpred EUR</td>
<td align="right">1285709</td>
<td align="right">1025578</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td align="left">HapMap3 (restricted)</td>
<td align="left">Yengo 2022 Height EUR PGSC</td>
<td align="right">1097307</td>
<td align="right">925857</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td align="left">HapMap3 (restricted)</td>
<td align="left">ExPRSweb Height raw DBSLMM</td>
<td align="right">12</td>
<td align="right">0</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td align="left">HapMap3 (restricted)</td>
<td align="left">PRSmixPlus Height EUR</td>
<td align="right">47507</td>
<td align="right">1</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td align="left">HapMap3 (restricted)</td>
<td align="left">Gunn 2024 Height ldpred EUR</td>
<td align="right">1285709</td>
<td align="right">1016685</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td align="left">HapMap3</td>
<td align="left">Yengo 2022 Height EUR PGSC</td>
<td align="right">1097307</td>
<td align="right">1000510</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td align="left">HapMap3</td>
<td align="left">ExPRSweb Height raw DBSLMM</td>
<td align="right">12</td>
<td align="right">0</td>
<td align="left">FALSE</td>
</tr>
<tr class="odd">
<td align="left">HapMap3</td>
<td align="left">PRSmixPlus Height EUR</td>
<td align="right">47507</td>
<td align="right">1</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td align="left">HapMap3</td>
<td align="left">Gunn 2024 Height ldpred EUR</td>
<td align="right">1285709</td>
<td align="right">1128089</td>
<td align="left">TRUE</td>
</tr>
</tbody>
</table>
</details>
</div>

<!-- footer.html -->
<hr/>

<div class="centered-container">
<div class="rounded-image-container" style="width: 500px;">
<img src="Images/logo/sponsors.png">
</div>
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
