---
title: OpenSNP Benchmark with dense SNP list
output: 
  html_document:
    theme: cosmo
    toc: true
    toc_float: true
    toc_depth: 2
    css: styles/styles.css
    includes:
      in_header: header.html
      after_body: footer.html

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(eval = FALSE)
library(knitr)
```

***

# Overview

We have previously used OpenSNP as a dataset for testing GenoPred. An important constraint within the default GenoPred implementation is its restriction to HapMap3 variants. This may negatively impact the R2 of PGS, particularly those leveraging functional genomic annotations. This is also particularly problematic when using externally derived score files, such as from the PGS catalogue, which may include variant variants that are not in the HapMap3 SNPlist. To remedy this we can use a reference dataset that includes more variant, and that is what we will test here.

We will run GenoPred on the OpenSNP data using the default reference data (restricted to HapMap3) and a denser reference (all common variants in the 1KG+HGDP dataset - dbSNP common variant list with MAF > 0.01 in at least of the 1KG populations). We will calculate PGS based on height GWAS summary statistics, but also a range of height PGS score files on the PGS catalogue, and see the difference in performance.

Note, the yengo et al 2022 GWAS of height was restricted to HapMap3 variants. To evaluate the impact of using the denser reference, we will need to find GWAS and PGS score files that are not restricted to HapMap3 variants.

***

# Prepare input

***

## Download genotypes

<details><summary>Show code</summary>

```{bash}
mkdir -p /users/k1806347/oliverpainfel/Data/OpenSNP/raw
cd /users/k1806347/oliverpainfel/Data/OpenSNP/raw
wget https://zenodo.org/records/1442755/files/CrowdAI_v3.tar.gz?download=1 
mv 'CrowdAI_v3.tar.gz?download=1' CrowdAI_v3.tar.gz
tar -xvzf CrowdAI_v3.tar.gz
rm CrowdAI_v3.tar.gz

# Use both training and testing data to maximise sample size for testing.
```

</details>

***

## Prepare phenotype

<details><summary>Show code</summary>

```{r}
library(data.table)

# Training 
train_dat <- fread('/users/k1806347/oliverpainfel/Data/OpenSNP/raw/CrowdAI_v3/training_set_details.txt')
train_dat$FID <- train_dat$id
train_dat$IID <- train_dat$id
train_dat <- train_dat[, c('FID', 'IID', 'height'), with = F]

# Test 
test_dat <- fread('/users/k1806347/oliverpainfel/Data/OpenSNP/raw/CrowdAI_v3/test_set_details_SECRET.txt')
test_dat$FID <- test_dat$id
test_dat$IID <- test_dat$id
test_dat <- test_dat[, c('FID', 'IID', 'height'), with = F]

all_dat<-rbind(train_dat, test_dat)
dir.create('/users/k1806347/oliverpainfel/Data/OpenSNP/processed/pheno', recursive = T)
write.table(
  all_dat,
  '/users/k1806347/oliverpainfel/Data/OpenSNP/processed/pheno/height.txt',
  col.names = T,
  row.names = F,
  quote = F
)
```

</details>

***

## Split VCF by chromosome

<details><summary>Show code</summary>

```{bash}
module add bcftools/1.12-gcc-13.2.0-python-3.11.6

# Create index
bcftools index /users/k1806347/oliverpainfel/Data/OpenSNP/raw/CrowdAI_v3/fullset/genotyping_data_fullset_train.vcf.gz

bcftools index /users/k1806347/oliverpainfel/Data/OpenSNP/raw/CrowdAI_v3/fullset/genotyping_data_fullset_test.vcf.gz

# Merge the vcfs for training and testing data
bcftools merge \
  /users/k1806347/oliverpainfel/Data/OpenSNP/raw/CrowdAI_v3/fullset/genotyping_data_fullset_train.vcf.gz \
  /users/k1806347/oliverpainfel/Data/OpenSNP/raw/CrowdAI_v3/fullset/genotyping_data_fullset_test.vcf.gz \
  -o /users/k1806347/oliverpainfel/Data/OpenSNP/raw/CrowdAI_v3/fullset/genotyping_data_fullset_merged.vcf.gz

# Now, split by chromosome using plink2
# Run on the command line within pipeline conda environment
mkdir /users/k1806347/oliverpainfel/Data/OpenSNP/processed/geno
for chr in $(seq 1 22);do
    /users/k1806347/oliverpainfel/Software/plink2 \
        --vcf /users/k1806347/oliverpainfel/Data/OpenSNP/raw/CrowdAI_v3/fullset/genotyping_data_fullset_merged.vcf.gz \
        --chr ${chr} \
        --out /users/k1806347/oliverpainfel/Data/OpenSNP/processed/geno/opensnp_merged.chr${chr} \
        --export vcf bgz
done
```

</details>

***

## Download height GWAS

<details><summary>Show code</summary>

```{bash}
mkdir -p /users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test

# Yengo 2022
wget --no-check-certificate -O /users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_eur.txt https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST90245001-GCST90246000/GCST90245992/GCST90245992_buildGRCh37.tsv

# Loh 2018
wget --no-check-certificate -O /users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/loh_2018_height_eur.txt http://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST90029001-GCST90030000/GCST90029008/GCST90029008_buildGRCh37.tsv

# Verma 2024
wget --no-check-certificate -O /users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/verma_2024_height_eur.txt http://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST90475001-GCST90476000/GCST90475359/GCST90475359.tsv.gz
```

</details>

***

## Create gwas_list, target_list and config

<details><summary>Show code</summary>

```{r}
setwd('/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline')
library(data.table)

# Create config file for sparse configuration
conf <- c(
  'outdir: /users/k1806347/oliverpainfel/Data/OpenSNP/GenoPred/sparse_test',
  'config_file: misc/opensnp/config_sparse.yaml',
  'gwas_list: misc/opensnp/gwas_list_dense_test.txt',
  'score_list: misc/opensnp/score_list_dense_test.txt',
  'target_list: misc/opensnp/target_list.txt',
  "pgs_scaling: ['discrete','continuous']",
  "target_populations: ['EUR']",
  "pgs_methods: ['ptclump','megaprs']"
)

dir.create('misc/opensnp/', recursive = T)
dir.create('/users/k1806347/oliverpainfel/Data/OpenSNP/GenoPred', recursive = T)
write.table(conf, 'misc/opensnp/config_sparse.yaml', col.names = F, row.names = F, quote = F)

# Create config file for sparse configuration but restricted to target variants
conf <- c(
  'outdir: /users/k1806347/oliverpainfel/Data/OpenSNP/GenoPred/sparse_restricted_test',
  'config_file: misc/opensnp/config_sparse_restricted.yaml',
  'restrict_to_target_variants: T',
  'gwas_list: misc/opensnp/gwas_list_dense_test.txt',
  'score_list: misc/opensnp/score_list_dense_test.txt',
  'target_list: misc/opensnp/target_list.txt',
  "pgs_scaling: ['discrete','continuous']",
  "target_populations: ['EUR']",
  "pgs_methods: ['ptclump','megaprs']"
)

dir.create('misc/opensnp/', recursive = T)
dir.create('/users/k1806347/oliverpainfel/Data/OpenSNP/GenoPred', recursive = T)
write.table(conf, 'misc/opensnp/config_sparse_restricted.yaml', col.names = F, row.names = F, quote = F)

# Create config file for dense configuration
conf <- c(
  'outdir: /users/k1806347/oliverpainfel/Data/OpenSNP/GenoPred/dense_test',
  'dense_reference: T',
  'config_file: misc/opensnp/config_dense.yaml',
  'gwas_list: misc/opensnp/gwas_list_dense_test.txt',
  'score_list: misc/opensnp/score_list_dense_test.txt',
  'target_list: misc/opensnp/target_list.txt',
  "pgs_scaling: ['discrete','continuous']",
  "target_populations: ['EUR']",
  "pgs_methods: ['ptclump','megaprs']"
)

dir.create('misc/opensnp/', recursive = T)
dir.create('/users/k1806347/oliverpainfel/Data/OpenSNP/GenoPred', recursive = T)
write.table(conf, 'misc/opensnp/config_dense.yaml', col.names = F, row.names = F, quote = F)

# Create config file for dense configuration including ambiguous variants
conf <- c(
  'outdir: /users/k1806347/oliverpainfel/Data/OpenSNP/GenoPred/dense_incl_ambig_test',
  'dense_reference: T',
  'keep_ambiguous: T',
  'config_file: misc/opensnp/config_dense_incl_ambig.yaml',
  'gwas_list: misc/opensnp/gwas_list_dense_test.txt',
  'score_list: misc/opensnp/score_list_dense_test.txt',
  'target_list: misc/opensnp/target_list.txt',
  "pgs_scaling: ['discrete','continuous']",
  "target_populations: ['EUR']",
  "pgs_methods: ['ptclump']"
)

dir.create('misc/opensnp/', recursive = T)
dir.create('/users/k1806347/oliverpainfel/Data/OpenSNP/GenoPred', recursive = T)
write.table(conf, 'misc/opensnp/config_dense_incl_ambig.yaml', col.names = F, row.names = F, quote = F)

# Create target_list
target_list <- fread('example_input/target_list.txt')
target_list <- rbind(target_list,
                     data.table(name = 'opensnp',
                                path = '/users/k1806347/oliverpainfel/Data/OpenSNP/processed/geno/opensnp_merged',
                                type = 'vcf',
                                indiv_report = F))

target_list <- target_list[target_list$name == 'opensnp', ]

write.table(target_list, 'misc/opensnp/target_list.txt', col.names = T, row.names = F, quote = F, sep = ' ')

# Create gwas_list
gwas_list <- NULL
gwas_list<-rbind(gwas_list, 
                 data.table(name='yengo_eur',
                            path = '/users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_eur.txt',
                            population = 'EUR',
                            n = NA,
                            sampling = NA,
                            prevalence = NA,
                            mean = NA,
                            sd = NA,
                            label = "\"Yengo 2022 Height EUR\""))

gwas_list<-rbind(gwas_list, 
                 data.table(name='loh_eur.txt',
                            path = '/users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/loh_2018_height_eur.txt',
                            population = 'EUR',
                            n = NA,
                            sampling = NA,
                            prevalence = NA,
                            mean = NA,
                            sd = NA,
                            label = "\"Loh 2018 Height EUR\""))

gwas_list<-rbind(gwas_list, 
                 data.table(name='verma_eur',
                            path = '/users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/verma_2024_height_eur.txt',
                            population = 'EUR',
                            n = NA,
                            sampling = NA,
                            prevalence = NA,
                            mean = NA,
                            sd = NA,
                            label = "\"Verma 2024 Height EUR\""))

write.table(gwas_list, 'misc/opensnp/gwas_list_dense_test.txt', col.names = T, row.names = F, quote = F, sep = ' ')

# Create score_list
score_list <- data.frame(
  name = 'PGS002804',
  path = NA,
  label = "\"Yengo 2022 Height EUR PGSC\""
)

score_list <- rbind(score_list,
                   data.table(
                     name = 'PGS002972',
                      path = NA,
                      label = "\"ExPRSweb Height raw DBSLMM\""))

score_list <- rbind(score_list,
                   data.table(
                     name = 'PGS004781',
                      path = NA,
                      label = "\"PRSmixPlus Height EUR\""))

score_list <- rbind(score_list,
                   data.table(
                     name = 'PGS004996',
                      path = NA,
                      label = "\"Gunn 2024 Height ldpred EUR\""))

write.table(score_list, 'misc/opensnp/score_list_dense_test.txt', col.names=T, row.names=F, quote=F, sep=' ')

```

</details>

***

# Run GenoPred

<details><summary>Show code</summary>

```{bash}
snakemake --profile slurm --use-conda --configfile=misc/opensnp/config_sparse.yaml output_all -n 
snakemake --profile slurm --use-conda --configfile=misc/opensnp/config_sparse_restricted.yaml output_all -n 
snakemake --profile slurm --use-conda --configfile=misc/opensnp/config_dense.yaml output_all -n 
snakemake --profile slurm --use-conda --configfile=misc/opensnp/config_dense_incl_ambig.yaml output_all -n 
```

</details>

***

## Evaluate PGS

Note. ~5M dense reference variants were present in OpenSNP. This leads to the external score files that go beyond HapMap3 variants to be dropped due to <75% of variants being present. This highlights that this issue is not just a consequence of using the default HapMap3 restricted reference data, but generally poor generalisability across target datasets due to missing variants.

<details><summary>Show code</summary>

```{r}
# Test correlation between PGS and observed height

setwd('/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline/')
library(data.table)
library(ggplot2)
library(cowplot)

source('../functions/misc.R')
source_all('../functions')

# Read in pheno data
pheno <- fread('/users/k1806347/oliverpainfel/Data/OpenSNP/processed/pheno/height.txt')

# Read in PGS
pgs <- list(
  sparse = read_pgs(config = 'misc/opensnp/config_sparse.yaml', name = 'opensnp')$opensnp,
  sparse_restricted = read_pgs(config = 'misc/opensnp/config_sparse_restricted.yaml', name = 'opensnp')$opensnp,
  dense = read_pgs(config = 'misc/opensnp/config_dense.yaml', name = 'opensnp')$opensnp
)

# Estimate correlation between pheno and pgs
cor <- NULL
for(test_i in names(pgs)){
  for(pop_i in names(pgs[[test_i]])){
    for(gwas_i in names(pgs[[test_i]][[pop_i]])){
      for(pgs_method_i in names(pgs[[test_i]][[pop_i]][[gwas_i]])){
        pgs_i <- pgs[[test_i]][[pop_i]][[gwas_i]][[pgs_method_i]]
        pheno_pgs<-merge(pheno, pgs_i, by = c('FID','IID'))
        
        for(model_i in names(pgs_i)[-1:-2]){
          y <- scale(pheno_pgs$height)
          x <- scale(pheno_pgs[[model_i]])
          
          if(all(is.na(x))){
            next
          }
          
          coef_i <- coef(summary(mod <- lm(y ~ x)))
          
          tmp <- data.table(
            test = test_i,
            pop = pop_i,
            gwas = gwas_i,
            pgs_method = pgs_method_i,
            name = model_i,
            r = coef_i[2,1],
            se = coef_i[2,2],
            p = coef_i[2,4],
            n = nobs(mod))
        cor <- rbind(cor, tmp)
        
        }
      }
    }
  }
}

# Save the results
dir.create('/scratch/prj/oliverpainfel/Data/OpenSNP/assoc')
write.csv(
  cor,
  '/scratch/prj/oliverpainfel/Data/OpenSNP/assoc/genopred-dense-assoc.csv',
  row.names = F
)

# The European sample is the only one large enough for interpretable results. All other populations are <20 individuals
# Subset EUR results
cor_eur <- cor[cor$pop == 'EUR', ]

# Restrict to best and and pseudoval only
cor_eur_subset <- NULL
for(test_i in unique(cor_eur$test)){
  for(pop_i in unique(cor_eur$pop)){
    for(gwas_i in unique(cor_eur$gwas[cor_eur$pop == pop_i])){
      for(pgs_method_i in unique(cor_eur$pgs_method[cor_eur$pop == pop_i & cor_eur$gwas == gwas_i])){
        
        # Subset relevant results
        cor_eur_i <- cor_eur[
          cor_eur$test == test_i &
          cor_eur$pop == pop_i &
          cor_eur$gwas == gwas_i &
          cor_eur$pgs_method == pgs_method_i,]
    
        # Top R
        if(pgs_method_i %in% c('ptclump','ldpred2','megaprs','prscs','lassosum','dbslmm')){
          top_i <- cor_eur_i[which(cor_eur_i$r == max(cor_eur_i$r, na.rm = T))[1],]
          top_i$model <- 'Top'
          cor_eur_subset <- rbind(cor_eur_subset, top_i)
        }

        # PseudoVal
        if(pgs_method_i %in% c('ptclump','sbayesr','ldpred2','megaprs','prscs','lassosum','dbslmm','prscsx')){
          pseudo_param <- find_pseudo(config = paste0('misc/opensnp/config_',test_i,'.yaml'), gwas = gwas_i, pgs_method = pgs_method_i)
          pseudo_i <- cor_eur_i[grepl(paste0(pseudo_param,'$'), cor_eur_i$name),]
          pseudo_i$model <- 'Pseudo'
          if(pgs_method_i %in% c('prscsx')){
            pseudo_i$pgs_method<-paste0(pseudo_i$pgs_method, " (", gsub('.*_','', gsub(paste0('_',pseudo_param), '', pseudo_i$name)), ")")
          }
          cor_eur_subset <- rbind(cor_eur_subset, pseudo_i)
        }
        
        # External
        external_tmp<-cor_eur_i[cor_eur_i$pgs_method == 'external',]
        external_tmp$model <- 'External'
        cor_eur_subset <- rbind(cor_eur_subset, external_tmp)
        
      }
    }
  }
}

# Plot the results
cor_eur_subset$model <- factor(cor_eur_subset$model, levels = c('Top','Pseudo','External'))
cor_eur_subset$test <-
  factor(
    cor_eur_subset$test,
    levels = c('sparse', 'sparse_restricted', 'dense'),
    labels = c('HapMap3', "HapMap3\n(restricted)", "Dense")
  )

cor_eur_subset$gwas <-
  factor(
    cor_eur_subset$gwas,
    levels = c(
      "loh_eur.txt",
      "verma_eur",
      "yengo_eur",
      "PGS004996",
      "PGS002804"
    ),
    labels = c("Loh 2018", "Verma 2024", "Yengo 2022", "Gunn 2024", "Yengo 2022\n(incl. 23andMe)")
  )

dir.create('/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/docs/Images/OpenSNP')

plot_obj_eur <- 
  ggplot(cor_eur_subset, aes(x = pgs_method, y = r, fill = model)) +
  geom_bar(stat = "identity", position = position_dodge2(preserve = "single"), width = 0.7) +
  geom_errorbar(
    aes(ymin = r - se, ymax = r + se),
    width = .2,
    position = position_dodge(width = 0.7)
  ) +
  labs(
    y = "Correlation (SE)",
    x = 'PGS Method',
    fill = 'Model'
  ) +
  coord_cartesian(ylim = c(0.2, NA)) +
  theme_half_open() +
  background_grid() +
  panel_border() +
  facet_grid(gwas ~ test) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        plot.title = element_text(hjust = 0.5, size=12))

png('/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/docs/Images/OpenSNP/genopred-dense_eur.png',
    units = 'px',
    width = 2000,
    height = 2500,
    res = 300)

  plot_obj_eur

dev.off()

```

</details>

***

# Results

In summary, the performance of the PGS are very similar whether using the HapMap3-only reference with or without restricting to variants in the target sample, and when using the dense reference. The external score files that are not restricted to HapMap3 variants are still dropped due to poor overlap with the target sample variants, highlighting that it is not just the density/coverage provided by the reference that is the issue.

<details><summary>Show results in OpenSNP</summary>

<div class="centered-container">
<div class="rounded-image-container" style="width: 60%;">
![Yengo - EUR PGS correlation with height in EUR OpenSNP ](Images/OpenSNP/genopred-dense_eur.png)
</div>
</div>

</details>

<details><summary>Show GWAS QC info</summary>

```{r, eval = T, echo = F}
gwas_data <- data.frame(
  Label = c("Dense", "Dense", "Dense",
            "HapMap3 (restricted)", "HapMap3 (restricted)", "HapMap3 (restricted)",
            "HapMap3", "HapMap3", "HapMap3"),
  Label_GWAS = c("Yengo 2022 Height EUR", "Loh 2018 Height EUR", "Verma 2024 Height EUR",
                 "Yengo 2022 Height EUR", "Loh 2018 Height EUR", "Verma 2024 Height EUR",
                 "Yengo 2022 Height EUR", "Loh 2018 Height EUR", "Verma 2024 Height EUR"),
  Population = rep("EUR", 9),
  NSNP_Original = c(1372608, 12007535, 19718047,
                    1372608, 12007535, 19718047,
                    1372608, 12007535, 19718047),
  NSNP_Final = c(956956, 4536885, 4961041,
                 943912, 1073008, 1076986,
                 1021951, 1159849, 1167206)
)

kable(gwas_data, 'markdown')
```

</details>

<details><summary>Show score file info</summary>

```{r, eval = T, echo = F}

score_data <- data.frame(
  Label = c(
    "Dense", "Dense", "Dense", "Dense",
    "HapMap3 (restricted)", "HapMap3 (restricted)", "HapMap3 (restricted)", "HapMap3 (restricted)",
    "HapMap3", "HapMap3", "HapMap3", "HapMap3"
  ),
  GWAS = c(
    "Yengo 2022 Height EUR PGSC",
    "ExPRSweb Height raw DBSLMM",
    "PRSmixPlus Height EUR",
    "Gunn 2024 Height ldpred EUR",
    "Yengo 2022 Height EUR PGSC",
    "ExPRSweb Height raw DBSLMM",
    "PRSmixPlus Height EUR",
    "Gunn 2024 Height ldpred EUR",
    "Yengo 2022 Height EUR PGSC",
    "ExPRSweb Height raw DBSLMM",
    "PRSmixPlus Height EUR",
    "Gunn 2024 Height ldpred EUR"
  ),
  NSNP_Original = c(
    1097307, 12, 47507, 1285709,
    1097307, 12, 47507, 1285709,
    1097307, 12, 47507, 1285709
  ),
  NSNP_Final = c(
    934335, 0, 1, 1025578,
    925857, 0, 1, 1016685,
    1000510, 0, 1, 1128089
  ),
  Pass = c(
    TRUE, FALSE, FALSE, TRUE,
    TRUE, FALSE, FALSE, TRUE,
    TRUE, FALSE, FALSE, TRUE
  )
)

kable(score_data, 'markdown')
```

</details>

***

# pgsc_calc


For comparison, try calculating the scores from the PGS catalogue using the PGS Catalogue tool, [pgsc_calc](https://pgsc-calc.readthedocs.io/en/latest/).

***

## Install

<details><summary>Show code</summary>

```{bash}
# Install nextflow (already on CREATE)
module load nextflow/24.10.2-gcc-13.2.0

# Install Singularity (already on CREATE)

# Run the pgsc_calc test profile
NXF_SINGULARITY_CACHEDIR=/users/k1806347/oliverpainfel/tmp
nextflow run pgscatalog/pgsc_calc -profile test,singularity

```

</details>

***

## Download reference

<details><summary>Show code</summary>

```{bash}

mkdir ~/oliverpainfel/Data/pgsc_calc
cd ~/oliverpainfel/Data/pgsc_calc
wget https://ftp.ebi.ac.uk/pub/databases/spot/pgs/resources/pgsc_HGDP+1kGP_v1.tar.zst

```

</details>

***

## Set up samplesheet

First, you need to describe the structure of your genomic data in a standardised way. To do this, set up a spreadsheet that looks like:

<details><summary>Show code</summary>

```{r}

samplesheet <- data.frame(
  sampleset = 'opensnp',
  path_prefix = paste0('/scratch/prj/oliverpainfel/Data/OpenSNP/processed/geno/opensnp_merged.chr', 1:22),
  chrom = 1:22,
  format = 'vcf'
)

dir.create('~/oliverpainfel/Data/OpenSNP/pgsc_calc')
write.csv(samplesheet, '~/oliverpainfel/Data/OpenSNP/pgsc_calc/samplesheet.csv', row.names=F, quote=F)

```

</details>

***

## Run the pipeline

<details><summary>Show code</summary>

```{bash}

nextflow run pgscatalog/pgsc_calc \
    -profile singularity \
    --input ~/oliverpainfel/Data/OpenSNP/pgsc_calc/samplesheet.csv \
    --target_build GRCh37 \
    --pgs_id PGS002804,PGS002972,PGS004781,PGS004996 \
    --run_ancestry ~/oliverpainfel/Data/pgsc_calc/pgsc_HGDP+1kGP_v1.tar.zst \
    --outdir ~/oliverpainfel/Data/OpenSNP/pgsc_calc/output \
    --min_overlap 0
    
```

</details>

Looking through the output, the results are very similar to those of GenoPred. Slightly more variants were matched when using pgsc_calc, not suprisingly due to denser reference. It took quite a while to run, considering it is only using precomputed score files, again probably due to the more dense reference. We should check the correlation between the PGS from GenoPred and pgsc_calc for the scores in common. All in all, I am feeling quite good about GenoPred in comparison. Also, reinforces that the issue of missing variants from external scores is not specific to GenoPred. 

<details><summary>Show code</summary>

```{r}
library(data.table)

pgsc<-fread('~/oliverpainfel/Data/OpenSNP/pgsc_calc/output/results/opensnp/score/opensnp_pgs.txt.gz')

# Create object similar to output of GenoPred
pgsc_list <- split(pgsc, pgsc$PGS)

# Merge with GenoPred scores and check correlation.
setwd('/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline/')
library(data.table)
library(ggplot2)
library(cowplot)

source('../functions/misc.R')
source_all('../functions')

# Read in PGS
genopred_pgs <- read_pgs(config = 'misc/opensnp/config_dense.yaml', name = 'opensnp')$opensnp


###
# Combine PGS from genopred and pgsc
###

both_list<-list()
for(i in c('PGS004996', 'PGS002804')){
  genopred_i<- genopred_pgs$EUR[[i]]$external
  names(genopred_i)<-c('FID','IID',paste0('genopred_',i))
  genopred_i$FID<-as.numeric(genopred_i$FID)
  genopred_i$IID<-as.numeric(genopred_i$IID)
  pgsc_i<- pgsc_list[[paste0(i,'_hmPOS_GRCh37')]]
  pgsc_i<-pgsc_i[, c('FID','IID','Z_MostSimilarPop'), with=F]
  names(pgsc_i)<-c('FID','IID',paste0('pgsc_',i))
  pgsc_i$FID<-as.numeric(pgsc_i$FID)
  pgsc_i$IID<-as.numeric(pgsc_i$IID)

  both_list[[i]] <- merge(genopred_i, pgsc_i, by = c('FID','IID'))
}

both_df = Reduce(function(...) merge(..., by = c('FID','IID')), both_list)

cor(both_df[,-1:-2])

#                    genopred_PGS004996 pgsc_PGS004996 genopred_PGS002804 pgsc_PGS002804
# genopred_PGS004996          1.0000000      0.9998863          0.6910604      0.6898466
# pgsc_PGS004996              0.9998863      1.0000000          0.6911306      0.6902942
# genopred_PGS002804          0.6910604      0.6911306          1.0000000      0.9992698
# pgsc_PGS002804              0.6898466      0.6902942          0.9992698      1.0000000

```

</details>

The correlation between PGS from PGSC and GenoPred are >/= 0.999, as expected. 
