---
title: Evaluation of PGS-hub
output: 
  html_document:
    theme: cosmo
    toc: true
    toc_float: true
    toc_depth: 2
    css: styles/styles.css
    includes:
      in_header: header.html
      after_body: footer.html

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(eval = FALSE)
```

```{css, echo=F}
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
```

***

A website called PGS-hub was recently developed for running PGS methods. I am curious to see how it works.

Create some GWAS summary statistics in the required format.

```{r}
# CHR, SNP, POSITION, A1, A2, A1_FREQ, BETA, SE, P, N, N_EFF. The file is tab-separated.
# Lets use the height GWAS from yengo EUR
# We have downloaded these before
# wget --no-check-certificate -O /users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_eur.txt https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST90245001-GCST90246000/GCST90245992/GCST90245992_buildGRCh37.tsv

library(data.table)
ss <- fread('/users/k1806347/oliverpainfel/Data/GWAS_sumstats/opensnp_test/yengo_2022_height_eur.txt')
ss <- ss[, c('chromosome','variant_id','base_pair_location','effect_allele','other_allele','effect_allele_frequency','beta','standard_error','p_value','n'), with = F]
names(ss) <- c('CHR', 'SNP', 'POSITION', 'A1', 'A2', 'A1_FREQ', 'BETA', 'SE', 'P', 'N')
ss$N_EFF <- ss$N

dir.create('~/oliverpainfel/Analyses/pgs_hub/input', recursive = T)
fwrite(ss, '~/oliverpainfel/Analyses/pgs_hub/input/yengo_2022_height_eur.txt.gz', sep = '\t', na = 'NA', quote = F)

ss$SNP <- paste(ss$CHR, ss$POSITION, ss$A1, ss$A2, sep=':')
fwrite(ss, '~/oliverpainfel/Analyses/pgs_hub/input/yengo_2022_height_eur.snp_edit.txt.gz', sep = '\t', na = 'NA', quote = F)

```
