<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>GenoPred Pipeline - Reference preparation</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />
<link rel="shortcut icon" href="Images/logo/favicon.ico">

<link rel="stylesheet" href="styles/night-mode.css" id="nightModeStylesheet">

<script>
function toggleNightMode() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (stylesheet.disabled) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
}
</script>

<label class="switch">
  <input type="checkbox" id="toggleNightMode" checked>
  <span class="slider round"></span>
</label>

<script>
document.getElementById('toggleNightMode').addEventListener('change', function() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (this.checked) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
});
</script>

<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-YR18ZB3PR3"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-YR18ZB3PR3');
</script>


<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="styles/styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img class="logo-img" src="Images/logo/Horizontal_white.png" style="height: 42px;" /></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pipeline
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="pipeline_overview.html">Overview</a>
    </li>
    <li>
      <a href="pipeline_readme.html">Instructions</a>
    </li>
    <li>
      <a href="pipeline_technical.html">Technical documentation</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Research
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="research_index.html">Overview</a>
    </li>
    <li>
      <a href="comparison_of_methods_summary.html">Polygenic Scoring Methods Comparison</a>
    </li>
    <li>
      <a href="Functionally_informed_prediction.html">Quantifying Polygenic Signal Mediated by Altered Gene Expression</a>
    </li>
    <li>
      <a href="Absolute_Conversion.html">Translating Polygenic Scores onto the Absolute Scale</a>
    </li>
  </ul>
</li>
<li>
  <a href="more_index.html">More</a>
</li>
<li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">GenoPred Pipeline - Reference
preparation</h1>

</div>


<hr />
<p>In this script we will prepare reference data for GenoPred based on a
merged version of the 1KG and HGDP datasets.</p>
<p>We are going to start with the plink2 format version of these data
created by the PGS catalogue team: <a
href="https://ftp.ebi.ac.uk/pub/databases/spot/pgs/resources/pgsc_HGDP+1kGP_v1.tar.zst"
class="uri">https://ftp.ebi.ac.uk/pub/databases/spot/pgs/resources/pgsc_HGDP+1kGP_v1.tar.zst</a></p>
<p>PGS catalogue team used the data downloaded from gnomAD (v3.1.2): <a
href="https://gnomad.broadinstitute.org/downloads#v3"
class="uri">https://gnomad.broadinstitute.org/downloads#v3</a></p>
<p>The original gnomAD publication for these data is here: <a
href="https://doi.org/10.1101%2F2023.01.23.525248"
class="uri">https://doi.org/10.1101%2F2023.01.23.525248</a></p>
<p>We are going to:</p>
<ul>
<li>Extract HapMap3 variants</li>
<li>Convert the data to plink1 format</li>
<li>Insert RSIDs</li>
<li>Remove related individuals</li>
<li>Split by chromosome</li>
<li>Create keep files for each population</li>
<li>Create freq files for each population</li>
<li>Prepare .rds files summarising SNP information</li>
</ul>
<hr />
<div id="download-and-decompress" class="section level1">
<h1>Download and decompress</h1>
<pre class="bash"><code># 1KG+HGDP data
cd /users/k1806347/oliverpainfel/Data/hgdp_1kg
wget https://ftp.ebi.ac.uk/pub/databases/spot/pgs/resources/pgsc_HGDP+1kGP_v1.tar.zst
tar --use-compress-program=unzstd -xvf pgsc_HGDP+1kGP_v1.tar.zst
unzstd GRCh37_HGDP+1kGP_ALL.pvar.zst

# Sample meta data
wget https://storage.googleapis.com/gcp-public-data--gnomad/release/3.1.2/vcf/genomes/gnomad.genomes.v3.1.2.hgdp_1kg_subset_sample_meta.tsv.bgz
gunzip -c gnomad.genomes.v3.1.2.hgdp_1kg_subset_sample_meta.tsv.bgz
</code></pre>
<hr />
</div>
<div id="format-and-output-required-reference-files"
class="section level1">
<h1>Format and output required reference files</h1>
<pre class="r"><code>setwd(&#39;/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline&#39;)
source(&#39;../functions/misc.R&#39;)
source_all(&#39;../functions&#39;)
library(data.table)
library(GenoUtils)

# Read in variant data for 1KG+HGDP
pvar &lt;- fread(&#39;/users/k1806347/oliverpainfel/Data/hgdp_1kg/GRCh37_HGDP+1kGP_ALL.pvar&#39;, nThread=5)

# Remove unneeded data and variant that aren&#39;t SNPs
pvar$INFO&lt;-NULL
pvar$FILTER&lt;-NULL
pvar&lt;-pvar[nchar(pvar$REF) == 1,]
pvar&lt;-pvar[nchar(pvar$ALT) == 1,]
names(pvar)&lt;-c(&#39;CHR&#39;,&#39;BP&#39;,&#39;SNP&#39;,&#39;A1&#39;,&#39;A2&#39;)
pvar&lt;-pvar[pvar$CHR %in% 1:22,]
pvar$CHR&lt;-as.numeric(pvar$CHR)

# Read in the hapmap snp data
# Use the previously prepared 1KG reference data
hm3&lt;-NULL
for(i in 1:22){
  hm3&lt;-rbind(hm3, readRDS(paste0(&#39;/users/k1806347/oliverpainfel/test/GenoPred/pipeline/resources/data/ref/ref.chr&#39;,i,&#39;.rds&#39;)))
}

hm3&lt;-hm3[,c(&#39;CHR&#39;,&#39;SNP&#39;,&#39;BP_GRCh36&#39;,&#39;BP_GRCh37&#39;,&#39;BP_GRCh38&#39;,&#39;A1&#39;,&#39;A2&#39;), with=F]
names(hm3) &lt;- c(&#39;CHR&#39;,&#39;SNP&#39;,&#39;BP_GRCh36&#39;,&#39;BP&#39;,&#39;BP_GRCh38&#39;,&#39;A1&#39;,&#39;A2&#39;)

# Merge by chromosome and BP data
pvar_hm3&lt;-merge(pvar, hm3, by=c(&#39;CHR&#39;,&#39;BP&#39;))

# Check alleles match
pvar_hm3$IUPAC.x&lt;-snp_iupac(pvar_hm3$A1.x, pvar_hm3$A2.x)
pvar_hm3$IUPAC.y&lt;-snp_iupac(pvar_hm3$A1.y, pvar_hm3$A2.y)
flip &lt;- detect_strand_flip(pvar_hm3$IUPAC.x, pvar_hm3$IUPAC.y)
sum(flip) # 0

# Retain variants that match IUPAC
pvar_hm3&lt;-pvar_hm3[pvar_hm3$IUPAC.x == pvar_hm3$IUPAC.y,]

# Save intermediate file
write.table(pvar_hm3,&#39;/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3_snps.txt&#39;, col.names = T, row.names = F, quote = F)

# Extract hm3 SNPs from plink2 files
dir.create(&#39;/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred&#39;)
write.table(pvar_hm3$SNP.x, &#39;/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/extract.snplist&#39;, col.names=F, row.names=F, quote=F)
system(&#39;plink2 --pfile /users/k1806347/oliverpainfel/Data/hgdp_1kg/GRCh37_HGDP+1kGP_ALL --make-pgen --extract /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/extract.snplist --out /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3 --threads 5&#39;)

# Insert RSIDs into new plink files
pvar &lt;- fread(&#39;/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3.pvar&#39;)
pvar &lt;- pvar[, c(&#39;#CHROM&#39;,&#39;POS&#39;,&#39;ID&#39;,&#39;REF&#39;,&#39;ALT&#39;), with=F]
pvar_hm3$ID &lt;- pvar_hm3$SNP.x
pvar[pvar_hm3, on=.(ID), SNP := i.SNP.y]
pvar&lt;-pvar[, c(&#39;#CHROM&#39;,&#39;POS&#39;,&#39;SNP&#39;,&#39;REF&#39;,&#39;ALT&#39;), with=F]
names(pvar)&lt;-c(&#39;#CHROM&#39;,&#39;POS&#39;,&#39;ID&#39;,&#39;REF&#39;,&#39;ALT&#39;)
fwrite(pvar, &#39;/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3.pvar&#39;, col.names=T, row.names=F, quote=F, sep=&#39; &#39;)

# Remove hard filtered individuals
samp_meta&lt;-fread(&#39;/users/k1806347/oliverpainfel/Data/hgdp_1kg/gnomad.genomes.v3.1.2.hgdp_1kg_subset_sample_meta.tsv.bgz&#39;)
samp_meta&lt;-samp_meta[grepl(&quot;hard_filtered\&quot;:false&quot;, samp_meta$gnomad_sample_filters), ]

# Remove PCA outliers
samp_meta&lt;-samp_meta[grepl(&quot;outlier\&quot;:false&quot;, samp_meta$hgdp_tgp_meta), ]

# Remove relatives (we will use the sample meta-data for this - Estimated using PC-relate)
library(jsonlite)
kin_dat &lt;- NULL
for(i in 1:nrow(samp_meta)){
  tmp &lt;- fromJSON(samp_meta$relatedness_inference[i])
  if(is.data.frame(tmp$related_samples)){
    tmp$related_samples$ID&lt;-samp_meta$s[i]
    kin_dat&lt;-rbind(kin_dat, tmp$related_samples)
  }
}

# Restrict table to individuals in the genetic data
psam &lt;- fread(&#39;/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3.psam&#39;)
psam&lt;-psam[psam$`#IID` %in% samp_meta$s,]
kin_dat&lt;-kin_dat[kin_dat$s %in% psam$`#IID` &amp; kin_dat$ID %in% psam$`#IID`, ]
relations&lt;-data.frame(id_1=kin_dat$s, id_2=kin_dat$ID, kin=kin_dat$kin)

# Use GreedyRelated to find maximum unrelated set
relations$Pair &lt;- apply(relations, 1, function(row) {
  # Construct a unique identifier for each pair regardless of order
  paste(sort(c(row[&#39;id_1&#39;], row[&#39;id_2&#39;])), collapse = &quot;_&quot;)
})
relations$Pair&lt;-as.numeric(factor(relations$Pair))
relations$Factor&lt;-relations$kin
relations$ID&lt;-relations$id_1
relations&lt;-relations[, c(&#39;ID&#39;, &#39;Pair&#39;, &#39;Factor&#39;)]
relations&lt;-relations[order(relations$Pair), ]
fwrite(relations, file=&#39;/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/relations.txt&#39;, sep=&#39; &#39;, quote=F)

system(&#39;/scratch/prj/ukbiobank/KCL_Data/Software/tools/GreedyRelated-master-v1.2.1/GreedyRelated -r /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/relations.txt -t 0.05 -s 1 &gt; /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/relations_remove.txt&#39;)

remove&lt;-fread(&#39;/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/relations_remove.txt&#39;)

# Create keep file of unrelaed individuals
psam_unrel &lt;- psam[!(psam$`#IID` %in% remove$V2),]
nrow(psam_unrel) # 3313 - This is 65 less than in the preprint Table S3. Maybe Greedy related is worse?

write.table(psam_unrel[, &#39;#IID&#39;, with=F], &#39;/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/unrel.keep&#39;, col.names=T, row.names=F, quote=F)

# Split by chromosome and retain unrelated individuals
for(i in 1:22){
  system(paste0(&#39;plink2 --pfile /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3 --keep /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/unrel.keep --chr &#39;,i,&#39; --make-pgen --out /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3.chr&#39;,i))
}

# Create keep files and calculate allele frequencies for each population
# Population data is stored withint the psam file
psam &lt;- fread(&#39;/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3.chr22.psam&#39;)
pop_dat&lt;-psam[, c(&#39;#IID&#39;,&#39;SuperPop&#39;)]
names(pop_dat)&lt;-c(&#39;#IID&#39;,&#39;POP&#39;)
write.table(pop_dat, &#39;/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/pop.txt&#39;, row.names=F, col.names=T, quote=F)

dir.create(&#39;/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/keep_files/&#39;)
for(pop in unique(pop_dat$POP)){
  write.table(pop_dat[pop_dat$POP == pop, &#39;#IID&#39;, with=F], paste0(&#39;/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/keep_files/&#39;,pop,&#39;.keep&#39;), col.names=F, row.names=F, quote=F)
  dir.create(paste0(&#39;/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/freq_files/&#39;, pop), recursive = T)
  for(i in 1:22){
    system(paste0(&#39;plink2 --pfile /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3.chr&#39;,i,&#39; --keep /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/keep_files/&#39;,pop,&#39;.keep --chr &#39;,i,&#39; --freq --out /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/freq_files/&#39;, pop,&#39;/hgdp_1kg_hm3.chr&#39;, i))
  }
}

# Create frequency files across all reference individuals (TRANS)
dir.create(paste0(&#39;/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/freq_files/TRANS&#39;), recursive = T)
for(i in 1:22){
  system(paste0(&#39;plink2 --pfile /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3.chr&#39;,i,&#39; --chr &#39;,i,&#39; --freq --out /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/freq_files/TRANS/hgdp_1kg_hm3.chr&#39;, i))
}

####
# Create .rds files for SNP stats
####

pvar_hm3&lt;-pvar_hm3[,c(&#39;CHR&#39;,&#39;BP&#39;,&#39;SNP.y&#39;,&#39;A1.x&#39;,&#39;A2.x&#39;,&#39;BP_GRCh36&#39;,&#39;BP_GRCh38&#39;), with=F]
names(pvar_hm3)&lt;-c(&#39;#CHROM&#39;,&#39;POS_GRCh37&#39;,&#39;ID&#39;,&#39;ALT&#39;,&#39;REF&#39;,&#39;POS_GRCh36&#39;,&#39;POS_GRCh38&#39;)

for(chr in 1:22){
    ref&lt;-pvar_hm3[pvar_hm3$`#CHROM` == chr,]
    ref$IUPAC&lt;-snp_iupac(ref$ALT, ref$REF)
    for(pop in unique(pop_dat$POP)){
        # Read in reference frequency data
        freq&lt;-fread(paste0(&#39;/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/freq_files/&#39;, pop,&#39;/hgdp_1kg_hm3.chr&#39;, chr,&#39;.afreq&#39;))

        # The freq files have come from the reference files, so we can assume they are on the same strand
        freq_match&lt;-merge(ref, freq[,c(&#39;ID&#39;,&#39;ALT&#39;,&#39;REF&#39;,&#39;ALT_FREQS&#39;), with=F], by=c(&#39;ID&#39;,&#39;ALT&#39;,&#39;REF&#39;))
        freq_swap&lt;-merge(ref, freq[,c(&#39;ID&#39;,&#39;ALT&#39;,&#39;REF&#39;,&#39;ALT_FREQS&#39;), with=F], by.x=c(&#39;ID&#39;,&#39;ALT&#39;,&#39;REF&#39;), by.y=c(&#39;ID&#39;,&#39;REF&#39;,&#39;ALT&#39;))
        freq_swap$ALT_FREQS&lt;-1-freq_swap$ALT_FREQS
        tmp_freq&lt;-rbind(freq_match, freq_swap)
        tmp_freq&lt;-tmp_freq[match(ref$ID, tmp_freq$ID),]

        ref[[paste0(&#39;REF.FRQ.&#39;,pop)]]&lt;-tmp_freq$ALT_FREQS
    }

    ref&lt;-ref[,c(&quot;#CHROM&quot;,&quot;ID&quot;,&quot;POS_GRCh36&quot;,&quot;POS_GRCh37&quot;,&quot;POS_GRCh38&quot;,&quot;ALT&quot;,&quot;REF&quot;,&quot;IUPAC&quot;,paste0(&#39;REF.FRQ.&#39;,unique(pop_dat$POP))), with=F]
    names(ref)&lt;-c(&quot;CHR&quot;,&quot;SNP&quot;,&quot;BP_GRCh36&quot;,&quot;BP_GRCh37&quot;,&quot;BP_GRCh38&quot;,&quot;A1&quot;,&quot;A2&quot;,&quot;IUPAC&quot;,paste0(&#39;REF.FRQ.&#39;,unique(pop_dat$POP)))
    saveRDS(ref, file = paste0(&#39;/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3.chr&#39;,chr,&#39;.rds&#39;))
}

# Create keeplist
keep_list&lt;-data.frame(POP=unique(pop_dat$POP),
                      PATH=paste0(&#39;resources/data/ref/keep_files/&#39;,unique(pop_dat$POP),&#39;.keep&#39;))

write.table(keep_list, &#39;/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref.keep.list&#39;, col.names=F, row.names=F, quote=F)</code></pre>
<hr />
</div>
<div id="package-for-zenodo" class="section level1">
<h1>Package for zenodo</h1>
<p><ref_dir> ├── ref.chr&lt;1-22&gt;.<pgen/pvar/psam> ├──
ref.chr&lt;1-22&gt;.rds ├── ref.pop.txt (#IID, POP - with header) ├──
ref.keep.list (POP and PATH - without header) ├── keep_files │
└──<POP>.keep (#IID - with header) └── freq_files └──<POP>
└──ref.<POP>.chr<CHR>.afreq # PLINK2 .afreq format</p>
<pre class="bash"><code>
# Copy over all the relavent files
mkdir /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref

for chr in $(seq 1 22); do
  for file in $(echo pgen pvar psam rds); do
    cp /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3.chr${chr}.${file} /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref/ref.chr${chr}.${file}
  done
done

cp /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/pop.txt /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref/ref.pop.txt

mkdir /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref/keep_files

cp /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/keep_files/*.keep /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref/keep_files/

cp -r /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/freq_files /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref/

for pop in $(ls /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref/freq_files); do
 for chr in $(seq 1 22); do
  mv /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref/freq_files/${pop}/hgdp_1kg_hm3.chr${chr}.afreq /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref/freq_files/${pop}/ref.${pop}.chr${chr}.afreq
  rm /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref/freq_files/${pop}/hgdp_1kg_hm3.chr${chr}.log
 done
done

cp /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref.keep.list /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref/ref.keep.list

# Compress folder and upload to zenodo
cd /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred
tar -czvf genopred_1kg_hgdp.tar.gz ref
</code></pre>
<hr />
</div>
<div id="create-subset-of-pan-ukb-ld-scores" class="section level1">
<h1>Create subset of Pan UKB LD scores</h1>
<p>More information on the data: <a
href="https://pan.ukbb.broadinstitute.org/downloads/index.html"
class="uri">https://pan.ukbb.broadinstitute.org/downloads/index.html</a>
Original citation: Pan-UKB team. <a
href="https://pan.ukbb.broadinstitute.org"
class="uri">https://pan.ukbb.broadinstitute.org</a>. 2020.</p>
<pre class="bash"><code>
mkdir -p /users/k1806347/oliverpainfel/Data/ld_scores
cd /users/k1806347/oliverpainfel/Data/ld_scores
wget --no-check-certificate https://pan-ukb-us-east-1.s3.amazonaws.com/ld_release/UKBB.ALL.ldscore.tar.gz
tar -xvf UKBB.ALL.ldscore.tar.gz
rm UKBB.ALL.ldscore.tar.gz
for population in $(echo EUR EAS CSA AMR MID AFR); do
  mv UKBB.ALL.ldscore/UKBB.$population.l2.M_5_50 UKBB.$population.rsid.l2.M_5_50
  mv UKBB.ALL.ldscore/UKBB.$population.rsid.l2.ldscore.gz UKBB.$population.rsid.l2.ldscore.gz
done
rm -r UKBB.ALL.ldscore
cd ..
tar -czvf ld_scores.tar.gz ld_scores
</code></pre>
</div>

<!-- footer.html -->
<hr/>

<div class="centered-container">
<div class="rounded-image-container" style="width: 500px;">
<img src="Images/logo/sponsors.png">
</div>
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
