<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>GenoPred Pipeline - Benchmark</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />
<link rel="shortcut icon" href="Images/logo/favicon.ico">

<link rel="stylesheet" href="styles/night-mode.css" id="nightModeStylesheet">

<script>
function toggleNightMode() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (stylesheet.disabled) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
}
</script>

<label class="switch">
  <input type="checkbox" id="toggleNightMode" checked>
  <span class="slider round"></span>
</label>

<script>
document.getElementById('toggleNightMode').addEventListener('change', function() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (this.checked) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
});
</script>

<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-YR18ZB3PR3"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-YR18ZB3PR3');
</script>


<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="styles/styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img class="logo-img" src="Images/logo/Horizontal_white.png" style="height: 42px;" /></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pipeline
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="pipeline_overview.html">Overview</a>
    </li>
    <li>
      <a href="pipeline_readme.html">Instructions</a>
    </li>
    <li>
      <a href="pipeline_technical.html">Technical documentation</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Research
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="research_index.html">Overview</a>
    </li>
    <li>
      <a href="comparison_of_methods_summary.html">Polygenic Scoring Methods Comparison</a>
    </li>
    <li>
      <a href="Functionally_informed_prediction.html">Quantifying Polygenic Signal Mediated by Altered Gene Expression</a>
    </li>
    <li>
      <a href="Absolute_Conversion.html">Translating Polygenic Scores onto the Absolute Scale</a>
    </li>
  </ul>
</li>
<li>
  <a href="more_index.html">More</a>
</li>
<li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">GenoPred Pipeline - Benchmark</h1>

</div>


<hr />
<p>In this document we will benchmark the time taken and max memory for
each part of the GenoPred pipeline. We are using the benchmark results
from the snakemake benchmark functionality. We will record the time
taken to run the pipeline restricted to chromosome 22 to save time, and
then results can be extrapolated.</p>
<p>We will test the pipeline using 10 GWAS, 10 external scoring files, 3
target samples with sample sizes 100, 1000 and 10000, with 10M SNPs
originally. The number of SNPs in external scoring files can vary a lot,
so we will use scoring files based on methods restricted to hapmap3
variants.</p>
<hr />
<div id="prepare-input-data" class="section level1">
<h1>Prepare input data</h1>
<pre class="bash"><code># Create directory for inputs
mkdir -p /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data</code></pre>
<hr />
<div id="target-data" class="section level2">
<h2>Target data</h2>
<details>
<summary>
Show code
</summary>
<pre class="r"><code># Read in previous population probabilities
library(data.table)
ukb_pop &lt;- fread(&#39;/scratch/prj/ukbiobank/usr/ollie_pain/ReQC/defining_ancestry/UKBB.Ancestry.model_pred&#39;)

# Insert project specific IDs
fam &lt;-fread(&#39;/scratch/prj/ukbiobank/ukb82087/genotyped/ukb82087_binary_pre_qc.fam&#39;)
fam$row&lt;-1:nrow(fam)
ukb_pop &lt;- merge(ukb_pop, fam[,c(&#39;V1&#39;,&#39;row&#39;), with=F], by.x=&#39;FID&#39;,by.y=&#39;row&#39;)

# Identify subset with EUR probability &gt; 0.995
ukb_pop_eur &lt;- ukb_pop[ukb_pop$EUR &gt; 0.995,]
ukb_pop_eur &lt;- data.table(FID = ukb_pop_eur$V1,
                          IID = ukb_pop_eur$V1)

# Remove withdrawals
psam &lt;- fread(&#39;/scratch/prj/ukbiobank/ukb82087/imputed/ukb82087_imp_chr22_MAF1_INFO4_v1.psam&#39;)
names(psam)[1:2]&lt;-c(&#39;FID&#39;,&#39;IID&#39;)
ukb_pop_eur &lt;- ukb_pop_eur[ukb_pop_eur$FID %in% psam$FID,]

# Save keep file listing 100, 1000, and 10000 individuals
write.table(ukb_pop_eur[1:100,], &#39;/users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb_eur_subset_1.keep&#39;, col.names = F, row.names = F, quote = F)
write.table(ukb_pop_eur[1:1000,], &#39;/users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb_eur_subset_2.keep&#39;, col.names = F, row.names = F, quote = F)
write.table(ukb_pop_eur[1:10000,], &#39;/users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb_eur_subset_3.keep&#39;, col.names = F, row.names = F, quote = F)</code></pre>
<pre class="bash"><code># Create symlinks to the pgen files resitricted to INFO 0.4 and MAF 0.01
ln -s /scratch/prj/ukbiobank/ukb82087/imputed/ukb82087_imp_chr22_MAF1_INFO4_v1.psam /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.chr22.psam
ln -s /scratch/prj/ukbiobank/ukb82087/imputed/ukb82087_imp_chr22_MAF1_INFO4_v1.bgen.bgi /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.chr22.bgen.bgi 
ln -s /scratch/prj/ukbiobank/ukb82087/imputed/ukb82087_imp_chr22_MAF1_INFO4_v1.pvar /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.chr22.pvar
ln -s /scratch/prj/ukbiobank/ukb82087/imputed/ukb82087_imp_chr22_MAF1_INFO4_v1.pgen /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.chr22.pgen
ln -s /scratch/prj/ukbiobank/ukb82087/imputed/ukb82087_imp_chr22_MAF1_INFO4_v1.sample /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.chr22.sample

# Subset the plink2 format files
plink2 --pfile /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.chr22 \
--keep /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb_eur_subset_3.keep \
--make-pgen \
--threads 1 \
--out /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.subset_3.chr22

plink2 --pfile /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.subset_3.chr22 \
--keep /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb_eur_subset_2.keep \
--make-pgen \
--threads 1 \
--out /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.subset_2.chr22

plink2 --pfile /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.subset_2.chr22 \
--keep /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb_eur_subset_1.keep \
--make-pgen \
--threads 1 \
--out /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.subset_1.chr22

# Convert to bgen format
plink2 --pfile /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.subset_3.chr22 \
--keep /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb_eur_subset_3.keep \
--export bgen-1.3 \
--threads 1 \
--out /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.subset_3.chr22

plink2 --pfile /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.subset_2.chr22 \
--keep /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb_eur_subset_2.keep \
--export bgen-1.3 \
--threads 1 \
--out /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.subset_2.chr22

plink2 --pfile /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.subset_1.chr22 \
--keep /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb_eur_subset_1.keep \
--export bgen-1.3 \
--threads 1 \
--out /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.subset_1.chr22

# Index the bgen files
/users/k1806347/oliverpainfel/Software/QCTOOL/bgen.tgz/build/apps/bgenix -index -g /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.subset_1.chr22.bgen

/users/k1806347/oliverpainfel/Software/QCTOOL/bgen.tgz/build/apps/bgenix -index -g /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.subset_2.chr22.bgen

/users/k1806347/oliverpainfel/Software/QCTOOL/bgen.tgz/build/apps/bgenix -index -g /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.subset_3.chr22.bgen

# Rename sample file
mv /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.subset_1.chr22.sample /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.subset_1.sample
mv /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.subset_2.chr22.sample /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.subset_2.sample
mv /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.subset_3.chr22.sample /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.subset_3.sample</code></pre>
<pre class="r"><code># Create a target_list
target_list &lt;- data.frame(
  name = c(&#39;subset_1_plink2&#39;, &#39;subset_2_plink2&#39;, &#39;subset_3_plink2&#39;, &#39;subset_1_bgen&#39;, &#39;subset_2_bgen&#39;, &#39;subset_3_bgen&#39;),
  path = c(&#39;/users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.subset_1&#39;, &#39;/users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.subset_2&#39;, &#39;/users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/ukb.subset_3&#39;),
  type = c(&#39;plink2&#39;,&#39;plink2&#39;,&#39;plink2&#39;,&#39;bgen&#39;,&#39;bgen&#39;,&#39;bgen&#39;),
  indiv_report = c(T, F, F, F, F, F)
)

write.table(target_list, &#39;/users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/target_list.txt&#39;, col.names=T, row.names=F, quote=F)</code></pre>
</details>
<hr />
</div>
<div id="gwas-data" class="section level2">
<h2>GWAS data</h2>
<details>
<summary>
Show code
</summary>
<pre class="r"><code># Create GWAS list
gwas_list &lt;-data.frame(name = c(&#39;DEPR06&#39;,&#39;COLL01&#39;,&#39;BODY04&#39;,&#39;HEIG03&#39;,&#39;DIAB05&#39;,&#39;COAD01&#39;,&#39;SCLE03&#39;,&#39;RHEU02&#39;,&#39;BRCA01&#39;,&#39;PRCA01&#39;),
                       path = NA,
                       population = &#39;EUR&#39;,
                       n = NA,
                       sampling = c(0.2698322, NA, NA, NA, 0.1675544, 0.3298934, 0.3599529, 0.2463970, 0.5371324, 0.5643190),
                       prevalence = c(0.15, NA, NA, NA, 0.05, 0.03, 0.00164, 0.005, 0.125, 0.125),
                       mean = c(NA,0,0,0,NA,NA,NA,NA,NA,NA),
                       sd = c(NA,1,1,1,NA,NA,NA,NA,NA,NA),
                       label = c(&#39;Major Depression&#39;,&#39;Intelligence&#39;,&#39;BMI&#39;,&#39;Height&#39;,&#39;T2D&#39;,&#39;CAD&#39;,&#39;MultiScler&#39;,&#39;RheuArth&#39;,&#39;Breast Cancer&#39;,&#39;Prostate Cancer&#39;))

gwas_list$path &lt;-paste0(&#39;/scratch/prj/gwas_sumstats/cleaned/&#39;, gwas_list$name,&#39;.gz&#39;)
gwas_list$label&lt;-paste0(&quot;\&quot;&quot;, gwas_list$label, &quot;\&quot;&quot;)
write.table(gwas_list, &#39;/users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/gwas_list.txt&#39;, col.names=T, row.names=F, quote=F)</code></pre>
</details>
<hr />
</div>
<div id="external-score-file" class="section level2">
<h2>External score file</h2>
<details>
<summary>
Show code
</summary>
<pre class="r"><code># Create a score list
score_list &lt;- data.frame(name=c(&#39;PGS003980&#39;,&#39;PGS003981&#39;,&#39;PGS003982&#39;,&#39;PGS003983&#39;,&#39;PGS003984&#39;),
                         path=NA,
                         label=c(&#39;body mass index&#39;,&#39;inflammatory bowel disease&#39;,&#39;type 2 diabetes mellitus&#39;,&#39;breast carcinoma&#39;,&#39;stroke&#39;))
score_list$label&lt;-paste0(&quot;\&quot;&quot;, score_list$label, &quot;\&quot;&quot;)
write.table(score_list, &#39;/users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/score_list.txt&#39;, col.names=T, row.names=F, quote=F)</code></pre>
</details>
<hr />
</div>
<div id="configfile" class="section level2">
<h2>Configfile</h2>
<details>
<summary>
Show code
</summary>
<pre class="r"><code>config &lt;- c(
  &quot;outdir: /users/k1806347/oliverpainfel/test/genopred_benchmark/output/ncores_1&quot;,
  &quot;config_file: /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/config_ncores1.yaml&quot;,
  &quot;gwas_list: /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/gwas_list.txt&quot;,
  &quot;target_list: /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/target_list.txt&quot;,
  &quot;score_list: /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/score_list.txt&quot;,
  &quot;pgs_methods: [&#39;ptclump&#39;,&#39;dbslmm&#39;,&#39;prscs&#39;,&#39;sbayesr&#39;,&#39;lassosum&#39;,&#39;ldpred2&#39;,&#39;megaprs&#39;]&quot;,
  &quot;testing: chr22&quot;,
  &quot;cores_prep_pgs: 1&quot;,
  &quot;cores_target_pgs: 1&quot;,
  &quot;cores_impute_23andme: 1&quot;,
  &quot;cores_outlier_detection: 1&quot;
)
write.table(config, &#39;/users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/config_ncores1.yaml&#39;, col.names=F, row.names=F, quote=F)

config &lt;- c(
  &quot;outdir: /users/k1806347/oliverpainfel/test/genopred_benchmark/output/ncores_5&quot;,
  &quot;config_file: /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/config_ncores5.yaml&quot;,
  &quot;gwas_list: /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/gwas_list.txt&quot;,
  &quot;target_list: /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/target_list.txt&quot;,
  &quot;score_list: /users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/score_list.txt&quot;,
  &quot;pgs_methods: [&#39;ptclump&#39;,&#39;dbslmm&#39;,&#39;prscs&#39;,&#39;sbayesr&#39;,&#39;lassosum&#39;,&#39;ldpred2&#39;,&#39;megaprs&#39;]&quot;,
  &quot;testing: chr22&quot;,
  &quot;cores_prep_pgs: 5&quot;,
  &quot;cores_target_pgs: 5&quot;,
  &quot;cores_impute_23andme: 5&quot;,
  &quot;cores_outlier_detection: 5&quot;
)

write.table(config, &#39;/users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/config_ncores5.yaml&#39;, col.names=F, row.names=F, quote=F)</code></pre>
</details>
<hr />
</div>
</div>
<div id="run-pipeline" class="section level1">
<h1>Run pipeline</h1>
<pre class="bash"><code>git describe --tags 
#v2.2.5-13-g3771778

# N cores = 1
snakemake --profile slurm -n --configfile=/users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/config_ncores1.yaml --use-conda output_all pc_projection outlier_detection

# N cores = 5
snakemake --profile slurm -n --configfile=/users/k1806347/oliverpainfel/test/genopred_benchmark/input_data/config_ncores5.yaml --use-conda output_all pc_projection outlier_detection
  </code></pre>
<hr />
</div>
<div id="collate-results" class="section level1">
<h1>Collate results</h1>
<details>
<summary>
Show code
</summary>
<pre class="r"><code>library(data.table)
library(ggplot2)
library(cowplot)

runs &lt;- c(&#39;ncores_1&#39;,&#39;ncores_5&#39;)

bm_dat_all &lt;- NULL
for (run_i in runs) {
  # Identify benchmark files
  bm_files_i &lt;-
    paste0(
      &#39;/users/k1806347/oliverpainfel/test/genopred_benchmark/output/&#39;,
      run_i,
      &#39;/reference/benchmarks/&#39;,
      list.files(
        paste0(
          &#39;/users/k1806347/oliverpainfel/test/genopred_benchmark/output/&#39;,
          run_i,
          &#39;/reference/benchmarks&#39;
        )
      )
    )
  
  # Read in benchmark files
  bm_dat_i &lt;- do.call(rbind, lapply(bm_files_i, function(file) {
    tmp &lt;- fread(file)
    tmp$file &lt;- basename(file)
    return(tmp)
  }))
  
  bm_dat_i$ncores &lt;- as.numeric(gsub(&#39;ncores_&#39;,&#39;&#39;,run_i))
  
  bm_dat_all&lt;-rbind(bm_dat_all, bm_dat_i)
  
}

# Read in benchmark data that is configuration independent
bm_files_i &lt;-
  paste0(
    &#39;resources/data/benchmarks/&#39;,
    list.files(&#39;resources/data/benchmarks&#39;)
  )

# Read in benchmark files
bm_dat_i &lt;- do.call(rbind, lapply(bm_files_i, function(file) {
  tmp &lt;- fread(file)
  tmp$file &lt;- basename(file)
  return(tmp)
}))

bm_dat_i$ncores &lt;- 1

bm_dat_all&lt;-rbind(bm_dat_all, bm_dat_i)

# Create rule, n, pgs_method and type columns
bm_dat_all$rule &lt;- gsub(&#39;-.*&#39;,&#39;&#39;,bm_dat_all$file)
bm_dat_all$n &lt;- NA
bm_dat_all$type &lt;- NA
bm_dat_all$pgs_method &lt;- NA

bm_dat_all$n[grepl(&#39;subset_1&#39;, bm_dat_all$file)] &lt;- 100
bm_dat_all$n[grepl(&#39;subset_2&#39;, bm_dat_all$file)] &lt;- 1000
bm_dat_all$n[grepl(&#39;subset_3&#39;, bm_dat_all$file)] &lt;- 10000

bm_dat_all$type[grepl(&#39;bgen&#39;, bm_dat_all$file)] &lt;- &#39;bgen&#39;
bm_dat_all$type[grepl(&#39;plink2&#39;, bm_dat_all$file)] &lt;- &#39;plink2&#39;

bm_dat_all$pgs_method[grepl(&#39;ptclump&#39;, bm_dat_all$file)] &lt;- &#39;ptclump&#39;
bm_dat_all$pgs_method[grepl(&#39;dbslmm&#39;, bm_dat_all$file)] &lt;- &#39;dbslmm&#39;
bm_dat_all$pgs_method[grepl(&#39;ldpred2&#39;, bm_dat_all$file)] &lt;- &#39;ldpred2&#39;
bm_dat_all$pgs_method[grepl(&#39;sbayesr&#39;, bm_dat_all$file)] &lt;- &#39;sbayesr&#39;
bm_dat_all$pgs_method[grepl(&#39;lassosum&#39;, bm_dat_all$file)] &lt;- &#39;lassosum&#39;
bm_dat_all$pgs_method[grepl(&#39;prscs&#39;, bm_dat_all$file)] &lt;- &#39;prscs&#39;
bm_dat_all$pgs_method[grepl(&#39;megaprs&#39;, bm_dat_all$file)] &lt;- &#39;megaprs&#39;
bm_dat_all$pgs_method[grepl(&#39;external&#39;, bm_dat_all$file)] &lt;- &#39;external&#39;

# Calculate mean time and memory for each rule
bm_dat_summary &lt;- NULL

####
# Split by ncores
####

for (rule_i in c(
  &#39;prep_pgs_ptclump_i&#39;,
  &#39;prep_pgs_dbslmm_i&#39;,
  &#39;prep_pgs_prscs_i&#39;,
  &#39;prep_pgs_megaprs_i&#39;,
  &#39;prep_pgs_lassosum_i&#39;,
  &#39;prep_pgs_ldpred2_i&#39;,
  &#39;prep_pgs_sbayesr_i&#39;
)) {
  bm_dat_all_tmp &lt;-
    bm_dat_all[bm_dat_all$rule == rule_i, ]

    for (ncores_i in unique(bm_dat_all_tmp$ncores)){
      bm_dat_all_tmp_2 &lt;-
        bm_dat_all_tmp[bm_dat_all_tmp$ncores == ncores_i, ]
      
      tmp &lt;- data.table(
        ncores = ncores_i,
        rule = rule_i,
        type = &#39;any&#39;,
        n = &#39;any&#39;,
        pgs_method = &#39;any&#39;,
        mean_time = mean(bm_dat_all_tmp_2$s),
        max_time = max(bm_dat_all_tmp_2$s),
        mean_mem = mean(bm_dat_all_tmp_2$max_rss),
        max_mem = max(bm_dat_all_tmp_2$max_rss)
      )
      bm_dat_summary &lt;- rbind(bm_dat_summary, tmp)
    }
}

####
# Split by ncores and n
####

# Sum across populations for each type

for (rule_i in c(
  &#39;target_pgs_i&#39;
)) {
  bm_dat_all_tmp &lt;-
    bm_dat_all[bm_dat_all$rule == rule_i, ]

    for (n_i in unique(bm_dat_all_tmp$n)) {
      for (ncores_i in unique(bm_dat_all_tmp$ncores)){
        bm_dat_all_tmp_2 &lt;-
          bm_dat_all_tmp[bm_dat_all_tmp$ncores == ncores_i &amp; bm_dat_all_tmp$n == n_i, ]
        for (type_i in unique(bm_dat_all_tmp$type)) {
          bm_dat_all_tmp_2$s[bm_dat_all_tmp_2$type == type_i]&lt;-sum(bm_dat_all_tmp_2$s[bm_dat_all_tmp_2$type == type_i])
        }
        
        tmp &lt;- data.table(
          ncores = ncores_i,
          rule = rule_i,
          type = &#39;any&#39;,
          n = n_i,
          pgs_method = &#39;all&#39;,
          mean_time = mean(bm_dat_all_tmp_2$s),
          max_time = max(bm_dat_all_tmp_2$s),
          mean_mem = mean(bm_dat_all_tmp_2$max_rss),
          max_mem = max(bm_dat_all_tmp_2$max_rss)
        )
        bm_dat_summary &lt;- rbind(bm_dat_summary, tmp)
      }
    }
  }

####
# no split
####

for (rule_i in c(
  &#39;ref_pca_i&#39;,
  &#39;sumstat_prep_i&#39;,
  &#39;download_pgs_external&#39;,
  &#39;prep_pgs_external_i&#39;,
  &#39;score_reporter.txt&#39;,
  &#39;indiv_report_i&#39;
)) {
  bm_dat_all_tmp &lt;-
    bm_dat_all[bm_dat_all$rule == rule_i, ]

  tmp &lt;- data.table(
    ncores = &#39;any&#39;,
    rule = rule_i,
    type = &#39;any&#39;,
    n = &#39;any&#39;,
    pgs_method = &#39;any&#39;,
    mean_time = mean(bm_dat_all_tmp$s),
    max_time = max(bm_dat_all_tmp$s),
    mean_mem = mean(bm_dat_all_tmp$max_rss),
    max_mem = max(bm_dat_all_tmp$max_rss)
  )
  bm_dat_summary &lt;- rbind(bm_dat_summary, tmp)
}

####
# split by n and type
####

rule_i &lt;- &#39;format_target_i&#39;
bm_dat_all_tmp &lt;-
  bm_dat_all[bm_dat_all$rule == rule_i, ]
for (n_i in unique(bm_dat_all_tmp$n)) {
  for (type_i in unique(bm_dat_all_tmp$type)) {
    bm_dat_all_tmp2 &lt;-
      bm_dat_all_tmp[
        bm_dat_all_tmp$n == n_i &amp; 
        bm_dat_all_tmp$type == type_i, ]
    
    tmp &lt;- data.table(
      ncores = &#39;any&#39;,
      rule = rule_i,
      type = type_i,
      n = n_i,
      pgs_method = &#39;any&#39;,
      mean_time = mean(bm_dat_all_tmp2$s),
      max_time = max(bm_dat_all_tmp2$s),
      mean_mem = mean(bm_dat_all_tmp2$max_rss),
      max_mem = max(bm_dat_all_tmp2$max_rss)
    )
    bm_dat_summary &lt;- rbind(bm_dat_summary, tmp)
  }
}

####
# split by n
####

for (rule_i in c(&#39;ancestry_inference_i&#39;,&#39;ancestry_reporter&#39;,&#39;pc_projection_i&#39;,&#39;sample_report_i&#39;,&#39;outlier_detection_i&#39;)) {
  bm_dat_all_tmp &lt;-
    bm_dat_all[bm_dat_all$rule == rule_i, ]
  for (n_i in unique(bm_dat_all_tmp$n)) {
    bm_dat_all_tmp2 &lt;-
      bm_dat_all_tmp[
        bm_dat_all_tmp$n == n_i, ]
    
    tmp &lt;- data.table(
      ncores = &#39;any&#39;,
      rule = rule_i,
      type = &#39;any&#39;,
      n = n_i,
      pgs_method = &#39;any&#39;,
      mean_time = mean(bm_dat_all_tmp2$s),
      max_time = max(bm_dat_all_tmp2$s),
      mean_mem = mean(bm_dat_all_tmp2$max_rss),
      max_mem = max(bm_dat_all_tmp2$max_rss)
    )
    bm_dat_summary &lt;- rbind(bm_dat_summary, tmp)
  }
}

#################
# Plot
#################
####
# split by ncores
####

bm_dat_summary_i &lt;- bm_dat_summary[bm_dat_summary$rule %in% c(
    &#39;prep_pgs_ptclump_i&#39;,
    &#39;prep_pgs_dbslmm_i&#39;,
    &#39;prep_pgs_prscs_i&#39;,
    &#39;prep_pgs_megaprs_i&#39;,
    &#39;prep_pgs_lassosum_i&#39;,
    &#39;prep_pgs_ldpred2_i&#39;,
    &#39;prep_pgs_sbayesr_i&#39;
  ),]

bm_dat_summary_i$method &lt;- gsub(&#39;_i&#39;, &#39;&#39;, gsub(&#39;prep_pgs_&#39;,&#39;&#39;,bm_dat_summary_i$rule))
bm_dat_summary_i&lt;-merge(bm_dat_summary_i, pgs_method_labels, by=&#39;method&#39;)

pgs_method_time &lt;-
  ggplot(bm_dat_summary_i, aes(x = label, y = mean_time, fill = ncores)) +
  geom_bar(stat = &quot;identity&quot;, position = &quot;dodge&quot;) +
  labs(x = &quot;PGS Method&quot;, y = &quot;Time (seconds)&quot;, fill=&#39;Cores&#39;) +
  theme_half_open() +
  background_grid() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

pgs_method_mem &lt;-
  ggplot(bm_dat_summary_i, aes(x = label, y = max_mem, fill = ncores)) +
  geom_bar(stat = &quot;identity&quot;, position=&quot;dodge&quot;) +
  labs(x = &quot;PGS Method&quot;, y = &quot;Max Memory (Mb)&quot;, fill=&#39;Cores&#39;) +
  theme_half_open() +
  background_grid() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

png(&#39;Images/Benchmark/pgs_methods.png&#39;, res = 300, width = 1800, height = 1800, units = &#39;px&#39;)
  plot_grid(pgs_method_time, pgs_method_mem, nrow=2)
dev.off()

####
# split by ncores and n
####

bm_dat_summary_i &lt;- bm_dat_summary[bm_dat_summary$rule %in% c(
    &#39;target_pgs_i&#39;
  ),]

target_pgs_time &lt;-
  ggplot(bm_dat_summary_i, aes(x = n, y = mean_time, fill = ncores)) +
  geom_bar(stat = &quot;identity&quot;, position = &quot;dodge&quot;) +
  labs(x = &quot;N&quot;, y = &quot;Time (seconds)&quot;, fill=&#39;Cores&#39;) +
  theme_half_open() +
  background_grid() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

target_pgs_mem &lt;-
  ggplot(bm_dat_summary_i, aes(x = n, y = max_mem, fill = ncores)) +
  geom_bar(stat = &quot;identity&quot;, position=&quot;dodge&quot;) +
  labs(x = &quot;N&quot;, y = &quot;Max Memory (Mb)&quot;, fill=&#39;Cores&#39;) +
  theme_half_open() +
  background_grid() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

png(&#39;Images/Benchmark/target_pgs.png&#39;, res = 300, width = 1800, height = 1800, units = &#39;px&#39;)
  plot_grid(target_pgs_time, target_pgs_mem, nrow=2)
dev.off()

####
# no split
####

bm_dat_summary_i &lt;- bm_dat_summary[bm_dat_summary$rule %in% c(
    &#39;download_pgs_external&#39;,
    &#39;sumstat_prep_i&#39;,
    &#39;prep_pgs_external_i&#39;,
    &#39;ref_pca_i&#39;,
    &#39;indiv_report_i&#39;
  ),]

no_split_time &lt;- 
  ggplot(bm_dat_summary_i, aes(x = rule, y = mean_time, fill = rule)) +
  geom_bar(stat = &quot;identity&quot;, position=&quot;dodge&quot;) +
  labs(x = &quot;Rule&quot;, y = &quot;Time (seconds)&quot;) +
  theme_half_open() +
  background_grid() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1), legend.position=&quot;none&quot;)

no_split_mem &lt;- 
  ggplot(bm_dat_summary_i, aes(x = rule, y = max_mem, fill = rule)) +
  geom_bar(stat = &quot;identity&quot;, position=&quot;dodge&quot;) +
  labs(x = &quot;Rule&quot;, y = &quot;Max Memory (Mb)&quot;) +
  theme_half_open() +
  background_grid() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1), legend.position=&quot;none&quot;)

png(&#39;Images/Benchmark/no_split.png&#39;, res = 300, width = 1800, height = 2200, units = &#39;px&#39;)
  plot_grid(no_split_time, no_split_mem, nrow=2)
dev.off()

####
# split by n and type
####

bm_dat_summary_i &lt;- bm_dat_summary[bm_dat_summary$rule %in% c(&#39;format_target_i&#39;),]

n_type_split_time &lt;- 
  ggplot(bm_dat_summary_i, aes(x = n, y = mean_time, fill = type)) +
  geom_bar(stat = &quot;identity&quot;, position=&quot;dodge&quot;) +
  labs(x = &quot;Rule&quot;, y = &quot;Time (seconds)&quot;, fill = &#39;Type&#39;) +
  theme_half_open() +
  background_grid() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

n_type_split_mem &lt;- 
  ggplot(bm_dat_summary_i, aes(x = n, y = max_mem, fill = type)) +
  geom_bar(stat = &quot;identity&quot;, position=&quot;dodge&quot;) +
  labs(x = &quot;Rule&quot;, y = &quot;Max Memory (Mb)&quot;, fill = &#39;Type&#39;) +
  theme_half_open() +
  background_grid() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

png(&#39;Images/Benchmark/n_type_split.png&#39;, res = 300, width = 1800, height = 1800, units = &#39;px&#39;)
  plot_grid(n_type_split_time, n_type_split_mem, nrow=2)
dev.off()

####
# split by n
####

bm_dat_summary_i &lt;-
  bm_dat_summary[bm_dat_summary$rule %in% c(
    &#39;ancestry_inference_i&#39;,
    &#39;pc_projection_i&#39;,
    &#39;sample_report_i&#39;,
    &#39;outlier_detection_i&#39;
  ), ]

n_split_time &lt;- 
  ggplot(bm_dat_summary_i, aes(x = rule, y = mean_time, fill = n)) +
  geom_bar(stat = &quot;identity&quot;, position=&quot;dodge&quot;) +
  labs(x = &quot;Rule&quot;, y = &quot;Time (seconds)&quot;) +
  theme_half_open() +
  background_grid() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

n_split_mem &lt;- 
  ggplot(bm_dat_summary_i, aes(x = rule, y = max_mem, fill = n)) +
  geom_bar(stat = &quot;identity&quot;, position=&quot;dodge&quot;) +
  labs(x = &quot;Rule&quot;, y = &quot;Max Memory (Mb)&quot;) +
  theme_half_open() +
  background_grid() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

png(&#39;Images/Benchmark/n_split.png&#39;, res = 300, width = 1800, height = 2200, units = &#39;px&#39;)
  plot_grid(n_split_time, n_split_mem, nrow=2)
dev.off()</code></pre>
</details>
<hr />
</div>
<div id="results" class="section level1">
<h1>Results</h1>
<p>The pipeline was run using chromosome 22 only.</p>
<hr />
<details>
<summary>
Show PGS Methods
</summary>
<div class="centered-container">
<div class="rounded-image-container" style="width: 50%;">
<div class="figure">
<img src="Images/Benchmark/pgs_methods.png" alt="" />
<p class="caption">Rules affected by N cores and PGS method</p>
</div>
</div>
</div>
<div class="note-box">
<p><strong>Note:</strong> SBayesR and DBSLMM would capitalise on
multiple cores when running on multiple chromosomes. pT+clump is not
implemented in parallel, since they are already fast.</p>
</div>
</details>
<details>
<summary>
Show target scoring
</summary>
<div class="centered-container">
<div class="rounded-image-container" style="width: 50%;">
<div class="figure">
<img src="Images/Benchmark/target_pgs.png" alt="" />
<p class="caption">Rules affected by N cores and target sample size</p>
</div>
</div>
</div>
<div class="note-box">
<p><strong>Note:</strong> target_pgs_i resources vary according to the
number of pgs_methods and scores requested. In this benchmark, 4586
scores were computed for each individual.</p>
</div>
</details>
<details>
<summary>
Show rules with constant time/memory
</summary>
<div class="centered-container">
<div class="rounded-image-container" style="width: 50%;">
<div class="figure">
<img src="Images/Benchmark/no_split.png" alt="" />
<p class="caption">Rules with constant time/memory</p>
</div>
</div>
</div>
</details>
<details>
<summary>
Show rules affected by N
</summary>
<div class="centered-container">
<div class="rounded-image-container" style="width: 50%;">
<div class="figure">
<img src="Images/Benchmark/n_split.png" alt="" />
<p class="caption">Rules affected by target sample size</p>
</div>
</div>
</div>
</details>
<details>
<summary>
Show rules affected by N and Type (format_target_i)
</summary>
<div class="centered-container">
<div class="rounded-image-container" style="width: 50%;">
<div class="figure">
<img src="Images/Benchmark/n_type_split.png" alt="" />
<p class="caption">Rules affected by target sample size and genotype
data format</p>
</div>
</div>
</div>
</details>
<hr />
</div>
<div id="required-storage" class="section level1">
<h1>Required storage</h1>
<p>This shows the disk space required for the reference data:</p>
<pre><code>38G     gctb_ref
17M     hm3_snplist
2.3G    ldak_bld
1.0K    ldak_highld
119M    ldak_map
459K    ld_blocks
7.2G    ldpred2_ref
44M     ldsc_ref
21M     logs
4.4G    prscs_ref
1.1G    ref</code></pre>
<p>In total this is ~50G, most of which is the GCTB reference required
for SBayesR.</p>
<hr />
</div>
<div id="genome-wide-benchmark" class="section level1">
<h1>Genome-wide benchmark</h1>
<p>We also performed a benchmark of PGS methods genome-wide using
European GWAS summary statistics from Yengo et al.Â This was carried out
using default settings in GenoPred, with 10 cores allocated to each PGS
method, except pT+clump and DBSLMM which are run on a single core.</p>
<details>
<summary>
Show PGS methods benchmark
</summary>
<div class="centered-container">
<div class="rounded-image-container" style="width: 50%;">
<div class="figure">
<img src="Images/OpenSNP/time_cpu_bench_pgs_methods.png" alt="" />
<p class="caption">PGS methods run genome-wide</p>
</div>
</div>
</div>
<div class="note-box">
<p><strong>Note:</strong> pT+clump is run using 1 core. All other
methods are run using 10 cores.</p>
</div>
</details>
</div>

<!-- footer.html -->
<hr/>

<div class="centered-container">
<div class="rounded-image-container" style="width: 500px;">
<img src="Images/logo/sponsors.png">
</div>
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
