<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>GenoPred Pipeline - Using 23andMe data</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />
<link rel="shortcut icon" href="Images/logo/favicon.ico">

<link rel="stylesheet" href="styles/night-mode.css" id="nightModeStylesheet">

<script>
function toggleNightMode() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (stylesheet.disabled) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
}
</script>

<label class="switch">
  <input type="checkbox" id="toggleNightMode" checked>
  <span class="slider round"></span>
</label>

<script>
document.getElementById('toggleNightMode').addEventListener('change', function() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (this.checked) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
});
</script>

<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-YR18ZB3PR3"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-YR18ZB3PR3');
</script>


<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="styles/styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img class="logo-img" src="Images/logo/Horizontal_white.png" style="height: 42px;" /></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pipeline
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="pipeline_overview.html">Overview</a>
    </li>
    <li>
      <a href="pipeline_readme.html">Instructions</a>
    </li>
    <li>
      <a href="pipeline_technical.html">Technical documentation</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Research
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="research_index.html">Overview</a>
    </li>
    <li>
      <a href="comparison_of_methods_summary.html">Polygenic Scoring Methods Comparison</a>
    </li>
    <li>
      <a href="Functionally_informed_prediction.html">Quantifying Polygenic Signal Mediated by Altered Gene Expression</a>
    </li>
    <li>
      <a href="Absolute_Conversion.html">Translating Polygenic Scores onto the Absolute Scale</a>
    </li>
  </ul>
</li>
<li>
  <a href="more_index.html">More</a>
</li>
<li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">GenoPred Pipeline - Using 23andMe data</h1>

</div>


<hr />
<p>This document is a simplified set of instructions for the specific
scenario where 23andMe data for an individual, and a set of score files
on the PGS catalogue is being used as the input to the pipeline. It also
provides a demonstration of using the genopred container.</p>
<hr />
<div id="setting-up" class="section level1">
<h1>Setting up</h1>
<div id="download-genopred-container" class="section level2">
<h2>Download GenoPred container</h2>
<p>Containers are available for both docker and singularity (see <a
href="pipeline_readme.html#containers">here</a>)</p>
<pre class="bash"><code>singularity \
  pull \
  --arch \
  amd64 \
  /users/k1806347/oliverpainfel/Software/singularity/genopred_pipeline_latest.sif \
  library://opain/genopred/genopred_pipeline:latest</code></pre>
<hr />
</div>
<div id="prepare-configuration" class="section level2">
<h2>Prepare configuration</h2>
<pre class="r"><code># Create directory to store configuration files, outputs and resources
workdir&lt;-&#39;/users/k1806347/oliverpainfel/test/genopred_23andme&#39;
dir.create(paste0(workdir, &#39;/config&#39;), recursive = T)
dir.create(paste0(workdir, &#39;/resources&#39;), recursive = T)
dir.create(paste0(workdir, &#39;/output&#39;), recursive = T)

# Create config file
config&lt;-c(
  paste0(&#39;outdir: &#39;, workdir, &#39;/output&#39;),
  paste0(&#39;resdir: &#39;, workdir, &#39;/resources&#39;),
  paste0(&#39;config_file: &#39;, workdir, &#39;/config/config.yaml&#39;),
  paste0(&#39;target_list: &#39;, workdir, &#39;/config/target_list.txt&#39;),
  paste0(&#39;score_list: &#39;, workdir, &#39;/config/score_list.txt&#39;),
  &#39;cores_target_pgs: 8&#39;,
  &#39;cores_impute_23andme: 8&#39;
)

write.table(config, paste0(workdir, &#39;/config/config.yaml&#39;), col.names = F, row.names = F, quote = F)

# Create target_list
target_list&lt;-data.frame(
  name=&#39;Joe_Bloggs&#39;,
  path=&#39;/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline/test_data/target/23andMe_individual/Joe_Bloggs_genome_0123456789.zip&#39;,
  type=&#39;23andMe&#39;,
  indiv_report=TRUE
)

write.table(target_list, paste0(workdir, &#39;/config/target_list.txt&#39;), col.names=T, row.names=F, quote=F)

# Create score_list
score_list&lt;-data.frame(
  name=&#39;PGS002804&#39;,
  path=NA,
  label=&#39;Height&#39;
)

score_list$label&lt;-paste0(&quot;\&quot;&quot;, score_list$label, &quot;\&quot;&quot;)

write.table(score_list, paste0(workdir, &#39;/config/score_list.txt&#39;), col.names=T, row.names=F, quote=F)</code></pre>
<hr />
</div>
</div>
<div id="run-pipeline" class="section level1">
<h1>Run pipeline</h1>
<pre class="bash"><code># Start container
singularity shell \
  --bind /scratch/prj/oliverpainfel:/scratch/prj/oliverpainfel \
  --writable-tmpfs \
  /users/k1806347/oliverpainfel/Software/singularity/genopred_pipeline_latest.sif
  
# Activate GenoPred environment
source /opt/mambaforge/etc/profile.d/conda.sh
conda activate genopred

# Go to GenoPred/pipeline directory
cd /tools/GenoPred/pipeline/

# Do a dry run to check the scheduled steps are expected (there should not be any steps saying &#39;download&#39;, and it should not be necessary to build the conda environment)
snakemake -j8 --use-conda --configfile=/users/k1806347/oliverpainfel/test/genopred_23andme/config/config.yaml output_all</code></pre>
<hr />
</div>
<div id="computational-benchmark" class="section level1">
<h1>Computational benchmark</h1>
<details>
<summary>
Show code
</summary>
<pre class="r"><code>bm_files_i&lt;-list.files(path=&#39;/users/k1806347/oliverpainfel/test/genopred_23andme/output/reference/benchmarks&#39;, full.names = T)

bm_dat_all &lt;- do.call(rbind, lapply(bm_files_i, function(file) {
  tmp &lt;- fread(file)
  tmp$file &lt;- basename(file)
  return(tmp)
}))

# Create rule column
bm_dat_all$rule &lt;- gsub(&#39;-.*&#39;,&#39;&#39;,bm_dat_all$file)

# Calculate total wall time
sum(bm_dat_all$s)/60/60 # 4.904476 hours

# Calculate wall time taken by rule
bm_dat_rule_time &lt;- NULL
for (i in unique(bm_dat_all$rule)) {
  bm_dat_rule_time &lt;- rbind(
    bm_dat_rule_time,
    data.frame(
      rule = i,
      time = sum(bm_dat_all$s[bm_dat_all$rule == i])))
}

# Tidy results
bm_dat_rule_time &lt;-
  bm_dat_rule_time[!(bm_dat_rule_time$rule %in% c(&#39;ancestry_reporter&#39;, &#39;score_reporter.txt&#39;)), ]

bm_dat_rule_time$step&lt;-gsub(&#39;_i$&#39;,&#39;&#39;, bm_dat_rule_time$rule)
bm_dat_rule_time$step&lt;-gsub(&#39;prep_pgs_&#39;,&#39;&#39;, bm_dat_rule_time$step)

bm_dat_rule_time$step[bm_dat_rule_time$rule == &#39;format_target_i&#39;]&lt;-&#39;Target QC&#39;
bm_dat_rule_time$step[bm_dat_rule_time$rule == &#39;impute_23andme_i&#39;]&lt;-&#39;Imputation&#39;
bm_dat_rule_time$step[bm_dat_rule_time$rule == &#39;sumstat_prep_i&#39;]&lt;-&#39;GWAS QC&#39;
bm_dat_rule_time$step[bm_dat_rule_time$rule == &#39;indiv_report_i&#39;]&lt;-&#39;Report Creation&#39;
bm_dat_rule_time$step[bm_dat_rule_time$rule == &#39;ancestry_inference_i&#39;]&lt;-&#39;Ancestry Inference&#39;
bm_dat_rule_time$step[bm_dat_rule_time$rule == &#39;target_pgs_i&#39;]&lt;-&#39;Target Scoring&#39;
bm_dat_rule_time$step[bm_dat_rule_time$rule == &#39;prep_pgs_external_i&#39;]&lt;-&#39;Score file QC&#39;
bm_dat_rule_time$step[bm_dat_rule_time$rule == &#39;download_pgs_external&#39;]&lt;-&#39;Download score file&#39;

#######
# Create a pie chart
#######

data &lt;- data.frame(
  category = bm_dat_rule_time$step,
  values = bm_dat_rule_time$time
)

data$category&lt;-factor(data$category, 
                      levels = c(&#39;Imputation&#39;,
                                 &#39;Target QC&#39;,
                                 &#39;Ancestry Inference&#39;,
                                 &#39;Download score file&#39;,
                                 &#39;Score file QC&#39;,
                                 &#39;Target Scoring&#39;,
                                 &#39;Report Creation&#39;))

data&lt;-data[order(data$category),]

data$perc&lt;-data$values/sum(data$values)*100
data$cum_perc&lt;-cumsum(data$perc)

for(i in 1:length(data$category)) {
  data$start[i] &lt;- ifelse(i == 1, 0, data$end[i - 1])
  data$end[i] &lt;- data$cum_perc[i]
  data$label_position[i] &lt;- data$cum_perc[i] - data$perc[i] / 2
}

data$time_clean&lt;-NA
data$time_clean[data$values &lt; 60] &lt;-
  paste0(round(data$values[data$values &lt; 60], 1), &#39; sec&#39;)
data$time_clean[data$values &gt; 60] &lt;-
  paste0(round(data$values[data$values &gt; 60] / 60, 1), &#39; min&#39;)
data$time_clean[data$values &gt; 3600] &lt;-
  paste0(round(data$values[data$values &gt; 3600] / 60 / 60, 1), &#39; hr&#39;)

data$label&lt;-paste0(data$category,
                   &quot;\n(&quot;, data$time_clean, &#39;, &#39;,
                   round(data$perc,1),&quot;%)&quot;)

library(ggplot2)
library(ggrepel)

png(&#39;~/oliverpainfel/Software/MyGit/GenoPred/docs/Images/Benchmark/23andme_benchmark.png&#39;, res=300, width=2500, height=1000, units = &#39;px&#39;)

set.seed(1)
ggplot(data,
       aes(
         xmin = start,
         xmax = end,
         ymin = 0,
         ymax = 0.1,
         fill = category
       )) +
  geom_rect(colour=&#39;black&#39;, size=0.1) +
  geom_text_repel(
    aes(
      x = label_position,
      y = 0.1,
      label = label,
      segment.square  = T,
      segment.inflect = T
    ),
    force = 100,
    nudge_y           = 0.15,
    hjust = 0.5,
    segment.size      = 0.3,
    segment.curvature = -0.1,
    segment.color=&#39;darkgrey&#39;,
    box.padding=0.6,
  ) +
  scale_x_continuous(breaks = seq(0, 100, by = 20)) +
  coord_cartesian(clip = &quot;off&quot;, xlim = c(-20, 120), ylim = c(0, 0.3)) +
  labs(x=&#39;Time (%)&#39;) +
  theme(
    axis.title.y = element_blank(),
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    axis.line.y = element_blank(),
    axis.line.x = element_line(color = &quot;black&quot;, size = 0.5),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    legend.position = &quot;none&quot;,
    plot.background = element_rect(fill = &quot;white&quot;, color = NA),
    panel.background = element_rect(fill = &quot;white&quot;, color = NA)
  )

dev.off()</code></pre>
</details>
<hr />
<div class="centered-container">
<div class="rounded-image-container">
<div class="figure">
<img src="Images/Benchmark/23andme_benchmark.png" alt="" />
<p class="caption">Time taken for each step</p>
</div>
</div>
</div>
</div>

<!-- footer.html -->
<hr/>

<div class="centered-container">
<div class="rounded-image-container" style="width: 500px;">
<img src="Images/logo/sponsors.png">
</div>
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
