<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>GenoPred Pipeline - Technical details</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />
<link rel="shortcut icon" href="Images/logo/favicon.ico">

<link rel="stylesheet" href="styles/night-mode.css" id="nightModeStylesheet">

<script>
function toggleNightMode() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (stylesheet.disabled) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
}
</script>

<label class="switch">
  <input type="checkbox" id="toggleNightMode" checked>
  <span class="slider round"></span>
</label>

<script>
document.getElementById('toggleNightMode').addEventListener('change', function() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (this.checked) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
});
</script>

<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-YR18ZB3PR3"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-YR18ZB3PR3');
</script>


<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="styles/styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img class="logo-img" src="Images/logo/Horizontal_white.png" style="height: 42px;" /></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pipeline
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="pipeline_overview.html">Overview</a>
    </li>
    <li>
      <a href="pipeline_readme.html">Instructions</a>
    </li>
    <li>
      <a href="pipeline_technical.html">Technical documentation</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Research
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="research_index.html">Overview</a>
    </li>
    <li>
      <a href="comparison_of_methods_summary.html">Polygenic Scoring Methods Comparison</a>
    </li>
    <li>
      <a href="Functionally_informed_prediction.html">Quantifying Polygenic Signal Mediated by Altered Gene Expression</a>
    </li>
    <li>
      <a href="Absolute_Conversion.html">Translating Polygenic Scores onto the Absolute Scale</a>
    </li>
  </ul>
</li>
<li>
  <a href="more_index.html">More</a>
</li>
<li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">GenoPred Pipeline - Technical details</h1>

</div>


<hr />
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>This document provides technical details of the GenoPred pipeline.
The GenoPred pipeline automates the process of calculating polygenic
scores. The pipeline aims to implement the current practises for
polygenic scoring. See <a href="pipeline_overview.html">here</a> more
general information regarding the GenoPred pipeline.</p>
<p>Please cite our preprint when using the pipeline:</p>
<ul>
<li>“Pain, O. et al. ”The GenoPred Pipeline: A Comprehensive and
Scalable Pipeline for Polygenic Scoring.” MedRxiv 2024. <a
href="https://doi.org/10.1101/2024.06.12.24308843"
class="uri">https://doi.org/10.1101/2024.06.12.24308843</a></li>
</ul>
<p>If relevant, please also cite our paper comparing polygenic scoring
methods and describing the reference-standardised approach:</p>
<ul>
<li>Pain, O. et al. “Evaluation of polygenic prediction methodology
within a reference-standardized framework.” PLoS genetics. <a
href="https://doi.org/10.1371/journal.pgen.1009021"
class="uri">https://doi.org/10.1371/journal.pgen.1009021</a></li>
</ul>
<p>Please also cite the relevant studies for the tools and data used by
the GenoPred pipeline.</p>
<hr />
</div>
<div id="software" class="section level1">
<h1>Software</h1>
<p>The pipeline, designed for both ease of use and reproducibility, is
constructed using Snakemake. This workflow management system integrates
a variety of scripts: R scripts, Python scripts, and bash scripts,
providing a comprehensive toolkit. To ensure a seamless experience and
maintain reproducibility, the entire pipeline operates within a Conda
environment. Moreover, for accessibility and collaboration, the pipeline
has been made publicly available on GitHub, allowing users and
developers alike to access, utilize, and contribute to its continuous
improvement.</p>
<hr />
</div>
<div id="workflow" class="section level1">
<h1>Workflow</h1>
<p>The pipeline can be separated into two halves, reference data
preparation and target data application. Reference data preparation
refers to steps that only require reference data, including the
reference genotype data, GWAS summary statistics, and externally
generated score files. Target data application refers to steps that
apply previously prepared reference data to the target datasets.</p>
<p>See <a href="pipeline_readme.html#pipeline-configuration">here</a>
for more information on pipeline configuration.</p>
<hr />
<div class="centered-container">
<div class="rounded-image-container">
<div class="figure">
<img src="Images/pipeline_readme/pipeline_schematic_lowdef.png"
alt="" />
<p class="caption">Figure 1. GenoPred pipeline schematic. Shows input
files, processes, outputs, and rules.</p>
</div>
</div>
</div>
<hr />
</div>
<div id="reference-data-preparation" class="section level1">
<h1>Reference data preparation</h1>
<hr />
<div id="reference-genetic-data" class="section level2">
<h2>Reference genetic data</h2>
<p>The default reference genotype data used by GenoPred is a previously
prepared dataset, which is based on the 1000 Genomes Phase 3 (1KG) and
Human Genome Diversity Project (HGDP) sample, restricted HapMap3
variants. This dataset contains 1204449 variants for 3313
individuals.</p>
<p>Users can provide there own reference data using the
<code>refdir</code> parameter in the <code>configfile</code>. The
reference data folder must have the following structure:</p>
<pre><code>[refdir]
  ├── ref.chr[1-22].[pgen/pvar/psam] (plink2 genotype data with RSIDs in SNP column)
  ├── ref.chr[1-22].rds (SNP data - refer to default ref data for format)
  ├── ref.pop.txt (Population data for reference individuals - with header)
  ├── ref.keep.list (lists keep files for each population - columns pop and path - no header)
  ├── keep_files
  │   └──[pop].keep (keep files for each population - no header)
  └── freq_files
      └──[pop]
          └──ref.[pop].chr[1-22].afrq (plink2 .afrq format)</code></pre>
<p><strong>Note:</strong> .psam, ref.pop.txt and keep_files must contain
IID, and can optionally include FID information. The ID information must
be consistent across these files.</p>
<hr />
</div>
<div id="reference-pca" class="section level2">
<h2>Reference PCA</h2>
<p>Principal components analysis (PCA) using the reference genetic data
is performed using the <a
href="https://github.com/opain/GenoPred/blob/master/Scripts/ref_pca/ref_pca.R">ref_pca.R
script</a>. First, PLINK v1.9 is used to retain variants with MAF &gt;
0.05, missingness &lt;0.02, and HWE p-value &gt; 1e-6. Then, long range
LD regions are removed (<a
href="https://www.cell.com/ajhg/fulltext/S0002-9297(08)00353-4">ref</a>),
and LD-pruning is applied using PLINK v1.9, using a 1000kb sliding
window, with step size 5, and r2 threshold of 0.2 (–indep-pairwise 1000
5 0.2). Then PCA is performed using plink v2, storing SNP-weights for
the first 6 PCs. The mean and SD of PCs within each reference population
are calculated.</p>
<hr />
</div>
<div id="gwas-sumstat-qc" class="section level2">
<h2>GWAS sumstat QC</h2>
<p>GWAS summary statistics, referred to as sumstats, submitted by the
user are interpreted, formatted, and cleaned using the <a
href="https://github.com/opain/GenoUtils/blob/main/inst/scripts/sumstat_cleaner.R">sumstat_cleaner.R</a>
script within the <a
href="https://github.com/opain/GenoUtils/tree/main">GenoUtils</a>
package. The GenoPred pipeline uses the default parameters for the
script.</p>
<p>GWAS sumstats are read into R using the data.table::fread function,
handling a range of column delimiters and compressions. The sumstat
column names are interpreted and standardised using a <a
href="https://github.com/opain/GenoUtils/blob/main/R/constants.R">dictionary</a>.
If sample size and allele frequency are reported for cases and controls,
the total sample size and mean allele frequency are inserted.</p>
<p>The variants within the sumstats are then aligned to the 1KG+HGDP
reference dataset. Strand ambiguous variants are removed. If chromosome
number (CHR) and basepair position (BP) data is present in the sumstats,
the genome build is identified through comparison of chromosome 22 data
to a range of reference builds, considering the alleles of each variant
(A1/A2) - <a
href="https://www.bioinformatics.org/sms/iupac.html">IUPAC</a> codes are
used throughout, to allow for strand flips between the sumstats and
reference. Then, sumstat variants are matched to reference variants
using CHR, BP, and IUPAC data. If CHR and BP is not present, or genome
build cannot be determined, RSIDs and IUPAC codes are used to match
sumstat and reference variants. After matching with the reference, the
reference allele frequency is inserted (REF.FREQ), using data from the
1KG+HGDP population matching that of the GWAS sample.</p>
<p>Variants in the summary statistics are then filtered according to the
following criteria:</p>
<ul>
<li>Imputation quality (INFO): By default, variants with an INFO score
&lt; 0.9 are removed (if available in summary statistics).</li>
<li>Minor allele frequency (MAF): MAF thresholds were set at 0.01. The
script filtered variants below this threshold in either the GWAS sample
(if available – FREQ) or reference data (REF.FREQ). Additionally, a MAF
difference threshold (default 0.2) was used to exclude variants with
substantial MAF discrepancies between GWAS and reference
populations.</li>
<li>P-Value: Variants with out-of-bound p-values (0 &lt; P ≤ 1) were
removed, ensuring statistical validity.</li>
<li>Duplicate SNP ID: All variants with a duplicated SNP ID are
removed.</li>
<li>Sample Size: Variants with sample size &gt;3SD from the median are
removed (if per variant sample size is available in summary
statistics).</li>
</ul>
<p>Finally, the script inserts a BETA coefficient column if it is not
present already, calculating as log(OR), or using signed Z-scores,
sample size and allele frequency estimates (<a
href="https://www.nature.com/articles/ng.3538">ref</a>). The standard
error (SE) is also inserted if it is not present, subsequently removing
variants with a SE equal to zero. If the SE is available, there is also
a check whether P values have been adjusted using genomic control – If
genomic control is detected, the p-value is recomputed using Z scores
(BETA/SE).</p>
<hr />
</div>
<div id="adjustment-of-gwas-effect-sizes" class="section level2">
<h2>Adjustment of GWAS effect sizes</h2>
<p>The GenoPred pipeline integrates various polygenic scoring methods
for adjusting effect sizes in GWAS summary statistics. It uses a
reference-standardized approach, adjusting GWAS effect sizes based on an
ancestry-matched reference sample, without considering specific target
sample characteristics.</p>
<hr />
<div id="single-ancestry-methods" class="section level3">
<h3>Single Ancestry Methods</h3>
<p>The GenoPred pipeline currently only implements PGS methods intended
for GWAS based on a single ancestral population. When using GWAS based
on mixed ancestry individuals, we suggest specifying the population that
matches most individuals in the GWAS.</p>
<div
style="border: 1px solid #ddd; padding: 0px; overflow-y: scroll; height:500px; overflow-x: scroll; width:100%; ">
<table class="table table-striped table-hover" style="color: black; width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
Table 1 provides a summary of each approach and its implementation in
GenoPred.
</caption>
<thead>
<tr>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
Method
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
Software
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
PubMed ID
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
Pseudovalidation Option
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
Parameters
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
MHC Region
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
LD Reference
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
CPU usage*
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
Memory Usage*
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
DBSLMM
</td>
<td style="text-align:left;">
DBSLMM
</td>
<td style="text-align:left;">
32330416
</td>
<td style="text-align:left;">
Yes (only option)
</td>
<td style="text-align:left;">
SNP-heritability estimated using LD Score Regression (on liability scale
for binary outcomes)
</td>
<td style="text-align:left;">
Not excluded
</td>
<td style="text-align:left;">
Population-matched 1KG+HGDP
</td>
<td style="text-align:left;">
40 seconds
</td>
<td style="text-align:left;">
450 Mb
</td>
</tr>
<tr>
<td style="text-align:left;">
lassosum
</td>
<td style="text-align:left;">
lassosum R package
</td>
<td style="text-align:left;">
28480976
</td>
<td style="text-align:left;">
Yes
</td>
<td style="text-align:left;">
s = 0.2, 0.5, 0.9, 1; lambda = exp(seq(log(0.001), log(0.1),
length.out=20))
</td>
<td style="text-align:left;">
Not excluded
</td>
<td style="text-align:left;">
Population-matched 1KG+HGDP
</td>
<td style="text-align:left;">
10 seconds
</td>
<td style="text-align:left;">
400 Mb
</td>
</tr>
<tr>
<td style="text-align:left;">
LDpred2
</td>
<td style="text-align:left;">
bigsnpr R package
</td>
<td style="text-align:left;">
33326037
</td>
<td style="text-align:left;">
Yes
</td>
<td style="text-align:left;">
Grid includes 126 combinations of heritability and non-zero effect
fractions (p).
</td>
<td style="text-align:left;">
Not excluded
</td>
<td style="text-align:left;">
EUR UKB (LDpred2-provided)
</td>
<td style="text-align:left;">
3 minutes
</td>
<td style="text-align:left;">
500 Mb
</td>
</tr>
<tr>
<td style="text-align:left;">
MegaPRS
</td>
<td style="text-align:left;">
LDAK
</td>
<td style="text-align:left;">
34234142
</td>
<td style="text-align:left;">
Yes
</td>
<td style="text-align:left;">
Fits lasso, ridge, bolt and BayesR models, with a total of 148 sets of
hyperparameters
</td>
<td style="text-align:left;">
Not excluded
</td>
<td style="text-align:left;">
Population-matched 1KG+HGDP
</td>
<td style="text-align:left;">
5 minutes
</td>
<td style="text-align:left;">
450Mb
</td>
</tr>
<tr>
<td style="text-align:left;">
PRS-CS
</td>
<td style="text-align:left;">
PRS-CS
</td>
<td style="text-align:left;">
30992449
</td>
<td style="text-align:left;">
Yes
</td>
<td style="text-align:left;">
phi = 1e-6, 1e-4, 1e-2, 1, auto
</td>
<td style="text-align:left;">
Not excluded
</td>
<td style="text-align:left;">
Population-matched UKB (PRS-CS provided)
</td>
<td style="text-align:left;">
35 minutes
</td>
<td style="text-align:left;">
350 Mb
</td>
</tr>
<tr>
<td style="text-align:left;">
pT+clump
</td>
<td style="text-align:left;">
PLINK
</td>
<td style="text-align:left;">
25722852
</td>
<td style="text-align:left;">
No
</td>
<td style="text-align:left;">
p-value thresholds: 1e-8, 1e-6, 1e-4, 1e-2, 0.1, 0.2, 0.3, 0.4, 0.5, 1;
Clumping: r2 = 0.1; window = 250kb
</td>
<td style="text-align:left;">
Only top variant retained
</td>
<td style="text-align:left;">
Population-matched 1KG+HGDP
</td>
<td style="text-align:left;">
5 seconds
</td>
<td style="text-align:left;">
100 Mb
</td>
</tr>
<tr>
<td style="text-align:left;">
SBayesR
</td>
<td style="text-align:left;">
GCTB
</td>
<td style="text-align:left;">
31704910
</td>
<td style="text-align:left;">
Yes (only option)
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
Excluded (as recommended)
</td>
<td style="text-align:left;">
EUR UKB (GCTB-provided)
</td>
<td style="text-align:left;">
3 minutes
</td>
<td style="text-align:left;">
500 Mb
</td>
</tr>
</tbody>
</table>
</div>
<p>* = Based on chromosome 22 using single core. See <a
href="pipeline_benchmark.html">here</a> for full benchmark
information.</p>
<hr />
<div id="dblsmm" class="section level4">
<h4>DBLSMM</h4>
<p>DBSLMM, a Bayesian shrinkage method, is implemented using the script
<a
href="https://github.com/opain/GenoPred/blob/master/Scripts/pgs_methods/dbslmm.R">pgs_methods/dbslmm.R</a>.
We use the ‘tuning’ version, where the SNP-based heritability parameter
is varied by a factor of 0.8, 1, and 1.2. When using a factor of 1, this
corresponds to the ‘default’ version of DBSLMM. The SNP-based
heritability (on the liability scale for binary outcomes) is estimated
using LD-score regression, using LD scores released by pan-UK (<a
href="https://pan.ukbb.broadinstitute.org/downloads">link</a>). GenoPred
clones DBSLMM (d158a144dd2f2dc883ad93d0ea71e8fc48e80bd3) from GitHub,
downloads the <a
href="https://drive.google.com/file/d/1eAbEyhF8rO_faOFL3jqRo9LmfgJNRH6K/view">precompiled
software binary</a> provided by the DBSLMM authors, and uses the
1KG+HGDP reference for LD estimation based on the super population of
the GWAS sample. LD block data is only provided for EUR, EAS and AFR
populations - If the GWAS are from CSA, AMR, MID populations, EUR LD
block data is used.</p>
<hr />
</div>
<div id="lassosum" class="section level4">
<h4>lassosum</h4>
<p>lassosum, an R package, applies the LASSO method for constructing
PGS, is implemented using the script <a
href="https://github.com/opain/GenoPred/blob/master/Scripts/pgs_methods/lassosum.R">pgs_methods/lassosum.R</a>.
It uses a range of lambda hyperparameters and offers a pseudovalidation
feature to deduce the optimal lambda value, eliminating the need for
external testing data. GenoPred uses lassosum v0.4.5, with the default
range of lambda values, and the 1KG+HGDP reference for LD estimation,
matching the population of the GWAS sample. LD block data for EUR, EAS
and AFR are provided by lassosum - If the GWAS are from CSA, AMR, MID
populations, EUR LD block data is used.</p>
<hr />
</div>
<div id="ldpred2" class="section level4">
<h4>LDpred2</h4>
<p>LDpred2, part of the bigsnpr R package, operates in ‘inf’, ‘grid’,
and ‘auto’ modes. The ‘inf’ mode is better suited for highly polygenic
traits, whereas ‘grid’ and ‘auto’ adjust effect sizes using various
hyperparameters, including SNP heritability. In GenoPred, LDpred2 is run
using the script <a
href="https://github.com/opain/GenoPred/blob/master/Scripts/pgs_methods/ldpred2.R">pgs_methods/ldpred2</a>.R.
GenoPred uses bigsnpr v1.12.2, with the default LDpred2 grid search, and
recommended GWAS quality control checks. GenoPred employs LDpred2’s
precomputed LD matrices based on the European individuals from the UK
Biobank, and it is applied only to GWAS based on a EUR population. If
the SNP-h2 estimated using LDSC is &lt;0.05, the SNP-heritability used
by LDpred2 is set to 0.05.</p>
<hr />
</div>
<div id="megaprs" class="section level4">
<h4>MegaPRS</h4>
<p>MegaPRS uses a range of priors (lasso, ridge, bolt, BayesR) for SNP
effects, run via LDAK software. In GenoPred, MegaPRS is run using the
script <a
href="https://github.com/opain/GenoPred/blob/master/Scripts/pgs_methods/megaprs.R">pgs_methods/megaprs.R</a>.
The GenoPred pipeline uses LDAK v5.1, with the BLD-LDAK heritability
model, where SNP variance is dependent on allele frequency, LD, and
functional annotations. It includes a pseudovalidation approach.
GenoPred employs the 1KG+HGDP reference for LD estimation, matching the
population of the GWAS sample.</p>
<hr />
</div>
<div id="prs-cs" class="section level4">
<h4>PRS-CS</h4>
<p>PRS-CS, a Bayesian method using a continuous shrinkage prior,
specifies a range of global shrinkage parameters (phi), generating
multiple sets of genetic effects for polygenic scoring. Its ‘auto’ model
estimates the optimal parameter directly from GWAS summary statistics,
negating the need for an external dataset. In GenoPred, PRS-CS is run
using the script <a
href="https://github.com/opain/GenoPred/blob/master/Scripts/pgs_methods/prscs.R">pgs_methods/prscs.R</a>.
By default, GenoPred specifies four phi parameters (1e-6, 1e-4, 1e-2, 1)
and the auto model, but the user can modify this behaviour using the
prscs_phi parameter in the configfile. By default, GenoPred uses the
PRS-CS provided 1KG-derived LD matrix data, matching the population of
the GWAS sample. The user can select the UKB-derived LD matrix data to
be used using the <code>prscs_ldref</code> parameter in the
<code>configfile</code>. 1KG is used by default as PGS based on Yengo et
al. sumstats performed significantly better in the OpenSNP target
sample, when using the 1KG reference data (this may differ for other
GWAS).</p>
<hr />
</div>
<div id="ptclump" class="section level4">
<h4>pT+clump</h4>
<p>In GenoPred, pT+clump is run using the script <a
href="https://github.com/opain/GenoPred/blob/master/Scripts/pgs_methods/ptclump.R">pgs_methods/ptclump.R</a>.
This method conducts LD-based clumping (using PLINK v1.9, r2 = 0.1;
window = 250kb) and shrinks effects to zero across a range of p-value
thresholds (1e-8, 1e-6, 1e-4, 1e-2, 0.1, 0.2, 0.3, 0.4, 0.5, 1). To
address long-range LD in the MHC/HLA region (Chr6:28-34Mb), only the
most associated variant is used for polygenic scoring. GenoPred uses the
1KG+HGDP reference for LD estimation, matching the population of the
GWAS sample.</p>
<hr />
</div>
<div id="sbayesr" class="section level4">
<h4>SBayesR</h4>
<p>In GenoPred, SBayesR is run using the script <a
href="https://github.com/opain/GenoPred/blob/master/Scripts/pgs_methods/sbayesr.R">pgs_methods/sbayesr.R</a>.
Implemented with GCTB software, SBayesR is a Bayesian method that uses
GWAS summary statistics to estimate key parameters. GenoPred uses GCTB
v2.03 with LD matrices from the GCTB authors, calculated using European
individuals in the UK Biobank, restricting its application to GWAS based
on EUR populations. GenoPred uses the robust parameterisation option in
SBayesR.</p>
<hr />
</div>
</div>
<div id="pseudovalidation" class="section level3">
<h3>Pseudovalidation</h3>
<p>Several polygenic scoring methods use a range of ‘tuning’ parameters
when adjusting the GWAS effect sizes for polygenic scoring, resulting in
multiple polygenic scores from a given GWAS. Traditionally, the user
would then evaluate the predictive utility of polygenic scores derived
using these various parameters, to select the best performing polygenic
score, using cross-validation or bootstrapping to estimating the effect
of the polygenic score without overfitting. However, several methods
offer a pseudovalidation approach, whereby the optimal tuning parameter
is estimated based on the GWAS summary statistics alone, convieniently
avoiding the need for a tuning sample. The performance of
pseudovalidation relative to formal validation varies between methods.
See our previous <a href="comparison_of_methods_summary.html">polygenic
scoring methods comparison</a> for more information.</p>
<hr />
</div>
</div>
<div id="external-score-file-qc" class="section level2">
<h2>External score file QC</h2>
<p>This step harmonises externally derived score files with the
reference genetic data. It is performed using the <a
href="https://github.com/opain/GenoPred/blob/master/Scripts/external_score_processor/external_score_processor.R">external_score_processor.R</a>
script. The script requires the <a
href="https://www.pgscatalog.org/downloads/#scoring_header">PGS
Catalogue header format</a>. First the script checks whether the PGS
file contains comments in the header, and if it does, records the
reported genome build. It then reads in the score files, retaining the
PGS catalogue harmonised columns if they are present. The variants
within the score file are then aligned to a reference dataset - This is
the same process used to <a href="#gwas-sumstat-qc">harmonise GWAS
summary statistics</a>, except the script will use the reported genome
build if present.</p>
<hr />
</div>
</div>
<div id="target-data-application" class="section level1">
<h1>Target data application</h1>
<p>The user can provide individual-level target genotype data in several
formats, including PLINK1, PLINK2, VCF, BGEN, or 23andMe. PLINK1,
PLINK2, VCF and BGEN datasets are assumed to have already undergone
genotype imputation and can include 1 or more individuals. 23andMe
format data contains data for a single individual and is assumed to be
unimputed. Therefore, prior to reference harmonisation, 23andMe data
undergoes imputation and is converted to PLINK1 format within the
GenoPred pipeline.</p>
<hr />
<div id="target-genotype-imputation" class="section level2">
<h2>Target genotype imputation</h2>
<p>Genotype imputation is only performed for target datasets in the
format of 23andMe. This functionality is included for polygenic scores
to be calculated for 23andMe datasets. GenoPred is not intended for
large-scale genotype imputation – Instead we would recommend using free
and publicly available imputation servers, which leverage large-scale
haplotype reference panels. GenoPred requires target datasets in formats
other than 23andMe to have been imputed in advance, or to data that does
not require imputation (such as whole genome sequence data).</p>
<p>Genotype imputation in GenoPred is performed using the <a
href="https://github.com/opain/GenoPred/blob/master/Scripts/23andMe_imputer/23andMe_imputer.R">23andMe_imputer.R</a>
script, which is largely based on the code written by Lasse Folkersen
for the website Impute.Me. In GenoPred, genotype imputation is performed
using shapeit and impute2, using the 1KG Phase 3 reference. The imputed
data is then converted to PLINK1 binary format, using PLINK2 and a
–hard-call-threshold of 0.1.</p>
<hr />
</div>
<div id="target-genotype-qc" class="section level2">
<h2>Target genotype QC</h2>
<p>Target genotype QC is performed using the <a
href="https://github.com/opain/GenoPred/blob/master/Scripts/format_target/format_target.R">format_target.R</a>
script. Each chromosome of the input data is processed in parallel.
Initially, reference and target sample variant data are read into R.
Then the genome build of the target data is identified through
comparison of CHR, BP, and IUPAC information in the reference data.
Then, variants in the target data are matched to those in the reference
data, identifying variants that are strand flipped, have mismatch RSIDs
and duplicate variants. PLINK2 is then used to create PLINK2 format
target genotype data, extracting variants in common with the reference,
updating IDs, and managing any needed strand flips. If the input target
genotype data is in VCF or BGEN format, either the PLINK –vcf-min-gq 10
or –hard-call-threshold 0.1 parameters are applied respectively.
Finally, reference variants that are not present in the target data are
inserted as missing, to allow reference allele frequency-based
imputation during downstream polygenic scoring.</p>
<hr />
</div>
<div id="ancestry-inference" class="section level2">
<h2>Ancestry Inference</h2>
<p>Target samples then undergo ancestry inference, using the <a
href="https://github.com/opain/GenoPred/blob/master/Scripts/Ancestry_identifier/Ancestry_identifier.R">Ancestry_identifier.R</a>
script, estimating the probability that each target individual matches
each reference population (AFR = African, AMR = Admixed American, EAS =
East Asian, EUR = European, CSA = Central and South Asian, MID = Middle
Eastern). Population membership was predicted using a reference trained
elastic net model consisting of the first six reference-projected
genetic principal components. Principal components were defined in the
reference dataset using variants present in the target dataset with a
minor allele frequency &gt;0.05, missingness &lt;0.02 and Hardy-Weinberg
p-value &gt;1×10-6 (if target sample size &lt;100, then only missingness
threshold is applied in the target). LD pruning for independent variants
is then performed in PLINK after removal of long-range LD regions (ref),
using a window size of 1000, step size of 5, and r2 threshold of 0.2.
The A multinomial elastic net model predicting super population
membership in the reference is derived in using the glmnet R package,
with model performance assessed using 5-fold cross validation. The
reference-derived principal components are then projected into the
target dataset, and the reference-derived elastic net model is used to
predict population membership in target. By default, target individuals
are assigned to a population if the predicted probability was &gt;0.95,
but the user can modify this threshold using the ancestry_prob_thresh
parameter in the config file. If an individual does not have a predicted
probability greater than the ancestry_prob_thresh parameter, then they
will be excluded from downstream polygenic scoring. If the
ancestry_prob_thresh parameter is low, then an individual may be
assigned to multiple reference populations, and they will have polygenic
scores that have been standardised according to each assigned reference
population. In this case, the individual-level report created by
GenoPred will present polygenic scores standardised according to the
reference population with the highest predicted probability.</p>
<hr />
</div>
<div id="within-target-qc" class="section level2">
<h2>Within-target QC</h2>
<p>GenoPred can also perform several quality control functions within
the target sample, using script <a
href="https://github.com/opain/GenoPred/blob/master/Scripts/outlier_detection/outlier_detection.R">outlier_detection.R</a>.
Note, this step is not reference-standardised, making it less suitable
for prediction modelling, but is still useful for accounting for
population stratification when inference is the aim.</p>
<p>If the user doesn’t provide a list of unrelated indiviuals using the
<code>unrel</code> column in the <code>target_list</code>, the
relatedness across the full target sample is estimated using the KING
estimator in PLINKv1.9. A list of unrelated individuals is defined using
a threshold of &gt;0.044. Then within each population with at least 100
individuals (assigned by the ancestry inference step), PCA is performed
to produce within-sample PCs (as opposed to reference-projected PCs) -
These may capture target sample specific structure that
reference-projected PCs will not, such as batch effects. PCA was applied
after LD-pruning and removal of long-range LD regions (ref). Using the
top 10 within-sample PCs, outliers are detected using the distance to
k-means centroids. PCs and outlier thresholds are defined using a
maximum of 1000 unrelated individuals, and then projected onto the full
target sample. The number of clusters within each population was defined
using the NbClust R package. Individuals were removed if their distance
from any one centroid was greater than Q3+30×IQR, where Q3 is the third
quartile, and IQR is the interquartile range. This threshold is
arbitrary, but was decided after visual inspection of centroid distance
distributions in UK Biobank and seems to generalise fairly well.</p>
<hr />
</div>
<div id="target-scoring" class="section level2">
<h2>Target scoring</h2>
<p>This step calculates scores in the target sample, based on scoring
files from the PGS methods, and from the reference PCA. It is carried
out using the <a
href="https://github.com/opain/GenoPred/blob/master/Scripts/target_scoring/target_scoring.R">target_scoring.R
script</a>. It is performed separately for each target sample
population. Scoring is performed using PLINK2 using the –read-freq
function to impute missing genotypes using the ancestry-matched
reference allele frequency. The scores are computed for each chromosome,
then multiplied by the number of non-missing variants to get the sum
score, and then added across chromosomes. Finally, the target sample
scores are standardised according to the mean and SD of the score in an
ancestry-matched reference.</p>
<hr />
</div>
<div id="report-creation" class="section level2">
<h2>Report creation</h2>
<hr />
<div id="individual-level" class="section level3">
<h3>Individual-level</h3>
<p>This step creates an .html report summarising the pipeline outputs
for each individual in the target sample. It simply reads in pipeline
outputs, and then tabulates and plots them. The only analysis it
performs is the conversion of polygenic scores onto the absolute scale.
It uses a <a href="https://pubmed.ncbi.nlm.nih.gov/34983942/">previously
published method</a>. The estimate of the PGS R2 come from the lassosum
pseudovalidation analysis, and the distribution in the general
population is provided by the user in the prev, mean and sd columns of
the gwas_list. Note: It does not convert PGS from externally derived
polygenic scores onto the absolutes scale. An example of the
individual-level report derived using the test data can be found
<a href="example_plink1-1_EUR.1_EUR-report.html" target="_blank">here</a>.</p>
<hr />
</div>
<div id="sample-level" class="section level3">
<h3>Sample-level</h3>
<p>This step creates an .html report summarising the pipeline outputs
for each target sample. It simply reads in pipeline outputs, and then
tabulates and plots them. An example of the sample-level report derived
using the test data can be found
<a href="example_plink1-report.html" target="_blank">here</a>.</p>
<hr />
</div>
</div>
</div>
<div id="any-questions" class="section level1">
<h1>Any questions?</h1>
<p>Please post questions as an issue on the GenoPred GitHub repo <a
href="https://github.com/opain/GenoPred/issues">here</a>.</p>
</div>

<!-- footer.html -->
<hr/>

<div class="centered-container">
<div class="rounded-image-container" style="width: 500px;">
<img src="Images/logo/sponsors.png">
</div>
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
