---
title: GenoPred Pipeline - Reference preparation
output: 
  html_document:
    theme: cosmo
    toc: true
    toc_float: true
    css: styles/styles.css
    includes:
      in_header: header.html
      after_body: footer.html
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(eval = FALSE)
library(knitr)
library(kableExtra)
library(data.table)
```

***

In this script we will prepare reference data for GenoPred based on a merged version of the 1KG and HGDP datasets.

We are going to start with the plink2 format version of these data created by the PGS catalogue team: https://ftp.ebi.ac.uk/pub/databases/spot/pgs/resources/pgsc_HGDP+1kGP_v1.tar.zst

PGS catalogue team used the data downloaded from gnomAD (v3.1.2): https://gnomad.broadinstitute.org/downloads#v3

The original gnomAD publication for these data is here: https://doi.org/10.1101%2F2023.01.23.525248

We are going to:

- Extract HapMap3 variants
- Convert the data to plink1 format
- Insert RSIDs
- Remove related individuals
- Split by chromosome
- Create keep files for each population
- Create freq files for each population
- Prepare .rds files summarising SNP information

***

# Download and decompress

```{bash}
# 1KG+HGDP data
cd /users/k1806347/oliverpainfel/Data/hgdp_1kg
wget https://ftp.ebi.ac.uk/pub/databases/spot/pgs/resources/pgsc_HGDP+1kGP_v1.tar.zst
tar --use-compress-program=unzstd -xvf pgsc_HGDP+1kGP_v1.tar.zst
unzstd GRCh37_HGDP+1kGP_ALL.pvar.zst

# Sample meta data
wget https://storage.googleapis.com/gcp-public-data--gnomad/release/3.1.2/vcf/genomes/gnomad.genomes.v3.1.2.hgdp_1kg_subset_sample_meta.tsv.bgz
gunzip -c gnomad.genomes.v3.1.2.hgdp_1kg_subset_sample_meta.tsv.bgz

```

***

# Format and output required reference files

```{r}

setwd('/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline')
source('../functions/misc.R')
source_all('../functions')
library(data.table)
library(GenoUtils)

# Read in variant data for 1KG+HGDP
pvar <- fread('/users/k1806347/oliverpainfel/Data/hgdp_1kg/GRCh37_HGDP+1kGP_ALL.pvar', nThread=5)

# Remove unneeded data and variant that aren't SNPs
pvar$INFO<-NULL
pvar$FILTER<-NULL
pvar<-pvar[nchar(pvar$REF) == 1,]
pvar<-pvar[nchar(pvar$ALT) == 1,]
names(pvar)<-c('CHR','BP','SNP','A1','A2')
pvar<-pvar[pvar$CHR %in% 1:22,]
pvar$CHR<-as.numeric(pvar$CHR)

# Read in the hapmap snp data
# Use the previously prepared 1KG reference data
hm3<-NULL
for(i in 1:22){
  hm3<-rbind(hm3, readRDS(paste0('/users/k1806347/oliverpainfel/test/GenoPred/pipeline/resources/data/ref/ref.chr',i,'.rds')))
}

hm3<-hm3[,c('CHR','SNP','BP_GRCh36','BP_GRCh37','BP_GRCh38','A1','A2'), with=F]
names(hm3) <- c('CHR','SNP','BP_GRCh36','BP','BP_GRCh38','A1','A2')

# Merge by chromosome and BP data
pvar_hm3<-merge(pvar, hm3, by=c('CHR','BP'))

# Check alleles match
pvar_hm3$IUPAC.x<-snp_iupac(pvar_hm3$A1.x, pvar_hm3$A2.x)
pvar_hm3$IUPAC.y<-snp_iupac(pvar_hm3$A1.y, pvar_hm3$A2.y)
flip <- detect_strand_flip(pvar_hm3$IUPAC.x, pvar_hm3$IUPAC.y)
sum(flip) # 0

# Retain variants that match IUPAC
pvar_hm3<-pvar_hm3[pvar_hm3$IUPAC.x == pvar_hm3$IUPAC.y,]

# Save intermediate file
write.table(pvar_hm3,'/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3_snps.txt', col.names = T, row.names = F, quote = F)

# Extract hm3 SNPs from plink2 files
dir.create('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred')
write.table(pvar_hm3$SNP.x, '/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/extract.snplist', col.names=F, row.names=F, quote=F)
system('plink2 --pfile /users/k1806347/oliverpainfel/Data/hgdp_1kg/GRCh37_HGDP+1kGP_ALL --make-pgen --extract /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/extract.snplist --out /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3 --threads 5')

# Insert RSIDs into new plink files
pvar <- fread('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3.pvar')
pvar <- pvar[, c('#CHROM','POS','ID','REF','ALT'), with=F]
pvar_hm3$ID <- pvar_hm3$SNP.x
pvar[pvar_hm3, on=.(ID), SNP := i.SNP.y]
pvar<-pvar[, c('#CHROM','POS','SNP','REF','ALT'), with=F]
names(pvar)<-c('#CHROM','POS','ID','REF','ALT')
fwrite(pvar, '/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3.pvar', col.names=T, row.names=F, quote=F, sep=' ')

# Remove hard filtered individuals
samp_meta<-fread('/users/k1806347/oliverpainfel/Data/hgdp_1kg/gnomad.genomes.v3.1.2.hgdp_1kg_subset_sample_meta.tsv.bgz')
samp_meta<-samp_meta[grepl("hard_filtered\":false", samp_meta$gnomad_sample_filters), ]

# Remove PCA outliers
samp_meta<-samp_meta[grepl("outlier\":false", samp_meta$hgdp_tgp_meta), ]

# Remove relatives (we will use the sample meta-data for this - Estimated using PC-relate)
library(jsonlite)
kin_dat <- NULL
for(i in 1:nrow(samp_meta)){
  tmp <- fromJSON(samp_meta$relatedness_inference[i])
  if(is.data.frame(tmp$related_samples)){
    tmp$related_samples$ID<-samp_meta$s[i]
    kin_dat<-rbind(kin_dat, tmp$related_samples)
  }
}

# Restrict table to individuals in the genetic data
psam <- fread('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3.psam')
psam<-psam[psam$`#IID` %in% samp_meta$s,]
kin_dat<-kin_dat[kin_dat$s %in% psam$`#IID` & kin_dat$ID %in% psam$`#IID`, ]
relations<-data.frame(id_1=kin_dat$s, id_2=kin_dat$ID, kin=kin_dat$kin)

# Use GreedyRelated to find maximum unrelated set
relations$Pair <- apply(relations, 1, function(row) {
  # Construct a unique identifier for each pair regardless of order
  paste(sort(c(row['id_1'], row['id_2'])), collapse = "_")
})
relations$Pair<-as.numeric(factor(relations$Pair))
relations$Factor<-relations$kin
relations$ID<-relations$id_1
relations<-relations[, c('ID', 'Pair', 'Factor')]
relations<-relations[order(relations$Pair), ]
fwrite(relations, file='/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/relations.txt', sep=' ', quote=F)

system('/scratch/prj/ukbiobank/KCL_Data/Software/tools/GreedyRelated-master-v1.2.1/GreedyRelated -r /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/relations.txt -t 0.05 -s 1 > /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/relations_remove.txt')

remove<-fread('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/relations_remove.txt')

# Create keep file of unrelaed individuals
psam_unrel <- psam[!(psam$`#IID` %in% remove$V2),]
nrow(psam_unrel) # 3313 - This is 65 less than in the preprint Table S3. Maybe Greedy related is worse?

write.table(psam_unrel[, '#IID', with=F], '/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/unrel.keep', col.names=T, row.names=F, quote=F)

# Split by chromosome and retain unrelated individuals
for(i in 1:22){
  system(paste0('plink2 --pfile /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3 --keep /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/unrel.keep --chr ',i,' --make-pgen --out /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3.chr',i))
}

# Create keep files and calculate allele frequencies for each population
# Population data is stored withint the psam file
psam <- fread('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3.chr22.psam')
pop_dat<-psam[, c('#IID','SuperPop')]
names(pop_dat)<-c('#IID','POP')
write.table(pop_dat, '/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/pop.txt', row.names=F, col.names=T, quote=F)

dir.create('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/keep_files/')
for(pop in unique(pop_dat$POP)){
  write.table(pop_dat[pop_dat$POP == pop, '#IID', with=F], paste0('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/keep_files/',pop,'.keep'), col.names=F, row.names=F, quote=F)
  dir.create(paste0('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/freq_files/', pop), recursive = T)
  for(i in 1:22){
    system(paste0('plink2 --pfile /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3.chr',i,' --keep /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/keep_files/',pop,'.keep --chr ',i,' --freq --out /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/freq_files/', pop,'/hgdp_1kg_hm3.chr', i))
  }
}

# Create frequency files across all reference individuals (TRANS)
dir.create(paste0('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/freq_files/TRANS'), recursive = T)
for(i in 1:22){
  system(paste0('plink2 --pfile /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3.chr',i,' --chr ',i,' --freq --out /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/freq_files/TRANS/hgdp_1kg_hm3.chr', i))
}

####
# Create .rds files for SNP stats
####

pvar_hm3<-pvar_hm3[,c('CHR','BP','SNP.y','A1.x','A2.x','BP_GRCh36','BP_GRCh38'), with=F]
names(pvar_hm3)<-c('#CHROM','POS_GRCh37','ID','ALT','REF','POS_GRCh36','POS_GRCh38')

for(chr in 1:22){
    ref<-pvar_hm3[pvar_hm3$`#CHROM` == chr,]
    ref$IUPAC<-snp_iupac(ref$ALT, ref$REF)
    for(pop in unique(pop_dat$POP)){
        # Read in reference frequency data
        freq<-fread(paste0('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/freq_files/', pop,'/hgdp_1kg_hm3.chr', chr,'.afreq'))

        # The freq files have come from the reference files, so we can assume they are on the same strand
        freq_match<-merge(ref, freq[,c('ID','ALT','REF','ALT_FREQS'), with=F], by=c('ID','ALT','REF'))
        freq_swap<-merge(ref, freq[,c('ID','ALT','REF','ALT_FREQS'), with=F], by.x=c('ID','ALT','REF'), by.y=c('ID','REF','ALT'))
        freq_swap$ALT_FREQS<-1-freq_swap$ALT_FREQS
        tmp_freq<-rbind(freq_match, freq_swap)
        tmp_freq<-tmp_freq[match(ref$ID, tmp_freq$ID),]

        ref[[paste0('REF.FRQ.',pop)]]<-tmp_freq$ALT_FREQS
    }

    ref<-ref[,c("#CHROM","ID","POS_GRCh36","POS_GRCh37","POS_GRCh38","ALT","REF","IUPAC",paste0('REF.FRQ.',unique(pop_dat$POP))), with=F]
    names(ref)<-c("CHR","SNP","BP_GRCh36","BP_GRCh37","BP_GRCh38","A1","A2","IUPAC",paste0('REF.FRQ.',unique(pop_dat$POP)))
    saveRDS(ref, file = paste0('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3.chr',chr,'.rds'))
}

# Create keeplist
keep_list<-data.frame(POP=unique(pop_dat$POP),
                      PATH=paste0('resources/data/ref/keep_files/',unique(pop_dat$POP),'.keep'))

write.table(keep_list, '/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref.keep.list', col.names=F, row.names=F, quote=F)

```

***

## Package for zenodo

<ref_dir>
├── ref.chr<1-22>.<pgen/pvar/psam>
├── ref.chr<1-22>.rds
├── ref.pop.txt (#IID, POP - with header)
├── ref.keep.list (POP and PATH - without header)
├── keep_files
│   └──<POP>.keep (#IID - with header)
└── freq_files
    └──<POP>
        └──ref.<POP>.chr<CHR>.afreq # PLINK2 .afreq format

```{bash}

# Copy over all the relavent files
mkdir /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref

for chr in $(seq 1 22); do
  for file in $(echo pgen pvar psam rds); do
    cp /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/hgdp_1kg_hm3.chr${chr}.${file} /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref/ref.chr${chr}.${file}
  done
done

cp /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/pop.txt /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref/ref.pop.txt

mkdir /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref/keep_files

cp /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/keep_files/*.keep /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref/keep_files/

cp -r /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/freq_files /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref/

for pop in $(ls /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref/freq_files); do
 for chr in $(seq 1 22); do
  mv /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref/freq_files/${pop}/hgdp_1kg_hm3.chr${chr}.afreq /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref/freq_files/${pop}/ref.${pop}.chr${chr}.afreq
  rm /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref/freq_files/${pop}/hgdp_1kg_hm3.chr${chr}.log
 done
done

cp /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref.keep.list /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/ref/ref.keep.list

# Compress folder and upload to zenodo
cd /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred
tar -czvf genopred_1kg_hgdp.tar.gz ref

```

***

# Create subset of Pan UKB LD scores

More information on the data: https://pan.ukbb.broadinstitute.org/downloads/index.html
Original citation: Pan-UKB team. https://pan.ukbb.broadinstitute.org. 2020.

```{bash}

mkdir -p /users/k1806347/oliverpainfel/Data/ld_scores
cd /users/k1806347/oliverpainfel/Data/ld_scores
wget --no-check-certificate https://pan-ukb-us-east-1.s3.amazonaws.com/ld_release/UKBB.ALL.ldscore.tar.gz
tar -xvf UKBB.ALL.ldscore.tar.gz
rm UKBB.ALL.ldscore.tar.gz
for population in $(echo EUR EAS CSA AMR MID AFR); do
  mv UKBB.ALL.ldscore/UKBB.$population.l2.M_5_50 UKBB.$population.rsid.l2.M_5_50
  mv UKBB.ALL.ldscore/UKBB.$population.rsid.l2.ldscore.gz UKBB.$population.rsid.l2.ldscore.gz
done
rm -r UKBB.ALL.ldscore
cd ..
tar -czvf ld_scores.tar.gz ld_scores

```

***

# Make dense version of the reference data

By default, GenoPred uses a reference dataset restricted to HapMap3 variants to improve SNP overlap across datasets, and also improve computational efficiency while maintaining decent coverage. However, it may be of interest to use a reference with denser coverage of the genome. For example, when using externally derived PGS score files, such as those on PGS catalogue, which are not restricted to HapMap3 variants. It could also be desired to instead restrict PGS analyses to variants present in the target dataset, although this departs from the reference standardised approach and may limit generalisability of models across datasets. Therefore, here we will create a reference dataset for GenoPred that provides denser coverage, including all autosomal nonambiguous SNPs with a MAF > 0.01 in any reference population.

***

## Common variants in dbSNP

dbSNP has a list of common variants, with a MAF > 0.01 in at least 1KG super population. Create an LD reference using this list.

***

### Download dbSNP data

```{bash}
# Download the dbSNP data for common variants (MAF > 0.01 in at least 1 population in 1KG)
mkdir -p ~/oliverpainfel/Data/dbSNP
cd ~/oliverpainfel/Data/dbSNP
wget https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh37p13/VCF/00-common_all.vcf.gz

# Subset CHR BP RSID A1 and A2 information
zgrep -v "^#" ~/oliverpainfel/Data/dbSNP/00-common_all.vcf.gz | awk '{print $1, $2, $3, $4, $5}' > ~/oliverpainfel/Data/dbSNP/00-common_all.vcf.mini

# Split multi alleleic variants onto seperate lines
# Retain SNPs only
awk '{
  n = split($5, alts, ",");
  for (i = 1; i <= n; i++) {
    if (length($4) == 1 && length(alts[i]) == 1) {
      print $1, $2, $3, $4, alts[i]
    }
  }
}' ~/oliverpainfel/Data/dbSNP/00-common_all.vcf.mini > ~/oliverpainfel/Data/dbSNP/00-common_all.vcf.mini.split

```

***

### Subset 1KG+HGDP data to common variants and insert RSIDs

```{r}

setwd('/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline')
source('../functions/misc.R')
source_all('../functions')
library(data.table)
library(GenoUtils)

dir.create('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense')

# Read in variant data for 1KG+HGDP
pvar <- fread('/users/k1806347/oliverpainfel/Data/hgdp_1kg/GRCh37_HGDP+1kGP_ALL.pvar', nThread=5)

# Remove unneeded data and variant that aren't non ambiguous autosomal SNPs
pvar$INFO<-NULL
pvar$FILTER<-NULL
pvar<-pvar[nchar(pvar$REF) == 1,]
pvar<-pvar[nchar(pvar$ALT) == 1,]
names(pvar)<-c('CHR','BP','SNP','A1','A2')
pvar<-pvar[pvar$CHR %in% 1:22,]
pvar$CHR<-as.numeric(pvar$CHR)
pvar$IUPAC<-snp_iupac(pvar$A1, pvar$A2)
pvar <- pvar[pvar$IUPAC %in% c('R','Y','K','M'),]

# Read in dbSNP data and filter in the same way
dbsnp<-fread('~/oliverpainfel/Data/dbSNP/00-common_all.vcf.mini.split', header=F)
names(dbsnp)<-c('CHR','BP','SNP','A1','A2')
dbsnp<-dbsnp[dbsnp$CHR %in% 1:22,]
dbsnp$CHR<-as.numeric(dbsnp$CHR)
dbsnp$IUPAC<-snp_iupac(dbsnp$A1, dbsnp$A2)
dbsnp <- dbsnp[dbsnp$IUPAC %in% c('R','Y','K','M'),]

# Check overlap with HapMap3 variants
hapmap3_ref <- read_pvar('~/oliverpainfel/Software/MyGit/GenoPred/pipeline/resources/data/ref/ref.chr')
sum(paste0(dbsnp$CHR,':', dbsnp$BP) %in% paste0(hapmap3_ref$CHR,':', hapmap3_ref$BP)) # 1303109
# There appears to be good coverage of HapMap3 variants

# Merge the pvar data with the dbSNP data allowing for strand flips
ref_target <- merge(pvar, dbsnp, by=c('CHR','BP'))
flip_logical<-detect_strand_flip(targ = ref_target$IUPAC.x, ref = ref_target$IUPAC.y)
sum(flip_logical) # There are no strand flips

# Identify SNPs that have matched IUPAC
ref_target<-ref_target[ref_target$IUPAC.x == ref_target$IUPAC.y,]

# Clean up pvar
pvar<-ref_target[, c('CHR','BP','SNP.x','SNP.y','A1.x','A2.x'), with=F]
names(pvar)<-c('CHR','BP','ID','SNP','A1','A2')

sum(paste0(pvar$CHR,':', pvar$BP) %in% paste0(hapmap3_ref$CHR,':', hapmap3_ref$BP)) # 1208667
sum(pvar$SNP %in% hapmap3_ref$SNP) # 1208618
# There still appears to be good coverage of HapMap3 variants

# Remove duplicates preferentially keeping the hapmap3 variants
hapmap3_ref$IUPAC <- snp_iupac(hapmap3_ref$A1, hapmap3_ref$A2)
pvar$IUPAC <- snp_iupac(pvar$A1, pvar$A2)
pvar<-merge(pvar, hapmap3_ref, all.x=T, by = 'SNP')
pvar$in_hm3<-pvar$IUPAC.x == pvar$IUPAC.y
pvar$in_hm3[is.na(pvar$in_hm3)] <- F
pvar<-pvar[rev(order(pvar$in_hm3)),]

nrow(pvar) # 27023331
pvar<-pvar[!duplicated(pvar$SNP),]
nrow(pvar) # 26828676
pvar<-pvar[!duplicated(paste0(pvar$CHR,':',pvar$BP)),]
nrow(pvar) # 26828676
sum(pvar$in_hm3) # 1204008 of 1204449. 99.96% of variants present.

pvar<-pvar[, c('CHR.x','BP.x','ID','SNP','A1.x','A2.x'), with=F]
names(pvar)<-c('CHR','BP','ID','SNP','A1','A2')

# Extract SNPs from plink2 files
write.table(pvar$ID, '/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/extract.snplist', col.names=F, row.names=F, quote=F)

# Save intermediate file
fwrite(pvar,'/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/hgdp_1kg_nonambig_snps.txt', sep = '\t', quote = F, na='NA')

```

```{bash}
for chr in $(seq 1 22); do
  sbatch --mem 10G -n 5 -p neurohack_cpu -t 1:00:00 --wrap="plink2 \
    --pfile /users/k1806347/oliverpainfel/Data/hgdp_1kg/GRCh37_HGDP+1kGP_ALL \
    --keep /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/unrel.keep \
    --extract /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/extract.snplist \
    --chr ${chr} \
    --threads 5 \
    --make-pgen \
    --out /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/hgdp_1kg.chr${chr}"
done
```

```{r}
library(data.table)

# Insert RSIDs into new plink files
ref <- fread('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/hgdp_1kg_nonambig_snps.txt', nThread=5)

for(i in 1:22){
  print(i)
  pvar <- fread(paste0('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/hgdp_1kg.chr',i,'.pvar'), nThread=5)
  pvar <- pvar[, c('#CHROM','POS','ID','REF','ALT'), with=F]
  pvar[ref, on=.(ID), ID := i.SNP]
  fwrite(
    pvar,
    paste0('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/hgdp_1kg.chr',i,'.pvar'),
    col.names = T,
    row.names = F,
    quote = F,
    sep = ' ',
    nThread=5
  )
  rm(pvar)
  gc()
}

```

```{bash}
# Calculate allele frequencies for each population
# Use keep files from previous reference preparation
for pop in $(echo EUR EAS AFR AMR CSA MID); do
  mkdir -p /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/freq_files/${pop}
  
  for chr in $(seq 1 22); do
    sbatch --mem 10G -n 1 -p neurohack_cpu -t 1:00:00 --wrap="plink2 \
      --pfile /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/hgdp_1kg.chr${chr} \
      --keep /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/keep_files/${pop}.keep \
      --freq \
      --out /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/freq_files/${pop}/hgdp_1kg.chr${chr}"
      
  done
done

# Create frequency files across all reference individuals (TRANS)
mkdir -p /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/freq_files/TRANS
for chr in $(seq 1 22); do
  sbatch --mem 10G -n 1 -p neurohack_cpu -t 1:00:00 --wrap="plink2 \
    --pfile /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/hgdp_1kg.chr${chr} \
    --freq \
    --out /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/freq_files/TRANS/hgdp_1kg.chr${chr}"
done
```

```{r}
####
# Create .rds files for SNP stats
####
setwd('/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline')
source('../functions/misc.R')
source_all('../functions')
library(data.table)
library(GenoUtils)

# Make liftover functions
make_executable <- function(exe) {
  Sys.chmod(exe, mode = (file.info(exe)$mode | "111"))
}

snp_modifyBuild_offline<-function (info_snp, liftOver, chain){
  tmp_folder<-tempdir()

  if (!all(c("chr", "pos") %in% names(info_snp)))
    stop2("Please use proper names for variables in 'info_snp'. Expected %s.",
          "'chr' and 'pos'")
  liftOver <- normalizePath(liftOver)
  make_executable(liftOver)
  BED <- tempfile(fileext = ".BED")
  info_BED <- with(info_snp, data.frame(paste0("chr", chr),
                                        pos0 = pos - 1L, pos, id = seq_len(nrow(info_snp))))
  fwrite(info_BED, BED, col.names = FALSE, sep = " ")
  lifted<-paste0(tmp_folder,'/tmp.lifted')
  unmapped<-paste0(tmp_folder,'/tmp.unmapped')
  system(paste(liftOver, BED, chain, lifted, unmapped))
  new_pos <- fread(lifted)
  bad <- grep("^#", readLines(unmapped), value = TRUE, invert = TRUE)
  print(paste0(length(bad)," variants have not been mapped."))
  info_snp$pos <- NA
  info_snp$pos[new_pos$V4] <- new_pos$V3
  info_snp
}

for(i in 1:22){
  print(i)
  pvar <- fread(paste0('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/hgdp_1kg.chr',i,'.pvar'), nThread=5)
  
  # Insert BP for GRCh36 and 38
  names(pvar) <- c('chr','pos','snp','a2','a1')

  pvar_grch38<-snp_modifyBuild_offline(pvar, liftOver='/users/k1806347/oliverpainfel/Software/MyGit/GenoDisc/pipeline/resources/software/liftover/liftover', chain='/users/k1806347/oliverpainfel/Software/MyGit/GenoDisc/pipeline/resources/data/liftover/hg19ToHg38.over.chain.gz')
  
  pvar_grch36<-snp_modifyBuild_offline(pvar, liftOver='/users/k1806347/oliverpainfel/Software/MyGit/GenoDisc/pipeline/resources/software/liftover/liftover', chain='/users/k1806347/oliverpainfel/Software/MyGit/GenoDisc/pipeline/resources/data/liftover/hg19ToHg18.over.chain.gz')

  pvar[pvar_grch36, on=.(snp), pos_grch36 := i.pos]
  pvar[pvar_grch38, on=.(snp), pos_grch38 := i.pos]
  
  rm(pvar_grch36)
  rm(pvar_grch38)
  gc()
  
  pvar <- pvar[, c('chr','snp','a1','a2','pos_grch36','pos','pos_grch38'), with=F]
  names(pvar) <- c('CHR','SNP','A1','A2','BP_GRCh36','BP_GRCh37','BP_GRCh38')
  
  pvar$IUPAC<-snp_iupac(pvar$A1, pvar$A2)

  # Read in the FRQ data
  for(pop in c('EUR','EAS','AFR','AMR','CSA','MID')){
    freq <-
      fread(
        paste0(
          '/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/freq_files/',
          pop,
          '/hgdp_1kg.chr',
          i,
          '.afreq'
        ), nThread=5
      )
    
    names(freq)<-c('CHR','SNP','A2','A1','FREQ','OBS')
    freq$IUPAC<-snp_iupac(freq$A1, freq$A2)
    
    pvar<-merge(pvar, freq[, c('SNP','A1','FREQ'), with=F], by = 'SNP', sort = F)
    pvar$FREQ[pvar$A1.x != pvar$A1.y] <- 1-pvar$FREQ[pvar$A1.x != pvar$A1.y]
    
    pvar$A1.y <- NULL
    names(pvar)[names(pvar) == 'A1.x'] <- 'A1'
    pvar$FREQ <- round(pvar$FREQ, 5)
    names(pvar)[names(pvar) == 'FREQ'] <- paste0('REF.FRQ.', pop)
  }
  
  pvar <- pvar[, c("CHR", "SNP", "BP_GRCh36", "BP_GRCh37", "BP_GRCh38", "A1", "A2", "IUPAC", "REF.FRQ.EUR", "REF.FRQ.EAS", "REF.FRQ.AFR", "REF.FRQ.AMR", "REF.FRQ.CSA", "REF.FRQ.MID"), with=F]
  
  saveRDS(pvar, file = paste0('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/hgdp_1kg.chr', i,'.rds'))
}

```

***

### Package for zenodo

<ref_dir>
├── ref.chr<1-22>.<pgen/pvar/psam>
├── ref.chr<1-22>.rds
├── ref.pop.txt (#IID, POP - with header)
├── keep_files
│   └──<POP>.keep (#IID - with header)
└── freq_files
    └──<POP>
        └──ref.<POP>.chr<CHR>.afreq # PLINK2 .afreq format

```{bash}

# Copy over all the relavent files
mkdir /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/ref

for chr in $(seq 1 22); do
  for file in $(echo pgen pvar psam rds); do
    cp /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/hgdp_1kg.chr${chr}.${file} /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/ref/ref.chr${chr}.${file}
  done
done

cp /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/pop.txt /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/ref/ref.pop.txt

mkdir /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/ref/keep_files

cp /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/keep_files/*.keep /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/ref/keep_files/

cp -r /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/freq_files /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/ref/

for pop in $(ls /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/ref/freq_files); do
 for chr in $(seq 1 22); do
  mv /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/ref/freq_files/${pop}/hgdp_1kg.chr${chr}.afreq /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/ref/freq_files/${pop}/ref.${pop}.chr${chr}.afreq
  rm /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/ref/freq_files/${pop}/hgdp_1kg.chr${chr}.log
 done
done

# Compress folder and upload to google drive
cd /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense
tar -czvf genopred_dense_1kg_hgdp.tar.gz ref
~/oliverpainfel/Software/gdrive files upload genopred_dense_1kg_hgdp.tar.gz
```

***

## All non-ambiguous variants

The 'dense' reference dataset created above, restricted to common biallelelic non-ambiguous SNPs, will still miss some variants that may be used in externally derived score files, such as those from the PGS Catalog. The ultimate solution would be to use a reference with all non-ambiguous variants present - incl. rarer variants, indels, and multi-allelic variants. Multi-allelic variants will require us to move away from RSIDs, and instead use CHR:BP:A1:A2, or RSID:A1:A2 IDs. We should still store RSIDs though for matching to external datasets without CHR and BP information. INDELS will require us to move away from IUPAC codes. Create the reference data, and then we can see what needs to be updated in GenoPred to cope with the changes.

***

### Prepare dbSNP data

```{bash}
# Download the full dbSNP data
mkdir -p ~/oliverpainfel/Data/dbSNP
cd ~/oliverpainfel/Data/dbSNP
wget https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh37p13/VCF/00-All.vcf.gz

# Index as VCF
module load bcftools
tabix -p vcf ~/oliverpainfel/Data/dbSNP/00-All.vcf.gz

# Split multi allelic variants onto separate rows
# Remove ambiguous SNPs
bcftools norm -m -any --threads 10 -Ou ~/oliverpainfel/Data/dbSNP/00-All.vcf.gz \
  | bcftools view -e 'strlen(REF)==1 && strlen(ALT)==1 && ((REF="A" && ALT="T") || (REF="T" && ALT="A") || (REF="C" && ALT="G") || (REF="G" && ALT="C"))' -Ou \
  | bcftools query -f '%CHROM\t%POS\t%ID\t%REF\t%ALT\n' \
  > ~/oliverpainfel/Data/dbSNP/00-All.nonambiguous.tsv
  
```

***

### Subset 1KG+HGDP data to non-ambiguous variants and insert RSIDs

```{r}

setwd('/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline')
source('../functions/misc.R')
source_all('../functions')
library(data.table)
library(GenoUtils)

dir.create('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel')

# Read in variant data for 1KG+HGDP
pvar <- fread('/users/k1806347/oliverpainfel/Data/hgdp_1kg/GRCh37_HGDP+1kGP_ALL.pvar', nThread=5)

# Remove ambiguous variants
pvar$INFO<-NULL
pvar$FILTER<-NULL
names(pvar)<-c('CHR','BP','ID','A1','A2')
pvar<-pvar[!(pvar$A1 == 'T' & pvar$A2 == 'A' | pvar$A1 == 'A' & pvar$A2 == 'T') &
           !(pvar$A1 == 'C' & pvar$A2 == 'G' | pvar$A1 == 'G' & pvar$A2 == 'C'),]

# Read in dbSNP data
dbsnp <- fread('~/oliverpainfel/Data/dbSNP/00-All.nonambiguous.tsv', nThread=5)
names(dbsnp)<-c('CHR','BP','RSID','A1','A2')

# Merge the pvar data with the dbSNP data allowing for strand flips
ref_target <- merge(pvar, dbsnp, by=c('CHR','BP'))

# Fast, vectorised reverse-complement (IUPAC-aware, uses stringi if present)
rc <- function(x){
  x <- chartr("ACGTacgt",
              "TGCAtgca", x)
  if (requireNamespace("stringi", quietly = TRUE)) stringi::stri_reverse(x)
  else vapply(strsplit(x, "", fixed = TRUE),
              function(z) paste0(rev(z), collapse = ""), "")
}

# 1) quick matches without RC
match_rows <- ref_target[, A1.x == A1.y & A2.x == A2.y]
swapped    <- ref_target[, A1.x == A2.y & A2.x == A1.y]
match_rows <- match_rows | swapped

# 2) only RC-check the rows still unmatched, with memoisation
idx <- which(!match_rows)
u <- unique(unlist(ref_target[idx, .(A1.x, A2.x)]))   # unique alleles to RC once
rc_map <- setNames(rc(u), u)
A1x_rc <- rc_map[ ref_target$A1.x[idx] ]
A2x_rc <- rc_map[ ref_target$A2.x[idx] ]

rc_same <- (A1x_rc == ref_target$A1.y[idx]) & (A2x_rc == ref_target$A2.y[idx])
rc_swap <- (A1x_rc == ref_target$A2.y[idx]) & (A2x_rc == ref_target$A1.y[idx])
match_rows[idx] <- rc_same | rc_swap
# Note. there were no reverse complements

ref_target<-ref_target[match_rows,]

# Clean up pvar
pvar<-ref_target[, c('CHR','BP','ID','RSID','A1.x','A2.x'), with=F]
names(pvar)<-c('CHR','BP','ID','RSID','A1','A2')

# Remove duplicated variants
pvar[, allele_pair := ifelse(A1 < A2,
                             paste(A1, A2, sep = ":"),
                             paste(A2, A1, sep = ":"))]
pvar<-pvar[!duplicated(paste0(pvar$RSID, ':', pvar$allele_pair)),]

# Note. some variants have multiple RSIDs. Keep them to improve overlap with datasets that only contain RSID.

# Extract SNPs from plink2 files
write.table(unique(pvar$ID), '/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/extract.snplist', col.names=F, row.names=F, quote=F)

# Save intermediate file
fwrite(pvar,'/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/hgdp_1kg_nonambig.txt', sep = '\t', quote = F, na='NA')

```

```{bash}
for chr in $(seq 1 22) X Y; do
  sbatch --mem 10G -n 5 -p neurohack_cpu -t 1:00:00 --wrap="plink2 \
    --pfile /users/k1806347/oliverpainfel/Data/hgdp_1kg/GRCh37_HGDP+1kGP_ALL \
    --keep /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/unrel.keep \
    --extract /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/extract.snplist \
    --chr ${chr} \
    --threads 5 \
    --make-pgen \
    --out /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/hgdp_1kg.chr${chr}"
done

# Delete the FILTER and INFO columns from the pvar files
for chr in $(seq 1 22) X Y; do
  awk 'BEGIN{FS=OFS="\t"} $0 !~ /^##/ {print $1,$2,$3,$4,$5}' /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/hgdp_1kg.chr${chr}.pvar > /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/hgdp_1kg.chr${chr}.pvar.tmp && mv /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/hgdp_1kg.chr${chr}.pvar.tmp /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/hgdp_1kg.chr${chr}.pvar
done

```

```{bash}
# Calculate allele frequencies for each population
# Use keep files from previous reference preparation
for pop in $(echo EUR EAS AFR AMR CSA MID); do
  mkdir -p /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/freq_files/${pop}
  
for chr in $(seq 1 22) X Y; do
    sbatch --mem 10G -n 1 -p neurohack_cpu -t 1:00:00 --wrap="plink2 \
      --pfile /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/hgdp_1kg.chr${chr} \
      --keep /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/keep_files/${pop}.keep \
      --freq \
      --out /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/freq_files/${pop}/hgdp_1kg.chr${chr}"
      
  done
done

# Create frequency files across all reference individuals (TRANS)
mkdir -p /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/freq_files/TRANS
for chr in $(seq 1 22) X Y; do
  sbatch --mem 10G -n 1 -p neurohack_cpu -t 1:00:00 --wrap="plink2 \
    --pfile /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/hgdp_1kg.chr${chr} \
    --freq \
    --out /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/freq_files/TRANS/hgdp_1kg.chr${chr}"
done
```

```{r}
####
# Create .rds files for SNP stats
####

setwd('/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline')
source('../functions/misc.R')
source_all('../functions')
library(data.table)
library(GenoUtils)

# Make liftover functions
make_executable <- function(exe) {
  Sys.chmod(exe, mode = (file.info(exe)$mode | "111"))
}

snp_modifyBuild_offline<-function (info_snp, liftOver, chain){
  tmp_folder<-tempdir()

  if (!all(c("chr", "pos") %in% names(info_snp)))
    stop2("Please use proper names for variables in 'info_snp'. Expected %s.",
          "'chr' and 'pos'")
  liftOver <- normalizePath(liftOver)
  make_executable(liftOver)
  BED <- tempfile(fileext = ".BED")
  info_BED <- with(info_snp, data.frame(paste0("chr", chr),
                                        pos0 = pos - 1L, pos, id = seq_len(nrow(info_snp))))
  fwrite(info_BED, BED, col.names = FALSE, sep = " ")
  lifted<-paste0(tmp_folder,'/tmp.lifted')
  unmapped<-paste0(tmp_folder,'/tmp.unmapped')
  system(paste(liftOver, BED, chain, lifted, unmapped))
  new_pos <- fread(lifted)
  bad <- grep("^#", readLines(unmapped), value = TRUE, invert = TRUE)
  print(paste0(length(bad)," variants have not been mapped."))
  info_snp$pos <- NA
  info_snp$pos[new_pos$V4] <- new_pos$V3
  info_snp
}

snp_info<-fread('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/hgdp_1kg_nonambig.txt', nThread=10)
snp_info<-snp_info[, c('ID','RSID'), with = F]

for(i in c(1:22, 'X', 'Y')){
  print(i)
  pvar <- fread(paste0('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/hgdp_1kg.chr',i,'.pvar'), nThread=10)
  
  # Insert BP for GRCh36 and 38
  names(pvar) <- c('chr','pos','snp','a2','a1')

  pvar_grch38<-snp_modifyBuild_offline(pvar, liftOver='/users/k1806347/oliverpainfel/Software/MyGit/GenoDisc/pipeline/resources/software/liftover/liftover', chain='/users/k1806347/oliverpainfel/Software/MyGit/GenoDisc/pipeline/resources/data/liftover/hg19ToHg38.over.chain.gz')
  
  pvar_grch36<-snp_modifyBuild_offline(pvar, liftOver='/users/k1806347/oliverpainfel/Software/MyGit/GenoDisc/pipeline/resources/software/liftover/liftover', chain='/users/k1806347/oliverpainfel/Software/MyGit/GenoDisc/pipeline/resources/data/liftover/hg19ToHg18.over.chain.gz')

  pvar[pvar_grch36, on=.(snp), pos_grch36 := i.pos]
  pvar[pvar_grch38, on=.(snp), pos_grch38 := i.pos]
  
  rm(pvar_grch36)
  rm(pvar_grch38)
  gc()
  
  pvar <- pvar[, c('chr','snp','a1','a2','pos_grch36','pos','pos_grch38'), with=F]
  names(pvar) <- c('CHR','ID','A1','A2','BP_GRCh36','BP_GRCh37','BP_GRCh38')
  
  # Insert RSID
  pvar <- snp_info[pvar, on=.(ID), allow.cartesian=TRUE]

  # Read in the FRQ data
  for(pop in c('EUR','EAS','AFR','AMR','CSA','MID')){
    freq <-
      fread(
        paste0(
          '/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/freq_files/',
          pop,
          '/hgdp_1kg.chr',
          i,
          '.afreq'
        ), nThread=10
      )
    
    names(freq)<-c('CHR','ID','A2','A1','FREQ','OBS')
    pvar<-merge(pvar, freq[, c('ID','A1','FREQ'), with=F], by = 'ID', sort = F)
    pvar$FREQ[pvar$A1.x != pvar$A1.y] <- 1-pvar$FREQ[pvar$A1.x != pvar$A1.y]
    
    pvar$A1.y <- NULL
    names(pvar)[names(pvar) == 'A1.x'] <- 'A1'
    pvar$FREQ <- round(pvar$FREQ, 5)
    names(pvar)[names(pvar) == 'FREQ'] <- paste0('REF.FRQ.', pop)
  }
  
  pvar <- pvar[, c("CHR", "ID", "RSID", "BP_GRCh36", "BP_GRCh37", "BP_GRCh38", "A1", "A2", "REF.FRQ.EUR", "REF.FRQ.EAS", "REF.FRQ.AFR", "REF.FRQ.AMR", "REF.FRQ.CSA", "REF.FRQ.MID"), with=F]
  
  saveRDS(pvar, file = paste0('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/hgdp_1kg.chr', i,'.rds'))
}

```

***

### Package for zenodo

<ref_dir>
├── ref.chr<1-22>.<pgen/pvar/psam>
├── ref.chr<1-22>.rds
├── ref.pop.txt (#IID, POP - with header)
├── keep_files
│   └──<POP>.keep (#IID - with header)
└── freq_files
    └──<POP>
        └──ref.<POP>.chr<CHR>.afreq # PLINK2 .afreq format

```{bash}

# Copy over all the relavent files
mkdir /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/ref

for chr in $(seq 1 22) X Y; do
  for file in $(echo pgen pvar psam rds); do
    cp /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/hgdp_1kg.chr${chr}.${file} /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/ref/ref.chr${chr}.${file}
  done
done

cp /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/pop.txt /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/ref/ref.pop.txt

mkdir /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/ref/keep_files

cp /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/keep_files/*.keep /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/ref/keep_files/

cp -r /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/freq_files /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/ref/

for pop in $(ls /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/ref/freq_files); do
  for chr in $(seq 1 22) X Y; do
  mv /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/ref/freq_files/${pop}/hgdp_1kg.chr${chr}.afreq /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/ref/freq_files/${pop}/ref.${pop}.chr${chr}.afreq
  rm /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel/ref/freq_files/${pop}/hgdp_1kg.chr${chr}.log
 done
done

# Compress folder and upload to google drive
cd /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense_multi_indel
tar -czvf genopred_dense_multi_indel_1kg_hgdp.tar.gz ref
~/oliverpainfel/Software/gdrive files upload genopred_dense_multi_indel_1kg_hgdp.tar.gz
```

***

# Make a full set of reference data

Here will will extract information for all SNPs present in the 1KG+HGDP reference. Ultimately, we want .rds files for each chromosome, containing CHR, BP (in builds GRCh36,37,38), SNP, A1, A2. This will be used to harmonise external datasets without loosing SNPs unnessecearily. 

***

## Make .rds files

```{bash}
# Download the full dbSNP data
mkdir -p ~/oliverpainfel/Data/dbSNP
cd ~/oliverpainfel/Data/dbSNP
wget https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh37p13/VCF/00-All.vcf.gz

# Index as VCF
module load bcftools
tabix -p vcf ~/oliverpainfel/Data/dbSNP/00-All.vcf.gz

# Split multi allelic variants onto separate rows
# Only retain non-ambiguous SNPs
bcftools norm -m -any --threads 10 -Ou ~/oliverpainfel/Data/dbSNP/00-All.vcf.gz \
  | bcftools view -v snps -Ou \
  | bcftools query -f '%CHROM\t%POS\t%ID\t%REF\t%ALT\n' \
  | awk '!($4$5 ~ /^(AT|TA|CG|GC)$/)' \
  > ~/oliverpainfel/Data/dbSNP/00-All.snps.nonambiguous.tsv

# Split into chr 1-22
for chr in {1..22}; do
  awk -v c="$chr" '$1 == c' ~/oliverpainfel/Data/dbSNP/00-All.snps.nonambiguous.tsv \
    > ~/oliverpainfel/Data/dbSNP/00-All.snps.nonambiguous.chr${chr}.tsv
done
```

```{r}
####
# Insert BP positions across builds (GRCh36, 37, 38) using liftover, and write out as .rds file
####
library(foreach)
library(doMC)
registerDoMC(5)
library(data.table)
setDTthreads(1)  # avoid oversubscription inside foreach workers
library(GenoUtils)

# Make liftover functions
make_executable <- function(exe) {
  Sys.chmod(exe, mode = (file.info(exe)$mode | "111"))
}

snp_modifyBuild_offline<-function (info_snp, liftOver, chain){
  if (!all(c("chr", "pos") %in% names(info_snp)))
    stop2("Please use proper names for variables in 'info_snp'. Expected %s.",
          "'chr' and 'pos'")
  liftOver <- normalizePath(liftOver)
  make_executable(liftOver)
  BED <- tempfile(fileext = ".BED")
  info_BED <- with(info_snp, data.frame(paste0("chr", chr),
                                        pos0 = pos - 1L, pos, id = seq_len(nrow(info_snp))))
  fwrite(info_BED, BED, col.names = FALSE, sep = " ")
  lifted<-tempfile(fileext = ".lifted")
  unmapped<-tempfile(fileext = ".unmapped")
  system(paste(liftOver, BED, chain, lifted, unmapped))
  new_pos <- fread(lifted)
  bad <- grep("^#", readLines(unmapped), value = TRUE, invert = TRUE)
  print(paste0(length(bad)," variants have not been mapped."))
  info_snp$pos <- NA
  info_snp$pos[new_pos$V4] <- new_pos$V3
  info_snp
}

# Write out as rds files split by chromosome
log <- foreach(i = 1:22, .combine = c, .options.multicore = list(preschedule = FALSE)) %dopar% {
  # Read in dbsnp data
  ref_i <- fread(paste0('~/oliverpainfel/Data/dbSNP/00-All.snps.nonambiguous.chr',i,'.tsv'))

  # Insert column names
  names(ref_i) <- c('chr','pos','snp','a2','a1')
  
  # Chunk only the liftover step
  chunk_rows <- 5e6
  n <- nrow(ref_i)
  liftover_bin <- '/users/k1806347/oliverpainfel/Software/MyGit/GenoDisc/pipeline/resources/software/liftover/liftover'
  chain_37_to_38 <- "/users/k1806347/oliverpainfel/Software/MyGit/GenoDisc/pipeline/resources/data/liftover/hg19ToHg38.over.chain.gz"
  chain_37_to_36 <- "/users/k1806347/oliverpainfel/Software/MyGit/GenoDisc/pipeline/resources/data/liftover/hg19ToHg18.over.chain.gz"
  ref_i$pos_grch36 <- NA
  ref_i$pos_grch38 <- NA
  for (start in seq(1, n, by = chunk_rows)) {
    end <- min(start + chunk_rows - 1, n)
    chunk <- ref_i[start:end,]

    # 37 -> 38
    res38 <- snp_modifyBuild_offline(chunk, liftOver = liftover_bin, chain = chain_37_to_38)
    ref_i$pos_grch38[start:end] <- res38$pos
    rm(res38); gc(FALSE)

    # 37 -> 36
    res36 <- snp_modifyBuild_offline(chunk, liftOver = liftover_bin, chain = chain_37_to_36)
    ref_i$pos_grch36[start:end] <- res36$pos
    rm(res36); gc(FALSE)
  }
  
  ref_i <- ref_i[, c('chr','snp','a1','a2','pos_grch36','pos','pos_grch38'), with=F]
  names(ref_i) <- c('CHR','SNP','A1','A2','BP_GRCh36','BP_GRCh37','BP_GRCh38')
  
  ref_i$IUPAC <- snp_iupac(ref_i$A1, ref_i$A2)

  saveRDS(ref_i, paste0('~/oliverpainfel/Data/dbSNP/00-All.snps.nonambiguous.chr',i,'.rds'))
}

```

***

# Calculate allele freqencies

For now just calculate frequency data for EUR.

```{bash}
module load plink2

for pop in $(echo EUR); do
  mkdir -p /users/k1806347/oliverpainfel/Data/hgdp_1kg/GRCh37_HGDP+1kGP_ALL/freq_files/${pop}
  
  sbatch --mem 20G -n 10 -p neurohack_cpu -t 1:00:00 --wrap="plink2 \
    --pfile /users/k1806347/oliverpainfel/Data/hgdp_1kg/GRCh37_HGDP+1kGP_ALL \
    --keep /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/keep_files/${pop}.keep \
    --freq \
    --threads 10 \
    --out /users/k1806347/oliverpainfel/Data/hgdp_1kg/GRCh37_HGDP+1kGP_ALL/freq_files/${pop}/GRCh37_HGDP+1kGP_ALL.${pop}"
done

```

***

# Archive

Code where I was trying to identify common variants from the full dbSNP data

***

### Identify common variants

Calculate MAF in each popualtion. Restrict to SNPs. 

```{bash}
module load plink2

mkdir -p /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_denser
for pop in $(echo EUR EAS AFR AMR CSA MID); do
    sbatch --mem 10G -n 5 -p neurohack_cpu,interruptible_cpu -t 1:00:00 --wrap="plink2 \
      --pfile /users/k1806347/oliverpainfel/Data/hgdp_1kg/GRCh37_HGDP+1kGP_ALL \
      --keep /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/keep_files/${pop}.keep \
      --maf 0.01 \
      --threads 5 \
      --snps-only 'just-acgt' \
      --make-just-pvar \
      --out /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_denser/hgdp_1kg.GW.${pop}"
done

# Identify list of variants with MAF > 0.01 in any population
cat /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_denser/hgdp_1kg.GW.*.pvar | grep -v "^#" | awk '{print $3}' | sort -u > /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_denser/union_variants.txt

# Subset full pvar to these variants only
sbatch --mem 10G -n 5 -p neurohack_cpu,interruptible_cpu -t 1:00:00 --wrap="plink2 \
  --pfile /users/k1806347/oliverpainfel/Data/hgdp_1kg/GRCh37_HGDP+1kGP_ALL \
  --extract /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_denser/union_variants.txt \
  --threads 5 \
  --make-just-pvar \
  --out /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_denser/hgdp_1kg.GW.ANY"

```

***

### Subset 1KG+HGDP to include common variants and insert RSIDs

```{bash}
# Subset dbsnp data to variant locations that are present in the 1KG+HGDP pvar with MAF 0.01
awk -F'\t' 'NR==FNR && $1 !~ /^#/ {
  key = $1":"$2;
  keep[key] = 1;
  next
}
BEGIN { OFS = " " }
{
  key = $1":"$2;
  if (key in keep) print
}' /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_denser/hgdp_1kg.GW.ANY.pvar ~/oliverpainfel/Data/dbSNP/00-All.vcf.mini.split > ~/oliverpainfel/Data/dbSNP/00-All.vcf.mini.split.filtered

```

```{r}

setwd('/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline')
source('../functions/misc.R')
source_all('../functions')
library(data.table)
library(GenoUtils)

dir.create('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense')

# Read in variant data for 1KG+HGDP
pvar <- fread('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_denser/hgdp_1kg.GW.ANY.pvar', nThread=10)

# Remove unneeded data and variant that aren't non ambiguous autosomal SNPs
pvar$INFO<-NULL
pvar$FILTER<-NULL
names(pvar)<-c('CHR','BP','SNP','A1','A2')
pvar<-pvar[pvar$CHR %in% 1:22,]
pvar$CHR<-as.numeric(pvar$CHR)
pvar$IUPAC<-snp_iupac(pvar$A1, pvar$A2)
pvar <- pvar[pvar$IUPAC %in% c('R','Y','K','M'),]
nrow(pvar) # 17706653

# Check overlap with HapMap3 variants
hapmap3_ref <- read_pvar('~/oliverpainfel/Software/MyGit/GenoPred/pipeline/resources/data/ref/ref.chr')
sum(paste0(pvar$CHR,':', pvar$BP) %in% paste0(hapmap3_ref$CHR,':', hapmap3_ref$BP)) # 1203510
nrow(hapmap3_ref) # 1204449
# 99.9% of HapMap3 variants are present. This is good.

# Read in dbSNP data and filter in the same way
dbsnp<-fread('~/oliverpainfel/Data/dbSNP/00-All.vcf.mini', nThread=10, header=F)
names(dbsnp)<-c('CHR','BP','SNP','A1','A2')
dbsnp<-dbsnp[dbsnp$CHR %in% 1:22,]
dbsnp$CHR<-as.numeric(dbsnp$CHR)
dbsnp$IUPAC<-snp_iupac(dbsnp$A1, dbsnp$A2)
dbsnp <- dbsnp[dbsnp$IUPAC %in% c('R','Y','K','M'),]

# Merge the pvar data with the dbSNP data allowing for strand flips
ref_target <- merge(pvar, dbsnp, by=c('CHR','BP'))
flip_logical<-detect_strand_flip(targ = ref_target$IUPAC.x, ref = ref_target$IUPAC.y)
sum(flip_logical) # There are no strand flips

# Identify SNPs that have matched IUPAC
ref_target<-ref_target[ref_target$IUPAC.x == ref_target$IUPAC.y,]

# Clean up pvar
pvar<-ref_target[, c('CHR','BP','SNP.x','SNP.y','A1.x','A2.x'), with=F]
names(pvar)<-c('CHR','BP','ID','SNP','A1','A2')

# Remove duplicates
pvar <- pvar[!duplicated(pvar$SNP),] # None removed
pvar <- pvar[!duplicated(paste0(pvar$CHR,':',pvar$BP)),] # 1075 removed
# 13,839,753 variants remain.

# Extract SNPs from plink2 files
write.table(pvar$ID, '/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/extract.snplist', col.names=F, row.names=F, quote=F)

# Save intermediate file
fwrite(pvar,'/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/hgdp_1kg_nonambig_snps.txt', sep = '\t', quote = F, na='NA')

###
# To maximise compatability with the methods that only contain HapMap3 variants in their ref, we should make sure we are keeping the HapMap3 variants.
# Read in the HapMap3 variants
hapmap3_ref <- read_pvar('~/oliverpainfel/Software/MyGit/GenoPred/pipeline/resources/data/ref/ref.chr')
sum(paste0(pvar$CHR,':', pvar$BP) %in% paste0(hapmap3_ref$CHR,':', hapmap3_ref$BP)) # 982075

# Check how many HapMap3 variants there are in the dbSNP common SNP list.
# Check by RSID to start
sum(dbsnp$SNP %in% hapmap3_ref$SNP) # 982031 of 1204449
sum(paste0(dbsnp$CHR,':', dbsnp$BP) %in% paste0(hapmap3_ref$CHR,':', hapmap3_ref$BP)) # 982075
# There are about 200k variants in the HapMap3 SNPlist missing from the dbsnp common variant list.

pvar <-fread('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/hgdp_1kg_nonambig_snps.txt')
sum(pvar$SNP %in% hapmap3_ref$SNP) # 981196 of 1204449
# Most of those variants are present in the 1KG+HGDP reference.
# Look up some variants that are missing to try and understand why.

missing <- hapmap3_ref[!(hapmap3_ref$SNP %in% pvar$SNP),]$SNP
freq<-readRDS('~/oliverpainfel/Software/MyGit/GenoPred/pipeline/resources/data/ref/ref.chr1.rds')
freq[freq$SNP %in% missing,]
# They are all common. This means the dbSNP common variant list is going to have poor overlap with common LD reference data, and also provide incomplete coverage of common variants more generally.
###

```

```{bash}
for chr in $(seq 1 22); do
  sbatch --mem 10G -n 5 -p neurohack_cpu,interruptible_cpu -t 1:00:00 --wrap="plink2 \
    --pfile /users/k1806347/oliverpainfel/Data/hgdp_1kg/GRCh37_HGDP+1kGP_ALL \
    --keep /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/unrel.keep \
    --extract /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/extract.snplist \
    --chr ${chr} \
    --threads 5 \
    --make-pgen \
    --out /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/hgdp_1kg.chr${chr}"
done
```

```{r}
library(data.table)

# Insert RSIDs into new plink files
ref <- fread('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/hgdp_1kg_nonambig_snps.txt')

for(i in 1:22){
  pvar <- fread(paste0('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/hgdp_1kg.chr',i,'.pvar'))
  pvar <- pvar[, c('#CHROM','POS','ID','REF','ALT'), with=F]
  pvar[ref, on=.(ID), ID := i.SNP]
  fwrite(
    pvar,
    paste0('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/hgdp_1kg.chr',i,'.pvar'),
    col.names = T,
    row.names = F,
    quote = F,
    sep = ' '
  )
}

```

```{bash}
# Calculate allele frequencies for each population
# Use keep files from previous reference preparation
for pop in $(echo EUR EAS AFR AMR CSA MID); do
  mkdir -p /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/freq_files/${pop}
  
  for chr in $(seq 1 22); do
    sbatch --mem 10G -n 1 -p neurohack_cpu,interruptible_cpu -t 1:00:00 --wrap="plink2 \
      --pfile /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/hgdp_1kg.chr${chr} \
      --keep /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred/keep_files/${pop}.keep \
      --freq \
      --out /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/freq_files/${pop}/hgdp_1kg.chr${chr}"
      
  done
done

# Create frequency files across all reference individuals (TRANS)
mkdir -p /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/freq_files/TRANS
for chr in $(seq 1 22); do
  sbatch --mem 10G -n 1 -p neurohack_cpu,interruptible_cpu -t 1:00:00 --wrap="plink2 \
    --pfile /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/hgdp_1kg.chr${chr} \
    --freq \
    --out /users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/freq_files/TRANS/hgdp_1kg.chr${chr}"
done
```

```{r}
####
# Create .rds files for SNP stats
####

# Make liftover functions
make_executable <- function(exe) {
  Sys.chmod(exe, mode = (file.info(exe)$mode | "111"))
}

snp_modifyBuild_offline<-function (info_snp, liftOver, chain){
  tmp_folder<-tempdir()

  if (!all(c("chr", "pos") %in% names(info_snp)))
    stop2("Please use proper names for variables in 'info_snp'. Expected %s.",
          "'chr' and 'pos'")
  liftOver <- normalizePath(liftOver)
  make_executable(liftOver)
  BED <- tempfile(fileext = ".BED")
  info_BED <- with(info_snp, data.frame(paste0("chr", chr),
                                        pos0 = pos - 1L, pos, id = seq_len(nrow(info_snp))))
  fwrite(info_BED, BED, col.names = FALSE, sep = " ")
  lifted<-paste0(tmp_folder,'/tmp.lifted')
  unmapped<-paste0(tmp_folder,'/tmp.unmapped')
  system(paste(liftOver, BED, chain, lifted, unmapped))
  new_pos <- fread(lifted)
  bad <- grep("^#", readLines(unmapped), value = TRUE, invert = TRUE)
  print(paste0(length(bad)," variants have not been mapped."))
  info_snp$pos <- NA
  info_snp$pos[new_pos$V4] <- new_pos$V3
  info_snp
}

for(i in 1:22){
  pvar <- fread(paste0('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/hgdp_1kg.chr',i,'.pvar'))
  
  # Insert BP for GRCh36 and 38
  names(pvar) <- c('chr','pos','snp','a2','a1')

  pvar_grch38<-snp_modifyBuild_offline(pvar, liftOver='/users/k1806347/oliverpainfel/Software/MyGit/GenoDisc/pipeline/resources/software/liftover/liftover', chain='/users/k1806347/oliverpainfel/Software/MyGit/GenoDisc/pipeline/resources/data/liftover/hg19ToHg38.over.chain.gz')
  
  pvar_grch36<-snp_modifyBuild_offline(pvar, liftOver='/users/k1806347/oliverpainfel/Software/MyGit/GenoDisc/pipeline/resources/software/liftover/liftover', chain='/users/k1806347/oliverpainfel/Software/MyGit/GenoDisc/pipeline/resources/data/liftover/hg19ToHg18.over.chain.gz')

  pvar[pvar_grch36, on=.(snp), pos_grch36 := i.pos]
  pvar[pvar_grch38, on=.(snp), pos_grch38 := i.pos]
  
  rm(pvar_grch36)
  rm(pvar_grch38)
  gc()
  
  pvar <- pvar[, c('chr','snp','a1','a2','pos_grch36','pos','pos_grch38'), with=F]
  names(pvar) <- c('CHR','SNP','A1','A2','BP_GRCh36','BP_GRCh37','BP_GRCh38')
  
  pvar$IUPAC<-snp_iupac(pvar$A1, pvar$A2)

  # Read in the FRQ data
  for(pop in c('EUR','EAS','AFR','AMR','CSA','MID')){
    freq <-
      fread(
        paste0(
          '/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/freq_files/',
          pop,
          '/hgdp_1kg.chr',
          i,
          '.afreq'
        )
      )
    
    names(freq)<-c('CHR','SNP','A2','A1','FREQ','OBS')
    freq$IUPAC<-snp_iupac(freq$A1, freq$A2)
    
    pvar<-merge(pvar, freq[, c('SNP','A1','FREQ'), with=F], by = 'SNP', sort = F)
    pvar$FREQ[pvar$A1.x != pvar$A1.y] <- 1-pvar$FREQ[pvar$A1.x != pvar$A1.y]
    
    pvar$A1.y <- NULL
    names(pvar)[names(pvar) == 'A1.x'] <- 'A1'
    pvar$FREQ <- round(pvar$FREQ, 5)
    names(pvar)[names(pvar) == 'FREQ'] <- paste0('REF.FRQ.', pop)
  }
  
  pvar <- pvar[, c("CHR", "SNP", "BP_GRCh36", "BP_GRCh37", "BP_GRCh38", "A1", "A2", "IUPAC", "REF.FRQ.EUR", "REF.FRQ.EAS", "REF.FRQ.AFR", "REF.FRQ.AMR", "REF.FRQ.CSA", "REF.FRQ.MID"), with=F]
  
  saveRDS(pvar, file = paste0('/users/k1806347/oliverpainfel/Data/hgdp_1kg/genopred_dense/hgdp_1kg.chr', i,'.rds'))
}

```


