<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Preparing LDpred2 LD matrices</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />
<link rel="shortcut icon" href="Images/logo/favicon.ico">

<link rel="stylesheet" href="styles/night-mode.css" id="nightModeStylesheet">

<script>
function toggleNightMode() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (stylesheet.disabled) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
}
</script>

<label class="switch">
  <input type="checkbox" id="toggleNightMode" checked>
  <span class="slider round"></span>
</label>

<script>
document.getElementById('toggleNightMode').addEventListener('change', function() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (this.checked) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
});
</script>

<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-YR18ZB3PR3"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-YR18ZB3PR3');
</script>


<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="styles/styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img class="logo-img" src="Images/logo/Horizontal_white.png" style="height: 42px;" /></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pipeline
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="pipeline_overview.html">Overview</a>
    </li>
    <li>
      <a href="pipeline_readme.html">Instructions</a>
    </li>
    <li>
      <a href="pipeline_technical.html">Technical documentation</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Research
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="research_index.html">Overview</a>
    </li>
    <li>
      <a href="comparison_of_methods_summary.html">Polygenic Scoring Methods Comparison</a>
    </li>
    <li>
      <a href="Functionally_informed_prediction.html">Quantifying Polygenic Signal Mediated by Altered Gene Expression</a>
    </li>
    <li>
      <a href="Absolute_Conversion.html">Translating Polygenic Scores onto the Absolute Scale</a>
    </li>
  </ul>
</li>
<li>
  <a href="more_index.html">More</a>
</li>
<li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Preparing LDpred2 LD matrices</h1>

</div>


<hr />
<p>Here we will calculate LD matrices for LDpred2 using HapMap3 SNPs in
the 1KG super populations. This is to allow fair comparison across
methods for a cross-population polygenic scoring study.</p>
<p>The code will follow that used to generate the LD matrices for EUR,
in this paper (<a
href="https://doi.org/10.1016/j.xhgg.2022.100136">link</a>).</p>
<pre class="bash"><code># Subset 1KG into super populations and merge across chromosomes
for pop in $(echo EUR EAS AFR); do
  mkdir -p /users/k1806347/oliverpainfel/Data/1KG/ldpred2/${pop}

  # Create mergelist
  rm -f /users/k1806347/oliverpainfel/Data/1KG/ldpred2/${pop}/mergelist.txt
  for chr in $(seq 1 22); do
    echo ~/oliverpainfel/Data/1KG/GenoPred/v2.0.0/ref.chr${chr} &gt;&gt; /users/k1806347/oliverpainfel/Data/1KG/ldpred2/${pop}/mergelist.txt
  done
  
  ~/oliverpainfel/Software/plink1.9/plink \
    --merge-list /users/k1806347/oliverpainfel/Data/1KG/ldpred2/${pop}/mergelist.txt \
    --make-bed \
    --keep /users/k1806347/oliverpainfel/Data/1KG/GenoPred/v2.0.0/keep_files/${pop}.keep \
    --out /users/k1806347/oliverpainfel/Data/1KG/ldpred2/${pop}/ref.${pop}.GW
done</code></pre>
<pre class="r"><code>library(bigsnpr)
library(bigreadr)

# Create object for genetic data in ldpred2 format, and impute missing values
for(pop in c(&#39;EUR&#39;, &#39;EAS&#39;, &#39;AFR&#39;)){
  # Read in reference data
  snp_readBed(
    paste0(
      &#39;/users/k1806347/oliverpainfel/Data/1KG/ldpred2/&#39;,
      pop,
      &#39;/ref.&#39;,
      pop,
      &#39;.GW.bed&#39;
    )
  )
  
  # Attach the ref object in R session
  ref &lt;-
    snp_attach(
      paste0(
        &#39;/scratch/prj/oliverpainfel/Data/1KG/ldpred2/&#39;,
        pop,
        &#39;/ref.&#39;,
        pop,
        &#39;.GW.rds&#39;
      )
    )
  G &lt;- ref$genotypes
  NCORES &lt;- nb_cores()
  
  #### Impute missing values (bigsnpr can&#39;t handle missing data in most functions)
  G_imp &lt;- snp_fastImputeSimple(G, method = &quot;mean2&quot;, ncores = NCORES)

  # Save imputed reference
  ref$genotypes&lt;-G_imp
  saveRDS(ref, paste0(
        &#39;/scratch/prj/oliverpainfel/Data/1KG/ldpred2/&#39;,
        pop,
        &#39;/ref.&#39;,
        pop,
        &#39;.GW.rds&#39;
      ))
}

################
# Now calculate LD matrices following code from https://doi.org/10.1016/j.xhgg.2022.100136

library(dplyr)
library(bigreadr)
library(bigsnpr)

for(pop in c(&#39;EUR&#39;, &#39;EAS&#39;, &#39;AFR&#39;)){
  obj.1000G &lt;- snp_attach(paste0(
          &#39;/scratch/prj/oliverpainfel/Data/1KG/ldpred2/&#39;,
          pop,
          &#39;/ref.&#39;,
          pop,
          &#39;.GW.rds&#39;
        ))
  G &lt;- obj.1000G$genotypes
  CHR &lt;- obj.1000G$map$chromosome
  POS &lt;- obj.1000G$map$physical.pos
  POS2 &lt;- snp_asGeneticPos(CHR, POS, dir =&#39;/users/k1806347/oliverpainfel/Data/Genetic_Map/CEU&#39;, ncores = NCORES) # Note this is not appropriate but I think Florian did the same as used distance from UKB
  MAF &lt;- snp_MAF(G, ncores = NCORES)
  keep &lt;- (MAF &gt; 0.02)

  all_final_grp &lt;- NULL
  for(chr in 1:22){
    ind.chr2 &lt;- which(CHR == chr &amp; keep)

    corr0 &lt;- runonce::save_run({
      snp_cor(
        G,
        ind.col = ind.chr2,
        infos.pos = POS2[ind.chr2],
        size = 3 / 1000,
        ncores = NCORES
      )
    }, file = paste0(
      &#39;/scratch/prj/oliverpainfel/Data/1KG/ldpred2/&#39;,
      pop,
      &#39;/LD_chr&#39;,
      chr,
      &quot;.rds&quot;
    ))
    
    # find nearly independent LD blocks
    m &lt;- length(ind.chr2)
    (SEQ &lt;- round(seq_log(m / 30, m / 5, length.out = 20)))
    splits &lt;- snp_ldsplit(corr0, thr_r2 = 0.05, min_size = 50, max_size = SEQ, max_r2 = 0.15)
    splits$cost2 &lt;- sapply(splits$all_size, function(sizes) sum(sizes^2))
  
    best_split &lt;- splits %&gt;%
      arrange(cost2 * sqrt(5 + cost)) %&gt;%
      print() %&gt;%
      slice(1) %&gt;%
      print()
  
    (all_size &lt;- best_split$all_size[[1]])
    best_grp &lt;- rep(seq_along(all_size), all_size)
  
    runonce::save_run({
      corr0T &lt;- as(corr0, &quot;dgTMatrix&quot;)
      corr0T@x &lt;-
        ifelse(best_grp[corr0T@i + 1L] == best_grp[corr0T@j + 1L], corr0T@x, 0)
      as(Matrix::drop0(corr0T), &quot;symmetricMatrix&quot;)
    }, file = paste0(
      &#39;/scratch/prj/oliverpainfel/Data/1KG/ldpred2/&#39;,
      pop,
      &#39;/LD_with_blocks_chr&#39;,
      chr,
      &quot;.rds&quot;
    ))
  
    # return
    all_final_grp &lt;- rbind(all_final_grp, tibble(best_split, ind = list(ind.chr2)))
  }
  saveRDS(
    all_final_grp,
    paste0(
      &#39;/scratch/prj/oliverpainfel/Data/1KG/ldpred2/&#39;,
      pop,
      &#39;/all_final_grp.rds&#39;
    )
  )
  
  ###
  # Create map.rds
  ###
  map &lt;- obj.1000G$map[keep,]

  # Compute allele frequency
  map$af&lt;-big_colstats(G, ind.col = which(keep), ncores = NCORES)$sum / (2 * nrow(G))
  
  # Compute LD scores
  map$ld &lt;- do.call(&#39;c&#39;, lapply(1:22, function(chr) {
    cat(chr, &quot;.. &quot;, sep = &quot;&quot;)
    corr_chr &lt;- readRDS(
      paste0(
        &#39;/scratch/prj/oliverpainfel/Data/1KG/ldpred2/&#39;,
        pop,
        &#39;/LD_chr&#39;,
        chr,
        &quot;.rds&quot;
      )
    )
    Matrix::colSums(corr_chr ^ 2)
  }))
  
  map &lt;- map[, c(&#39;chromosome&#39;,&#39;physical.pos&#39;,&#39;allele2&#39;,&#39;allele1&#39;,&#39;marker.ID&#39;,&#39;af&#39;,&#39;ld&#39;)]
  names(map) &lt;- c(&#39;chr&#39;,&#39;pos&#39;,&#39;a0&#39;,&#39;a1&#39;,&#39;rsid&#39;,&#39;af&#39;,&#39;ld&#39;)
  
  saveRDS(
    map,
    paste0(
      &#39;/scratch/prj/oliverpainfel/Data/1KG/ldpred2/&#39;,
      pop,
      &#39;/map.rds&#39;
    )
  )
}</code></pre>

<!-- footer.html -->
<hr/>

<div class="centered-container">
<div class="rounded-image-container" style="width: 500px;">
<img src="Images/logo/sponsors.png">
</div>
</div>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
