<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>research_index.knit</title>

<script src="site_libs/header-attrs-2.23/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.0/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.0/css/v4-shims.min.css" rel="stylesheet" />
<!-- header.html -->
<style>
  .navbar {
    min-height: 80px; /* Adjust as needed */
  }

  .navbar-brand {
    padding: 10px 15px; /* Adjust as needed */
  }
</style>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img id="logo" style="height: 60px;" src="Horizontal.png" /></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="research_index.html">Research</a>
</li>
<li>
  <a href="tools_index.html">Tools</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">




</div>


<p><br/></p>
<div id="genopred-research" class="section level1">
<h1>GenoPred Research</h1>
<p><br/></p>
<div id="comparison-of-existing-polygenic-scoring-methodology"
class="section level2">
<h2>1. Comparison of existing polygenic scoring methodology</h2>
<div style="display: flex; align-items: flex-start;">
<p style="padding-right: 20px; flex: 1;">
Polygenic scoring methodology is rapidly evolving. GenoPred benchmarks
leading polygenic scoring methods within a controlled environment,
providing a fair comparison of methods across a range of
settings.<br/><br/><strong>Overview and code:</strong>
<a href="temp.html" target="_blank"> Click here
</a><br/><strong>Paper:</strong>
<a href="https://doi.org/10.1371/journal.pgen.1009021" target="_blank">
Pain et al. Evaluation of polygenic prediction methodology within a
reference-standardized framework. PLoS genetics 17.5 (2021): e1009021.
</a>
</p>
<p><img src="comparison_of_methods.png" style="width: 300px; flex: 0 0 auto;"/></p>
</div>
<p><br/></p>
</div>
<div
id="quantifying-polygenic-signal-mediated-by-altered-gene-expression"
class="section level2">
<h2>2. Quantifying Polygenic Signal Mediated by Altered Gene
Expression</h2>
<div style="display: flex; align-items: flex-start;">
<p style="padding-right: 20px; flex: 1;">
Many genetic associations fall within regulatory regions of the genome.
We use imputed gene expression scores to evaluate the proportion of
polygenic signals linked to altered expression of
genes.<br/><br/><strong>Overview and code:</strong>
<a href="temp.html" target="_blank"> Click here
</a><br/><strong>Paper:</strong>
<a href="https://doi.org/10.1093/hmg/ddab053" target="_blank"> Pain et
al. Imputed gene expression risk scores: a functionally informed
component of polygenic risk. Human Molecular Genetics 30.8 (2021):
727-738.</a>
</p>
<p><img src="cis_pts.png" style="width: 300px; flex: 0 0 auto;"/></p>
</div>
<p><br/></p>
</div>
<div id="translating-polygenic-scores-onto-the-absolute-scale"
class="section level2">
<h2>3. Translating Polygenic Scores onto the Absolute Scale</h2>
<div style="display: flex; align-items: flex-start;">
<p style="padding-right: 20px; flex: 1;">
The interpretability of polygenic scores is essential for their use
within a clinical setting. We developed a simple approach and tool for
converting polygenic scores from the relative scale onto the absolute
scale.<br/><br/><strong>Overview and code:</strong>
<a href="temp.html" target="_blank"> Click
here</a><br/><strong>Paper:</strong>
<a href="https://doi.org/10.1038/s41431-021-01028-z" target="_blank">
Pain et al. A tool for translating polygenic scores onto the absolute
scale using summary statistics. European Journal of Human Genetics 30.3
(2022): 339-348.</a>
</p>
<p><img src="interp.png" style="width: 300px; flex: 0 0 auto;"/></p>
</div>
<p><br/></p>
</div>
<div id="polytranscriptomic-scoring" class="section level2">
<h2>4. Polytranscriptomic Scoring</h2>
<div style="display: flex; align-items: flex-start;">
<p style="padding-right: 20px; flex: 1;">
Polytranscriptomic scores leverage large-scale transcriptome-wide
association study (TWAS) results and observed expression levels in a
target samples. We find these polytranscriptomic scores can be used to
predict disease risk and clinical
characteristics.<br/><br/><strong>Overview and code:</strong>
<a href="temp.html" target="_blank"> Click
here</a><br/><strong>Preprint:</strong>
<a href="https://doi.org/10.1101/2023.01.18.23284589" target="_blank">
Pain et al. Harnessing Transcriptomic Signals for Amyotrophic Lateral
Sclerosis to Identify Novel Drugs and Enhance Risk Prediction. medRxiv
(2023). </a>
</p>
<p><img src="obs_pts.png" style="width: 300px; flex: 0 0 auto;"/></p>
</div>
<p><br/></p>
</div>
<div id="polygenic-prediction-of-molecular-traits"
class="section level2">
<h2>5. Polygenic Prediction of Molecular Traits</h2>
<div style="display: flex; align-items: flex-start;">
<p style="padding-right: 20px; flex: 1;">
Models using genetic variation to predict molecular traits, such as gene
expression, are useful for modelling the mechanistic effect of variants
identified using GWAS. However, meta-analysis of QTL datasets
necessitates the use of summary statistic-based methods for generating
these models. We evaluate a range of summary statistic methods for
prediting gene expression.<br/><br/><strong>Overview and code:</strong>
<a href="temp.html" target="_blank"> Click
here</a><br/><strong>Paper:</strong>
<a href="https://doi.org/10.1101/2022.11.23.517213" target="_blank">
Pain et al. Polygenic Prediction of Molecular Traits using Large-Scale
Meta-analysis Summary Statistics. bioRxiv (2022): 2022-11. </a>
</p>
<p><img src="pred_mol.png" style="width: 300px; flex: 0 0 auto;"/></p>
</div>
<p><br/></p>
<p><br/></p>
</div>
</div>

<!-- footer.html -->
<div style="background-color: #f5f5f5; padding: 20px; display: flex; align-items: center; justify-content: space-between;">
  <p>Created by Oliver Pain with colleagues at King's College London.</p>
  <div>
    <img src="kcl.svg" alt="Kings College London" style="height: 60px;">
    <img src="Wellcome_logo.png" alt="The Wellcome Trust" style="height: 60px;">
    <img src="nihr-logo.png" alt="National Institute for Health and Care Research" style="height: 60px;">
  </div>
</div>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
