---
title: GenoPred Research
output: 
  html_document:
    theme: cosmo
    toc: true
    toc_float: true
    css: styles/styles.css
    includes:
      in_header: header.html
      after_body: footer.html

---

***

# 1. Polygenic Scoring Methods Comparison

<div style="display: flex; flex-wrap: wrap; align-items: center; justify-content: center;">
  <p style="padding-right: 20px; flex: 1;">Polygenic scoring methodology is rapidly evolving. GenoPred benchmarks leading polygenic scoring methods within a controlled environment, providing a fair comparison of methods across a range of settings.<br/><br/><strong>Overview and code:</strong> <a href="comparison_of_methods_summary.html"> Click here
</a><br/><strong>Paper:</strong> <a href="https://doi.org/10.1371/journal.pgen.1009021" target="_blank"> Pain et al. Evaluation of polygenic prediction methodology within a reference-standardized framework. PLoS genetics 17.5 (2021): e1009021.
</a></p>
  
  <img src="Images/research_summary/comparison_of_methods.png" style="width: 300px; flex: 0 0 auto;"/>
</div>

<br/>

# 2. Quantifying Polygenic Signal Mediated by Altered Gene Expression

<div style="display: flex; flex-wrap: wrap; align-items: center; justify-content: center;">
  <p style="padding-right: 20px; flex: 1;">Many genetic associations fall within regulatory regions of the genome. We use imputed gene expression scores to evaluate the proportion of polygenic signals linked to altered expression of genes.<br/><br/><strong>Overview and code:</strong> <a href="Functionally_informed_prediction.html"> Click here
</a><br/><strong>Paper:</strong> <a href="https://doi.org/10.1093/hmg/ddab053" target="_blank"> Pain et al. Imputed gene expression risk scores: a functionally informed component of polygenic risk. Human Molecular Genetics 30.8 (2021): 727-738.</a></p>
  <img src="Images/research_summary/cis_pts.png" style="width: 300px; flex: 0 0 auto;"/>
</div>

<br/>

# 3. Translating Polygenic Scores onto the Absolute Scale

<div style="display: flex; flex-wrap: wrap; align-items: center; justify-content: center;">
  <p style="padding-right: 20px; flex: 1;">The interpretability of polygenic scores is essential for their use within a clinical setting. We developed a simple approach and tool for converting polygenic scores from the relative scale onto the absolute scale.<br/><br/><strong>Overview and code:</strong> <a href="Absolute_Conversion.html"> Click here</a><br/><strong>Paper:</strong> <a href="https://doi.org/10.1038/s41431-021-01028-z" target="_blank"> Pain et al. A tool for translating polygenic scores onto the absolute scale using summary statistics. European Journal of Human Genetics 30.3 (2022): 339-348.</a></p>
  <img src="Images/research_summary/interp.png" style="width: 300px; flex: 0 0 auto;"/>
</div>

<br/>

# 4. Polytranscriptomic Scoring

<div style="display: flex; flex-wrap: wrap; align-items: center; justify-content: center;">
  <p style="padding-right: 20px; flex: 1;">Polytranscriptomic scores leverage large-scale transcriptome-wide association study (TWAS) results and observed expression levels in a target samples. We find these polytranscriptomic scores can be used to predict disease risk and clinical characteristics.<br/><br/><strong>Overview and code:</strong> Not yet available<br/><strong>Preprint:</strong> <a href="https://doi.org/10.1101/2023.01.18.23284589" target="_blank"> Pain et al. Harnessing Transcriptomic Signals for Amyotrophic Lateral Sclerosis to Identify Novel Drugs and Enhance Risk Prediction. medRxiv (2023).
</a></p>
  <img src="Images/research_summary/obs_pts.png" style="width: 300px; flex: 0 0 auto;"/>
</div>

<br/>

# 5. Polygenic Prediction of Molecular Traits

<div style="display: flex; flex-wrap: wrap; align-items: center; justify-content: center;">
  <p style="padding-right: 20px; flex: 1;">Models using genetic variation to predict molecular traits, such as gene expression, are useful for modelling the mechanistic effect of variants identified using GWAS. However, meta-analysis of QTL datasets necessitates the use of summary statistic-based methods for generating these models. We evaluate a range of summary statistic methods for prediting gene expression.<br/><br/><strong>Overview and code:</strong> <a href="https://github.com/opain/eQTL_to_TWAS#readme" target="_blank"> Click here</a><br/><strong>Preprint:</strong> <a href="https://doi.org/10.1101/2022.11.23.517213" target="_blank"> Pain et al. Polygenic Prediction of Molecular Traits using Large-Scale Meta-analysis Summary Statistics. bioRxiv (2022): 2022-11.
</a></p>
  <img src="Images/research_summary/pred_mol.png" style="width: 300px; flex: 0 0 auto;"/>
</div>

<br/>

# 6. Cross-Ancestry Polygenic Prediction

<div style="display: flex; flex-wrap: wrap; align-items: center; justify-content: center;">
  <p style="padding-right: 20px; flex: 1;">
    As GWAS become available across diverse populations, polygenic scoring methods that combine multiple sources are essential for equitable prediction. 
    This study benchmarked single- and multi-source PGS methods across African, East Asian, and European ancestries using public GWAS and UK Biobank target data. 
    A novel application of the LEOPARD method enables accurate linear combinations of ancestry-specific PGS using only summary statistics, 
    offering a scalable solution to optimise prediction in underrepresented groups. 
    <br/><br/><strong>Overview and code:</strong> 
    <a href="CrossPop.html" target="_blank">Click here</a> 
    <br/><strong>Preprint:</strong> 
    <a href="https://doi.org/10.1101/2025.03.27.25324773" target="_blank">
      Pain. Leveraging Global Genetics Resources to Enhance Polygenic Prediction Across Ancestrally Diverse Populations. medRxiv (2025).
    </a>
  </p>
  <img src="Images/research_summary/cross_pop_2.png" style="width: 300px; flex: 0 0 auto;"/>
</div>

<br/>
<br/>

