<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>GenoPred Pipeline - Running Offline</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />
<link rel="shortcut icon" href="Images/logo/favicon.ico">

<link rel="stylesheet" href="styles/night-mode.css" id="nightModeStylesheet">

<script>
function toggleNightMode() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (stylesheet.disabled) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
}
</script>

<label class="switch">
  <input type="checkbox" id="toggleNightMode" checked>
  <span class="slider round"></span>
</label>

<script>
document.getElementById('toggleNightMode').addEventListener('change', function() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (this.checked) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
});
</script>

<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-YR18ZB3PR3"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-YR18ZB3PR3');
</script>


<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="styles/styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img class="logo-img" src="Images/logo/Horizontal_white.png" style="height: 42px;" /></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pipeline
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="pipeline_overview.html">Overview</a>
    </li>
    <li>
      <a href="pipeline_readme.html">Instructions</a>
    </li>
    <li>
      <a href="pipeline_technical.html">Technical documentation</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Research
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="research_index.html">Overview</a>
    </li>
    <li>
      <a href="comparison_of_methods_summary.html">Polygenic Scoring Methods Comparison</a>
    </li>
    <li>
      <a href="Functionally_informed_prediction.html">Quantifying Polygenic Signal Mediated by Altered Gene Expression</a>
    </li>
    <li>
      <a href="Absolute_Conversion.html">Translating Polygenic Scores onto the Absolute Scale</a>
    </li>
  </ul>
</li>
<li>
  <a href="more_index.html">More</a>
</li>
<li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">GenoPred Pipeline - Running Offline</h1>

</div>


<hr />
<p>This document describes how to run the GenoPred pipeline offline (an
environment without access to the internet).</p>
<p>First, the dependencies of the pipeline will need to prepared within
an environment that <strong>does</strong> have access to the internet.
These include the software container and additional pipeline resources.
Then these resources can be transferred to the offline environment, and
the pipeline can be run.</p>
<p>Below, I provide an example of this process.</p>
<hr />
<div id="download-software-container" class="section level1">
<h1>Download software container</h1>
<p>The GenoPred image has already been built using the Dockerfile <a
href="https://github.com/opain/GenoPred/blob/master/pipeline/misc/docker/Dockerfile">here</a>.
This image is hosted on dockerhub and the singularity library.</p>
<pre class="bash"><code># Docker
docker \
  pull \
  opaino/genopred_pipeline:latest

# Singularity
singularity \
  pull \
  --arch \
  amd64 \
  /users/k1806347/oliverpainfel/Software/singularity/genopred_pipeline_latest.sif \
  library://opain/genopred/genopred_pipeline:latest</code></pre>
</details>
<hr />
</div>
<div id="downloading-pipeline-resources" class="section level1">
<h1>Downloading pipeline resources</h1>
<p>The resources required by the pipeline depends on the analyses
requested by the user. I have provided rules to download required
resources for two scenarios:</p>
<ul>
<li><code>get_key_resources</code>: Allows for <strong>most</strong> PGS
methods (<code>dbslmm</code>, <code>lassosum</code>,
<code>megaprs</code>, <code>ptclump</code>)</li>
<li><code>get_all_resources</code>: Allows for <strong>all</strong> PGS
methods (additionally incl. <code>ldpred2</code>, <code>prscs</code>,
<code>prscsx</code>, <code>quickprs</code>, <code>sbayesr</code>,
<code>sbayesrc</code>, <code>xwing</code>)</li>
</ul>
<p>Alternatively, the user can download only the data required for
<code>ldpred2</code>, <code>prscs</code>, <code>prscsx</code>,
<code>quickprs</code>, <code>sbayesr</code>, <code>sbayesrc</code>,
<code>xwing</code>:</p>
<ul>
<li><code>get_ldpred2_resources</code>: Allows for
<strong>ldpred2</strong></li>
<li><code>get_prscs_resources</code>: Allows for
<strong>prscs</strong></li>
<li><code>get_prscsx_resources</code>: Allows for
<strong>prscsx</strong></li>
<li><code>get_quickprs_resources</code>: Allows for
<strong>quickprs</strong></li>
<li><code>get_sbayesr_resources</code>: Allows for
<strong>sbayesr</strong></li>
<li><code>get_sbayesrc_resources</code>: Allows for
<strong>sbayesrc</strong></li>
<li><code>get_xwing_resources</code>: Allows for
<strong>xwing</strong></li>
</ul>
<p><strong>Note.</strong> <code>23andMe</code> format target data will
only be allowed for if the <code>download_impute2_data</code> rule is
run, as reference data for imputation is required.</p>
<hr />
<p>In this example, I will run the pipeline using the example
configuration with the test data (running <code>ptclump</code>,
<code>dbslmm</code>, and <code>lassosum</code>), so I will use the
<code>get_key_resources</code> rule to download the required
resources</p>
<pre class="bash"><code># Create a configuration file specifying directory to save the required resources
echo &quot;resdir: genopred_resources&quot; &gt; config_offline.yaml

# Run GenoPred pipeline using the get_key_resources rule
cd /users/k1806347/oliverpainfel/Software/MyGit/GenoPred/pipeline
conda activate genopred
snakemake \
  --profile slurm \
  --configfile=/users/k1806347/oliverpainfel/test/offline_example/config.yaml \
  get_key_resources</code></pre>
<hr />
</div>
<div id="download-test-data" class="section level1">
<h1>Download test data</h1>
<p>For demonstration purposes, we will use the test data for the
GenoPred pipeline. This is the same data as is used for the <a
href="https://opain.github.io/GenoPred/pipeline_readme.html#Run_using_test_data">main
pipeline tutorial page</a>.</p>
<pre class="bash"><code>cd /users/k1806347/oliverpainfel/test/offline_example

# Download from Zenodo
wget -O test_data.tar.gz https://zenodo.org/records/10640650/files/test_data.tar.gz?download=1

# Decompress
tar -xf test_data.tar.gz

# Once decompressed, delete compressed version to save space
rm test_data.tar.gz</code></pre>
<hr />
</div>
<div id="run-pipeline-offline" class="section level1">
<h1>Run pipeline offline</h1>
<p>I will now start an interactive session in the downloaded container.
I will mount a folder within the container so I can read and write files
outside of the container. See <a
href="https://docs.docker.com/">docker</a> and <a
href="https://sylabs.io/docs/">singularity</a> websites for general
documentation on their use.</p>
<pre class="bash"><code>######
# Start interactive session within the container
######
# When using singularity or docker, we must mount folders we want to access within the container

# Singularity 
singularity shell \
  --bind /scratch/prj/oliverpainfel:/scratch/prj/oliverpainfel \
  --writable-tmpfs \
  /users/k1806347/oliverpainfel/Software/singularity/genopred_pipeline_latest.sif

# Docker
docker run \
  -it \
  -v /users/k1806347/oliverpainfel:/users/k1806347/oliverpainfel \
  genopred_pipeline:v0.4</code></pre>
<hr />
<p>Once the container has been started, we can use the pipeline as
normal. A previously downloaded version of the GenoPred repo will be in
the folder <code>/tools/GenoPred</code>, and the <code>genopred</code>
conda environment will already be available.</p>
<p>In this example, before running the pipeline, I first get the
configuration set up to the use the test data, the example configuration
files, and run in an offline environment. First I create a symbolic link
to the previously downloaded test_data, to align with the
example_configuration files. Then, I update the example configuration to
run in an offline environment:</p>
<ul>
<li>Update the <code>resdir</code> parameter configfile to use the
previously downloaded resources</li>
<li>Update the <code>score_list</code> to only include locally stored
score files</li>
</ul>
<pre class="bash"><code># Activate GenoPred environment
source /opt/mambaforge/etc/profile.d/conda.sh
conda activate genopred

# Create symbolic for test_data inside the pipeline folder to work with the example configuration files
cd /tools/GenoPred/pipeline
ln -s /users/k1806347/oliverpainfel/test/offline_example/test_data ./test_data

# Update configuration files to run offline
# 1. Remove score files requiring direct access to PGS catalogue
awk -F&#39; &#39; &#39;$2 != &quot;NA&quot; {print}&#39; example_input/score_list.txt &gt; example_input/score_list_2.txt &amp;&amp; mv example_input/score_list_2.txt example_input/score_list.txt
# 2. Update resdir to previously downloaded resources
echo &quot;resdir: /scratch/prj/oliverpainfel/test/offline_example/genopred_resources&quot; &gt;&gt; example_input/config.yaml

# Do a dry run to check the scheduled steps are expected (there should not be any steps saying &#39;download&#39;, and it should not be necessary to build the conda environment)
snakemake -n --use-conda --configfile=example_input/config.yaml output_all

# Run pipeline using test data and example configuration
snakemake -j4 --use-conda --configfile=example_input/config.yaml output_all</code></pre>
<hr />
</div>
<div id="any-questions" class="section level1">
<h1>Any questions?</h1>
<p>Please post questions as an issue on the GenoPred GitHub repo <a
href="https://github.com/opain/GenoPred/issues">here</a>.</p>
</div>

<!-- footer.html -->
<hr/>

<div class="centered-container">
<div class="rounded-image-container" style="width: 500px;">
<img src="Images/logo/sponsors.png">
</div>
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
