<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Running GenoPred on DNAnexus</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />
<link rel="shortcut icon" href="Images/logo/favicon.ico">

<link rel="stylesheet" href="styles/night-mode.css" id="nightModeStylesheet">

<script>
function toggleNightMode() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (stylesheet.disabled) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
}
</script>

<label class="switch">
  <input type="checkbox" id="toggleNightMode" checked>
  <span class="slider round"></span>
</label>

<script>
document.getElementById('toggleNightMode').addEventListener('change', function() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (this.checked) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
});
</script>

<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-YR18ZB3PR3"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-YR18ZB3PR3');
</script>


<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="styles/styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img class="logo-img" src="Images/logo/Horizontal_white.png" style="height: 42px;" /></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pipeline
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="pipeline_overview.html">Overview</a>
    </li>
    <li>
      <a href="pipeline_readme.html">Instructions</a>
    </li>
    <li>
      <a href="pipeline_technical.html">Technical documentation</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Research
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="research_index.html">Overview</a>
    </li>
    <li>
      <a href="comparison_of_methods_summary.html">Polygenic Scoring Methods Comparison</a>
    </li>
    <li>
      <a href="Functionally_informed_prediction.html">Quantifying Polygenic Signal Mediated by Altered Gene Expression</a>
    </li>
    <li>
      <a href="Absolute_Conversion.html">Translating Polygenic Scores onto the Absolute Scale</a>
    </li>
  </ul>
</li>
<li>
  <a href="more_index.html">More</a>
</li>
<li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Running GenoPred on DNAnexus</h1>

</div>


<hr />
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>The UK Biobank (UKB) has recently updated its data access policy,
requiring researchers to access and analyze UKB data exclusively via the
UKB Research Analysis Platform (UKB-RAP), hosted by DNAnexus. This
represents a significant shift in workflows, moving from institutional
servers to a cloud-computing environment. DNAnexus and similar
cloud-computing systems are likely to become standard for future
datasets, so this guide aims to provide instructions on running GenoPred
on DNAnexus, with a particular focus on analyzing UKB data. The UKB
dataset presents unique challenges due to its size, making efficient
analysis essential.</p>
<p>Cloud computing involves requesting access to an instance (or virtual
machine) with a specified amount of resources, such as disk space, RAM,
and the number of CPU cores. Once access is granted, users must create
the required software and data environment within the instance—a step
that is often unfamiliar to many. After completing the desired analyses,
users must export any outputs they wish to retain, as any data or
software left on the instance will be deleted once it is terminated.</p>
<p>This process of transferring software and data into and out of the
instance is a significant departure from the experience of working on
personal computers or institutional servers, where you can resume work
seamlessly each time you log in.</p>
<hr />
</div>
<div id="rstudio-vs.-cloud-workstation" class="section level1">
<h1>RStudio vs. Cloud Workstation</h1>
<p>Currently, the easiest way to run GenoPred on DNAnexus is
interactively, using either RStudio or a Cloud Workstation with the
GenoPred software container. Each approach offers unique advantages.</p>
<p>RStudio mounts files directly from a DNAnexus project, eliminating
the need to import data. This feature saves time and disk space,
particularly with large datasets like UKB genetic data.</p>
<p>Unlike RStudio, the cloud workstation requires input data to be
manually imported using dx download, which can be time-consuming and
requires additional storage space. While dxfuse allows project folders
to be mounted (even across projects), it is prone to instability during
long analyses, such as when using UKB.</p>
<p>Another distinction is that cloud workstations can be connected to
via VScode, which some people may prefer to Rstudio.</p>
<p>Given the advantage of stable project folder mounting in RStudio, I
will demonstrate the workflow in that context. For the demo, minimal
resources are required, so I will request an
<code>mem1_ssd1_v2_x2</code> instance type. A similar workflow can also
be applied in the Cloud Workstation, substituting the Singularity
container with the GenoPred Docker container.</p>
<hr />
</div>
<div id="demonstration-using-rstudio" class="section level1">
<h1>Demonstration using Rstudio</h1>
<hr />
<div id="step-0-prepare-test-data" class="section level2">
<h2>Step 0: Prepare test data</h2>
<p>To make the demonstration as similar as possible to working with UKB
data, I will first upload a version of the GenoPred test data into a
DNAnexus project. While it would be simpler to download the test data
directly into the instance, this approach will better mimic the set up
when working with UKB data.</p>
<pre class="bash"><code># Step 1: Download the GenoPred test data from Zenodo
wget -O test_data.tar.gz https://zenodo.org/records/10640650/files/test_data.tar.gz?download=1

# Step 2: Decompress the downloaded file
tar -xf test_data.tar.gz

# Step 3: Extract only the necessary files for the demonstration
mkdir genopred_test_data
mv test_data/target/imputed_sample_plink2/example.chr22* genopred_test_data/
mv test_data/reference/gwas_sumstats/BODY04.gz genopred_test_data/

# Step 4: Load the Conda environment with dxpy installed
# The Conda environment file for installing dxpy is available in the GenoPred repository
# Path: GenoPred/pipeline/misc/dnanexus/dxpy_env.yml
conda activate dxpy_env

# Step 5: Log in to DNAnexus using an API token, if not already logged in
# API tokens can be created on the DNAnexus website (https://platform.dnanexus.com/)
dx login --token &lt;token&gt;

# Step 6: Select the desired DNAnexus project
dx select genopred_demo

# Step 7: Upload the prepared data to DNAnexus
dx upload -r genopred_test_data

# Step 8: Clean up temporary files to save space
rm -r test_data test_data.tar.gz genopred_test_data</code></pre>
<hr />
</div>
<div
id="step-1.-install-singularity-and-download-the-genopred-container"
class="section level2">
<h2>Step 1. Install Singularity and Download the GenoPred Container</h2>
<p>Once your RStudio session has started, install Singularity via the
terminal. Follow the commands below to set up Singularity, along with
other essential tools:</p>
<pre class="bash"><code># Configure keyboard layout (optional step to avoid prompts)
echo &#39;keyboard-configuration keyboard-configuration/layout select us&#39; | sudo debconf-set-selections
echo &#39;keyboard-configuration keyboard-configuration/variant select English (US)&#39; | sudo debconf-set-selections

# Install required dependencies (e.g., tmux, build tools, and libraries for Singularity)
sudo DEBIAN_FRONTEND=noninteractive apt update &amp;&amp; \
sudo DEBIAN_FRONTEND=noninteractive apt install -y build-essential libseccomp-dev pkg-config squashfs-tools cryptsetup golang tmux

# Set the Singularity version to install
export VERSION=3.11.0

# Download and extract the Singularity source code
wget https://github.com/sylabs/singularity/releases/download/v${VERSION}/singularity-ce-${VERSION}.tar.gz
tar -xvzf singularity-ce-${VERSION}.tar.gz
cd singularity-ce-${VERSION}

# Configure and build Singularity without SUID
./mconfig --without-suid
make -C builddir

# Install Singularity system-wide
sudo make -C builddir install

# Return to the home directory
cd ~</code></pre>
<p>After installing Singularity, download the GenoPred container using
the following command:</p>
<pre class="bash"><code># Pull the GenoPred container
singularity pull library://opain/genopred/genopred_pipeline:latest</code></pre>
<div class="note-box">
<p><strong>Note:</strong> To make your workflow more efficient and
reproducible, consider saving the downloaded container file in your
project folder. This allows you to re-use the container in future
analyses without needing to download it again.</p>
</div>
<hr />
</div>
<div id="step-2.-import-input-data" class="section level2">
<h2>Step 2. Import input data</h2>
<p>In a DNAnexus project, data is automatically mounted within the
RStudio session and can be accessed from the /mnt/project directory.
However, if the data you need is not located in the project folder, you
will need to import it manually using the dx download command.</p>
<p>For this demonstration, we are working with the small test dataset,
so I will show how to import it. When working with larger datasets, such
as UKB genetic data, it is often more efficient to use the mounted
version of the dispensed data. This approach saves time and disk space
by avoiding the need to download and store large files within the
RStudio instance.</p>
<pre class="bash"><code># View the files in your current project folder
dx ls

# More instructions on using the dx commands can be found here:
# https://documentation.dnanexus.com/getting-started/cli-quickstart

# Import the test data into the instance
dx download -r genopred_test_data</code></pre>
<hr />
</div>
<div id="step-3.-set-up-pipeline-configuration" class="section level2">
<h2>Step 3. Set up pipeline configuration</h2>
<p>Now, we will create the configuration files required to run the
GenoPred pipeline. Note that the <code>outdir</code> and
<code>resdir</code> must be set to directories located outside the
container to ensure proper access and storage.</p>
<pre class="r"><code># Create directories for configuration and output files
dir.create(&#39;/home/rstudio-server/genopred/config&#39;, recursive = TRUE)
dir.create(&#39;/home/rstudio-server/genopred/output&#39;, recursive = TRUE)

# Create gwas_list configuration
gwas_list &lt;- data.frame(
  name = &#39;BODY04&#39;, 
  path = &#39;/home/rstudio-server/genopred_test_data/BODY04.gz&#39;,
  population = &#39;EUR&#39;,
  n = NA,
  sampling = NA,
  prevalence = NA,
  mean = 0,
  sd = 1,
  label = &#39;&quot;Body Mass Index&quot;&#39;
)

write.table(
  gwas_list, 
  &#39;/home/rstudio-server/genopred/config/gwas_list.txt&#39;, 
  col.names = TRUE, 
  row.names = FALSE, 
  quote = FALSE, 
  sep = &#39; &#39;
)

# Create target_list configuration
target_list &lt;- data.frame(
  name = &#39;example_plink1&#39;,
  path = &#39;/home/rstudio-server/genopred_test_data/example&#39;,
  type = &#39;plink2&#39;,
  indiv_report = FALSE
)

write.table(
  target_list, 
  &#39;/home/rstudio-server/genopred/config/target_list.txt&#39;, 
  col.names = TRUE, 
  row.names = FALSE, 
  quote = FALSE, 
  sep = &#39; &#39;
)

# Create main configuration file
conf &lt;- c(
  &#39;outdir: /home/rstudio-server/genopred/output&#39;,
  &#39;resdir: /home/rstudio-server/genopred/resources&#39;,
  &#39;config_file: /home/rstudio-server/genopred/config/config.yaml&#39;,
  &#39;gwas_list: /home/rstudio-server/genopred/config/gwas_list.txt&#39;,
  &#39;target_list: /home/rstudio-server/genopred/config/target_list.txt&#39;,
  &quot;pgs_methods: [&#39;ptclump&#39;]&quot;,
  &#39;testing: chr22&#39;
)

writeLines(
  conf, 
  &#39;/home/rstudio-server/genopred/config/config.yaml&#39;
)</code></pre>
<hr />
</div>
<div id="step-4.-run-the-pipeline-in-the-container"
class="section level2">
<h2>Step 4. Run the pipeline in the container</h2>
<p>Now we can run the GenoPred pipeline. This can be done either
interactively within the container or by executing the desired commands
directly. Running the pipeline interactively is often more convenient
for performing a dry run before launching the full analysis.</p>
<hr />
<div id="start-an-interactive-session-in-the-container"
class="section level3">
<h3>Start an Interactive Session in the Container</h3>
<p>To begin, start an interactive session within the Singularity
container. Make sure to mount the home directory within the RStudio
session to store the outputs:</p>
<pre class="bash"><code>singularity shell \
  --bind /home/rstudio-server:/home/rstudio-server \
  --writable-tmpfs \
  /home/rstudio-server/genopred_pipeline_latest.sif</code></pre>
<hr />
</div>
<div id="run-genopred-inside-the-container" class="section level3">
<h3>Run GenoPred Inside the Container</h3>
<p>Once inside the container, you can use the GenoPred pipeline as
usual:</p>
<pre class="bash"><code># Activate the GenoPred Environment:
source /opt/mambaforge/etc/profile.d/conda.sh
conda activate genopred

# Navigate to the Pipeline Folder:
cd /tools/GenoPred/pipeline

# Perform a Dry Run: 
# A dry run checks the pipeline&#39;s steps without executing them, helping you identify any missing dependencies or issues:
snakemake -n --use-conda --configfile=/home/rstudio-server/genopred/config/config.yaml output_all

# Run the Pipeline: 
# Once satisfied with the dry run, execute the pipeline:
snakemake -j1 --use-conda --configfile=/home/rstudio-server/genopred/config/config.yaml output_all</code></pre>
<p>After the analysis is complete, you can leave the container by
typing:</p>
<pre class="bash"><code>exit</code></pre>
<hr />
</div>
</div>
<div id="step-5.-export-ouputs-to-the-project-folder"
class="section level2">
<h2>Step 5. Export ouputs to the project folder</h2>
<p>To avoid losing the outputs of your analysis when the RStudio session
is terminated, you need to export the results to your DNAnexus project
folder. Both the resdir (resources) and outdir (outputs) should be saved
for future analyses.</p>
<p>For simplicity and efficiency, we will compress the outputs and
resources into a single tar file and then upload it to the DNAnexus
project. If you plan to reuse the resources (e.g., for different
pipeline configurations), you may choose to store them in separate tar
files.</p>
<pre class="bash"><code># Compress the GenoPred working directory
cd /home/rstudio-server
tar -cvf test_run_genopred.tar.gz genopred

# Upload the GenoPred container
dx upload genopred_pipeline_latest.sif

# Upload the tar file containing pipeline resources and outputs
dx upload test_run_genopred.tar.gz</code></pre>
<p>Once the files are uploaded, you can safely terminate the RStudio
session. Ensure the session is fully terminated by checking the Monitor
tab in your DNAnexus project folder.</p>
<hr />
</div>
</div>
<div id="extending-your-analysis-in-the-future" class="section level1">
<h1>Extending your analysis in the future</h1>
<p>If you want to extend your analysis without rerunning steps that have
already completed, you can start a new RStudio session, import the
outputs from a previous run, and resume the pipeline from within the
container. Note that you will also need to import the input data used in
the previous analysis.</p>
<pre class="bash"><code># Download the Outputs from the Previous Run:
dx download test_run_genopred.tar.gz
tar -xvf test_run_genopred.tar.gz

# Download the Singularity Container:
dx download genopred_pipeline_latest.sif

# Download the Input Data:
dx download -r genopred_test_data</code></pre>
<p>When using dx download for files that are not part of a tar archive,
the original timestamps are lost. This may confuse GenoPred, as it will
interpret the files as being updated. To fix this, reset the timestamps
of the input files:</p>
<pre class="bash"><code>find /home/rstudio-server/genopred_test_data/ -type f -exec touch -t 200001010101.01 {} +</code></pre>
<p>If the input data is accessed via the automatic mount in
/mnt/project, this step is unnecessary, as the timestamps are preserved.
This another advantage of mounting the input data.</p>
<pre class="bash"><code># Start an Interactive Session in the Container:
singularity shell \
  --bind /home/rstudio-server:/home/rstudio-server \
  --writable-tmpfs \
  /home/rstudio-server/genopred_pipeline_latest.sif

# Activate the GenoPred Environment:
source /opt/mambaforge/etc/profile.d/conda.sh
conda activate genopred

# Navigate to the Pipeline Folder:
cd /tools/GenoPred/pipeline

### Check the Pipeline State: Run a dry run to verify the current state of the pipeline:
snakemake -n --use-conda --configfile=/home/rstudio-server/genopred/config/config.yaml output_all</code></pre>
<p>The pipeline will indicate that there is nothing to be done if the
configuration has not changed and all outputs are up to date.</p>
<p>If an input file is updated (e.g., by changing its timestamp), the
pipeline will automatically rerun only the necessary steps:</p>
<pre class="bash"><code>touch /home/rstudio-server/genopred_test_data/BODY04.gz
snakemake -n --use-conda --configfile=/home/rstudio-server/genopred/config/config.yaml output_all</code></pre>
<p>This dry run will show which steps need to be re-executed due to the
update.</p>
<hr />
</div>
<div id="troubleshooting" class="section level1">
<h1>Troubleshooting</h1>
<p>Please post questions as an issue on the GenoPred GitHub repo <a
href="https://github.com/opain/GenoPred/issues">here</a>.</p>
</div>

<!-- footer.html -->
<hr/>

<div class="centered-container">
<div class="rounded-image-container" style="width: 500px;">
<img src="Images/logo/sponsors.png">
</div>
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
