<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Running GenoPred on DNAnexus</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />
<link rel="shortcut icon" href="Images/logo/favicon.ico">

<link rel="stylesheet" href="styles/night-mode.css" id="nightModeStylesheet">

<script>
function toggleNightMode() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (stylesheet.disabled) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
}
</script>

<label class="switch">
  <input type="checkbox" id="toggleNightMode" checked>
  <span class="slider round"></span>
</label>

<script>
document.getElementById('toggleNightMode').addEventListener('change', function() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (this.checked) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
});
</script>

<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-YR18ZB3PR3"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-YR18ZB3PR3');
</script>


<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="styles/styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img class="logo-img" src="Images/logo/Horizontal_white.png" style="height: 42px;" /></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pipeline
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="pipeline_overview.html">Overview</a>
    </li>
    <li>
      <a href="pipeline_readme.html">Instructions</a>
    </li>
    <li>
      <a href="pipeline_technical.html">Technical documentation</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Research
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="research_index.html">Overview</a>
    </li>
    <li>
      <a href="comparison_of_methods_summary.html">Polygenic Scoring Methods Comparison</a>
    </li>
    <li>
      <a href="Functionally_informed_prediction.html">Quantifying Polygenic Signal Mediated by Altered Gene Expression</a>
    </li>
    <li>
      <a href="Absolute_Conversion.html">Translating Polygenic Scores onto the Absolute Scale</a>
    </li>
  </ul>
</li>
<li>
  <a href="more_index.html">More</a>
</li>
<li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Running GenoPred on DNAnexus</h1>

</div>


<hr />
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>The UK Biobank (UKB) has recently updated its data access policy,
requiring researchers to access and analyze UKB data exclusively via the
UKB Research Analysis Platform (UKB-RAP), hosted by DNAnexus. This
represents a significant shift in workflows, moving from institutional
servers to a cloud-computing environment. DNAnexus and similar
cloud-computing systems are likely to become standard for future
datasets, so this guide aims to provide instructions on running GenoPred
on DNAnexus, with a particular focus on analyzing UKB data. The UKB
dataset presents unique challenges due to its size, making efficient
analysis essential.</p>
<p>Cloud computing involves requesting access to an instance (or virtual
machine) with a specified amount of resources, such as disk space, RAM,
and the number of CPU cores. Once access is granted, users must create
the required software and data environment within the instance—a step
that is often unfamiliar to many. After completing the desired analyses,
users must export any outputs they wish to retain, as any data or
software left on the instance will be deleted once it is terminated.</p>
<p>This process of transferring software and data into and out of the
instance is a significant departure from the experience of working on
personal computers or institutional servers, where you can resume work
seamlessly each time you log in.</p>
<hr />
</div>
<div id="rstudio-vs.-cloud-workstation" class="section level1">
<h1>RStudio vs. Cloud Workstation</h1>
<p>Currently, the easiest way to run GenoPred on DNAnexus is
interactively, using either RStudio or a Cloud Workstation with the
GenoPred software container. Each approach offers unique advantages.</p>
<p>RStudio mounts files directly from a DNAnexus project, eliminating
the need to import data. This feature saves time and disk space,
particularly with large datasets like UKB genetic data.</p>
<p>Unlike RStudio, the cloud workstation requires input data to be
manually imported using dx download, which can be time-consuming and
requires additional storage space. While dxfuse allows project folders
to be mounted (even across projects), it is prone to instability during
long analyses, such as when using UKB.</p>
<p>Another distinction is that cloud workstations can be connected to
via VScode, which some people may prefer to Rstudio.</p>
<p>Given the advantage of stable project folder mounting in RStudio, I
will demonstrate the workflow in that context. For the demo, minimal
resources are required, so I will request an
<code>mem1_ssd1_v2_x2</code> instance type. A similar workflow can also
be applied in the Cloud Workstation, substituting the Singularity
container with the GenoPred Docker container.</p>
<hr />
</div>
<div id="demonstration-using-rstudio" class="section level1">
<h1>Demonstration using Rstudio</h1>
<hr />
<div id="step-0-prepare-test-data" class="section level2">
<h2>Step 0: Prepare test data</h2>
<p>To make the demonstration as similar as possible to working with UKB
data, I will first upload a version of the GenoPred test data into a
DNAnexus project. While it would be simpler to download the test data
directly into the instance, this approach will better mimic the set up
when working with UKB data.</p>
<pre class="bash"><code># Step 1: Download the GenoPred test data from Zenodo
wget -O test_data.tar.gz https://zenodo.org/records/10640650/files/test_data.tar.gz?download=1

# Step 2: Decompress the downloaded file
tar -xf test_data.tar.gz

# Step 3: Extract only the necessary files for the demonstration
mkdir genopred_test_data
mv test_data/target/imputed_sample_plink2/example.chr22* genopred_test_data/
mv test_data/reference/gwas_sumstats/BODY04.gz genopred_test_data/

# Step 4: Load the Conda environment with dxpy installed
# The Conda environment file for installing dxpy is available in the GenoPred repository
# Path: GenoPred/pipeline/misc/dnanexus/dxpy_env.yml
conda activate dxpy_env

# Step 5: Log in to DNAnexus using an API token, if not already logged in
# API tokens can be created on the DNAnexus website (https://platform.dnanexus.com/)
dx login --token &lt;token&gt;

# Step 6: Select the desired DNAnexus project
dx select genopred_demo

# Step 7: Upload the prepared data to DNAnexus
dx upload -r genopred_test_data

# Step 8: Clean up temporary files to save space
rm -r test_data test_data.tar.gz genopred_test_data</code></pre>
<hr />
</div>
<div
id="step-1.-install-singularity-and-download-the-genopred-container"
class="section level2">
<h2>Step 1. Install Singularity and Download the GenoPred Container</h2>
<p>Once your RStudio session has started, install Singularity via the
terminal. Follow the commands below to set up Singularity, along with
other essential tools:</p>
<pre class="bash"><code># Configure keyboard layout (optional step to avoid prompts)
echo &#39;keyboard-configuration keyboard-configuration/layout select us&#39; | sudo debconf-set-selections
echo &#39;keyboard-configuration keyboard-configuration/variant select English (US)&#39; | sudo debconf-set-selections

# Install required dependencies (e.g., tmux, build tools, and libraries for Singularity)
sudo DEBIAN_FRONTEND=noninteractive apt update &amp;&amp; \
sudo DEBIAN_FRONTEND=noninteractive apt install -y build-essential libseccomp-dev pkg-config squashfs-tools cryptsetup golang tmux

# Set the Singularity version to install
export VERSION=3.11.0

# Download and extract the Singularity source code
wget https://github.com/sylabs/singularity/releases/download/v${VERSION}/singularity-ce-${VERSION}.tar.gz
tar -xvzf singularity-ce-${VERSION}.tar.gz
cd singularity-ce-${VERSION}

# Configure and build Singularity without SUID
./mconfig --without-suid
make -C builddir

# Install Singularity system-wide
sudo make -C builddir install

# Return to the home directory
cd ~</code></pre>
<div class="note-box">
<p><strong>Note:</strong> This will need to be done for every new
instance of Rstudio that you want to run GenoPred. For convenience, you
can store the code in a shell script.</p>
</div>
<p>After installing Singularity, download the GenoPred container using
the following command:</p>
<pre class="bash"><code># Pull the GenoPred container
singularity pull library://opain/genopred/genopred_pipeline:latest</code></pre>
<div class="note-box">
<p><strong>Note:</strong> To make your workflow more efficient and
reproducible, consider saving the downloaded container file in your
project folder. This allows you to re-use the container in future
analyses without needing to download it again.</p>
</div>
<hr />
</div>
<div id="step-2.-import-input-data" class="section level2">
<h2>Step 2. Import input data</h2>
<p>In a DNAnexus project, data is automatically mounted within the
RStudio session and can be accessed from the /mnt/project directory.
However, if the data you need is not located in the project folder, you
will need to import it manually using the dx download command.</p>
<p>For this demonstration, we are working with the small test dataset,
so I will show how to import it. When working with larger datasets, such
as UKB genetic data, it is often more efficient to use the mounted
version of the dispensed data. This approach saves time and disk space
by avoiding the need to download and store large files within the
RStudio instance.</p>
<pre class="bash"><code># View the files in your current project folder
dx ls

# More instructions on using the dx commands can be found here:
# https://documentation.dnanexus.com/getting-started/cli-quickstart

# Import the test data into the instance
dx download -r genopred_test_data</code></pre>
<hr />
</div>
<div id="step-3.-set-up-pipeline-configuration" class="section level2">
<h2>Step 3. Set up pipeline configuration</h2>
<p>Now, we will create the configuration files required to run the
GenoPred pipeline. Note that the <code>outdir</code> and
<code>resdir</code> must be set to directories located outside the
container to ensure proper access and storage.</p>
<pre class="r"><code># Create directories for configuration and output files
dir.create(&#39;/home/rstudio-server/genopred/config&#39;, recursive = TRUE)
dir.create(&#39;/home/rstudio-server/genopred/output&#39;, recursive = TRUE)

# Create gwas_list configuration
gwas_list &lt;- data.frame(
  name = &#39;BODY04&#39;, 
  path = &#39;/home/rstudio-server/genopred_test_data/BODY04.gz&#39;,
  population = &#39;EUR&#39;,
  n = NA,
  sampling = NA,
  prevalence = NA,
  mean = 0,
  sd = 1,
  label = &#39;&quot;Body Mass Index&quot;&#39;
)

write.table(
  gwas_list, 
  &#39;/home/rstudio-server/genopred/config/gwas_list.txt&#39;, 
  col.names = TRUE, 
  row.names = FALSE, 
  quote = FALSE, 
  sep = &#39; &#39;
)

# Create target_list configuration
target_list &lt;- data.frame(
  name = &#39;example_plink1&#39;,
  path = &#39;/home/rstudio-server/genopred_test_data/example&#39;,
  type = &#39;plink2&#39;,
  indiv_report = FALSE
)

write.table(
  target_list, 
  &#39;/home/rstudio-server/genopred/config/target_list.txt&#39;, 
  col.names = TRUE, 
  row.names = FALSE, 
  quote = FALSE, 
  sep = &#39; &#39;
)

# Create main configuration file
conf &lt;- c(
  &#39;outdir: /home/rstudio-server/genopred/output&#39;,
  &#39;resdir: /home/rstudio-server/genopred/resources&#39;,
  &#39;config_file: /home/rstudio-server/genopred/config/config.yaml&#39;,
  &#39;gwas_list: /home/rstudio-server/genopred/config/gwas_list.txt&#39;,
  &#39;target_list: /home/rstudio-server/genopred/config/target_list.txt&#39;,
  &quot;pgs_methods: [&#39;ptclump&#39;]&quot;,
  &#39;testing: chr22&#39;
)

writeLines(
  conf, 
  &#39;/home/rstudio-server/genopred/config/config.yaml&#39;
)</code></pre>
<hr />
</div>
<div id="step-4.-run-the-pipeline-in-the-container"
class="section level2">
<h2>Step 4. Run the pipeline in the container</h2>
<p>Now we can run the GenoPred pipeline. This can be done either
interactively within the container or by executing the desired commands
directly. Running the pipeline interactively is often more convenient
for performing a dry run before launching the full analysis.</p>
<hr />
<div id="start-an-interactive-session-in-the-container"
class="section level3">
<h3>Start an Interactive Session in the Container</h3>
<p>To ensure that your analysis persists even when the RStudio server
tab is closed, it is recommended to start the container within a tmux
session. This will allow you to detach and reattach to the session as
needed.</p>
<p>Start a tmux session within the terminal by running:</p>
<pre class="bash"><code>tmux</code></pre>
<p>This will take you into a <code>tmux</code> session. You can ‘detach’
from the <code>tmux</code> session by pressing <code>Ctrl+b d</code>,
and reattach to the session in the future by typing:</p>
<pre class="bash"><code>tmux attach</code></pre>
<p>Further instructions on using <code>tmux</code> can be found <a
href="https://www.redhat.com/en/blog/introduction-tmux-linux">here</a>.</p>
<p>To begin, start an interactive session within the Singularity
container. Make sure to mount the home directory within the RStudio
session to store the outputs:</p>
<pre class="bash"><code>singularity shell \
  --bind /home/rstudio-server:/home/rstudio-server \
  --writable-tmpfs \
  /home/rstudio-server/genopred_pipeline_latest.sif</code></pre>
<hr />
</div>
<div id="run-genopred-inside-the-container" class="section level3">
<h3>Run GenoPred Inside the Container</h3>
<p>Once inside the container, you can use the GenoPred pipeline as
usual:</p>
<pre class="bash"><code># Activate the GenoPred Environment:
source /opt/mambaforge/etc/profile.d/conda.sh
conda activate genopred

# Navigate to the Pipeline Folder:
cd /tools/GenoPred/pipeline

# Perform a Dry Run: 
# A dry run checks the pipeline&#39;s steps without executing them, helping you identify any missing dependencies or issues:
snakemake -n --use-conda --configfile=/home/rstudio-server/genopred/config/config.yaml output_all

# Run the Pipeline: 
# Once satisfied with the dry run, execute the pipeline:
snakemake -j1 --use-conda --configfile=/home/rstudio-server/genopred/config/config.yaml output_all</code></pre>
<p>While this analysis is running, you can detach from the
<code>tmux</code> session, close the RStudio tab, and close your
browser. When you reopen the RStudio app, you may see that your session
appears suspended. Do not worry—your analysis will continue running in
the background.</p>
<p>By using <code>tmux</code>, your analysis will continue to run even
if the terminal session or RStudio server tab is closed.</p>
<p>After the analysis is complete, you can leave the container by
typing:</p>
<pre class="bash"><code>exit</code></pre>
<hr />
</div>
</div>
<div id="step-5.-export-ouputs-to-the-project-folder"
class="section level2">
<h2>Step 5. Export ouputs to the project folder</h2>
<p>To avoid losing the outputs of your analysis when the RStudio session
is terminated, you need to export the results to your DNAnexus project
folder. Both the resdir (resources) and outdir (outputs) should be saved
for future analyses.</p>
<p>For simplicity and efficiency, we will compress the outputs and
resources into a single tar file and then upload it to the DNAnexus
project. If you plan to reuse the resources (e.g., for different
pipeline configurations), you may choose to store them in separate tar
files.</p>
<pre class="bash"><code># Compress the GenoPred working directory
cd /home/rstudio-server
tar -cvf test_run_genopred.tar genopred

# Upload the GenoPred container
dx upload genopred_pipeline_latest.sif

# Upload the tar file containing pipeline resources and outputs
dx upload test_run_genopred.tar</code></pre>
<p>Once the files are uploaded, you can safely terminate the RStudio
session. Ensure the session is fully terminated by checking the Monitor
tab in your DNAnexus project folder.</p>
<hr />
</div>
</div>
<div id="extending-your-analysis-in-the-future" class="section level1">
<h1>Extending your analysis in the future</h1>
<p>If you want to extend your analysis without rerunning steps that have
already completed, you can start a new RStudio session, import the
outputs from a previous run, and resume the pipeline from within the
container. Note that you will also need to import the input data used in
the previous analysis.</p>
<pre class="bash"><code># Download the Outputs from the Previous Run:
dx download test_run_genopred.tar
tar -xvf test_run_genopred.tar

# Download the Singularity Container:
dx download genopred_pipeline_latest.sif

# Download the Input Data:
dx download -r genopred_test_data</code></pre>
<p>When using dx download for files that are not part of a tar archive,
the original timestamps are lost. This may confuse GenoPred, as it will
interpret the files as being updated. To fix this, reset the timestamps
of the input files:</p>
<pre class="bash"><code>find /home/rstudio-server/genopred_test_data/ -type f -exec touch -t 200001010101.01 {} +</code></pre>
<p>If the input data is accessed via the automatic mount in
/mnt/project, this step is unnecessary, as the timestamps are preserved.
This another advantage of mounting the input data.</p>
<pre class="bash"><code># Start an Interactive Session in the Container:
singularity shell \
  --bind /home/rstudio-server:/home/rstudio-server \
  --writable-tmpfs \
  /home/rstudio-server/genopred_pipeline_latest.sif

# Activate the GenoPred Environment:
source /opt/mambaforge/etc/profile.d/conda.sh
conda activate genopred

# Navigate to the Pipeline Folder:
cd /tools/GenoPred/pipeline

### Check the Pipeline State: Run a dry run to verify the current state of the pipeline:
snakemake -n --use-conda --configfile=/home/rstudio-server/genopred/config/config.yaml output_all</code></pre>
<p>The pipeline will indicate that there is nothing to be done if the
configuration has not changed and all outputs are up to date.</p>
<p>If an input file is updated (e.g., by changing its timestamp), the
pipeline will automatically rerun only the necessary steps:</p>
<pre class="bash"><code>touch /home/rstudio-server/genopred_test_data/BODY04.gz
snakemake -n --use-conda --configfile=/home/rstudio-server/genopred/config/config.yaml output_all</code></pre>
<p>This dry run will show which steps need to be re-executed due to the
update.</p>
<hr />
</div>
<div id="running-genopred-with-uk-biobank" class="section level1">
<h1>Running GenoPred with UK Biobank</h1>
<p>The UKB imputed genetic data is provided without post-imputation QC,
resulting in large files. The <code>format_target</code> step of
GenoPred, which reformats the target genetic data, is time-intensive but
reduces file size to ~86GB, making future analyses faster and cheaper.
Storing this output for reuse is highly recommended. Additionally,
selecting instances with appropriate resources for each pipeline stage
ensures cost efficiency, as some steps utilize multiple cores while
others do not.</p>
<hr />
<div id="key-adjustments-for-ukb-data" class="section level2">
<h2>Key Adjustments for UKB Data</h2>
<ul>
<li>Mounted UKB Data: Avoid duplicating the large UKB data files by
using a mounted version to save time and disk space.</li>
<li>File Naming with Symbolic Links: Update file names as required by
GenoPred using symbolic links to avoid duplicating data.</li>
<li>High-Resource Instance for <code>format_target</code>: Use a
<code>mem3_ssd2_v2_x8</code> instance, which supports 8 processes and
provides sufficient RAM and disk space for processing the large UKB
files efficiently. Other steps of the pipeline will have other resources
requirements.</li>
</ul>
<hr />
</div>
<div id="workflow" class="section level2">
<h2>Workflow</h2>
<ul>
<li>Run <code>format_target</code> on a High-Resource Instance: Use the
<code>mem3_ssd2_v2_x8</code> instance to efficiently complete the
<code>format_target</code> step.</li>
<li>Export and Store Processed Data: Save the reformatted data for
future use, avoiding repeated processing.</li>
<li>Terminate the High-Resource Instance: After completing
<code>format_target</code>, terminate the instance to minimize
costs.</li>
<li>Continue Analysis on Cost-Effective Instances: For downstream steps,
switch to an instance with resources tailored to those stages.</li>
<li>This approach balances efficiency and cost when working with
large-scale UKB genetic data in the GenoPred pipeline.</li>
</ul>
<hr />
</div>
<div id="step-1.-install-singularity" class="section level2">
<h2>Step 1. Install Singularity</h2>
<p>Once your RStudio session has started, install Singularity via the
terminal. We will use the shell script we created <a
href="#step-1-install-singularity-and-download-the-genopred-container">here</a>:</p>
<pre class="bash"><code># Download, update permissions and run script to install singularity
dx download install_singularity.sh
chmod a+x install_singularity.sh
./install_singularity</code></pre>
<hr />
</div>
<div id="step-2.-prepare-input-data" class="section level2">
<h2>Step 2. Prepare input data</h2>
<p>We will use a mounted version of the UKB genetic data to save time
and disk space. To meet GenoPred’s file name requirements without
duplicating data, we will create symbolic links. Given the size of the
UKB genetic data, we will also request an instance with sufficient
resources.</p>
<pre class="bash"><code># Create symlinks to the dispensed imputed genetic data
mkdir -p /home/rstudio-server/ukb/ukb_symlinks

# Link bgen and bgen.bgi files for all chromosomes
for chr in $(seq 1 22); do
  for file in $(echo bgen bgen.bgi); do
    ln -s /mnt/project/Bulk/Imputation/UKB\ imputation\ from\ genotype/ukb22828_c${chr}_b0_v3.${file} /home/rstudio-server/ukb/ukb_symlinks/ukb_imp.chr${chr}.${file}
  done
done

# Link the sample file (same for all chromosomes)
ln -s /mnt/project/Bulk/Imputation/UKB\ imputation\ from\ genotype/ukb22828_c1_b0_v3.sample /home/rstudio-server/ukb/ukb_symlinks/ukb_imp.sample</code></pre>
<hr />
</div>
<div id="step-3.-set-up-pipeline-configuration-1"
class="section level2">
<h2>Step 3. Set up pipeline configuration</h2>
<p>Next, create the configuration files required to run the GenoPred
pipeline. Ensure that <code>outdir</code> and <code>resdir</code> are
set to directories outside the container for proper access and
storage.</p>
<pre class="r"><code># Create directories for configuration and output files
dir.create(&#39;/home/rstudio-server/genopred/config/ukb/basic&#39;, recursive = T)
dir.create(&#39;/home/rstudio-server/genopred/output&#39;, recursive = T)

# Create target list
# We are specifying the symbolic links we made for the UKB data
target_list &lt;- data.frame(
  name=&#39;ukb&#39;,
  path=&#39;/home/rstudio-server/ukb/ukb_symlinks/ukb_imp&#39;,
  type=&#39;bgen&#39;,
  indiv_report=F
)

write.table(
  target_list,
  &#39;/home/rstudio-server/genopred/config/ukb/basic/target_list.txt&#39;,
  col.names = T,
  row.names = F,
  quote = F
)

# Create config file
conf &lt;- c(
  &#39;outdir: /home/rstudio-server/genopred/output&#39;,
  &#39;config_file: /home/rstudio-server/genopred/config/ukb/basic/config.yaml&#39;,
  &#39;resdir: /home/rstudio-server/genopred/resources&#39;,
  &#39;target_list: /home/rstudio-server/genopred/config/ukb/basic/target_list.txt&#39;
)

write.table(
  conf,
  &#39;/home/rstudio-server/genopred/config/ukb/basic/config.yaml&#39;,
  col.names = F,
  row.names = F,
  quote = F
)</code></pre>
<hr />
</div>
<div id="step-4.-run-the-pipeline-in-the-container-1"
class="section level2">
<h2>Step 4. Run the pipeline in the container</h2>
<p>Now we can run the GenoPred pipeline. This can be done either
interactively within the container or by executing the desired commands
directly. Running the pipeline interactively is often more convenient
for performing a dry run before launching the full analysis.</p>
<hr />
<div id="start-an-interactive-session-in-the-container-1"
class="section level3">
<h3>Start an Interactive Session in the Container</h3>
<p>To ensure that your analysis persists even when the RStudio server
tab is closed, it is recommended to start the container within a tmux
session. This will allow you to detach and reattach to the session as
needed.</p>
<p>Start a tmux session within the terminal by running:</p>
<pre class="bash"><code>tmux</code></pre>
<p>This will take you into a <code>tmux</code> session. You can ‘detach’
from the <code>tmux</code> session by pressing <code>Ctrl+b d</code>,
and reattach to the session in the future by typing:</p>
<pre class="bash"><code>tmux attach</code></pre>
<p>Further instructions on using <code>tmux</code> can be found <a
href="https://www.redhat.com/en/blog/introduction-tmux-linux">here</a>.</p>
<p>To begin, start an interactive session within the Singularity
container. Make sure to mount the home directory within the RStudio
session and the directory that the symbolic links point to, to store the
pipeline outputs and access the input data within the container.</p>
<pre class="bash"><code>singularity shell \
  --bind /home/rstudio-server:/home/rstudio-server \
  --bind /mnt/project/Bulk/Imputation/UKB\ imputation\ from\ genotype:/mnt/project/Bulk/Imputation/UKB\ imputation\ from\ genotype \
  --writable-tmpfs \
  /mnt/project/genopred_pipeline_latest.sif</code></pre>
<hr />
</div>
<div id="run-genopred-inside-the-container-1" class="section level3">
<h3>Run GenoPred Inside the Container</h3>
<p>Once inside the container, you can use the GenoPred pipeline as
usual. The resources provided by this instance
(<code>mem3_ssd2_v2_x8</code>) will not be required for all steps in the
GenoPred workflow, so to be cost efficient, I am just carrying out the
<code>format_target</code> step within this instance. I will then export
the data, terminate this instance, continue my analysis using a new
instance with appropriate resources for downstream steps.</p>
<pre class="bash"><code># Activate the GenoPred Environment:
source /opt/mambaforge/etc/profile.d/conda.sh
conda activate genopred

# Navigate to the Pipeline Folder:
cd /tools/GenoPred/pipeline

# Perform a Dry Run: 
# A dry run checks the pipeline&#39;s steps without executing them, helping you identify any missing dependencies or issues:
snakemake -n --use-conda --configfile=/home/rstudio-server/genopred/config/ukb/basic/config.yaml format_target

# Run the Pipeline: 
# Once satisfied with the dry run, execute the pipeline:
snakemake -j8 --use-conda --configfile=/home/rstudio-server/genopred/config/ukb/basic/config.yaml format_target</code></pre>
<div class="note-box">
<p><strong>Note:</strong> Since the instance has 8 cores available, I
use the <code>-j8</code> parameter when running <code>snakemake</code>
to ensure it utilizes all 8 cores efficiently.</p>
</div>
<p>After the analysis is complete, you can leave the container by
typing:</p>
<pre class="bash"><code>exit</code></pre>
<p>In total, this analysis took ~18 hours, costing ~£6.</p>
<hr />
</div>
</div>
<div id="step-5.-export-ouputs-to-the-project-folder-1"
class="section level2">
<h2>Step 5. Export ouputs to the project folder</h2>
<p>To avoid losing the outputs of your analysis when the RStudio session
is terminated, you need to export the results to your DNAnexus project
folder. Both the resdir (resources) and outdir (outputs) should be saved
for future analyses.</p>
<p>For simplicity and efficiency, we will compress the outputs and
resources into a single tar file and then upload it to the DNAnexus
project. If you plan to reuse the resources (e.g., for different
pipeline configurations), you may choose to store them in separate tar
files.</p>
<pre class="bash"><code># Compress and upload the GenoPred working directory
cd /home/rstudio-server
tar -cvf ukb_genopred.tar genopred
dx upload ukb_genopred.tar</code></pre>
<p>I will also tar and upload the symlinks created for the UKB data.
While these could be recreated, this approach conveniently preserves the
timestamps, making it easier to resume the analysis seamlessly when
extending it in the future.</p>
<pre class="bash"><code># Compress and upload the ukb directory containing symlinks
cd /home/rstudio-server
tar -cvf ukb_symlinks.tar ukb
dx upload ukb_symlinks.tar</code></pre>
<p>Once the files are uploaded, you can safely terminate the RStudio
session. Ensure the session is fully terminated by checking the Monitor
tab in your DNAnexus project folder.</p>
<hr />
</div>
<div id="step-6.-ancestry-inference-and-within-target-qc"
class="section level2">
<h2>Step 6. Ancestry Inference and Within Target QC</h2>
<p>The ancestry inference step is required prior to polygenic scoring,
so we will do this now. In the same session, we will also perform the
within-sample QC and project reference principal components, which
generate other useful outputs.</p>
<hr />
<div id="start-an-rstudio-session" class="section level3">
<h3>Start an Rstudio session</h3>
<p>Neither of the these steps require much RAM. The within-sample QC can
leverage multiple cores, but ancestry inference doesn’t. We need enough
disk space to import the output from the previous run, but not much
more. In this demonstration I am using a <code>mem2_ssd1_v2_x8</code>
instance, which seemed to work well.</p>
<hr />
</div>
<div id="install-singularity" class="section level3">
<h3>Install Singularity</h3>
<p>Once your RStudio session has started, install Singularity via the
terminal. We will use the shell script we created <a
href="#step-1-install-singularity-and-download-the-genopred-container">here</a>:</p>
<pre class="bash"><code># Download, update permissions and run script to install singularity
dx download install_singularity.sh
chmod a+x install_singularity.sh
./install_singularity</code></pre>
<hr />
</div>
<div id="import-the-data-from-previous-session" class="section level3">
<h3>Import the data from previous session</h3>
<p>We are going to extend our previous analysis UKB using GenoPred. We
need to recreate the environment we had before.</p>
<pre class="bash"><code># Download and decompress the symlinks previous run of GenoPred
# Decompressing the mounted data save time and disk space
tar -xvf /mnt/project/ukb_symlinks.tar -C ~/
tar -xvf /mnt/project/ukb_genopred.tar -C ~/</code></pre>
<hr />
</div>
<div id="estimate-relatedness" class="section level3">
<h3>Estimate relatedness</h3>
<p>Although GenoPred can estimate relatedness from scratch, UKB is very
large and it will be time consuming. Instead, we will use the kinship
matrix released by UKBto identify a list of unrelated individuals, using
the software <a
href="https://gitlab.com/choishingwan/GreedyRelated">GreedyRelated</a>,
a script I have written previously <a
href="https://github.com/opain/GenoPred/tree/master/Scripts/ukb_relative_remover/ukb_relative_remover.R">ukb_relative_remover.R</a>.</p>
<pre class="bash"><code># Install optparse in R
Rscript -e &quot;install.packages(&#39;optparse&#39;, repos=&#39;https://cloud.r-project.org/&#39;)&quot;

# Download and give permission to GreedyRelated binary for linux
wget https://gitlab.com/-/project/14754196/uploads/6fdc44072a8a866cb77c2cf91f68d662/GreedyRelated_linux
chmod a+x GreedyRelated_linux

# Identify list of related individuals (using threshold of 0.044)
Rscript ukb_relative_remover.R \
  --rel_file /mnt/project/Bulk/Genotype\ Results/Genotype\ calls/ukb_rel.dat \
  --rel_thresh 0.044 \
  --seed 1 \
  --GreedyRelated ./GreedyRelated_linux \
  --output ukb/ukb82087</code></pre>
<pre class="r"><code># Identify a list of unrelated individuals by removing this list of related indivduals from the full list of individuals in the UKB .fam file
library(data.table)
related &lt;- fread(&#39;ukb/ukb82087.related&#39;)$V1
fam&lt;-fread(&#39;/mnt/project/Bulk/Genotype Results/Genotype calls/ukb22418_c1_b0_v2.fam&#39;)$V1

unrelated&lt;-fam[!(fam %in% related)]
write.table(unrelated, &#39;ukb/ukb82087.unrelated&#39;, col.names=F, row.names = F, quote = F)</code></pre>
<hr />
</div>
<div id="update-configuration-for-relatedness" class="section level3">
<h3>Update configuration for relatedness</h3>
<p>We must update the target_list configuration file for GenoPred to
indicate the location of the file indicating unrelated individuals.</p>
<pre class="r"><code>target_list &lt;- data.frame(
  name=&#39;ukb&#39;,
  path=&#39;/home/rstudio-server/ukb/ukb_symlinks/ukb_imp&#39;,
  type=&#39;bgen&#39;,
  indiv_report=F,
  unrel=&#39;/home/rstudio-server/ukb/ukb82087.unrelated&#39;
)

write.table(
  target_list,
  &#39;/home/rstudio-server/genopred/config/ukb/basic/target_list.txt&#39;,
  col.names = T,
  row.names = F,
  quote = F
)</code></pre>
<hr />
</div>
<div id="run-pipeline" class="section level3">
<h3>Run pipeline</h3>
<pre class="bash"><code># Start a tmux session to ensure the analysis persists even the connection is lost
tmux

# Start an interactive session inside the GenoPred container
singularity shell \
  --bind /home/rstudio-server:/home/rstudio-server \
  --bind /mnt/project/Bulk/Imputation/UKB\ imputation\ from\ genotype:/mnt/project/Bulk/Imputation/UKB\ imputation\ from\ genotype \
  --writable-tmpfs \
  /mnt/project/genopred_pipeline_latest.sif

# Activate the GenoPred Environment:
source /opt/mambaforge/etc/profile.d/conda.sh
conda activate genopred

# Navigate to the Pipeline Folder:
cd /tools/GenoPred/pipeline

# Perform a Dry Run: 
# A dry run checks the pipeline&#39;s steps without executing them, helping you identify any missing dependencies or issues
# We can see that GenoPred will pick up where it left off, and won&#39;t rerun steps it ran before.
snakemake -n --use-conda --configfile=/home/rstudio-server/genopred/config/ukb/basic/config.yaml ancestry_inference outlier_detection pc_projection

# Run the analysis. Here I am using 8 cores since I am using an instance with 8 cores available (mem2_ssd1_v2_x8).
snakemake -j8 --use-conda --configfile=/home/rstudio-server/genopred/config/ukb/basic/config.yaml ancestry_inference outlier_detection pc_projection
</code></pre>
<hr />
</div>
<div id="package-and-export-outputs" class="section level3">
<h3>Package and export outputs</h3>
<p>Once the analysis is complete, we will compress and
<code>dx upload</code> the output. The <code>genopred</code> output
folder will contain the contents of the previous run as well (the
reformatted UKB data), so we can delete the old version of the
<code>ukb_genopred.tar</code> file in our project folder after the
upload of the new version is complete. The same goes for the
<code>ukb</code>/<code>ukb_symlinks.tar</code> folder - We need to
reupload this since we are now storing the list of unrelated individuals
in there.</p>
<pre class="bash"><code># Compress and upload the GenoPred working directory
cd /home/rstudio-server
tar -cvf ukb_genopred.tar genopred
dx upload ukb_genopred.tar

# Compress and upload the ukb directory containing symlinks and list of unrelated individuals
tar -cvf ukb_symlinks.tar ukb
dx upload ukb_symlinks.tar</code></pre>
<hr />
</div>
</div>
<div id="step-7.-generating-score-files" class="section level2">
<h2>Step 7. Generating score files</h2>
<p>Score files can be generated using GenoPred either on DNAnexus or on
other platforms, as this step does not require access to UK Biobank
(UKB) data.</p>
<p>Notably, score files generated in one instance of GenoPred (or with
other software) can be reused as input for another instance of GenoPred.
For example, you can:</p>
<ol style="list-style-type: decimal">
<li><p>Generate score files using GenoPred on an institutional server
(e.g., for free or with existing resources).</p></li>
<li><p>Copy these score files to DNAnexus and use them to perform target
sample scoring in the UKB dataset on DNAnexus.</p></li>
</ol>
<p>There are already several demonstrations of running GenoPred on
DNAnexus in this document. The same setup can be used to generate score
files. So, I will focus on demonstrating the more common scenario of
importing PGS score files from a previous run of GenoPred, to be used
for target sample scoring in UKB on DNAnexus.</p>
<p>There are two approaches for using scores files from a different run
of GenoPred. The score file can be reformated to the PGS catalogue
format, and included in the <code>score_list</code>, but this requires
one set of weights per file, which is inefficient, and looses some
functionality downstream such as the function to return the
pseudovalidated score (<code>find_pseudo()</code>). An alternative
solution, is to copy the input GWAS sumstats, QC’d sumstats, and the PGS
score files, which provides full downstream functionality.</p>
<p>The most convenient solution will depend on your needs. If you just
want to use a relatively fast PGS method, then you might as well run on
DNAnexus as it won’t cost much. If you want to use computationally
intensive methods, then you may want to save money by running on your
institutional server, and then importing the output to DNAnexus. If you
only want to use a single score from the computationally intensive
method, then exporting that score alone and specifying it using the
<code>score_list</code> will be most convenient. However, if you want
full functionality of GenoPred, whilst running PGS methods on a
different server, then copying the entire GWAS and score file
directories from a previous run onto DNAnexus is needed. I will
demonstrate only the final scenario as it is the most convoluted.</p>
<hr />
</div>
<div id="step-8.-target-sampling-scoring" class="section level2">
<h2>Step 8. Target sampling scoring</h2>
<p>Here I will use a score file generated using GenoPred previously. The
score file was generated using the an coronary artery disease GWAS and
the <code>ptclump</code> method. I have uploaded it to my DNAnexus
project from my institutional server using the dx upload function
(similar to I did in <a href="(#step-0-prepare-test-data)">this
section</a>).</p>
<pre class="bash"><code># Package and upload the required sumstats and score files to the DNAnexus project folder.
mkdir -p genopred_scores/gwas_sumstat
mkdir -p genopred_scores/pgs_score_files/sbayesr

cp ~/oliverpainfel/GenoPred/pipeline/example_input/gwas_list.txt genopred_scores/
cp -r ~/oliverpainfel/GenoPred/pipeline/test_data/output/test1/reference/gwas_sumstat/COAD01 genopred_scores/gwas_sumstat/
cp -r ~/oliverpainfel/GenoPred/pipeline/test_data/output/test1/reference/pgs_score_files/ptclump/COAD01 genopred_scores/pgs_score_files/ptclump/

tar -cvf genopred_scores.tar genopred_scores
dx upload genopred_scores.tar</code></pre>
<p>Now I will spin up a new instance in Rstudio to perform target sample
scoring in UKB. Using <code>mem2_ssd1_v2_x4</code> instance.</p>
<ul>
<li>Install Singularity (same as before - I put it in a shell script to
make it easier to run)</li>
</ul>
<pre class="bash"><code>dx download install_singularity.sh
chmod a+x install_singularity.sh
./install_singularity.sh</code></pre>
<ul>
<li>Import the data from previous session (Decompress from the mounted
version to save disk space and time)</li>
</ul>
<pre class="bash"><code># Import GenoPred inputs relating to UKB
tar -xvf /mnt/project/ukb_symlinks.tar -C ~/
tar -xvf /mnt/project/ukb_genopred.tar -C ~/

# Import GenoPred outputs from PGS methods, and move into the apropriate genopred folder 
tar -xvf /mnt/project/genopred_scores.tar -C ~/
mv genopred_scores/gwas_sumstat ~/genopred/output/reference/
mv genopred_scores/pgs_score_files ~/genopred/output/reference/</code></pre>
<ul>
<li>Update configuration to specify the GWAS, score files and PGS
methods to use.</li>
</ul>
<p>The gwas_list and pgs_methods should match the configuration used to
generate the score files. However, we will need to create empty files to
represent the original raw GWAS summary statistics, which we did not
copy over to DNAnexus.</p>
<pre class="r"><code># Make an empty file to represent the unQC&#39;d sumstats
library(data.table)
gwas_list&lt;-fread(&#39;~/genopred_scores/gwas_list.txt&#39;)
gwas_list&lt;-gwas_list[gwas_list$name == &#39;COAD01&#39;,]
dir.create(&#39;/home/rstudio-server/raw_sumstats&#39;)
for(i in 1:nrow(gwas_list)){
  path &lt;- paste0(&#39;/home/rstudio-server/raw_sumstats/&#39;, gwas_list$name[i],&#39;.txt&#39;)
  file.create(path)
  gwas_list$path[i] &lt;- path
}

gwas_list$label&lt;-paste0(&quot;\&quot;&quot;, gwas_list$label, &quot;\&quot;&quot;)

dir.create(&#39;/home/rstudio-server/genopred/config/ukb/demo&#39;)

write.table(
  gwas_list,
  &#39;/home/rstudio-server/genopred/config/ukb/demo/gwas_list.txt&#39;,
  col.names = T,
  row.names = F,
  quote = F
)

# Create config file
conf &lt;- c(
  &#39;outdir: /home/rstudio-server/genopred/output&#39;,
  &#39;config_file: /home/rstudio-server/genopred/config/ukb/demo/config.yaml&#39;,
  &#39;resdir: /home/rstudio-server/genopred/resources&#39;,
  &#39;gwas_list: /home/rstudio-server/genopred/config/ukb/demo/gwas_list.txt&#39;,
  &quot;pgs_methods: [&#39;ptclump&#39;]&quot;,
  &#39;target_list: /home/rstudio-server/genopred/config/ukb/basic/target_list.txt&#39;,
  &#39;cores_target_pgs: 1&#39;
)

write.table(
  conf,
  &#39;/home/rstudio-server/genopred/config/ukb/demo/config.yaml&#39;,
  col.names = F,
  row.names = F,
  quote = F
)</code></pre>
<ul>
<li>Run pipeline</li>
</ul>
<pre class="bash"><code># Start a tmux session
tmux

# Start interactive session in container
singularity shell \
  --bind /home/rstudio-server:/home/rstudio-server \
  --bind /mnt/project/Bulk/Imputation/UKB\ imputation\ from\ genotype:/mnt/project/Bulk/Imputation/UKB\ imputation\ from\ genotype \
  --writable-tmpfs \
  /mnt/project/genopred_pipeline_latest.sif

# Activate the GenoPred Environment:
source /opt/mambaforge/etc/profile.d/conda.sh
conda activate genopred

# Navigate to the Pipeline Folder:
cd /tools/GenoPred/pipeline

# It will think the score files need to be recreated due to the sumstat paths changing. So touch the outputs of prep_pgs
# This just updates the file timestamps for step prior to prep_pgs so the pipeline doesn&#39;t think it needs to recreate them due to the raw sumstats being newer than the score files.
snakemake -t -j1 --use-conda --configfile=/home/rstudio-server/genopred/config/ukb/demo/config.yaml prep_pgs

# Perform a Dry Run: 
# A dry run checks the pipeline&#39;s steps without executing them, helping you identify any missing dependencies or issues.
snakemake -n --use-conda --configfile=/home/rstudio-server/genopred/config/ukb/demo/config.yaml output_all

# We can see it will only run target scoring and downstream steps, as it should.
# Now we can run using four cores, matching the resources available in our instance (mem2_ssd1_v2_x4)
snakemake -j4 --use-conda --configfile=/home/rstudio-server/genopred/config/ukb/demo/config.yaml output_all</code></pre>
<ul>
<li>Package and export PGS</li>
</ul>
<p>You could tar and export the entire <code>genopred</code> folder
again, or you could export just the files you need, such as the score
files and the report.</p>
<pre class="bash"><code># Upload report
dx upload genopred/output/ukb/reports/ukb-report.html</code></pre>
<p>It could also be convenient to store the PGS in an .RDS file, and
then export that file.</p>
<pre class="bash"><code>export LC_ALL=C
export LANG=C</code></pre>
<pre class="r"><code>setwd(&#39;/tools/GenoPred/pipeline&#39;)
library(data.table)

source(&#39;../functions/misc.R&#39;)
source_all(&#39;../functions&#39;)

# Read in PGS
pgs &lt;- read_pgs(config = &#39;/home/rstudio-server/genopred/config/ukb/demo/config.yaml&#39;)

saveRDS(pgs, file = &quot;/home/rstudio-server/ukb_pgs_COAD01.rds&quot;)</code></pre>
<pre class="bash"><code>dx upload ukb_pgs_COAD01.rds</code></pre>
<hr />
</div>
</div>
<div id="troubleshooting" class="section level1">
<h1>Troubleshooting</h1>
<p>Please post questions as an issue on the GenoPred GitHub repo <a
href="https://github.com/opain/GenoPred/issues">here</a>.</p>
</div>

<!-- footer.html -->
<hr/>

<div class="centered-container">
<div class="rounded-image-container" style="width: 500px;">
<img src="Images/logo/sponsors.png">
</div>
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
