<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Standardised vs. Non-Standardised</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-sm-12 col-md-4 col-lg-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-sm-12 col-md-8 col-lg-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GenoPred</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fas fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Standardised vs. Non-Standardised</h1>

</div>


<style>
p.caption {
  font-size: 1.5em;
}
</style>
<style type="text/css">
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
</style>
<hr />
<p>This page compares my reference standardised polygenic scores to non-reference standardised polygenic scores.</p>
<div id="original-scores" class="section level1">
<h1><span class="header-section-number">1</span> Original scores</h1>
<p>Polygenic scores for UKB were derived using PRSice, genotyped SNPs only, including the MHC region, with LD-estimates from the UKB sample. My polygenic scores were derived without using the pT+clump method but not using PRSice, I used HapMap3 variants from the UKB imputed genetic data, I used only the single top SNP from the MHC region, and LD estimates from the EUR popultion in 1KG Phase 3.</p>
<p>Check the correlation between the two types of polygenic scores and how well they predict outcomes.</p>
<details>
<p><summary>Show code</summary></p>
<pre class="r"><code>library(data.table)

pheno_name&lt;-c(&#39;BMI&#39;,&#39;CAD&#39;,&#39;Intelligence&#39;,&#39;IBD&#39;,&#39;T2D&#39;,&#39;Height&#39;,&#39;RheuArth&#39;,&#39;MultiScler&#39;)
gwas&lt;-c(&#39;BODY03&#39;,&#39;COAD01&#39;,&#39;COLL01&#39;,&#39;CROH01&#39;,&#39;DIAB05&#39;,&#39;HEIG03&#39;,&#39;RHEU01&#39;,&#39;SCLE02&#39;)

Results&lt;-NULL
cor_res&lt;-NULL
for(i in 1:length(gwas)){
  PRS_geno&lt;-fread(paste0(&#39;/mnt/lustre/groups/ukbiobank/sumstats/PRS/ukb18177_glanville/PRS_for_use/&#39;,gwas[i],&#39;_header.all.score&#39;))
  PRS_stand&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Data/UKBB/PRS_for_comparison/1KG_ref/pt_clump/&#39;,gwas[i],&#39;/UKBB.subset.w_hm3.&#39;,gwas[i],&#39;.profiles&#39;))
  PRS_geno$FID&lt;-NULL
  PRS_stand$FID&lt;-NULL

  # Extract individuals in common between the two PRSs
  overlap&lt;-intersect(PRS_geno$IID, PRS_stand$IID)
  PRS_geno&lt;-PRS_geno[PRS_geno$IID %in% overlap,]
  PRS_stand&lt;-PRS_stand[PRS_stand$IID %in% overlap,]

  # Extract individuals surviving QC
  keep&lt;-fread(&#39;/users/k1806347/brc_scratch/Analyses/PRS_comparison/UKBB_outcomes_for_prediction/ukb18177_glanville_post_qc_id_list.UpdateIDs.fam&#39;)
  PRS_geno&lt;-PRS_geno[PRS_geno$IID %in% keep$V2,]
  PRS_stand&lt;-PRS_stand[PRS_stand$IID %in% keep$V2,]

  # Extract pTs in common
  overlap_pT&lt;-intersect(names(PRS_stand), names(PRS_geno))
  pTs&lt;-overlap_pT[grepl(gwas[i], overlap_pT)]
  PRS_geno&lt;-PRS_geno[,names(PRS_geno) %in% overlap_pT, with=F]
  PRS_stand&lt;-PRS_stand[,names(PRS_stand) %in% overlap_pT, with=F]

  # Read in the phenotype file
  pheno&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Data/UKBB/Phenotype/PRS_comp_subset/UKBB.&#39;,pheno_name[i],&#39;.txt&#39;))
  pheno$FID&lt;-NULL
  pheno_var&lt;-names(pheno)[2]

  # Merge pheno and PRS
  PRS_geno_pheno&lt;-merge(pheno, PRS_geno, by=&#39;IID&#39;)
  PRS_stand_pheno&lt;-merge(pheno, PRS_stand, by=&#39;IID&#39;)

  # Run regressions
  for(k in 1:length(pTs)){
    geno_PRS_mod&lt;-lm(as.formula(paste0(&#39;scale(&#39;,pheno_var,&#39;) ~ scale(&#39;,pTs[k],&#39;)&#39;)), data=PRS_geno_pheno)
    geno_PRS_mod_summary&lt;-summary(geno_PRS_mod)
    stand_PRS_mod&lt;-lm(as.formula(paste0(&#39;scale(&#39;,pheno_var,&#39;) ~ scale(&#39;,pTs[k],&#39;)&#39;)), data=PRS_stand_pheno)
    stand_PRS_mod_summary&lt;-summary(stand_PRS_mod)

    Results&lt;-rbind(Results, data.frame( Phenotype=rep(pheno_name[i],2),
                                        pT=rep(gsub(paste0(gwas[i],&#39;_&#39;),&#39;&#39;,pTs[k]),2),
                                        Type=c(&#39;Current   &#39;,&#39;Standardised&#39;),
                                        BETA=c(coef(geno_PRS_mod_summary)[2,1],coef(stand_PRS_mod_summary)[2,1]),
                                        SE=c(coef(geno_PRS_mod_summary)[2,2],coef(stand_PRS_mod_summary)[2,2])))
  }
  
  # Calculate correlation between the two PRS
  names(PRS_stand)[-1]&lt;-paste0(names(PRS_stand),&#39;_geno&#39;)[-1]
  prs_both&lt;-merge(PRS_geno,PRS_stand,by=&#39;IID&#39;)
  
  pT&lt;-gsub(&#39;.*_&#39;,&#39;&#39;,names(PRS_geno)[-1])
    
  for(k in pT){
    res&lt;-cor(prs_both[[paste0(gwas[i],&#39;_&#39;,k)]], prs_both[[paste0(gwas[i],&#39;_&#39;,k,&#39;_geno&#39;)]])
    cor_res&lt;-rbind(cor_res, data.frame( GWAS=gwas[i],
                                                            pT=k,
                                        cor=res))
  }
  
  # Fit model with both PRS
  both_PRS_mod_pheno&lt;-merge(pheno, prs_both, by=&#39;IID&#39;)

  for(k in 1:length(pTs)){
    both_PRS_mod&lt;-lm(as.formula(paste0(&#39;scale(&#39;,pheno_var,&#39;) ~ scale(&#39;,pTs[k],&#39;) + scale(&#39;,pTs[k],&#39;_geno)&#39;)), data=both_PRS_mod_pheno)
    
    both_PRS_mod_summary&lt;-summary(lm(as.formula(paste0(&#39;scale(both_PRS_mod_pheno$&#39;,pheno_var,&#39;) ~ scale(predict(both_PRS_mod, newdata = both_PRS_mod_pheno))&#39;))))
    
    Results&lt;-rbind(Results, data.frame( Phenotype=pheno_name[i],
                                        pT=gsub(paste0(gwas[i],&#39;_&#39;),&#39;&#39;,pTs[k]),
                                        Type=c(&#39;Both   &#39;),
                                        BETA=coef(both_PRS_mod_summary)[2,1],
                                        SE=coef(both_PRS_mod_summary)[2,2]))
  }

}

# Plot the results
library(ggplot2)
library(cowplot)

Plots&lt;-list()
for(i in 1:length(gwas)){
Plots[[pheno_name[i]]]&lt;-ggplot(Results[Results$Phenotype == pheno_name[i],], aes(x=pT, y=BETA, fill=Type)) +
                        geom_bar(position=position_dodge(), stat=&quot;identity&quot;) +
                        geom_errorbar(aes(ymin=BETA-SE, ymax=BETA+SE), width=.2, position=position_dodge(.9)) +
                        labs(title=pheno_name[i], y=&#39;Correlation&#39;, x=paste0(&#39;pT for &#39;,gwas[i],&#39; PRS&#39;)) +
                        theme(axis.text.x = element_text(angle = 45, hjust = 1), legend.position = &#39;top&#39;, legend.justification = c(0.5, 0), legend.title=element_blank())
}

png(&#39;/users/k1806347/brc_scratch/Analyses/PRS_comparison/UKBB_outcomes_for_prediction/Current_PRS_vs_standardised_PRS.png&#39;, unit=&#39;px&#39;, width=2000, height=3500, res=300)
plot_grid(plotlist=Plots, labels = &quot;AUTO&quot;, ncol=2)
dev.off()

write.csv(cor_res, &#39;/users/k1806347/brc_scratch/Analyses/PRS_comparison/UKBB_outcomes_for_prediction/Current_PRS_vs_standardised_PRS_cor.csv&#39;, row.names=F, quote=F)</code></pre>
</details>
<details>
<p><summary>Show correlation between phenotype and different PRS</summary></p>
<center>
<div class="figure">
<img src="Images/standardised_vs_nonstandardised/Current_PRS_vs_standardised_PRS.png" alt="Correlation between phenotype and PRS" />
<p class="caption">Correlation between phenotype and PRS</p>
</div>

</details>
<details>
<p><summary>Show correlation between PRS</summary></p>
<table>
<caption>Correlation between standardised and non-standardised PRS</caption>
<thead>
<tr class="header">
<th align="left">GWAS</th>
<th align="right">pT</th>
<th align="right">cor</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">BODY03</td>
<td align="right">0.01</td>
<td align="right">0.7202845</td>
</tr>
<tr class="even">
<td align="left">BODY03</td>
<td align="right">0.10</td>
<td align="right">0.7066447</td>
</tr>
<tr class="odd">
<td align="left">BODY03</td>
<td align="right">0.20</td>
<td align="right">0.7136640</td>
</tr>
<tr class="even">
<td align="left">BODY03</td>
<td align="right">0.30</td>
<td align="right">0.7195138</td>
</tr>
<tr class="odd">
<td align="left">BODY03</td>
<td align="right">0.40</td>
<td align="right">0.7233397</td>
</tr>
<tr class="even">
<td align="left">BODY03</td>
<td align="right">0.50</td>
<td align="right">0.7258802</td>
</tr>
<tr class="odd">
<td align="left">BODY03</td>
<td align="right">1.00</td>
<td align="right">0.7270496</td>
</tr>
<tr class="even">
<td align="left">COAD01</td>
<td align="right">0.01</td>
<td align="right">0.6116274</td>
</tr>
<tr class="odd">
<td align="left">COAD01</td>
<td align="right">0.10</td>
<td align="right">0.6273054</td>
</tr>
<tr class="even">
<td align="left">COAD01</td>
<td align="right">0.20</td>
<td align="right">0.6441036</td>
</tr>
<tr class="odd">
<td align="left">COAD01</td>
<td align="right">0.30</td>
<td align="right">0.6574784</td>
</tr>
<tr class="even">
<td align="left">COAD01</td>
<td align="right">0.40</td>
<td align="right">0.6656764</td>
</tr>
<tr class="odd">
<td align="left">COAD01</td>
<td align="right">0.50</td>
<td align="right">0.6673727</td>
</tr>
<tr class="even">
<td align="left">COAD01</td>
<td align="right">1.00</td>
<td align="right">0.6694118</td>
</tr>
<tr class="odd">
<td align="left">COLL01</td>
<td align="right">0.01</td>
<td align="right">0.6222449</td>
</tr>
<tr class="even">
<td align="left">COLL01</td>
<td align="right">0.10</td>
<td align="right">0.7127647</td>
</tr>
<tr class="odd">
<td align="left">COLL01</td>
<td align="right">0.20</td>
<td align="right">0.7431230</td>
</tr>
<tr class="even">
<td align="left">COLL01</td>
<td align="right">0.30</td>
<td align="right">0.7562313</td>
</tr>
<tr class="odd">
<td align="left">COLL01</td>
<td align="right">0.40</td>
<td align="right">0.7642572</td>
</tr>
<tr class="even">
<td align="left">COLL01</td>
<td align="right">0.50</td>
<td align="right">0.7685698</td>
</tr>
<tr class="odd">
<td align="left">COLL01</td>
<td align="right">1.00</td>
<td align="right">0.7729352</td>
</tr>
<tr class="even">
<td align="left">CROH01</td>
<td align="right">0.01</td>
<td align="right">0.7065433</td>
</tr>
<tr class="odd">
<td align="left">CROH01</td>
<td align="right">0.10</td>
<td align="right">0.7361608</td>
</tr>
<tr class="even">
<td align="left">CROH01</td>
<td align="right">0.20</td>
<td align="right">0.7564487</td>
</tr>
<tr class="odd">
<td align="left">CROH01</td>
<td align="right">0.30</td>
<td align="right">0.7694719</td>
</tr>
<tr class="even">
<td align="left">CROH01</td>
<td align="right">0.40</td>
<td align="right">0.7747197</td>
</tr>
<tr class="odd">
<td align="left">CROH01</td>
<td align="right">0.50</td>
<td align="right">0.7781611</td>
</tr>
<tr class="even">
<td align="left">CROH01</td>
<td align="right">1.00</td>
<td align="right">0.7811400</td>
</tr>
<tr class="odd">
<td align="left">DIAB05</td>
<td align="right">0.01</td>
<td align="right">0.5495322</td>
</tr>
<tr class="even">
<td align="left">DIAB05</td>
<td align="right">0.10</td>
<td align="right">0.5899781</td>
</tr>
<tr class="odd">
<td align="left">DIAB05</td>
<td align="right">0.20</td>
<td align="right">0.6086436</td>
</tr>
<tr class="even">
<td align="left">DIAB05</td>
<td align="right">0.30</td>
<td align="right">0.6163363</td>
</tr>
<tr class="odd">
<td align="left">DIAB05</td>
<td align="right">0.40</td>
<td align="right">0.6209179</td>
</tr>
<tr class="even">
<td align="left">DIAB05</td>
<td align="right">0.50</td>
<td align="right">0.6234401</td>
</tr>
<tr class="odd">
<td align="left">DIAB05</td>
<td align="right">1.00</td>
<td align="right">0.6255095</td>
</tr>
<tr class="even">
<td align="left">HEIG03</td>
<td align="right">0.01</td>
<td align="right">0.7915908</td>
</tr>
<tr class="odd">
<td align="left">HEIG03</td>
<td align="right">0.10</td>
<td align="right">0.8224246</td>
</tr>
<tr class="even">
<td align="left">HEIG03</td>
<td align="right">0.20</td>
<td align="right">0.8319358</td>
</tr>
<tr class="odd">
<td align="left">HEIG03</td>
<td align="right">0.30</td>
<td align="right">0.8354708</td>
</tr>
<tr class="even">
<td align="left">HEIG03</td>
<td align="right">0.40</td>
<td align="right">0.8387046</td>
</tr>
<tr class="odd">
<td align="left">HEIG03</td>
<td align="right">0.50</td>
<td align="right">0.8398742</td>
</tr>
<tr class="even">
<td align="left">HEIG03</td>
<td align="right">1.00</td>
<td align="right">0.8404782</td>
</tr>
<tr class="odd">
<td align="left">RHEU01</td>
<td align="right">0.01</td>
<td align="right">0.3635614</td>
</tr>
<tr class="even">
<td align="left">RHEU01</td>
<td align="right">0.10</td>
<td align="right">0.5217018</td>
</tr>
<tr class="odd">
<td align="left">RHEU01</td>
<td align="right">0.20</td>
<td align="right">0.5668218</td>
</tr>
<tr class="even">
<td align="left">RHEU01</td>
<td align="right">0.30</td>
<td align="right">0.5859811</td>
</tr>
<tr class="odd">
<td align="left">RHEU01</td>
<td align="right">0.40</td>
<td align="right">0.5971442</td>
</tr>
<tr class="even">
<td align="left">RHEU01</td>
<td align="right">0.50</td>
<td align="right">0.6014540</td>
</tr>
<tr class="odd">
<td align="left">RHEU01</td>
<td align="right">1.00</td>
<td align="right">0.6055250</td>
</tr>
<tr class="even">
<td align="left">SCLE02</td>
<td align="right">0.01</td>
<td align="right">0.5444079</td>
</tr>
<tr class="odd">
<td align="left">SCLE02</td>
<td align="right">0.10</td>
<td align="right">0.5444104</td>
</tr>
<tr class="even">
<td align="left">SCLE02</td>
<td align="right">0.20</td>
<td align="right">0.5740674</td>
</tr>
<tr class="odd">
<td align="left">SCLE02</td>
<td align="right">0.30</td>
<td align="right">0.5922845</td>
</tr>
<tr class="even">
<td align="left">SCLE02</td>
<td align="right">0.40</td>
<td align="right">0.5992493</td>
</tr>
<tr class="odd">
<td align="left">SCLE02</td>
<td align="right">0.50</td>
<td align="right">0.6052226</td>
</tr>
<tr class="even">
<td align="left">SCLE02</td>
<td align="right">1.00</td>
<td align="right">0.6120867</td>
</tr>
</tbody>
</table>
</details>
<p><br/></p>
<p>The results show the PRS calcualted using these two approaches predict phenotypes to a similar degree, and although he scores are not highly correlated, their covariance with phenotypic data is higlighy correlated as a joint model does not improve prediction substantially.</p>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
