<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>UK Biobank Benchmark</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />
<link rel="stylesheet" href="styles/night-mode.css" id="nightModeStylesheet">

<script>
function toggleNightMode() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (stylesheet.disabled) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
}
</script>

<label class="switch">
  <input type="checkbox" id="toggleNightMode" checked>
  <span class="slider round"></span>
</label>

<script>
document.getElementById('toggleNightMode').addEventListener('change', function() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (this.checked) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="styles/styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img class="logo-img" src="Images/logo/Horizontal_white.png" style="height: 42px;" /></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pipeline
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="pipeline_overview.html">Overview</a>
    </li>
    <li>
      <a href="pipeline_readme.html">Instructions</a>
    </li>
    <li>
      <a href="pipeline_technical.html">Technical documentation</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Research
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="research_index.html">Overview</a>
    </li>
    <li>
      <a href="comparison_of_methods_summary.html">Polygenic Scoring Methods Comparison</a>
    </li>
    <li>
      <a href="Functionally_informed_prediction.html">Quantifying Polygenic Signal Mediated by Altered Gene Expression</a>
    </li>
    <li>
      <a href="Absolute_Conversion.html">Translating Polygenic Scores onto the Absolute Scale</a>
    </li>
  </ul>
</li>
<li>
  <a href="more_index.html">More</a>
</li>
<li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">UK Biobank Benchmark</h1>

</div>


<hr />
<p>Here we will use the GenoPred pipeline to derive polygenic scores in
UK Biobank.</p>
<hr />
<div id="summary" class="section level1">
<h1>Summary</h1>
<hr />
</div>
<div id="prepare-config" class="section level1">
<h1>Prepare config</h1>
<hr />
<div id="target_list" class="section level2">
<h2>target_list</h2>
<div class="shallow-break">

</div>
<details>
<summary>
Show code
</summary>
<p><br/></p>
<h3>
Create symlinks
</h3>
<p>We will create symlinks to the imputed genotype data for UKB. We will
use the pgen format data for computationl efficiency and those
restricted to MAF &gt;= 1% and INFO &gt;= 0.4. We are using genetic data
that is not application specific, so the data doesnâ€™t need to be
reprocessed for each application. Therefore we will use row number IDs
for the .psam file so they can be connected to application specific data
downstream.</p>
<pre class="bash"><code>mkdir -p /scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/ukb_symlinks

# pgen and pvar files
for chr in $(seq 1 22);do
  for file in $(echo pgen pvar);do
    ln -s /datasets/ukbiobank/June2017/Imputed/ukb_imp_chr${chr}_v3_MAF1_INFO4.${file} /scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/ukb_symlinks/ukb_imp_maf1_info4.chr${chr}.${file}
  done
done</code></pre>
<pre class="r"><code># Make .psam 
n = 487409
psam &lt;- data.frame(FID = 1:487409,
                   IID = 1:487409)
names(psam)[1]&lt;-&#39;#FID&#39;
write.table(psam, &#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/ukb_symlinks/rownumber.psam&#39;, col.names=T, row.names = F, quote = F)</code></pre>
<pre class="bash"><code>for chr in $(seq 1 22);do
  ln -s /scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/ukb_symlinks/rownumber.psam /scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/ukb_symlinks/ukb_imp_maf1_info4.chr${chr}.psam
done</code></pre>
<hr />
<h3>
Create target_list
</h3>
<pre class="bash"><code>mkdir -p /scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/configs/basic</code></pre>
<pre class="r"><code>target_list &lt;- data.frame(
  name=&#39;ukb&#39;,
  path=&#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/ukb_symlinks/ukb_imp_maf1_info4&#39;,
  type=&#39;plink2&#39;,
  indiv_report=F
)

write.table(target_list, &#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/configs/basic/target_list.txt&#39;, col.names=T, row.names=F, quote=F)</code></pre>
</details>
<hr />
</div>
<div id="gwas_list" class="section level2">
<h2>gwas_list</h2>
<p>Use the same GWAS as used in the INTERVENE paper (<a
href="https://doi.org/10.1101/2023.11.20.23298215">link</a>), except
exclude AD as overlaps with UKB, and use Yengo Height with UKB excluded.
After closer inspection, the RA GWAS (Ho et al) is actually majority EAS
ancestry. For this initial evaluation, lets prioritise EUR GWAS, so for
RA I will use a more recent GWAS by Ishigaki et al (GCST90132223). Also
the sample size specified for breast cancer GWAS seems to be wrong, the
sumstats on GWAS catalogue are the meta-analysis (122,977 cases and
105,974 controls).</p>
<details>
<summary>
Show code
</summary>
<h3>
Download sumstats from the GWAS catalogue
</h3>
<pre class="bash"><code>mkdir -p /scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/gwas_sumstats</code></pre>
<pre class="r"><code>gwasc_table &lt;- data.frame(
  id = c(
    &#39;GCST005838&#39;,
    &#39;GCST002783&#39;,
    &#39;GCST004988&#39;,
    &#39;GCST006085&#39;,
    &#39;GCST008059&#39;,
    &#39;GCST90132223&#39;,
    &#39;GCST004131&#39;,
    &#39;GCST007140&#39;,
    &#39;GCST90013445&#39;,
    &#39;GCST008972&#39;,
    &#39;GCST007954&#39;,
    &#39;GCST004773&#39;,
    &#39;GCST90018959_exclukb&#39;
  ),
  url = c(
    &#39;https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST005001-GCST006000/GCST005838/harmonised/29531354-GCST005838-EFO_0000712.h.tsv.gz&#39;,
    &#39;https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST002001-GCST003000/GCST002783/harmonised/25673413-GCST002783-EFO_0004340.h.tsv.gz&#39;,
    &#39;https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST004001-GCST005000/GCST004988/harmonised/29059683-GCST004988-EFO_0000305.h.tsv.gz&#39;,
    &#39;https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST006001-GCST007000/GCST006085/harmonised/29892016-GCST006085-EFO_0001663.h.tsv.gz&#39;,
    &#39;https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST008001-GCST009000/GCST008059/20171017_MW_eGFR_overall_EA_nstud42.dbgap.txt.gz&#39;,
    &#39;https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST90132001-GCST90133000/GCST90132223/GCST90132223_buildGRCh37.tsv.gz&#39;,
    &#39;https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST004001-GCST005000/GCST004131/harmonised/28067908-GCST004131-EFO_0003767.h.tsv.gz&#39;,
    &#39;https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST007001-GCST008000/GCST007140/GERA-sqrtHDL.tsv.gz&#39;,
    &#39;https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST90013001-GCST90014000/GCST90013445/GCST90013445_buildGRCh38.tsv&#39;,
    &#39;https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST008001-GCST009000/GCST008972/urate_chr1_22_LQ_IQ06_mac10_all_741_rsid.txt.gz&#39;,
    &#39;https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST007001-GCST008000/GCST007954/HbA1c_METAL_European.txt.gz&#39;,
    &#39;https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST004001-GCST005000/GCST004773/METAANALYSIS_DIAGRAM_SE1.txt&#39;,
    &#39;https://portals.broadinstitute.org/collaboration/giant/images/8/8e/GIANT_HEIGHT_YENGO_2022_GWAS_SUMMARY_STATS_EUR_excluding_UKB.gz&#39;
  )
)

gwasc_table$ending&lt;-ifelse(grepl(&#39;\\.gz$&#39;, gwasc_table$url), &#39;.gz&#39;,&#39;&#39;)

for (i in 1:nrow(gwasc_table)) {
  system(
    paste0(
      &#39;wget --no-check-certificate -O /scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/gwas_sumstats/&#39;,
      gwasc_table$id[i],
      &#39;.tsv&#39;, gwasc_table$ending[i], &#39; &#39;,
      gwasc_table$url[i]
    )
  )
}

# Check the headers are expected
library(data.table)
for(i in 1:nrow(gwasc_table)){
  tmp &lt;-
    fread(
      paste0(
        &#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/gwas_sumstats/&#39;,
        gwasc_table$id[i],
        &#39;.tsv&#39;,
        gwasc_table$ending[i]
      ),
      nrows=10
    )
  
  print(gwasc_table$id[i])
  print(head(tmp, 2))
}

# Update header of &quot;GCST008059&quot;
i&lt;-which(gwasc_table$id == &#39;GCST008059&#39;)
tmp &lt;-
  fread(
    paste0(
      &#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/gwas_sumstats/&#39;,
      gwasc_table$id[i],
      &#39;.tsv&#39;,
      gwasc_table$ending[i]
    )
  )

names(tmp)[names(tmp) == &quot;n_total_sum&quot;] &lt;- &#39;N&#39;
tmp$Allele1 &lt;- toupper(tmp$Allele1)
tmp$Allele2 &lt;- toupper(tmp$Allele2)

fwrite(
  tmp,
  paste0(
    &#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/gwas_sumstats/&#39;,
    gwasc_table$id[i],
    &#39;.tsv&#39;,
    gwasc_table$ending[i]
  ),
  sep = &#39;\t&#39;,
  quote = F,
  na=&#39;NA&#39;
)

# Update header of &quot;GCST007140&quot;
i&lt;-which(gwasc_table$id == &#39;GCST007140&#39;)
tmp &lt;-
  fread(
    paste0(
      &#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/gwas_sumstats/&#39;,
      gwasc_table$id[i],
      &#39;.tsv&#39;,
      gwasc_table$ending[i]
    )
  )

names(tmp) &lt;-
  c(&quot;RSID&quot; ,
    &quot;CHR&quot; ,
    &quot;BP&quot; ,
    &quot;A1&quot; ,
    &quot;A2&quot; ,
    &quot;Effect allele (EA)&quot; ,
    &quot;FREQ&quot; ,
    &quot;N&quot; ,
    &quot;BETA&quot; ,
    &quot;P&quot;)

tmp &lt;- tmp[, c(&quot;RSID&quot; ,
               &quot;CHR&quot; ,
               &quot;BP&quot; ,
               &quot;A1&quot; ,
               &quot;A2&quot; ,
               &quot;FREQ&quot; ,
               &quot;N&quot; ,
               &quot;BETA&quot; ,
               &quot;P&quot;), with=F]

fwrite(
  tmp,
  paste0(
    &#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/gwas_sumstats/&#39;,
    gwasc_table$id[i],
    &#39;.tsv&#39;,
    gwasc_table$ending[i]
  ),
  sep = &#39;\t&#39;,
  quote = F,
  na=&#39;NA&#39;
)

# Update header of &quot;GCST008972&quot;
i&lt;-which(gwasc_table$id == &#39;GCST008972&#39;)

# There is missing data which screws up fread
system(&quot;zcat /scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/gwas_sumstats/GCST008972.tsv.gz | sed &#39;s/  / NA /g&#39; | gzip &gt; /scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/gwas_sumstats/GCST008972_modified.tsv.gz; mv /scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/gwas_sumstats/GCST008972_modified.tsv.gz /scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/gwas_sumstats/GCST008972.tsv.gz&quot;)

tmp &lt;-
  fread(
    &#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/gwas_sumstats/GCST008972.tsv.gz&#39;
  )

names(tmp)[names(tmp) == &quot;n_total_sum&quot;] &lt;- &#39;N&#39;

tmp$Allele1 &lt;- toupper(tmp$Allele1)
tmp$Allele2 &lt;- toupper(tmp$Allele2)

fwrite(
  tmp,
  paste0(
    &#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/gwas_sumstats/&#39;,
    gwasc_table$id[i],
    &#39;.tsv&#39;,
    gwasc_table$ending[i]
  ),
  sep = &#39;\t&#39;,
  quote = F,
  na=&#39;NA&#39;
)

# Update header of &quot;GCST004773&quot;
i&lt;-which(gwasc_table$id == &#39;GCST004773&#39;)

tmp &lt;-
  fread(
    paste0(
      &#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/gwas_sumstats/&#39;,
      gwasc_table$id[i],
      &#39;.tsv&#39;,
      gwasc_table$ending[i]
    )
  )

tmp$CHR&lt;-gsub(&#39;:.*&#39;,&#39;&#39;,tmp$`Chr:Position`)
tmp$BP&lt;-gsub(&#39;.*:&#39;,&#39;&#39;,tmp$`Chr:Position`)

fwrite(
  tmp,
  paste0(
    &#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/gwas_sumstats/&#39;,
    gwasc_table$id[i],
    &#39;.tsv&#39;,
    gwasc_table$ending[i]
  ),
  sep = &#39;\t&#39;,
  quote = F,
  na=&#39;NA&#39;
)</code></pre>
<h3>
Create gwas_list
</h3>
<p>Get the GWAS info and prevalence data from INTERVENE paper.</p>
<pre class="bash"><code>mkdir -p /scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/configs/benchmark</code></pre>
<pre class="r"><code>gwas_dat&lt;-fread(&#39;https://raw.githubusercontent.com/intervene-EU-H2020/prspipe/main/config/studies_for_methods_comparison.tsv&#39;)

prev_dat&lt;-fread(&#39;https://raw.githubusercontent.com/intervene-EU-H2020/prspipe/main/config/pop_prevalence.tsv&#39;)
names(prev_dat)&lt;-c(&#39;name&#39;,&#39;prev&#39;)
gwas_dat&lt;-merge(gwas_dat, prev_dat, by=&#39;name&#39;, all.x=T)

gwas_dat$ftp_address&lt;-NULL
gwas_dat$local_path&lt;-NULL

gwas_dat&lt;-merge(gwasc_table, gwas_dat, by.x=&#39;id&#39;, by.y=&#39;study_id&#39;, all.x=T)

# Update labels to those used in the INTERVENE paper table
gwas_labels&lt;-data.frame(
  id=c(
    &#39;GCST005838&#39;, &#39;GCST90012877&#39;, &#39;GCST90013534&#39;, &#39;GCST004773&#39;, &#39;GCST004988&#39;, &#39;GCST006085&#39;, &#39;GCST90013445&#39;, &#39;GCST004131&#39;, &#39;GCST008059&#39;, &#39;GCST90018959&#39;, &#39;GCST008972&#39;, &#39;GCST002783&#39;, &#39;GCST007140&#39;, &#39;GCST007954&#39;),
  label=c(&#39;Stroke&#39;, &#39;AD or family history of AD&#39;, &#39;Rheumatoid arthritis&#39;, &#39;T2D&#39;, &#39;Breast cancer&#39;, &#39;Prostate cancer&#39;, &#39;T1D&#39;, &#39;IBD&#39;, &#39;eGFR&#39;, &#39;Height&#39;, &#39;Urate levels&#39;, &#39;BMI&#39;, &#39;HDL&#39;, &#39;HbA1c&#39;)
)
gwas_dat&lt;-merge(gwas_dat, gwas_labels, by=&#39;id&#39;, all.x=T)

# Fill in information for Height
gwas_dat$label[gwas_dat$id == &#39;GCST90018959_exclukb&#39;] &lt;- &#39;Height&#39;
gwas_dat$binary[gwas_dat$id == &#39;GCST90018959_exclukb&#39;] &lt;- &#39;no&#39;

# Fill in information for RA
gwas_dat$label[gwas_dat$id == &#39;GCST90132223&#39;] &lt;- &#39;Rheumatoid arthritis&#39;
gwas_dat$binary[gwas_dat$id == &#39;GCST90132223&#39;] &lt;- &#39;yes&#39;
gwas_dat$n_cases[gwas_dat$id == &#39;GCST90132223&#39;] &lt;- 22350
gwas_dat$n_controls[gwas_dat$id == &#39;GCST90132223&#39;] &lt;- 74823
gwas_dat$prev[gwas_dat$id == &#39;GCST90132223&#39;] &lt;- 0.01

# Insert correct ncase and ncontrol data for breast cancer
gwas_dat$n_cases[gwas_dat$id == &#39;GCST004988&#39;] &lt;- 122977
gwas_dat$n_controls[gwas_dat$id == &#39;GCST004988&#39;] &lt;- 105974

# Format as gwas_list
gwas_list &lt;- data.frame(
  name = gwas_dat$id,
  path = paste0(
    &#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/gwas_sumstats/&#39;,
    gwas_dat$id,
    &#39;.tsv&#39;,
    gwas_dat$ending
  ),
  population = &#39;EUR&#39;,
  n = gwas_dat$n_cases+gwas_dat$n_controls,
  sampling = gwas_dat$n_cases/(gwas_dat$n_controls+gwas_dat$n_cases),
  prevalence = gwas_dat$prev,
  mean = NA,
  sd = NA,
  label = gwas_dat$label
)

gwas_list$sampling[is.na(gwas_list$prevalence)]&lt;-NA
gwas_list$sampling&lt;-round(gwas_list$sampling, 3)
gwas_list$mean[is.na(gwas_list$prevalence)]&lt;-0
gwas_list$sd[is.na(gwas_list$prevalence)]&lt;-1

gwas_list$label&lt;-paste0(&quot;\&quot;&quot;, gwas_list$label, &quot;\&quot;&quot;)

write.table(gwas_list, &#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/configs/benchmark/gwas_list.txt&#39;, col.names=T, row.names=F, quote=F)</code></pre>
</details>
<hr />
</div>
<div id="configfile" class="section level2">
<h2>configfile</h2>
<details>
<summary>
Show code
</summary>
<pre class="r"><code># Create config file
conf &lt;- c(
  &#39;outdir: /scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/output&#39;,
  &#39;config_file: /scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/configs/benchmark/config.yaml&#39;,
  &#39;gwas_list: /scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/configs/benchmark/gwas_list.txt&#39;,
  &#39;target_list: /scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/configs/basic/target_list.txt&#39;,
  &quot;pgs_methods: [&#39;ptclump&#39;,&#39;dbslmm&#39;,&#39;prscs&#39;,&#39;sbayesr&#39;,&#39;lassosum&#39;,&#39;ldpred2&#39;,&#39;megaprs&#39;]&quot;
)

write.table(conf, &#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/configs/benchmark/config.yaml&#39;, col.names = F, row.names = F, quote = F)</code></pre>
</details>
<hr />
</div>
</div>
<div id="run-pipeline" class="section level1">
<h1>Run pipeline</h1>
<pre class="bash"><code>snakemake --profile slurm --use-conda --configfile=../../usr/k1806347/configs/benchmark/config.yaml output_all -n </code></pre>
<hr />
<div id="computational-benchmark" class="section level2">
<h2>Computational benchmark</h2>
<details>
<summary>
Show code
</summary>
<pre class="r"><code>library(data.table)
library(ggplot2)
library(cowplot)

# read in gwas_list
gwas_list&lt;-fread(&#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/configs/benchmark/gwas_list.txt&#39;)

# Read in configuration specific benchmark files
bm_files_i &lt;-
  paste0(
    &#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/output/reference/benchmarks/&#39;,
    list.files(&#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/output/reference/benchmarks/&#39;)
  )

# Read in benchmark files
bm_dat_all &lt;- do.call(rbind, lapply(bm_files_i, function(file) {
  tmp &lt;- fread(file)
  tmp$file &lt;- basename(file)
  return(tmp)
}))

# Create rule column
bm_dat_all$rule &lt;- gsub(&#39;-.*&#39;,&#39;&#39;,bm_dat_all$file)

# Summarise memory and time
bm_dat_summary &lt;- NULL

#######
# Rules that are only run once
#######

tmp &lt;- bm_dat_all[bm_dat_all$rule %in% c(&#39;ancestry_inference_i&#39;,
                                         &#39;outlier_detection_i&#39;,
                                         &#39;sample_report_i&#39;)]

tmp &lt;- data.frame(
  rule = tmp$rule,
  pgs_method = NA,
  pop = NA,
  mean_time = tmp$s,
  max_time = tmp$s,
  mean_mem = tmp$max_rss,
  max_mem = tmp$max_rss
)

bm_dat_summary &lt;-
  rbind(bm_dat_summary, tmp)

#######
# Sum the format_target time
#######

tmp &lt;- bm_dat_all[bm_dat_all$rule %in% c(&#39;format_target_i&#39;)]

tmp &lt;- data.frame(
  rule = tmp$rule[1],
  pgs_method = NA,
  pop = NA,
  mean_time = mean(tmp$s),
  max_time = max(tmp$s),
  mean_mem = mean(tmp$max_rss),
  max_mem = max(tmp$max_rss)
)

bm_dat_summary &lt;-
  rbind(bm_dat_summary, tmp)


#######
# Average across GWAS
#######
for (rule_i in c(
  &#39;prep_pgs_ptclump_i&#39;,
  &#39;prep_pgs_dbslmm_i&#39;,
  &#39;prep_pgs_prscs_i&#39;,
  &#39;prep_pgs_megaprs_i&#39;,
  &#39;prep_pgs_lassosum_i&#39;,
  &#39;prep_pgs_ldpred2_i&#39;,
  &#39;prep_pgs_sbayesr_i&#39;,
  &#39;sumstat_prep_i&#39;
)) {
    bm_dat_all_tmp &lt;-
    bm_dat_all[bm_dat_all$rule == rule_i, ]
    
    bm_dat_all_tmp$gwas &lt;- gsub(&#39;.*-&#39;,&#39;&#39;, gsub(&#39;.txt&#39;,&#39;&#39;, bm_dat_all_tmp$file))
    bm_dat_all_tmp&lt;-bm_dat_all_tmp[bm_dat_all_tmp$gwas %in% gwas_list$name,]
    
    tmp &lt;- data.table(
      rule = rule_i,
      pgs_method = NA,
      pop = NA,
      mean_time = mean(bm_dat_all_tmp$s),
      max_time = max(bm_dat_all_tmp$s),
      mean_mem = mean(bm_dat_all_tmp$max_rss),
      max_mem = max(bm_dat_all_tmp$max_rss)
    )
    
    bm_dat_summary&lt;-rbind(bm_dat_summary, tmp)
}
  
####
# Average target_pgs_i across PGS method and GWAS
####

rule_i&lt;-&#39;target_pgs_i&#39; 
bm_dat_all_tmp &lt;- bm_dat_all[bm_dat_all$rule == rule_i, ]

rule_split&lt;-strsplit(bm_dat_all_tmp$file, &#39;-&#39;)
bm_dat_all_tmp$pgs_method &lt;- lapply(rule_split, function(x) x[3])
bm_dat_all_tmp$pop &lt;- lapply(rule_split, function(x) x[4])

for(pgs_method_i in unique(bm_dat_all_tmp$pgs_method)){
  for(pop_i in unique(bm_dat_all_tmp$pop)){
    bm_dat_all_tmp_2&lt;-bm_dat_all_tmp[bm_dat_all_tmp$pgs_method == pgs_method_i &amp; bm_dat_all_tmp$pop == pop_i,]
    
    tmp &lt;- data.table(
      rule = rule_i,
      pgs_method = pgs_method_i,
      pop = pop_i,
      mean_time = mean(bm_dat_all_tmp_2$s),
      max_time = max(bm_dat_all_tmp_2$s),
      mean_mem = mean(bm_dat_all_tmp_2$max_rss),
      max_mem = max(bm_dat_all_tmp_2$max_rss)
    )

    bm_dat_summary&lt;-rbind(bm_dat_summary, tmp)

  }
}

write.csv(bm_dat_summary, &#39;/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/docs/Images/bm_res.csv&#39;)

#################
# Plot
#################

bm_dat_summary_i &lt;- bm_dat_summary[bm_dat_summary$rule %in% c(
    &#39;prep_pgs_ptclump_i&#39;,
    &#39;prep_pgs_dbslmm_i&#39;,
    &#39;prep_pgs_prscs_i&#39;,
    &#39;prep_pgs_megaprs_i&#39;,
    &#39;prep_pgs_lassosum_i&#39;,
    &#39;prep_pgs_ldpred2_i&#39;,
    &#39;prep_pgs_sbayesr_i&#39;
  ),]

bm_dat_summary_i$method &lt;- gsub(&#39;_i&#39;, &#39;&#39;, gsub(&#39;prep_pgs_&#39;,&#39;&#39;,bm_dat_summary_i$rule))
bm_dat_summary_i&lt;-merge(bm_dat_summary_i, pgs_method_labels, by=&#39;method&#39;)

pgs_method_time &lt;-
  ggplot(bm_dat_summary_i, aes(x = label, y = mean_time)) +
  geom_bar(stat = &quot;identity&quot;, position = &quot;dodge&quot;) +
  labs(x = &quot;PGS Method&quot;, y = &quot;Time (seconds)&quot;, fill=&#39;Cores&#39;) +
  theme_half_open() +
  background_grid() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

pgs_method_mem &lt;-
  ggplot(bm_dat_summary_i, aes(x = label, y = max_mem)) +
  geom_bar(stat = &quot;identity&quot;, position=&quot;dodge&quot;) +
  labs(x = &quot;PGS Method&quot;, y = &quot;Max Memory (Mb)&quot;, fill=&#39;Cores&#39;) +
  theme_half_open() +
  background_grid() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

png(&#39;/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/docs/Images/UKB/bm_pgs_methods.png&#39;, res = 300, width = 1800, height = 1800, units = &#39;px&#39;)
  plot_grid(pgs_method_time, pgs_method_mem, nrow=2)
dev.off()</code></pre>
</details>
<hr />
</div>
</div>
<div id="evaluate-scores" class="section level1">
<h1>Evaluate scores</h1>
<hr />
<div id="prepare-phenotype-data" class="section level2">
<h2>Prepare phenotype data</h2>
<p>We will use application 82087 for this analysis, and use <a
href="https://kenhanscombe.github.io/ukbkings/index.html">ukbkings</a>
to extract the relevant data.</p>
<hr />
<div id="binary-outcomes" class="section level3">
<h3>Binary outcomes</h3>
<details>
<summary>
Show code
</summary>
<pre class="r"><code>library(ukbkings)
library(dplyr)
library(stringr)

# Extract ICD code and sex from UKB
project_dir &lt;- &quot;/datasets/ukbiobank/ukb82087&quot;
f &lt;- bio_field(project_dir)
f %&gt;%
    select(field, name) %&gt;%
    filter(str_detect(field, &quot;^31-0.0|41202|41203|41204|41205&quot;)) %&gt;%
    bio_field_add(&quot;/scratch/prj/ukbiobank/usr/ollie_pain/phenotypes/icd_field_subset.txt&quot;)

bio_phen(
    project_dir,
    field = &quot;/scratch/prj/ukbiobank/usr/ollie_pain/phenotypes/icd_field_subset.txt&quot;,
    out = &quot;/scratch/prj/ukbiobank/usr/ollie_pain/phenotypes/icd_field_subset&quot;
)

system(&quot;ls -lh /scratch/prj/ukbiobank/usr/ollie_pain/phenotypes/icd_field_subset.rds&quot;)
df &lt;- readRDS(&quot;/scratch/prj/ukbiobank/usr/ollie_pain/phenotypes/icd_field_subset.rds&quot;)

# Identify individuals with no ICD codes
rows_with_na &lt;- apply(df[, !(names(df) %in% c(&#39;eid&#39;,&#39;31-0.0&#39;))], 1, function(x) all(is.na(x)))
df_no_icd&lt;-df[rows_with_na,]

# Change to long format
library(tidyr)

# Split into ICD9 and ICD10
df_long &lt;- df %&gt;%
  pivot_longer(cols = starts_with(&#39;412&#39;), names_to = &quot;diagnosis_event&quot;, values_to = &quot;icd_code&quot;) %&gt;%
  drop_na(icd_code)

# Reinsert people with no ICD codes (so they are included as controls)
df_no_icd&lt;-df_no_icd[, c(&#39;eid&#39;,&#39;31-0.0&#39;)]
df_no_icd$diagnosis_event&lt;-NA
df_no_icd$icd_code&lt;-NA
df_long&lt;-rbind(df_long, df_no_icd)

# Create icd checker function
icd_checker &lt;- function(dat, sex = NULL, incl, excl = NULL){

  out&lt;-data.frame(
    eid=unique(dat$eid),
    endpoint=NA
  )
  
  # Subset by inclusion
  dat_in &lt;- unique(dat$eid[grepl(paste(paste0(&#39;^&#39;,incl), collapse=&#39;|&#39;), dat$icd_code)])
  out$endpoint&lt;-ifelse(out$eid %in% dat_in, 1, 0)
  
  if(!is.null(excl)){
    # Remove those with exclusions
    dat_ex &lt;- unique(dat$eid[grepl(paste(paste0(&#39;^&#39;,excl), collapse=&#39;|&#39;), dat$icd_code)])
    out$endpoint[out$eid %in% dat_in &amp; out$eid %in% dat_ex]&lt;-0
  }
  
  if(!is.null(sex)){
    # Subset relevant sex
    dat_na &lt;- unique(dat$eid[dat$`31-0.0` != sex])
    out$endpoint[out$eid %in% dat_na]&lt;-NA
  }
  
  return(out)
}

###
# Breast cancer
###

bc&lt;-icd_checker(dat = df_long, sex = 0, incl = c(&#39;C50&#39;,&#39;174&#39;))

###
# Chronic kidney disease
###

ckd&lt;-icd_checker(dat = df_long, incl = c(&#39;N18&#39;,&#39;585&#39;))
N14_DIALYSIS&lt;-icd_checker(dat = df_long, incl = c(&#39;Z992&#39;,&#39;Y841&#39;))
ckd$endpoint[ckd$eid %in% N14_DIALYSIS$eid[N14_DIALYSIS$endpoint == 1]]&lt;-1

###
# Gout
###

gout&lt;-icd_checker(dat = df_long, incl = c(&#39;M10&#39;,&#39;2740&#39;))

###
# IBD
###

ibd&lt;-icd_checker(dat = df_long, incl = c(&#39;K50&#39;,&#39;K51&#39;,&#39;555&#39;,&#39;556&#39;))

###
# Prostate cancer
###

pc&lt;-icd_checker(dat = df_long, sex = 1, incl = c(&#39;C61&#39;,&#39;185&#39;))

###
# Rheumatoid arthritis
###

ra&lt;-icd_checker(dat = df_long, incl = c(&#39;M058&#39;,&#39;M059&#39;,&#39;7140A&#39;))

###
# Stroke exluding SAH
###

stroke&lt;-icd_checker(dat = df_long, incl = c(&#39;I61&#39;, &#39;I63&#39;, &#39;I64&#39;, &#39;431&#39;, &#39;4330A&#39;, &#39;4331A&#39;, &#39;4339A&#39;, &#39;4340A&#39;, &#39;4341A&#39;, &#39;4349A&#39;, &#39;436&#39;), excl=&#39;I636&#39;)

###
# Type 1 diabetes
###

t1d&lt;-icd_checker(dat = df_long, incl = c(&#39;E10&#39;,&#39;25001&#39;,&#39;25011&#39;))
E4_DM2&lt;-icd_checker(dat = df_long, incl = c(&#39;E110&#39;, &#39;E111&#39;, &#39;E112&#39;, &#39;E113&#39;, &#39;E114&#39;, &#39;E115&#39;, &#39;E116&#39;, &#39;E117&#39;, &#39;E118&#39;, &#39;E119&#39;, &#39;2502A&#39;, &#39;2501A&#39;, &#39;2503A&#39;, &#39;2504A&#39;, &#39;2505A&#39;, &#39;2506A&#39;, &#39;2507A&#39;, &#39;2508A&#39;, &#39;2500A&#39;), excl=&#39;E11&#39;)
t1d$endpoint[t1d$eid %in% E4_DM2$eid[E4_DM2$endpoint == 1]]&lt;-0

###
# Type 2 diabetes
###

t2d&lt;-icd_checker(dat = df_long, incl = c(&#39;E11&#39;,&#39;25000&#39;,&#39;25010&#39;))
E4_DM1&lt;-icd_checker(dat = df_long, incl = c(&#39;E100&#39;, &#39;E101&#39;, &#39;E102&#39;, &#39;E103&#39;, &#39;E104&#39;, &#39;E105&#39;, &#39;E106&#39;, &#39;E107&#39;, &#39;E108&#39;, &#39;E109&#39;, &#39;2502B&#39;, &#39;2501B&#39;, &#39;2503B&#39;, &#39;2504B&#39;, &#39;2505B&#39;, &#39;2506B&#39;, &#39;2507B&#39;, &#39;2508B&#39;, &#39;2500B&#39;), excl=&#39;E10&#39;)
t2d$endpoint[t2d$eid %in% E4_DM1$eid[E4_DM1$endpoint == 1]]&lt;-0

# Create table of N
endpoints&lt;-list(bc=bc, 
                ckd=ckd, 
                gout=gout, 
                ibd=ibd, 
                pc=pc, 
                ra=ra, 
                stroke=stroke, 
                t1d=t1d, 
                t2d=t2d)

# Save the phenotype data
library(data.table)
for(i in names(endpoints)){
  tmp&lt;-endpoints[[i]]
  tmp&lt;-tmp[complete.cases(tmp),]
  fwrite(
    tmp,
    paste0(
      &#39;/scratch/prj/ukbiobank/usr/ollie_pain/phenotypes/&#39;,
      i,
      &#39;.txt&#39;
    ),
    row.names = F,
    quote = F,
    na = &#39;NA&#39;,
    sep = &#39;\t&#39;
  )
}</code></pre>
</details>
<hr />
</div>
<div id="continuous-outcomes" class="section level3">
<h3>Continuous outcomes</h3>
<details>
<summary>
Show code
</summary>
<pre class="r"><code>library(ukbkings)
library(dplyr)
library(stringr)

# Extract ICD code and sex from UKB
project_dir &lt;- &quot;/datasets/ukbiobank/ukb82087&quot;
f &lt;- bio_field(project_dir)
f %&gt;%
    select(field, name) %&gt;%
    filter(str_detect(field, &quot;^50-|^21001-|^30700-|^30750-|^30880-|^30760-&quot;)) %&gt;%
    bio_field_add(&quot;/scratch/prj/ukbiobank/usr/ollie_pain/phenotypes/continuous_field_subset.txt&quot;)

bio_phen(
    project_dir,
    field = &quot;/scratch/prj/ukbiobank/usr/ollie_pain/phenotypes/continuous_field_subset.txt&quot;,
    out = &quot;/scratch/prj/ukbiobank/usr/ollie_pain/phenotypes/continuous_field_subset&quot;
)

system(&quot;ls -lh /scratch/prj/ukbiobank/usr/ollie_pain/phenotypes/continuous_field_subset.rds&quot;)
df &lt;- readRDS(&quot;/scratch/prj/ukbiobank/usr/ollie_pain/phenotypes/continuous_field_subset.rds&quot;)

# Take the first observation of each outcome
library(tidyr)
df_long &lt;- df %&gt;%
  pivot_longer(cols = names(df)[!grepl(&#39;eid&#39;, names(df))], names_to = &quot;variable&quot;, values_to = &quot;outcome&quot;) %&gt;%
  drop_na(outcome)
df_long$variable&lt;-gsub(&#39;-.*&#39;,&#39;&#39;, df_long$variable)
df_long&lt;-df_long[!duplicated(df_long[,c(&#39;eid&#39;,&#39;variable&#39;)]),]

var_pheno&lt;-data.frame(
  var=c(&#39;50&#39;,&#39;21001&#39;,&#39;30700&#39;,&#39;30750&#39;,&#39;30880&#39;,&#39;30760&#39;),
  pheno=c(&#39;height&#39;,&#39;bmi&#39;,&#39;egfr&#39;,&#39;hba1c&#39;,&#39;urate&#39;,&#39;hdl&#39;)
)

library(data.table)

for(i in 1:nrow(var_pheno)){
  tmp &lt;- df_long[df_long$variable == var_pheno$var[i],]
  tmp &lt;- data.frame(
    eid = tmp$eid,
    outcome = tmp$outcome
  )
  
  fwrite(
    tmp,
    paste0(
      &#39;/scratch/prj/ukbiobank/usr/ollie_pain/phenotypes/&#39;,
      var_pheno$pheno[i],
      &#39;.txt&#39;
    ),
    row.names = F,
    quote = F,
    na = &#39;NA&#39;,
    sep = &#39;\t&#39;
  )
}</code></pre>
</details>
<hr />
</div>
<div id="remove-related-individuals" class="section level3">
<h3>Remove related individuals</h3>
<p>Relatedness has already been estimated in UK Biobank. Using greedy
related, identify a list of unrelated individuals for population outlier
detection, and for each phenotype. A function for this is available in
ukbkings.</p>
<details>
<summary>
Show code
</summary>
<pre class="r"><code>library(ukbkings)
library(data.table)

psam&lt;-fread(&#39;/scratch/prj/ukbiobank/ukb82087/imputed/ukb82087_imp_chr1_MAF1_INFO4_v1.psam&#39;)
psam$rn&lt;-1:nrow(psam)

project_dir &lt;- &quot;/datasets/ukbiobank/ukb82087&quot;
greedy_related &lt;- &quot;/scratch/prj/ukbiobank/KCL_Data/Software/tools/GreedyRelated-master-v1.2.1/GreedyRelated&quot;

# Create a list of unrelated individuals irrespective of a phenotype
psam_unrel_all &lt;- psam[!(
  psam$IID %in% bio_gen_related_remove(
    project_dir = project_dir,
    greedy_related = greedy_related,
    thresh = 0.044,
    seed = 1
  )$eid
), ]

write.table(psam_unrel_all$IID, &#39;/scratch/prj/ukbiobank/usr/ollie_pain/phenotypes/unrelated.txt&#39;, row.names=F, col.names=F, quote=F)
write.table(psam_unrel_all$rn, &#39;/scratch/prj/ukbiobank/usr/ollie_pain/phenotypes/unrelated.row_number.txt&#39;, row.names=F, col.names=F, quote=F)

n_table&lt;-data.frame(
  pheno=NA,
  subset=c(&#39;geno&#39;,&#39;unrel&#39;),
  n=c(nrow(psam), nrow(psam_unrel_all)),
  n_cas=NA,
  n_con=NA
)

# Create a list of unrelated individuals after removing individuals with missing data for each phenotype
bin_phenos &lt;- c(&#39;bc&#39;, &#39;ckd&#39;, &#39;gout&#39;, &#39;ibd&#39;, &#39;pc&#39;, &#39;ra&#39;, &#39;stroke&#39;, &#39;t1d&#39;, &#39;t2d&#39;)
con_phenos &lt;- c(&#39;height&#39;,&#39;bmi&#39;,&#39;egfr&#39;,&#39;hba1c&#39;,&#39;urate&#39;,&#39;hdl&#39;)

for(i in c(bin_phenos, con_phenos)){
  tmp&lt;-fread(paste0(
      &#39;/scratch/prj/ukbiobank/usr/ollie_pain/phenotypes/&#39;,
      i,
      &#39;.txt&#39;
    ))
  
  tmp&lt;-tmp[tmp$eid %in% psam$IID,]
  tmp_unrel &lt;- tmp[!(
    tmp$eid %in% bio_gen_related_remove(
      project_dir = project_dir,
      greedy_related = greedy_related,
      keep = tmp$eid,
      thresh = 0.044,
      seed = 1
    )$eid
  ), ]
  
  fwrite(
    tmp_unrel,
    paste0(
      &#39;/scratch/prj/ukbiobank/usr/ollie_pain/phenotypes/&#39;,
      i,
      &#39;.unrel.txt&#39;
    ),
    row.names = F,
    quote = F,
    na = &#39;NA&#39;,
    sep = &#39;\t&#39;
  )
  
  names(tmp)&lt;-c(&#39;IID&#39;,&#39;PHENO&#39;)
  names(tmp_unrel)&lt;-c(&#39;IID&#39;,&#39;PHENO&#39;)
    
  if(i %in% bin_phenos){
    n_i&lt;-data.frame(
      pheno=i,
      subset=c(&#39;geno&#39;,&#39;unrel&#39;),
      n=c(nrow(tmp), nrow(tmp_unrel)),
      n_cas=c(sum(tmp$PHENO == 1), sum(tmp_unrel$PHENO == 1)),
      n_con=c(sum(tmp$PHENO == 0), sum(tmp_unrel$PHENO == 0))
    )
  } else {
    n_i&lt;-data.frame(
      pheno=i,
      subset=c(&#39;geno&#39;,&#39;unrel&#39;),
      n=c(nrow(tmp), nrow(tmp_unrel)),
      n_cas=NA,
      n_con=NA
    )
  }
  
  n_table&lt;-rbind(n_table, n_i)
  print(n_table)
}

write.csv(n_table, &#39;/scratch/prj/ukbiobank/usr/ollie_pain/phenotypes/pheno_n.csv&#39;, row.names=F)</code></pre>
</details>
<hr />
</div>
</div>
<div id="check-correlation" class="section level2">
<h2>Check correlation</h2>
<details>
<summary>
Show code
</summary>
<pre class="r"><code># Test correlation between PGS and observed height

setwd(&#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/GenoPred/pipeline/&#39;)
library(data.table)
library(ggplot2)
library(cowplot)

source(&#39;../functions/misc.R&#39;)
source_all(&#39;../functions&#39;)

# Define pgs_methods used
pgs_methods &lt;- read_param(config = &#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/configs/benchmark/config.yaml&#39;, param = &#39;pgs_methods&#39;, return_obj = F)

# Read in PGS
pgs &lt;- read_pgs(config = &#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/configs/benchmark/config.yaml&#39;, name = &#39;ukb&#39;)$ukb

# Read in psam to get row number based IDs
psam&lt;-fread(&#39;/scratch/prj/ukbiobank/ukb82087/imputed/ukb82087_imp_chr1_MAF1_INFO4_v1.psam&#39;)
psam$rn&lt;-1:nrow(psam)

# Read in gwas_list
gwas_list&lt;-read_param(config = &#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/configs/benchmark/config.yaml&#39;, param = &#39;gwas_list&#39;)

# Create column containing the phenotypes corresponding to each GWAS
gwas_list$pheno&lt;-tolower(gwas_list$label)
gwas_list$pheno[gwas_list$pheno == &#39;breast cancer&#39;]&lt;-&#39;bc&#39;
gwas_list$pheno[gwas_list$pheno == &#39;prostate cancer&#39;]&lt;-&#39;pc&#39;
gwas_list$pheno[gwas_list$pheno == &#39;egfr&#39;]&lt;-&#39;egfr,ckd&#39;
gwas_list$pheno[gwas_list$pheno == &#39;urate levels&#39;]&lt;-&#39;urate,gout&#39;
gwas_list$pheno[gwas_list$pheno == &#39;rheumatoid arthritis&#39;]&lt;-&#39;ra&#39;

bin_phenos &lt;- c(&#39;bc&#39;, &#39;ckd&#39;, &#39;gout&#39;, &#39;ibd&#39;, &#39;pc&#39;, &#39;ra&#39;, &#39;stroke&#39;, &#39;t1d&#39;, &#39;t2d&#39;)
con_phenos &lt;- c(&#39;height&#39;,&#39;bmi&#39;,&#39;egfr&#39;,&#39;hba1c&#39;,&#39;urate&#39;,&#39;hdl&#39;)
phenos&lt;-c(bin_phenos, con_phenos)

cor&lt;-NULL
for(i in phenos){
  # Read in pheno data
  pheno_i &lt;- fread(paste0(
      &#39;/scratch/prj/ukbiobank/usr/ollie_pain/phenotypes/&#39;,
      i,
      &#39;.unrel.txt&#39;
    ))
  
  names(pheno_i)&lt;-c(&#39;IID&#39;,&#39;PHENO&#39;)
  # Update to row number based IDs
  pheno_i&lt;-merge(psam[,c(&#39;IID&#39;,&#39;rn&#39;), with=F], pheno_i, by=&#39;IID&#39;)
  pheno_i$IID&lt;-pheno_i$rn
  pheno_i$rn&lt;-NULL
  pheno_i$FID&lt;-pheno_i$IID
  pheno_i&lt;-pheno_i[, c(&#39;FID&#39;,&#39;IID&#39;,&#39;PHENO&#39;), with=F]
  
  # Estimate correlation between pheno and pgs from relevant gwas
  for(pop_i in names(pgs)){
    gwas_i &lt;- gwas_list$name[grepl(i, gwas_list$pheno)]
    for(pgs_method_i in names(pgs[[pop_i]][[gwas_i]])){
      pgs_i &lt;- pgs[[pop_i]][[gwas_i]][[pgs_method_i]]
      pheno_pgs&lt;-merge(pheno_i, pgs_i, by = c(&#39;FID&#39;,&#39;IID&#39;))
      
      for(model_i in names(pgs_i)[-1:-2]){
        x &lt;- scale(pheno_pgs[[model_i]])
        
        if(all(is.na(x))){
          next
        }
        
        
        if(i %in% bin_phenos){
          y &lt;- pheno_pgs$PHENO
          coef_i &lt;- coef(summary(mod &lt;- glm(y ~ x, family=&#39;binomial&#39;)))
          
          tmp &lt;- data.table(
            pheno = i,
            pop = pop_i,
            gwas = gwas_i,
            pgs_method = pgs_method_i,
            name = model_i,
            r = coef_i[2,1],
            se = coef_i[2,2],
            p = coef_i[2,4],
            n = nobs(mod),
            n_cas = sum(pheno_pgs$PHENO == 1),
            n_con = sum(pheno_pgs$PHENO == 0))
        } else {
          y &lt;- scale(pheno_pgs$PHENO)
          coef_i &lt;- coef(summary(mod &lt;- lm(y ~ x)))
          
          tmp &lt;- data.table(
            pheno = i,
            pop = pop_i,
            gwas = gwas_i,
            pgs_method = pgs_method_i,
            name = model_i,
            r = coef_i[2,1],
            se = coef_i[2,2],
            p = coef_i[2,4],
            n = nobs(mod),
            n_cas = NA,
            n_con = NA)
        }
        
        cor &lt;- rbind(cor, tmp)
      
      }
    }
  }
}

saveRDS(cor, &#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/assoc/cor.RDS&#39;)
cor&lt;-readRDS(&#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/assoc/cor.RDS&#39;)

# Subset results to best and pseudoval
cor_subset&lt;-NULL
for(pheno_i in phenos){
  for(pop_i in names(pgs)){
    for(gwas_i in unique(cor$gwas[cor$pheno == pheno_i &amp; cor$pop == pop_i])){
      for(pgs_method_i in unique(cor$pgs_method[cor$pheno == pheno_i &amp; cor$pop == pop_i &amp; cor$gwas == gwas_i])){
        
        # Subset relevant results
        cor_pop_i &lt;- cor[
          cor$pheno == pheno_i &amp;
          cor$pop == pop_i &amp;
          cor$gwas == gwas_i &amp;
          cor$pgs_method == pgs_method_i,]
        
        # Remove results with large SE
        cor_pop_i&lt;-cor_pop_i[which(cor_pop_i$se &lt; 10),]
    
        # Top R
        if(pgs_method_i %in% c(&#39;ptclump&#39;,&#39;ldpred2&#39;,&#39;megaprs&#39;,&#39;prscs&#39;,&#39;lassosum&#39;,&#39;dbslmm&#39;)){
          top_i &lt;- cor_pop_i[which(abs(cor_pop_i$r) == max(abs(cor_pop_i$r), na.rm = T))[1],]
          top_i$model &lt;- &#39;Top&#39;
          cor_subset &lt;- rbind(cor_subset, top_i)
        }
    
        # PseudoVal
        if(pgs_method_i %in% c(&#39;ptclump&#39;,&#39;sbayesr&#39;,&#39;ldpred2&#39;,&#39;megaprs&#39;,&#39;prscs&#39;,&#39;lassosum&#39;,&#39;dbslmm&#39;)){
          cor_pop_i$name &lt;- gsub(paste0(gwas_i, &#39;_&#39;), &#39;&#39;, cor_pop_i$name)
          pseudo_param &lt;- find_pseudo(config = &#39;/scratch/prj/ukbiobank/usr/ollie_pain/GenoPredPipe/usr/k1806347/configs/benchmark/config.yaml&#39;, gwas = gwas_i, pgs_method = pgs_method_i)
          pseudo_i &lt;- cor_pop_i[cor_pop_i$name == pseudo_param,]
          pseudo_i$model &lt;- &#39;Pseudo&#39;
          cor_subset &lt;- rbind(cor_subset, pseudo_i)
        }
      }
    }
  }
}

# Plot the results
cor_subset$model &lt;- factor(cor_subset$model, levels = c(&#39;Top&#39;,&#39;Pseudo&#39;))
dir.create(&#39;/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/docs/Images/UKB&#39;)

png(&#39;/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/docs/Images/UKB/cor_all.png&#39;, res=300, width=4000, height=6000, units = &#39;px&#39;)
ggplot(cor_subset, aes(x = pgs_method, y = r, fill = model)) +
  geom_errorbar(
    aes(ymin = r - se, ymax = r + se),
    width = .2,
    position = position_dodge(width = 0.7)
  ) +
  geom_point(stat=&quot;identity&quot;, position=position_dodge(0.7), size=2, shape=23) +
  labs(
    y = &quot;BETA (SE)&quot;,
    x = &#39;PGS Method&#39;,
    fill = &#39;Model&#39;
  ) +
  theme_half_open() +
  background_grid() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        plot.title = element_text(hjust = 0.5, size=12)) +
  facet_grid(pheno ~ pop , scales=&quot;free_y&quot;) +
  panel_border()
dev.off()

png(&#39;/users/k1806347/oliverpainfel/Software/MyGit/GenoPred/docs/Images/UKB/cor_eur.png&#39;, res=300, width=4500, height=3000, units = &#39;px&#39;)
ggplot(cor_subset[cor_subset$pop == &#39;EUR&#39;], aes(x = pgs_method, y = r, fill = model)) +
  geom_errorbar(
    aes(ymin = r - se, ymax = r + se),
    width = .2,
    position = position_dodge(width = 0.7)
  ) +
  geom_point(stat=&quot;identity&quot;, position=position_dodge(0.7), size=2, shape=23) +
  labs(
    y = &quot;BETA (SE)&quot;,
    x = &#39;PGS Method&#39;,
    fill = &#39;Model&#39;
  ) +
  theme_half_open() +
  background_grid() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        plot.title = element_text(hjust = 0.5, size=12)) +
  facet_wrap(pheno ~ . , scales=&quot;free&quot;, nrow = 3) +
  panel_border()
dev.off()</code></pre>
</details>
<details>
<summary>
Show associations for all populations
</summary>
<div class="centered-container">
<div class="rounded-image-container">
<div class="figure">
<img src="Images/UKB/cor_all.png" alt="" />
<p class="caption">UK Biobank - All populations</p>
</div>
</div>
</div>
</details>
<details>
<summary>
Show associations for European population
</summary>
<div class="centered-container">
<div class="rounded-image-container">
<div class="figure">
<img src="Images/UKB/cor_eur.png" alt="" />
<p class="caption">UK Biobank - All populations</p>
</div>
</div>
</div>
</details>
<p></br></p>
<hr />
</div>
</div>
<div id="conclusion" class="section level1">
<h1>Conclusion</h1>
<p>This preliminary analysis indicates the pipeline is working properly.
For proper comparison methods and evaluation of score, a cross
validation framework should be used.</p>
</div>

<!-- footer.html -->
<hr/>

<div class="centered-container">
<div class="rounded-image-container" style="width: 500px;">
<img src="Images/logo/sponsors.png">
</div>
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
